<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Model Fitting in Spatial Transcriptomics - Code &amp; Concept â€¢ SPATA2</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Model Fitting in Spatial Transcriptomics - Code &amp; Concept">
<meta property="og:description" content="SPATA2">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SPATA2</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/spata-v2-installation.html">Installation</a>
</li>
<li>
  <a href="../articles/spata-v2-spata-data.html">SPATAData</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">Around the spata2 object</li>
    <li>
      <a href="../articles/spata-v2-object-initiation-and-manipulation.html">Object Set Up and Manipulation</a>
    </li>
    <li>
      <a href="../articles/spata-v2-platform-compatibility.html">Platform Compatibility</a>
    </li>
    <li>
      <a href="../articles/spata-v2-autoencoder.html">Autoencoder</a>
    </li>
    <li>
      <a href="../articles/spata-v2-genes-gene-sets-and-features.html">Genes, Gene Sets and Features</a>
    </li>
    <li class="dropdown-header">Image Analysis</li>
    <li>
      <a href="../articles/spata-v2-image-handling.html">Image Handling</a>
    </li>
    <li>
      <a href="../articles/spata-v2-image-annotations.html">Image Annotations</a>
    </li>
    <li>
      <a href="../articles/spata-v2-spatial-segmentation.html">Spatial Segmentation</a>
    </li>
    <li class="dropdown-header">Analysis of Gene Expression</li>
    <li>
      <a href="../articles/spata-v2-cnv-analysis.html">Copy Number Variations</a>
    </li>
    <li>
      <a href="../articles/spata-v2-clustering.html">Clustering</a>
    </li>
    <li>
      <a href="../articles/spata-v2-dea.html">Differential Expression</a>
    </li>
    <li>
      <a href="../articles/spata-v2-gsea.html">Gene Set Enrichment</a>
    </li>
    <li class="dropdown-header">Expression Gradients &amp; Gradient Screening</li>
    <li>
      <a href="../articles/spata-v2-spatial-trajectories.html">Spatial Trajectories</a>
    </li>
    <li>
      <a href="../articles/spata-v2-spatial-trajectory-screening.html">Spatial Trajectory Screening</a>
    </li>
    <li>
      <a href="../articles/spata-v2-image-annotation-screening.html">Image Annotation Screening</a>
    </li>
    <li>
      <a href="../articles/spata-v2-scd.html">Integrate Single Cell Deconvolution</a>
    </li>
    <li class="dropdown-header">Visualization</li>
    <li>
      <a href="../articles/spata-v2-plotting-surface.html">Surface Plotting</a>
    </li>
    <li>
      <a href="../articles/spata-v2-plotting-misc.html">Miscellaneous</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Code &amp; Concept
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/cc-spatial-measures.html">Spatial Measures</a>
    </li>
    <li>
      <a href="../articles/cc-spatial-segmentation-vs-image-annotations.html">Spatial Segmentation vs. Image Annotations</a>
    </li>
    <li>
      <a href="../articles/cc-spatial-trajectory-screening.html">Spatial Trajectory Screening</a>
    </li>
    <li>
      <a href="../articles/cc-image-annotation-screening.html">Image Annotation Screening</a>
    </li>
    <li>
      <a href="../articles/cc-model-fitting-in-spatial-transcriptomic-studies.html">Model Fitting</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Model Fitting in Spatial Transcriptomics - Code
&amp; Concept</h1>
                        <h4 data-toc-skip class="author">Jan
Kueckelhaus</h4>
            
            <h4 data-toc-skip class="date">2022-08-20</h4>
      
      
      <div class="hidden name"><code>cc-model-fitting-in-spatial-transcriptomic-studies.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="prerequisites">1. Prerequisites<a class="anchor" aria-label="anchor" href="#prerequisites"></a>
</h2>
<p>Make sure to be familiar with the following vignettes:</p>
<ul>
<li><a href="cc-spatial-trajectory-screening.html">Spatial Trajectory
Screening - Code &amp; Concept</a></li>
<li><a href="cc-image-annotation-screening.html">Image Annotation
Screening - Code &amp; Concept</a></li>
</ul>
</div>
<div class="section level2">
<h2 id="introduction-overview">2. Introduction &amp; Overview<a class="anchor" aria-label="anchor" href="#introduction-overview"></a>
</h2>
<p>Model fitting is the mathematical procedure used in SPATA2 to screen
for genes with specific expression patterns. How the expression changes
of genes are inferred is explained in the vignettes mentioned in 1.
Prerequisites. This vignette explains step by step how the inferred
expression changes are fitted to predefined models that reflect
biologically relevant dynamics. To evaulate a gene-model-fit
<strong>SPATA2</strong> uses the concept of <em>Pearson correlation</em>
and/or <em>Area under the curve (AUC)</em>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://themilolab.com/" class="external-link">SPATA2</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SPATAData</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="the-data-input">3. The data input<a class="anchor" aria-label="anchor" href="#the-data-input"></a>
</h2>
<p>Both, spatial trajectory screening (STS) and image annotation
screening (IAS), infer gene expression changes by binning barcode-spots
and aligning the mean gene expression of each bin along the screening
direction. As this results in basically identical data structures the
subsequent model fitting works exactly the same for IAS and STS.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># spatial trajectory screening input </span>
<span class="co"># displayed on the left below</span>
<span class="va">traj_df_wide</span> <span class="op">&lt;-</span> 
  <span class="fu"><a href="../reference/getStsDf.html">getStsDf</a></span><span class="op">(</span>
    object <span class="op">=</span> <span class="va">object_t269</span>, <span class="co"># using object from previous tutorial</span>
    id <span class="op">=</span> <span class="st">"horizontal_mid"</span>, 
    variables <span class="op">=</span> <span class="st">"EGFR"</span>, 
    n_bins <span class="op">=</span> <span class="fl">50</span>, 
    format <span class="op">=</span> <span class="st">"wide"</span>
  <span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> 
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span>
    <span class="va">trajectory_order</span>, <span class="co"># numeric variable indicating the order of the bins / direction</span>
    <span class="va">proj_length_binned</span>, <span class="co"># the actual names of the bins</span>
    <span class="va">EGFR</span> <span class="co"># the gene expression along the direction</span>
    <span class="op">)</span>

<span class="co"># image annotation screening input</span>
<span class="co"># displayed on the right below</span>
<span class="va">ias_df_wide</span> <span class="op">&lt;-</span> 
  <span class="fu"><a href="../reference/getIasDf.html">getIasDf</a></span><span class="op">(</span>
    object <span class="op">=</span> <span class="va">object_t313</span>, <span class="co"># using object from previous tutorial</span>
    id <span class="op">=</span> <span class="st">"necrotic_center"</span>, 
    distance <span class="op">=</span> <span class="st">"2.25mm"</span>, 
    binwidth <span class="op">=</span> <span class="st">"112.5um"</span>,
    variables <span class="op">=</span> <span class="st">"FN1"</span>, 
    summarize_by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"bins_circle"</span>, <span class="st">"bins_angle"</span><span class="op">)</span>
  <span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> 
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span>
    <span class="va">bins_order</span>, <span class="co"># numeric variable indicating the order of the bins / direction</span>
    <span class="va">bins_circle</span>, <span class="co"># the actual names of the bins</span>
    <span class="va">FN1</span> <span class="co"># the gene expression along the direction</span>
  <span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> 
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">bins_circle</span> <span class="op">!=</span> <span class="st">"Core"</span><span class="op">)</span></code></pre></div>
<div class="row">
<div class="col-md-6">
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 50 x 3</span></span>
<span class="co">##    trajectory_order proj_length_binned  EGFR</span>
<span class="co">##               <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>              <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="co">## <span style="color: #BCBCBC;"> 1</span>                1 (-0.888,25.1]      0.900</span>
<span class="co">## <span style="color: #BCBCBC;"> 2</span>                2 (25.1,49.9]        0.978</span>
<span class="co">## <span style="color: #BCBCBC;"> 3</span>                3 (49.9,74.7]        0.948</span>
<span class="co">## <span style="color: #BCBCBC;"> 4</span>                4 (74.7,99.5]        0.940</span>
<span class="co">## <span style="color: #BCBCBC;"> 5</span>                5 (99.5,124]         0.977</span>
<span class="co">## <span style="color: #BCBCBC;"> 6</span>                6 (124,149]          0.973</span>
<span class="co">## <span style="color: #BCBCBC;"> 7</span>                7 (149,174]          0.971</span>
<span class="co">## <span style="color: #BCBCBC;"> 8</span>                8 (174,199]          0.980</span>
<span class="co">## <span style="color: #BCBCBC;"> 9</span>                9 (199,223]          0.939</span>
<span class="co">## <span style="color: #BCBCBC;">10</span>               10 (223,248]          0.971</span>
<span class="co">## <span style="color: #949494;"># i 40 more rows</span></span></code></pre>
</div>
<div class="col-md-6">
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 21 x 3</span></span>
<span class="co">##    bins_order bins_circle   FN1</span>
<span class="co">##         <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="co">## <span style="color: #BCBCBC;"> 1</span>          1 Circle 1    0.931</span>
<span class="co">## <span style="color: #BCBCBC;"> 2</span>          2 Circle 2    1    </span>
<span class="co">## <span style="color: #BCBCBC;"> 3</span>          3 Circle 3    0.929</span>
<span class="co">## <span style="color: #BCBCBC;"> 4</span>          4 Circle 4    0.915</span>
<span class="co">## <span style="color: #BCBCBC;"> 5</span>          5 Circle 5    0.886</span>
<span class="co">## <span style="color: #BCBCBC;"> 6</span>          6 Circle 6    0.812</span>
<span class="co">## <span style="color: #BCBCBC;"> 7</span>          7 Circle 7    0.769</span>
<span class="co">## <span style="color: #BCBCBC;"> 8</span>          8 Circle 8    0.723</span>
<span class="co">## <span style="color: #BCBCBC;"> 9</span>          9 Circle 9    0.651</span>
<span class="co">## <span style="color: #BCBCBC;">10</span>         10 Circle 10   0.592</span>
<span class="co">## <span style="color: #949494;"># i 11 more rows</span></span></code></pre>
</div>
</div>
<p>Although the data.frames contain results of different methods and the
names of the columns differ slightly the data structure is identical.
The most important parts of both data.frames are the variable indicating
the direction (<em>trajectory_order</em>, <em>bins_order</em>) and the
gene expression variables that are named like the gene.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">traj_df_wide</span>, mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">trajectory_order</span>, y <span class="op">=</span> <span class="va">EGFR</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_smooth.html" class="external-link">geom_smooth</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"loess"</span>, span <span class="op">=</span> <span class="fl">0.25</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>subtitle <span class="op">=</span> <span class="st">"STS - Sample 269"</span><span class="op">)</span>

<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">ias_df_wide</span>, mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">bins_order</span>, y <span class="op">=</span> <span class="va">FN1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_smooth.html" class="external-link">geom_smooth</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"loess"</span>, span <span class="op">=</span> <span class="fl">0.25</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>subtitle <span class="op">=</span> <span class="st">"IAS - Sample 313"</span><span class="op">)</span></code></pre></div>
<p><img src="cc-model-fitting-in-spatial-transcriptomic-studies_files/figure-html/unnamed-chunk-6-1.png" width="50%"><img src="cc-model-fitting-in-spatial-transcriptomic-studies_files/figure-html/unnamed-chunk-6-2.png" width="50%"></p>
</div>
<div class="section level2">
<h2 id="the-model-input">4. The model input<a class="anchor" aria-label="anchor" href="#the-model-input"></a>
</h2>
<p>In previous vignettes we describe how the expression changes from
genes along a spatial trajectory or in relation to an image annotation
are captured in numeric vectors (as plotted above). These vectors can be
fitted against models.</p>
<div class="section level3">
<h3 id="inbuilt-models">4.1 Inbuilt models<a class="anchor" aria-label="anchor" href="#inbuilt-models"></a>
</h3>
<p><strong>SPATA2</strong> provides the following predefined models.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># the number of bins (the resolution)</span>
<span class="va">n_bins</span> <span class="op">&lt;-</span> <span class="fl">50</span>

<span class="co"># plot models </span>
<span class="fu"><a href="../reference/showModels.html">showModels</a></span><span class="op">(</span>input <span class="op">=</span> <span class="va">n_bins</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>strip.background <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Bins"</span><span class="op">)</span></code></pre></div>
<p><img src="cc-model-fitting-in-spatial-transcriptomic-studies_files/figure-html/the-models-1.png" width="100%"></p>
<p>The corresponding data can be created with the function
<code><a href="../reference/create_model_df.html">create_model_df()</a></code>.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># create data.frame with model variables</span>
<span class="va">model_df</span> <span class="op">&lt;-</span> 
  <span class="fu"><a href="../reference/create_model_df.html">create_model_df</a></span><span class="op">(</span>
    input <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="va">n_bins</span>, <span class="co"># = 50</span>
    var_order <span class="op">=</span> <span class="st">"trajectory_order"</span>,
    model_subset <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"peak"</span>, <span class="st">"ascending"</span>, <span class="st">"descending"</span><span class="op">)</span>, <span class="co"># only use models that contain these catchphrases</span>
    model_remove <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"rev"</span>, <span class="st">"immediate"</span>, <span class="st">"two"</span><span class="op">)</span> <span class="co"># remove models that contain these catchphrases</span>
    <span class="op">)</span>

<span class="co"># show results </span>
<span class="va">model_df</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 50 x 11</span></span>
<span class="co">##    trajectory_order one_peak sharp_peak early_peak late_peak late_ascending</span>
<span class="co">##               <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="co">## <span style="color: #BCBCBC;"> 1</span>                1  0                0     0              0        0      </span>
<span class="co">## <span style="color: #BCBCBC;"> 2</span>                2  0.004<span style="text-decoration: underline;">11</span>          0     0.017<span style="text-decoration: underline;">0</span>         0        0.005<span style="text-decoration: underline;">16</span></span>
<span class="co">## <span style="color: #BCBCBC;"> 3</span>                3  0.016<span style="text-decoration: underline;">4</span>           0     0.067<span style="text-decoration: underline;">0</span>         0        0.010<span style="text-decoration: underline;">4</span> </span>
<span class="co">## <span style="color: #BCBCBC;"> 4</span>                4  0.036<span style="text-decoration: underline;">6</span>           0     0.146          0        0.015<span style="text-decoration: underline;">8</span> </span>
<span class="co">## <span style="color: #BCBCBC;"> 5</span>                5  0.064<span style="text-decoration: underline;">4</span>           0     0.250          0        0.021<span style="text-decoration: underline;">3</span> </span>
<span class="co">## <span style="color: #BCBCBC;"> 6</span>                6  0.099<span style="text-decoration: underline;">4</span>           0     0.371          0        0.026<span style="text-decoration: underline;">9</span> </span>
<span class="co">## <span style="color: #BCBCBC;"> 7</span>                7  0.141            0     0.5            0        0.032<span style="text-decoration: underline;">7</span> </span>
<span class="co">## <span style="color: #BCBCBC;"> 8</span>                8  0.188            0     0.629          0        0.038<span style="text-decoration: underline;">6</span> </span>
<span class="co">## <span style="color: #BCBCBC;"> 9</span>                9  0.241            0     0.75           0        0.044<span style="text-decoration: underline;">6</span> </span>
<span class="co">## <span style="color: #BCBCBC;">10</span>               10  0.298            0     0.854          0        0.050<span style="text-decoration: underline;">7</span> </span>
<span class="co">## <span style="color: #949494;"># i 40 more rows</span></span>
<span class="co">## <span style="color: #949494;"># i 5 more variables: linear_ascending &lt;dbl&gt;, abrupt_ascending &lt;dbl&gt;,</span></span>
<span class="co">## <span style="color: #949494;">#   late_descending &lt;dbl&gt;, linear_descending &lt;dbl&gt;, abrupt_descending &lt;dbl&gt;</span></span></code></pre>
<p>Note that models are always provided with a range of values from 0 to
1.</p>
</div>
<div class="section level3">
<h3 id="user-defined-models">4.2 User defined models<a class="anchor" aria-label="anchor" href="#user-defined-models"></a>
</h3>
<p>The set of models that is included in the screening process can be
expanded with user defined models. User defined models can either be
provided as numeric vectors or as formulas that create the numeric
vectors.</p>
<div class="section level4">
<h4 id="models-as-numeric-vectors">4.2.1 Models as numeric vectors<a class="anchor" aria-label="anchor" href="#models-as-numeric-vectors"></a>
</h4>
<p>For a model to be valid it must live up to two requirements:</p>
<ol style="list-style-type: decimal">
<li><p>The model must be of the same length as the variable that
contains the inferred gene expression change. (Length of the vector must
be equal to the number of bins.)</p></li>
<li><p>Itâ€™s values must range from 0 to 1.</p></li>
</ol>
<p>The example below creates the model <em>linear_ascending</em>.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># create the linear_ascending model </span>
<span class="va">linear_ascending_model</span> <span class="op">&lt;-</span> <span class="fl">1</span><span class="op">:</span><span class="va">n_bins</span> <span class="co"># n_bins = 50</span>

<span class="va">linear_ascending_model</span> <span class="op">&lt;-</span> <span class="fu">scales</span><span class="fu">::</span><span class="fu"><a href="https://scales.r-lib.org/reference/rescale.html" class="external-link">rescale</a></span><span class="op">(</span><span class="va">linear_ascending_model</span>, to <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span><span class="op">)</span>

<span class="va">linear_ascending_model</span></code></pre></div>
<pre><code><span class="co">##  [1] 0.00000000 0.02040816 0.04081633 0.06122449 0.08163265 0.10204082</span>
<span class="co">##  [7] 0.12244898 0.14285714 0.16326531 0.18367347 0.20408163 0.22448980</span>
<span class="co">## [13] 0.24489796 0.26530612 0.28571429 0.30612245 0.32653061 0.34693878</span>
<span class="co">## [19] 0.36734694 0.38775510 0.40816327 0.42857143 0.44897959 0.46938776</span>
<span class="co">## [25] 0.48979592 0.51020408 0.53061224 0.55102041 0.57142857 0.59183673</span>
<span class="co">## [31] 0.61224490 0.63265306 0.65306122 0.67346939 0.69387755 0.71428571</span>
<span class="co">## [37] 0.73469388 0.75510204 0.77551020 0.79591837 0.81632653 0.83673469</span>
<span class="co">## [43] 0.85714286 0.87755102 0.89795918 0.91836735 0.93877551 0.95918367</span>
<span class="co">## [49] 0.97959184 1.00000000</span></code></pre>
<p>The validity of the model input can be checked by adding it to the
models plotted with <code><a href="../reference/showModels.html">showModels()</a></code>. If the model is included
in the plot without any errors or warning messages the input is valid.
In general, models must be provided in a named list.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/showModels.html">showModels</a></span><span class="op">(</span>
  input <span class="op">=</span> <span class="va">n_bins</span>, 
  model_subset <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"linear_descending"</span>, <span class="st">"my_linear_ascending_model"</span><span class="op">)</span>,
  model_add <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>my_linear_ascending_model <span class="op">=</span> <span class="va">linear_ascending_model</span><span class="op">)</span>
<span class="op">)</span></code></pre></div>
<p><img src="cc-model-fitting-in-spatial-transcriptomic-studies_files/figure-html/unnamed-chunk-8-1.png" width="100%"></p>
<p>Providing models as numeric variables can be used to use specific
gene expression changes as models.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">traj_df_wide</span> <span class="op">&lt;-</span> 
  <span class="fu"><a href="../reference/getStsDf.html">getStsDf</a></span><span class="op">(</span>
    object <span class="op">=</span> <span class="va">object_t269</span>,
    id <span class="op">=</span> <span class="st">"horizontal_mid"</span>,
    variables <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"EGFR"</span>, <span class="st">"MAG"</span><span class="op">)</span>, 
    n_bins <span class="op">=</span> <span class="va">n_bins</span>, <span class="co"># = 100 </span>
    format <span class="op">=</span> <span class="st">"wide"</span>
  <span class="op">)</span>

<span class="co"># show results</span>
<span class="va">traj_df_wide</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 50 x 6</span></span>
<span class="co">##    trajectory_part proj_length_binned trajectory_order trajectory_part_order</span>
<span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>                         <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>                 <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span>
<span class="co">## <span style="color: #BCBCBC;"> 1</span> Part 1          (-0.888,25.1]                     1                     1</span>
<span class="co">## <span style="color: #BCBCBC;"> 2</span> Part 1          (25.1,49.9]                       2                     2</span>
<span class="co">## <span style="color: #BCBCBC;"> 3</span> Part 1          (49.9,74.7]                       3                     3</span>
<span class="co">## <span style="color: #BCBCBC;"> 4</span> Part 1          (74.7,99.5]                       4                     4</span>
<span class="co">## <span style="color: #BCBCBC;"> 5</span> Part 1          (99.5,124]                        5                     5</span>
<span class="co">## <span style="color: #BCBCBC;"> 6</span> Part 1          (124,149]                         6                     6</span>
<span class="co">## <span style="color: #BCBCBC;"> 7</span> Part 1          (149,174]                         7                     7</span>
<span class="co">## <span style="color: #BCBCBC;"> 8</span> Part 1          (174,199]                         8                     8</span>
<span class="co">## <span style="color: #BCBCBC;"> 9</span> Part 1          (199,223]                         9                     9</span>
<span class="co">## <span style="color: #BCBCBC;">10</span> Part 1          (223,248]                        10                    10</span>
<span class="co">## <span style="color: #949494;"># i 40 more rows</span></span>
<span class="co">## <span style="color: #949494;"># i 2 more variables: EGFR &lt;dbl&gt;, MAG &lt;dbl&gt;</span></span></code></pre>
<p>Selecting only the columns with the gene exprssion and converting the
data.frame to a list makes it a valid input for the argument
<code>model_add</code>.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/showModels.html">showModels</a></span><span class="op">(</span>
  input <span class="op">=</span> <span class="va">n_bins</span>, 
  model_subset <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"linear"</span>, <span class="st">"EGFR"</span>, <span class="st">"MAG"</span><span class="op">)</span>,
  model_add <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">as.list</a></span><span class="op">(</span><span class="va">traj_df_wide</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"EGFR"</span>, <span class="st">"MAG"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span>
<span class="op">)</span></code></pre></div>
<p><img src="cc-model-fitting-in-spatial-transcriptomic-studies_files/figure-html/unnamed-chunk-10-1.png" width="100%"></p>
</div>
<div class="section level4">
<h4 id="models-as-formulas">4.2.2 Models as formulas<a class="anchor" aria-label="anchor" href="#models-as-formulas"></a>
</h4>
<p>If the number of bins changes, the length of the numeric vector must
change, too. Therefore, instead of specifying the numeric vector over
and over again a function can be written which is then added as a
formula.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># write function that creates a linear_ascending model</span>

<span class="va">create_linear_ascending_model</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span>
  
  <span class="co"># take length of the input as length for the </span>
  <span class="co"># output vector</span>
  <span class="va">n_bins</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span>
  
  <span class="va">out</span> <span class="op">&lt;-</span> <span class="fu">scales</span><span class="fu">::</span><span class="fu"><a href="https://scales.r-lib.org/reference/rescale.html" class="external-link">rescale</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="va">n_bins</span>, to <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span><span class="op">)</span>
  
  <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">out</span><span class="op">)</span>
  
<span class="op">}</span>

<span class="co"># show results</span>
<span class="fu">create_linear_ascending_model</span><span class="op">(</span>x <span class="op">=</span> <span class="va">traj_df_wide</span><span class="op">[[</span><span class="st">"trajectory_order"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">##  [1] 0.00000000 0.02040816 0.04081633 0.06122449 0.08163265 0.10204082</span>
<span class="co">##  [7] 0.12244898 0.14285714 0.16326531 0.18367347 0.20408163 0.22448980</span>
<span class="co">## [13] 0.24489796 0.26530612 0.28571429 0.30612245 0.32653061 0.34693878</span>
<span class="co">## [19] 0.36734694 0.38775510 0.40816327 0.42857143 0.44897959 0.46938776</span>
<span class="co">## [25] 0.48979592 0.51020408 0.53061224 0.55102041 0.57142857 0.59183673</span>
<span class="co">## [31] 0.61224490 0.63265306 0.65306122 0.67346939 0.69387755 0.71428571</span>
<span class="co">## [37] 0.73469388 0.75510204 0.77551020 0.79591837 0.81632653 0.83673469</span>
<span class="co">## [43] 0.85714286 0.87755102 0.89795918 0.91836735 0.93877551 0.95918367</span>
<span class="co">## [49] 0.97959184 1.00000000</span></code></pre>
<p>Again, the formula must be provided in a named list.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/showModels.html">showModels</a></span><span class="op">(</span>
  input <span class="op">=</span> <span class="va">n_bins</span>, 
  model_subset <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sinus"</span>, <span class="st">"EGFR"</span>, <span class="st">"my_linear_model"</span><span class="op">)</span>,
  model_add <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>
    my_linear_model <span class="op">=</span> <span class="op">~</span> <span class="fu">create_linear_ascending_model</span><span class="op">(</span>x <span class="op">=</span> <span class="va">.x</span><span class="op">)</span>, <span class="co"># set to = .x</span>
    EGFR <span class="op">=</span> <span class="va">traj_df_wide</span><span class="op">[[</span><span class="st">"EGFR"</span><span class="op">]</span><span class="op">]</span> 
  <span class="op">)</span>
<span class="op">)</span></code></pre></div>
<p><img src="cc-model-fitting-in-spatial-transcriptomic-studies_files/figure-html/unnamed-chunk-12-1.png" width="100%"></p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="modelling-expression-changes">5. Modelling expression changes<a class="anchor" aria-label="anchor" href="#modelling-expression-changes"></a>
</h2>
<p>Evaluating a gene-model-fit works the same for both methods, STS and
IAS. The rest of this vignette uses a spatial trajectory as an
example.</p>
<p>We are using the same sample of the central nervous system malignancy
that we used in the <a href="">previous vignette</a> which features
three different, adjacent histological areas: Tumor, a transition zone
as well as infiltrated cortex. The spatial trajectory stays the same,
too.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">object_t269</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/downloadSpataObject.html">downloadSpataObject</a></span><span class="op">(</span>sample_name <span class="op">=</span> <span class="st">"269_T"</span><span class="op">)</span>

<span class="co"># creates trajectory manually </span>
<span class="co"># alternatively you can draw it interactively with `createSpatialTrajectories()`</span>
<span class="va">object_t269</span> <span class="op">&lt;-</span> 
  <span class="fu"><a href="../reference/createSpatialTrajectories.html">addSpatialTrajectory</a></span><span class="op">(</span>
    object <span class="op">=</span> <span class="va">object_t269</span>, 
    id <span class="op">=</span> <span class="st">"horizontal_mid"</span>,
    width <span class="op">=</span> <span class="fl">100</span>,
    start <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fl">100</span>, y <span class="op">=</span> <span class="fl">300</span><span class="op">)</span>,
    end <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fl">500</span>, y <span class="op">=</span> <span class="fl">300</span><span class="op">)</span>
  <span class="op">)</span></code></pre></div>
<p>We will use four genes as examples. Three of them follow an
interesting expression change along the trajectory, namely
<em>EGFR</em>, <em>MAG</em> and <em>SNAP25</em>. The last one,
<em>BCL9</em>, does not.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">genes_of_interest</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"EGFR"</span>, <span class="st">"MAG"</span>, <span class="st">"SNAP25"</span>, <span class="st">"BCL9"</span><span class="op">)</span>

<span class="fu"><a href="../reference/plotSurfaceComparison.html">plotSurfaceComparison</a></span><span class="op">(</span>
  object <span class="op">=</span> <span class="va">object_t269</span>, 
  color_by <span class="op">=</span> <span class="va">genes_of_interest</span>
<span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="../reference/ggpLayerTrajectories.html">ggpLayerTrajectories</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">object_t269</span>, ids <span class="op">=</span> <span class="st">"horizontal_mid"</span>, size <span class="op">=</span> <span class="fl">1.5</span><span class="op">)</span></code></pre></div>
<p><img src="cc-model-fitting-in-spatial-transcriptomic-studies_files/figure-html/example%20genes-1.png" width="100%"></p>
<p>The inferred gene expression changes look like this.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotTrajectoryLineplot.html">plotTrajectoryLineplot</a></span><span class="op">(</span>
  object <span class="op">=</span> <span class="va">object_t269</span>, 
  id <span class="op">=</span> <span class="st">"horizontal_mid"</span>,
  variables <span class="op">=</span> <span class="va">genes_of_interest</span>,
  n_bins <span class="op">=</span> <span class="va">n_bins</span>
<span class="op">)</span> </code></pre></div>
<p><img src="cc-model-fitting-in-spatial-transcriptomic-studies_files/figure-html/lineplots-1.png" width="100%"></p>
<p>The inferred expression changes along spatial trajectories are
captured in a data.frame that can be obtained via
<code><a href="../reference/getStsDf.html">getStsDf()</a></code>. Note that the inferred gene expression is
normalized to values from 0-1. 0 being the lowest expression and one the
highest within the scope of the trajectory.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">traj_df_example</span> <span class="op">&lt;-</span> 
  <span class="fu"><a href="../reference/getStsDf.html">getStsDf</a></span><span class="op">(</span>
    object <span class="op">=</span> <span class="va">object_t269</span>, 
    id <span class="op">=</span> <span class="st">"horizontal_mid"</span>,
    variables <span class="op">=</span> <span class="va">genes_of_interest</span>, 
    n_bins <span class="op">=</span> <span class="va">n_bins</span>,
    format <span class="op">=</span> <span class="st">"wide"</span>
  <span class="op">)</span>

<span class="co"># show results</span>
<span class="va">traj_df_example</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 50 x 8</span></span>
<span class="co">##    trajectory_part proj_length_binned trajectory_order trajectory_part_order</span>
<span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>                         <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>                 <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span>
<span class="co">## <span style="color: #BCBCBC;"> 1</span> Part 1          (-0.888,25.1]                     1                     1</span>
<span class="co">## <span style="color: #BCBCBC;"> 2</span> Part 1          (25.1,49.9]                       2                     2</span>
<span class="co">## <span style="color: #BCBCBC;"> 3</span> Part 1          (49.9,74.7]                       3                     3</span>
<span class="co">## <span style="color: #BCBCBC;"> 4</span> Part 1          (74.7,99.5]                       4                     4</span>
<span class="co">## <span style="color: #BCBCBC;"> 5</span> Part 1          (99.5,124]                        5                     5</span>
<span class="co">## <span style="color: #BCBCBC;"> 6</span> Part 1          (124,149]                         6                     6</span>
<span class="co">## <span style="color: #BCBCBC;"> 7</span> Part 1          (149,174]                         7                     7</span>
<span class="co">## <span style="color: #BCBCBC;"> 8</span> Part 1          (174,199]                         8                     8</span>
<span class="co">## <span style="color: #BCBCBC;"> 9</span> Part 1          (199,223]                         9                     9</span>
<span class="co">## <span style="color: #BCBCBC;">10</span> Part 1          (223,248]                        10                    10</span>
<span class="co">## <span style="color: #949494;"># i 40 more rows</span></span>
<span class="co">## <span style="color: #949494;"># i 4 more variables: EGFR &lt;dbl&gt;, MAG &lt;dbl&gt;, SNAP25 &lt;dbl&gt;, BCL9 &lt;dbl&gt;</span></span></code></pre>
<div class="section level3">
<h3 id="evaluating-fits-with-pearson-correlation">5.1 Evaluating fits with pearson correlation<a class="anchor" aria-label="anchor" href="#evaluating-fits-with-pearson-correlation"></a>
</h3>
<p>One option to evaluate a model fit is using pearson correlation.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># combine gene and model variables to a new data.frame</span>
<span class="va">combined_df</span> <span class="op">&lt;-</span> 
  <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span>
    <span class="va">traj_df_example</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"proj_length_binned"</span>, <span class="st">"trajectory_order"</span>,<span class="va">genes_of_interest</span><span class="op">)</span><span class="op">]</span>,
    <span class="va">model_df</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"abrupt_descending"</span>, <span class="st">"abrupt_ascending"</span>, <span class="st">"sharp_peak"</span><span class="op">)</span><span class="op">]</span>
    <span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> 
  <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span><span class="op">)</span>

<span class="co"># show results </span>
<span class="va">combined_df</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 50 x 9</span></span>
<span class="co">##    proj_length_binned trajectory_order  EGFR    MAG  SNAP25   BCL9</span>
<span class="co">##    <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>                         <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="co">## <span style="color: #BCBCBC;"> 1</span> (-0.888,25.1]                     1 0.900 0      0.014<span style="text-decoration: underline;">0</span>  0.183 </span>
<span class="co">## <span style="color: #BCBCBC;"> 2</span> (25.1,49.9]                       2 0.978 0.024<span style="text-decoration: underline;">5</span> 0.004<span style="text-decoration: underline;">75</span> 0.093<span style="text-decoration: underline;">1</span></span>
<span class="co">## <span style="color: #BCBCBC;"> 3</span> (49.9,74.7]                       3 0.948 0.017<span style="text-decoration: underline;">7</span> 0.004<span style="text-decoration: underline;">32</span> 0.509 </span>
<span class="co">## <span style="color: #BCBCBC;"> 4</span> (74.7,99.5]                       4 0.940 0      0.004<span style="text-decoration: underline;">48</span> 0.528 </span>
<span class="co">## <span style="color: #BCBCBC;"> 5</span> (99.5,124]                        5 0.977 0      0.004<span style="text-decoration: underline;">40</span> 0.605 </span>
<span class="co">## <span style="color: #BCBCBC;"> 6</span> (124,149]                         6 0.973 0.035<span style="text-decoration: underline;">3</span> 0.009<span style="text-decoration: underline;">14</span> 0.448 </span>
<span class="co">## <span style="color: #BCBCBC;"> 7</span> (149,174]                         7 0.971 0.034<span style="text-decoration: underline;">0</span> 0       0.173 </span>
<span class="co">## <span style="color: #BCBCBC;"> 8</span> (174,199]                         8 0.980 0.023<span style="text-decoration: underline;">1</span> 0       0.352 </span>
<span class="co">## <span style="color: #BCBCBC;"> 9</span> (199,223]                         9 0.939 0.022<span style="text-decoration: underline;">7</span> 0.004<span style="text-decoration: underline;">40</span> 0.691 </span>
<span class="co">## <span style="color: #BCBCBC;">10</span> (223,248]                        10 0.971 0.034<span style="text-decoration: underline;">7</span> 0.004<span style="text-decoration: underline;">48</span> 0.880 </span>
<span class="co">## <span style="color: #949494;"># i 40 more rows</span></span>
<span class="co">## <span style="color: #949494;"># i 3 more variables: abrupt_descending &lt;dbl&gt;, abrupt_ascending &lt;dbl&gt;,</span></span>
<span class="co">## <span style="color: #949494;">#   sharp_peak &lt;dbl&gt;</span></span></code></pre>
<p>Every observation/row in the data.frame <code>combined_df</code>
corresponds to one of the 50 bins to which the barcode-spots were
assigned. By comparing the course of <em>EGFR</em> along the trajectory
to the three models <em>abrupt_descending</em>,
<em>abrupt_ascending</em> and <em>sharp_peak</em> it becomes apparent
that that the expression change of <em>EGFR</em> resembles an abrupt
descending course.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">combined_df</span>, mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">trajectory_order</span>, y <span class="op">=</span> <span class="va">EGFR</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_smooth.html" class="external-link">geom_smooth</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"loess"</span>, span <span class="op">=</span> <span class="fl">0.25</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>subtitle <span class="op">=</span> <span class="st">"a) EGFR - Course"</span><span class="op">)</span>

<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">combined_df</span>, mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">trajectory_order</span>, y<span class="op">=</span> <span class="va">abrupt_descending</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_smooth.html" class="external-link">geom_smooth</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"loess"</span>, span <span class="op">=</span> <span class="fl">0.25</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>subtitle <span class="op">=</span> <span class="st">"b) Abrupt descending"</span><span class="op">)</span>

<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">combined_df</span>, mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">trajectory_order</span>, y <span class="op">=</span> <span class="va">abrupt_ascending</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_smooth.html" class="external-link">geom_smooth</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"loess"</span>, span <span class="op">=</span> <span class="fl">0.25</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>subtitle <span class="op">=</span> <span class="st">"c) Abrupt ascending"</span><span class="op">)</span>

<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">combined_df</span>, mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">trajectory_order</span>, y <span class="op">=</span> <span class="va">sharp_peak</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_smooth.html" class="external-link">geom_smooth</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"loess"</span>, span <span class="op">=</span> <span class="fl">0.25</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>subtitle <span class="op">=</span> <span class="st">"d) Sharp peak"</span><span class="op">)</span></code></pre></div>
<p><img src="cc-model-fitting-in-spatial-transcriptomic-studies_files/figure-html/unnamed-chunk-13-1.png" width="50%"><img src="cc-model-fitting-in-spatial-transcriptomic-studies_files/figure-html/unnamed-chunk-13-2.png" width="50%"><img src="cc-model-fitting-in-spatial-transcriptomic-studies_files/figure-html/unnamed-chunk-13-3.png" width="50%"><img src="cc-model-fitting-in-spatial-transcriptomic-studies_files/figure-html/unnamed-chunk-13-4.png" width="50%">
Every point in the plot above corresponds to a bin of barcode-spots
along the trajectory (here 50 bins). Itâ€™s position on the x-axis
displays the binâ€™s positioning in relation to the spatial direction of
the trajectory. Itâ€™s position on the y-axis displays the mean-expression
of gene <em>EGFR</em> of all barcode-spots that fall into the bin.</p>
<p>Plotting the mean-expression of each bin <strong>against the
trajectory order</strong> results in the inferred gene expression
changes, which, in case of EGFR, is best characterized with the model
<em>abrupt_descending</em>.</p>
<p>Plotting the mean-expression of each bin directly <strong>against the
models</strong> results in the figures below and classifies the
relationship between the expression change of a gene and the model.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Obviously, plotting *EGFR* against *EGFR* results in perfect correlation as both variables are identical. It serves as an example for what a perfect gene-model fit would look like</span>
<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">combined_df</span>, mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">EGFR</span>, y <span class="op">=</span> <span class="va">EGFR</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_smooth.html" class="external-link">geom_smooth</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"lm"</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>subtitle <span class="op">=</span> <span class="st">"a) EGFR - EGFR"</span><span class="op">)</span>

<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">combined_df</span>, mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">abrupt_descending</span>, y <span class="op">=</span> <span class="va">EGFR</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_smooth.html" class="external-link">geom_smooth</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"lm"</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span>  <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>subtitle <span class="op">=</span> <span class="st">"b) EGFR - Abrupt descending"</span><span class="op">)</span>

<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">combined_df</span>, mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">abrupt_ascending</span>, y <span class="op">=</span> <span class="va">EGFR</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_smooth.html" class="external-link">geom_smooth</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"lm"</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>subtitle <span class="op">=</span> <span class="st">"c) EGFR - Abrupt ascending"</span><span class="op">)</span>

<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">combined_df</span>, mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">sharp_peak</span>, y<span class="op">=</span> <span class="va">EGFR</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_smooth.html" class="external-link">geom_smooth</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"lm"</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>subtitle <span class="op">=</span> <span class="st">"c) EGFR - Sharp peak"</span><span class="op">)</span></code></pre></div>
<p><img src="cc-model-fitting-in-spatial-transcriptomic-studies_files/figure-html/unnamed-chunk-14-1.png" width="50%"><img src="cc-model-fitting-in-spatial-transcriptomic-studies_files/figure-html/unnamed-chunk-14-2.png" width="50%"><img src="cc-model-fitting-in-spatial-transcriptomic-studies_files/figure-html/unnamed-chunk-14-3.png" width="50%"><img src="cc-model-fitting-in-spatial-transcriptomic-studies_files/figure-html/unnamed-chunk-14-4.png" width="50%">
A positive slope of the fitted line indicates positive correlation
values. A negative slope of the fitted line indicates negative
correlation values. A slope that equals 0 indicates correlation values
of 0 and thus bad fits.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">breaks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">0.2</span>,<span class="fl">0.4</span>,<span class="fl">0.6</span>,<span class="fl">0.8</span>,<span class="fl">1</span><span class="op">)</span>

<span class="co"># shift data.frame for plotting and subsequent evaluation</span>
<span class="va">shifted_df</span> <span class="op">&lt;-</span> 
  <span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html" class="external-link">pivot_longer</a></span><span class="op">(</span>
    data <span class="op">=</span> <span class="va">combined_df</span>, 
    cols <span class="op">=</span> <span class="fu"><a href="https://tidyselect.r-lib.org/reference/all_of.html" class="external-link">any_of</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"abrupt_ascending"</span>, <span class="st">"abrupt_descending"</span>, <span class="st">"sharp_peak"</span><span class="op">)</span><span class="op">)</span>, 
    names_to <span class="op">=</span> <span class="st">"models"</span>,
    values_to <span class="op">=</span> <span class="st">"model_values"</span>
  <span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> 
  <span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html" class="external-link">pivot_longer</a></span><span class="op">(</span>
    cols <span class="op">=</span> <span class="fu"><a href="https://tidyselect.r-lib.org/reference/all_of.html" class="external-link">any_of</a></span><span class="op">(</span><span class="va">genes_of_interest</span><span class="op">)</span>,
    names_to <span class="op">=</span> <span class="st">"genes"</span>, 
    values_to <span class="op">=</span> <span class="st">"gene_values"</span>
  <span class="op">)</span>

<span class="co"># show results</span>
<span class="va">shifted_df</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 600 x 6</span></span>
<span class="co">##    proj_length_binned trajectory_order models     model_values genes gene_values</span>
<span class="co">##    <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>                         <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="co">## <span style="color: #BCBCBC;"> 1</span> (-0.888,25.1]                     1 abrupt_as~            0 EGFR       0.900 </span>
<span class="co">## <span style="color: #BCBCBC;"> 2</span> (-0.888,25.1]                     1 abrupt_as~            0 MAG        0     </span>
<span class="co">## <span style="color: #BCBCBC;"> 3</span> (-0.888,25.1]                     1 abrupt_as~            0 SNAP~      0.014<span style="text-decoration: underline;">0</span></span>
<span class="co">## <span style="color: #BCBCBC;"> 4</span> (-0.888,25.1]                     1 abrupt_as~            0 BCL9       0.183 </span>
<span class="co">## <span style="color: #BCBCBC;"> 5</span> (-0.888,25.1]                     1 abrupt_de~            1 EGFR       0.900 </span>
<span class="co">## <span style="color: #BCBCBC;"> 6</span> (-0.888,25.1]                     1 abrupt_de~            1 MAG        0     </span>
<span class="co">## <span style="color: #BCBCBC;"> 7</span> (-0.888,25.1]                     1 abrupt_de~            1 SNAP~      0.014<span style="text-decoration: underline;">0</span></span>
<span class="co">## <span style="color: #BCBCBC;"> 8</span> (-0.888,25.1]                     1 abrupt_de~            1 BCL9       0.183 </span>
<span class="co">## <span style="color: #BCBCBC;"> 9</span> (-0.888,25.1]                     1 sharp_peak            0 EGFR       0.900 </span>
<span class="co">## <span style="color: #BCBCBC;">10</span> (-0.888,25.1]                     1 sharp_peak            0 MAG        0     </span>
<span class="co">## <span style="color: #949494;"># i 590 more rows</span></span></code></pre>
<p>While variable <em>trajectory_order</em> contains the number of the
bin, <em>gene_values</em> contains the mean expression value of the gene
of each bin. The variable <em>model_values</em> contains the value a bin
would need to have in order to fit into the course of the model.
Plotting the relation between these gene-model-pairs and fitting a line
to it gives an impression in how far the inferred expression changes
resemble a model.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># plot gene-model-fits</span>
<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">shifted_df</span>, mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">model_values</span>, y <span class="op">=</span> <span class="va">gene_values</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_grid.html" class="external-link">facet_grid</a></span><span class="op">(</span> cols <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/vars.html" class="external-link">vars</a></span><span class="op">(</span><span class="va">models</span><span class="op">)</span>, rows <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/vars.html" class="external-link">vars</a></span><span class="op">(</span><span class="va">genes</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_smooth.html" class="external-link">geom_smooth</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"lm"</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_continuous</a></span><span class="op">(</span>
    breaks <span class="op">=</span> <span class="va">breaks</span>, 
    labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">breaks</span><span class="op">)</span>, 
    limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span>
  <span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_continuous</a></span><span class="op">(</span>
    breaks <span class="op">=</span> <span class="va">breaks</span>,
    labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">breaks</span><span class="op">)</span>, 
    limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span>
    <span class="op">)</span><span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p><img src="cc-model-fitting-in-spatial-transcriptomic-studies_files/figure-html/unnamed-chunk-16-1.png" width="100%"></p>
<p>Again, a positive slope of the fitted line indicates positive
correlation values (good fits, e.g.Â EGFR - abrupt_descending, MAG -
sharp_peak). A negative slope of the fitted line indicates negative
correlation values (inverse fits e.g.Â EGFR - abrupt_ascending). A slope
that equals 0 indicates correlation values of 0 and thus bad fits. E.g.
the inferred expression change of <em>BCL9</em> does not resemble any
model.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># good fit results in positive correlation values</span>
<span class="fu"><a href="https://rdrr.io/r/stats/cor.test.html" class="external-link">cor.test</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">combined_df</span><span class="op">[[</span><span class="st">"EGFR"</span><span class="op">]</span><span class="op">]</span>, y <span class="op">=</span> <span class="va">combined_df</span><span class="op">[[</span><span class="st">"abrupt_descending"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## </span>
<span class="co">##  Pearson's product-moment correlation</span>
<span class="co">## </span>
<span class="co">## data:  combined_df[["EGFR"]] and combined_df[["abrupt_descending"]]</span>
<span class="co">## t = 39.264, df = 48, p-value &lt; 2.2e-16</span>
<span class="co">## alternative hypothesis: true correlation is not equal to 0</span>
<span class="co">## 95 percent confidence interval:</span>
<span class="co">##  0.9732079 0.9913832</span>
<span class="co">## sample estimates:</span>
<span class="co">##       cor </span>
<span class="co">## 0.9847866</span></code></pre>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># inverse fit results in negative correlation values</span>
<span class="fu"><a href="https://rdrr.io/r/stats/cor.test.html" class="external-link">cor.test</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">combined_df</span><span class="op">[[</span><span class="st">"EGFR"</span><span class="op">]</span><span class="op">]</span>, y <span class="op">=</span> <span class="va">combined_df</span><span class="op">[[</span><span class="st">"abrupt_ascending"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## </span>
<span class="co">##  Pearson's product-moment correlation</span>
<span class="co">## </span>
<span class="co">## data:  combined_df[["EGFR"]] and combined_df[["abrupt_ascending"]]</span>
<span class="co">## t = -32.492, df = 48, p-value &lt; 2.2e-16</span>
<span class="co">## alternative hypothesis: true correlation is not equal to 0</span>
<span class="co">## 95 percent confidence interval:</span>
<span class="co">##  -0.9875288 -0.9613812</span>
<span class="co">## sample estimates:</span>
<span class="co">##       cor </span>
<span class="co">## -0.978014</span></code></pre>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># bad fit results in correlation values around 0 </span>
<span class="fu"><a href="https://rdrr.io/r/stats/cor.test.html" class="external-link">cor.test</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">combined_df</span><span class="op">[[</span><span class="st">"BCL9"</span><span class="op">]</span><span class="op">]</span>, y <span class="op">=</span> <span class="va">combined_df</span><span class="op">[[</span><span class="st">"sharp_peak"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## </span>
<span class="co">##  Pearson's product-moment correlation</span>
<span class="co">## </span>
<span class="co">## data:  combined_df[["BCL9"]] and combined_df[["sharp_peak"]]</span>
<span class="co">## t = -2.8295, df = 48, p-value = 0.006785</span>
<span class="co">## alternative hypothesis: true correlation is not equal to 0</span>
<span class="co">## 95 percent confidence interval:</span>
<span class="co">##  -0.5939340 -0.1114769</span>
<span class="co">## sample estimates:</span>
<span class="co">##        cor </span>
<span class="co">## -0.3780927</span></code></pre>
<p>The output of <code><a href="../reference/spatialTrajectoryScreening.html">spatialTrajectoryScreening()</a></code> and
<code><a href="../reference/imageAnnotationScreening.html">imageAnnotationScreening()</a></code> comes in form of a data.frame
where every observation/row corresponds to a gene-model-fit
evaulation.</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">eval_with_corr</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">input_model</span>, <span class="va">input_gene_expression</span><span class="op">)</span><span class="op">{</span>
  
  <span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/cor.test.html" class="external-link">cor.test</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">input_model</span>, y <span class="op">=</span> <span class="va">input_gene_expression</span><span class="op">)</span>
  
  <span class="va">out</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_c.html" class="external-link">str_c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">estimate</span>,<span class="fl">2</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">p.value</span>,<span class="fl">2</span><span class="op">)</span>, sep <span class="op">=</span> <span class="st">"_"</span><span class="op">)</span>
  
  <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">out</span><span class="op">)</span>
  
<span class="op">}</span>

<span class="va">corr_results</span> <span class="op">&lt;-</span> 
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span>.data <span class="op">=</span> <span class="va">shifted_df</span>, <span class="va">models</span>, <span class="va">genes</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> 
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarize</a></span><span class="op">(</span>
    res <span class="op">=</span> <span class="fu">eval_with_corr</span><span class="op">(</span>input_model <span class="op">=</span> <span class="va">model_values</span>, input_gene_expression <span class="op">=</span> <span class="va">gene_values</span><span class="op">)</span>
  <span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> 
  <span class="fu"><a href="https://tidyr.tidyverse.org/reference/separate.html" class="external-link">separate</a></span><span class="op">(</span>col <span class="op">=</span> <span class="va">res</span>, into <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"corr"</span>, <span class="st">"pvalue"</span><span class="op">)</span>, sep <span class="op">=</span> <span class="st">"_"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> 
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">genes</span>, <span class="va">models</span>, <span class="va">corr</span>, <span class="va">pvalue</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> 
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="va">genes</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> 
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">genes</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> 
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/desc.html" class="external-link">desc</a></span><span class="op">(</span><span class="va">corr</span><span class="op">)</span>, .by_group <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
  
<span class="va">corr_results</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 12 x 4</span></span>
<span class="co">## <span style="color: #949494;"># Groups:   genes [4]</span></span>
<span class="co">##    genes  models            corr  pvalue</span>
<span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> </span>
<span class="co">## <span style="color: #BCBCBC;"> 1</span> BCL9   abrupt_descending 0.41  0     </span>
<span class="co">## <span style="color: #BCBCBC;"> 2</span> BCL9   abrupt_ascending  -0.44 0     </span>
<span class="co">## <span style="color: #BCBCBC;"> 3</span> BCL9   sharp_peak        -0.38 0.01  </span>
<span class="co">## <span style="color: #BCBCBC;"> 4</span> EGFR   abrupt_descending 0.98  0     </span>
<span class="co">## <span style="color: #BCBCBC;"> 5</span> EGFR   sharp_peak        0.11  0.44  </span>
<span class="co">## <span style="color: #BCBCBC;"> 6</span> EGFR   abrupt_ascending  -0.98 0     </span>
<span class="co">## <span style="color: #BCBCBC;"> 7</span> MAG    sharp_peak        0.97  0     </span>
<span class="co">## <span style="color: #BCBCBC;"> 8</span> MAG    abrupt_ascending  0.15  0.31  </span>
<span class="co">## <span style="color: #BCBCBC;"> 9</span> MAG    abrupt_descending -0.08 0.6   </span>
<span class="co">## <span style="color: #BCBCBC;">10</span> SNAP25 abrupt_ascending  0.97  0     </span>
<span class="co">## <span style="color: #BCBCBC;">11</span> SNAP25 abrupt_descending -0.98 0     </span>
<span class="co">## <span style="color: #BCBCBC;">12</span> SNAP25 sharp_peak        -0.2  0.15</span></code></pre>
</div>
<div class="section level3">
<h3 id="evaluating-fits-via-area-under-the-curve-auc">5.2 Evaluating fits via area under the curve (AUC)<a class="anchor" aria-label="anchor" href="#evaluating-fits-via-area-under-the-curve-auc"></a>
</h3>
<p>Evaluating gene-model-fits via area under the curve (AUC) contains
three steps.</p>
<ul>
<li><ol style="list-style-type: decimal">
<li>Calculating the residuals of each model fit</li>
</ol></li>
<li><ol start="2" style="list-style-type: decimal">
<li>Calculating the residuals area under the curve (RAUC)</li>
</ol></li>
<li><ol start="3" style="list-style-type: decimal">
<li>Normalizing RAUC to create the <em>Residuals Area Over the
Curve</em> (RAOC)</li>
</ol></li>
</ul>
<p>For this part of the vignette we are using the gene <em>MAG</em>.</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># create appropriate theme</span>
<span class="va">theme_modelling</span> <span class="op">&lt;-</span> 
  <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span><span class="op">)</span>,
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>
      axis.line.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_line</a></span><span class="op">(</span>
        arrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/arrow.html" class="external-link">arrow</a></span><span class="op">(</span>length <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fl">0.075</span>, <span class="st">"inches"</span><span class="op">)</span>, type <span class="op">=</span> <span class="st">"closed"</span><span class="op">)</span>
      <span class="op">)</span>
    <span class="op">)</span>
  <span class="op">)</span>

<span class="va">MAG_plot</span> <span class="op">&lt;-</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">traj_df_example</span>, mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">trajectory_order</span>, y <span class="op">=</span> <span class="va">MAG</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> 
  <span class="co"># use points to plot every bin</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> 
  <span class="co"># use lines to connect</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html" class="external-link">geom_path</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> 
  <span class="va">theme_modelling</span>

<span class="va">peak_plot</span> <span class="op">&lt;-</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">model_df</span>, mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">trajectory_order</span>, y <span class="op">=</span> <span class="va">sharp_peak</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html" class="external-link">geom_path</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="va">theme_modelling</span>

<span class="co"># plot results</span>
<span class="va">MAG_plot</span>

<span class="va">peak_plot</span></code></pre></div>
<p><img src="cc-model-fitting-in-spatial-transcriptomic-studies_files/figure-html/unnamed-chunk-19-1.png" width="50%"><img src="cc-model-fitting-in-spatial-transcriptomic-studies_files/figure-html/unnamed-chunk-19-2.png" width="50%"></p>
<div class="section level4">
<h4 id="calculating-the-residuals">5.2.1 Calculating the residuals<a class="anchor" aria-label="anchor" href="#calculating-the-residuals"></a>
</h4>
<p>The residuals are calculated by subtracting the inferred gene
expression from the model variable of interest and taking the absolute
value from that via <code><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">base::abs()</a></code>.</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># join both data.frames</span>
<span class="co"># note that both data.frames contain the same number of rows</span>
<span class="co"># as arguments `input` (create_model_df()) and `n_bins` (getStsDf())</span>
<span class="co"># were specified with value 50</span>
<span class="va">joined_df</span> <span class="op">&lt;-</span> 
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">left_join</a></span><span class="op">(</span>
    x <span class="op">=</span> <span class="va">traj_df_example</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"proj_length_binned"</span>, <span class="st">"trajectory_order"</span>, <span class="st">"MAG"</span><span class="op">)</span><span class="op">]</span>,
    y <span class="op">=</span> <span class="va">model_df</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"trajectory_order"</span>, <span class="st">"sharp_peak"</span>, <span class="st">"abrupt_descending"</span><span class="op">)</span><span class="op">]</span>, 
    by <span class="op">=</span> <span class="st">"trajectory_order"</span>
  <span class="op">)</span>

<span class="co"># show results</span>
<span class="va">joined_df</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 50 x 5</span></span>
<span class="co">##    proj_length_binned trajectory_order    MAG sharp_peak abrupt_descending</span>
<span class="co">##    <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>                         <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="co">## <span style="color: #BCBCBC;"> 1</span> (-0.888,25.1]                     1 0               0                 1</span>
<span class="co">## <span style="color: #BCBCBC;"> 2</span> (25.1,49.9]                       2 0.024<span style="text-decoration: underline;">5</span>          0                 1</span>
<span class="co">## <span style="color: #BCBCBC;"> 3</span> (49.9,74.7]                       3 0.017<span style="text-decoration: underline;">7</span>          0                 1</span>
<span class="co">## <span style="color: #BCBCBC;"> 4</span> (74.7,99.5]                       4 0               0                 1</span>
<span class="co">## <span style="color: #BCBCBC;"> 5</span> (99.5,124]                        5 0               0                 1</span>
<span class="co">## <span style="color: #BCBCBC;"> 6</span> (124,149]                         6 0.035<span style="text-decoration: underline;">3</span>          0                 1</span>
<span class="co">## <span style="color: #BCBCBC;"> 7</span> (149,174]                         7 0.034<span style="text-decoration: underline;">0</span>          0                 1</span>
<span class="co">## <span style="color: #BCBCBC;"> 8</span> (174,199]                         8 0.023<span style="text-decoration: underline;">1</span>          0                 1</span>
<span class="co">## <span style="color: #BCBCBC;"> 9</span> (199,223]                         9 0.022<span style="text-decoration: underline;">7</span>          0                 1</span>
<span class="co">## <span style="color: #BCBCBC;">10</span> (223,248]                        10 0.034<span style="text-decoration: underline;">7</span>          0                 1</span>
<span class="co">## <span style="color: #949494;"># i 40 more rows</span></span></code></pre>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># compute residuals</span>
<span class="va">residuals_df</span> <span class="op">&lt;-</span> 
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>
    .data <span class="op">=</span> <span class="va">joined_df</span>,
    <span class="co"># use base::abs() to ensure positive values</span>
    sharp_peak_res <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="op">(</span><span class="va">sharp_peak</span> <span class="op">-</span> <span class="va">MAG</span><span class="op">)</span><span class="op">)</span>,
    abrupt_descending_res <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="op">(</span><span class="va">abrupt_descending</span> <span class="op">-</span> <span class="va">MAG</span><span class="op">)</span><span class="op">)</span>
  <span class="op">)</span>

<span class="co"># show results</span>
<span class="va">residuals_df</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 50 x 7</span></span>
<span class="co">##    proj_length_binned trajectory_order    MAG sharp_peak abrupt_descending</span>
<span class="co">##    <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>                         <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="co">## <span style="color: #BCBCBC;"> 1</span> (-0.888,25.1]                     1 0               0                 1</span>
<span class="co">## <span style="color: #BCBCBC;"> 2</span> (25.1,49.9]                       2 0.024<span style="text-decoration: underline;">5</span>          0                 1</span>
<span class="co">## <span style="color: #BCBCBC;"> 3</span> (49.9,74.7]                       3 0.017<span style="text-decoration: underline;">7</span>          0                 1</span>
<span class="co">## <span style="color: #BCBCBC;"> 4</span> (74.7,99.5]                       4 0               0                 1</span>
<span class="co">## <span style="color: #BCBCBC;"> 5</span> (99.5,124]                        5 0               0                 1</span>
<span class="co">## <span style="color: #BCBCBC;"> 6</span> (124,149]                         6 0.035<span style="text-decoration: underline;">3</span>          0                 1</span>
<span class="co">## <span style="color: #BCBCBC;"> 7</span> (149,174]                         7 0.034<span style="text-decoration: underline;">0</span>          0                 1</span>
<span class="co">## <span style="color: #BCBCBC;"> 8</span> (174,199]                         8 0.023<span style="text-decoration: underline;">1</span>          0                 1</span>
<span class="co">## <span style="color: #BCBCBC;"> 9</span> (199,223]                         9 0.022<span style="text-decoration: underline;">7</span>          0                 1</span>
<span class="co">## <span style="color: #BCBCBC;">10</span> (223,248]                        10 0.034<span style="text-decoration: underline;">7</span>          0                 1</span>
<span class="co">## <span style="color: #949494;"># i 40 more rows</span></span>
<span class="co">## <span style="color: #949494;"># i 2 more variables: sharp_peak_res &lt;dbl&gt;, abrupt_descending_res &lt;dbl&gt;</span></span></code></pre>
<p>Plotting the residuals along the trajectoryâ€™s course visualizes the
quality of the fit. The variable <em>trajectory_order</em> refers to the
ordering of the bins and indicates the direction of the trajectory.</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">linesize</span> <span class="op">&lt;-</span> <span class="fl">1.25</span>

<span class="co"># good fit </span>
<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">residuals_df</span>, mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">trajectory_order</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> 
  <span class="co"># course of MAG</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html" class="external-link">geom_line</a></span><span class="op">(</span>mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>y <span class="op">=</span> <span class="va">MAG</span><span class="op">)</span>, color <span class="op">=</span> <span class="st">"forestgreen"</span>, size <span class="op">=</span> <span class="va">linesize</span><span class="op">)</span> <span class="op">+</span> 
  <span class="co"># course of the model</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html" class="external-link">geom_line</a></span><span class="op">(</span>mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>y <span class="op">=</span> <span class="va">sharp_peak</span><span class="op">)</span>, color <span class="op">=</span> <span class="st">"blue"</span>, size <span class="op">=</span> <span class="va">linesize</span><span class="op">)</span> <span class="op">+</span> 
  <span class="co"># residuals</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_ribbon.html" class="external-link">geom_area</a></span><span class="op">(</span>mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>y <span class="op">=</span> <span class="va">sharp_peak_res</span><span class="op">)</span>, fill <span class="op">=</span> <span class="st">"red"</span>, alpha <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>y <span class="op">=</span> <span class="cn">NULL</span>, subtitle <span class="op">=</span> <span class="st">"a) Good Fit"</span><span class="op">)</span> <span class="op">+</span>
  <span class="va">theme_modelling</span>


<span class="co"># bad fit fit </span>
<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">residuals_df</span>, mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">trajectory_order</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> 
  <span class="co"># course of MAG</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html" class="external-link">geom_line</a></span><span class="op">(</span>mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>y <span class="op">=</span> <span class="va">MAG</span><span class="op">)</span>, color <span class="op">=</span> <span class="st">"forestgreen"</span>, size <span class="op">=</span> <span class="va">linesize</span><span class="op">)</span> <span class="op">+</span> 
  <span class="co"># course of the model</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html" class="external-link">geom_line</a></span><span class="op">(</span>mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>y <span class="op">=</span> <span class="va">abrupt_descending</span><span class="op">)</span>, color <span class="op">=</span> <span class="st">"blue"</span>, size <span class="op">=</span> <span class="va">linesize</span><span class="op">)</span> <span class="op">+</span> 
  <span class="co"># residuals</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_ribbon.html" class="external-link">geom_area</a></span><span class="op">(</span>mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>y <span class="op">=</span> <span class="va">abrupt_descending_res</span><span class="op">)</span>, fill <span class="op">=</span> <span class="st">"red"</span>, alpha <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>y <span class="op">=</span> <span class="cn">NULL</span>, subtitle <span class="op">=</span> <span class="st">"b) Bad Fit"</span><span class="op">)</span> <span class="op">+</span>
  <span class="va">theme_modelling</span></code></pre></div>
<p><img src="cc-model-fitting-in-spatial-transcriptomic-studies_files/figure-html/unnamed-chunk-21-1.png" width="50%"><img src="cc-model-fitting-in-spatial-transcriptomic-studies_files/figure-html/unnamed-chunk-21-2.png" width="50%"></p>
</div>
<div class="section level4">
<h4 id="calculating-the-residuals-area-under-the-curve-rauc">5.2.2 Calculating the residuals area under the curve (RAUC)<a class="anchor" aria-label="anchor" href="#calculating-the-residuals-area-under-the-curve-rauc"></a>
</h4>
<p>The AUC of the residuals is calculated with the R function
<code><a href="https://rdrr.io/pkg/pracma/man/trapz.html" class="external-link">pracma::trapz()</a></code> which implements trapezoidal
integration.</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># good fit (low AUC)</span>
<span class="va">auc_peak</span> <span class="op">&lt;-</span> <span class="fu">pracma</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/pracma/man/trapz.html" class="external-link">trapz</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">residuals_df</span><span class="op">[[</span><span class="st">"trajectory_order"</span><span class="op">]</span><span class="op">]</span>, y <span class="op">=</span> <span class="va">residuals_df</span><span class="op">[[</span><span class="st">"sharp_peak_res"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>

<span class="va">auc_peak</span></code></pre></div>
<pre><code><span class="co">## [1] 4.711071</span></code></pre>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># bad fit (high AUC)</span>
<span class="va">auc_desc</span> <span class="op">&lt;-</span> <span class="fu">pracma</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/pracma/man/trapz.html" class="external-link">trapz</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">residuals_df</span><span class="op">[[</span><span class="st">"trajectory_order"</span><span class="op">]</span><span class="op">]</span>, y <span class="op">=</span> <span class="va">residuals_df</span><span class="op">[[</span><span class="st">"abrupt_descending_res"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>

<span class="va">auc_desc</span></code></pre></div>
<pre><code><span class="co">## [1] 23.96847</span></code></pre>
<p>Results of this concept can be visualized with the function
<code><a href="../reference/plotTrajectoryLineplotFitted.html">plotTrajectoryLineplotFitted()</a></code> for every gene with every
model. Each of the three genes has at least one model that it fits well
to. The gene BCL9 has not.</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">models</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"abrupt_descending"</span>, <span class="st">"sharp_peak"</span>, <span class="st">"abrupt_ascending"</span><span class="op">)</span>

<span class="fu"><a href="../reference/plotTrajectoryLineplotFitted.html">plotTrajectoryLineplotFitted</a></span><span class="op">(</span>
  object <span class="op">=</span> <span class="va">object_t269</span>,
  id <span class="op">=</span> <span class="st">"horizontal_mid"</span>,
  variable <span class="op">=</span> <span class="st">"EGFR"</span>,
  n_bins <span class="op">=</span> <span class="fl">50</span>,
  model_subset <span class="op">=</span> <span class="va">models</span>, 
  nrow <span class="op">=</span> <span class="fl">1</span>
<span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="../reference/legendBottom.html">legendNone</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>subtitle <span class="op">=</span> <span class="st">"a) EGFR"</span><span class="op">)</span>

<span class="fu"><a href="../reference/plotTrajectoryLineplotFitted.html">plotTrajectoryLineplotFitted</a></span><span class="op">(</span>
  object <span class="op">=</span> <span class="va">object_t269</span>,
  id <span class="op">=</span> <span class="st">"horizontal_mid"</span>,
  variable <span class="op">=</span> <span class="st">"MAG"</span>,
  n_bins <span class="op">=</span> <span class="fl">50</span>,
  model_subset <span class="op">=</span> <span class="va">models</span>, 
  nrow <span class="op">=</span> <span class="fl">1</span>
<span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="../reference/legendBottom.html">legendNone</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>subtitle <span class="op">=</span> <span class="st">"b) MAG"</span><span class="op">)</span>

<span class="fu"><a href="../reference/plotTrajectoryLineplotFitted.html">plotTrajectoryLineplotFitted</a></span><span class="op">(</span>
  object <span class="op">=</span> <span class="va">object_t269</span>,
  id <span class="op">=</span> <span class="st">"horizontal_mid"</span>,
  variable <span class="op">=</span> <span class="st">"SNAP25"</span>,
  n_bins <span class="op">=</span> <span class="fl">50</span>,
  model_subset <span class="op">=</span> <span class="va">models</span>, 
  nrow <span class="op">=</span> <span class="fl">1</span>
<span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="../reference/legendBottom.html">legendNone</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>subtitle <span class="op">=</span> <span class="st">"c) SNAP25"</span><span class="op">)</span>

 <span class="fu"><a href="../reference/plotTrajectoryLineplotFitted.html">plotTrajectoryLineplotFitted</a></span><span class="op">(</span>
  object <span class="op">=</span> <span class="va">object_t269</span>,
  id <span class="op">=</span> <span class="st">"horizontal_mid"</span>,
  variable <span class="op">=</span> <span class="st">"BCL9"</span>,
  n_bins <span class="op">=</span> <span class="fl">50</span>,
  model_subset <span class="op">=</span> <span class="va">models</span>, 
  nrow <span class="op">=</span> <span class="fl">1</span>
<span class="op">)</span> <span class="op">+</span> 
   <span class="fu"><a href="../reference/legendBottom.html">legendNone</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> 
   <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>subtitle <span class="op">=</span> <span class="st">"d) BCL9"</span><span class="op">)</span></code></pre></div>
<p><img src="cc-model-fitting-in-spatial-transcriptomic-studies_files/figure-html/unnamed-chunk-23-1.png" width="50%"><img src="cc-model-fitting-in-spatial-transcriptomic-studies_files/figure-html/unnamed-chunk-23-2.png" width="50%"><img src="cc-model-fitting-in-spatial-transcriptomic-studies_files/figure-html/unnamed-chunk-23-3.png" width="50%"><img src="cc-model-fitting-in-spatial-transcriptomic-studies_files/figure-html/unnamed-chunk-23-4.png" width="50%"></p>
</div>
<div class="section level4">
<h4 id="normalizing-rauc-into-the-residuals-area-over-the-curve-raoc">5.2.3 Normalizing RAUC into the <em>Residuals Area Over the
Curve</em> (RAOC)<a class="anchor" aria-label="anchor" href="#normalizing-rauc-into-the-residuals-area-over-the-curve-raoc"></a>
</h4>
<p>Drawback of AUC is that the evaluation score depends on the number of
bins. The maximum AUC if <code>n_bins</code> = 50 is 50 (50x1 = 50,
50x<strong>1</strong> because the expression values will always range
from 0-1). Therefore, in case of <code>n_bins</code> = 50 an AUC of 24
is bad but if <code>n_bins</code> is set to 500 an AUC of 24 indicates
an excellent fit. Additionally, it is not intuitive that a low score
indicates a good fit and vice versa.</p>
<p>For that matter, we suggest to use the <em>Residuals Area Over the
Curve</em> for evaluation. Two steps are needed to convert the RAUC to
RAOC.</p>
<p>First, the AUC value is normalized against the number of bins such
that it ranges from 0 to 1 (Normalization, nAUC). Second, the difference
between 1 and nAUC is calculated.</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">nAUC_peak</span> <span class="op">&lt;-</span> <span class="va">auc_peak</span> <span class="op">/</span> <span class="va">n_bins</span> <span class="co"># n_bins = 50</span>

<span class="va">RAOC_peak</span> <span class="op">&lt;-</span> <span class="fl">1</span> <span class="op">-</span> <span class="va">nAUC_peak</span>

<span class="co"># good fit</span>
<span class="va">RAOC_peak</span></code></pre></div>
<pre><code><span class="co">## [1] 0.9057786</span></code></pre>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">nAUC_desc</span> <span class="op">&lt;-</span> <span class="va">auc_desc</span> <span class="op">/</span> <span class="va">n_bins</span> <span class="co"># n_bins = 50</span>

<span class="va">RAOC_desc</span> <span class="op">&lt;-</span> <span class="fl">1</span> <span class="op">-</span> <span class="va">nAUC_desc</span>

<span class="co"># bad fit</span>
<span class="va">RAOC_desc</span></code></pre></div>
<pre><code><span class="co">## [1] 0.5206306</span></code></pre>
<p>This concept is visualized below. The bigger the area over the curve
the better the fit.</p>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">normalized_residuals_df</span> <span class="op">&lt;-</span> 
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>
    .data <span class="op">=</span> <span class="va">residuals_df</span>,
    trajectory_order <span class="op">=</span> <span class="va">trajectory_order</span><span class="op">/</span><span class="fl">50</span>, 
    sharp_peak_res_above <span class="op">=</span> <span class="fl">1</span><span class="op">-</span><span class="op">(</span><span class="va">sharp_peak_res</span><span class="op">/</span><span class="fl">50</span><span class="op">)</span>
  <span class="op">)</span>

<span class="co"># plot with area under the curve</span>
<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">normalized_residuals_df</span>, mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">trajectory_order</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> 
  <span class="co"># course of MAG</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html" class="external-link">geom_line</a></span><span class="op">(</span>mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>y <span class="op">=</span> <span class="va">MAG</span><span class="op">)</span>, color <span class="op">=</span> <span class="st">"black"</span>, size <span class="op">=</span> <span class="va">linesize</span><span class="op">)</span> <span class="op">+</span> 
  <span class="co"># course of the model</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html" class="external-link">geom_line</a></span><span class="op">(</span>mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>y <span class="op">=</span> <span class="va">sharp_peak</span><span class="op">)</span>, color <span class="op">=</span> <span class="st">"blue"</span>, size <span class="op">=</span> <span class="va">linesize</span><span class="op">)</span> <span class="op">+</span> 
  <span class="co"># residuals</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_ribbon.html" class="external-link">geom_area</a></span><span class="op">(</span>mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>y <span class="op">=</span> <span class="va">sharp_peak_res</span><span class="op">)</span>, fill <span class="op">=</span> <span class="st">"tomato"</span>, alpha <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>y <span class="op">=</span> <span class="cn">NULL</span>, subtitle <span class="op">=</span> <span class="st">"a) Residuals Area Under the Curve"</span><span class="op">)</span> <span class="op">+</span>
  <span class="va">theme_modelling</span>


<span class="co"># plot with area over the curve </span>
<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">normalized_residuals_df</span>, mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">trajectory_order</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> 
  <span class="co"># course of MAG</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html" class="external-link">geom_line</a></span><span class="op">(</span>mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>y <span class="op">=</span> <span class="va">MAG</span><span class="op">)</span>, color <span class="op">=</span> <span class="st">"black"</span>, size <span class="op">=</span> <span class="va">linesize</span><span class="op">)</span> <span class="op">+</span> 
  <span class="co"># course of the model</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html" class="external-link">geom_line</a></span><span class="op">(</span>mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>y <span class="op">=</span> <span class="va">sharp_peak</span><span class="op">)</span>, color <span class="op">=</span> <span class="st">"blue"</span>, size <span class="op">=</span> <span class="va">linesize</span><span class="op">)</span> <span class="op">+</span> 
  <span class="co"># residuals, but over the curve</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_ribbon.html" class="external-link">geom_ribbon</a></span><span class="op">(</span>
    mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>
      ymin <span class="op">=</span> <span class="va">sharp_peak_res</span>,
      ymax <span class="op">=</span> <span class="fl">1</span>
      <span class="op">)</span>, fill <span class="op">=</span> <span class="st">"forestgreen"</span>, alpha <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>y <span class="op">=</span> <span class="cn">NULL</span>, subtitle <span class="op">=</span> <span class="st">"b) Residuals Area Over the Curve"</span><span class="op">)</span> <span class="op">+</span>
  <span class="va">theme_modelling</span></code></pre></div>
<p><img src="cc-model-fitting-in-spatial-transcriptomic-studies_files/figure-html/unnamed-chunk-25-1.png" width="50%"><img src="cc-model-fitting-in-spatial-transcriptomic-studies_files/figure-html/unnamed-chunk-25-2.png" width="50%"></p>
<p>Note that the variable <em>trajectory_order</em> has been scaled to
range from 0 - 1. The maximum area over the curve indicating a perfect
fit becomes 1 (1x1) regardless of the number of bins chosen.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Jan Kueckelhaus, Dieter-Henrik Heiland, Simon Frerich.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
