<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Copy Number Variations (CNV) • SPATA2</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Copy Number Variations (CNV)">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SPATA2</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">3.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/installation.html">Installation</a>
</li>
<li>
  <a href="../articles/spata-data.html">SPATAData</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">Initiating a SPATA2 object</li>
    <li>
      <a href="../articles/initiation-and-preprocessing-customized.html">Initiation &amp; Preprocessing with individual data</a>
    </li>
    <li>
      <a href="../articles/initiation-and-preprocessing-merfish.html">Initiation &amp; Preprocessing with MERFISH data</a>
    </li>
    <li>
      <a href="../articles/initiation-and-preprocessing-visium.html">Initiation &amp; Preprocessing with Visium data</a>
    </li>
    <li>
      <a href="../articles/initiation-and-preprocessing-visium-hd.html">Initiation &amp; Preprocessing with VisiumHD data</a>
    </li>
    <li class="dropdown-header">Basic Manipulation of a SPATA2 object</li>
    <li>
      <a href="../articles/adding-data.html">Adding Data to SPATA2 objects</a>
    </li>
    <li>
      <a href="../articles/package-compatibility.html">Platform Compatibility</a>
    </li>
    <li>
      <a href="../articles/molecular-variables.html">Molecular Variables</a>
    </li>
    <li>
      <a href="../articles/image-handling.html">Image Handling</a>
    </li>
    <li>
      <a href="../articles/spatial-segmentation.html">Spatial Segmentation</a>
    </li>
    <li class="dropdown-header">Implemented Algorithms &amp; Concepts</li>
    <li>
      <a href="../articles/cnv.html">Copy Number Variations</a>
    </li>
    <li>
      <a href="../articles/clustering.html">Clustering</a>
    </li>
    <li>
      <a href="../articles/dimensional-reduction.html">Dimensional Reduction</a>
    </li>
    <li>
      <a href="../articles/dea.html">Differential Expression</a>
    </li>
    <li>
      <a href="../articles/gsea.html">Gene Set Enrichment</a>
    </li>
    <li>
      <a href="../articles/spatial-measures.html">Spatial Measures</a>
    </li>
    <li class="dropdown-header">Spatial Gradient Screening</li>
    <li>
      <a href="../articles/creating-spatial-annotations.html">Creating Spatial Annotations</a>
    </li>
    <li>
      <a href="../articles/using-spatial-annotations.html">Using Spatial Annotations</a>
    </li>
    <li>
      <a href="../articles/spatial-annotation-screening.html">Spatial Annotation Screening</a>
    </li>
    <li>
      <a href="../articles/creating-spatial-trajectories.html">Creating Spatial Trajectories</a>
    </li>
    <li>
      <a href="../articles/spatial-trajectory-screening.html">Spatial Trajectory Screening</a>
    </li>
    <li class="dropdown-header">Visualization</li>
    <li>
      <a href="../articles/plotting-surface.html">Surface Plotting</a>
    </li>
  </ul>
</li>
<li>
  <a href="../articles/newslog.html">Newslog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Copy Number Variations (CNV)</h1>
            
      

      <div class="hidden name"><code>cnv.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="introduction">1. Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>InferCNV is used to identify large-scale chromosomal copy number
alterations in single cell RNA-Seq data including gains or deletions of
chromosomes or large segments of such. Several publications have
leveraged this technique. (<a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4123637/" class="external-link">Patel et
al., 2014</a>, <a href="https://pubmed.ncbi.nlm.nih.gov/27124452/" class="external-link">Tirosh et al.,
2016</a>, <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5519096/" class="external-link">Venteicher
et al., 2017</a> )</p>
<p>Note: We are currently encountering problems with the function
<code>infercnv:::plot_cnv()</code>. It does not prevent the copy number
variation analysis. Sometimes, however, the heatmap is not plotted.
Checkout your R console for hints in how far the function did not work
properly. We are working on a solution for that. If you encounter
warnings raised from <code>runCNV</code>()` <a href="https://github.com/theMILOlab/SPATA2/issues" class="external-link">let us know</a>.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r code-chunk"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="co"># load required packages</span></span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="fu">library</span>(SPATA2)</span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a><span class="fu">library</span>(SPATAData)</span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a></span>
<span id="cb1-6"><a href="#cb1-6" tabindex="-1"></a><span class="co"># load SPATA2 object</span></span>
<span id="cb1-7"><a href="#cb1-7" tabindex="-1"></a>object_t313 <span class="ot">&lt;-</span> <span class="fu">downloadSpataObject</span>(<span class="at">sample_name =</span> <span class="st">"UKF269T"</span>)</span>
<span id="cb1-8"><a href="#cb1-8" tabindex="-1"></a></span>
<span id="cb1-9"><a href="#cb1-9" tabindex="-1"></a><span class="co"># alternatively, use diet version (results might differ slightly)</span></span>
<span id="cb1-10"><a href="#cb1-10" tabindex="-1"></a>object_t313 <span class="ot">&lt;-</span> <span class="fu">loadExampleObject</span>(<span class="at">sample_name =</span> <span class="st">"UKF269T"</span>)</span>
<span id="cb1-11"><a href="#cb1-11" tabindex="-1"></a></span>
<span id="cb1-12"><a href="#cb1-12" tabindex="-1"></a><span class="co"># only histology</span></span>
<span id="cb1-13"><a href="#cb1-13" tabindex="-1"></a><span class="fu">plotSurface</span>(object, <span class="ot">=</span> object_t269, <span class="at">pt_alpha =</span> <span class="dv">0</span>)</span>
<span id="cb1-14"><a href="#cb1-14" tabindex="-1"></a></span>
<span id="cb1-15"><a href="#cb1-15" tabindex="-1"></a><span class="co"># histological grouping</span></span>
<span id="cb1-16"><a href="#cb1-16" tabindex="-1"></a><span class="fu">plotSurface</span>(<span class="at">object =</span> object_t269, <span class="at">color_by =</span> <span class="st">"histology"</span>)</span></code></pre></div>
<p><img src="cnv_files/figure-html/unnamed-chunk-2-1.png" width="50%"><img src="cnv_files/figure-html/unnamed-chunk-2-2.png" width="50%"></p>
<p>SPATA2 implements the package <a href="https://github.com/broadinstitute/inferCNV/wiki" class="external-link"><code>infercnv</code></a>
published and maintained by Broadinstitute and allows to integrate this
technique in your workflow of analyzing spatial trancsriptomic data
derived from malignancies.</p>
</div>
<div class="section level2">
<h2 id="running-cnv-analysis">2. Running CNV Analysis<a class="anchor" aria-label="anchor" href="#running-cnv-analysis"></a>
</h2>
<p>As with all other functions prefixed with <code>run*()</code> the
function <code><a href="../reference/runCNV.html">runCNV()</a></code> is a wrapper around all necessary
functions needed to conduct copy-number-variation-analysis. The results
needed for subsequent analysis steps are stored in the specified
<code>SPATA2</code> object (slot: <span class="citation">@cnv</span>).
The infercnv-object is stored in the folder specified in the argument
<code>directory_cnv_folder</code>.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">object_t269</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu"><a href="../reference/runCNV.html">runCNV</a></span><span class="op">(</span></span>
<span>    object <span class="op">=</span> <span class="va">object_t269</span>,</span>
<span>    <span class="co"># the directory must exist in your file system</span></span>
<span>    directory_cnv_folder <span class="op">=</span> <span class="st">"data-gbm269/cnv-results"</span>, </span>
<span>    cnv_prefix <span class="op">=</span> <span class="st">"Chr"</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
<p>If your desired set up deviates from the default you can reach any
function of the <code>inferncnv</code> pipeline by entering it’s name as
an argument and specify it’s input as a list of arguments with which you
want it to be called.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r code-chunk"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="co"># change input</span></span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a>object_t269 <span class="ot">&lt;-</span> </span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a>  <span class="fu">runCNV</span>(</span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a>    <span class="at">object =</span> object_t269</span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a>    <span class="at">directory_cnv_folder =</span> <span class="st">"data-gbm269/cnv-results"</span>, </span>
<span id="cb3-6"><a href="#cb3-6" tabindex="-1"></a>    <span class="at">clear_noise_via_ref_mean_sd =</span> <span class="fu">list</span>(<span class="at">sd_amplifier =</span> <span class="dv">2</span>)</span>
<span id="cb3-7"><a href="#cb3-7" tabindex="-1"></a>    )</span></code></pre></div>
<p>Copy number variation analysis requires reference data. This includes
a count matrix from healthy tissue, an annotation file as well as a
data.frame that contains information about the chromosome positions. We
provide reference data in the list <code><a href="../reference/cnv_ref.html">SPATA2::cnv_ref</a></code>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">cnv_ref</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "annotation" "mtr"        "regions"</span></span></code></pre>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">cnv_ref</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##            Length   Class      Mode</span></span>
<span><span class="co">## annotation        1 data.frame list</span></span>
<span><span class="co">## mtr        71067022 dgCMatrix  S4  </span></span>
<span><span class="co">## regions           4 data.frame list</span></span></code></pre>
<p><code><a href="../reference/runCNV.html">runCNV()</a></code> defaults to the content of this list. The
documentation of <code><a href="../reference/runCNV.html">runCNV()</a></code> contains a detailed description
of the requirements each reference input must meet in order for the
function to work.</p>
</div>
<div class="section level2">
<h2 id="cnv-results">3. CNV Results<a class="anchor" aria-label="anchor" href="#cnv-results"></a>
</h2>
<p>The results are stored in a list inside the <code>SPATA2</code>
object. This list can be obtained via <code><a href="../reference/getCnvResults.html">getCnvResults()</a></code>.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cnv_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getCnvResults.html">getCnvResults</a></span><span class="op">(</span><span class="va">object_t269</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">cnv_results</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "prefix"      "cnv_df"      "cnv_mtr"     "gene_pos_df" "regions_df"</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="visualization">4. Visualization<a class="anchor" aria-label="anchor" href="#visualization"></a>
</h2>
<p>Note, that the plots below might differ slightly from what you obtain
after copy and pasting this code since you are working with a reduced
version of the original UKF269T sample. To run this tutorial with the
complete data set download sample <em>UKF269T</em> with the <a href="spata-data.html">SPATAData</a> package.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># if you want the complete object</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SPATAData</span><span class="op">)</span></span>
<span></span>
<span><span class="va">object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/downloadSpataObject.html">downloadSpataObject</a></span><span class="op">(</span><span class="st">"UKF269T"</span><span class="op">)</span></span>
<span><span class="va">object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runCNV.html">runCNV</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="heatmap">4.1 Heatmap<a class="anchor" aria-label="anchor" href="#heatmap"></a>
</h3>
<p>The gains and losses of chromosomal segments can be displayed via
<code><a href="../reference/plotCnvHeatmap.html">plotCnvHeatmap()</a></code>. Without any meta data CNV heatmaps are
not that insightful. If you want to visualize the results across certain
groups to highlight differences across groups make use of the
<code>across</code> and <code>across_subset</code> arguments.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotCnvHeatmap.html">plotCnvHeatmap</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">object_t269</span>, across <span class="op">=</span> <span class="st">"histology"</span>, clrp <span class="op">=</span> <span class="st">"npg"</span><span class="op">)</span></span></code></pre></div>
<p><img src="cnv_files/figure-html/unnamed-chunk-8-1.png" width="100%"></p>
<p>You can add additional meta variables to the legend, numeric and
categorical ones alike. Furthermore, you can subset the chromosomes to
be displayed as well as the groups across which the results are shown.
CNV heatmaps in SPATA2 actually consist of multiple ggplots assembled
via the <code>aplot</code> package. You can adjust them by referring to
them by name within the <code>ggpLayer</code> argument. For a detailed
documentation of how to create the CNV heatmap you desire please refer
to the extensive documentation via <code><a href="../reference/plotCnvHeatmap.html">plotCnvHeatmap()</a></code>.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># a more complex set up</span></span>
<span><span class="fu"><a href="../reference/plotCnvHeatmap.html">plotCnvHeatmap</a></span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">object_t269</span>, </span>
<span>  across <span class="op">=</span> <span class="st">"histology"</span>, </span>
<span>  across_subset <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"infiltrated"</span>, <span class="st">"tumor"</span><span class="op">)</span>, <span class="co"># dont show the transition part</span></span>
<span>  meta_vars <span class="op">=</span> <span class="st">"SNAP25"</span>,  <span class="co"># visualize SNAP25 expression on the left</span></span>
<span>  meta_vars_clrs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"SNAP25"</span> <span class="op">=</span> <span class="st">"inferno"</span><span class="op">)</span>, <span class="co"># with the inferno color spectrum </span></span>
<span>  chrom_subset <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"1"</span>, <span class="st">"3"</span>, <span class="st">"7"</span>, <span class="st">"10"</span>, <span class="st">"14"</span>, <span class="st">"15"</span>, <span class="st">"18"</span>, <span class="st">"22"</span><span class="op">)</span>, <span class="co"># only show these chromosomes</span></span>
<span>  ggpLayers <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>arm <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="../reference/legendBottom.html">legendNone</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="co"># remove the chrom arm legend</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="co"># right plot</span></span>
<span><span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span><span class="va">object_t269</span>, color_by <span class="op">=</span> <span class="st">"SNAP25"</span>, pt_clrsp <span class="op">=</span> <span class="st">"inferno"</span><span class="op">)</span></span></code></pre></div>
<p><img src="cnv_files/figure-html/unnamed-chunk-9-1.png" width="50%"><img src="cnv_files/figure-html/unnamed-chunk-9-2.png" width="50%"></p>
</div>
<div class="section level3">
<h3 id="lineplot">4.2 Lineplot<a class="anchor" aria-label="anchor" href="#lineplot"></a>
</h3>
<p>An additional option to visualize the results provides
<code><a href="../reference/plotCnvLineplot.html">plotCnvLineplot()</a></code>.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotCnvLineplot.html">plotCnvLineplot</a></span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">object_t269</span>,</span>
<span>  across <span class="op">=</span> <span class="st">"histology"</span>, </span>
<span>  n_bins_bcsp <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>  n_bins_genes <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>  nrow <span class="op">=</span> <span class="fl">3</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="cnv_files/figure-html/unnamed-chunk-10-1.png" width="100%"></p>
</div>
<div class="section level3">
<h3 id="surface">4.3 Surface<a class="anchor" aria-label="anchor" href="#surface"></a>
</h3>
<p>The numeric values by which the copy number variations of each
chromosome are represented are immediately transferred to the
<code>SPATA2</code> object’s feature data and are thus accessible for
all functions that work with numeric variables. The character string
specified in argument <code>cnv_prefix</code> combined with the
chromosomes number determines the name by which you can refer to these
feature variables.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># cnv feature names</span></span>
<span><span class="fu"><a href="../reference/getCnvFeatureNames.html">getCnvFeatureNames</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">object_t269</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Chr1" "Chr2" "Chr3" "Chr4" "Chr5" "Chr6"</span></span></code></pre>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># are part of all feature names</span></span>
<span><span class="fu"><a href="../reference/getFeatureNames.html">getFeatureNames</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">object_t269</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##            factor           integer           numeric           numeric </span></span>
<span><span class="co">##      "orig.ident"    "nFeature_RNA"      "percent.mt"      "percent.RB" </span></span>
<span><span class="co">##            factor            factor </span></span>
<span><span class="co">## "RNA_snn_res.0.8"     "bayes_space"</span></span></code></pre>
<p>Use <code><a href="../reference/plotSurface.html">plotSurface()</a></code> to visualize chromosomal alterations
directly on the histology.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">object_t269</span>, </span>
<span>  color_by <span class="op">=</span> <span class="st">"Chr7"</span>, </span>
<span>  pt_clrsp <span class="op">=</span> <span class="st">"Reds"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">object_t269</span>, </span>
<span>  color_by <span class="op">=</span> <span class="st">"Chr10"</span>, </span>
<span>  pt_clrsp <span class="op">=</span> <span class="st">"Oslo"</span> </span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="cnv_files/figure-html/unnamed-chunk-11-1.png" width="50%"><img src="cnv_files/figure-html/unnamed-chunk-11-2.png" width="50%"></p>
</div>
<div class="section level3">
<h3 id="gradients">4.4 Gradients<a class="anchor" aria-label="anchor" href="#gradients"></a>
</h3>
<p>As showcased in our corresponding vignette about <a href="spata-v2-spatial-trajectories.html">spatial trajectories</a>
gradients of numeric variables can be displayed. Chromosomal alterations
are numeric variables.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSpatialTrajectories.html">plotSpatialTrajectories</a></span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">object_t269</span>, </span>
<span>  ids <span class="op">=</span> <span class="st">"horizontal_mid"</span>, </span>
<span>  color_by <span class="op">=</span> <span class="st">"Chr7"</span>, </span>
<span>  pt_clrsp <span class="op">=</span> <span class="st">"Reds 3"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plotSpatialTrajectories.html">plotSpatialTrajectories</a></span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">object_t269</span>, </span>
<span>  ids <span class="op">=</span> <span class="st">"horizontal_mid"</span>, </span>
<span>  color_by <span class="op">=</span> <span class="st">"Chr10"</span>, </span>
<span>  pt_clrsp <span class="op">=</span> <span class="st">"Oslo"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plotStsLineplot.html">plotStsLineplot</a></span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">object_t269</span>, </span>
<span>  id <span class="op">=</span> <span class="st">"horizontal_mid"</span>, </span>
<span>  variables <span class="op">=</span> <span class="st">"Chr7"</span> </span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plotStsLineplot.html">plotStsLineplot</a></span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">object_t269</span>, </span>
<span>  id <span class="op">=</span> <span class="st">"horizontal_mid"</span>, </span>
<span>  variables <span class="op">=</span> <span class="st">"Chr10"</span>, </span>
<span>  line_color <span class="op">=</span> <span class="st">"blue"</span> </span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="cnv_files/figure-html/unnamed-chunk-12-1.png" width="50%"><img src="cnv_files/figure-html/unnamed-chunk-12-2.png" width="50%"><img src="cnv_files/figure-html/unnamed-chunk-12-3.png" width="50%"><img src="cnv_files/figure-html/unnamed-chunk-12-4.png" width="50%">
(Note that trajectory lineplots always rescale variables to 0-1 or to
low-high.)</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Jan Kueckelhaus, Dieter-Henrik Heiland.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

      </footer>
</div>






  </body>
</html>
