<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Initiation of individual SPATA2 objects • SPATA2</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Initiation of individual SPATA2 objects">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SPATA2</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">3.1.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/installation.html">Installation</a>
</li>
<li>
  <a href="../articles/spata-data.html">SPATAData</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">Initiating a SPATA2 object</li>
    <li>
      <a href="../articles/initiation-and-preprocessing-customized.html">Initiation &amp; Preprocessing with individual data</a>
    </li>
    <li>
      <a href="../articles/initiation-and-preprocessing-merfish.html">Initiation &amp; Preprocessing with MERFISH data</a>
    </li>
    <li>
      <a href="../articles/initiation-and-preprocessing-visium.html">Initiation &amp; Preprocessing with Visium data</a>
    </li>
    <li>
      <a href="../articles/initiation-and-preprocessing-visium-hd.html">Initiation &amp; Preprocessing with VisiumHD data</a>
    </li>
    <li class="dropdown-header">Basic Manipulation of a SPATA2 object</li>
    <li>
      <a href="../articles/adding-data.html">Adding Data to SPATA2 objects</a>
    </li>
    <li>
      <a href="../articles/package-compatibility.html">Platform Compatibility</a>
    </li>
    <li>
      <a href="../articles/molecular-variables.html">Molecular Variables</a>
    </li>
    <li>
      <a href="../articles/image-handling.html">Image Handling</a>
    </li>
    <li>
      <a href="../articles/spatial-segmentation.html">Spatial Segmentation</a>
    </li>
    <li class="dropdown-header">Implemented Algorithms &amp; Concepts</li>
    <li>
      <a href="../articles/cnv.html">Copy Number Variations</a>
    </li>
    <li>
      <a href="../articles/clustering.html">Clustering</a>
    </li>
    <li>
      <a href="../articles/dimensional-reduction.html">Dimensional Reduction</a>
    </li>
    <li>
      <a href="../articles/dea.html">Differential Expression</a>
    </li>
    <li>
      <a href="../articles/gsea.html">Gene Set Enrichment</a>
    </li>
    <li>
      <a href="../articles/spatial-measures.html">Spatial Measures</a>
    </li>
    <li class="dropdown-header">Spatial Gradient Screening</li>
    <li>
      <a href="../articles/creating-spatial-annotations.html">Creating Spatial Annotations</a>
    </li>
    <li>
      <a href="../articles/using-spatial-annotations.html">Using Spatial Annotations</a>
    </li>
    <li>
      <a href="../articles/spatial-annotation-screening.html">Spatial Annotation Screening</a>
    </li>
    <li>
      <a href="../articles/creating-spatial-trajectories.html">Creating Spatial Trajectories</a>
    </li>
    <li>
      <a href="../articles/spatial-trajectory-screening.html">Spatial Trajectory Screening</a>
    </li>
    <li class="dropdown-header">Visualization</li>
    <li>
      <a href="../articles/plotting-surface.html">Surface Plotting</a>
    </li>
  </ul>
</li>
<li>
  <a href="../articles/newslog.html">Newslog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Initiation of individual SPATA2 objects</h1>
            
      

      <div class="hidden name"><code>initiation-and-preprocessing-customized.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="introduction">1. Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>While SPATA2 provides functions to initiate <code>SPATA2</code>
objects with standardized output from common platforms like <a href="initiation-and-preprocessing-visium.html">Visium</a> or <a href="initiation-and-preprocessing-merfish.html">MERFISH</a>, you can
initiate a SPATA2 object from scratch with individual data as long as
your data meets the following criteria:</p>
<ol style="list-style-type: decimal">
<li>There must be a clearly identified and quantifiable unit of data
like gene- or protein expression, metabolite counts.</li>
<li>There must be a clearly identified observational unit to which the
data is mapped like barcoded spots, single cells or beads.</li>
<li>There must be numeric information about the position of each
observation in 2D space in form of x- and y- coordinates.</li>
</ol>
<p>Everything else is optional. This tutorial guides you through the
required steps and best practices.</p>
</div>
<div class="section level2">
<h2 id="initiate-the-object">2. Initiate the object<a class="anchor" aria-label="anchor" href="#initiate-the-object"></a>
</h2>
<p>The function to use is <code><a href="../reference/initiateSpataObject.html">initiateSpataObject()</a></code>. Minimal
requirements to actually create a <code>SPATA2</code> object
include:</p>
<ol style="list-style-type: lower-alpha">
<li>Input for <code>sample_name</code>: A character value which names
your object.</li>
<li>Input for <code>modality</code>: A character value which best
describes the molecular modality you have data for. Read more <a href="molecular-variables.html">here</a>.</li>
<li>Input for <code>count_mtr</code>: A numeric m x n matrix with
rownames corresponding to molecule names and column names corresponding
to the identifiers of your observations, in SPATA2 terms: the
barcodes.</li>
<li>Input for <code>coords_df</code>: A data.frame with at least three
variables. <em>barcodes</em>, <em>x</em> or <em>x_orig</em> and
<em>y</em> or <em>y_orig</em>.</li>
</ol>
<div class="section level3">
<h3 id="with-minimal-requirements">2.1 With minimal requirements<a class="anchor" aria-label="anchor" href="#with-minimal-requirements"></a>
</h3>
<p>This code chunks creates a <code>SPATA2</code> object from scratch
with minimal requirements.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## use SPATA2 intern example data</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"example_data"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">count_mtr</span> <span class="op">&lt;-</span> <span class="va">example_data</span><span class="op">$</span><span class="va">count_mtr</span></span>
<span><span class="va">coords_df</span> <span class="op">&lt;-</span> <span class="va">example_data</span><span class="op">$</span><span class="va">coords_df</span></span>
<span></span>
<span><span class="co"># show count mtr structure</span></span>
<span><span class="va">count_mtr</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## 5 x 5 sparse Matrix of class "dgCMatrix"</span></span>
<span><span class="co">##       AAACAAGTATCTCCCA-1 AAACACCAATAACTGC-1 AAACAGAGCGACTCCT-1</span></span>
<span><span class="co">## MT1X                   1                  2                  3</span></span>
<span><span class="co">## ADM                    .                  1                  1</span></span>
<span><span class="co">## IGKC                   .                  .                  .</span></span>
<span><span class="co">## VEGFA                  1                  .                  .</span></span>
<span><span class="co">## DDIT3                  1                  1                  1</span></span>
<span><span class="co">##       AAACAGCTTTCAGAAG-1 AAACAGGGTCTATATT-1</span></span>
<span><span class="co">## MT1X                   2                  3</span></span>
<span><span class="co">## ADM                    .                  .</span></span>
<span><span class="co">## IGKC                   .                  .</span></span>
<span><span class="co">## VEGFA                  .                  1</span></span>
<span><span class="co">## DDIT3                  1                  .</span></span></code></pre>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># show coords data.frame structure</span></span>
<span><span class="va">coords_df</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 3,733 × 3</span></span></span>
<span><span class="co">##    barcodes           x_orig y_orig</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>               <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> AAACAAGTATCTCCCA-1   435.  226. </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> AAACACCAATAACTGC-1   132.  171. </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> AAACAGAGCGACTCCT-1   407.  455. </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> AAACAGCTTTCAGAAG-1    96   273  </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> AAACAGGGTCTATATT-1   110.  247. </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> AAACAGTGTTCCTGGG-1   219.   81.8</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> AAACATTTCCCGGATT-1   416.  157. </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> AAACCGGGTAGGTACC-1   165.  279. </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> AAACCGTTCGTCCAGG-1   216.  215. </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> AAACCTAAGCAGCCGG-1   365.  132. </span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 3,723 more rows</span></span></span></code></pre>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">object_minimal</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="../reference/initiateSpataObject.html">initiateSpataObject</a></span><span class="op">(</span></span>
<span>    sample_name <span class="op">=</span> <span class="st">"my_object"</span>, <span class="co"># req. a</span></span>
<span>    modality <span class="op">=</span> <span class="st">"gene"</span>, <span class="co"># req. b</span></span>
<span>    count_mtr <span class="op">=</span> <span class="va">count_mtr</span>, <span class="co"># req. c</span></span>
<span>    coords_df <span class="op">=</span> <span class="va">coords_df</span> <span class="co"># req. d</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="co"># left plot</span></span>
<span><span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span><span class="va">object_minimal</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># right plot</span></span>
<span><span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span><span class="va">object_minimal</span>, color_by <span class="op">=</span> <span class="st">"METRN"</span><span class="op">)</span></span></code></pre></div>
<p><img src="initiation-and-preprocessing-customized_files/figure-html/unnamed-chunk-2-1.png" width="50%"><img src="initiation-and-preprocessing-customized_files/figure-html/unnamed-chunk-2-2.png" width="50%"></p>
</div>
<div class="section level3">
<h3 id="including-images">2.2 Including images<a class="anchor" aria-label="anchor" href="#including-images"></a>
</h3>
<p>You can either register images after creating the object or do it
during the initiation. Whether you have it in your global environment or
whether you read it from a directory, the image should be compatible
with the <a href="https://bioconductor.org/packages/devel/bioc/vignettes/EBImage/inst/doc/EBImage-introduction.html" class="external-link">EBImage</a>
package - which is the case for almost all images. If you use
<code>img_dir</code> the directory of the image is stored, too, so you
can conveniently switch between images if you have registered
multiple.</p>
<div class="section level4">
<h4 id="aligned-images">2.2.1 Aligned images<a class="anchor" aria-label="anchor" href="#aligned-images"></a>
</h4>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># get an image from the example data (or read in your own)</span></span>
<span><span class="va">img_normres</span> <span class="op">&lt;-</span> <span class="va">example_data</span><span class="op">$</span><span class="va">img_normres</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">img_normres</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 576 600   3</span></span></code></pre>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># initiate the object with an image</span></span>
<span><span class="va">object_init_with_img</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="../reference/initiateSpataObject.html">initiateSpataObject</a></span><span class="op">(</span></span>
<span>    sample_name <span class="op">=</span> <span class="st">"my_object"</span>, </span>
<span>    modality <span class="op">=</span> <span class="st">"gene"</span>, </span>
<span>    count_mtr <span class="op">=</span> <span class="va">count_mtr</span>, </span>
<span>    coords_df <span class="op">=</span> <span class="va">coords_df</span>, </span>
<span>    img <span class="op">=</span> <span class="va">img_normres</span>, <span class="co"># provide the image</span></span>
<span>    img_name <span class="op">=</span> <span class="st">"norm_res"</span> <span class="co"># name the image</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="co"># register the image afterwards</span></span>
<span><span class="va">object_minimal</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="../reference/registerImage.html">registerImage</a></span><span class="op">(</span></span>
<span>    object <span class="op">=</span> <span class="va">object_minimal</span>, <span class="co"># object from previous code chunk</span></span>
<span>    img <span class="op">=</span> <span class="va">img_normres</span>, </span>
<span>    img_name <span class="op">=</span> <span class="st">"norm_res"</span>,</span>
<span>    unload <span class="op">=</span> <span class="cn">FALSE</span></span>
<span>    <span class="op">)</span></span>
<span></span>
<span><span class="co"># both result in the same output</span></span>
<span></span>
<span><span class="co"># left plot</span></span>
<span><span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span><span class="va">object_init_with_img</span>, display_image <span class="op">=</span> <span class="cn">T</span><span class="op">)</span> </span>
<span></span>
<span><span class="co"># right plot</span></span>
<span><span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span><span class="va">object_minimal</span>, display_image <span class="op">=</span> <span class="cn">T</span><span class="op">)</span> </span></code></pre></div>
<p><img src="initiation-and-preprocessing-customized_files/figure-html/unnamed-chunk-3-1.png" width="50%"><img src="initiation-and-preprocessing-customized_files/figure-html/unnamed-chunk-3-2.png" width="50%"></p>
</div>
<div class="section level4">
<h4 id="not-aligned-images">2.2.2 Not-aligned images<a class="anchor" aria-label="anchor" href="#not-aligned-images"></a>
</h4>
<p>What if your image is not aligned with the coordinates of your
observations which are stored in the <code>coords_df</code>? There are
two kinds of problems that could arise.</p>
</div>
<div class="section level4">
<h4 id="bad-scaling">Bad Scaling<a class="anchor" aria-label="anchor" href="#bad-scaling"></a>
</h4>
<p>One problem could be that the x- and y-coordinates of
<code>coords_df</code> are scaled to a different resolution than the
image. In the example below, the <code>SPATA2</code> object is initiated
with an image that is smaller in dimensions compared to the scaling of
the coordinates of the observations.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># an image</span></span>
<span><span class="va">img_lowres</span> <span class="op">&lt;-</span> <span class="va">example_data</span><span class="op">$</span><span class="va">img_lowres</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">img_lowres</span><span class="op">)</span></span>
<span><span class="co">## [1] "Image"</span></span>
<span><span class="co">## attr(,"package")</span></span>
<span><span class="co">## [1] "EBImage"</span></span>
<span></span>
<span><span class="co"># dimensions lowres image</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">img_lowres</span><span class="op">)</span></span>
<span><span class="co">## [1] 200 208   3</span></span>
<span></span>
<span><span class="co"># dimensions coordinates</span></span>
<span><span class="va">coords_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getCoordsDf.html">getCoordsDf</a></span><span class="op">(</span><span class="va">object_minimal</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># maxima of x- and y-coordinates surpass their respective image axes</span></span>
<span><span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html" class="external-link">map</a></span><span class="op">(</span>.x <span class="op">=</span> <span class="va">coords_df</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"x"</span>, <span class="st">"y"</span><span class="op">)</span><span class="op">]</span>, .f <span class="op">=</span> <span class="va">range</span><span class="op">)</span></span>
<span><span class="co">## $x</span></span>
<span><span class="co">## [1]  70.32 485.76</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $y</span></span>
<span><span class="co">## [1]  74.16 513.24</span></span>
<span></span>
<span><span class="va">object_badly_scaled</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="../reference/initiateSpataObject.html">initiateSpataObject</a></span><span class="op">(</span></span>
<span>    sample_name <span class="op">=</span> <span class="st">"my_object"</span>, </span>
<span>    modality <span class="op">=</span> <span class="st">"gene"</span>, </span>
<span>    count_mtr <span class="op">=</span> <span class="va">count_mtr</span>, </span>
<span>    coords_df <span class="op">=</span> <span class="va">coords_df</span>, </span>
<span>    img <span class="op">=</span> <span class="va">img_lowres</span>, </span>
<span>    img_name <span class="op">=</span> <span class="st">"lowres"</span> </span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="co"># the plotted observations -&gt; bad alignment</span></span>
<span><span class="co"># left plot (axes range is scaled to observation range by default)</span></span>
<span><span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span><span class="va">object_badly_scaled</span>, pt_clr <span class="op">=</span> <span class="st">"red"</span>, display_image <span class="op">=</span> <span class="cn">T</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># right plot with increased axes scale </span></span>
<span><span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span><span class="va">object_badly_scaled</span>, pt_clr <span class="op">=</span> <span class="st">"red"</span>, display_image <span class="op">=</span> <span class="cn">T</span>, xrange <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">600</span><span class="op">)</span>, yrange <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">600</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="initiation-and-preprocessing-customized_files/figure-html/unnamed-chunk-4-1.png" width="50%"><img src="initiation-and-preprocessing-customized_files/figure-html/unnamed-chunk-4-2.png" width="50%"></p>
<p>In this case a scale factor is needed.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># compute an image scale factor (or obtain it otherwise)</span></span>
<span><span class="va">isf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">img_lowres</span><span class="op">)</span><span class="op">)</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">img_normres</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">isf</span><span class="op">)</span></span>
<span><span class="co">## [1] 0.3466667</span></span>
<span></span>
<span><span class="va">object_well_scaled</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="../reference/initiateSpataObject.html">initiateSpataObject</a></span><span class="op">(</span></span>
<span>    sample_name <span class="op">=</span> <span class="st">"my_object"</span>, </span>
<span>    modality <span class="op">=</span> <span class="st">"gene"</span>, </span>
<span>    count_mtr <span class="op">=</span> <span class="va">count_mtr</span>, </span>
<span>    coords_df <span class="op">=</span> <span class="va">coords_df</span>, </span>
<span>    img <span class="op">=</span> <span class="va">img_lowres</span>, </span>
<span>    img_name <span class="op">=</span> <span class="st">"lowres"</span>, </span>
<span>    scale_factors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>image <span class="op">=</span> <span class="va">isf</span><span class="op">)</span> <span class="co"># provide the scale factor!</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="co"># the image scale factor is used to scale the coordinates to the image that</span></span>
<span><span class="co"># is currently active, which is the image 'lowres' since object_well_scaled contains</span></span>
<span><span class="co"># only that</span></span>
<span><span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span><span class="va">object_well_scaled</span>, pt_clr <span class="op">=</span> <span class="st">"red"</span>, display_image <span class="op">=</span> <span class="cn">T</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="initiation-and-preprocessing-customized_files/figure-html/unnamed-chunk-5-1.png" width="50%"></p>
</div>
<div class="section level4">
<h4 id="bad-justification">Bad justification<a class="anchor" aria-label="anchor" href="#bad-justification"></a>
</h4>
<p>If the image needs additional justification in terms horizontal or
vertical translation, rotation or stretching, please refer to the
vignette on <a href="image-handling.html">image handling</a>.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># for some image adjustments</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/aoles/EBImage" class="external-link">EBImage</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># mess up the image justification</span></span>
<span><span class="va">img_bad_just</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/EBImage/man/spatial.html" class="external-link">flip</a></span><span class="op">(</span><span class="va">example_data</span><span class="op">$</span><span class="va">img_normres</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/EBImage/man/spatial.html" class="external-link">rotate</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">47</span>, bg.col <span class="op">=</span> <span class="st">"white"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/EBImage/man/spatial.html" class="external-link">translate</a></span><span class="op">(</span>v <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">20</span>, <span class="fl">55</span><span class="op">)</span>, bg.col <span class="op">=</span> <span class="st">"white"</span><span class="op">)</span></span>
<span>  </span>
<span><span class="co"># initiate the SPATA2 object with it</span></span>
<span><span class="va">object_bad_just</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="../reference/initiateSpataObject.html">initiateSpataObject</a></span><span class="op">(</span></span>
<span>    sample_name <span class="op">=</span> <span class="st">"my_object"</span>, </span>
<span>    modality <span class="op">=</span> <span class="st">"gene"</span>, </span>
<span>    count_mtr <span class="op">=</span> <span class="va">count_mtr</span>, </span>
<span>    coords_df <span class="op">=</span> <span class="va">coords_df</span>, </span>
<span>    img <span class="op">=</span> <span class="va">img_bad_just</span>, </span>
<span>    img_name <span class="op">=</span> <span class="st">"bad_just"</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span><span class="va">object_bad_just</span>, pt_clr <span class="op">=</span> <span class="st">"red"</span>, display_image <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span></code></pre></div>
<p><img src="initiation-and-preprocessing-customized_files/figure-html/unnamed-chunk-6-1.png" width="50%"></p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="si-units-and-the-pixel-scale-factor">3. SI units and the pixel scale factor<a class="anchor" aria-label="anchor" href="#si-units-and-the-pixel-scale-factor"></a>
</h2>
<p>In SPATA2, the pixel scale factor is used to transform the
coordinates from pixel (or loose numeric values) to SI units. It is a
numeric value that comes with an attribute which indicates the SI unit
to which the pixel- or the loose numeric value - is scaled (e.g. mm /
px). It is applied after the coordinates have been scaled to the image
resolution.</p>
<p>Note, that the name <em>pixel scale factor</em> has evolved
historically, since SPATA2 was developed with the Visium platform in
mind. The term does not fit perfectly for platforms such as MERFISH or
Xenium experiments since they do not provide an image and the
coordinates are provided in SI units (despite being dealt with as simple
numeric values in R). A more fitting name would be
<em>loose-numeric-value-to-SI-factor</em> - that’s quite long, though.
Furthermore, the SI unit system of SPATA2 works stable and we don’t want
to touch it. Therefore, the name pixel scale factor remains.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># we computed that beforehand...</span></span>
<span><span class="va">psf</span> <span class="op">&lt;-</span> <span class="fl">0.01368328</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="va">psf</span>, which <span class="op">=</span> <span class="st">"unit"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"mm/px"</span></span>
<span></span>
<span><span class="kw">if</span><span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span><span class="op">{</span></span>
<span>  </span>
<span>  <span class="co"># this is the scale factor MERFISH and Xenium objects are inititated with</span></span>
<span>  <span class="co"># since the coordinates already come in um units</span></span>
<span>  <span class="va">psf</span> <span class="op">&lt;-</span> <span class="fl">1</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="va">psf</span>, which <span class="op">=</span> <span class="st">"unit"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"um/px"</span></span>
<span>  </span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># usage of the pixel scale factor translated: </span></span>
<span><span class="co"># multiply the x- and y-coordinates with this factor to obtain </span></span>
<span><span class="co"># x- and y-coordinates in mm scale</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">psf</span><span class="op">)</span></span>
<span><span class="co">## [1] 0.01368328</span></span>
<span><span class="co">## attr(,"unit")</span></span>
<span><span class="co">## [1] "mm/px"</span></span>
<span></span>
<span><span class="va">object_advanced</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="../reference/initiateSpataObject.html">initiateSpataObject</a></span><span class="op">(</span></span>
<span>    sample_name <span class="op">=</span> <span class="st">"my_object"</span>, </span>
<span>    modality <span class="op">=</span> <span class="st">"gene"</span>, </span>
<span>    count_mtr <span class="op">=</span> <span class="va">count_mtr</span>, </span>
<span>    coords_df <span class="op">=</span> <span class="va">coords_df</span>, </span>
<span>    img <span class="op">=</span> <span class="va">img_normres</span>, <span class="co"># provide the image</span></span>
<span>    img_name <span class="op">=</span> <span class="st">"norm_res"</span>, <span class="co"># name the image</span></span>
<span>    scale_factors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>pixel <span class="op">=</span> <span class="va">psf</span><span class="op">)</span> <span class="co"># the pixel scale factor for image norm_res</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># object_advanced has been created with a pixel scale factor</span></span>
<span><span class="fu"><a href="../reference/containsScaleFactor.html">containsScaleFactor</a></span><span class="op">(</span><span class="va">object_advanced</span>, fct_name <span class="op">=</span> <span class="st">"pixel"</span><span class="op">)</span></span>
<span><span class="co">## [1] TRUE</span></span>
<span></span>
<span><span class="co"># object_minimal has not (!) been created with a pixel scale factor</span></span>
<span><span class="fu"><a href="../reference/containsScaleFactor.html">containsScaleFactor</a></span><span class="op">(</span><span class="va">object_minimal</span>, fct_name <span class="op">=</span> <span class="st">"pixel"</span><span class="op">)</span></span>
<span><span class="co">## [1] FALSE</span></span>
<span></span>
<span><span class="co"># note how the factor works for all kinds of  units</span></span>
<span><span class="fu"><a href="../reference/getPixelScaleFactor.html">getPixelScaleFactor</a></span><span class="op">(</span><span class="va">object_advanced</span>, unit <span class="op">=</span> <span class="st">"um"</span><span class="op">)</span></span>
<span><span class="co">## [1] 13.68328</span></span>
<span><span class="co">## attr(,"unit")</span></span>
<span><span class="co">## [1] "um/px"</span></span></code></pre></div>
<p>The pixel scale factor must not necessarily be provided. The function
<code><a href="../reference/computePixelScaleFactor.html">computePixelScaleFactor()</a></code> can compute it if a certain
criteria is met. Namely that there is a fixed center-to-center distance
inherent to the method or platform that underlies the data set, as is
the case for the Visium platform. For that, we need the
<code>SpatialMethod</code> class.</p>
</div>
<div class="section level2">
<h2 id="the-spatial-method">4. The spatial method<a class="anchor" aria-label="anchor" href="#the-spatial-method"></a>
</h2>
<p>So far we have created <code>SPATA2</code> objects without giving any
thought to the spatial method underlying the data. However, the spatial
method or platform based on which the data was created often contains
specifics that are important and useful for some functions. SPATA2
stores information around that in an S4 object of class
<code>SpatialMethod</code>. In case of
<code><a href="../reference/initiateSpataObject.html">initiateSpataObject()</a></code> which does not read data from a
predefined platform but with individual data the argument
<code>spatial_method</code> defaults to <em>‘Undefined’</em>. The
<code>SpatialMethod</code> class provides information about the
experiment set up and tells certain functions whether it is valid to use
them with your data set or not.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># show overview</span></span>
<span><span class="va">object_minimal</span></span></code></pre></div>
<pre><code><span><span class="co">## An object of class SPATA2</span></span>
<span><span class="co">## Sample: my_object </span></span>
<span><span class="co">## Size: 3733 x 2000 (undefined observations x molecules)</span></span>
<span><span class="co">## Memory: 84.15 Mb</span></span>
<span><span class="co">## Platform: Undefined </span></span>
<span><span class="co">## Molecular assays (1):</span></span>
<span><span class="co">## 1. Assay</span></span>
<span><span class="co">##  Molecular modality: gene</span></span>
<span><span class="co">##  Distinct molecules: 2000</span></span>
<span><span class="co">##  Matrices (1):</span></span>
<span><span class="co">##  -counts (active)</span></span>
<span><span class="co">## Registered images (1):</span></span>
<span><span class="co">## - norm_res (576x600 px, active, loaded)</span></span>
<span><span class="co">## Meta variables (1): sample</span></span></code></pre>
<div class="section level3">
<h3 id="the-concept">4.1 The concept<a class="anchor" aria-label="anchor" href="#the-concept"></a>
</h3>
<p>You can read more about this class with <code>SpatialMethod</code>.
Every platform known to SPATA2 has such a <code>SpatialMethod</code>
object defined. Consider the object for the Visium platform with the
6.5mm x 6.5mm capture area, in SPATA2 referred to as
<code>VisiumSmall</code>.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># show the class</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">VisiumSmall</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "SpatialMethod"</span></span>
<span><span class="co">## attr(,"package")</span></span>
<span><span class="co">## [1] "SPATA2"</span></span></code></pre>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># show the slot names </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/methods/slot.html" class="external-link">slotNames</a></span><span class="op">(</span><span class="va">VisiumSmall</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "info"               "method_specifics"   "name"              </span></span>
<span><span class="co">## [4] "unit"               "observational_unit" "version"</span></span></code></pre>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># show the content of each slot as a list </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/methods/show.html" class="external-link">show</a></span><span class="op">(</span><span class="va">VisiumSmall</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## An object of class 'SpatialMethod'.</span></span></code></pre>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># contrast this with the spatial method with which </span></span>
<span><span class="co"># undefined SPATA2 objects are created by default</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/methods/show.html" class="external-link">show</a></span><span class="op">(</span><span class="va">spatial_methods</span><span class="op">$</span><span class="va">Undefined</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## An object of class 'SpatialMethod'.</span></span></code></pre>
<p>Depending on the content of this object, in particular the slot <span class="citation">@method_specifics</span>, certain functions in SPATA2
work with you data and certain functions do not. For instance, if
information about the center-to-center distance in slot
<code>@method_specifics$ccd</code> is missing the function
<code><a href="../reference/computePixelScaleFactor.html">computePixelScaleFactor()</a></code> does not work, since it needs the
center-to-center distance.</p>
</div>
<div class="section level3">
<h3 id="create-it-yourself">4.2 Create it yourself<a class="anchor" aria-label="anchor" href="#create-it-yourself"></a>
</h3>
<p>To create a valid <code>SpatialMethod</code> object, you need only
three aspects: the observational unit, the name of the method and the
default unit. Everything else is optional.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sp_method</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="../reference/createSpatialMethod.html">createSpatialMethod</a></span><span class="op">(</span></span>
<span>    observational_unit <span class="op">=</span> <span class="st">"spot"</span>, </span>
<span>    name <span class="op">=</span> <span class="st">"MyPlatform"</span>,</span>
<span>    unit <span class="op">=</span> <span class="st">"mm"</span>, </span>
<span>    method_specifics <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"ccd"</span> <span class="op">=</span> <span class="st">"100um"</span><span class="op">)</span> <span class="co"># optional  </span></span>
<span>    <span class="op">)</span></span>
<span></span>
<span><span class="co"># create a SPATA2 object with a defined spatial method</span></span>
<span><span class="va">object_advanced</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="../reference/initiateSpataObject.html">initiateSpataObject</a></span><span class="op">(</span></span>
<span>    sample_name <span class="op">=</span> <span class="st">"my_object"</span>, </span>
<span>    modality <span class="op">=</span> <span class="st">"gene"</span>, </span>
<span>    count_mtr <span class="op">=</span> <span class="va">count_mtr</span>, </span>
<span>    coords_df <span class="op">=</span> <span class="va">coords_df</span>, </span>
<span>    img <span class="op">=</span> <span class="va">img_normres</span>, </span>
<span>    img_name <span class="op">=</span> <span class="st">"norm_res"</span>, </span>
<span>    spatial_method <span class="op">=</span> <span class="va">sp_method</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="co"># the platform appears in the overview</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/methods/show.html" class="external-link">show</a></span><span class="op">(</span><span class="va">object_advanced</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## An object of class SPATA2</span></span>
<span><span class="co">## Sample: my_object </span></span>
<span><span class="co">## Size: 3733 x 2000 (spots x molecules)</span></span>
<span><span class="co">## Memory: 84.21 Mb</span></span>
<span><span class="co">## Platform: MyPlatform </span></span>
<span><span class="co">## Molecular assays (1):</span></span>
<span><span class="co">## 1. Assay</span></span>
<span><span class="co">##  Molecular modality: gene</span></span>
<span><span class="co">##  Distinct molecules: 2000</span></span>
<span><span class="co">##  Matrices (1):</span></span>
<span><span class="co">##  -counts (active)</span></span>
<span><span class="co">## Registered images (1):</span></span>
<span><span class="co">## - norm_res (576x600 px, active, loaded)</span></span>
<span><span class="co">## Meta variables (1): sample</span></span></code></pre>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># the SPATA2 object contains information about the spatial method underlying its data</span></span>
<span><span class="co"># which qualifies it for automatic computation of the pixel scale factor</span></span>
<span><span class="fu"><a href="../reference/containsCCD.html">containsCCD</a></span><span class="op">(</span><span class="va">object_advanced</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] TRUE</span></span></code></pre>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">object_advanced</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/computePixelScaleFactor.html">computePixelScaleFactor</a></span><span class="op">(</span><span class="va">object_advanced</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># looks familiar?</span></span>
<span><span class="fu"><a href="../reference/getPixelScaleFactor.html">getPixelScaleFactor</a></span><span class="op">(</span><span class="va">object_advanced</span>, unit <span class="op">=</span> <span class="st">"mm"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 0.01368328</span></span>
<span><span class="co">## attr(,"unit")</span></span>
<span><span class="co">## [1] "mm/px"</span></span></code></pre>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Jan Kueckelhaus, Dieter-Henrik Heiland.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

      </footer>
</div>






  </body>
</html>
