<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Initiation &amp; Preprocessing: Visium • SPATA2</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Initiation &amp; Preprocessing: Visium">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SPATA2</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">3.1.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/installation.html">Installation</a>
</li>
<li>
  <a href="../articles/spata-data.html">SPATAData</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">Initiating a SPATA2 object</li>
    <li>
      <a href="../articles/initiation-and-preprocessing-customized.html">Initiation &amp; Preprocessing with individual data</a>
    </li>
    <li>
      <a href="../articles/initiation-and-preprocessing-merfish.html">Initiation &amp; Preprocessing with MERFISH data</a>
    </li>
    <li>
      <a href="../articles/initiation-and-preprocessing-visium.html">Initiation &amp; Preprocessing with Visium data</a>
    </li>
    <li>
      <a href="../articles/initiation-and-preprocessing-visium-hd.html">Initiation &amp; Preprocessing with VisiumHD data</a>
    </li>
    <li class="dropdown-header">Basic Manipulation of a SPATA2 object</li>
    <li>
      <a href="../articles/adding-data.html">Adding Data to SPATA2 objects</a>
    </li>
    <li>
      <a href="../articles/package-compatibility.html">Platform Compatibility</a>
    </li>
    <li>
      <a href="../articles/molecular-variables.html">Molecular Variables</a>
    </li>
    <li>
      <a href="../articles/image-handling.html">Image Handling</a>
    </li>
    <li>
      <a href="../articles/spatial-segmentation.html">Spatial Segmentation</a>
    </li>
    <li class="dropdown-header">Implemented Algorithms &amp; Concepts</li>
    <li>
      <a href="../articles/cnv.html">Copy Number Variations</a>
    </li>
    <li>
      <a href="../articles/clustering.html">Clustering</a>
    </li>
    <li>
      <a href="../articles/dimensional-reduction.html">Dimensional Reduction</a>
    </li>
    <li>
      <a href="../articles/dea.html">Differential Expression</a>
    </li>
    <li>
      <a href="../articles/gsea.html">Gene Set Enrichment</a>
    </li>
    <li>
      <a href="../articles/spatial-measures.html">Spatial Measures</a>
    </li>
    <li class="dropdown-header">Spatial Gradient Screening</li>
    <li>
      <a href="../articles/creating-spatial-annotations.html">Creating Spatial Annotations</a>
    </li>
    <li>
      <a href="../articles/using-spatial-annotations.html">Using Spatial Annotations</a>
    </li>
    <li>
      <a href="../articles/spatial-annotation-screening.html">Spatial Annotation Screening</a>
    </li>
    <li>
      <a href="../articles/creating-spatial-trajectories.html">Creating Spatial Trajectories</a>
    </li>
    <li>
      <a href="../articles/spatial-trajectory-screening.html">Spatial Trajectory Screening</a>
    </li>
    <li class="dropdown-header">Visualization</li>
    <li>
      <a href="../articles/plotting-surface.html">Surface Plotting</a>
    </li>
  </ul>
</li>
<li>
  <a href="../articles/newslog.html">Newslog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Initiation &amp; Preprocessing: Visium</h1>
            
      

      <div class="hidden name"><code>initiation-and-preprocessing-visium.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="initiation">1. Initiation<a class="anchor" aria-label="anchor" href="#initiation"></a>
</h2>
<p>To initiate a <code>SPATA2</code> object directly from the Visium
output use the function <code><a href="../reference/initiateSpataObjectVisium.html">initiateSpataObjectVisium()</a></code>. It
works for both slide types, those with a capture area of 6.5mm x 6.5mm
(referred to as <code>VisiumSmall</code> in SPATA2) and of 11mm x 11mm
(referred to as <code>VisiumLarge</code> in SPATA2). This example
vignette uses data from a 6.5mm x 6.5mm. You can download the folder <a href="https://drive.google.com/drive/folders/1CC1HUDpoZpgX1U9Hzcr-074_8lgdxu4q?usp=sharing" class="external-link">here</a>.
Save the folder under a directory that you later provide as input for
<code>directory_visium</code>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://themilolab.com/" class="external-link">SPATA2</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">object</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="../reference/initiateSpataObjectVisium.html">initiateSpataObjectVisium</a></span><span class="op">(</span></span>
<span>    sample_name <span class="op">=</span> <span class="st">"UKF269T"</span>, </span>
<span>    directory_visium <span class="op">=</span> <span class="st">"initiate_VisiumSmall"</span> <span class="co"># adjust to your liking </span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="co"># show overview</span></span>
<span><span class="va">object</span></span></code></pre></div>
<pre><code><span><span class="co">## An object of class SPATA2</span></span>
<span><span class="co">## Sample: UKF269T </span></span>
<span><span class="co">## Size: 3217 x 33538 (spots x molecules)</span></span>
<span><span class="co">## Memory: 220.31 Mb</span></span>
<span><span class="co">## Platform: VisiumSmall </span></span>
<span><span class="co">## Molecular assays (1):</span></span>
<span><span class="co">## 1. Assay</span></span>
<span><span class="co">##  Molecular modality: gene</span></span>
<span><span class="co">##  Distinct molecules: 33538</span></span>
<span><span class="co">##  Matrices (1):</span></span>
<span><span class="co">##  -counts (active)</span></span>
<span><span class="co">## Registered images (1):</span></span>
<span><span class="co">## - lowres (582x600 px, active, loaded)</span></span>
<span><span class="co">## Meta variables (1): sample</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="image-processing">2. Image processing<a class="anchor" aria-label="anchor" href="#image-processing"></a>
</h2>
<p>(Beta; still in progress since it does not work as well on images
with fluent tissue background transition.)</p>
<p>Image processing is not required. However, it facilitates the
integration of histological features as displayed by the histology
image, the Visium platform allows to integrate. The goal of image
processing is to identify the precise spatial outline of the tissue on
the histology slide. The function <code><a href="../reference/processImage.html">processImage()</a></code> is a
wrapper around <code><a href="../reference/identifyPixelContent.html">identifyPixelContent()</a></code> and
<code>identifyTissueOutline(..., method = "image")</code>. Please refer
to the documentation of either function to obtain more information.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/identifyPixelContent.html">identifyPixelContent</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span></code></pre></div>
<p>The results of <code><a href="../reference/identifyPixelContent.html">identifyPixelContent()</a></code> can be plotted
with <code><a href="../reference/plotImageMask.html">plotImageMask()</a></code> and
<code><a href="../reference/plotImageMask.html">plotPixelContent()</a></code>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotImageMask.html">plotImageMask</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plotImageMask.html">plotPixelContent</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span></code></pre></div>
<p><img src="initiation-and-preprocessing-visium_files/figure-html/unnamed-chunk-4-1.png" width="50%"><img src="initiation-and-preprocessing-visium_files/figure-html/unnamed-chunk-4-2.png" width="50%"></p>
<p>The results of
<code>identifyTissueOutline(..., method = "image")</code> are best
visualized by setting <code>outline = TRUE</code> with the
<code><a href="../reference/plotImage.html">plotImage()</a></code> function.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotImage.html">plotImage</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plotImage.html">plotImage</a></span><span class="op">(</span><span class="va">object</span>, outline <span class="op">=</span> <span class="cn">TRUE</span>, line_size <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p><img src="initiation-and-preprocessing-visium_files/figure-html/unnamed-chunk-5-1.png" width="50%"><img src="initiation-and-preprocessing-visium_files/figure-html/unnamed-chunk-5-2.png" width="50%"></p>
</div>
<div class="section level2">
<h2 id="spatial-processing">3. Spatial processing<a class="anchor" aria-label="anchor" href="#spatial-processing"></a>
</h2>
<p>This step should not be skipped! Many functions in SPATA2 need to
know where the edge of the tissue section is and they need to know if
there are multiple tissue sections. This kind of data is not provided
with the standard output of most platforms and needs to be computed.
With spatial processing we particularly refer to the identification of
spatial outliers - observations that are part of the data set but lie
too far away from the contiguous tissue section to be considered part of
the data set that is of actual interest. In case of the Visium platform
they are usually artefacts. The function
<code>identifyTissueOutline(..., method = "obs")</code> uses the <a href="https://en.wikipedia.org/wiki/DBSCAN" class="external-link">DBSCAN</a> algorithm to
identify potential spatial outliers. The results are stored in a
variable called <em>tissue_section</em> which also contains information
to which tissue section each observation was assigned in case of
multiple tissue sections.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># this is the default input for the visium platform and has already been </span></span>
<span><span class="co"># called in initiateSpataObjectVisium(). </span></span>
<span><span class="co"># if the results do not satisfy you, you can run it over and over again with </span></span>
<span><span class="co"># different parameter inputs </span></span>
<span><span class="va">object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/identifyTissueOutline.html">identifyTissueOutline</a></span><span class="op">(</span><span class="va">object</span>, method <span class="op">=</span> <span class="st">"obs"</span>, eps <span class="op">=</span> <span class="st">"125um"</span>, minPts <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span><span class="va">object</span>, color_by <span class="op">=</span> <span class="st">"tissue_section"</span>, pt_clrp <span class="op">=</span> <span class="st">"tab20"</span><span class="op">)</span></span></code></pre></div>
<p><img src="initiation-and-preprocessing-visium_files/figure-html/unnamed-chunk-6-1.png" width="50%"></p>
<div class="section level3">
<h3 id="tissue-outline-parameters">3.1 Tissue outline parameters<a class="anchor" aria-label="anchor" href="#tissue-outline-parameters"></a>
</h3>
<p>DBSCAN requires two important parameters, namely <code>eps</code> and
<code>minPts</code>. Depending on the platform SPATA2 defaults to
different parameter inputs - those that have worked well for us in the
past. You can, however, change the parameter input. Consider this
example mouse brain data set with two tissue sections obviously
separated.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">object_example</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/loadExampleObject.html">loadExampleObject</a></span><span class="op">(</span><span class="st">"LMU_MCI"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">object_example</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/identifyTissueOutline.html">identifyTissueOutline</a></span><span class="op">(</span><span class="va">object_example</span>, eps <span class="op">=</span> <span class="st">"125um"</span>, minPts <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span><span class="va">object_example</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span><span class="va">object_example</span>, color_by <span class="op">=</span> <span class="st">"tissue_section"</span><span class="op">)</span></span></code></pre></div>
<p><img src="initiation-and-preprocessing-visium_files/figure-html/unnamed-chunk-7-1.png" width="50%"><img src="initiation-and-preprocessing-visium_files/figure-html/unnamed-chunk-7-2.png" width="50%"></p>
<p>There are no spatial outliers in contrast to the glioma sample
<em>UKF269T</em>. But what happens if we change the parameter input?</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># example 1: increase eps</span></span>
<span><span class="va">object_example</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/identifyTissueOutline.html">identifyTissueOutline</a></span><span class="op">(</span><span class="va">object_example</span>, eps <span class="op">=</span>  <span class="st">"1.5mm"</span>, minPts <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span>
<span></span>
<span><span class="va">out_example_1</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span><span class="va">object_example</span>, color_by <span class="op">=</span> <span class="st">"tissue_section"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="../reference/ggpLayerScaleBarSI.html">ggpLayerScaleBarSI</a></span><span class="op">(</span><span class="va">object_example</span>, sb_dist <span class="op">=</span> <span class="st">"1.5mm"</span>, sb_pos <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">200</span>, <span class="fl">100</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>subtitle <span class="op">=</span> <span class="st">"eps = 1.5mm"</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># example 2: increase minPts</span></span>
<span><span class="va">object_example</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/identifyTissueOutline.html">identifyTissueOutline</a></span><span class="op">(</span><span class="va">object_example</span>, eps <span class="op">=</span> <span class="st">"125um"</span>, minPts <span class="op">=</span> <span class="fl">6</span><span class="op">)</span></span>
<span></span>
<span><span class="va">out_example_2</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span></span>
<span>    object <span class="op">=</span> <span class="va">object_example</span>,</span>
<span>    color_by <span class="op">=</span> <span class="st">"tissue_section"</span>,</span>
<span>    clrp_adjust <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"tissue_section_0"</span> <span class="op">=</span> <span class="st">"black"</span><span class="op">)</span></span>
<span>    <span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="../reference/ggpLayerScaleBarSI.html">ggpLayerScaleBarSI</a></span><span class="op">(</span><span class="va">object_example</span>, sb_dist <span class="op">=</span> <span class="st">"125um"</span>, sb_pos <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">200</span>, <span class="fl">100</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>subtitle <span class="op">=</span> <span class="st">"minPts = 6"</span><span class="op">)</span></span>
<span>  </span>
<span><span class="co"># left </span></span>
<span><span class="va">out_example_1</span> </span>
<span></span>
<span><span class="co"># right </span></span>
<span><span class="va">out_example_2</span></span></code></pre></div>
<p><img src="initiation-and-preprocessing-visium_files/figure-html/unnamed-chunk-8-1.png" width="50%"><img src="initiation-and-preprocessing-visium_files/figure-html/unnamed-chunk-8-2.png" width="50%"></p>
<p>In the first example, we increase the <code>eps</code> parameter -
the radius starting from an observation in which is screened for
adjacent neighbors. 1.5 mm though is way too much. The lower spots of
the upper tissue section find <em>neighbors</em> in the upper spots of
the lower section. In the second example we increase <code>minPts</code>
to a value of 6. Given the arrangement of Visium spots each should have
6 neighbors, right? No! Spots on the edge belong to the tissue but they
do not have 6 neighbors within they reach of <code>eps</code> (defaults
to 125um) but usually only 3 to 4 spots.<br>
Are the black spots outliers or do they belong to the tissue section?
You can choose to keep or remove them.</p>
</div>
<div class="section level3">
<h3 id="spatial-outliers">3.2 Spatial outliers<a class="anchor" aria-label="anchor" href="#spatial-outliers"></a>
</h3>
<p>What has been defined as <em>tissue_section_0</em> could not be
assigned to a tissue section and is considered a spatial outlier. Use
<code><a href="../reference/identifySpatialOutliers.html">identifySpatialOutliers()</a></code> to make it official. In case of
the glioma sample <em>UKF269T</em> it is quite obvious that the spots
marked as outliers do not belong to the tissue.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># uses the results of identifyTissueOutline() to create a logical variable called sp_outlier</span></span>
<span><span class="va">object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/identifySpatialOutliers.html">identifySpatialOutliers</a></span><span class="op">(</span><span class="va">object</span>, method <span class="op">=</span> <span class="st">"obs"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">plot_with_outliers</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span><span class="va">object</span>, color_by <span class="op">=</span> <span class="st">"sp_outlier"</span>, clrp_adjust <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"TRUE"</span> <span class="op">=</span> <span class="st">"blue"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># remove where sp_outlier == TRUE</span></span>
<span><span class="va">object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/removeSpatialOutliers.html">removeSpatialOutliers</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span>
<span></span>
<span><span class="va">plot_without_outliers</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span><span class="va">object</span>, color_by <span class="op">=</span> <span class="st">"sp_outlier"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># left plot</span></span>
<span><span class="va">plot_with_outliers</span></span>
<span></span>
<span><span class="co"># right plot</span></span>
<span><span class="va">plot_without_outliers</span></span></code></pre></div>
<p><img src="initiation-and-preprocessing-visium_files/figure-html/unnamed-chunk-9-1.png" width="50%"><img src="initiation-and-preprocessing-visium_files/figure-html/unnamed-chunk-9-2.png" width="50%"></p>
<p>Note, <code><a href="../reference/identifySpatialOutliers.html">identifySpatialOutliers()</a></code> can also identify
outliers based on the tissue outline identified with
<code>identifyTissueOutline(..., method = "image")</code>. Also both
methods, <em>image</em> and <em>obs</em> can be combined. Refer to the
documentation of the function for more information.</p>
</div>
</div>
<div class="section level2">
<h2 id="data-processing">4. Data processing<a class="anchor" aria-label="anchor" href="#data-processing"></a>
</h2>
<p>These steps are about additional noise removal as well as about
processing raw counts.</p>
<div class="section level3">
<h3 id="data-cleaning">4.1. Data cleaning<a class="anchor" aria-label="anchor" href="#data-cleaning"></a>
</h3>
<p>First you might want to remove certain genes from the raw count
matrix. There are wrappers for certain steps like
<code><a href="../reference/removeMolecules.html">removeGenesStress()</a></code> and
<code><a href="../reference/removeMolecules.html">removeGenesZeroCounts()</a></code>. Individual genes can always be
removed with <code><a href="../reference/removeMolecules.html">removeGenes()</a></code>.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># before</span></span>
<span><span class="fu"><a href="../reference/nMolecules.html">nGenes</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span>
<span><span class="co">## [1] 33538</span></span>
<span></span>
<span><span class="co"># removes stress genes</span></span>
<span><span class="va">object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/removeMolecules.html">removeGenesStress</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># removes genes that were not detected in any of the observations</span></span>
<span><span class="va">object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/removeMolecules.html">removeGenesZeroCounts</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># afterwards</span></span>
<span><span class="fu"><a href="../reference/nMolecules.html">nGenes</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span>
<span><span class="co">## [1] 21556</span></span></code></pre></div>
<p>In some cases there are observations - in case of Visium barcoded
spots - left with no counts at all. If this is the case
<code><a href="../reference/removeObs.html">removeObsZeroCounts()</a></code> removes them. If there are none
nothing happens.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># before</span></span>
<span><span class="fu"><a href="../reference/nObs.html">nObs</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span>
<span><span class="co">## [1] 3213</span></span>
<span></span>
<span><span class="co"># check for and remove observations with zero counts</span></span>
<span><span class="va">object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/removeObs.html">removeObsZeroCounts</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># afterwards </span></span>
<span><span class="fu"><a href="../reference/nObs.html">nObs</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span>
<span><span class="co">## [1] 3213</span></span></code></pre></div>
<p>Afterwards, you can compute meta data for the observations.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/computeMetaFeatures.html">computeMetaFeatures</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># plot left</span></span>
<span><span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span><span class="va">object</span>, color_by <span class="op">=</span> <span class="st">"n_counts_gene"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># plot right</span></span>
<span><span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span><span class="va">object</span>, color_by <span class="op">=</span> <span class="st">"n_distinct_gene"</span><span class="op">)</span></span></code></pre></div>
<p><img src="initiation-and-preprocessing-visium_files/figure-html/unnamed-chunk-12-1.png" width="50%"><img src="initiation-and-preprocessing-visium_files/figure-html/unnamed-chunk-12-2.png" width="50%"></p>
</div>
<div class="section level3">
<h3 id="matrix-processing">4.2 Matrix processing<a class="anchor" aria-label="anchor" href="#matrix-processing"></a>
</h3>
<p>The <code>SPATA2</code> object is initiated with a raw count matrix.
For almost all downstream analysis steps it is recommended to use
processed matrices. The first step is usually log-normalization. To
create a normalized matrix use <code><a href="../reference/normalizeCounts.html">normalizeCounts()</a></code>. It uses
<code><a href="https://satijalab.org/seurat/reference/NormalizeData.html" class="external-link">Seurat::NormalizeData()</a></code> in the background. The input
options for <code>method</code> correspond to the options in this
function from the Seurat package. By default, the normalized matrix is
named after input for <code>method</code>, activated and thus used by
default in downstream analysis and visualization. The function
<code><a href="../reference/normalizeCounts.html">normalizeCounts()</a></code> can be called multiple times with
different inputs for <code>method</code> which populates the list of
processed matrices in the respective assay. Furthermore, processed
matrices can be added with <code><a href="../reference/addProcessedMatrix.html">addProcessedMatrix()</a></code> if you want
to create them with SPATA2-extern functions. The default matrix that is
used can be set with <code><a href="../reference/activateMatrix.html">activateMatrix()</a></code>. By default,
<code><a href="../reference/normalizeCounts.html">normalizeCounts()</a></code> activates the processed matrix it has
created.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># obtain matrix names prior to normalization</span></span>
<span><span class="fu"><a href="../reference/getMatrixNames.html">getMatrixNames</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span>
<span><span class="co">## [1] "counts"</span></span>
<span></span>
<span><span class="va">plot_before</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span><span class="va">object</span>, color_by <span class="op">=</span> <span class="st">"MAG"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>color <span class="op">=</span> <span class="st">"MAG\n(Counts)"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># create log normalized matrix</span></span>
<span><span class="va">object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/normalizeCounts.html">normalizeCounts</a></span><span class="op">(</span><span class="va">object</span>, method <span class="op">=</span> <span class="st">"LogNormalize"</span><span class="op">)</span></span>
<span><span class="co">## Normalizing layer: counts</span></span>
<span><span class="co">## 02:03:11 Active matrix in assay 'gene': 'LogNormalize'</span></span>
<span></span>
<span><span class="co"># obtain matrix names after normalization</span></span>
<span><span class="fu"><a href="../reference/getMatrixNames.html">getMatrixNames</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span>
<span><span class="co">## [1] "counts"       "LogNormalize"</span></span>
<span></span>
<span><span class="co"># check active matrix </span></span>
<span><span class="fu"><a href="../reference/activateMatrix.html">activeMatrix</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span>
<span><span class="co">## [1] "LogNormalize"</span></span>
<span></span>
<span><span class="va">plot_afterwards</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span><span class="va">object</span>, color_by <span class="op">=</span> <span class="st">"MAG"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>color <span class="op">=</span> <span class="st">"MAG\n(logNorm)"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># left plot</span></span>
<span><span class="va">plot_before</span></span>
<span></span>
<span><span class="co"># right plot</span></span>
<span><span class="va">plot_afterwards</span></span></code></pre></div>
<p><img src="initiation-and-preprocessing-visium_files/figure-html/unnamed-chunk-13-1.png" width="50%"><img src="initiation-and-preprocessing-visium_files/figure-html/unnamed-chunk-13-2.png" width="50%"></p>
</div>
</div>
<div class="section level2">
<h2 id="spatially-variable-genes">5. Spatially variable genes<a class="anchor" aria-label="anchor" href="#spatially-variable-genes"></a>
</h2>
<p>Since spatial transcriptomics is all about spatial pattern of gene
expression you might want to identify genes with a spatial pattern that
is non-random. We recommend the prefiltering for these kind of genes,
for instance, in our SPATA2 intern Spatial Annotation Screening
algorithm. Spatially variable genes can, for instance, be identified
using the wrapper around SPARKX (Zhu et al., 2021).</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># results are stored inside the SPATA2 object</span></span>
<span><span class="va">object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runSPARKX.html">runSPARKX</a></span><span class="op">(</span><span class="va">object</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## ## ===== SPARK-X INPUT INFORMATION ==== </span></span>
<span><span class="co">## ## number of total samples: 3213 </span></span>
<span><span class="co">## ## number of total genes: 21556 </span></span>
<span><span class="co">## ## Running with single core, may take some time </span></span>
<span><span class="co">## ## Testing With Projection Kernel</span></span>
<span><span class="co">## ## Testing With Gaussian Kernel 1</span></span>
<span><span class="co">## ## Testing With Gaussian Kernel 2</span></span>
<span><span class="co">## ## Testing With Gaussian Kernel 3</span></span>
<span><span class="co">## ## Testing With Gaussian Kernel 4</span></span>
<span><span class="co">## ## Testing With Gaussian Kernel 5</span></span>
<span><span class="co">## ## Testing With Cosine Kernel 1</span></span>
<span><span class="co">## ## Testing With Cosine Kernel 2</span></span>
<span><span class="co">## ## Testing With Cosine Kernel 3</span></span>
<span><span class="co">## ## Testing With Cosine Kernel 4</span></span>
<span><span class="co">## ## Testing With Cosine Kernel 5</span></span></code></pre>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># get genes with a p-value &lt; 0.01</span></span>
<span><span class="va">sparkx_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runSPARKX.html">getSparkxGenes</a></span><span class="op">(</span><span class="va">object</span>, threshold_pval <span class="op">=</span> <span class="fl">0.01</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">sparkx_genes</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  chr [1:13950] "RPL22" "ID3" "MARCKSL1" "PHC2" "RPS8" "WLS" "GNG5" "RPL5" ...</span></span></code></pre>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># visualize in space</span></span>
<span><span class="fu"><a href="../reference/plotSurfaceComparison.html">plotSurfaceComparison</a></span><span class="op">(</span><span class="va">object</span>, color_by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">sparkx_genes</span>, <span class="fl">6</span><span class="op">)</span>, nrow <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="initiation-and-preprocessing-visium_files/figure-html/unnamed-chunk-14-1.png" width="100%"></p>
</div>
<div class="section level2">
<h2 id="conclusion-and-more-data-sets">6. Conclusion and more data sets<a class="anchor" aria-label="anchor" href="#conclusion-and-more-data-sets"></a>
</h2>
<p>That’s it. The object can be used for any downstream analyses such as
<a href="dimensional-reduction.html">dimensional reduction</a>, <a href="clustering.html">clustering</a>, <a href="spatial-annotation-screening.html">spatial annotation
screening</a> or <a href="spatial-trajectory-screening.html">spatial
trajectory screening</a>. Refer to tab Tutorials for more links.
Furthermore, you can skim our curated data base of spatial data sets for
those of platform VisiumSmall or VisiumLarge using <a href="spata-data.html">SPATAData</a>.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># load package</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SPATAData</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># filter for samples from platform VisiumHD</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SPATAData/man/sourceDataFrame.html" class="external-link">sourceDataFrame</a></span><span class="op">(</span><span class="va">platform</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"VisiumSmall"</span>, <span class="st">"VisiumLarge"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 161 × 33</span></span></span>
<span><span class="co">##    sample_name  comment donor_id donor_species grade histo_class histo_class_sub</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>          </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> HumanBreast… <span style="color: #BB0000;">NA</span>      <span style="color: #BB0000;">NA</span>       Homo sapiens  <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>          <span style="color: #BB0000;">NA</span>             </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> HumanGliobl… <span style="color: #BB0000;">NA</span>      <span style="color: #BB0000;">NA</span>       Homo sapiens  IV    Glioblasto… <span style="color: #BB0000;">NA</span>             </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> HumanTonsil… <span style="color: #BB0000;">NA</span>      <span style="color: #BB0000;">NA</span>       Homo sapiens  <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>          <span style="color: #BB0000;">NA</span>             </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> HumanColore… <span style="color: #BB0000;">NA</span>      <span style="color: #BB0000;">NA</span>       Homo sapiens  <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>          <span style="color: #BB0000;">NA</span>             </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> HumanGliobl… <span style="color: #BB0000;">NA</span>      <span style="color: #BB0000;">NA</span>       Homo sapiens  <span style="color: #BB0000;">NA</span>    Glioblasto… <span style="color: #BB0000;">NA</span>             </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> HumanKidney… <span style="color: #BB0000;">NA</span>      <span style="color: #BB0000;">NA</span>       Homo sapiens  <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>          <span style="color: #BB0000;">NA</span>             </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> HumanLungCa… <span style="color: #BB0000;">NA</span>      <span style="color: #BB0000;">NA</span>       Homo sapiens  <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>          <span style="color: #BB0000;">NA</span>             </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> HumanOvaria… <span style="color: #BB0000;">NA</span>      <span style="color: #BB0000;">NA</span>       Homo sapiens  <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>          <span style="color: #BB0000;">NA</span>             </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> Breast_Canc… <span style="color: #BB0000;">NA</span>      <span style="color: #BB0000;">NA</span>       Homo sapiens  <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>          <span style="color: #BB0000;">NA</span>             </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> Breast_Canc… <span style="color: #BB0000;">NA</span>      <span style="color: #BB0000;">NA</span>       Homo sapiens  <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>          <span style="color: #BB0000;">NA</span>             </span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 151 more rows</span></span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 26 more variables: institution &lt;chr&gt;, lm_source &lt;dttm&gt;, organ &lt;chr&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   organ_part &lt;chr&gt;, organ_side &lt;chr&gt;, pathology &lt;chr&gt;, platform &lt;chr&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   pub_citation &lt;chr&gt;, pub_doi &lt;chr&gt;, pub_journal &lt;chr&gt;, pub_year &lt;dbl&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   sex &lt;chr&gt;, source &lt;chr&gt;, tags &lt;chr&gt;, tissue_age &lt;dbl&gt;, web_link &lt;chr&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   workgroup &lt;chr&gt;, mean_counts &lt;dbl&gt;, median_counts &lt;dbl&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   modality_gene &lt;lgl&gt;, modality_metabolite &lt;lgl&gt;, modality_protein &lt;lgl&gt;, …</span></span></span></code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Jan Kueckelhaus, Dieter-Henrik Heiland.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

      </footer>
</div>






  </body>
</html>
