<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Initiation &amp; Preprocessing: VisiumHD • SPATA2</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Initiation &amp; Preprocessing: VisiumHD">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SPATA2</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">3.1.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/installation.html">Installation</a>
</li>
<li>
  <a href="../articles/spata-data.html">SPATAData</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">Initiating a SPATA2 object</li>
    <li>
      <a href="../articles/initiation-and-preprocessing-customized.html">Initiation &amp; Preprocessing with individual data</a>
    </li>
    <li>
      <a href="../articles/initiation-and-preprocessing-merfish.html">Initiation &amp; Preprocessing with MERFISH data</a>
    </li>
    <li>
      <a href="../articles/initiation-and-preprocessing-visium.html">Initiation &amp; Preprocessing with Visium data</a>
    </li>
    <li>
      <a href="../articles/initiation-and-preprocessing-visium-hd.html">Initiation &amp; Preprocessing with VisiumHD data</a>
    </li>
    <li class="dropdown-header">Basic Manipulation of a SPATA2 object</li>
    <li>
      <a href="../articles/adding-data.html">Adding Data to SPATA2 objects</a>
    </li>
    <li>
      <a href="../articles/package-compatibility.html">Platform Compatibility</a>
    </li>
    <li>
      <a href="../articles/molecular-variables.html">Molecular Variables</a>
    </li>
    <li>
      <a href="../articles/image-handling.html">Image Handling</a>
    </li>
    <li>
      <a href="../articles/spatial-segmentation.html">Spatial Segmentation</a>
    </li>
    <li class="dropdown-header">Implemented Algorithms &amp; Concepts</li>
    <li>
      <a href="../articles/cnv.html">Copy Number Variations</a>
    </li>
    <li>
      <a href="../articles/clustering.html">Clustering</a>
    </li>
    <li>
      <a href="../articles/dimensional-reduction.html">Dimensional Reduction</a>
    </li>
    <li>
      <a href="../articles/dea.html">Differential Expression</a>
    </li>
    <li>
      <a href="../articles/gsea.html">Gene Set Enrichment</a>
    </li>
    <li>
      <a href="../articles/spatial-measures.html">Spatial Measures</a>
    </li>
    <li class="dropdown-header">Spatial Gradient Screening</li>
    <li>
      <a href="../articles/creating-spatial-annotations.html">Creating Spatial Annotations</a>
    </li>
    <li>
      <a href="../articles/using-spatial-annotations.html">Using Spatial Annotations</a>
    </li>
    <li>
      <a href="../articles/spatial-annotation-screening.html">Spatial Annotation Screening</a>
    </li>
    <li>
      <a href="../articles/creating-spatial-trajectories.html">Creating Spatial Trajectories</a>
    </li>
    <li>
      <a href="../articles/spatial-trajectory-screening.html">Spatial Trajectory Screening</a>
    </li>
    <li class="dropdown-header">Visualization</li>
    <li>
      <a href="../articles/plotting-surface.html">Surface Plotting</a>
    </li>
  </ul>
</li>
<li>
  <a href="../articles/newslog.html">Newslog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Initiation &amp; Preprocessing: VisiumHD</h1>
            
      

      <div class="hidden name"><code>initiation-and-preprocessing-visium-hd.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="initiation">1. Initiation<a class="anchor" aria-label="anchor" href="#initiation"></a>
</h2>
<p>To initiate a <code>SPATA2</code> object directly from the Visium
output use the function <code><a href="../reference/initiateSpataObjectVisiumHD.html">initiateSpataObjectVisiumHD()</a></code>. This
example vignette uses data from an <a href="https://www.10xgenomics.com/datasets/visium-hd-cytassist-gene-expression-libraries-human-pancreas" class="external-link">example
data</a> set provided by 10X Genomics. You can download the folder <a href="https://drive.google.com/drive/folders/1Ym5Fj7MFh3G1F1Hd_pshuQ5Q0ee7RACD?usp=sharing" class="external-link">here</a>.</p>
<p>Note: It is crucial to install the package arrow in a way that
<code><a href="https://arrow.apache.org/docs/r/reference/read_parquet.html" class="external-link">arrow::read_parquet()</a></code> works. There are several ways.
Installing the package with
<code>install.packages('arrow', repos = 'https://apache.r-universe.dev')</code>
worked reliably for us.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://themilolab.com/" class="external-link">SPATA2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">object</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="../reference/initiateSpataObjectVisiumHD.html">initiateSpataObjectVisiumHD</a></span><span class="op">(</span></span>
<span>    sample_name <span class="op">=</span> <span class="st">"HumanPancreasHD"</span>, </span>
<span>    directory_visium <span class="op">=</span> <span class="st">"my/path/to/VisiumHD_folder"</span> <span class="co"># adjust to your liking </span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="co"># show overview</span></span>
<span><span class="va">object</span></span></code></pre></div>
<pre><code><span><span class="co">## An object of class SPATA2</span></span>
<span><span class="co">## Sample: HumanPancreasHD </span></span>
<span><span class="co">## Size: 106574 x 18085 (spots x molecules)</span></span>
<span><span class="co">## Memory: 663.77 Mb</span></span>
<span><span class="co">## Platform: VisiumHD (Resolution: 16um) </span></span>
<span><span class="co">## Molecular assays (1):</span></span>
<span><span class="co">## 1. Assay</span></span>
<span><span class="co">##  Molecular modality: gene</span></span>
<span><span class="co">##  Distinct molecules: 18085</span></span>
<span><span class="co">##  Matrices (1):</span></span>
<span><span class="co">##  -counts (active)</span></span>
<span><span class="co">## Registered images (2):</span></span>
<span><span class="co">## - hires (6000x5968 px, not loaded)</span></span>
<span><span class="co">## - lowres (600x597 px, active, loaded)</span></span>
<span><span class="co">## Meta variables (2): sample, tissue_section</span></span></code></pre>
<p>While VisiumHD offers standard resolutions with 16um, 8um and 2um,
the squared nature of the grid of spots allows to analyze data in
individual resolutions. Hence, if you want to analyze your data in
resolution 10um, 24um, 30um or 40um (etc.) you can specify that and the
function <code><a href="../reference/reduceResolutionVisiumHD.html">reduceResolutionVisiumHD()</a></code> works in the background
of <code><a href="../reference/initiateSpataObjectVisiumHD.html">initiateSpataObjectVisiumHD()</a></code> to create the resolution
of your liking. Read more on that in this <a href="individual-visiumHD-resolution.html">vignette</a>.</p>
<p>Furthermore, you can adjust the resolutions with which images are
used. Particularly, the hires image of the VisiumHD platform is often
too large for R to handle smoothly since R is not particularly efficient
when it comes to handling images of a certain size. This resizing
functionality allows you to adjust the size in which the image is
handled when used in order to optimize the ratio between image
resolution and computational performance. The function required is
<code><a href="../reference/resizeImage.html">resizeImage()</a></code> which can be called after initiation or used
during initiation with the <code>resize_images</code> argument.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># example genes </span></span>
<span><span class="va">example_genes</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>    <span class="st">'INS'</span>, <span class="st">'REG3A'</span>, <span class="st">'TTR'</span>, <span class="st">'GCG'</span>, <span class="st">'GHRL'</span>, <span class="st">'SST'</span>, <span class="st">'VIP'</span>, <span class="st">'PPY'</span>, <span class="st">'CARTPT'</span>, <span class="st">'REG1B'</span>, </span>
<span>    <span class="st">'NPY'</span>, <span class="st">'REG3G'</span>, <span class="st">'ALDOB'</span>, <span class="st">'PCSK1N'</span>, <span class="st">'IAPP'</span>, <span class="st">'IGKC'</span>, <span class="st">'DKK4'</span>, <span class="st">'CHGA'</span>, <span class="st">'IGHA1'</span>, </span>
<span>    <span class="st">'CPA3'</span>, <span class="st">'PLA2G2A'</span>, <span class="st">'GAL'</span>, <span class="st">'RBP4'</span>, <span class="st">'SLC30A8'</span>, <span class="st">'HBA2'</span></span>
<span>    <span class="op">)</span></span>
<span></span>
<span><span class="co"># example object using the resize options</span></span>
<span><span class="va">object_resized</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="../reference/initiateSpataObjectVisiumHD.html">initiateSpataObjectVisiumHD</a></span><span class="op">(</span></span>
<span>    sample_name <span class="op">=</span> <span class="st">"HumanPancreasHD"</span>, </span>
<span>    directory_visium <span class="op">=</span> <span class="st">"my/path/to/VisiumHD_folder"</span>, </span>
<span>    square_res <span class="op">=</span> <span class="st">"20um"</span>,</span>
<span>    genes <span class="op">=</span> <span class="va">example_genes</span>, </span>
<span>    unload <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>    resize_images <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>hires <span class="op">=</span> <span class="fl">0.5</span>, lowres <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="co"># show object</span></span>
<span><span class="va">object_resized</span></span></code></pre></div>
<pre><code><span><span class="co">## An object of class SPATA2</span></span>
<span><span class="co">## Sample: HumanPancreasHD </span></span>
<span><span class="co">## Size: 68558 x 25 (spots x molecules)</span></span>
<span><span class="co">## Memory: 1.31 Gb</span></span>
<span><span class="co">## Platform: VisiumHD (Resolution: 20um) </span></span>
<span><span class="co">## Molecular assays (1):</span></span>
<span><span class="co">## 1. Assay</span></span>
<span><span class="co">##  Molecular modality: gene</span></span>
<span><span class="co">##  Distinct molecules: 25</span></span>
<span><span class="co">##  Matrices (1):</span></span>
<span><span class="co">##  -counts (active)</span></span>
<span><span class="co">## Registered images (2):</span></span>
<span><span class="co">## - hires (3000x2984 px, loaded)</span></span>
<span><span class="co">## - lowres (300x298.5 px, active, loaded)</span></span>
<span><span class="co">## Meta variables (5): sample, square_exp, square_count, square_perc, tissue_section</span></span></code></pre>
<p>Notice the different image resolutions between <code>object</code>
and <code>object_resized</code> in the printed object summary. Images
can be plotted with <code><a href="../reference/plotImage.html">plotImage()</a></code>.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># left plot</span></span>
<span><span class="fu"><a href="../reference/plotImage.html">plotImage</a></span><span class="op">(</span><span class="va">object_resized</span>, img_name <span class="op">=</span> <span class="st">"lowres"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># right plot</span></span>
<span><span class="fu"><a href="../reference/plotImage.html">plotImage</a></span><span class="op">(</span><span class="va">object_resized</span>, img_name <span class="op">=</span> <span class="st">"hires"</span><span class="op">)</span></span></code></pre></div>
<p><img src="initiation-and-preprocessing-visium-hd_files/figure-html/unnamed-chunk-5-1.png" width="50%"><img src="initiation-and-preprocessing-visium-hd_files/figure-html/unnamed-chunk-5-2.png" width="50%"></p>
<p>For this vignette, we’ll use the 16um object, simply stored in
<code>object</code>.</p>
</div>
<div class="section level2">
<h2 id="image-processing">2. Image processing<a class="anchor" aria-label="anchor" href="#image-processing"></a>
</h2>
<p>Image processing is not required. However, it facilitates the
integration of histological features as displayed by the histology
image, the Visium platform allows to integrate. The goal of image
processing is to identify the precise spatial outline of the tissue on
the histology slide. The function <code><a href="../reference/processImage.html">processImage()</a></code> is a
wrapper around <code><a href="../reference/identifyPixelContent.html">identifyPixelContent()</a></code> and
<code>identifyTissueOutline(..., method = "image")</code>. Please refer
to the documentation of either function to obtain more information.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># by default, the active image is always used</span></span>
<span><span class="co"># you might need to adjust sigma, frgmt_threshold and other parameters to optimize results </span></span>
<span><span class="va">object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/identifyPixelContent.html">identifyPixelContent</a></span><span class="op">(</span><span class="va">object</span>, frgmt_threshold <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.01</span>,<span class="fl">0.05</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/identifyTissueOutline.html">identifyTissueOutline</a></span><span class="op">(</span><span class="va">object</span>, method <span class="op">=</span> <span class="st">"image"</span><span class="op">)</span></span></code></pre></div>
<p>The results of <code><a href="../reference/identifyPixelContent.html">identifyPixelContent()</a></code> can be plotted
with <code><a href="../reference/plotImageMask.html">plotImageMask()</a></code> and
<code><a href="../reference/plotImageMask.html">plotPixelContent()</a></code>.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotImageMask.html">plotImageMask</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plotImageMask.html">plotPixelContent</a></span><span class="op">(</span><span class="va">object</span>, clrp <span class="op">=</span> <span class="st">"jco"</span><span class="op">)</span></span></code></pre></div>
<p><img src="initiation-and-preprocessing-visium-hd_files/figure-html/unnamed-chunk-7-1.png" width="50%"><img src="initiation-and-preprocessing-visium-hd_files/figure-html/unnamed-chunk-7-2.png" width="50%"></p>
<p>The results of
<code>identifyTissueOutline(..., method = "image")</code> are best
visualized by setting <code>outline = TRUE</code> with the
<code><a href="../reference/plotImage.html">plotImage()</a></code> function.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># get colors of one of the known color palettes</span></span>
<span><span class="va">cvec</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/color_vector.html">color_vector</a></span><span class="op">(</span><span class="st">"jco"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># left plot</span></span>
<span><span class="fu"><a href="../reference/plotImage.html">plotImage</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># right plot</span></span>
<span><span class="fu"><a href="../reference/plotImage.html">plotImage</a></span><span class="op">(</span><span class="va">object</span>, outline <span class="op">=</span> <span class="cn">TRUE</span>, line_size <span class="op">=</span> <span class="fl">1</span>, line_color <span class="op">=</span> <span class="va">cvec</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, fragments <span class="op">=</span> <span class="va">cvec</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p><img src="initiation-and-preprocessing-visium-hd_files/figure-html/unnamed-chunk-8-1.png" width="50%"><img src="initiation-and-preprocessing-visium-hd_files/figure-html/unnamed-chunk-8-2.png" width="50%"></p>
<p>Note, that the tissue fragment identified on the left of the image is
located within the capture area of the Visium slide. Hence, it won’t be
of importance when it comes to analyzing gene expression.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># left plot</span></span>
<span><span class="fu"><a href="../reference/plotImage.html">plotImage</a></span><span class="op">(</span><span class="va">object</span>, outline <span class="op">=</span> <span class="cn">TRUE</span>, line_size <span class="op">=</span> <span class="fl">1</span>, line_color <span class="op">=</span> <span class="va">cvec</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, fragments <span class="op">=</span> <span class="va">cvec</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="../reference/ggpLayerPoints.html">ggpLayerPoints</a></span><span class="op">(</span><span class="va">object</span>, pt_clr <span class="op">=</span> <span class="va">cvec</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, pt_alpha <span class="op">=</span> <span class="fl">0.5</span>, use_scattermore <span class="op">=</span> <span class="cn">T</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="../reference/ggpLayerFrameByCoords.html">ggpLayerFrameByImage</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="../reference/ggpLayerCaptureArea.html">ggpLayerCaptureArea</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span> </span>
<span></span>
<span><span class="co"># right plot</span></span>
<span><span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span><span class="va">object</span>, pt_clr <span class="op">=</span> <span class="va">cvec</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, pt_alpha <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="../reference/ggpLayerCaptureArea.html">ggpLayerCaptureArea</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span></code></pre></div>
<p><img src="initiation-and-preprocessing-visium-hd_files/figure-html/unnamed-chunk-9-1.png" width="50%"><img src="initiation-and-preprocessing-visium-hd_files/figure-html/unnamed-chunk-9-2.png" width="50%"></p>
</div>
<div class="section level2">
<h2 id="spatial-processing">3. Spatial processing<a class="anchor" aria-label="anchor" href="#spatial-processing"></a>
</h2>
<p>This step should not be skipped! Many functions in SPATA2 need to
know where the edge of the tissue section is and they need to know if
there are multiple tissue sections. This kind of data is not provided
with the standard output of most platforms and needs to be computed.
With spatial processing we particularly refer to the identification of
the tissue edge and spatial outliers - observations that are part of the
data set but lie too far away from the contiguous tissue section to be
considered part of the data set that is of actual interest. In case of
the VisiumHD platform they are usually artefacts. The function
<code>identifyTissueOutline(..., method = "obs")</code> uses the <a href="https://en.wikipedia.org/wiki/DBSCAN" class="external-link">DBSCAN</a> algorithm to
identify potential spatial outliers. The tutorial for <a href="initiate-and-preprocessing-visium.html">non-HD Visium</a>
exemplifies that.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># this is the default input for the visium platform and has already been </span></span>
<span><span class="co"># called in initiateSpataObjectVisiumHD(). </span></span>
<span><span class="co"># if the results do not satisfy you, you can run it over and over again with </span></span>
<span><span class="co"># different parameter inputs </span></span>
<span><span class="va">object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/identifyTissueOutline.html">identifyTissueOutline</a></span><span class="op">(</span><span class="va">object</span>, method <span class="op">=</span> <span class="st">"obs"</span>, eps <span class="op">=</span> <span class="st">"20um"</span>, minPts <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span>
<span></span>
<span><span class="va">xrange</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"9mm"</span>, <span class="st">"11mm"</span><span class="op">)</span></span>
<span><span class="va">yrange</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"7mm"</span>, <span class="st">"9mm"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">axes_layer</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ggpLayerAxesSI.html">ggpLayerAxesSI</a></span><span class="op">(</span><span class="va">object</span>, breaks <span class="op">=</span> <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_c.html" class="external-link">str_c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">3</span>, <span class="fl">5</span>, <span class="fl">7</span>, <span class="fl">9</span>, <span class="fl">11</span><span class="op">)</span>, <span class="st">"mm"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">rect_layer</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ggpLayerRect.html">ggpLayerRect</a></span><span class="op">(</span><span class="va">object</span>, xrange <span class="op">=</span> <span class="va">xrange</span>, yrange <span class="op">=</span> <span class="va">yrange</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># left plot</span></span>
<span><span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span><span class="va">object</span>, color_by <span class="op">=</span> <span class="st">"tissue_section"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="va">axes_layer</span> <span class="op">+</span></span>
<span>  <span class="va">rect_layer</span></span>
<span></span>
<span><span class="co"># right plot (zoomed in on rectangular)</span></span>
<span><span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span><span class="va">object</span>, color_by <span class="op">=</span> <span class="st">"tissue_section"</span>, xrange <span class="op">=</span> <span class="va">xrange</span>, yrange <span class="op">=</span> <span class="va">yrange</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="va">axes_layer</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="../reference/legendBottom.html">legendNone</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="initiation-and-preprocessing-visium-hd_files/figure-html/unnamed-chunk-10-1.png" width="50%"><img src="initiation-and-preprocessing-visium-hd_files/figure-html/unnamed-chunk-10-2.png" width="50%"></p>
<p>What has been defined as <em>tissue_section_0</em> could not be
assigned to a tissue section and is always considered a spatial outlier.
Furthermore, observations of tissue sections that are too small are
labeled as spatial outliers, too. What defines <em>too small</em> can be
set with <code>min_section</code> which takes a numeric value defining
the minimal number of observations in order for a tissue section to be
considered large enough. Note that you can also use the results of
<code>identifyTissueOutline(..., method = 'image')</code> for spatial
outlier detection and removal or combine the results of both methods by
setting <code>method = 'image'</code> or
<code>method = c('image', 'obs')</code>. You can play around with these
parameters using <code><a href="../reference/identifySpatialOutliers.html">identifySpatialOutliers()</a></code> over and over
again since the resulting <em>sp_outlier</em> meta variable is simply
overwritten. Manual adjustment of this variable is always possible using
<code><a href="../reference/getMetaDf.html">getMetaDf()</a></code> and
<code>addFeatures(..., overwrite = TRUE)</code>.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># min_section = 1% of all observations</span></span>
<span><span class="va">min_section</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/nObs.html">nObs</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span><span class="op">*</span><span class="fl">0.01</span></span>
<span></span>
<span><span class="va">min_section</span></span>
<span><span class="co">## [1] 1065.74</span></span>
<span></span>
<span><span class="co"># uses the results of identifyTissueOutline() to create a logical variable called sp_outlier</span></span>
<span><span class="va">object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/identifySpatialOutliers.html">identifySpatialOutliers</a></span><span class="op">(</span><span class="va">object</span>, method <span class="op">=</span> <span class="st">"obs"</span>, min_section <span class="op">=</span> <span class="va">min_section</span><span class="op">)</span></span>
<span></span>
<span><span class="va">plot_with_outliers</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span><span class="va">object</span>, color_by <span class="op">=</span> <span class="st">"sp_outlier"</span>, clrp_adjust <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"TRUE"</span> <span class="op">=</span> <span class="st">"blue"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># remove where sp_outlier == TRUE</span></span>
<span><span class="va">object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/removeSpatialOutliers.html">removeSpatialOutliers</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span>
<span></span>
<span><span class="va">plot_without_outliers</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span><span class="va">object</span>, color_by <span class="op">=</span> <span class="st">"sp_outlier"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># left plot</span></span>
<span><span class="va">plot_with_outliers</span></span>
<span></span>
<span><span class="co"># right plot</span></span>
<span><span class="va">plot_without_outliers</span></span></code></pre></div>
<p><img src="initiation-and-preprocessing-visium-hd_files/figure-html/unnamed-chunk-11-1.png" width="50%"><img src="initiation-and-preprocessing-visium-hd_files/figure-html/unnamed-chunk-11-2.png" width="50%"></p>
</div>
<div class="section level2">
<h2 id="data-processing">4. Data processing<a class="anchor" aria-label="anchor" href="#data-processing"></a>
</h2>
<p>These steps are about additional noise removal as well as about
processing raw counts.</p>
<div class="section level3">
<h3 id="data-cleaning">4.1. Data cleaning<a class="anchor" aria-label="anchor" href="#data-cleaning"></a>
</h3>
<p>First you might want to remove certain genes from the raw count
matrix. There are wrappers for certain steps like
<code><a href="../reference/removeMolecules.html">removeGenesStress()</a></code> and
<code><a href="../reference/removeMolecules.html">removeGenesZeroCounts()</a></code>. Individual genes can always be
removed with <code><a href="../reference/removeMolecules.html">removeGenes()</a></code>.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># before</span></span>
<span><span class="fu"><a href="../reference/nMolecules.html">nGenes</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span>
<span><span class="co">## [1] 18085</span></span>
<span></span>
<span><span class="co"># removes stress genes</span></span>
<span><span class="va">object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/removeMolecules.html">removeGenesStress</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># removes genes that were not detected in any of the observations</span></span>
<span><span class="va">object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/removeMolecules.html">removeGenesZeroCounts</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># afterwards</span></span>
<span><span class="fu"><a href="../reference/nMolecules.html">nGenes</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span>
<span><span class="co">## [1] 18044</span></span></code></pre></div>
<p>In some cases there are observations - in case of Visium barcoded
spots - left with no counts at all. If this is the case
<code><a href="../reference/removeObs.html">removeObsZeroCounts()</a></code> removes them. If there are none
nothing happens.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># before</span></span>
<span><span class="fu"><a href="../reference/nObs.html">nObs</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span>
<span><span class="co">## [1] 106360</span></span>
<span></span>
<span><span class="co"># check for and remove observations with zero counts</span></span>
<span><span class="va">object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/removeObs.html">removeObsZeroCounts</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># afterwards </span></span>
<span><span class="fu"><a href="../reference/nObs.html">nObs</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span>
<span><span class="co">## [1] 106360</span></span></code></pre></div>
<p>Afterwards, you can compute meta data for the observations.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/computeMetaFeatures.html">computeMetaFeatures</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># plot left</span></span>
<span><span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span><span class="va">object</span>, color_by <span class="op">=</span> <span class="st">"n_counts_gene"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># plot right</span></span>
<span><span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span><span class="va">object</span>, color_by <span class="op">=</span> <span class="st">"n_distinct_gene"</span><span class="op">)</span></span></code></pre></div>
<p><img src="initiation-and-preprocessing-visium-hd_files/figure-html/unnamed-chunk-14-1.png" width="50%"><img src="initiation-and-preprocessing-visium-hd_files/figure-html/unnamed-chunk-14-2.png" width="50%"></p>
</div>
<div class="section level3">
<h3 id="matrix-processing">4.2 Matrix processing<a class="anchor" aria-label="anchor" href="#matrix-processing"></a>
</h3>
<p>The <code>SPATA2</code> object is initiated with a raw count matrix.
For almost all downstream analysis steps it is recommended to use
processed matrices. The first step is usually log-normalization. To
create a normalized matrix use <code><a href="../reference/normalizeCounts.html">normalizeCounts()</a></code>. It uses
<code><a href="https://satijalab.org/seurat/reference/NormalizeData.html" class="external-link">Seurat::NormalizeData()</a></code> in the background. The input
options for <code>method</code> correspond to the options in this
function from the Seurat package. By default, the normalized matrix is
named after input for <code>method</code>, activated and thus used by
default in downstream analysis and visualization. The function
<code><a href="../reference/normalizeCounts.html">normalizeCounts()</a></code> can be called multiple times with
different inputs for <code>method</code> which populates the list of
processed matrices in the respective assay. Furthermore, processed
matrices can be added with <code><a href="../reference/addProcessedMatrix.html">addProcessedMatrix()</a></code> if you want
to create them with SPATA2-extern functions. The default matrix that is
used can be set with <code><a href="../reference/activateMatrix.html">activateMatrix()</a></code>. By default,
<code><a href="../reference/normalizeCounts.html">normalizeCounts()</a></code> activates the processed matrix it has
created.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># obtain matrix names prior to normalization</span></span>
<span><span class="fu"><a href="../reference/getMatrixNames.html">getMatrixNames</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span>
<span><span class="co">## [1] "counts"</span></span>
<span></span>
<span><span class="va">plot_with_raw_counts</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span><span class="va">object</span>, color_by <span class="op">=</span> <span class="st">"INS"</span>,  pt_clrsp <span class="op">=</span> <span class="st">"Reds 3"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>color <span class="op">=</span> <span class="st">"INS\n(Counts)"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># create log normalized matrix</span></span>
<span><span class="va">object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/normalizeCounts.html">normalizeCounts</a></span><span class="op">(</span><span class="va">object</span>, method <span class="op">=</span> <span class="st">"LogNormalize"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># obtain matrix names after normalization</span></span>
<span><span class="fu"><a href="../reference/getMatrixNames.html">getMatrixNames</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span>
<span><span class="co">## [1] "counts"       "LogNormalize"</span></span>
<span></span>
<span><span class="co"># check active matrix</span></span>
<span><span class="fu"><a href="../reference/activateMatrix.html">activeMatrix</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span>
<span><span class="co">## [1] "LogNormalize"</span></span>
<span></span>
<span><span class="co"># uses the processed matrix LogNormalize</span></span>
<span><span class="co"># use alpha_by to scale transparency to INS as well</span></span>
<span><span class="va">plot_with_proc_data</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span><span class="va">object</span>, color_by <span class="op">=</span> <span class="st">"INS"</span>, alpha_by <span class="op">=</span> <span class="st">"INS"</span>, pt_clrsp <span class="op">=</span> <span class="st">"Reds 3"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>color <span class="op">=</span> <span class="st">"INS\n(logNorm)"</span><span class="op">)</span> </span>
<span></span>
<span><span class="co"># left plot</span></span>
<span><span class="va">plot_with_raw_counts</span></span>
<span></span>
<span><span class="co"># right plot</span></span>
<span><span class="va">plot_with_proc_data</span></span></code></pre></div>
<p><img src="initiation-and-preprocessing-visium-hd_files/figure-html/unnamed-chunk-15-1.png" width="50%"><img src="initiation-and-preprocessing-visium-hd_files/figure-html/unnamed-chunk-15-2.png" width="50%"></p>
</div>
</div>
<div class="section level2">
<h2 id="variable-genes">5. Variable genes<a class="anchor" aria-label="anchor" href="#variable-genes"></a>
</h2>
<p>Genes of high variability can be identified with wrappers around the
some Seurat functions.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># identifies molecules of high variability in the default assay (= gene)</span></span>
<span><span class="va">object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/identifyVariableMolecules.html">identifyVariableMolecules</a></span><span class="op">(</span><span class="va">object</span>, method <span class="op">=</span> <span class="st">"vst"</span>, n_mol <span class="op">=</span> <span class="fl">2500</span><span class="op">)</span></span>
<span></span>
<span><span class="va">var_mols</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getVariableMolecules.html">getVariableMolecules</a></span><span class="op">(</span><span class="va">object</span>, method <span class="op">=</span> <span class="st">"vst"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">var_mols</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  chr [1:2500] "INS" "REG3A" "TTR" "GCG" "GHRL" "SST" "VIP" "PPY" "CARTPT" ...</span></span></code></pre>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># example plots</span></span>
<span><span class="fu"><a href="../reference/plotSurfaceComparison.html">plotSurfaceComparison</a></span><span class="op">(</span><span class="va">object</span>, color_by <span class="op">=</span> <span class="va">var_mols</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span>, pt_clrsp <span class="op">=</span> <span class="st">"Reds 3"</span>,  alpha_by <span class="op">=</span> <span class="cn">TRUE</span>, nrow <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="initiation-and-preprocessing-visium-hd_files/figure-html/unnamed-chunk-16-1.png" width="100%"></p>
</div>
<div class="section level2">
<h2 id="conclusion-and-more-data-sets">6. Conclusion and more data sets<a class="anchor" aria-label="anchor" href="#conclusion-and-more-data-sets"></a>
</h2>
<p>That’s it. The object can be used for any downstream analyses such as
<a href="dimensional-reduction.html">dimensional reduction</a>, <a href="clustering.html">clustering</a>, <a href="spatial-annotation-screening.html">spatial annotation
screening</a> or <a href="spatial-trajectory-screening.html">spatial
trajectory screening</a>. Refer to tab Tutorials for more links.
Furthermore, you can skim our curated data base of spatial data sets for
those of platform VisiumHD using <a href="spata-data.html">SPATAData</a>.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># load package</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SPATAData</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># filter for samples from platform VisiumHD</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SPATAData/man/sourceDataFrame.html" class="external-link">sourceDataFrame</a></span><span class="op">(</span><span class="va">platform</span> <span class="op">==</span> <span class="st">"VisiumHD"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 2 × 18</span></span></span>
<span><span class="co">##   sample_name      donor_species institution lm_source           organ pathology</span></span>
<span><span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;dttm&gt;</span>              <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span> HumanLungCancer… Homo sapiens  10X Genomi… 2024-08-04 <span style="color: #949494;">22:03:33</span> Lung  tumor    </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span> HumanPancreasHD  Homo sapiens  10X Genomi… 2024-08-04 <span style="color: #949494;">22:03:33</span> Panc… <span style="color: #BB0000;">NA</span>       </span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 12 more variables: platform &lt;chr&gt;, source &lt;chr&gt;, web_link &lt;chr&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   mean_counts &lt;dbl&gt;, median_counts &lt;dbl&gt;, modality_gene &lt;lgl&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   modality_metabolite &lt;lgl&gt;, modality_protein &lt;lgl&gt;, n_obs &lt;int&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   n_tissue_sections &lt;int&gt;, obs_unit &lt;chr&gt;, obj_size &lt;lbstr_by&gt;</span></span></span></code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Jan Kueckelhaus, Dieter-Henrik Heiland.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

      </footer>
</div>






  </body>
</html>
