<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Newslog • SPATA2</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Newslog">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SPATA2</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">3.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/installation.html">Installation</a>
</li>
<li>
  <a href="../articles/spata-data.html">SPATAData</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">Initiating a SPATA2 object</li>
    <li>
      <a href="../articles/initiation-and-preprocessing-customized.html">Initiation &amp; Preprocessing with individual data</a>
    </li>
    <li>
      <a href="../articles/initiation-and-preprocessing-merfish.html">Initiation &amp; Preprocessing with MERFISH data</a>
    </li>
    <li>
      <a href="../articles/initiation-and-preprocessing-visium.html">Initiation &amp; Preprocessing with Visium data</a>
    </li>
    <li>
      <a href="../articles/initiation-and-preprocessing-visium-hd.html">Initiation &amp; Preprocessing with VisiumHD data</a>
    </li>
    <li class="dropdown-header">Basic Manipulation of a SPATA2 object</li>
    <li>
      <a href="../articles/adding-data.html">Adding Data to SPATA2 objects</a>
    </li>
    <li>
      <a href="../articles/package-compatibility.html">Platform Compatibility</a>
    </li>
    <li>
      <a href="../articles/molecular-variables.html">Molecular Variables</a>
    </li>
    <li>
      <a href="../articles/image-handling.html">Image Handling</a>
    </li>
    <li>
      <a href="../articles/spatial-segmentation.html">Spatial Segmentation</a>
    </li>
    <li class="dropdown-header">Implemented Algorithms &amp; Concepts</li>
    <li>
      <a href="../articles/cnv.html">Copy Number Variations</a>
    </li>
    <li>
      <a href="../articles/clustering.html">Clustering</a>
    </li>
    <li>
      <a href="../articles/dimensional-reduction.html">Dimensional Reduction</a>
    </li>
    <li>
      <a href="../articles/dea.html">Differential Expression</a>
    </li>
    <li>
      <a href="../articles/gsea.html">Gene Set Enrichment</a>
    </li>
    <li>
      <a href="../articles/spatial-measures.html">Spatial Measures</a>
    </li>
    <li class="dropdown-header">Spatial Gradient Screening</li>
    <li>
      <a href="../articles/creating-spatial-annotations.html">Creating Spatial Annotations</a>
    </li>
    <li>
      <a href="../articles/using-spatial-annotations.html">Using Spatial Annotations</a>
    </li>
    <li>
      <a href="../articles/spatial-annotation-screening.html">Spatial Annotation Screening</a>
    </li>
    <li>
      <a href="../articles/creating-spatial-trajectories.html">Creating Spatial Trajectories</a>
    </li>
    <li>
      <a href="../articles/spatial-trajectory-screening.html">Spatial Trajectory Screening</a>
    </li>
    <li class="dropdown-header">Visualization</li>
    <li>
      <a href="../articles/plotting-surface.html">Surface Plotting</a>
    </li>
  </ul>
</li>
<li>
  <a href="../articles/newslog.html">Newslog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Newslog</h1>
            
      

      <div class="hidden name"><code>newslog.Rmd</code></div>

    </div>

    
    
<p>To update your object to be suitable for the newest version, just
run:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># the output is an object fit for the version of SPATA2 you have installed</span></span>
<span><span class="va">my_object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/updateSpataObject.html">updateSpataObject</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">my_object</span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="spata2-v2-0-4---v3-0-0">SPATA2 v2.0.4 -&gt; v3.0.0<a class="anchor" aria-label="anchor" href="#spata2-v2-0-4---v3-0-0"></a>
</h2>
<p>With the publication of our manuscript <em>Kueckelhaus et
al. 2024</em> we are delighted to release Version 3.0.0 of SPATA2. With
it come several changes and improvements regarding the analysis approach
to gene expression as gradients as well as S4 object structure, function
and arguments naming as well as compatibility with platforms beyond
Visium.</p>
<div class="section level3">
<h3 id="s4-architecture">S4 architecture<a class="anchor" aria-label="anchor" href="#s4-architecture"></a>
</h3>
<p>Similar to Seurat, <code>SPATA2</code> is centered around an S4
object that contains data and tracks progress in preprocessing and
downstream analysis. In version 3.0.0, the object architecture has been
significantly revised, introducing a new S4 class. The previous
<code>spata2</code> class, designed in 2019, has been replaced by the
new <code>SPATA2</code> class. This update includes new and renamed
slots, and the spatial multi-omic study components are now more
elegantly divided into new S4 classes, inspired partly by Seurat’s
structure (e.g., creating an S4 object for each raw count matrix as an
<code>MolecularAssay</code> object).</p>
<p>Understanding the exact architecture of the <code>SPATA2</code>
objects is complex and generally unnecessary. However, detailed
documentation is available for those interested. Each class and its
subclasses are extensively documented. Use the question mark operator
like this <code><a href="../reference/SPATA2.html">?SPATA2</a></code>, <code><a href="../reference/MolecularAssay-class.html">?MolecularAssay</a></code>,
<code><a href="../reference/SpatialData-class.html">?SpatialData</a></code> to get more details, or visit the S4-classes
section on the <a href="reference.html">reference</a> page.</p>
</div>
<div class="section level3">
<h3 id="platform-support">Platform support<a class="anchor" aria-label="anchor" href="#platform-support"></a>
</h3>
<p>SPATA2 v3.0.0 offers enhanced support for platforms like MERFISH,
SlideSeq, Xenium, and more. The new S4 structure and classes SPATA2 uses
provide a clearer framework for creating individual <code>SPATA2</code>
objects with specific spatial data sets. For each platform, we have
created specialized tutorials for initiation and suggested processing
steps. Note that all <code>initiateSpataObject_*()</code> functions from
version 2.0.4 have been deprecated in favor of new initiation functions.
Please refer to the respective tutorials via the dropdown button:
Articles &gt;&gt; Object Initiation &amp; Processing.</p>
</div>
<div class="section level3">
<h3 id="deprecated-functions-renaming-and-new-naming-conventions">Deprecated functions, renaming and new naming conventions<a class="anchor" aria-label="anchor" href="#deprecated-functions-renaming-and-new-naming-conventions"></a>
</h3>
<p><em>“There are only two hard things in Computer Science: cache
invalidation and naming things.”</em> - Phil Carlton</p>
<p>With the improvements made to SPATA2 v3.0.0, multiple aspects and
concepts have been renamed or the naming has been expanded. This has
consequences for the function naming and thus for the code. While we
deprecated multiple functions, we did so in a way that it automatically
calls the new function in the background and prints out a warning that
the function you just used is no longer in use and likely to disappear
in the near future.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co">## the term expression matrix has been abandoned in favor of processed matrix</span></span>
<span><span class="co"># this function throws a warning...</span></span>
<span><span class="fu"><a href="../reference/getExpressionMatrix.html">getExpressionMatrix</a></span><span class="op">(</span><span class="va">object</span>, mtr_name <span class="op">=</span> <span class="st">"LogNormalize"</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span>
<span><span class="co">## Warning in confuns::give_feedback(msg = msg, fdb.fn = fdb_fn, with.time =</span></span>
<span><span class="co">## FALSE): Function `getExpressionMatrix()` is deprecated and will be deleted in</span></span>
<span><span class="co">## the near future. Please use `getProcessedMatrix() or getMatrix()` instead.</span></span>
<span><span class="co">## 5 x 5 sparse Matrix of class "dgCMatrix"</span></span>
<span><span class="co">##            AAACAAGTATCTCCCA-1 AAACAATCTACTAGCA-1 AAACACCAATAACTGC-1</span></span>
<span><span class="co">## AL627309.1                  .                  .                  .</span></span>
<span><span class="co">## AL627309.3                  .                  .                  .</span></span>
<span><span class="co">## AL669831.5                  .                  .                  .</span></span>
<span><span class="co">## FAM87B                      .                  .                  .</span></span>
<span><span class="co">## LINC00115                   .                  .                  .</span></span>
<span><span class="co">##            AAACAGAGCGACTCCT-1 AAACAGCTTTCAGAAG-1</span></span>
<span><span class="co">## AL627309.1                  .                  .</span></span>
<span><span class="co">## AL627309.3                  .                  .</span></span>
<span><span class="co">## AL669831.5                  .                  .</span></span>
<span><span class="co">## FAM87B                      .                  .</span></span>
<span><span class="co">## LINC00115                   .                  .</span></span>
<span></span>
<span><span class="co"># ... and calls this one right after - your code should still work</span></span>
<span><span class="fu"><a href="../reference/getProcessedMatrix.html">getProcessedMatrix</a></span><span class="op">(</span><span class="va">object</span>, mtr_name <span class="op">=</span> <span class="st">"LogNormalize"</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span>
<span><span class="co">## 5 x 5 sparse Matrix of class "dgCMatrix"</span></span>
<span><span class="co">##            AAACAAGTATCTCCCA-1 AAACAATCTACTAGCA-1 AAACACCAATAACTGC-1</span></span>
<span><span class="co">## AL627309.1                  .                  .                  .</span></span>
<span><span class="co">## AL627309.3                  .                  .                  .</span></span>
<span><span class="co">## AL669831.5                  .                  .                  .</span></span>
<span><span class="co">## FAM87B                      .                  .                  .</span></span>
<span><span class="co">## LINC00115                   .                  .                  .</span></span>
<span><span class="co">##            AAACAGAGCGACTCCT-1 AAACAGCTTTCAGAAG-1</span></span>
<span><span class="co">## AL627309.1                  .                  .</span></span>
<span><span class="co">## AL627309.3                  .                  .</span></span>
<span><span class="co">## AL669831.5                  .                  .</span></span>
<span><span class="co">## FAM87B                      .                  .</span></span>
<span><span class="co">## LINC00115                   .                  .</span></span></code></pre></div>
<p>However, multiple changes have been made and there might be code you
wrote in the past that does not work any longer with SPATA2 v3.0.0. We
apologize for that. If the automatically generated warning messages do
not guide you or the function you want to use simply does not exist any
longer, please raise an issue including the code of SPATA2 v2.0.4 and we
will help you to write the code to obtain the same results with SPATA2
v3.0.0.</p>
<div class="section level4">
<h4 id="active-and-inactive">Active and inactive<a class="anchor" aria-label="anchor" href="#active-and-inactive"></a>
</h4>
<p>The new <code>SPATA2</code> object facilitates handling multiple data
modalities. Often, the <code>SPATA2</code> object contains more than one
image (e.g., lowres and hires) or more than one data matrix (e.g. raw
counts, log normalized counts, scaled). Functions that access data must
choose one by default. Whenever there are two or more options, the
function will default to what has been declared as active by the
<code>activate*()</code> functions. The current active state can be
checked with the corresponding <code>active*()</code> functions.</p>
<p>In previous versions, there was no uniform naming convention for
functions dealing with activation. Now, there is. Consequently,
functions like <code>setActive*()</code> or <code>getActive*()</code>
are deprecated.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># names of all registered images</span></span>
<span><span class="fu"><a href="../reference/getImageNames.html">getImageNames</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span>
<span><span class="co">## [1] "lowres" "hires"</span></span>
<span></span>
<span><span class="co"># currently active image is "lowres"</span></span>
<span><span class="fu"><a href="../reference/activateImage.html">activeImage</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span>
<span><span class="co">## [1] "lowres"</span></span>
<span></span>
<span><span class="co"># activate different image</span></span>
<span><span class="va">object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/activateImage.html">activateImage</a></span><span class="op">(</span><span class="va">object</span>, img_name <span class="op">=</span> <span class="st">"hires"</span><span class="op">)</span></span>
<span><span class="co">## 14:22:01 Unloading image lowres.</span></span>
<span><span class="co">## 14:22:01 Loading image hires.</span></span>
<span><span class="co">## 14:22:02 Active image: 'hires'.</span></span>
<span></span>
<span><span class="co"># currently active image has changed to "hires"</span></span>
<span><span class="fu"><a href="../reference/activateImage.html">activeImage</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span>
<span><span class="co">## [1] "hires"</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="spatial-annotation-screening-previously-image-annotation-screening">Spatial Annotation Screening (previously Image Annotation
Screening)<a class="anchor" aria-label="anchor" href="#spatial-annotation-screening-previously-image-annotation-screening"></a>
</h4>
<p>The concept has been expanded to include screening based on
annotations that highlight not only image-related histomorphological
features but also areas with very high (or very low) expression of
numeric features or specific clusters. Image annotations are now part of
the broader term Spatial Annotations, which includes Image Annotations,
Numeric Annotations, and Group Annotations. Each type provides
high-level functions to create necessary spatial reference outlines for
gradient-based analysis. Consequently, <em>Image Annotation Screening
(IAS)</em> has been renamed to <em>Spatial Annotation Screening
(SAS)</em>. This affects every function, changing <code>Ias</code> to
<code>Sas</code>, and <code>binwidth</code> to <code>resolution</code>.
See our new tutorial <a href="creating-spatial-annotations.html">here</a>.</p>
<p>Additionally, the mathematical approach for obtaining gradients has
changed during the review process. This includes how distance to the
spatial outline is integrated, allowing the integration of multiple
annotations of the same kind (see Main Figure 4a-b of <em>Kueckelhaus et
al. 2024</em>). The parameter <code>binwidth</code> has been replaced by
<code>resolution</code>, with the input remaining the same. The term
binwidth no longer fits the approach, hence the change. For a detailed
description of how gradients are obtained and p-values computed, refer
to the methods section and supplementary figures of the manuscript
<em>Kueckelhaus et al. 2024</em>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co">## the algorithm</span></span>
<span><span class="co"># previously (now deprecated)</span></span>
<span><span class="va">ias_out</span> <span class="op">&lt;-</span> <span class="fu">imageAnnotationScreening</span><span class="op">(</span><span class="va">object</span>, ids <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"first_id"</span>, <span class="st">"sec_id"</span><span class="op">)</span>, binwidth <span class="op">=</span> <span class="st">"100um"</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># now</span></span>
<span><span class="va">sas_out</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/spatialAnnotationScreening.html">spatialAnnotationScreening</a></span><span class="op">(</span><span class="va">object</span>, ids <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"first_id"</span>, <span class="st">"sec_id"</span><span class="op">)</span>, resolution <span class="op">=</span> <span class="st">"100um"</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## plotting functions</span></span>
<span><span class="co"># previously (now deprecated)</span></span>
<span><span class="fu">plotIasLineplot</span><span class="op">(</span><span class="va">object</span>, ids  <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"first_id"</span>, <span class="st">"second_id"</span><span class="op">)</span>, variables <span class="op">=</span> <span class="st">"EGFR"</span>, binwidth <span class="op">=</span> <span class="st">"100um"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># now</span></span>
<span><span class="fu"><a href="../reference/plotSasLineplot.html">plotSasLineplot</a></span><span class="op">(</span><span class="va">object</span>, ids <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"first_id"</span>, <span class="st">"second_id"</span><span class="op">)</span>, variables <span class="op">=</span> <span class="st">"EGFR"</span>, resolution <span class="op">=</span> <span class="st">"100um"</span><span class="op">)</span></span></code></pre></div>
<p>Apart from the renaming of <code>binwidth</code> to
<code>resolution</code>, Spatial Trajectory Screening is largely
unaffected by these changes. The methodology for calculating gradients
based on spatial trajectories has been adjusted accordingly, but other
than the <code>binwidth</code> to <code>resolution</code> change, not
much has been altered. Furthermore, the overall concept of inferring
expression gradients related to spatial reference features (Spatial
Trajectories or Spatial Annotations) is now termed Spatial Gradient
Screening.</p>
</div>
<div class="section level4">
<h4 id="molecular-assays">Molecular assays<a class="anchor" aria-label="anchor" href="#molecular-assays"></a>
</h4>
<p>The core of any spatial transcriptomcis data set is a raw integer
matrix of mRNA counts with rows that correspond to gene names and
columns corresponding to observation identifiers - in SPATA2 terms
<em>barcodes</em>. This count matrix can be processed in multiple
different ways and downstream analysis can follow multiple steps. While
gene expression is of paramount importance in spatial omic studies,
spatial analysis can also focus on or integrate metabolomics and
proteomic data. Hence, we expanded the naming and the concept to
molecules to allow future integration of these kinds of data, too. The
assay name corresponds to the data modality of the assay. SPATA2 knows
<em>gene</em>, <em>protein</em> and <em>metabolite</em>.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co">## assume that you have a SPATA2 object from the Visium platform</span></span>
<span><span class="co"># and you vertically integrated protein expression data </span></span>
<span></span>
<span><span class="co"># works as usual</span></span>
<span><span class="va">gene_names</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getMolecules.html">getGenes</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># results in the same output </span></span>
<span><span class="va">gene_names</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getMolecules.html">getMolecules</a></span><span class="op">(</span><span class="va">object</span>, assay_name <span class="op">=</span> <span class="st">"gene"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># hypothetical data set contains also an assay of spatially resolved protein data</span></span>
<span><span class="va">protein_names</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getMolecules.html">getProteins</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span>
<span></span>
<span><span class="va">protein_names</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getMolecules.html">getMolecules</a></span><span class="op">(</span><span class="va">object</span>, assay_name <span class="op">=</span> <span class="st">"protein"</span><span class="op">)</span></span></code></pre></div>
<p>Molecular data matrices used to be stored in a loose list in slot
<code>@data</code>. Within the new <code>SPATA2</code> object, molecular
data is now structured in objects of class <code>MolecularAssay</code>
(partly inspired by the Seurat <code>Assay</code>-class). Among other
things these objects contain a slot called <code>@mtr_counts</code>
which contains the raw count matrix and a slot called
<code>@mtr_proc</code>, a named list of matrices that resulted from
different processing steps. Hence, SPATA2 differentiates between raw
count matrix as obtained via <code><a href="../reference/getCountMatrix.html">getCountMatrix()</a></code> and processed
matrices as obtained via <code><a href="../reference/getProcessedMatrix.html">getProcessedMatrix()</a></code>. In previous
versions of SPATA2, we used the term expression matrix to refer to
processed gene expression matrices. This naming convention has been
abandoned in favor of raw counts vs. processed.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co">## extraction of raw coutns remains the same apart from the new assay_name argument which defaults</span></span>
<span><span class="co"># to the name of the active assay as obtained by activeAssay(object)</span></span>
<span><span class="va">count_mtr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getCountMatrix.html">getCountMatrix</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## extract a normalized matrix</span></span>
<span><span class="co"># previously (now deprecated)</span></span>
<span><span class="va">norm_mtr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getExpressionMatrix.html">getExpressionMatrix</a></span><span class="op">(</span><span class="va">object</span>, mtr_name <span class="op">=</span> <span class="st">"LogNormalize"</span><span class="op">)</span></span>
<span><span class="co">## Warning in confuns::give_feedback(msg = msg, fdb.fn = fdb_fn, with.time =</span></span>
<span><span class="co">## FALSE): Function `getExpressionMatrix()` is deprecated and will be deleted in</span></span>
<span><span class="co">## the near future. Please use `getProcessedMatrix() or getMatrix()` instead.</span></span>
<span></span>
<span><span class="co"># now</span></span>
<span><span class="va">norm_mtr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getProcessedMatrix.html">getProcessedMatrix</a></span><span class="op">(</span><span class="va">object</span>, mtr_name <span class="op">=</span> <span class="st">"LogNormalize"</span><span class="op">)</span></span></code></pre></div>
<p>Analysis results that are directly linked to the molecular data such
as SPARKX, <a href="dea.html">DEA</a>, <a href="gsea.html">GSEA</a>, <a href="cnv.html">CNV</a> etc. are stored in slot <code>@analysis</code>
of the assay. The functions to extract or plot the data remain the same
like <code><a href="../reference/runSPARKX.html">getSparkxGeneDf()</a></code>, <code><a href="../reference/getDeaResultsDf.html">getDeaResultsDf()</a></code>,
etc.</p>
</div>
<div class="section level4">
<h4 id="meta-data-previously-feature-data">Meta data (previously feature data)<a class="anchor" aria-label="anchor" href="#meta-data-previously-feature-data"></a>
</h4>
<p>Previously, we stored meta variables like clustering and quality
control variables in the slot <code>@fdata</code>, short for feature
data, based on the idea of “features that are not directly related to
molecular counts.” However, in other packages like Seurat or Squidpy,
the term <em>feature</em> often refers to numeric variables, including
gene expression, causing understandable confusion. To address this, we
renamed the slot. As shown in the documentation of the new
<code>SPATA2</code> class, the slot is now called
<code>@meta_obs</code>, short for metadata of the SPATA2 object’s
observations (barcoded spots, cells, etc.).</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># previously (now deprecated)</span></span>
<span><span class="fu"><a href="../reference/getFeatureDf.html">getFeatureDf</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span>
<span><span class="co">## Warning in confuns::give_feedback(msg = msg, fdb.fn = fdb_fn, with.time =</span></span>
<span><span class="co">## FALSE): Function `getFeatureDf()` is deprecated and will be deleted in the near</span></span>
<span><span class="co">## future. Please use `getMetaDf()` instead.</span></span>
<span><span class="co">## <span style="color: #949494;"># A tibble: 3,517 × 128</span></span></span>
<span><span class="co">##    barcodes         sample nCount_Spatial nFeature_Spatial percent.mt percent.RB</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> AAACAAGTATCTCCC… T313             <span style="text-decoration: underline;">4</span>319             <span style="text-decoration: underline;">2</span>164      2.17        6.54</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> AAACAATCTACTAGC… T313             <span style="text-decoration: underline;">6</span>005             <span style="text-decoration: underline;">2</span>796      2.10        8.46</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> AAACACCAATAACTG… T313              349              254      4.18        7.74</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> AAACAGAGCGACTCC… T313            <span style="text-decoration: underline;">11</span>394             <span style="text-decoration: underline;">4</span>248      2.85        8.63</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> AAACAGCTTTCAGAA… T313              902              616      1.72        6.33</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> AAACAGGGTCTATAT… T313              175              144      5.49        6.33</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> AAACATGGTGAGAGG… T313             <span style="text-decoration: underline;">8</span>327             <span style="text-decoration: underline;">2</span>958      0.870       8.11</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> AAACCCGAACGAAAT… T313             <span style="text-decoration: underline;">2</span>940             <span style="text-decoration: underline;">1</span>595      2.21        7.21</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> AAACCGGGTAGGTAC… T313               96               83      1.57       10.2 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> AAACCGTTCGTCCAG… T313              671              394      1.99        8.18</span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 3,507 more rows</span></span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 122 more variables: Spatial_snn_res.0.8 &lt;fct&gt;, seurat_clusters &lt;fct&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   AC_like &lt;dbl&gt;, AC_like_Prolif &lt;dbl&gt;, MES_like_hypoxia_independent &lt;dbl&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   MES_like_hypoxia_MHC &lt;dbl&gt;, NPC_like_neural &lt;dbl&gt;, NPC_like_OPC &lt;dbl&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   NPC_like_Prolif &lt;dbl&gt;, OPC_like &lt;dbl&gt;, OPC_like_Prolif &lt;dbl&gt;, cDC1 &lt;dbl&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   cDC2 &lt;dbl&gt;, DC1 &lt;dbl&gt;, DC2 &lt;dbl&gt;, DC3 &lt;dbl&gt;, Mast &lt;dbl&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   Mono_anti_infl &lt;dbl&gt;, Mono_hypoxia &lt;dbl&gt;, Mono_naive &lt;dbl&gt;, pDC &lt;dbl&gt;, …</span></span></span>
<span></span>
<span><span class="co"># now </span></span>
<span><span class="fu"><a href="../reference/getMetaDf.html">getMetaDf</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span>
<span><span class="co">## <span style="color: #949494;"># A tibble: 3,517 × 128</span></span></span>
<span><span class="co">##    barcodes         sample nCount_Spatial nFeature_Spatial percent.mt percent.RB</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> AAACAAGTATCTCCC… T313             <span style="text-decoration: underline;">4</span>319             <span style="text-decoration: underline;">2</span>164      2.17        6.54</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> AAACAATCTACTAGC… T313             <span style="text-decoration: underline;">6</span>005             <span style="text-decoration: underline;">2</span>796      2.10        8.46</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> AAACACCAATAACTG… T313              349              254      4.18        7.74</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> AAACAGAGCGACTCC… T313            <span style="text-decoration: underline;">11</span>394             <span style="text-decoration: underline;">4</span>248      2.85        8.63</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> AAACAGCTTTCAGAA… T313              902              616      1.72        6.33</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> AAACAGGGTCTATAT… T313              175              144      5.49        6.33</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> AAACATGGTGAGAGG… T313             <span style="text-decoration: underline;">8</span>327             <span style="text-decoration: underline;">2</span>958      0.870       8.11</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> AAACCCGAACGAAAT… T313             <span style="text-decoration: underline;">2</span>940             <span style="text-decoration: underline;">1</span>595      2.21        7.21</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> AAACCGGGTAGGTAC… T313               96               83      1.57       10.2 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> AAACCGTTCGTCCAG… T313              671              394      1.99        8.18</span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 3,507 more rows</span></span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 122 more variables: Spatial_snn_res.0.8 &lt;fct&gt;, seurat_clusters &lt;fct&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   AC_like &lt;dbl&gt;, AC_like_Prolif &lt;dbl&gt;, MES_like_hypoxia_independent &lt;dbl&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   MES_like_hypoxia_MHC &lt;dbl&gt;, NPC_like_neural &lt;dbl&gt;, NPC_like_OPC &lt;dbl&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   NPC_like_Prolif &lt;dbl&gt;, OPC_like &lt;dbl&gt;, OPC_like_Prolif &lt;dbl&gt;, cDC1 &lt;dbl&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   cDC2 &lt;dbl&gt;, DC1 &lt;dbl&gt;, DC2 &lt;dbl&gt;, DC3 &lt;dbl&gt;, Mast &lt;dbl&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   Mono_anti_infl &lt;dbl&gt;, Mono_hypoxia &lt;dbl&gt;, Mono_naive &lt;dbl&gt;, pDC &lt;dbl&gt;, …</span></span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="management-of-spatial-data">Management of spatial data<a class="anchor" aria-label="anchor" href="#management-of-spatial-data"></a>
</h3>
<p>SPATA2 focuses on spatial data and integrating multiple spatial
analysis approaches. Consequently, spatial data now has its own slot
<code>@spatial</code>, previously used for random miscellaneous spatial
data, now a clearly structured S4 object called
<code>SpatialData</code>. For more details on the class, call
<code><a href="../reference/SpatialData-class.html">?SpatialData</a></code>. This new architecture improves the management
of spatial aspects and alignment challenges. Here are the most important
aspects:</p>
<div class="section level4">
<h4 id="image">Image<a class="anchor" aria-label="anchor" href="#image"></a>
</h4>
<p>The <code>SPATA2</code> object allows the storage of multiple images
by registering them individually using <code><a href="../reference/registerImage.html">registerImage()</a></code>.
Each registered image creates a container object of class
<code>HistoImage</code>, storing the image or its file directory, along
with additional information and data from image processing steps (e.g.,
tissue outline, alignment).</p>
<p>Working with multiple images, alongside the coordinates of data
points and spatial reference features (SpatialAnnotation,
SpatialTrajectory), requires careful alignment. Alignment involves
matching image resolution and adjusting images for angle, horizontal or
vertical translation, and stretching. This is crucial for images of
neighboring tissue sections that are similar but not perfectly
overlapping.</p>
<div class="section level5">
<h5 id="the-reference-image">The Reference Image<a class="anchor" aria-label="anchor" href="#the-reference-image"></a>
</h5>
<p>To facilitate alignment and integration of multiple images,
coordinates, and spatial reference features, a reference image is
designated. By default, this is the first image loaded into the SPATA2
object. SPATA2 assumes that coordinates and the reference image align
perfectly in terms of vertical and horizontal justification (scaling may
still be needed). Aligning additional images to the reference image
ensures alignment with the data point coordinates. Additionally, the
reference image allows automatic transfer of scale factors to newly
registered images.</p>
</div>
<div class="section level5">
<h5 id="the-active-image">The Active Image<a class="anchor" aria-label="anchor" href="#the-active-image"></a>
</h5>
<p>The active image is the default image used in functions requiring
image input, such as <code><a href="../reference/createImageAnnotations.html">createImageAnnotations()</a></code> or
<code><a href="../reference/plotSpatialAnnotations.html">plotSpatialAnnotations()</a></code>, and for extracting coordinates.
During extraction, coordinates are scaled to the resolution of the
currently active image. The reference image can also serve as the active
image.</p>
</div>
</div>
<div class="section level4">
<h4 id="scaling-coordinates">Scaling coordinates<a class="anchor" aria-label="anchor" href="#scaling-coordinates"></a>
</h4>
<p><code>SPATA2</code> objects allow for the simultaneous storage of
multiple images, which may differ in resolution, requiring different
scale factors to align spatial data coordinates accurately with the
images. In version 3.0.0, coordinate variables are differentiated as
<em>x_orig</em>, <em>y_orig</em> (original coordinates) and <em>x</em>,
<em>y</em> (scaled coordinates).</p>
<p>The rationale: Whenever spatial data (coordinates of barcode spots,
coordinates of cells, tissue outlines, spatial annotations, spatial
trajectories, etc.) are stored, they are recorded as <em>x_orig</em> and
<em>y_orig</em>. Even if you are currently working within the resolution
of a different image the coordinates are scaled back to the “original
resolution” of your reference image. Upon extraction, a scaling process
is applied to any spatial data, such as in
<code><a href="../reference/getSpatAnnOutlineDf.html">getSpatAnnOutlineDf()</a></code> or <code><a href="../reference/getTissueOutlineDf.html">getTissueOutlineDf()</a></code>.
This process ensures that coordinates are appropriately scaled to the
resolution of the active image. For instance, this scaling occurs in the
background when extracting or plotting barcoded spots of a Visium data
set in the context of a hypothetical image, <em>lowres</em>.</p>
<p>First, the original data.frame is obtained. `</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># `as_is` = TRUE, skips any processing steps</span></span>
<span><span class="va">coords_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getCoordsDf.html">getCoordsDf</a></span><span class="op">(</span><span class="va">object</span>, as_is <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, <span class="op">]</span></span>
<span></span>
<span><span class="va">coords_df</span></span>
<span><span class="co">## <span style="color: #949494;"># A tibble: 10 × 7</span></span></span>
<span><span class="co">##    barcodes           x_orig y_orig   row   col section          section_dbscan</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>               <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>         </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> TAACCGTCCAGTTCAT-1   <span style="text-decoration: underline;">4</span>002   <span style="text-decoration: underline;">2</span>342     1    15 tissue_section_1 1             </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> CGCGTGCTATCAACGA-1   <span style="text-decoration: underline;">5</span>189   <span style="text-decoration: underline;">2</span>340     1    27 tissue_section_1 1             </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> TGGTGTGACAGACGAT-1   <span style="text-decoration: underline;">2</span>716   <span style="text-decoration: underline;">2</span>516     2     2 tissue_section_1 1             </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> ATCTATCGATGATCAA-1   <span style="text-decoration: underline;">2</span>815   <span style="text-decoration: underline;">2</span>688     3     3 tissue_section_1 1             </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> CGGTAACAAGATACAT-1   <span style="text-decoration: underline;">2</span>914   <span style="text-decoration: underline;">2</span>516     2     4 tissue_section_1 1             </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> TCGCCGGAGAGTCTTA-1   <span style="text-decoration: underline;">3</span>013   <span style="text-decoration: underline;">2</span>688     3     5 tissue_section_1 1             </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> GGAGGAGTGTGTTTAT-1   <span style="text-decoration: underline;">3</span>112   <span style="text-decoration: underline;">2</span>515     2     6 tissue_section_1 1             </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> TTAGGTGTGACTGGTC-1   <span style="text-decoration: underline;">3</span>211   <span style="text-decoration: underline;">2</span>687     3     7 tissue_section_1 1             </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> CAGGGCTAACGAAACC-1   <span style="text-decoration: underline;">3</span>310   <span style="text-decoration: underline;">2</span>515     2     8 tissue_section_1 1             </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> CCCGTGGGTTAATTGA-1   <span style="text-decoration: underline;">3</span>409   <span style="text-decoration: underline;">2</span>687     3     9 tissue_section_1 1</span></span></code></pre></div>
<p>Second, the image scale factor of image <em>lwores</em> is
extracted.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># extract an image scale factor for the image named "lowres"</span></span>
<span><span class="va">isf</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getScaleFactor.html">getScaleFactor</a></span><span class="op">(</span><span class="va">object</span>, img_name <span class="op">=</span> <span class="st">"lowres"</span>, fct_name <span class="op">=</span> <span class="st">"image"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">isf</span></span>
<span><span class="co">## [1] 0.03460208</span></span></code></pre></div>
<p>Third, the variables are transformed (scaled) to x and y.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># create variables x and y by applying the scale factor to the original coordinates</span></span>
<span><span class="va">coords_df</span><span class="op">$</span><span class="va">x</span> <span class="op">&lt;-</span> <span class="va">coords_df</span><span class="op">$</span><span class="va">x_orig</span> <span class="op">*</span> <span class="va">isf</span></span>
<span><span class="va">coords_df</span><span class="op">$</span><span class="va">y</span> <span class="op">&lt;-</span> <span class="va">coords_df</span><span class="op">$</span><span class="va">y_orig</span> <span class="op">*</span> <span class="va">isf</span></span>
<span></span>
<span><span class="va">coords_df</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"barcodes"</span>, <span class="st">"x_orig"</span>, <span class="st">"y_orig"</span>, <span class="st">"x"</span>, <span class="st">"y"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="co">## <span style="color: #949494;"># A tibble: 10 × 5</span></span></span>
<span><span class="co">##    barcodes           x_orig y_orig     x     y</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>               <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> TAACCGTCCAGTTCAT-1   <span style="text-decoration: underline;">4</span>002   <span style="text-decoration: underline;">2</span>342 138.   81.0</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> CGCGTGCTATCAACGA-1   <span style="text-decoration: underline;">5</span>189   <span style="text-decoration: underline;">2</span>340 180.   81.0</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> TGGTGTGACAGACGAT-1   <span style="text-decoration: underline;">2</span>716   <span style="text-decoration: underline;">2</span>516  94.0  87.1</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> ATCTATCGATGATCAA-1   <span style="text-decoration: underline;">2</span>815   <span style="text-decoration: underline;">2</span>688  97.4  93.0</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> CGGTAACAAGATACAT-1   <span style="text-decoration: underline;">2</span>914   <span style="text-decoration: underline;">2</span>516 101.   87.1</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> TCGCCGGAGAGTCTTA-1   <span style="text-decoration: underline;">3</span>013   <span style="text-decoration: underline;">2</span>688 104.   93.0</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> GGAGGAGTGTGTTTAT-1   <span style="text-decoration: underline;">3</span>112   <span style="text-decoration: underline;">2</span>515 108.   87.0</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> TTAGGTGTGACTGGTC-1   <span style="text-decoration: underline;">3</span>211   <span style="text-decoration: underline;">2</span>687 111.   93.0</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> CAGGGCTAACGAAACC-1   <span style="text-decoration: underline;">3</span>310   <span style="text-decoration: underline;">2</span>515 115.   87.0</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> CCCGTGGGTTAATTGA-1   <span style="text-decoration: underline;">3</span>409   <span style="text-decoration: underline;">2</span>687 118.   93.0</span></span></code></pre></div>
<p>If the object does not contain an image, as with MERFISH or Xenium,
the <code>@images</code> slot of the <code>SpatialData</code> object is
empty. Without images, there is no need for alignment or image scale
factors. In this case, the <em>x</em> and <em>y</em> variables in the
data frame will be equal to <em>x_orig</em> and <em>y_orig</em>. This
might sound redundant or unnecessary. However, to accommodate potential
future options for aligning images of adjacent tissue slides to datasets
without original images, the *_orig coordinate structure is maintained,
even if no images are present during the initial object creation.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Jan Kueckelhaus, Dieter-Henrik Heiland.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

      </footer>
</div>






  </body>
</html>
