<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Single Cell Deconvolution • SPATA2</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Single Cell Deconvolution">
<meta property="og:description" content="SPATA2">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SPATA2</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.0.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/spata-v2-installation.html">Installation</a>
</li>
<li>
  <a href="../articles/spata-v2-spata-data.html">SPATAData</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">Around the spata2 object</li>
    <li>
      <a href="../articles/spata-v2-object-initiation-and-manipulation.html">Object Set Up and Manipulation</a>
    </li>
    <li>
      <a href="../articles/spata-v2-platform-compatibility.html">Platform Compatibility</a>
    </li>
    <li>
      <a href="../articles/spata-v2-autoencoder.html">Autoencoder</a>
    </li>
    <li>
      <a href="../articles/spata-v2-genes-gene-sets-and-features.html">Genes, Gene Sets and Features</a>
    </li>
    <li class="dropdown-header">Image Analysis</li>
    <li>
      <a href="../articles/spata-v2-image-handling.html">Image Handling</a>
    </li>
    <li>
      <a href="../articles/spata-v2-image-annotations.html">Image Annotations</a>
    </li>
    <li>
      <a href="../articles/spata-v2-spatial-segmentation.html">Spatial Segmentation</a>
    </li>
    <li class="dropdown-header">Analysis of Gene Expression</li>
    <li>
      <a href="../articles/spata-v2-cnv-analysis.html">Copy Number Variations</a>
    </li>
    <li>
      <a href="../articles/spata-v2-clustering.html">Clustering</a>
    </li>
    <li>
      <a href="../articles/spata-v2-dea.html">Differential Expression</a>
    </li>
    <li>
      <a href="../articles/spata-v2-gsea.html">Gene Set Enrichment</a>
    </li>
    <li class="dropdown-header">Expression Gradients &amp; Gradient Screening</li>
    <li>
      <a href="../articles/spata-v2-spatial-trajectories.html">Spatial Trajectories</a>
    </li>
    <li>
      <a href="../articles/spata-v2-spatial-trajectory-screening.html">Spatial Trajectory Screening</a>
    </li>
    <li>
      <a href="../articles/spata-v2-image-annotation-screening.html">Image Annotation Screening</a>
    </li>
    <li>
      <a href="../articles/spata-v2-scd.html">Integrate Single Cell Deconvolution</a>
    </li>
    <li class="dropdown-header">Visualization</li>
    <li>
      <a href="../articles/spata-v2-plotting-surface.html">Surface Plotting</a>
    </li>
    <li>
      <a href="../articles/spata-v2-plotting-misc.html">Miscellaneous</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Code &amp; Concept
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/cc-spatial-measures.html">Spatial Measures</a>
    </li>
    <li>
      <a href="../articles/cc-spatial-segmentation-vs-image-annotations.html">Spatial Segmentation vs. Image Annotations</a>
    </li>
    <li>
      <a href="../articles/cc-spatial-trajectory-screening.html">Spatial Trajectory Screening</a>
    </li>
    <li>
      <a href="../articles/cc-image-annotation-screening.html">Image Annotation Screening</a>
    </li>
    <li>
      <a href="../articles/cc-model-fitting-in-spatial-transcriptomic-studies.html">Model Fitting</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Single Cell Deconvolution</h1>
            
      
      
      <div class="hidden name"><code>spata-v2-scd.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="prerequisites">1. Prerequisites<a class="anchor" aria-label="anchor" href="#prerequisites"></a>
</h2>
<p>None.</p>
</div>
<div class="section level2">
<h2 id="introduction">1. Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Multiple algorithms for single cell deconvolution have been published
that allow to map single cell sequencing data sets in the spatial
dimensions of a Visium data set generated from the same tissue.
<code>SPATA2</code> offers functions to visualize that and to integrate
the positioning of cells in gradient screening. The data set used is an
injured mouse cortex slice as has been published by <a href="https://www.biorxiv.org/content/10.1101/2023.02.24.529840v1" class="external-link">Koupourtidou
et al. 2023</a>.</p>
<p>The injuries have been integrated in our analysis using the <a href="spata-v2-image-annotations.html">image annotations system</a>.
They were called <em>inj1</em> and <em>inj2</em>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://themilolab.com/" class="external-link">SPATA2</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span>

<span class="co"># load data set </span>
<span class="va">object_mci</span> <span class="op">&lt;-</span> <span class="fu">downloadloadPubExample</span><span class="op">(</span><span class="st">"MCI_LMU"</span><span class="op">)</span>

<span class="co"># set default and image annotations</span>
<span class="va">object_mci</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/setDefault.html">setDefault</a></span><span class="op">(</span><span class="va">object_mci</span>, clrp <span class="op">=</span> <span class="st">"sifre"</span>, pt_clrp <span class="op">=</span> <span class="st">"sifre"</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"image_annotations"</span><span class="op">)</span>

<span class="va">object_mci</span> <span class="op">&lt;-</span> 
  <span class="fu"><a href="../reference/setImageAnnotation.html">setImageAnnotations</a></span><span class="op">(</span>
    object <span class="op">=</span> <span class="va">object_mci</span>,
    img_anns <span class="op">=</span> <span class="va">image_annotations</span><span class="op">[[</span><span class="st">"MCI_LMU"</span><span class="op">]</span><span class="op">]</span>, 
    overwrite <span class="op">=</span> <span class="cn">TRUE</span>
    <span class="op">)</span>

<span class="va">injury_add_on</span> <span class="op">&lt;-</span> 
  <span class="fu"><a href="../reference/ggpLayerImgAnnOutline.html">ggpLayerImgAnnOutline</a></span><span class="op">(</span><span class="va">object_mci</span>, ids <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"inj1"</span>, <span class="st">"inj2"</span><span class="op">)</span><span class="op">)</span>

<span class="co"># get single cell data set </span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"sc_deconvolution"</span><span class="op">)</span>
<span class="va">sc_mci</span> <span class="op">&lt;-</span> <span class="va">sc_deconvolution</span><span class="op">[[</span><span class="st">"MCI_LMU"</span><span class="op">]</span><span class="op">]</span>

<span class="co"># plot example data</span>
<span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span><span class="va">object_mci</span>, color_by <span class="op">=</span> <span class="st">"clusters"</span><span class="op">)</span> <span class="op">+</span> 
  <span class="va">injury_add_on</span>

<span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span><span class="va">sc_mci</span>, color_by <span class="op">=</span> <span class="st">"cell_type"</span>, pt_clrp <span class="op">=</span> <span class="st">"sifre"</span>, pt_size <span class="op">=</span> <span class="fl">1.5</span><span class="op">)</span> <span class="op">+</span> 
  <span class="va">injury_add_on</span></code></pre></div>
<div class="figure">
<img src="spata-v2-scd_files/figure-html/unnamed-chunk-2-1.png" alt="Fig.1 The injured mouse cortex sample clustered and with deconvolution single cell data." width="50%"><img src="spata-v2-scd_files/figure-html/unnamed-chunk-2-2.png" alt="Fig.1 The injured mouse cortex sample clustered and with deconvolution single cell data." width="50%"><p class="caption">
Fig.1 The injured mouse cortex sample clustered and with deconvolution
single cell data.
</p>
</div>
</div>
<div class="section level2">
<h2 id="single-cell-input">2. Single Cell input<a class="anchor" aria-label="anchor" href="#single-cell-input"></a>
</h2>
<p>Most function that integrate single cell deconvolution take the
single cell input via the argument <code>sc_input</code>. It expects a
data.frame that contains at least the columns <em>x</em> and <em>y</em>
as numeric pixel coordinates and a factor column called
<em>cell_type</em>.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sc_mci</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 4,356 x 7</span></span>
<span class="co">##        x     y cell_type             spot_id              barcodes   x_mm y_mm</span>
<span class="co">##    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>                 <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494;">[mm]</span> <span style="color: #949494;">[mm]</span></span>
<span class="co">## <span style="color: #BCBCBC;"> 1</span>  442.  378. Neurons               AAACAAGTATCTCCCA-1_0 AAACAAGTA~ 5.83 4.98</span>
<span class="co">## <span style="color: #BCBCBC;"> 2</span>  412.  140. Mural cells           AAACAGAGCGACTCCT-1_0 AAACAGAGC~ 5.42 1.85</span>
<span class="co">## <span style="color: #BCBCBC;"> 3</span>  412.  137. Astrocytes            AAACAGAGCGACTCCT-1_1 AAACAGAGC~ 5.43 1.81</span>
<span class="co">## <span style="color: #BCBCBC;"> 4</span>  413.  142. Astrocytes            AAACAGAGCGACTCCT-1_2 AAACAGAGC~ 5.44 1.87</span>
<span class="co">## <span style="color: #BCBCBC;"> 5</span>  412.  138. Microglia             AAACAGAGCGACTCCT-1_3 AAACAGAGC~ 5.43 1.82</span>
<span class="co">## <span style="color: #BCBCBC;"> 6</span>  424.  447. Monocytes/Macrophages AAACATTTCCCGGATT-1_0 AAACATTTC~ 5.58 5.89</span>
<span class="co">## <span style="color: #BCBCBC;"> 7</span>  425.  452. Microglia             AAACATTTCCCGGATT-1_1 AAACATTTC~ 5.60 5.96</span>
<span class="co">## <span style="color: #BCBCBC;"> 8</span>  426.  450. Microglia             AAACATTTCCCGGATT-1_2 AAACATTTC~ 5.62 5.93</span>
<span class="co">## <span style="color: #BCBCBC;"> 9</span>  492.  341. OPCs                  AAACCCGAACGAAATC-1_0 AAACCCGAA~ 6.48 4.50</span>
<span class="co">## <span style="color: #BCBCBC;">10</span>  494.  346. OPCs                  AAACCCGAACGAAATC-1_1 AAACCCGAA~ 6.51 4.56</span>
<span class="co">## <span style="color: #949494;"># i 4,346 more rows</span></span></code></pre>
<p>Further columns are optional. For this vignette the three mentioned
above suffice.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cell_types</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Astrocytes"</span>, <span class="st">"Monocytes/Macrophages"</span>, <span class="st">"Neurons"</span>, <span class="st">"Microglia"</span><span class="op">)</span>

<span class="va">sc_mci</span> <span class="op">&lt;-</span> 
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">sc_mci</span>, <span class="va">cell_type</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="op">{</span><span class="op">{</span><span class="va">cell_types</span><span class="op">}</span><span class="op">}</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> 
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>cell_type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/droplevels.html" class="external-link">droplevels</a></span><span class="op">(</span><span class="va">cell_type</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> 
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span>, <span class="va">cell_type</span><span class="op">)</span>

<span class="va">sc_mci</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 3,221 x 3</span></span>
<span class="co">##        x     y cell_type            </span>
<span class="co">##    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>                </span>
<span class="co">## <span style="color: #BCBCBC;"> 1</span>  442.  378. Neurons              </span>
<span class="co">## <span style="color: #BCBCBC;"> 2</span>  412.  137. Astrocytes           </span>
<span class="co">## <span style="color: #BCBCBC;"> 3</span>  413.  142. Astrocytes           </span>
<span class="co">## <span style="color: #BCBCBC;"> 4</span>  412.  138. Microglia            </span>
<span class="co">## <span style="color: #BCBCBC;"> 5</span>  424.  447. Monocytes/Macrophages</span>
<span class="co">## <span style="color: #BCBCBC;"> 6</span>  425.  452. Microglia            </span>
<span class="co">## <span style="color: #BCBCBC;"> 7</span>  426.  450. Microglia            </span>
<span class="co">## <span style="color: #BCBCBC;"> 8</span>  215.  391. Monocytes/Macrophages</span>
<span class="co">## <span style="color: #BCBCBC;"> 9</span>  215.  390. Astrocytes           </span>
<span class="co">## <span style="color: #BCBCBC;">10</span>  215.  393. Astrocytes           </span>
<span class="co">## <span style="color: #949494;"># i 3,211 more rows</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="visualization">3. Visualization<a class="anchor" aria-label="anchor" href="#visualization"></a>
</h2>
<p>To visualize the cells in space use <code><a href="../reference/plotSurfaceSC.html">plotSurfaceSC()</a></code>.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">injury_add_on_thin</span> <span class="op">&lt;-</span> 
  <span class="fu"><a href="../reference/ggpLayerImgAnnOutline.html">ggpLayerImgAnnOutline</a></span><span class="op">(</span><span class="va">object_mci</span>, ids <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"inj1"</span>, <span class="st">"inj2"</span><span class="op">)</span>, line_size <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span>

<span class="fu"><a href="../reference/plotSurfaceSC.html">plotSurfaceSC</a></span><span class="op">(</span>
  object <span class="op">=</span> <span class="va">object_mci</span>, 
  sc_input <span class="op">=</span> <span class="va">sc_mci</span>, 
  cell_types <span class="op">=</span> <span class="va">cell_types</span>, 
  display_image <span class="op">=</span> <span class="cn">TRUE</span>, 
  display_density <span class="op">=</span> <span class="cn">FALSE</span>,
  display_facets <span class="op">=</span> <span class="cn">TRUE</span>, 
  pt_size <span class="op">=</span> <span class="fl">0.5</span>, 
<span class="op">)</span> <span class="op">+</span> 
  <span class="va">injury_add_on_thin</span> <span class="op">+</span> 
  <span class="fu"><a href="../reference/legendBottom.html">legendNone</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<div class="figure">
<img src="spata-v2-scd_files/figure-html/unnamed-chunk-5-1.png" alt="Fig.2 Single cells of different cell types plotted on histology." width="100%"><p class="caption">
Fig.2 Single cells of different cell types plotted on histology.
</p>
</div>
<p>2d density plots might be more insightful.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotSurfaceSC.html">plotSurfaceSC</a></span><span class="op">(</span>
  object <span class="op">=</span> <span class="va">object_mci</span>, 
  sc_input <span class="op">=</span> <span class="va">sc_mci</span>, 
  cell_types <span class="op">=</span> <span class="va">cell_types</span>, 
  display_image <span class="op">=</span> <span class="cn">TRUE</span>, 
  display_density <span class="op">=</span> <span class="cn">TRUE</span>,
  display_points <span class="op">=</span> <span class="cn">FALSE</span>,
  display_facets <span class="op">=</span> <span class="cn">TRUE</span>, 
  line_size <span class="op">=</span> <span class="fl">0.5</span>
<span class="op">)</span> <span class="op">+</span> 
  <span class="va">injury_add_on_thin</span> <span class="op">+</span> 
  <span class="fu"><a href="../reference/legendBottom.html">legendNone</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<div class="figure">
<img src="spata-v2-scd_files/figure-html/unnamed-chunk-6-1.png" alt="Fig.3 2D kernel density estimates of single cells on the histology." width="100%"><p class="caption">
Fig.3 2D kernel density estimates of single cells on the histology.
</p>
</div>
<p>The density of both, microglia and monocytes, seems to decrease
rapidly with the distance to the image annotations. The contrary seems
to apply to neurons. The density of astrocytes seems not to be
significantly affected.</p>
</div>
<div class="section level2">
<h2 id="spatial-relation-to-image-annotations">4. Spatial relation to image annotations<a class="anchor" aria-label="anchor" href="#spatial-relation-to-image-annotations"></a>
</h2>
<p>The density of specific cells might stand in relation to the distance
to specific areas in the tissue. Figure 3 indicates high density of
microglia and monocytes around the injury zones. As we have shown in our
vignettes about <a href="spata-v2-image-annotation-screening.html">gene
expression and image annotations</a> one can plot the abundance of X
against the distance to image annotations, thus specific histological
microstructures.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotImageGgplot.html">plotImageGgplot</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">object_mci</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="../reference/ggpLayerEncirclingIAS.html">ggpLayerEncirclingIAS</a></span><span class="op">(</span>
    object <span class="op">=</span> <span class="va">object_mci</span>, 
    id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"inj1"</span>, <span class="st">"inj2"</span><span class="op">)</span>, 
    distance <span class="op">=</span> <span class="st">"1mm"</span>
  <span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="../reference/ggpLayerScaleBarSI.html">ggpLayerScaleBarSI</a></span><span class="op">(</span>
    object <span class="op">=</span> <span class="va">object_mci</span>, 
    sb_dist <span class="op">=</span> <span class="st">"1mm"</span>, 
    sb_pos <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"1.5mm"</span>, <span class="st">"6.75mm"</span><span class="op">)</span>, 
    sgmt_size <span class="op">=</span> <span class="fl">0.75</span>, 
    text_size <span class="op">=</span> <span class="fl">7.5</span>
  <span class="op">)</span> </code></pre></div>
<div class="figure">
<img src="spata-v2-scd_files/figure-html/unnamed-chunk-7-1.png" alt="Fig.4 Visualization of the conducted screening of the area around the injury annotations." width="100%"><p class="caption">
Fig.4 Visualization of the conducted screening of the area around the
injury annotations.
</p>
</div>
<p>Let X be the density of cells.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotIasRidgeplot.html">plotIasRidgeplotSC</a></span><span class="op">(</span>
  object <span class="op">=</span> <span class="va">object_mci</span>, 
  id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"inj1"</span>, <span class="st">"inj2"</span><span class="op">)</span>, <span class="co"># infer for both and then take the average</span>
  distance <span class="op">=</span> <span class="st">"1mm"</span>,
  sc_input <span class="op">=</span> <span class="va">sc_mci</span>, 
  free_y <span class="op">=</span> <span class="cn">TRUE</span>,
  display_border <span class="op">=</span> <span class="cn">TRUE</span>,
  include_area <span class="op">=</span> <span class="cn">TRUE</span> <span class="co"># include the area of the injury annotations</span>
<span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>subtitle <span class="op">=</span> <span class="st">"a) Free y-axis"</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="../reference/legendBottom.html">legendNone</a></span><span class="op">(</span><span class="op">)</span>

<span class="fu"><a href="../reference/plotIasRidgeplot.html">plotIasRidgeplotSC</a></span><span class="op">(</span>
  object <span class="op">=</span> <span class="va">object_mci</span>, 
  id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"inj1"</span>, <span class="st">"inj2"</span><span class="op">)</span>, 
  distance <span class="op">=</span> <span class="st">"1mm"</span>,
  sc_input <span class="op">=</span> <span class="va">sc_mci</span>,
  free_y <span class="op">=</span> <span class="cn">FALSE</span>, 
  display_border <span class="op">=</span> <span class="cn">TRUE</span>,
  include_area <span class="op">=</span> <span class="cn">TRUE</span> 
<span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>subtitle <span class="op">=</span> <span class="st">"b) Fixed y-axis"</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="../reference/legendBottom.html">legendNone</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<div class="figure">
<img src="spata-v2-scd_files/figure-html/unnamed-chunk-8-1.png" alt="Fig.4 Inferred density of cell types as a function of distance to the injury annotations." width="100%"><img src="spata-v2-scd_files/figure-html/unnamed-chunk-8-2.png" alt="Fig.4 Inferred density of cell types as a function of distance to the injury annotations." width="100%"><p class="caption">
Fig.4 Inferred density of cell types as a function of distance to the
injury annotations.
</p>
</div>
<p>Use the function <code><a href="../reference/inferSingleCellGradient.html">inferSingleCellGradient()</a></code> to obtain the
results in a data.frame.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/inferSingleCellGradient.html">inferSingleCellGradient</a></span><span class="op">(</span>
  object <span class="op">=</span> <span class="va">object_mci</span>, 
  sc_input <span class="op">=</span> <span class="va">sc_mci</span>, 
  id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"inj1"</span>, <span class="st">"inj2"</span><span class="op">)</span>, 
  distance <span class="op">=</span> <span class="st">"1mm"</span>,
  calculate <span class="op">=</span> <span class="st">"density"</span>, <span class="co"># compute density </span>
  area_unit <span class="op">=</span> <span class="st">"mm2"</span>, <span class="co"># with unit count/mm2</span>
  normalize <span class="op">=</span> <span class="cn">FALSE</span> <span class="co"># do not rescale to 0-1</span>
<span class="op">)</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 11 x 7</span></span>
<span class="co">##    bins_circle bins_order bins_angle Neurons Astrocytes Microglia</span>
<span class="co">##    <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="co">## <span style="color: #BCBCBC;"> 1</span> Core                 0 (0,360]        0        114.      240. </span>
<span class="co">## <span style="color: #BCBCBC;"> 2</span> Circle 1             1 (0,360]       42.1       87.9     202. </span>
<span class="co">## <span style="color: #BCBCBC;"> 3</span> Circle 2             2 (0,360]       26.9       83.7     105. </span>
<span class="co">## <span style="color: #BCBCBC;"> 4</span> Circle 3             3 (0,360]       47.2       97.4      57.4</span>
<span class="co">## <span style="color: #BCBCBC;"> 5</span> Circle 4             4 (0,360]       48.1      108.       73.7</span>
<span class="co">## <span style="color: #BCBCBC;"> 6</span> Circle 5             5 (0,360]       35.3       84.2      74.3</span>
<span class="co">## <span style="color: #BCBCBC;"> 7</span> Circle 6             6 (0,360]       52.1      108.       47.3</span>
<span class="co">## <span style="color: #BCBCBC;"> 8</span> Circle 7             7 (0,360]       58.2      119.       45.5</span>
<span class="co">## <span style="color: #BCBCBC;"> 9</span> Circle 8             8 (0,360]       49.8       87.0      67.8</span>
<span class="co">## <span style="color: #BCBCBC;">10</span> Circle 9             9 (0,360]       56.4       79.5      40.8</span>
<span class="co">## <span style="color: #BCBCBC;">11</span> Circle 10           10 (0,360]       42.0       64.4      52.4</span>
<span class="co">## <span style="color: #949494;"># i 1 more variable: `Monocytes/Macrophages` &lt;dbl&gt;</span></span></code></pre>
<p>The columns named by the cell type contain the density values
calculated for each distance/circular bin in the unit specified via
<code>area_unit</code>. Use the argument <code>as_models = TRUE</code>
to obtain the results as valid input for <code>add_models</code> in <a href="spata-v2-image-annotation-screening.html">image annotation
screening</a>.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/inferSingleCellGradient.html">inferSingleCellGradient</a></span><span class="op">(</span>
  object <span class="op">=</span> <span class="va">object_mci</span>, 
  sc_input <span class="op">=</span> <span class="va">sc_mci</span>, 
  id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"inj1"</span>, <span class="st">"inj2"</span><span class="op">)</span>, 
  distance <span class="op">=</span> <span class="st">"1mm"</span>,
  calculate <span class="op">=</span> <span class="st">"density"</span>,
  as_models <span class="op">=</span> <span class="cn">TRUE</span>
<span class="op">)</span></code></pre></div>
<pre><code><span class="co">## $Neurons</span>
<span class="co">##  [1] 0.0000000 0.7229662 0.4624053 0.8106138 0.8261313 0.6065198 0.8960836</span>
<span class="co">##  [8] 1.0000000 0.8564700 0.9695214 0.7225793</span>
<span class="co">## </span>
<span class="co">## $Astrocytes</span>
<span class="co">##  [1] 0.9022810 0.4304072 0.3522829 0.6027643 0.7881667 0.3631900 0.7889645</span>
<span class="co">##  [8] 1.0000000 0.4141949 0.2756438 0.0000000</span>
<span class="co">## </span>
<span class="co">## $Microglia</span>
<span class="co">##  [1] 1.00000000 0.80944945 0.32133099 0.08314482 0.16503405 0.16838363</span>
<span class="co">##  [7] 0.03270634 0.02345116 0.13558303 0.00000000 0.05814409</span>
<span class="co">## </span>
<span class="co">## $`Monocytes/Macrophages`</span>
<span class="co">##  [1] 1.00000000 0.45156204 0.11913118 0.09565253 0.11272233 0.08934293</span>
<span class="co">##  [7] 0.05290540 0.03235082 0.00000000 0.07798454 0.03589546</span></code></pre>
</div>
<div class="section level2">
<h2 id="accounting-for-multiple-sections">5. Accounting for multiple sections<a class="anchor" aria-label="anchor" href="#accounting-for-multiple-sections"></a>
</h2>
<p>Note that this sample contains two sections that, albeit being from
the same organ, are independent of each other. Relating cells from one
section to the image annotation of another section does not make any
sense. The function <code><a href="../reference/add_dbscan_variable.html">add_tissue_section_variable()</a></code> and
<code><a href="../reference/include_tissue_outline.html">include_tissue_outline()</a></code> account for that as is exemplified
in this visualization with a distance set to 6mm which would go beyond
the capture frame of the Visium slide.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># visualizes the result of include_tissue_outline()</span>
<span class="va">tissue_outline</span> <span class="op">&lt;-</span> 
  <span class="fu"><a href="../reference/ggpLayerTissueOutline.html">ggpLayerTissueOutline</a></span><span class="op">(</span><span class="va">object_mci</span>, line_color <span class="op">=</span> <span class="st">"black"</span>, line_size <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>

<span class="co"># uses include_tissue_outline()</span>
<span class="va">many_expansions</span> <span class="op">&lt;-</span> 
  <span class="fu"><a href="../reference/ggpLayerEncirclingIAS.html">ggpLayerEncirclingIAS</a></span><span class="op">(</span>
    object <span class="op">=</span> <span class="va">object_mci</span>,
    id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"inj1"</span>, <span class="st">"inj2"</span><span class="op">)</span>, 
    distance <span class="op">=</span> <span class="st">"6mm"</span>, 
    line_size <span class="op">=</span> <span class="fl">1</span>
    <span class="op">)</span>

<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="../reference/ggpLayerImage.html">ggpLayerImage</a></span><span class="op">(</span><span class="va">object_mci</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_equal</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="va">tissue_outline</span> <span class="op">+</span> 
  <span class="fu"><a href="../reference/ggpLayerThemeCoords.html">ggpLayerThemeCoords</a></span><span class="op">(</span><span class="op">)</span>

<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="../reference/ggpLayerImage.html">ggpLayerImage</a></span><span class="op">(</span><span class="va">object_mci</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_equal</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="va">tissue_outline</span> <span class="op">+</span>
  <span class="va">many_expansions</span> <span class="op">+</span> 
  <span class="fu"><a href="../reference/ggpLayerThemeCoords.html">ggpLayerThemeCoords</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<div class="figure">
<img src="spata-v2-scd_files/figure-html/unnamed-chunk-11-1.png" alt="Fig.6 Visualization of how SPATA2 considers the outline of actual tissue sections within algorithms." width="50%"><img src="spata-v2-scd_files/figure-html/unnamed-chunk-11-2.png" alt="Fig.6 Visualization of how SPATA2 considers the outline of actual tissue sections within algorithms." width="50%"><p class="caption">
Fig.6 Visualization of how SPATA2 considers the outline of actual tissue
sections within algorithms.
</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Jan Kueckelhaus, Dieter-Henrik Heiland, Simon Frerich.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
