<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Autoencoder Denoising • SPATA2</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Autoencoder Denoising">
<meta property="og:description" content="SPATA2">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SPATA2</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/spata-v2-installation.html">Installation</a>
</li>
<li>
  <a href="../articles/spata-v2-spata-data.html">SPATAData</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">Around the spata2 object</li>
    <li>
      <a href="../articles/spata-v2-object-initiation-and-manipulation.html">Object Set Up and Manipulation</a>
    </li>
    <li>
      <a href="../articles/spata-v2-platform-compatibility.html">Platform Compatibility</a>
    </li>
    <li>
      <a href="../articles/spata-v2-autoencoder.html">Autoencoder</a>
    </li>
    <li>
      <a href="../articles/spata-v2-genes-gene-sets-and-features.html">Genes, Gene Sets and Features</a>
    </li>
    <li class="dropdown-header">Image Analysis</li>
    <li>
      <a href="../articles/spata-v2-image-handling.html">Image Handling</a>
    </li>
    <li>
      <a href="../articles/spata-v2-image-annotations.html">Image Annotations</a>
    </li>
    <li>
      <a href="../articles/spata-v2-spatial-segmentation.html">Spatial Segmentation</a>
    </li>
    <li class="dropdown-header">Analysis of Gene Expression</li>
    <li>
      <a href="../articles/spata-v2-cnv-analysis.html">Copy Number Variations</a>
    </li>
    <li>
      <a href="../articles/spata-v2-clustering.html">Clustering</a>
    </li>
    <li>
      <a href="../articles/spata-v2-dea.html">Differential Expression</a>
    </li>
    <li>
      <a href="../articles/spata-v2-gsea.html">Gene Set Enrichment</a>
    </li>
    <li class="dropdown-header">Expression Gradients &amp; Gradient Screening</li>
    <li>
      <a href="../articles/spata-v2-spatial-trajectories.html">Spatial Trajectories</a>
    </li>
    <li>
      <a href="../articles/spata-v2-spatial-trajectory-screening.html">Spatial Trajectory Screening</a>
    </li>
    <li>
      <a href="../articles/spata-v2-image-annotation-screening.html">Image Annotation Screening</a>
    </li>
    <li>
      <a href="../articles/spata-v2-scd.html">Integrate Single Cell Deconvolution</a>
    </li>
    <li class="dropdown-header">Visualization</li>
    <li>
      <a href="../articles/spata-v2-plotting-surface.html">Surface Plotting</a>
    </li>
    <li>
      <a href="../articles/spata-v2-plotting-misc.html">Miscellaneous</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Code &amp; Concept
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/cc-spatial-measures.html">Spatial Measures</a>
    </li>
    <li>
      <a href="../articles/cc-spatial-segmentation-vs-image-annotations.html">Spatial Segmentation vs. Image Annotations</a>
    </li>
    <li>
      <a href="../articles/cc-spatial-trajectory-screening.html">Spatial Trajectory Screening</a>
    </li>
    <li>
      <a href="../articles/cc-image-annotation-screening.html">Image Annotation Screening</a>
    </li>
    <li>
      <a href="../articles/cc-model-fitting-in-spatial-transcriptomic-studies.html">Model Fitting</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Autoencoder Denoising</h1>
            
      
      
      <div class="hidden name"><code>spata-v2-autoencoder.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="introduction">1. Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Sequencing data from single-cell RNA sequencing (scRNA-seq) and from
spatial transcriptomic experiments alike are prone to noise and
technical artifacts that might obstruct downstram analysis. As recently
proposed by <a href="https://www.nature.com/articles/s41467-018-07931-2" class="external-link">Gökcen et al.,
2019</a> autoencoder networks work against that. <strong>SPATA2</strong>
offers a similar approach to denoise data. Apart from data that makes
more sense (see Figure 3.2 and 3.3) denoising your data often results in
more insightful visualization.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://themilolab.com/" class="external-link">SPATA2</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SPATAData</span><span class="op">)</span>

<span class="va">spata_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/downloadSpataObject.html">downloadSpataObject</a></span><span class="op">(</span><span class="st">"275_T"</span><span class="op">)</span></code></pre></div>
<p>Note: Before you use the functions <strong>SPATA2</strong> provides
please make sure that the R-packages <code>keras</code> and
<code>tensorflow</code> are installed on your device. Installing both
packages requires some steps in addition to what R does by default.
Therefore the installation of <strong>SPATA2</strong> does not include
these two packages. Click <a href="https://keras.rstudio.com/reference/install_keras.html" class="external-link">here</a>
for a guide on how to setup both packages.</p>
</div>
<div class="section level2">
<h2 id="denoising-your-data">2. Denoising your data<a class="anchor" aria-label="anchor" href="#denoising-your-data"></a>
</h2>
<p>Although there es only one function needed to denoise your data,
namely <code><a href="../reference/runAutoencoderDenoising.html">runAutoencoderDenoising()</a></code>, you might want to assess
how you set up the neural network that does the denoising - see chapter
2.2 for that matter.</p>
<div class="section level3">
<h3 id="using-the-neural-network">2.1 Using the neural network<a class="anchor" aria-label="anchor" href="#using-the-neural-network"></a>
</h3>
<p>Denoising your data in <strong>SPATA2</strong> is as simple as it
gets. The function <code>runAutoenconderDenoising()</code> constructs
the neural network according to your adjustments (see Chapter 2.2 on how
to optimize that) and stores the resulting denoised expression matrix
named <em>denoised</em> next to the already existing expression matrix
named <em>scaled</em> by using the function
<code><a href="../reference/addExpressionMatrix.html">addExpressionMatrix()</a></code>.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># all expression matrices before denoising</span>
<span class="fu"><a href="../reference/getExpressionMatrixNames.html">getExpressionMatrixNames</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">spata_obj</span><span class="op">)</span>

<span class="co"># active expression matrix before denoising</span>
<span class="fu"><a href="../reference/getActiveMatrixName.html">getActiveMatrixName</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">spata_obj</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] "scaled"</span></code></pre>
<pre><code><span class="co">## [1] "scaled"</span></code></pre>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># denoising your data </span>
<span class="va">spata_obj</span> <span class="op">&lt;-</span>
  <span class="fu"><a href="../reference/runAutoencoderDenoising.html">runAutoencoderDenoising</a></span><span class="op">(</span>
    object <span class="op">=</span> <span class="va">spata_obj</span>, 
    activation <span class="op">=</span> <span class="st">"selu"</span>, 
    bottleneck <span class="op">=</span> <span class="fl">56</span>, 
    epochs <span class="op">=</span> <span class="fl">20</span>, 
    layers <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">128</span>, <span class="fl">64</span>, <span class="fl">32</span><span class="op">)</span>, 
    dropout <span class="op">=</span> <span class="fl">0.1</span>
    <span class="op">)</span>

<span class="co"># all expression matrices after denoising</span>
<span class="fu"><a href="../reference/getExpressionMatrixNames.html">getExpressionMatrixNames</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">spata_obj</span><span class="op">)</span>

<span class="co"># active expression matrix after denoising</span>
<span class="fu"><a href="../reference/getActiveMatrixName.html">getActiveMatrixName</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">spata_obj</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] "scaled"   "denoised"</span></code></pre>
<pre><code><span class="co">## [1] "denoised"</span></code></pre>
<p>Use <code><a href="../reference/printAutoencoderSummary.html">printAutoencoderSummary()</a></code> if you want to recall the
way you set up the neural network that denoised your scaled data.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># print a summary </span>
<span class="fu"><a href="../reference/printAutoencoderSummary.html">printAutoencoderSummary</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">spata_obj</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## The neural network that generated matrix 'denoised' was constructed with the following adjustments: </span>
<span class="co">## </span>
<span class="co">## Activation function: selu</span>
<span class="co">## Bottleneck neurons: 56</span>
<span class="co">## Dropout: 0.1</span>
<span class="co">## Epochs: 20</span>
<span class="co">## Layers: 128, 64 and 32</span></code></pre>
<p>By default all functions in <strong>SPATA2</strong> (like
<code><a href="../reference/plotSurface.html">plotSurface()</a></code> for example) use the expression matrix that
is set to the active expression matrix. Use
<code><a href="../reference/setActiveExpressionMatrix.html">setActiveMatrix()</a></code> to switch between the expression matrices
that are currently stored inside your spata-object.</p>
</div>
<div class="section level3">
<h3 id="the-optimal-neural-network-set-up">2.2 The optimal neural network set up<a class="anchor" aria-label="anchor" href="#the-optimal-neural-network-set-up"></a>
</h3>
<p>How neural networks are set up and how they work in detail is beyond
the scope of this tutorial. Among other things they differ in the number
of <em>bottleneck neurons</em> as well as in the <em>activation
function</em> they use. With <code><a href="../reference/runAutoencoderAssessment.html">runAutoencoderAssessment()</a></code> you
can iterate over all combinations you are interested in and assess the
results by the total variance of the principal component analysis
results.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">spata_obj</span> <span class="op">&lt;-</span>
  <span class="fu"><a href="../reference/runAutoencoderAssessment.html">runAutoencoderAssessment</a></span><span class="op">(</span>
    object <span class="op">=</span> <span class="va">spata_obj</span>,
    activations <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"relu"</span>, <span class="st">"selu"</span>, <span class="st">"sigmoid"</span><span class="op">)</span>, 
    bottlenecks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">32</span>, <span class="fl">40</span>, <span class="fl">48</span>, <span class="fl">56</span>, <span class="fl">64</span><span class="op">)</span>,
    epochs <span class="op">=</span> <span class="fl">20</span>, 
    layers <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">128</span>, <span class="fl">64</span>, <span class="fl">32</span><span class="op">)</span>, 
    dropout <span class="op">=</span> <span class="fl">0.1</span>
    <span class="op">)</span>

<span class="fu"><a href="../reference/plotAutoencoderAssessment.html">plotAutoencoderAssessment</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">spata_obj</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Additional set up of neural network: </span>
<span class="co">## </span>
<span class="co">## Epochs: 20</span>
<span class="co">## Dropout: 0.1</span>
<span class="co">## Layers: 128, 64 and 32</span></code></pre>
<div class="figure">
<img src="spata-v2-autoencoder_files/figure-html/runAutoencoderAssessment()-invis-1.png" alt="Figure 2.1 Visualization of total variance of different autoencoder set ups" width="100%"><p class="caption">
Figure 2.1 Visualization of total variance of different autoencoder set
ups
</p>
</div>
<p>Sticking to the set up that resulted in the highest total variance is
a good start. If you want to extract the assessment results use
<code><a href="../reference/getAutoencoderAssessment.html">getAutoencoderAssessment()</a></code>. It returns a list of the total
variance results as well as the components the neural networks were
composed of.</p>
</div>
<div class="section level3">
<h3 id="storing-several-expression-matrices">3.3 Storing several expression matrices<a class="anchor" aria-label="anchor" href="#storing-several-expression-matrices"></a>
</h3>
<p>You can store as many denoised expression matrices as you want in
your spata-object. Use the argument <code>mtr_name_output</code> for
that matter. This makes <code><a href="../reference/runAutoencoderDenoising.html">runAutoencoderDenoising()</a></code> add the
denoised expression matrix under the name you specified instead of the
default name <em>denoised</em>. Keep in mind that the size of your
spata-object is mainly determined by the data matrices it contains and
that it can quickly grow to several gigabytes which might considerably
affect the speed with which your R session runs.</p>
</div>
</div>
<div class="section level2">
<h2 id="examples">3. Examples<a class="anchor" aria-label="anchor" href="#examples"></a>
</h2>
<p>Figure 3.1 exemplifies how autoencoder denoising captures the main
pattern and how this pattern is emphasized in the surface plot using the
expression of gene <em>MARCH6</em>.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">spata_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/setActiveExpressionMatrix.html">setActiveMatrix</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">spata_obj</span>, mtr_name <span class="op">=</span> <span class="st">"scaled"</span><span class="op">)</span>

<span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span>
  object <span class="op">=</span> <span class="va">spata_obj</span>, 
  color_by <span class="op">=</span> <span class="st">"MARCH6"</span>, 
  pt_size <span class="op">=</span> <span class="fl">1.8</span>
<span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Scaled (not denoised)"</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="../reference/legendBottom.html">legendNone</a></span><span class="op">(</span><span class="op">)</span>

<span class="va">spata_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/setActiveExpressionMatrix.html">setActiveMatrix</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">spata_obj</span>, mtr_name <span class="op">=</span> <span class="st">"denoised"</span><span class="op">)</span>

<span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span>
  object <span class="op">=</span> <span class="va">spata_obj</span>, 
  color_by <span class="op">=</span> <span class="st">"MARCH6"</span>,
  pt_size <span class="op">=</span> <span class="fl">1.8</span>
  <span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Denoised"</span><span class="op">)</span></code></pre></div>
<div class="figure">
<img src="spata-v2-autoencoder_files/figure-html/example-surface-1.png" alt="Figure 3.1 First visual demonstration of scaled vs. denoised data" width="50%"><img src="spata-v2-autoencoder_files/figure-html/example-surface-2.png" alt="Figure 3.1 First visual demonstration of scaled vs. denoised data" width="50%"><p class="caption">
Figure 3.1 First visual demonstration of scaled vs. denoised data
</p>
</div>
<p>Both, Figure 3.2 and Figure 3.3 go one step further. The example
genes <em>AIF1</em> and <em>CD68</em> are both expressed in macrophages
which, logically, should result in an almost perfect spatial overlap of
both genes. Still, technical artifacts might interfere.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">example_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"AIF1"</span>, <span class="st">"CD68"</span><span class="op">)</span>

<span class="va">spata_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/setActiveExpressionMatrix.html">setActiveMatrix</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">spata_obj</span>, mtr_name <span class="op">=</span> <span class="st">"scaled"</span><span class="op">)</span>

<span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span>
  object <span class="op">=</span> <span class="va">spata_obj</span>,
  color_by <span class="op">=</span> <span class="va">example_genes</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>,
  smooth <span class="op">=</span> <span class="cn">FALSE</span>, 
  pt_size <span class="op">=</span> <span class="fl">1.8</span>
  <span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Scaled (not denoised)"</span><span class="op">)</span> 

<span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span>
  object <span class="op">=</span> <span class="va">spata_obj</span>,
  color_by <span class="op">=</span> <span class="va">example_genes</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>,
  smooth <span class="op">=</span> <span class="cn">FALSE</span>,
  pt_size <span class="op">=</span> <span class="fl">1.8</span>
  <span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Scaled (not denoised)"</span><span class="op">)</span></code></pre></div>
<div class="figure">
<img src="spata-v2-autoencoder_files/figure-html/example-surface2-1.png" alt="Figure 3.2 Not denoised expression of genes AIF1 and CD68" width="50%"><img src="spata-v2-autoencoder_files/figure-html/example-surface2-2.png" alt="Figure 3.2 Not denoised expression of genes AIF1 and CD68" width="50%"><p class="caption">
Figure 3.2 Not denoised expression of genes AIF1 and CD68
</p>
</div>
<p>In addition to that, the fact that both genes naturally differ in
expression levels might affect the colors used to represent these
expression levels which additionally obscure the exclusive coexpression
of <em>AIF1</em> and <em>CD68</em> on the same cells. Denoising does not
lead to perfectly matching expression but it does a sufficient job in
capturing the underlying pattern.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">spata_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/setActiveExpressionMatrix.html">setActiveMatrix</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">spata_obj</span>, mtr_name <span class="op">=</span> <span class="st">"denoised"</span><span class="op">)</span>

<span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span>
  object <span class="op">=</span> <span class="va">spata_obj</span>,
  color_by <span class="op">=</span> <span class="va">example_genes</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>,
  smooth <span class="op">=</span> <span class="cn">FALSE</span>, 
  pt_size <span class="op">=</span> <span class="fl">1.8</span>
  <span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Denoised"</span><span class="op">)</span> 

<span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span>
  object <span class="op">=</span> <span class="va">spata_obj</span>,
  color_by <span class="op">=</span> <span class="va">example_genes</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>,
  smooth <span class="op">=</span> <span class="cn">FALSE</span>, 
  pt_size <span class="op">=</span> <span class="fl">1.8</span>
  <span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Denoised"</span><span class="op">)</span></code></pre></div>
<div class="figure">
<img src="spata-v2-autoencoder_files/figure-html/example_genes-surface3-1.png" alt="Figure 3.3 Denoised expression of genes AIF1 and CD68" width="50%"><img src="spata-v2-autoencoder_files/figure-html/example_genes-surface3-2.png" alt="Figure 3.3 Denoised expression of genes AIF1 and CD68" width="50%"><p class="caption">
Figure 3.3 Denoised expression of genes AIF1 and CD68
</p>
</div>
<p>The prominent changes of the coloring between surface plots from the
scaled expression matrix (Figure 3.2) and from the denoised one (Figure
3.3) results from the autoencoder network returning normally distributed
values while the majority of results from count normalization and
scaling are not normally distributed. To visualize differences between
the scaled and the denoised expression matrix you can either use
functions that visualize the distribution of numeric variables (such as
<code><a href="../reference/plotBoxplot.html">plotDensityplot()</a></code>, <code><a href="../reference/plotBoxplot.html">plotViolinplot()</a></code> etc.) in
combination with <code><a href="../reference/setActiveExpressionMatrix.html">setActiveMatrix()</a></code> as shown in Figure 3.4
or you can use the function <code><a href="../reference/plotAutoencoderResults.html">plotAutoencoderResults()</a></code> as
shown in Figure 3.5.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">spata_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/setActiveExpressionMatrix.html">setActiveMatrix</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">spata_obj</span>, mtr_name <span class="op">=</span> <span class="st">"scaled"</span><span class="op">)</span>

<span class="fu"><a href="../reference/plotBoxplot.html">plotViolinplot</a></span><span class="op">(</span>
  object <span class="op">=</span> <span class="va">spata_obj</span>,
  variables <span class="op">=</span> <span class="va">example_genes</span>,
  display_facets <span class="op">=</span> <span class="cn">FALSE</span>
  <span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Scaled (not denoised)"</span><span class="op">)</span>

<span class="va">spata_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/setActiveExpressionMatrix.html">setActiveMatrix</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">spata_obj</span>, mtr_name <span class="op">=</span> <span class="st">"denoised"</span><span class="op">)</span>

<span class="fu"><a href="../reference/plotBoxplot.html">plotViolinplot</a></span><span class="op">(</span>
  object <span class="op">=</span> <span class="va">spata_obj</span>,
  variables <span class="op">=</span> <span class="va">example_genes</span>,
  display_facets <span class="op">=</span> <span class="cn">FALSE</span>
  <span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Denoised"</span><span class="op">)</span></code></pre></div>
<div class="figure">
<img src="spata-v2-autoencoder_files/figure-html/example_genes-distribution-1.png" alt="Figure 3.4 Differences in distribution between scaled and denoised expression (violinplot)" width="50%"><img src="spata-v2-autoencoder_files/figure-html/example_genes-distribution-2.png" alt="Figure 3.4 Differences in distribution between scaled and denoised expression (violinplot)" width="50%"><p class="caption">
Figure 3.4 Differences in distribution between scaled and denoised
expression (violinplot)
</p>
</div>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotAutoencoderResults.html">plotAutoencoderResults</a></span><span class="op">(</span>
  object <span class="op">=</span> <span class="va">spata_obj</span>,
  genes <span class="op">=</span> <span class="va">example_genes</span>,
  pt_alpha <span class="op">=</span> <span class="fl">0.4</span>,
  pt_size <span class="op">=</span> <span class="fl">1.25</span>
  <span class="op">)</span></code></pre></div>
<div class="figure">
<img src="spata-v2-autoencoder_files/figure-html/example_genes-correlation-1.png" alt="Figure 3.5 Differences in distribution between scaled and denoised expression (scatterplot)" width="100%"><p class="caption">
Figure 3.5 Differences in distribution between scaled and denoised
expression (scatterplot)
</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Jan Kueckelhaus, Dieter-Henrik Heiland, Simon Frerich.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
