<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Differential Expression Analysis (DEA) • SPATA2</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Differential Expression Analysis (DEA)">
<meta property="og:description" content="SPATA2">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SPATA2</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/spata-v2-installation.html">Installation</a>
</li>
<li>
  <a href="../articles/spata-v2-spata-data.html">SPATAData</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">Around the Spata-Object</li>
    <li>
      <a href="../articles/spata-v2-object-initiation-and-manipulation.html">Object Set Up and Manipulation</a>
    </li>
    <li>
      <a href="../articles/">Genes, Gene Sets and Features</a>
    </li>
    <li class="dropdown-header">Integrating Different Platforms</li>
    <li>
      <a href="../articles/spata-v2-cnv-analysis.html">Copy Number Variations</a>
    </li>
    <li>
      <a href="../articles/spata-v2-segmentation.html">Spatial Segmentation (Manual Annotation)</a>
    </li>
    <li>
      <a href="../articles/spata-v2-dea.html">Differential Expression Analysis</a>
    </li>
    <li>
      <a href="../articles/spata-v2-gsea.html">Gene Set Enrichment Analysis</a>
    </li>
    <li class="dropdown-header">Gradient Screening</li>
    <li>
      <a href="../articles/spata-v2-trajectory-analysis.html">Spatial Trajectories</a>
    </li>
    <li>
      <a href="../articles/spata-v2-spatial-trajectory-screening.html">Spatial Trajectory Screening</a>
    </li>
    <li>
      <a href="../articles/spata-v2-image-annotations.html">Image Annotations</a>
    </li>
    <li>
      <a href="../articles/spata-v2-image-annotation-screening.html">Image Annotation Screening</a>
    </li>
    <li class="dropdown-header">Visualization</li>
    <li>
      <a href="../articles/spata-v2-plotting-surface.html">Surface Plotting</a>
    </li>
    <li>
      <a href="../articles/spata-v2-plotting-misc.html">Miscellaneous</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Code &amp; Concept
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/cc-spatial-measures.html">Spatial Measures</a>
    </li>
    <li>
      <a href="../articles/cc-spatial-segmentation-vs-image-annotation.html">Spatial Segmentation vs. Image Annotations</a>
    </li>
    <li>
      <a href="../articles/cc-spatial-trajectory-screening.html">Spatial Trajectory Screening</a>
    </li>
    <li>
      <a href="../articles/cc-image-annotation-screening.html">Image Annotation Screening</a>
    </li>
    <li>
      <a href="../articles/cc-model-fitting-in-spatial-transcriptomic-studies.html">Model Fitting</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Differential Expression Analysis (DEA)</h1>
            
      
      
      <div class="hidden name"><code>spata-v2-dea.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="prerequisites">1. Prerequisites<a class="anchor" aria-label="anchor" href="#prerequisites"></a>
</h2>
<p>None.</p>
</div>
<div class="section level2">
<h2 id="introduction">2. Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Differential expression analysis (DEA) aims to discover quantitative
changes in gene expression levels between defined experimental groups.
Grouping information is stored in form of grouping variables in the
feature data of <code>spata2</code> object. This includes all
spata-intern generated groups such as <a href="spata-v2-spatial-segmentation.html">spatial segmentation</a> and
<a href="spata-v2-clustering.html">clustering</a> as well as any other
grouping variable that has been added via
<code><a href="../reference/addFeatures.html">addFeatures()</a></code>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://themilolab.com/" class="external-link">SPATA2</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SPATAData</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span>

<span class="co"># download sample 269_T </span>
<span class="va">object_t269</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SPATAData/man/downloadSpataObject.html" class="external-link">downloadSpataObject</a></span><span class="op">(</span>sample_name <span class="op">=</span> <span class="st">"269_T"</span><span class="op">)</span>

<span class="va">object_t269</span> <span class="op">&lt;-</span> 
  <span class="fu">adjustDefaultInstructions</span><span class="op">(</span>
    object <span class="op">=</span> <span class="va">object_t269</span>, 
    display_image <span class="op">=</span> <span class="cn">TRUE</span>,
    clrp <span class="op">=</span> <span class="st">"npg"</span>,
    pt_clrp <span class="op">=</span> <span class="st">"npg"</span>
  <span class="op">)</span>

<span class="co"># download sample 275_T </span>
<span class="va">object_t275</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SPATAData/man/downloadSpataObject.html" class="external-link">downloadSpataObject</a></span><span class="op">(</span>sample_name <span class="op">=</span> <span class="st">"275_T"</span><span class="op">)</span>

<span class="va">object_t275</span> <span class="op">&lt;-</span> 
  <span class="fu">adjustDefaultInstructions</span><span class="op">(</span>
    object <span class="op">=</span> <span class="va">object_t275</span>, 
    display_image <span class="op">=</span> <span class="cn">TRUE</span>,
    clrp <span class="op">=</span> <span class="st">"jco"</span>,
    pt_clrp <span class="op">=</span> <span class="st">"jco"</span>
  <span class="op">)</span>

<span class="co"># plot histology images </span>
<span class="fu"><a href="../reference/plotImageGgplot.html">plotImageGgplot</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">object_t269</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>subtitle <span class="op">=</span> <span class="st">"Sample 269_T"</span><span class="op">)</span>

<span class="fu"><a href="../reference/plotImageGgplot.html">plotImageGgplot</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">object_t275</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>subtitle <span class="op">=</span> <span class="st">"Sample 275_T"</span><span class="op">)</span></code></pre></div>
<div class="figure">
<img src="spata-v2-dea_files/figure-html/unnamed-chunk-2-1.png" alt="Fig. 1 The histology of the two example samples." width="50%"><img src="spata-v2-dea_files/figure-html/unnamed-chunk-2-2.png" alt="Fig. 1 The histology of the two example samples." width="50%"><p class="caption">
Fig. 1 The histology of the two example samples.
</p>
</div>
<p>The grouping variables for this tutorial have been prepared before
and can be added like the code shows below.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># load histological/spatial segmentation list</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"spatial_segmentations"</span><span class="op">)</span>

<span class="co"># add histology grouping to spata object </span>
<span class="va">object_t269</span> <span class="op">&lt;-</span> 
  <span class="fu"><a href="../reference/addFeatures.html">addFeatures</a></span><span class="op">(</span>
    object <span class="op">=</span> <span class="va">object_t269</span>, 
    feature_df <span class="op">=</span> <span class="va">spatial_segmentations</span><span class="op">[[</span><span class="st">"269_T"</span><span class="op">]</span><span class="op">]</span>, 
    overwrite <span class="op">=</span> <span class="cn">TRUE</span>
  <span class="op">)</span>

<span class="fu"><a href="../reference/getGroupingOptions.html">getGroupingOptions</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">object_t269</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">##            factor            factor </span>
<span class="co">## "seurat_clusters"       "histology"</span></code></pre>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># load clustering list</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"snn_clustering"</span><span class="op">)</span>

<span class="co"># add cluster grouping to spata object</span>
<span class="va">object_t275</span> <span class="op">&lt;-</span> 
  <span class="fu"><a href="../reference/addFeatures.html">addFeatures</a></span><span class="op">(</span>
    object <span class="op">=</span> <span class="va">object_t275</span>, 
    feature_df <span class="op">=</span> <span class="va">snn_clustering</span><span class="op">[[</span><span class="st">"275_T"</span><span class="op">]</span><span class="op">]</span>, 
    overwrite <span class="op">=</span> <span class="cn">TRUE</span>
  <span class="op">)</span>

<span class="fu"><a href="../reference/getGroupingOptions.html">getGroupingOptions</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">object_t275</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">##            factor            factor            factor </span>
<span class="co">##     "bayes_space" "seurat_clusters"             "snn"</span></code></pre>
<p>Visualize the grouping variables with <code><a href="../reference/plotSurface.html">plotSurface()</a></code>.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span>
  object <span class="op">=</span> <span class="va">object_t269</span>, 
  color_by <span class="op">=</span> <span class="st">"histology"</span>
<span class="op">)</span> 

<span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span>
  object <span class="op">=</span> <span class="va">object_t275</span>, 
  color_by <span class="op">=</span> <span class="st">"snn"</span>
<span class="op">)</span></code></pre></div>
<div class="figure">
<img src="spata-v2-dea_files/figure-html/unnamed-chunk-4-1.png" alt="Fig.2 The grouping of the two examples." width="50%"><img src="spata-v2-dea_files/figure-html/unnamed-chunk-4-2.png" alt="Fig.2 The grouping of the two examples." width="50%"><p class="caption">
Fig.2 The grouping of the two examples.
</p>
</div>
</div>
<div class="section level2">
<h2 id="running-the-analysis">3. Running the analysis<a class="anchor" aria-label="anchor" href="#running-the-analysis"></a>
</h2>
<p>DEA in <strong>SPATA2</strong> uses the function
<code><a href="https://satijalab.org/seurat/reference/FindAllMarkers.html" class="external-link">Seurat::FindAllMarkers()</a></code>. It’s output is a data.frame in
which each row corresponds to a gene that turned out to be a marker gene
for one of the identity groups. Additional variables provide information
about it’s p-value, adjusted p-value, logFold change etc.
<code><a href="../reference/runDEA.html">runDEA()</a></code> does not alter the output but stores it in the
<code>spata2</code> object, more precisely in the <code>@dea</code>
slot. With <code><a href="../reference/printDeaOverview.html">printDeaOverview()</a></code> you can check across which
groups and according to which method DEA has already been conducted.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">object_t269</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runDEA.html">runDEA</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">object_t269</span>, across <span class="op">=</span> <span class="st">"histology"</span>, method_de <span class="op">=</span> <span class="st">"wilcox"</span><span class="op">)</span>

<span class="fu"><a href="../reference/printDeaOverview.html">printDeaOverview</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">object_t269</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## DEA results exist for grouping variables:</span>
<span class="co">## </span>
<span class="co">## - 'seurat_clusters' with methods: wilcox</span>
<span class="co">## - 'histology' with methods: wilcox</span></code></pre>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">object_t275</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runDEA.html">runDEA</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">object_t275</span>, across <span class="op">=</span> <span class="st">"snn"</span>, method_de <span class="op">=</span> <span class="st">"wilcox"</span><span class="op">)</span>

<span class="fu"><a href="../reference/printDeaOverview.html">printDeaOverview</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">object_t275</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## DEA results exist for grouping variables:</span>
<span class="co">## </span>
<span class="co">## - 'snn' with methods: wilcox</span>
<span class="co">## - 'bayes_space' with methods: wilcox</span></code></pre>
</div>
<div class="section level2">
<h2 id="extracting-results">3. Extracting results<a class="anchor" aria-label="anchor" href="#extracting-results"></a>
</h2>
<p>There are two main functions with which you can manually extract the
DEA results desired. First, <code><a href="../reference/getDeaResultsDf.html">getDeaResultsDf()</a></code> returns the
original resulting data.frame of <code><a href="https://satijalab.org/seurat/reference/FindAllMarkers.html" class="external-link">Seurat::FindAllMarkers()</a></code>.
<code><a href="../reference/getDeaResultsDf.html">getDeaGenes()</a></code> returns a vector of gene names.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># extract the complete data.frame</span>
<span class="fu"><a href="../reference/getDeaResultsDf.html">getDeaResultsDf</a></span><span class="op">(</span>
  object <span class="op">=</span> <span class="va">object_t275</span>, 
  across <span class="op">=</span> <span class="st">"snn"</span>, 
  method_de <span class="op">=</span> <span class="st">"wilcox"</span>
<span class="op">)</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 3,786 x 7</span></span>
<span class="co">##        p_val avg_log2FC pct.1 pct.2 p_val_adj snn   gene  </span>
<span class="co">##        <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> </span>
<span class="co">## <span style="color: #BCBCBC;"> 1</span> 7.77<span style="color: #949494;">e</span><span style="color: #BB0000;">-182</span>      1.27  0.799 0.277 1.67<span style="color: #949494;">e</span><span style="color: #BB0000;">-177</span> 0     POSTN </span>
<span class="co">## <span style="color: #BCBCBC;"> 2</span> 6.72<span style="color: #949494;">e</span><span style="color: #BB0000;">-221</span>      1.22  1     0.982 1.44<span style="color: #949494;">e</span><span style="color: #BB0000;">-216</span> 0     PTN   </span>
<span class="co">## <span style="color: #BCBCBC;"> 3</span> 1.89<span style="color: #949494;">e</span><span style="color: #BB0000;">-233</span>      1.22  1     0.996 4.06<span style="color: #949494;">e</span><span style="color: #BB0000;">-229</span> 0     CST3  </span>
<span class="co">## <span style="color: #BCBCBC;"> 4</span> 4.55<span style="color: #949494;">e</span><span style="color: #BB0000;">-180</span>      1.18  0.95  0.583 9.75<span style="color: #949494;">e</span><span style="color: #BB0000;">-176</span> 0     LANCL2</span>
<span class="co">## <span style="color: #BCBCBC;"> 5</span> 1.14<span style="color: #949494;">e</span><span style="color: #BB0000;">-151</span>      0.997 0.925 0.553 2.44<span style="color: #949494;">e</span><span style="color: #BB0000;">-147</span> 0     TTYH3 </span>
<span class="co">## <span style="color: #BCBCBC;"> 6</span> 1.68<span style="color: #949494;">e</span><span style="color: #BB0000;">-165</span>      0.996 0.995 0.885 3.60<span style="color: #949494;">e</span><span style="color: #BB0000;">-161</span> 0     HOPX  </span>
<span class="co">## <span style="color: #BCBCBC;"> 7</span> 2.08<span style="color: #949494;">e</span><span style="color: #BB0000;">-158</span>      0.983 0.96  0.699 4.47<span style="color: #949494;">e</span><span style="color: #BB0000;">-154</span> 0     METRN </span>
<span class="co">## <span style="color: #BCBCBC;"> 8</span> 3.64<span style="color: #949494;">e</span><span style="color: #BB0000;">-177</span>      0.980 0.995 0.9   7.80<span style="color: #949494;">e</span><span style="color: #BB0000;">-173</span> 0     RAMP1 </span>
<span class="co">## <span style="color: #BCBCBC;"> 9</span> 5.41<span style="color: #949494;">e</span><span style="color: #BB0000;">-138</span>      0.924 0.994 0.758 1.16<span style="color: #949494;">e</span><span style="color: #BB0000;">-133</span> 0     RCAN1 </span>
<span class="co">## <span style="color: #BCBCBC;">10</span> 5.57<span style="color: #949494;">e</span><span style="color: #BB0000;">-157</span>      0.885 0.997 0.751 1.19<span style="color: #949494;">e</span><span style="color: #BB0000;">-152</span> 0     IGFBP2</span>
<span class="co">## <span style="color: #949494;"># i 3,776 more rows</span></span></code></pre>
<p>Using the arguments <code>across_subset</code>, <code>min_lfc</code>,
<code>n_highest_lfc</code>, <code>max_adj_pval</code>,
<code>n_lowest_pval</code> the output of the function can be adjusted to
specific questions.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># e.g. top 10 genes for histology area 'tumor' </span>
<span class="fu"><a href="../reference/getDeaResultsDf.html">getDeaResultsDf</a></span><span class="op">(</span>
  object <span class="op">=</span> <span class="va">object_t269</span>, 
  across <span class="op">=</span> <span class="st">"histology"</span>,
  across_subset <span class="op">=</span> <span class="st">"tumor"</span>, <span class="co"># the group name(s) of interest,</span>
  n_highest_lfc <span class="op">=</span> <span class="fl">10</span>, <span class="co"># top ten genes</span>
  max_adj_pval <span class="op">=</span> <span class="fl">0.01</span> <span class="co"># pval must be lower or equal than 0.01</span>
<span class="op">)</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 10 x 7</span></span>
<span class="co">##    p_val avg_log2FC pct.1 pct.2 p_val_adj histology gene    </span>
<span class="co">##    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   </span>
<span class="co">## <span style="color: #BCBCBC;"> 1</span>     0       2.30 1     0.917         0 tumor     CCT6A   </span>
<span class="co">## <span style="color: #BCBCBC;"> 2</span>     0       2.08 1     0.998         0 tumor     SEC61G  </span>
<span class="co">## <span style="color: #BCBCBC;"> 3</span>     0       2.03 1     0.879         0 tumor     NIPSNAP2</span>
<span class="co">## <span style="color: #BCBCBC;"> 4</span>     0       1.86 0.999 0.909         0 tumor     PTN     </span>
<span class="co">## <span style="color: #BCBCBC;"> 5</span>     0       1.82 1     0.834         0 tumor     CD99    </span>
<span class="co">## <span style="color: #BCBCBC;"> 6</span>     0       1.79 1     0.977         0 tumor     SPP1    </span>
<span class="co">## <span style="color: #BCBCBC;"> 7</span>     0       1.76 1     0.928         0 tumor     PTPRZ1  </span>
<span class="co">## <span style="color: #BCBCBC;"> 8</span>     0       1.75 1     0.865         0 tumor     SPARC   </span>
<span class="co">## <span style="color: #BCBCBC;"> 9</span>     0       1.72 0.995 0.815         0 tumor     APOC1   </span>
<span class="co">## <span style="color: #BCBCBC;">10</span>     0       1.60 1     0.844         0 tumor     EGFR</span></code></pre>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># e.g. all upregulated genes</span>
<span class="fu"><a href="../reference/getDeaResultsDf.html">getDeaResultsDf</a></span><span class="op">(</span>
  object <span class="op">=</span> <span class="va">object_t275</span>,
  across <span class="op">=</span> <span class="st">"snn"</span>, 
  min_lfc <span class="op">=</span> <span class="fl">0.1</span> <span class="co"># genes with lfc bigger than or equal to 0.1</span>
<span class="op">)</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 1,953 x 7</span></span>
<span class="co">##        p_val avg_log2FC pct.1 pct.2 p_val_adj snn   gene  </span>
<span class="co">##        <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> </span>
<span class="co">## <span style="color: #BCBCBC;"> 1</span> 7.77<span style="color: #949494;">e</span><span style="color: #BB0000;">-182</span>      1.27  0.799 0.277 1.67<span style="color: #949494;">e</span><span style="color: #BB0000;">-177</span> 0     POSTN </span>
<span class="co">## <span style="color: #BCBCBC;"> 2</span> 6.72<span style="color: #949494;">e</span><span style="color: #BB0000;">-221</span>      1.22  1     0.982 1.44<span style="color: #949494;">e</span><span style="color: #BB0000;">-216</span> 0     PTN   </span>
<span class="co">## <span style="color: #BCBCBC;"> 3</span> 1.89<span style="color: #949494;">e</span><span style="color: #BB0000;">-233</span>      1.22  1     0.996 4.06<span style="color: #949494;">e</span><span style="color: #BB0000;">-229</span> 0     CST3  </span>
<span class="co">## <span style="color: #BCBCBC;"> 4</span> 4.55<span style="color: #949494;">e</span><span style="color: #BB0000;">-180</span>      1.18  0.95  0.583 9.75<span style="color: #949494;">e</span><span style="color: #BB0000;">-176</span> 0     LANCL2</span>
<span class="co">## <span style="color: #BCBCBC;"> 5</span> 1.14<span style="color: #949494;">e</span><span style="color: #BB0000;">-151</span>      0.997 0.925 0.553 2.44<span style="color: #949494;">e</span><span style="color: #BB0000;">-147</span> 0     TTYH3 </span>
<span class="co">## <span style="color: #BCBCBC;"> 6</span> 1.68<span style="color: #949494;">e</span><span style="color: #BB0000;">-165</span>      0.996 0.995 0.885 3.60<span style="color: #949494;">e</span><span style="color: #BB0000;">-161</span> 0     HOPX  </span>
<span class="co">## <span style="color: #BCBCBC;"> 7</span> 2.08<span style="color: #949494;">e</span><span style="color: #BB0000;">-158</span>      0.983 0.96  0.699 4.47<span style="color: #949494;">e</span><span style="color: #BB0000;">-154</span> 0     METRN </span>
<span class="co">## <span style="color: #BCBCBC;"> 8</span> 3.64<span style="color: #949494;">e</span><span style="color: #BB0000;">-177</span>      0.980 0.995 0.9   7.80<span style="color: #949494;">e</span><span style="color: #BB0000;">-173</span> 0     RAMP1 </span>
<span class="co">## <span style="color: #BCBCBC;"> 9</span> 5.41<span style="color: #949494;">e</span><span style="color: #BB0000;">-138</span>      0.924 0.994 0.758 1.16<span style="color: #949494;">e</span><span style="color: #BB0000;">-133</span> 0     RCAN1 </span>
<span class="co">## <span style="color: #BCBCBC;">10</span> 5.57<span style="color: #949494;">e</span><span style="color: #BB0000;">-157</span>      0.885 0.997 0.751 1.19<span style="color: #949494;">e</span><span style="color: #BB0000;">-152</span> 0     IGFBP2</span>
<span class="co">## <span style="color: #949494;"># i 1,943 more rows</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="visualize-results">4. Visualize results<a class="anchor" aria-label="anchor" href="#visualize-results"></a>
</h2>
<div class="section level3">
<h3 id="heatmaps">4.1 Heatmaps<a class="anchor" aria-label="anchor" href="#heatmaps"></a>
</h3>
<p><code><a href="../reference/plotDeaHeatmap.html">plotDeaHeatmap()</a></code> visualizes the results of DEA by using
to the results you would extract with
<code><a href="../reference/getDeaResultsDf.html">getDeaResultsDf()</a></code>.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">hm</span> <span class="op">&lt;-</span> 
  <span class="fu"><a href="../reference/plotDeaHeatmap.html">plotDeaHeatmap</a></span><span class="op">(</span>
    object <span class="op">=</span> <span class="va">object_t275</span>, 
    across <span class="op">=</span> <span class="st">"snn"</span>,
    across_subset <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"0"</span>, <span class="st">"2"</span>, <span class="st">"4"</span>, <span class="st">"6"</span><span class="op">)</span>,
    method_de <span class="op">=</span> <span class="st">"wilcox"</span>,
    n_highest_lfc <span class="op">=</span> <span class="fl">10</span>, 
    n_bcsp <span class="op">=</span> <span class="fl">100</span>
  <span class="op">)</span>

<span class="va">hm</span></code></pre></div>
<div class="figure">
<img src="spata-v2-dea_files/figure-html/unnamed-chunk-12-1.png" alt="Fig.3 DEA-Heatmap." width="100%"><p class="caption">
Fig.3 DEA-Heatmap.
</p>
</div>
</div>
<div class="section level3">
<h3 id="dotplots">4.2 Dotplots<a class="anchor" aria-label="anchor" href="#dotplots"></a>
</h3>
<p><code><a href="../reference/plotDeaDotPlot.html">plotDeaDotPlot()</a></code> visualizes the results of DEA either by
group…</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotDeaDotPlot.html">plotDeaDotPlot</a></span><span class="op">(</span>
  object <span class="op">=</span> <span class="va">object_t275</span>, 
  across <span class="op">=</span> <span class="st">"snn"</span>,
  across_subset <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"0"</span>, <span class="st">"2"</span>, <span class="st">"4"</span><span class="op">)</span>, 
  n_highest_lfc <span class="op">=</span> <span class="fl">5</span>,
  by_group <span class="op">=</span> <span class="cn">TRUE</span>, 
  scales <span class="op">=</span> <span class="st">"free_y"</span>, 
  ncol <span class="op">=</span> <span class="fl">1</span>
<span class="op">)</span></code></pre></div>
<div class="figure">
<img src="spata-v2-dea_files/figure-html/unnamed-chunk-13-1.png" alt="Fig.4 DEA-Dotplot by group." width="100%"><p class="caption">
Fig.4 DEA-Dotplot by group.
</p>
</div>
<p>… or with all groups together.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotDeaDotPlot.html">plotDeaDotPlot</a></span><span class="op">(</span>
  object <span class="op">=</span> <span class="va">object_t269</span>, 
  across <span class="op">=</span> <span class="st">"histology"</span>,
  color_by <span class="op">=</span> <span class="st">"histology"</span>,
  n_highest_lfc <span class="op">=</span> <span class="fl">5</span>, 
  by_group <span class="op">=</span> <span class="cn">FALSE</span>
<span class="op">)</span></code></pre></div>
<div class="figure">
<img src="spata-v2-dea_files/figure-html/unnamed-chunk-14-1.png" alt="Fig.5 DEA-Dotplot for all groups." width="100%"><p class="caption">
Fig.5 DEA-Dotplot for all groups.
</p>
</div>
</div>
<div class="section level3">
<h3 id="box--and-violinplots">4.3 Box- and Violinplots<a class="anchor" aria-label="anchor" href="#box--and-violinplots"></a>
</h3>
<p>There are additional ways to visualize the results of your DEA. As
with almost all plotting functions in <strong>SPATA2</strong> a vector
of gene names is necessary for the function to know which genes to plot.
<code><a href="../reference/getDeaResultsDf.html">getDeaGenes()</a></code> is the second function to extract DEA results
and a wrapper around <code><a href="../reference/getDeaResultsDf.html">getDeaResultsDf()</a></code> that returns a
vector gene names.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">genes_of_interest</span> <span class="op">&lt;-</span>
  <span class="fu"><a href="../reference/getDeaResultsDf.html">getDeaGenes</a></span><span class="op">(</span>
    object <span class="op">=</span> <span class="va">object_t269</span>,
    across <span class="op">=</span> <span class="st">"histology"</span>, <span class="co"># the grouping variable</span>
    method_de <span class="op">=</span> <span class="st">"wilcox"</span>, <span class="co"># the method with which the results were computed</span>
    n_highest_lfc <span class="op">=</span> <span class="fl">10</span>, 
    max_adj_pval <span class="op">=</span> <span class="fl">0.01</span>
  <span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">genes_of_interest</span><span class="op">)</span> <span class="co"># first six</span></code></pre></div>
<pre><code><span class="co">##      tumor      tumor      tumor      tumor      tumor      tumor </span>
<span class="co">##    "CCT6A"   "SEC61G" "NIPSNAP2"      "PTN"     "CD99"     "SPP1"</span></code></pre>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">tail</a></span><span class="op">(</span><span class="va">genes_of_interest</span><span class="op">)</span> <span class="co"># last six</span></code></pre></div>
<pre><code><span class="co">## infiltrated infiltrated infiltrated infiltrated infiltrated infiltrated </span>
<span class="co">##      "RTN1"     "NPAS4"    "SLC1A2"      "BEX1"      "ENC1"      "NRGN"</span></code></pre>
<p>A vector of gene names as returned by <code><a href="../reference/getDeaResultsDf.html">getDeaGenes()</a></code> is a
perfectly valid input for other plotting functions.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">top_5_markers_269</span> <span class="op">&lt;-</span> 
  <span class="fu"><a href="../reference/getDeaResultsDf.html">getDeaGenes</a></span><span class="op">(</span>
    object <span class="op">=</span> <span class="va">object_t269</span>,
    across <span class="op">=</span> <span class="st">"histology"</span>,
    n_lowest_pval <span class="op">=</span> <span class="fl">5</span>, 
    min_lfc <span class="op">=</span> <span class="fl">0.1</span> <span class="co"># set min_lfc! else downregulated genes are included</span>
    <span class="op">)</span> 

<span class="va">top_5_markers_269</span></code></pre></div>
<pre><code><span class="co">##       tumor       tumor       tumor       tumor       tumor  transition </span>
<span class="co">##     "CCT6A"    "SEC61G"  "NIPSNAP2"       "PTN"      "CD99"      "PLP1" </span>
<span class="co">##  transition  transition  transition  transition infiltrated infiltrated </span>
<span class="co">##       "MBP"        "TF"   "SELENOP"       "MAG"    "SNAP25"     "VSNL1" </span>
<span class="co">## infiltrated infiltrated infiltrated </span>
<span class="co">##       "CCK"     "OLFM1"      "RTN1"</span></code></pre>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># plot results for t269</span>
<span class="fu"><a href="../reference/plotBoxplot.html">plotBoxplot</a></span><span class="op">(</span>
  object <span class="op">=</span> <span class="va">object_t269</span>,
  variables <span class="op">=</span> <span class="va">top_5_markers_269</span>,
  across <span class="op">=</span> <span class="st">"histology"</span>, 
  nrow <span class="op">=</span> <span class="fl">3</span>
  <span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>, axis.ticks.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="../reference/legendBottom.html">legendTop</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<div class="figure">
<img src="spata-v2-dea_files/figure-html/unnamed-chunk-16-1.png" alt="Fig.7 DEA-Boxplots" width="100%"><p class="caption">
Fig.7 DEA-Boxplots
</p>
</div>
</div>
<div class="section level3">
<h3 id="surface-plots">4.3 Surface plots<a class="anchor" aria-label="anchor" href="#surface-plots"></a>
</h3>
<p>Of course, another way to visualize your results are surface
plots.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotSurfaceComparison.html">plotSurfaceComparison</a></span><span class="op">(</span>
  object <span class="op">=</span> <span class="va">object_t269</span>, 
  color_by <span class="op">=</span> <span class="va">top_5_markers_269</span> ,
  nrow <span class="op">=</span> <span class="fl">3</span>
  <span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="../reference/legendBottom.html">legendNone</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<div class="figure">
<img src="spata-v2-dea_files/figure-html/unnamed-chunk-17-1.png" alt="Fig.8 Surface plots with DEA results for sample 269_T." width="100%"><p class="caption">
Fig.8 Surface plots with DEA results for sample 269_T.
</p>
</div>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">top_5_markers_275</span> <span class="op">&lt;-</span> 
  <span class="fu"><a href="../reference/getDeaResultsDf.html">getDeaGenes</a></span><span class="op">(</span>
    object <span class="op">=</span> <span class="va">object_t275</span>,
    across <span class="op">=</span> <span class="st">"snn"</span>,
    across_subset <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"0"</span>, <span class="st">"2"</span>, <span class="st">"4"</span><span class="op">)</span>,
    n_lowest_pval <span class="op">=</span> <span class="fl">5</span>, 
    min_lfc <span class="op">=</span> <span class="fl">0.1</span> <span class="co"># set min_lfc! else downregulated genes are included</span>
    <span class="op">)</span>

<span class="fu"><a href="../reference/plotSurfaceComparison.html">plotSurfaceComparison</a></span><span class="op">(</span>
  object <span class="op">=</span> <span class="va">object_t275</span>, 
  color_by <span class="op">=</span> <span class="va">top_5_markers_275</span> ,
  nrow <span class="op">=</span> <span class="fl">3</span>
  <span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="../reference/legendBottom.html">legendNone</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<div class="figure">
<img src="spata-v2-dea_files/figure-html/unnamed-chunk-18-1.png" alt="Fig.9 Surface plots with DEA results for sample 275_T." width="100%"><p class="caption">
Fig.9 Surface plots with DEA results for sample 275_T.
</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Jan Kueckelhaus, Dieter-Henrik Heiland.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
