<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Copy Number Variations (CNV) â€¢ SPATA2</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Copy Number Variations (CNV)">
<meta property="og:description" content="SPATA2">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SPATA2</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/spata-v2-installation.html">Installation</a>
</li>
<li>
  <a href="../articles/spata-v2-spata-data.html">SPATAData</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">Around the spata2 object</li>
    <li>
      <a href="../articles/spata-v2-object-initiation-and-manipulation.html">Object Set Up and Manipulation</a>
    </li>
    <li>
      <a href="../articles/spata-v2-platform-compatibility.html">Platform Compatibility</a>
    </li>
    <li>
      <a href="../articles/spata-v2-autoencoder.html">Autoencoder</a>
    </li>
    <li>
      <a href="../articles/spata-v2-genes-gene-sets-and-features.html">Genes, Gene Sets and Features</a>
    </li>
    <li class="dropdown-header">Image Analysis</li>
    <li>
      <a href="../articles/spata-v2-image-handling.html">Image Handling</a>
    </li>
    <li>
      <a href="../articles/spata-v2-image-annotations.html">Image Annotations</a>
    </li>
    <li>
      <a href="../articles/spata-v2-spatial-segmentation.html">Spatial Segmentation</a>
    </li>
    <li class="dropdown-header">Analysis of Gene Expression</li>
    <li>
      <a href="../articles/spata-v2-cnv-analysis.html">Copy Number Variations</a>
    </li>
    <li>
      <a href="../articles/spata-v2-clustering.html">Clustering</a>
    </li>
    <li>
      <a href="../articles/spata-v2-dea.html">Differential Expression</a>
    </li>
    <li>
      <a href="../articles/spata-v2-gsea.html">Gene Set Enrichment</a>
    </li>
    <li class="dropdown-header">Expression Gradients &amp; Gradient Screening</li>
    <li>
      <a href="../articles/spata-v2-spatial-trajectories.html">Spatial Trajectories</a>
    </li>
    <li>
      <a href="../articles/spata-v2-spatial-trajectory-screening.html">Spatial Trajectory Screening</a>
    </li>
    <li>
      <a href="../articles/spata-v2-image-annotation-screening.html">Image Annotation Screening</a>
    </li>
    <li>
      <a href="../articles/spata-v2-scd.html">Integrate Single Cell Deconvolution</a>
    </li>
    <li class="dropdown-header">Visualization</li>
    <li>
      <a href="../articles/spata-v2-plotting-surface.html">Surface Plotting</a>
    </li>
    <li>
      <a href="../articles/spata-v2-plotting-misc.html">Miscellaneous</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Code &amp; Concept
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/cc-spatial-measures.html">Spatial Measures</a>
    </li>
    <li>
      <a href="../articles/cc-spatial-segmentation-vs-image-annotations.html">Spatial Segmentation vs. Image Annotations</a>
    </li>
    <li>
      <a href="../articles/cc-spatial-trajectory-screening.html">Spatial Trajectory Screening</a>
    </li>
    <li>
      <a href="../articles/cc-image-annotation-screening.html">Image Annotation Screening</a>
    </li>
    <li>
      <a href="../articles/cc-model-fitting-in-spatial-transcriptomic-studies.html">Model Fitting</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Copy Number Variations (CNV)</h1>
            
      
      
      <div class="hidden name"><code>spata-v2-cnv-analysis.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="prerequisites">1. Prerequisites<a class="anchor" aria-label="anchor" href="#prerequisites"></a>
</h2>
<p>None.</p>
</div>
<div class="section level2">
<h2 id="introduction">2. Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>InferCNV is used to identify large-scale chromosomal copy number
alterations in single cell RNA-Seq data including gains or deletions of
chromosomes or large segments of such. Several publications have
leveraged this technique. (<a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4123637/" class="external-link">Patel et
al., 2014</a>, <a href="https://pubmed.ncbi.nlm.nih.gov/27124452/" class="external-link">Tirosh et al.,
2016</a>, <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5519096/" class="external-link">Venteicher
et al., 2017</a> )</p>
<p>Note: We are currently encountering problems with the function
<code>infercnv:::plot_cnv()</code>. It does not prevent the copy number
variation analysis. Sometimes, however, the heatmap is not plotted.
Checkout your R console for hints in how far the function did not work
properly. We are working on a solution for that. If you encounter
warnings raised from <code><a href="../reference/runCnvAnalysis.html">runCnvAnalysis()</a></code> <a href="https://github.com/theMILOlab/SPATA2/issues" class="external-link">let us know</a>.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r code-chunk"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co"># load package</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(SPATA2)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(SPATAData)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>object_t269 <span class="ot">&lt;-</span> <span class="fu">downloadSpataObject</span>(<span class="at">sample_name =</span> <span class="st">"269_T"</span>)</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="co"># load histolgoy segmentations</span></span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(<span class="st">"spatial_segmentations"</span>)</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a>object_t269 <span class="ot">&lt;-</span> </span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">addFeatures</span>(</span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a>    <span class="at">object =</span> object_t269, </span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a>    <span class="at">feature_df =</span> spatial_segmentations[[<span class="st">"269_T"</span>]], </span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a>    <span class="at">overwrite =</span> <span class="cn">TRUE</span></span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb1-17"><a href="#cb1-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-18"><a href="#cb1-18" aria-hidden="true" tabindex="-1"></a><span class="co"># only histology</span></span>
<span id="cb1-19"><a href="#cb1-19" aria-hidden="true" tabindex="-1"></a><span class="fu">plotSurface</span>(object, <span class="ot">=</span> object_t269, <span class="at">pt_alpha =</span> <span class="dv">0</span>)</span>
<span id="cb1-20"><a href="#cb1-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-21"><a href="#cb1-21" aria-hidden="true" tabindex="-1"></a><span class="co"># histological grouping</span></span>
<span id="cb1-22"><a href="#cb1-22" aria-hidden="true" tabindex="-1"></a><span class="fu">plotSurface</span>(<span class="at">object =</span> object_t269, <span class="at">color_by =</span> <span class="st">"histology"</span>)</span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># only histology</span>
<span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">object_t269</span>, pt_alpha <span class="op">=</span> <span class="fl">0</span><span class="op">)</span>

<span class="co"># histological grouping</span>
<span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">object_t269</span>, color_by <span class="op">=</span> <span class="st">"histology"</span>, pt_clrp <span class="op">=</span> <span class="st">"npg"</span><span class="op">)</span></code></pre></div>
<div class="figure">
<img src="spata-v2-cnv-analysis_files/figure-html/unnamed-chunk-3-1.png" alt="Fig.1 The example sample raw and colored by histological classification." width="50%"><img src="spata-v2-cnv-analysis_files/figure-html/unnamed-chunk-3-2.png" alt="Fig.1 The example sample raw and colored by histological classification." width="50%"><p class="caption">
Fig.1 The example sample raw and colored by histological classification.
</p>
</div>
<p><strong>SPATA2</strong> implements the package <a href="https://github.com/broadinstitute/inferCNV/wiki" class="external-link"><code>infercnv</code></a>
published and maintained by Broadinstitute and allows to integrate this
technique in your workflow of analyzing spatial trancsriptomic data
derived from malignancies.</p>
</div>
<div class="section level2">
<h2 id="running-cnv-analysis">3. Running CNV-Analysis<a class="anchor" aria-label="anchor" href="#running-cnv-analysis"></a>
</h2>
<p>As with all other functions prefixed with <code>run*()</code> the
function <code><a href="../reference/runCnvAnalysis.html">runCnvAnalysis()</a></code> is a wrapper around all necessary
functions needed to conduct copy-number-variation-analysis. The results
needed for subsequent analysis steps are stored in the specified
<code>spata2</code> object (slot: <span class="citation">@cnv</span>).
The infercnv-object is stored in the folder specified in the argument
<code>directory_cnv_folder</code>.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">object_t269</span> <span class="op">&lt;-</span>
  <span class="fu"><a href="../reference/runCnvAnalysis.html">runCnvAnalysis</a></span><span class="op">(</span>
    object <span class="op">=</span> <span class="va">object_t269</span>,
    directory_cnv_folder <span class="op">=</span> <span class="st">"data-gbm269/cnv-results"</span>, <span class="co"># example directory</span>
    cnv_prefix <span class="op">=</span> <span class="st">"Chr"</span>
    <span class="op">)</span></code></pre></div>
<p>If your desired set up deviates from the default you can reach any
function of the <code>inferncnv</code> pipeline by entering itâ€™s name as
an argument and specify itâ€™s input as a list of arguments with which you
want it to be called.</p>
<p>Copy number variation analysis requires reference data. This includes
a count matrix from healthy tissue, an annotation file as well as a
data.frame that contains information about the chromosome positions. We
provide reference data in the list <code><a href="../reference/cnv_ref.html">SPATA2::cnv_ref</a></code>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">cnv_ref</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] "annotation" "mtr"        "regions"</span></code></pre>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">cnv_ref</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">##            Length   Class      Mode   </span>
<span class="co">## annotation        1 data.frame list   </span>
<span class="co">## mtr        71067022 -none-     numeric</span>
<span class="co">## regions           4 data.frame list</span></code></pre>
<p><code><a href="../reference/runCnvAnalysis.html">runCnvAnalysis()</a></code> defaults to the content of this list.
The documentation of <code><a href="../reference/runCnvAnalysis.html">runCnvAnalysis()</a></code> contains a detailed
description of the requirements each reference input must meet in order
for the function to work.</p>
</div>
<div class="section level2">
<h2 id="cnv-results">4. CNV-Results<a class="anchor" aria-label="anchor" href="#cnv-results"></a>
</h2>
<p>The results are stored in a list inside the <code>spata2</code>
object. This list can be obtained via <code><a href="../reference/getCnvResults.html">getCnvResults()</a></code>.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cnv_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getCnvResults.html">getCnvResults</a></span><span class="op">(</span><span class="va">object_t269</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">cnv_results</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] "prefix"      "cnv_df"      "cnv_mtr"     "gene_pos_df" "regions_df"</span></code></pre>
</div>
<div class="section level2">
<h2 id="visualization">5. Visualization<a class="anchor" aria-label="anchor" href="#visualization"></a>
</h2>
<div class="section level3">
<h3 id="heatmap">5.1 Heatmap<a class="anchor" aria-label="anchor" href="#heatmap"></a>
</h3>
<p>The gains and losses of chromosomal segments can be displayed via
<code><a href="../reference/plotCnvHeatmap.html">plotCnvHeatmap()</a></code>.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotCnvHeatmap.html">plotCnvHeatmap</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">object_t269</span><span class="op">)</span></code></pre></div>
<div class="figure">
<img src="spata-v2-cnv-analysis_files/figure-html/unnamed-chunk-8-1.png" alt="Fig.2 Basic CNV-Heatmap." width="100%"><p class="caption">
Fig.2 Basic CNV-Heatmap.
</p>
</div>
<p>If you want to visualize the results across certain groups to
identify subclones make use of the <code>across</code>-argument.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotCnvHeatmap.html">plotCnvHeatmap</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">object_t269</span>, across <span class="op">=</span> <span class="st">"histology"</span>, clrp <span class="op">=</span> <span class="st">"npg"</span><span class="op">)</span></code></pre></div>
<div class="figure">
<img src="spata-v2-cnv-analysis_files/figure-html/unnamed-chunk-9-1.png" alt="Fig.3 CNV-Heatmap with meta data displayes chromosomal alterations across histological classification." width="100%"><p class="caption">
Fig.3 CNV-Heatmap with meta data displayes chromosomal alterations
across histological classification.
</p>
</div>
</div>
<div class="section level3">
<h3 id="lineplot">5.2 Lineplot<a class="anchor" aria-label="anchor" href="#lineplot"></a>
</h3>
<p>An additional option to visualize the results provides
<code><a href="../reference/plotCnvLineplot.html">plotCnvLineplot()</a></code>.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotCnvLineplot.html">plotCnvLineplot</a></span><span class="op">(</span>
  object <span class="op">=</span> <span class="va">object_t269</span>,
  across <span class="op">=</span> <span class="st">"histology"</span>, 
  n_bins_bcsp <span class="op">=</span> <span class="fl">1000</span>,
  n_bins_genes <span class="op">=</span> <span class="fl">1000</span>,
  nrow <span class="op">=</span> <span class="fl">3</span>
<span class="op">)</span></code></pre></div>
<div class="figure">
<img src="spata-v2-cnv-analysis_files/figure-html/unnamed-chunk-10-1.png" alt="Fig.4 CNV-Lineplot displays chromosomal changes across histological classification." width="100%"><p class="caption">
Fig.4 CNV-Lineplot displays chromosomal changes across histological
classification.
</p>
</div>
</div>
<div class="section level3">
<h3 id="surface">5.3 Surface<a class="anchor" aria-label="anchor" href="#surface"></a>
</h3>
<p>The numeric values by which the copy number variations of each
chromosome are represented are immediately transferred to the
<code>spata2</code> objectâ€™s feature data and are thus accessible for
all functions that work with numeric variables. The character string
specified in argument <code>cnv_prefix</code> combined with the
chromosomes number determines the name by which you can refer to these
feature variables.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># cnv feature names</span>
<span class="fu"><a href="../reference/getCnvFeatureNames.html">getCnvFeatureNames</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">object_t269</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">##  [1] "Chr1"  "Chr2"  "Chr3"  "Chr4"  "Chr5"  "Chr6"  "Chr7"  "Chr8"  "Chr9" </span>
<span class="co">## [10] "Chr10" "Chr11" "Chr12" "Chr13" "Chr14" "Chr15" "Chr16" "Chr17" "Chr18"</span>
<span class="co">## [19] "Chr19" "Chr20" "Chr21" "Chr22"</span></code></pre>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># are part of all feature names</span>
<span class="fu"><a href="../reference/getFeatureNames.html">getFeatureNames</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">object_t269</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">##            numeric            integer             factor            numeric </span>
<span class="co">##   "nCount_Spatial" "nFeature_Spatial"  "seurat_clusters"             "Chr0" </span>
<span class="co">##            numeric            numeric            numeric            numeric </span>
<span class="co">##             "Chr1"            "Chr10"            "Chr11"            "Chr12" </span>
<span class="co">##            numeric            numeric            numeric            numeric </span>
<span class="co">##            "Chr13"            "Chr14"            "Chr15"            "Chr16" </span>
<span class="co">##            numeric            numeric            numeric            numeric </span>
<span class="co">##            "Chr17"            "Chr18"            "Chr19"             "Chr2" </span>
<span class="co">##            numeric            numeric            numeric            numeric </span>
<span class="co">##            "Chr20"            "Chr21"            "Chr22"            "Chr23" </span>
<span class="co">##            numeric            numeric            numeric            numeric </span>
<span class="co">##             "Chr3"             "Chr4"             "Chr5"             "Chr6" </span>
<span class="co">##            numeric            numeric            numeric             factor </span>
<span class="co">##             "Chr7"             "Chr8"             "Chr9"        "histology"</span></code></pre>
<p>Use <code><a href="../reference/plotSurface.html">plotSurface()</a></code> to visualize chromosomal alterations
directly on the histology.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span>
  object <span class="op">=</span> <span class="va">object_t269</span>, 
  color_by <span class="op">=</span> <span class="st">"Chr7"</span>, 
  pt_clrsp <span class="op">=</span> <span class="st">"Reds"</span>
<span class="op">)</span>

<span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span>
  object <span class="op">=</span> <span class="va">object_t269</span>, 
  color_by <span class="op">=</span> <span class="st">"Chr10"</span>, 
  pt_clrsp <span class="op">=</span> <span class="st">"Oslo"</span> 
<span class="op">)</span></code></pre></div>
<div class="figure">
<img src="spata-v2-cnv-analysis_files/figure-html/unnamed-chunk-11-1.png" alt="Fig.5 Surface plots display chromosomal alterations directory on the histology." width="50%"><img src="spata-v2-cnv-analysis_files/figure-html/unnamed-chunk-11-2.png" alt="Fig.5 Surface plots display chromosomal alterations directory on the histology." width="50%"><p class="caption">
Fig.5 Surface plots display chromosomal alterations directory on the
histology.
</p>
</div>
</div>
<div class="section level3">
<h3 id="gradients">5.4 Gradients<a class="anchor" aria-label="anchor" href="#gradients"></a>
</h3>
<p>As showcased in our corresponding vignette about <a href="spata-v2-spatial-trajectories.html">spatial trajectories</a>
gradients of numeric variables can be displayed. Chromosomal alterations
are numeric variables.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">object_t269</span> <span class="op">&lt;-</span> 
    <span class="fu"><a href="../reference/setTrajectory.html">setTrajectory</a></span><span class="op">(</span>
      object <span class="op">=</span> <span class="va">object_t269</span>, 
      trajectory <span class="op">=</span> <span class="va">spatial_trajectories</span><span class="op">[[</span><span class="st">"269_T"</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="st">"horizontal_mid"</span><span class="op">]</span><span class="op">]</span>, 
      overwrite <span class="op">=</span> <span class="cn">TRUE</span>
    <span class="op">)</span>

<span class="fu"><a href="../reference/plotSpatialTrajectories.html">plotSpatialTrajectories</a></span><span class="op">(</span>
  object <span class="op">=</span> <span class="va">object_t269</span>, 
  ids <span class="op">=</span> <span class="st">"horizontal_mid"</span>, 
  color_by <span class="op">=</span> <span class="st">"Chr7"</span>, 
  pt_clrsp <span class="op">=</span> <span class="st">"Reds 3"</span>
<span class="op">)</span>

<span class="fu"><a href="../reference/plotSpatialTrajectories.html">plotSpatialTrajectories</a></span><span class="op">(</span>
  object <span class="op">=</span> <span class="va">object_t269</span>, 
  ids <span class="op">=</span> <span class="st">"horizontal_mid"</span>, 
  color_by <span class="op">=</span> <span class="st">"Chr10"</span>, 
  pt_clrsp <span class="op">=</span> <span class="st">"Oslo"</span>
<span class="op">)</span>

<span class="fu"><a href="../reference/plotTrajectoryLineplot.html">plotTrajectoryLineplot</a></span><span class="op">(</span>
  object <span class="op">=</span> <span class="va">object_t269</span>, 
  id <span class="op">=</span> <span class="st">"horizontal_mid"</span>, 
  variables <span class="op">=</span> <span class="st">"Chr7"</span>, 
  line_color <span class="op">=</span> <span class="st">"red"</span>
<span class="op">)</span>

<span class="fu"><a href="../reference/plotTrajectoryLineplot.html">plotTrajectoryLineplot</a></span><span class="op">(</span>
  object <span class="op">=</span> <span class="va">object_t269</span>, 
  id <span class="op">=</span> <span class="st">"horizontal_mid"</span>, 
  variables <span class="op">=</span> <span class="st">"Chr10"</span>, 
  line_color <span class="op">=</span> <span class="st">"blue"</span>, 
  x_nth <span class="op">=</span> <span class="fl">4</span>
<span class="op">)</span></code></pre></div>
<div class="figure">
<img src="spata-v2-cnv-analysis_files/figure-html/unnamed-chunk-12-1.png" alt="Fig.6 Gradients of Chromosomal alterations along a trajectory in space." width="50%"><img src="spata-v2-cnv-analysis_files/figure-html/unnamed-chunk-12-2.png" alt="Fig.6 Gradients of Chromosomal alterations along a trajectory in space." width="50%"><img src="spata-v2-cnv-analysis_files/figure-html/unnamed-chunk-12-3.png" alt="Fig.6 Gradients of Chromosomal alterations along a trajectory in space." width="50%"><img src="spata-v2-cnv-analysis_files/figure-html/unnamed-chunk-12-4.png" alt="Fig.6 Gradients of Chromosomal alterations along a trajectory in space." width="50%"><p class="caption">
Fig.6 Gradients of Chromosomal alterations along a trajectory in space.
</p>
</div>
<p>(Note that trajectory lineplots always rescale variables to 0-1 or to
low-high.)</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Jan Kueckelhaus, Dieter-Henrik Heiland, Simon Frerich.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
