<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Image Annotation Screening • SPATA2</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Image Annotation Screening">
<meta property="og:description" content="SPATA2">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SPATA2</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/spata-v2-installation.html">Installation</a>
</li>
<li>
  <a href="../articles/spata-v2-spata-data.html">SPATAData</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">Around the spata2 object</li>
    <li>
      <a href="../articles/spata-v2-object-initiation-and-manipulation.html">Object Set Up and Manipulation</a>
    </li>
    <li>
      <a href="../articles/spata-v2-platform-compatibility.html">Platform Compatibility</a>
    </li>
    <li>
      <a href="../articles/spata-v2-autoencoder.html">Autoencoder</a>
    </li>
    <li>
      <a href="../articles/spata-v2-genes-gene-sets-and-features.html">Genes, Gene Sets and Features</a>
    </li>
    <li class="dropdown-header">Image Analysis</li>
    <li>
      <a href="../articles/spata-v2-image-handling.html">Image Handling</a>
    </li>
    <li>
      <a href="../articles/spata-v2-image-annotations.html">Image Annotations</a>
    </li>
    <li>
      <a href="../articles/spata-v2-spatial-segmentation.html">Spatial Segmentation</a>
    </li>
    <li class="dropdown-header">Analysis of Gene Expression</li>
    <li>
      <a href="../articles/spata-v2-cnv-analysis.html">Copy Number Variations</a>
    </li>
    <li>
      <a href="../articles/spata-v2-clustering.html">Clustering</a>
    </li>
    <li>
      <a href="../articles/spata-v2-dea.html">Differential Expression</a>
    </li>
    <li>
      <a href="../articles/spata-v2-gsea.html">Gene Set Enrichment</a>
    </li>
    <li class="dropdown-header">Expression Gradients &amp; Gradient Screening</li>
    <li>
      <a href="../articles/spata-v2-spatial-trajectories.html">Spatial Trajectories</a>
    </li>
    <li>
      <a href="../articles/spata-v2-spatial-trajectory-screening.html">Spatial Trajectory Screening</a>
    </li>
    <li>
      <a href="../articles/spata-v2-image-annotation-screening.html">Image Annotation Screening</a>
    </li>
    <li>
      <a href="../articles/spata-v2-scd.html">Integrate Single Cell Deconvolution</a>
    </li>
    <li class="dropdown-header">Visualization</li>
    <li>
      <a href="../articles/spata-v2-plotting-surface.html">Surface Plotting</a>
    </li>
    <li>
      <a href="../articles/spata-v2-plotting-misc.html">Miscellaneous</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Code &amp; Concept
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/cc-spatial-measures.html">Spatial Measures</a>
    </li>
    <li>
      <a href="../articles/cc-spatial-segmentation-vs-image-annotations.html">Spatial Segmentation vs. Image Annotations</a>
    </li>
    <li>
      <a href="../articles/cc-spatial-trajectory-screening.html">Spatial Trajectory Screening</a>
    </li>
    <li>
      <a href="../articles/cc-image-annotation-screening.html">Image Annotation Screening</a>
    </li>
    <li>
      <a href="../articles/cc-model-fitting-in-spatial-transcriptomic-studies.html">Model Fitting</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Image Annotation Screening</h1>
            
      
      
      <div class="hidden name"><code>spata-v2-image-annotation-screening.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="requisites">1. Requisites<a class="anchor" aria-label="anchor" href="#requisites"></a>
</h2>
<p>Make sure to be familiar with the following vignettes:</p>
<ul>
<li><a href="spata-v2-image-annotations.html">Image Annotations</a></li>
<li><a href="cc-spatial-segmentation-vs-image-annotations.html">Spatial
Segmentation vs. Image Annotations - Code &amp; Concept</a></li>
<li><a href="cc-image-annotation-screening.html">Image Annotation
Screening - Code &amp; Concept</a></li>
<li><a href="cc-model-fitting-in-spatial-transcriptomics-studies.html">Model
Fitting in Spatial Transcriptomic Studies - Code &amp; Concept</a></li>
</ul>
<p>This one might help, too:</p>
<ul>
<li><a href="cc-spatial-measures.html">Spatial Measures</a></li>
</ul>
</div>
<div class="section level2">
<h2 id="introduction-overview">2. Introduction &amp; overview<a class="anchor" aria-label="anchor" href="#introduction-overview"></a>
</h2>
<p>Many algorithms have been developed to identify genes in spatial
transcriptomic studies that are expressed in a spatially meaningful
manner. E.g. <a href="https://xzhoulab.github.io/SPARK/" class="external-link">SPARKX</a>, <a href="https://github.com/edsgard/trendsceek" class="external-link">Trendseek</a>, <a href="https://github.com/Teichlab/SpatialDE" class="external-link">SpatialDE</a>. Drawback of
these algorithms is that they only <strong>detect</strong> genes of
interest but rarely come with additional information that allow
biological interpretation. Additionally, they always take the sample as
a whole and do not allow the user to specify histological regions of
interest that require further investigation regarding the impact they
have on the rest of the sample. To address this, we have created
<em>Image annotations screening (IAS)</em>. In combination with our
recently introduced <a href="spata-v2-image-annotations.html">image
annotation</a> system, IAS allows to screen for genes that feature
interesting expression patterns <strong>in relation to histological
structures or user defined regions</strong> as a function of distance to
them.</p>
<p>This tutorial guides you through all of its main functions. For a
more detailed explanation on how IAS works click <a href="">here</a> to
be guided to the code and concept behind image annotation screening.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://themilolab.com/" class="external-link">SPATA2</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SPATAData</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span>

<span class="va">object_t313</span> <span class="op">&lt;-</span> <span class="fu">SPATAData</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SPATAData/man/downloadSpataObject.html" class="external-link">downloadSpataObject</a></span><span class="op">(</span>sample_name <span class="op">=</span> <span class="st">"313_T"</span>, file <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span>

<span class="va">object_t313</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/setDefault.html">setDefault</a></span><span class="op">(</span><span class="va">object_t313</span>, pt_clrsp <span class="op">=</span> <span class="st">"Reds 3"</span>, display_image <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="setting-up-the-algorithm">3. Setting up the algorithm<a class="anchor" aria-label="anchor" href="#setting-up-the-algorithm"></a>
</h2>
<p>Image annotation screening requires four aspects must be
specified.</p>
<ul>
<li><ol style="list-style-type: decimal">
<li>The image annotation that encircles the area of interest, using
argument <code>id</code>.</li>
</ol></li>
<li><ol start="2" style="list-style-type: decimal">
<li>The area surrounding the image annotation that is screened. This is
done with a combination of arguments <code>distance</code>,
<code>binwidth</code> and <code>n_bins_circle</code>.</li>
</ol></li>
<li><ol start="3" style="list-style-type: decimal">
<li>Further confinement of the area by angle, using argument
<code>angle_span</code>.</li>
</ol></li>
<li><ol start="4" style="list-style-type: decimal">
<li>The resolution of the screening, using argument
<code>n_bins_angle</code>.</li>
</ol></li>
</ul>
<div class="section level3">
<h3 id="the-image-annotation-of-interest">3.1 The image annotation of interest<a class="anchor" aria-label="anchor" href="#the-image-annotation-of-interest"></a>
</h3>
<p>At first, the area of interest must be captured within an image
annotation. This is done with <code>createImageAnnotation()</code>. The
sample that will be used in this tutorial is the same as the one that’s
used in the <a href="">image annotation tutorial</a>. Example image
annotations are provided in the list <code>image_annotations</code>and
can be directly added using <code><a href="../reference/setImageAnnotation.html">setImageAnnotation()</a></code>. The image
annotation that is used throughout this vignette encircles a necrotic
area that is located in the center of the Glioblastoma. It carries the
ID <em>necrotic_center</em>.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># load example list</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"image_annotations"</span><span class="op">)</span>

<span class="co"># load example or create interactively with createImageAnnotations()</span>
<span class="va">object_t313</span> <span class="op">&lt;-</span> 
  <span class="fu"><a href="../reference/setImageAnnotation.html">setImageAnnotation</a></span><span class="op">(</span>
    object <span class="op">=</span> <span class="va">object_t313</span>, 
    img_ann <span class="op">=</span> <span class="va">image_annotations</span><span class="op">[[</span><span class="st">"313_T"</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="st">"necrotic_center"</span><span class="op">]</span><span class="op">]</span>,
    overwrite <span class="op">=</span> <span class="cn">TRUE</span>
  <span class="op">)</span>

<span class="fu"><a href="../reference/plotImageGgplot.html">plotImageGgplot</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">object_t313</span><span class="op">)</span>

<span class="fu"><a href="../reference/plotImageAnnotations.html">plotImageAnnotations</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">object_t313</span>, ids <span class="op">=</span> <span class="st">"necrotic_center"</span>, expand <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span></code></pre></div>
<div class="figure">
<img src="spata-v2-image-annotation-screening_files/figure-html/unnamed-chunk-3-1.png" alt="Fig.1 Glioblastoma T313 with it's prominent necrosis region." width="50%"><img src="spata-v2-image-annotation-screening_files/figure-html/unnamed-chunk-3-2.png" alt="Fig.1 Glioblastoma T313 with it's prominent necrosis region." width="50%"><p class="caption">
Fig.1 Glioblastoma T313 with it’s prominent necrosis region.
</p>
</div>
</div>
<div class="section level3">
<h3 id="the-area-surrounding-the-image-annotation">3.2 The area surrounding the image annotation<a class="anchor" aria-label="anchor" href="#the-area-surrounding-the-image-annotation"></a>
</h3>
<p>After deciding on the image annotation of interest with the argument
<code>id</code>, here <code>id = 'necrotic_center'</code>, the area
<strong>around the image annotation</strong> that is included in the
screening process must be set up. To set up the area that is screened
two of the three arguments <code>distance</code>, <code>binwidth</code>
and <code>n_bins_circle</code> must be specified. The set up can be
checked and visualized with the function
<code><a href="../reference/plotSurfaceIAS.html">plotSurfaceIAS()</a></code>.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">bcsp_dist</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getCCD.html">getCCD</a></span><span class="op">(</span><span class="va">object_t313</span>, unit <span class="op">=</span> <span class="st">"mm"</span><span class="op">)</span>

<span class="co"># show results</span>
<span class="va">bcsp_dist</span></code></pre></div>
<pre><code><span class="co">## 0.1 [mm]</span></code></pre>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotSurfaceIAS.html">plotSurfaceIAS</a></span><span class="op">(</span>
  object <span class="op">=</span> <span class="va">object_t313</span>, 
  id <span class="op">=</span> <span class="st">"necrotic_center"</span>, <span class="co"># the ID of the image annotation of interest</span>
  distance <span class="op">=</span> <span class="st">"1mm"</span>,
  binwidth <span class="op">=</span> <span class="va">bcsp_dist</span>, 
  pt_clrp <span class="op">=</span> <span class="st">"milo"</span>, <span class="co"># use colorful panel to highlight the bins</span>
  display_bins_angle <span class="op">=</span> <span class="cn">FALSE</span>, 
  ggpLayers <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>subtitle <span class="op">=</span> <span class="st">"a)"</span><span class="op">)</span>
<span class="op">)</span>

<span class="fu"><a href="../reference/plotSurfaceIAS.html">plotSurfaceIAS</a></span><span class="op">(</span>
  object <span class="op">=</span> <span class="va">object_t313</span>, 
  id <span class="op">=</span> <span class="st">"necrotic_center"</span>, <span class="co"># the ID of the image annotation of interest</span>
  distance <span class="op">=</span> <span class="st">"2.25mm"</span>, <span class="co"># covers the whole sample</span>
  binwidth <span class="op">=</span> <span class="va">bcsp_dist</span><span class="op">*</span><span class="fl">2</span>, <span class="co"># two layers of spots per bine</span>
  display_bins_angle <span class="op">=</span> <span class="cn">FALSE</span>,
  ggpLayers <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>subtitle <span class="op">=</span> <span class="st">"b)"</span><span class="op">)</span>
<span class="op">)</span> </code></pre></div>
<div class="figure">
<img src="spata-v2-image-annotation-screening_files/figure-html/unnamed-chunk-4-1.png" alt="Fig.2 Visualize the area that is included in the screening process." width="50%"><img src="spata-v2-image-annotation-screening_files/figure-html/unnamed-chunk-4-2.png" alt="Fig.2 Visualize the area that is included in the screening process." width="50%"><p class="caption">
Fig.2 Visualize the area that is included in the screening process.
</p>
</div>
</div>
<div class="section level3">
<h3 id="further-confining-the-area-by-angle">3.3 Further confining the area by angle<a class="anchor" aria-label="anchor" href="#further-confining-the-area-by-angle"></a>
</h3>
<p>To further confine the screened area you can use
<code>angle_span</code>.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotSurfaceIAS.html">plotSurfaceIAS</a></span><span class="op">(</span>
  object <span class="op">=</span> <span class="va">object_t313</span>, 
  id <span class="op">=</span> <span class="st">"necrotic_center"</span>, <span class="co"># the ID of the image annotation of interest</span>
  distance <span class="op">=</span> <span class="st">"2.25mm"</span>,
  binwidth <span class="op">=</span> <span class="va">bcsp_dist</span>, 
  angle_span <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">45</span>,<span class="fl">225</span><span class="op">)</span>, <span class="co"># use angle_span to confine the area based on angle</span>
  display_bins_angle <span class="op">=</span> <span class="cn">FALSE</span>,
  ggpLayers <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>subtitle <span class="op">=</span> <span class="st">"a)"</span><span class="op">)</span>
<span class="op">)</span> 

<span class="fu"><a href="../reference/plotSurfaceIAS.html">plotSurfaceIAS</a></span><span class="op">(</span>
  object <span class="op">=</span> <span class="va">object_t313</span>, 
  id <span class="op">=</span> <span class="st">"necrotic_center"</span>, <span class="co"># the ID of the image annotation of interest</span>
  distance <span class="op">=</span> <span class="st">"2.25mm"</span>,
  binwidth <span class="op">=</span> <span class="va">bcsp_dist</span>, 
  angle_span <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">45</span>, <span class="fl">225</span><span class="op">)</span>,
  display_bins_angle <span class="op">=</span> <span class="cn">FALSE</span>, 
  ggpLayers <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>subtitle <span class="op">=</span> <span class="st">"b)"</span><span class="op">)</span>
<span class="op">)</span> </code></pre></div>
<div class="figure">
<img src="spata-v2-image-annotation-screening_files/figure-html/unnamed-chunk-5-1.png" alt="Fig.3 Confining the area by an angle span." width="50%"><img src="spata-v2-image-annotation-screening_files/figure-html/unnamed-chunk-5-2.png" alt="Fig.3 Confining the area by an angle span." width="50%"><p class="caption">
Fig.3 Confining the area by an angle span.
</p>
</div>
</div>
<div class="section level3">
<h3 id="the-resolution">3.4 The resolution<a class="anchor" aria-label="anchor" href="#the-resolution"></a>
</h3>
<p>Gene expression changes area inferred by binning the barcode-spots by
their distance to the area of interest and by aligning mean gene
expression of each bin. To increase the resolution, the barcode-spots
can be additionally binned by angle. Argument <code>n_bins_angle</code>
defaults to 1.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotSurfaceIAS.html">plotSurfaceIAS</a></span><span class="op">(</span>
  object <span class="op">=</span> <span class="va">object_t313</span>, 
  id <span class="op">=</span> <span class="st">"necrotic_center"</span>,
  distance <span class="op">=</span> <span class="st">"2.25mm"</span>, 
  n_bins_angle <span class="op">=</span> <span class="fl">1</span>
  <span class="op">)</span></code></pre></div>
<div class="figure">
<img src="spata-v2-image-annotation-screening_files/figure-html/unnamed-chunk-6-1.png" alt="Fig.4 One angle bin." width="100%"><p class="caption">
Fig.4 One angle bin.
</p>
</div>
<p>By increasing <code>n_bins_angle</code> inferring gene expression
changes and the model fitting takes place within each angle-bin. The
final evaluation that tells whether a gene is expressed in an
e.g. descending way starting from the image annotation consists of a
summary of all gene-model-fits per angle bin.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotSurfaceIAS.html">plotSurfaceIAS</a></span><span class="op">(</span>
  object <span class="op">=</span> <span class="va">object_t313</span>, 
  id <span class="op">=</span> <span class="st">"necrotic_center"</span>,
  distance <span class="op">=</span> <span class="st">"2.25mm"</span>, 
  n_bins_angle <span class="op">=</span> <span class="fl">12</span>
  <span class="op">)</span></code></pre></div>
<div class="figure">
<img src="spata-v2-image-annotation-screening_files/figure-html/unnamed-chunk-7-1.png" alt="Fig.5 Twelve angle bins." width="100%"><p class="caption">
Fig.5 Twelve angle bins.
</p>
</div>
<p>Increasing the resolution ensures that detected genes follow a
certain trend <strong>everywhere</strong> around the image annotation
which results in fewer false-positives but also in fewer true-positives.
This is because genes might be excluded due to a bad evaluation in one
angle-bin that might affect the overall evaluation in a way that it is
discarded although within the remaining angle-bins the evaluation were
good.</p>
</div>
</div>
<div class="section level2">
<h2 id="running-the-algorithm">4. Running the algorithm<a class="anchor" aria-label="anchor" href="#running-the-algorithm"></a>
</h2>
<p>Once you’ve decided on the screening parameters, specify them within
the function <code><a href="../reference/imageAnnotationScreening.html">imageAnnotationScreening()</a></code> which runs the
algorithm. The parameter <code>variables</code> takes the numeric
variables that are included in the screening process. Usually it’s gene
names. Image annotation screening, however, is not restricted to genes
as virtually every numeric variable (e.g. <em>nCount_Genes</em>,
<em>nCount_Features</em>, gene sets, etc.) can be fitted to spatially
meaningful patterns, too. Therefore, the argument for the input is
simply called <code>variables</code>.</p>
<p>Here, we are using the genes that were already identified as
spatially variable by <a href="https://xzhoulab.github.io/SPARK/" class="external-link">SPARKX</a>. The goal is to
further analyze which of the genes are expressed in a spatial relation
to the necrotic center. In particular, we want to find <strong>genes
associated with necrosis</strong> using the descending models and
<strong>genes repelled by necrosis</strong> using the ascending
models.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># this is a wrapper around SPARK::sparkx()</span>
<span class="va">object_t313</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runSparkx.html">runSparkx</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">object_t313</span><span class="op">)</span>

<span class="va">spark_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runSparkx.html">getSparkxGeneDf</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">object_t313</span>, threshold_pval <span class="op">=</span> <span class="fl">0.01</span><span class="op">)</span>

<span class="co"># extract genes with a sparkx pvalue of 0.01 or lower</span>
<span class="va">sparkx_genes</span> <span class="op">&lt;-</span> <span class="va">spark_df</span><span class="op">[[</span><span class="st">"genes"</span><span class="op">]</span><span class="op">]</span>

<span class="co"># show results</span>
<span class="va">spark_df</span>

<span class="co"># show results (&gt; 10 000 genes detected as spatially variable with a p-value of &lt; 0.01)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">sparkx_genes</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 10,804 x 3</span></span>
<span class="co">##    genes   combinedPval adjustedPval</span>
<span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="co">## <span style="color: #BCBCBC;"> 1</span> B2M        7.11<span style="color: #949494;">e</span><span style="color: #BB0000;">-275</span>    1.46<span style="color: #949494;">e</span><span style="color: #BB0000;">-269</span></span>
<span class="co">## <span style="color: #BCBCBC;"> 2</span> SERF2      1.97<span style="color: #949494;">e</span><span style="color: #BB0000;">-261</span>    2.02<span style="color: #949494;">e</span><span style="color: #BB0000;">-256</span></span>
<span class="co">## <span style="color: #BCBCBC;"> 3</span> HLA-B      1.18<span style="color: #949494;">e</span><span style="color: #BB0000;">-247</span>    8.07<span style="color: #949494;">e</span><span style="color: #BB0000;">-243</span></span>
<span class="co">## <span style="color: #BCBCBC;"> 4</span> ACTB       3.33<span style="color: #949494;">e</span><span style="color: #BB0000;">-243</span>    1.71<span style="color: #949494;">e</span><span style="color: #BB0000;">-238</span></span>
<span class="co">## <span style="color: #BCBCBC;"> 5</span> MYL6       8.77<span style="color: #949494;">e</span><span style="color: #BB0000;">-241</span>    3.60<span style="color: #949494;">e</span><span style="color: #BB0000;">-236</span></span>
<span class="co">## <span style="color: #BCBCBC;"> 6</span> UBC        4.00<span style="color: #949494;">e</span><span style="color: #BB0000;">-238</span>    1.37<span style="color: #949494;">e</span><span style="color: #BB0000;">-233</span></span>
<span class="co">## <span style="color: #BCBCBC;"> 7</span> ATP5F1E    1.03<span style="color: #949494;">e</span><span style="color: #BB0000;">-231</span>    3.02<span style="color: #949494;">e</span><span style="color: #BB0000;">-227</span></span>
<span class="co">## <span style="color: #BCBCBC;"> 8</span> ACTG1      7.41<span style="color: #949494;">e</span><span style="color: #BB0000;">-229</span>    1.91<span style="color: #949494;">e</span><span style="color: #BB0000;">-224</span></span>
<span class="co">## <span style="color: #BCBCBC;"> 9</span> H3F3A      2.66<span style="color: #949494;">e</span><span style="color: #BB0000;">-227</span>    6.07<span style="color: #949494;">e</span><span style="color: #BB0000;">-223</span></span>
<span class="co">## <span style="color: #BCBCBC;">10</span> S100A11    2.29<span style="color: #949494;">e</span><span style="color: #BB0000;">-226</span>    4.71<span style="color: #949494;">e</span><span style="color: #BB0000;">-222</span></span>
<span class="co">## <span style="color: #949494;"># i 10,794 more rows</span></span></code></pre>
<pre><code><span class="co">##  chr [1:10804] "B2M" "SERF2" "HLA-B" "ACTB" "MYL6" "UBC" "ATP5F1E" "ACTG1" ...</span></code></pre>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># models that screen for genes that ...</span>
<span class="va">models_of_interest</span> <span class="op">&lt;-</span> 
  <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"linear_ascending"</span>, <span class="co"># ... increase gradually with the distance to the necrosis (repelled)</span>
    <span class="st">"late_ascending"</span>, <span class="co"># ... increase logarithmically with the distance to necrosis (stronger repelled)</span>
    <span class="st">"immediate_descending"</span>, <span class="co"># ... decrease logarithmically with the distance to necrosis (strongly associated)</span>
    <span class="st">"linear_descending"</span> <span class="co"># ... decrease gradually with the distance to necrosis (associated)</span>
    <span class="op">)</span>

<span class="va">n_bins_circle</span> <span class="op">&lt;-</span> <span class="fl">20</span>

<span class="fu"><a href="../reference/showModels.html">showModels</a></span><span class="op">(</span>
  input <span class="op">=</span> <span class="va">n_bins_circle</span>, <span class="co"># length of the model vector</span>
  model_subset <span class="op">=</span> <span class="va">models_of_interest</span>
<span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Circle Bins (Distance to Necrosis)"</span><span class="op">)</span></code></pre></div>
<div class="figure">
<img src="spata-v2-image-annotation-screening_files/figure-html/unnamed-chunk-10-1.png" alt="Fig.6 The models against which gene expression is fitted." width="100%"><p class="caption">
Fig.6 The models against which gene expression is fitted.
</p>
</div>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># run the algorithm and save the output in a new object</span>
<span class="va">IAS_T313</span> <span class="op">&lt;-</span> 
  <span class="fu"><a href="../reference/imageAnnotationScreening.html">imageAnnotationScreening</a></span><span class="op">(</span>
    object <span class="op">=</span> <span class="va">object_t313</span>, <span class="co"># the spata object</span>
    id <span class="op">=</span> <span class="st">"necrotic_center"</span>, <span class="co"># the image annotation of interest</span>
    variables <span class="op">=</span> <span class="va">sparkx_genes</span>, <span class="co"># the variables/genes to screen</span>
    n_bins_circle <span class="op">=</span> <span class="fl">20</span>, 
    distance <span class="op">=</span> <span class="st">"2.25mm"</span>,
    model_subset <span class="op">=</span> <span class="va">models_of_interest</span> <span class="co"># the models against which all genes are fitted </span>
  <span class="op">)</span></code></pre></div>
<pre><code><span class="co">## 20:57:54 Specified `distance` = 2.25mm and `n_bins_circle` = 20. Calculated `binwidth` = 0.1125mm.</span></code></pre>
<p><strong>Note:</strong> The output of
<code><a href="../reference/imageAnnotationScreening.html">imageAnnotationScreening()</a></code> is not saved in the
<code>spata2</code> object but returned in a separate S4 object of class
<code>ImageAnnotationScreening</code>. Do not overwrite the
<code>spata2</code> object by writing
<code>object_t313 &lt;- imageAnnotationScreening(object = object_t313, id = "necrotic_center", ...)</code>.</p>
</div>
<div class="section level2">
<h2 id="results">5. Results<a class="anchor" aria-label="anchor" href="#results"></a>
</h2>
<div class="section level3">
<h3 id="visualization">5.1 Visualization<a class="anchor" aria-label="anchor" href="#visualization"></a>
</h3>
<p>To get an overview of the screening as well as a first visualization
of the results use <code>plotOverview</code>. It sorts the genes by
their best model-fit and plots the evaluation score against the p-value
for every model.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotOverview.html">plotOverview</a></span><span class="op">(</span>
  object <span class="op">=</span> <span class="va">IAS_T313</span>,
  label_vars <span class="op">=</span> <span class="fl">5</span>, 
  label_size <span class="op">=</span> <span class="fl">3</span>
  <span class="op">)</span></code></pre></div>
<div class="figure">
<img src="spata-v2-image-annotation-screening_files/figure-html/unnamed-chunk-13-1.png" alt="Fig.7 Visualize gene evaluation by model." width="100%"><p class="caption">
Fig.7 Visualize gene evaluation by model.
</p>
</div>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># using ten circle bins to visualize the concept</span>
<span class="va">ias_layer_bins</span> <span class="op">&lt;-</span> 
  <span class="fu"><a href="../reference/ggpLayerEncirclingIAS.html">ggpLayerEncirclingIAS</a></span><span class="op">(</span>
    object <span class="op">=</span> <span class="va">object_t313</span>,
    distance <span class="op">=</span> <span class="st">"2.25mm"</span>, 
    n_bins_circle <span class="op">=</span> <span class="fl">10</span>, 
    id <span class="op">=</span> <span class="st">"necrotic_center"</span>,
    line_size <span class="op">=</span> <span class="fl">1</span>
  <span class="op">)</span>

<span class="co"># using only one circle bin to visualize the screening area</span>
<span class="va">ias_layer_range</span> <span class="op">&lt;-</span> 
  <span class="fu"><a href="../reference/ggpLayerEncirclingIAS.html">ggpLayerEncirclingIAS</a></span><span class="op">(</span>
    object <span class="op">=</span> <span class="va">object_t313</span>,
    distance <span class="op">=</span> <span class="st">"2.25mm"</span>, 
    n_bins_circle <span class="op">=</span> <span class="fl">1</span>, 
    id <span class="op">=</span> <span class="st">"necrotic_center"</span>,
    line_size <span class="op">=</span> <span class="fl">1</span>
  <span class="op">)</span>

<span class="va">example_genes</span> <span class="op">&lt;-</span> 
  <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>
    imm_desc <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"IGFBP3"</span>, <span class="st">"CD44"</span><span class="op">)</span>,
    lin_desc <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"FN1"</span>, <span class="st">"COL6A3"</span><span class="op">)</span>,
    lin_asc <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"SERF2"</span>, <span class="st">"MYL6"</span><span class="op">)</span>,
    late_asc <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"TRMT112"</span>, <span class="st">"LAPTM5"</span><span class="op">)</span>
    <span class="op">)</span> 

<span class="fu"><a href="../reference/plotSurfaceComparison.html">plotSurfaceComparison</a></span><span class="op">(</span>
  object <span class="op">=</span> <span class="va">object_t313</span>,
  color_by <span class="op">=</span> <span class="va">example_genes</span><span class="op">$</span><span class="va">imm_desc</span>, 
  nrow <span class="op">=</span> <span class="fl">1</span> 
<span class="op">)</span> <span class="op">+</span> 
  <span class="va">ias_layer_bins</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>subtitle <span class="op">=</span> <span class="st">"a) Immediate Descending"</span><span class="op">)</span>

<span class="fu"><a href="../reference/plotSurfaceComparison.html">plotSurfaceComparison</a></span><span class="op">(</span>
  object <span class="op">=</span> <span class="va">object_t313</span>,
  color_by <span class="op">=</span> <span class="va">example_genes</span><span class="op">$</span><span class="va">lin_desc</span>, 
  nrow <span class="op">=</span> <span class="fl">1</span> 
<span class="op">)</span> <span class="op">+</span> 
  <span class="va">ias_layer_range</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>subtitle <span class="op">=</span> <span class="st">"b) Linear descending"</span><span class="op">)</span>

<span class="fu"><a href="../reference/plotSurfaceComparison.html">plotSurfaceComparison</a></span><span class="op">(</span>
  object <span class="op">=</span> <span class="va">object_t313</span>,
  color_by <span class="op">=</span> <span class="va">example_genes</span><span class="op">$</span><span class="va">lin_asc</span>, 
  nrow <span class="op">=</span> <span class="fl">1</span> 
<span class="op">)</span> <span class="op">+</span> 
  <span class="va">ias_layer_range</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>subtitle <span class="op">=</span> <span class="st">"c) Linear ascending"</span><span class="op">)</span>

<span class="fu"><a href="../reference/plotSurfaceComparison.html">plotSurfaceComparison</a></span><span class="op">(</span>
  object <span class="op">=</span> <span class="va">object_t313</span>,
  color_by <span class="op">=</span> <span class="va">example_genes</span><span class="op">$</span><span class="va">late_asc</span>, 
  nrow <span class="op">=</span> <span class="fl">1</span> 
<span class="op">)</span> <span class="op">+</span> 
  <span class="va">ias_layer_range</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>subtitle <span class="op">=</span> <span class="st">"d) Late ascending"</span><span class="op">)</span></code></pre></div>
<div class="figure">
<img src="spata-v2-image-annotation-screening_files/figure-html/unnamed-chunk-14-1.png" alt="Fig.8 Visualize gene expression on the surface." width="50%"><img src="spata-v2-image-annotation-screening_files/figure-html/unnamed-chunk-14-2.png" alt="Fig.8 Visualize gene expression on the surface." width="50%"><img src="spata-v2-image-annotation-screening_files/figure-html/unnamed-chunk-14-3.png" alt="Fig.8 Visualize gene expression on the surface." width="50%"><img src="spata-v2-image-annotation-screening_files/figure-html/unnamed-chunk-14-4.png" alt="Fig.8 Visualize gene expression on the surface." width="50%"><p class="caption">
Fig.8 Visualize gene expression on the surface.
</p>
</div>
<p>The inferred gene expression can be visualized with
<code><a href="../reference/plotIasLineplot.html">plotIasLineplot()</a></code>.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">genes_vec</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://purrr.tidyverse.org/reference/flatten.html" class="external-link">flatten_chr</a></span><span class="op">(</span><span class="va">example_genes</span><span class="op">)</span>

<span class="fu"><a href="../reference/plotIasLineplot.html">plotIasLineplot</a></span><span class="op">(</span>
  object <span class="op">=</span> <span class="va">object_t313</span>, 
  id <span class="op">=</span> <span class="st">"necrotic_center"</span>,
  distance <span class="op">=</span> <span class="st">"2.25mm"</span>,
  n_bins_circle <span class="op">=</span> <span class="fl">20</span>,
  variables <span class="op">=</span> <span class="va">genes_vec</span>, 
  include_area <span class="op">=</span> <span class="cn">FALSE</span>,
  line_color <span class="op">=</span> <span class="st">"red"</span>,
  display_border <span class="op">=</span> <span class="cn">TRUE</span>,
  border_linesize <span class="op">=</span> <span class="fl">0.75</span>,
  border_linetype <span class="op">=</span> <span class="st">"solid"</span> <span class="co"># corresponds to the border of the image annotation</span>
<span class="op">)</span></code></pre></div>
<div class="figure">
<img src="spata-v2-image-annotation-screening_files/figure-html/unnamed-chunk-15-1.png" alt="Fig.9 Lineplots used to visualize the inferred gene expression." width="100%"><p class="caption">
Fig.9 Lineplots used to visualize the inferred gene expression.
</p>
</div>
<p>The function <code><a href="../reference/plotVolcano.html">plotVolcano()</a></code> makes use of the concept of
volcano plots as often used in DE-analyis to visualize the significance
as well as the log fold change of differentially expressed genes. Here
the left and the right side of the plot is used to display genes that
have inverse gene expression pattern.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotVolcano.html">plotVolcano</a></span><span class="op">(</span>
  object <span class="op">=</span> <span class="va">IAS_T313</span>,
  left <span class="op">=</span> <span class="st">"late_ascending"</span>, 
  right <span class="op">=</span> <span class="st">"immediate_descending"</span>,
  eval <span class="op">=</span> <span class="st">"ias_score"</span>,
  pval <span class="op">=</span> <span class="st">"p_value_mean_adjusted"</span>,
  threshold_eval <span class="op">=</span> <span class="fl">0.75</span>,
  threshold_pval <span class="op">=</span> <span class="fl">0.05</span>, 
  label_vars <span class="op">=</span> <span class="fl">5</span>, <span class="co"># label the top 5 genes </span>
  label_size <span class="op">=</span> <span class="fl">3</span>
  <span class="op">)</span></code></pre></div>
<div class="figure">
<img src="spata-v2-image-annotation-screening_files/figure-html/volcano_IAS-1.png" alt="Fig.10 Visualize genes that have an inverse spatial expression pattern." width="100%"><p class="caption">
Fig.10 Visualize genes that have an inverse spatial expression pattern.
</p>
</div>
</div>
<div class="section level3">
<h3 id="extract-results">5.2 Extract results<a class="anchor" aria-label="anchor" href="#extract-results"></a>
</h3>
<p>Results can be extracted as data.frames with
<code><a href="../reference/getResultsDf.html">getResultsDf()</a></code> or as character vectors of variable names
with <code><a href="../reference/getResultsVec.html">getResultsVec()</a></code>.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># complete results (left)</span>
<span class="va">res_df_all</span> <span class="op">&lt;-</span>
  <span class="fu"><a href="../reference/getResultsDf.html">getResultsDf</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">IAS_T313</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> 
  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fl">200</span><span class="op">)</span>

<span class="co"># subset results (right)</span>
<span class="va">res_df_best</span> <span class="op">&lt;-</span> 
  <span class="fu"><a href="../reference/getResultsDf.html">getResultsDf</a></span><span class="op">(</span>
    object <span class="op">=</span> <span class="va">IAS_T313</span>,
    threshold_eval <span class="op">=</span> <span class="fl">0.75</span>,
    threshold_pval <span class="op">=</span> <span class="fl">0.05</span>, 
    best_only <span class="op">=</span> <span class="cn">TRUE</span> <span class="co"># only best model fit</span>
  <span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> 
  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fl">200</span><span class="op">)</span></code></pre></div>
<div class="row">
<div class="col-md-6">
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 200 x 16</span></span>
<span class="co">## <span style="color: #949494;"># Groups:   models [1]</span></span>
<span class="co">##    variables models        ias_score n_bins_angle corr_mean corr_median corr_min</span>
<span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="co">## <span style="color: #BCBCBC;"> 1</span> IGFBP3    immediate_de~     0.907            1     0.927       0.927    0.927</span>
<span class="co">## <span style="color: #BCBCBC;"> 2</span> TNC       immediate_de~     0.894            1     0.904       0.904    0.904</span>
<span class="co">## <span style="color: #BCBCBC;"> 3</span> CD44      immediate_de~     0.886            1     0.934       0.934    0.934</span>
<span class="co">## <span style="color: #BCBCBC;"> 4</span> THBS2     immediate_de~     0.879            1     0.887       0.887    0.887</span>
<span class="co">## <span style="color: #BCBCBC;"> 5</span> GREM1     immediate_de~     0.878            1     0.979       0.979    0.979</span>
<span class="co">## <span style="color: #BCBCBC;"> 6</span> P4HA2     immediate_de~     0.877            1     0.917       0.917    0.917</span>
<span class="co">## <span style="color: #BCBCBC;"> 7</span> ARL4C     immediate_de~     0.876            1     0.918       0.918    0.918</span>
<span class="co">## <span style="color: #BCBCBC;"> 8</span> CTSK      immediate_de~     0.874            1     0.849       0.849    0.849</span>
<span class="co">## <span style="color: #BCBCBC;"> 9</span> MMP2      immediate_de~     0.873            1     0.869       0.869    0.869</span>
<span class="co">## <span style="color: #BCBCBC;">10</span> LOXL2     immediate_de~     0.870            1     0.846       0.846    0.846</span>
<span class="co">## <span style="color: #949494;"># i 190 more rows</span></span>
<span class="co">## <span style="color: #949494;"># i 9 more variables: corr_max &lt;dbl&gt;, corr_sd &lt;dbl&gt;, raoc_mean &lt;dbl&gt;,</span></span>
<span class="co">## <span style="color: #949494;">#   p_value_mean &lt;dbl&gt;, p_value_median &lt;dbl&gt;, p_value_combined &lt;dbl&gt;,</span></span>
<span class="co">## <span style="color: #949494;">#   p_value_mean_adjusted &lt;dbl&gt;, p_value_median_adjusted &lt;dbl&gt;,</span></span>
<span class="co">## <span style="color: #949494;">#   p_value_combined_adjusted &lt;dbl&gt;</span></span></code></pre>
</div>
<div class="col-md-6">
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 200 x 16</span></span>
<span class="co">## <span style="color: #949494;"># Groups:   models [2]</span></span>
<span class="co">##    variables models        ias_score n_bins_angle corr_mean corr_median corr_min</span>
<span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="co">## <span style="color: #BCBCBC;"> 1</span> IGFBP3    immediate_de~     0.907            1     0.927       0.927    0.927</span>
<span class="co">## <span style="color: #BCBCBC;"> 2</span> TNC       immediate_de~     0.894            1     0.904       0.904    0.904</span>
<span class="co">## <span style="color: #BCBCBC;"> 3</span> CD44      immediate_de~     0.886            1     0.934       0.934    0.934</span>
<span class="co">## <span style="color: #BCBCBC;"> 4</span> THBS2     immediate_de~     0.879            1     0.887       0.887    0.887</span>
<span class="co">## <span style="color: #BCBCBC;"> 5</span> GREM1     immediate_de~     0.878            1     0.979       0.979    0.979</span>
<span class="co">## <span style="color: #BCBCBC;"> 6</span> P4HA2     immediate_de~     0.877            1     0.917       0.917    0.917</span>
<span class="co">## <span style="color: #BCBCBC;"> 7</span> ARL4C     immediate_de~     0.876            1     0.918       0.918    0.918</span>
<span class="co">## <span style="color: #BCBCBC;"> 8</span> CTSK      immediate_de~     0.874            1     0.849       0.849    0.849</span>
<span class="co">## <span style="color: #BCBCBC;"> 9</span> MMP2      immediate_de~     0.873            1     0.869       0.869    0.869</span>
<span class="co">## <span style="color: #BCBCBC;">10</span> LOXL2     immediate_de~     0.870            1     0.846       0.846    0.846</span>
<span class="co">## <span style="color: #949494;"># i 190 more rows</span></span>
<span class="co">## <span style="color: #949494;"># i 9 more variables: corr_max &lt;dbl&gt;, corr_sd &lt;dbl&gt;, raoc_mean &lt;dbl&gt;,</span></span>
<span class="co">## <span style="color: #949494;">#   p_value_mean &lt;dbl&gt;, p_value_median &lt;dbl&gt;, p_value_combined &lt;dbl&gt;,</span></span>
<span class="co">## <span style="color: #949494;">#   p_value_mean_adjusted &lt;dbl&gt;, p_value_median_adjusted &lt;dbl&gt;,</span></span>
<span class="co">## <span style="color: #949494;">#   p_value_combined_adjusted &lt;dbl&gt;</span></span></code></pre>
</div>
</div>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">imm_desc_genes</span> <span class="op">&lt;-</span> 
  <span class="fu"><a href="../reference/getResultsVec.html">getResultsVec</a></span><span class="op">(</span>
    object <span class="op">=</span> <span class="va">IAS_T313</span>,
    model_subset <span class="op">=</span> <span class="st">"immediate_descending"</span>,
    threshold_eval <span class="op">=</span> <span class="fl">0.75</span>,
    threshold_pval <span class="op">=</span> <span class="fl">0.05</span>
  <span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> 
  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fl">8</span><span class="op">)</span> <span class="co"># keep top 8</span>

<span class="co">#show results</span>
<span class="va">imm_desc_genes</span></code></pre></div>
<pre><code><span class="co">## immediate_descending immediate_descending immediate_descending </span>
<span class="co">##             "IGFBP3"                "TNC"               "CD44" </span>
<span class="co">## immediate_descending immediate_descending immediate_descending </span>
<span class="co">##              "THBS2"              "GREM1"              "P4HA2" </span>
<span class="co">## immediate_descending immediate_descending </span>
<span class="co">##              "ARL4C"               "CTSK"</span></code></pre>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># plot results</span>
<span class="fu"><a href="../reference/plotSurfaceComparison.html">plotSurfaceComparison</a></span><span class="op">(</span>
  object <span class="op">=</span> <span class="va">object_t313</span>,
  color_by <span class="op">=</span> <span class="va">imm_desc_genes</span>,
  nrow <span class="op">=</span> <span class="fl">2</span>
<span class="op">)</span> <span class="op">+</span> 
  <span class="va">ias_layer_range</span></code></pre></div>
<div class="figure">
<img src="spata-v2-image-annotation-screening_files/figure-html/unnamed-chunk-19-1.png" alt="Fig.11 Top 8 genes that were identified as closely associated with necrosis." width="100%"><p class="caption">
Fig.11 Top 8 genes that were identified as closely associated with
necrosis.
</p>
</div>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">late_asc_genes</span> <span class="op">&lt;-</span> 
  <span class="fu"><a href="../reference/getResultsVec.html">getResultsVec</a></span><span class="op">(</span>
    object <span class="op">=</span> <span class="va">IAS_T313</span>,
    model_subset <span class="op">=</span> <span class="st">"late_ascending"</span>,
    threshold_eval <span class="op">=</span> <span class="fl">0.75</span>,
    threshold_pval <span class="op">=</span> <span class="fl">0.05</span>
  <span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> 
  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fl">8</span><span class="op">)</span> <span class="co"># keep top 8</span>

<span class="co">#show results</span>
<span class="va">late_asc_genes</span></code></pre></div>
<pre><code><span class="co">## late_ascending late_ascending late_ascending late_ascending late_ascending </span>
<span class="co">##         "UBL5"      "TRMT112"       "LAPTM5"       "UQCR11"        "PSME1" </span>
<span class="co">## late_ascending late_ascending late_ascending </span>
<span class="co">##       "PET100"     "C12orf57"    "TNFRSF12A"</span></code></pre>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># plot results</span>
<span class="fu"><a href="../reference/plotSurfaceComparison.html">plotSurfaceComparison</a></span><span class="op">(</span>
  object <span class="op">=</span> <span class="va">object_t313</span>,
  color_by <span class="op">=</span> <span class="va">late_asc_genes</span>,
  nrow <span class="op">=</span> <span class="fl">2</span>
<span class="op">)</span> <span class="op">+</span>
  <span class="va">ias_layer_range</span></code></pre></div>
<div class="figure">
<img src="spata-v2-image-annotation-screening_files/figure-html/unnamed-chunk-20-1.png" alt="Fig.12 Top 8 genes that were identified as repelled with necrosis." width="100%"><p class="caption">
Fig.12 Top 8 genes that were identified as repelled with necrosis.
</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="the-s4-imageannotationscreening-class">6. The S4 <code>ImageAnnotationScreening</code> class<a class="anchor" aria-label="anchor" href="#the-s4-imageannotationscreening-class"></a>
</h2>
<p>Image annotation screening results are stored in an S4 object of
class <code>ImageAnnotationScreening</code>. Use
<code><a href="../reference/ImageAnnotationScreening-class.html">?ImageAnnotationScreening</a></code> to read the description. Note
that <strong>I</strong>mageAnnotationScreening is the class name.
<strong>i</strong>mageAnnotationScreening() is the function that runs
the algorithm.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">IAS_T313</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] "ImageAnnotationScreening"</span>
<span class="co">## attr(,"package")</span>
<span class="co">## [1] "SPATA2"</span></code></pre>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">slotNames</span><span class="op">(</span><span class="va">IAS_T313</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">##  [1] "angle_span"      "binwidth"        "coords"          "distance"       </span>
<span class="co">##  [5] "img_annotation"  "info"            "method_padj"     "models"         </span>
<span class="co">##  [9] "n_bins_angle"    "n_bins_circle"   "results_primary" "results"        </span>
<span class="co">## [13] "sample"          "summarize_with"  "bcsp_exclude"</span></code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Jan Kueckelhaus, Dieter-Henrik Heiland, Simon Frerich.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
