<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Spatial Trajectory Screening • SPATA2</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Spatial Trajectory Screening">
<meta property="og:description" content="SPATA2">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SPATA2</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/spata-v2-installation.html">Installation</a>
</li>
<li>
  <a href="../articles/spata-v2-spata-data.html">SPATAData</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">Around the Spata-Object</li>
    <li>
      <a href="../articles/spata-v2-object-initiation-and-manipulation.html">Object Set Up and Manipulation</a>
    </li>
    <li>
      <a href="../articles/">Genes, Gene Sets and Features</a>
    </li>
    <li class="dropdown-header">Integrating Different Platforms</li>
    <li>
      <a href="../articles/spata-v2-cnv-analysis.html">Copy Number Variations</a>
    </li>
    <li>
      <a href="../articles/spata-v2-segmentation.html">Spatial Segmentation (Manual Annotation)</a>
    </li>
    <li>
      <a href="../articles/spata-v2-dea.html">Differential Expression Analysis</a>
    </li>
    <li>
      <a href="../articles/spata-v2-gsea.html">Gene Set Enrichment Analysis</a>
    </li>
    <li class="dropdown-header">Gradient Screening</li>
    <li>
      <a href="../articles/spata-v2-trajectory-analysis.html">Spatial Trajectories</a>
    </li>
    <li>
      <a href="../articles/spata-v2-spatial-trajectory-screening.html">Spatial Trajectory Screening</a>
    </li>
    <li>
      <a href="../articles/spata-v2-image-annotations.html">Image Annotations</a>
    </li>
    <li>
      <a href="../articles/spata-v2-image-annotation-screening.html">Image Annotation Screening</a>
    </li>
    <li class="dropdown-header">Visualization</li>
    <li>
      <a href="../articles/spata-v2-plotting-surface.html">Surface Plotting</a>
    </li>
    <li>
      <a href="../articles/spata-v2-plotting-misc.html">Miscellaneous</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Code &amp; Concept
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/cc-spatial-measures.html">Spatial Measures</a>
    </li>
    <li>
      <a href="../articles/cc-spatial-segmentation-vs-image-annotation.html">Spatial Segmentation vs. Image Annotations</a>
    </li>
    <li>
      <a href="../articles/cc-spatial-trajectory-screening.html">Spatial Trajectory Screening</a>
    </li>
    <li>
      <a href="../articles/cc-image-annotation-screening.html">Image Annotation Screening</a>
    </li>
    <li>
      <a href="../articles/cc-model-fitting-in-spatial-transcriptomic-studies.html">Model Fitting</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Spatial Trajectory Screening</h1>
            
      
      
      <div class="hidden name"><code>spata-v2-spatial-trajectory-screening.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="prerequisites">1. Prerequisites<a class="anchor" aria-label="anchor" href="#prerequisites"></a>
</h2>
<p>Make sure to be familiar with the following vignettes:</p>
<ul>
<li><a href="spata-v2-spatial-trajectories.html">Spatial
Trajectories</a></li>
<li><a href="cc-spatial-trajectory-screening.html">Spatial Trajectory
Screening - Code &amp; Concept</a></li>
<li><a href="cc-model-fitting-in-spatial-transcriptomics-studies.html">Model
Fitting in Spatial Transcriptomic Studies - Code &amp; Concept</a></li>
</ul>
</div>
<div class="section level2">
<h2 id="introduction-overview">2. Introduction &amp; overview<a class="anchor" aria-label="anchor" href="#introduction-overview"></a>
</h2>
<p>Spatial trajectory screening (STS) fits inferred expression changes
of genes to predefined models to screen for genes that follow
biologically relevant dynamics along a spatial trajectory. This tutorial
gives on overview about the most important functions.</p>
<p>Throughout the tutorial we are using the same example sample that we
used in the tutorial on <a href="">spatial trajectories</a> - the
glioblastoma sample T269.</p>
<p>As an example we are using a spatial transcriptomic sample of a
central nervous system malignancy that features three different,
adjacent histological areas: Tumor, a transition zone as well as
infiltrated cortex.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://themilolab.com/" class="external-link">SPATA2</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SPATAData</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span>

<span class="va">object_t269</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SPATAData/man/downloadSpataObject.html" class="external-link">downloadSpataObject</a></span><span class="op">(</span>sample_name <span class="op">=</span> <span class="st">"269_T"</span><span class="op">)</span>

<span class="co"># load example image annotations</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"image_annotations"</span><span class="op">)</span>

<span class="va">object_t269</span> <span class="op">&lt;-</span> 
  <span class="fu"><a href="../reference/setImageAnnotation.html">setImageAnnotations</a></span><span class="op">(</span>
    object <span class="op">=</span> <span class="va">object_t269</span>, 
    img_anns <span class="op">=</span> <span class="va">image_annotations</span><span class="op">[[</span><span class="st">"269_T"</span><span class="op">]</span><span class="op">]</span>,
    overwrite <span class="op">=</span> <span class="cn">TRUE</span>
  <span class="op">)</span>

<span class="co"># plot results</span>
<span class="fu"><a href="../reference/plotImageGgplot.html">plotImageGgplot</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">object_t269</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="../reference/ggpLayerFrameByCoords.html">ggpLayerFrameByCoords</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">object_t269</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="../reference/ggpLayerThemeCoords.html">ggpLayerThemeCoords</a></span><span class="op">(</span><span class="op">)</span>

<span class="fu"><a href="../reference/plotImageAnnotations.html">plotImageAnnotations</a></span><span class="op">(</span>
  object <span class="op">=</span> <span class="va">object_t269</span>,
  tags <span class="op">=</span> <span class="st">"hist_example"</span>, 
  square <span class="op">=</span> <span class="cn">TRUE</span>,
  expand <span class="op">=</span> <span class="fl">0.5</span>, 
  encircle <span class="op">=</span> <span class="cn">FALSE</span>, 
  nrow <span class="op">=</span> <span class="fl">2</span>, 
  display_caption <span class="op">=</span> <span class="cn">FALSE</span>
<span class="op">)</span></code></pre></div>
<div class="figure">
<img src="spata-v2-spatial-trajectory-screening_files/figure-html/unnamed-chunk-2-1.png" alt="Fig.1 Example sample T269." width="50%"><img src="spata-v2-spatial-trajectory-screening_files/figure-html/unnamed-chunk-2-2.png" alt="Fig.1 Example sample T269." width="50%"><p class="caption">
Fig.1 Example sample T269.
</p>
</div>
<p>The trajectory stays the same, too.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># adds the trajectory that is stored in `spatial_trajectories[["269_T"]][["horizontal_mid"]]`</span>
<span class="va">object_t269</span> <span class="op">&lt;-</span> 
    <span class="fu"><a href="../reference/createSpatialTrajectories.html">addSpatialTrajectory</a></span><span class="op">(</span>
    object <span class="op">=</span> <span class="va">object_t269</span>,
    id <span class="op">=</span> <span class="st">"horizontal_mid"</span>,
    start <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"1.5mm"</span>, <span class="st">"3.5mm"</span><span class="op">)</span> ,
    end <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"6.5mm"</span>, <span class="st">"3.5mm"</span><span class="op">)</span> ,
    width <span class="op">=</span> <span class="st">"2mm"</span>,
    overwrite <span class="op">=</span> <span class="cn">TRUE</span>
  <span class="op">)</span>

<span class="fu"><a href="../reference/plotSpatialTrajectories.html">plotSpatialTrajectories</a></span><span class="op">(</span>
  object <span class="op">=</span> <span class="va">object_t269</span>, 
  ids <span class="op">=</span> <span class="st">"horizontal_mid"</span>,
  color_by <span class="op">=</span> <span class="st">"MBP"</span>
<span class="op">)</span></code></pre></div>
<div class="figure">
<img src="spata-v2-spatial-trajectory-screening_files/figure-html/unnamed-chunk-3-1.png" alt="Fig.2 Example spatial trajectory." width="100%"><p class="caption">
Fig.2 Example spatial trajectory.
</p>
</div>
</div>
<div class="section level2">
<h2 id="run-the-algorithm">3. Run the algorithm<a class="anchor" aria-label="anchor" href="#run-the-algorithm"></a>
</h2>
<p>The function to use is called
<code><a href="../reference/spatialTrajectoryScreening.html">spatialTrajectoryScreening()</a></code>. The parameter
<code>variables</code> takes the numeric variables that are supposed to
be included in the screening process. Usually it’s gene names. Spatial
trajectory screening, however, is not restricted to genes as virtually
every numeric variable (e.g. <em>nCount_Genes</em>,
<em>nCount_Features</em>, <em>gene sets</em>, etc.) can be fitted to
spatially meaningful models, too. Therefore, the argument for the input
is simply called variables.</p>
<p>Here, we are using the genes that were already identified as
spatially variable by <code>SPARKX</code>. The goal is to further
analyze which of the genes are expressed in a meaningful way along the
trajectory.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># this is a wrapper around SPARK::sparkx()</span>
<span class="va">object_t269</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runSparkx.html">runSparkx</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">object_t269</span><span class="op">)</span>

<span class="va">spark_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runSparkx.html">getSparkxGeneDf</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">object_t269</span>, threshold_pval <span class="op">=</span> <span class="fl">0.01</span><span class="op">)</span>

<span class="co"># extract genes with a sparkx pvalue of 0.01 or lower</span>
<span class="va">sparkx_genes</span> <span class="op">&lt;-</span> <span class="va">spark_df</span><span class="op">[[</span><span class="st">"genes"</span><span class="op">]</span><span class="op">]</span>

<span class="co"># show results</span>
<span class="va">spark_df</span>

<span class="co"># show results (&gt; 10 000 genes detected as spatially variable with a p-value of &lt; 0.01)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">sparkx_genes</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 13,839 x 3</span></span>
<span class="co">##    genes    combinedPval adjustedPval</span>
<span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="co">## <span style="color: #BCBCBC;"> 1</span> ID3                 0            0</span>
<span class="co">## <span style="color: #BCBCBC;"> 2</span> MARCKSL1            0            0</span>
<span class="co">## <span style="color: #BCBCBC;"> 3</span> PHC2                0            0</span>
<span class="co">## <span style="color: #BCBCBC;"> 4</span> GNG5                0            0</span>
<span class="co">## <span style="color: #BCBCBC;"> 5</span> CNN3                0            0</span>
<span class="co">## <span style="color: #BCBCBC;"> 6</span> RHOC                0            0</span>
<span class="co">## <span style="color: #BCBCBC;"> 7</span> TXNIP               0            0</span>
<span class="co">## <span style="color: #BCBCBC;"> 8</span> S100A10             0            0</span>
<span class="co">## <span style="color: #BCBCBC;"> 9</span> S100A16             0            0</span>
<span class="co">## <span style="color: #BCBCBC;">10</span> C1orf61             0            0</span>
<span class="co">## <span style="color: #949494;"># i 13,829 more rows</span></span></code></pre>
<pre><code><span class="co">##  chr [1:13839] "ID3" "MARCKSL1" "PHC2" "GNG5" "CNN3" "RHOC" "TXNIP" ...</span></code></pre>
<p>Once you’ve decided on the input you can run the algorithm.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># catchphrases to subset the models that are of interest (use showModels() to check) </span>
<span class="va">model_subset</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sharp_peak"</span>, <span class="st">"abrupt_ascending"</span>, <span class="st">"abrupt_descending"</span><span class="op">)</span> 

<span class="va">STS_T269</span> <span class="op">&lt;-</span> 
  <span class="fu"><a href="../reference/spatialTrajectoryScreening.html">spatialTrajectoryScreening</a></span><span class="op">(</span>
    object <span class="op">=</span> <span class="va">object_t269</span>, 
    id <span class="op">=</span> <span class="st">"horizontal_mid"</span>, <span class="co"># ID of the spatial trajectory</span>
    variables <span class="op">=</span> <span class="va">sparkx_genes</span>, <span class="co"># the variables/genes to scree </span>
    model_subset <span class="op">=</span> <span class="va">model_subset</span> 
  <span class="op">)</span></code></pre></div>
<p><strong>Note</strong>: The output of
<code><a href="../reference/spatialTrajectoryScreening.html">spatialTrajectoryScreening()</a></code> is <strong>not</strong> saved
in the <code>spata2</code> object but returned in a separate S4 object
of class <code>SpatialTrajectoryScreening</code>. Do
<strong>not</strong> overwrite the <code>spata2</code> object by writing
<code>object_t269 &lt;- spatialTrajectoryScreening(object = object_t269, id = "horizontal_mid", ...)</code>.</p>
</div>
<div class="section level2">
<h2 id="results">4. Results<a class="anchor" aria-label="anchor" href="#results"></a>
</h2>
<div class="section level3">
<h3 id="visualization">4.1 Visualization<a class="anchor" aria-label="anchor" href="#visualization"></a>
</h3>
<p>To get an overview of the screening as well as a first visualization
of the results use <code><a href="../reference/plotOverview.html">plotOverview()</a></code>. It sorts genes by their
best model-fit and plots the evaluation score against the p-value for
every model.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotOverview.html">plotOverview</a></span><span class="op">(</span>
  object <span class="op">=</span> <span class="va">STS_T269</span>,
  label_vars <span class="op">=</span> <span class="fl">4</span>, <span class="co"># label top 4 variables/genes per model</span>
  label_size <span class="op">=</span> <span class="fl">3</span>
<span class="op">)</span></code></pre></div>
<div class="figure">
<img src="spata-v2-spatial-trajectory-screening_files/figure-html/unnamed-chunk-8-1.png" alt="Fig.4 Overview of the screening results." width="100%"><p class="caption">
Fig.4 Overview of the screening results.
</p>
</div>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">traj_layer</span> <span class="op">&lt;-</span> 
  <span class="fu"><a href="../reference/ggpLayerTrajectories.html">ggpLayerTrajectories</a></span><span class="op">(</span>
    object <span class="op">=</span> <span class="va">object_t269</span>, 
    ids <span class="op">=</span> <span class="st">"horizontal_mid"</span>, 
    size <span class="op">=</span> <span class="fl">2.5</span>
    <span class="op">)</span>

<span class="co"># abrupt descending</span>
<span class="fu"><a href="../reference/plotSurfaceComparison.html">plotSurfaceComparison</a></span><span class="op">(</span>
  object <span class="op">=</span> <span class="va">object_t269</span>, 
  color_by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"EGFR"</span>, <span class="st">"DDR1"</span><span class="op">)</span>
<span class="op">)</span> <span class="op">+</span> 
  <span class="va">traj_layer</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>subtitle <span class="op">=</span> <span class="st">"a) Abrupt descending"</span><span class="op">)</span>

<span class="co"># sharp peak</span>
<span class="fu"><a href="../reference/plotSurfaceComparison.html">plotSurfaceComparison</a></span><span class="op">(</span>
  object <span class="op">=</span> <span class="va">object_t269</span>, 
  color_by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CNTN2"</span>, <span class="st">"SLC31A2"</span><span class="op">)</span>
<span class="op">)</span> <span class="op">+</span> 
  <span class="va">traj_layer</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>subtitle <span class="op">=</span> <span class="st">"b) Sharp peak"</span><span class="op">)</span>

<span class="co"># abrupt ascending</span>
<span class="fu"><a href="../reference/plotSurfaceComparison.html">plotSurfaceComparison</a></span><span class="op">(</span>
  object <span class="op">=</span> <span class="va">object_t269</span>, 
  color_by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"BEX4"</span>, <span class="st">"KIF1A"</span><span class="op">)</span>
<span class="op">)</span> <span class="op">+</span> 
  <span class="va">traj_layer</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>subtitle <span class="op">=</span> <span class="st">"c) Abrupt ascending"</span><span class="op">)</span></code></pre></div>
<div class="figure">
<img src="spata-v2-spatial-trajectory-screening_files/figure-html/unnamed-chunk-9-1.png" alt="Fig.5 Surface plots of screening results." width="100%"><img src="spata-v2-spatial-trajectory-screening_files/figure-html/unnamed-chunk-9-2.png" alt="Fig.5 Surface plots of screening results." width="100%"><img src="spata-v2-spatial-trajectory-screening_files/figure-html/unnamed-chunk-9-3.png" alt="Fig.5 Surface plots of screening results." width="100%"><p class="caption">
Fig.5 Surface plots of screening results.
</p>
</div>
<p>The inferred gene expression can be visualized with
<code><a href="../reference/plotTrajectoryLineplot.html">plotTrajectoryLineplot()</a></code>.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotTrajectoryLineplot.html">plotTrajectoryLineplot</a></span><span class="op">(</span>
  object <span class="op">=</span> <span class="va">object_t269</span>, 
  id <span class="op">=</span> <span class="st">"horizontal_mid"</span>, 
  variables <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"EGFR"</span>, <span class="st">"DDR1"</span>, <span class="st">"CNTN2"</span>, <span class="st">"SLC31A2"</span>, <span class="st">"BEX4"</span>, <span class="st">"KIF1A"</span><span class="op">)</span>, 
  clrp <span class="op">=</span> <span class="st">"sifre"</span>,
  nrow <span class="op">=</span> <span class="fl">2</span>
<span class="op">)</span></code></pre></div>
<div class="figure">
<img src="spata-v2-spatial-trajectory-screening_files/figure-html/unnamed-chunk-10-1.png" alt="Fig.6 Trajectory lineplots of inferred expression changes." width="100%"><p class="caption">
Fig.6 Trajectory lineplots of inferred expression changes.
</p>
</div>
<p>The used models can always be visualized with
<code><a href="../reference/showColors.html">showModels()</a></code>.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/showColors.html">showModels</a></span><span class="op">(</span>model_subset <span class="op">=</span> <span class="va">model_subset</span><span class="op">)</span></code></pre></div>
<div class="figure">
<img src="spata-v2-spatial-trajectory-screening_files/figure-html/unnamed-chunk-11-1.png" alt="Fig.7 The models that were screened for." width="100%"><p class="caption">
Fig.7 The models that were screened for.
</p>
</div>
</div>
<div class="section level3">
<h3 id="extract-results">4.2 Extract results<a class="anchor" aria-label="anchor" href="#extract-results"></a>
</h3>
<p>Results can be extracted as data.frames with
<code><a href="../reference/getResultsDf.html">getResultsDf()</a></code> or as a character vector of variable/gene
names with <code><a href="../reference/getResultsVec.html">getResultsVec()</a></code>.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">res_df</span> <span class="op">&lt;-</span> 
  <span class="fu"><a href="../reference/getResultsDf.html">getResultsDf</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">STS_T269</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> 
  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fl">200</span><span class="op">)</span>

<span class="va">res_df</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 200 x 8</span></span>
<span class="co">## <span style="color: #949494;"># Groups:   models [1]</span></span>
<span class="co">##    variables models        sts_score  corr  raoc  p_value  rauc p_value_adjusted</span>
<span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="co">## <span style="color: #BCBCBC;"> 1</span> KIF1A     abrupt_ascen~     0.942 0.970 0.914 3.73<span style="color: #949494;">e</span><span style="color: #BB0000;">-13</span>  1.81         3.44<span style="color: #949494;">e</span><span style="color: #BB0000;">-10</span></span>
<span class="co">## <span style="color: #BCBCBC;"> 2</span> SLC1A2    abrupt_ascen~     0.940 0.969 0.912 5.28<span style="color: #949494;">e</span><span style="color: #BB0000;">-13</span>  1.85         4.39<span style="color: #949494;">e</span><span style="color: #BB0000;">-10</span></span>
<span class="co">## <span style="color: #BCBCBC;"> 3</span> CKB       abrupt_ascen~     0.936 0.969 0.903 5.83<span style="color: #949494;">e</span><span style="color: #BB0000;">-13</span>  2.04         4.66<span style="color: #949494;">e</span><span style="color: #BB0000;">-10</span></span>
<span class="co">## <span style="color: #BCBCBC;"> 4</span> BEX4      abrupt_ascen~     0.931 0.974 0.887 9.47<span style="color: #949494;">e</span><span style="color: #BB0000;">-14</span>  2.37         1.27<span style="color: #949494;">e</span><span style="color: #BB0000;">-10</span></span>
<span class="co">## <span style="color: #BCBCBC;"> 5</span> ATP1B1    abrupt_ascen~     0.929 0.962 0.895 3.27<span style="color: #949494;">e</span><span style="color: #BB0000;">-12</span>  2.21         1.63<span style="color: #949494;">e</span><span style="color: #BB0000;">- 9</span></span>
<span class="co">## <span style="color: #BCBCBC;"> 6</span> RTN4      abrupt_ascen~     0.928 0.973 0.883 1.61<span style="color: #949494;">e</span><span style="color: #BB0000;">-13</span>  2.45         1.89<span style="color: #949494;">e</span><span style="color: #BB0000;">-10</span></span>
<span class="co">## <span style="color: #BCBCBC;"> 7</span> BEX2      abrupt_ascen~     0.928 0.963 0.893 2.90<span style="color: #949494;">e</span><span style="color: #BB0000;">-12</span>  2.25         1.52<span style="color: #949494;">e</span><span style="color: #BB0000;">- 9</span></span>
<span class="co">## <span style="color: #BCBCBC;"> 8</span> SLC17A7   abrupt_ascen~     0.927 0.959 0.896 7.33<span style="color: #949494;">e</span><span style="color: #BB0000;">-12</span>  2.19         2.83<span style="color: #949494;">e</span><span style="color: #BB0000;">- 9</span></span>
<span class="co">## <span style="color: #BCBCBC;"> 9</span> RAB11FIP4 abrupt_ascen~     0.924 0.948 0.900 7.31<span style="color: #949494;">e</span><span style="color: #BB0000;">-11</span>  2.10         1.56<span style="color: #949494;">e</span><span style="color: #BB0000;">- 8</span></span>
<span class="co">## <span style="color: #BCBCBC;">10</span> CALM2     abrupt_ascen~     0.924 0.956 0.891 1.41<span style="color: #949494;">e</span><span style="color: #BB0000;">-11</span>  2.29         4.54<span style="color: #949494;">e</span><span style="color: #BB0000;">- 9</span></span>
<span class="co">## <span style="color: #949494;"># i 190 more rows</span></span></code></pre>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># top 3 peaking genes</span>
<span class="va">peaking_genes</span> <span class="op">&lt;-</span> 
  <span class="fu"><a href="../reference/getResultsVec.html">getResultsVec</a></span><span class="op">(</span>
    object <span class="op">=</span> <span class="va">STS_T269</span>, 
    threshold_eval <span class="op">=</span> <span class="fl">0.5</span>,
    threshold_pval <span class="op">=</span> <span class="fl">0.05</span>,
    model_subset <span class="op">=</span> <span class="st">"sharp_peak"</span>
  <span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> 
  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span>

<span class="fu"><a href="../reference/plotSurfaceComparison.html">plotSurfaceComparison</a></span><span class="op">(</span>
  object <span class="op">=</span> <span class="va">object_t269</span>, 
  color_by <span class="op">=</span> <span class="va">peaking_genes</span>
<span class="op">)</span></code></pre></div>
<div class="figure">
<img src="spata-v2-spatial-trajectory-screening_files/figure-html/unnamed-chunk-13-1.png" alt="Fig.8 Surface plots of screening results." width="100%"><p class="caption">
Fig.8 Surface plots of screening results.
</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="the-s4-spatialtrajectoryscreening-class">5. The S4 <code>SpatialTrajectoryScreening</code> class<a class="anchor" aria-label="anchor" href="#the-s4-spatialtrajectoryscreening-class"></a>
</h2>
<p>Spatial trajectory screening results are provided in an S4 object of
class <code>SpatialTrajectoryScreening</code>. Use
<code><a href="../reference/SpatialTrajectoryScreening-class.html">?SpatialTrajectoryScreening</a></code> to read the description. Note
that <strong>S</strong>patialTrajectoryScreening is the class name.
<strong>s</strong>patialTrajectoryScreening() is the function that runs
the algorithm.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">STS_T269</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] "SpatialTrajectoryScreening"</span>
<span class="co">## attr(,"package")</span>
<span class="co">## [1] "SPATA2"</span></code></pre>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">slotNames</span><span class="op">(</span><span class="va">STS_T269</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">##  [1] "binwidth"           "coords"             "id"                </span>
<span class="co">##  [4] "method_padj"        "models"             "n_bins"            </span>
<span class="co">##  [7] "results"            "sample"             "summarize_with"    </span>
<span class="co">## [10] "spatial_trajectory"</span></code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Jan Kueckelhaus, Dieter-Henrik Heiland.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
