<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Spatial Segmentation • SPATA2</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Spatial Segmentation">
<meta property="og:description" content="SPATA2">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SPATA2</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/spata-v2-installation.html">Installation</a>
</li>
<li>
  <a href="../articles/spata-v2-spata-data.html">SPATAData</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">Around the Spata-Object</li>
    <li>
      <a href="../articles/spata-v2-object-initiation-and-manipulation.html">Object Set Up and Manipulation</a>
    </li>
    <li>
      <a href="../articles/">Genes, Gene Sets and Features</a>
    </li>
    <li class="dropdown-header">Integrating Different Platforms</li>
    <li>
      <a href="../articles/spata-v2-cnv-analysis.html">Copy Number Variations</a>
    </li>
    <li>
      <a href="../articles/spata-v2-segmentation.html">Spatial Segmentation (Manual Annotation)</a>
    </li>
    <li>
      <a href="../articles/spata-v2-dea.html">Differential Expression Analysis</a>
    </li>
    <li>
      <a href="../articles/spata-v2-gsea.html">Gene Set Enrichment Analysis</a>
    </li>
    <li class="dropdown-header">Gradient Screening</li>
    <li>
      <a href="../articles/spata-v2-trajectory-analysis.html">Spatial Trajectories</a>
    </li>
    <li>
      <a href="../articles/spata-v2-spatial-trajectory-screening.html">Spatial Trajectory Screening</a>
    </li>
    <li>
      <a href="../articles/spata-v2-image-annotations.html">Image Annotations</a>
    </li>
    <li>
      <a href="../articles/spata-v2-image-annotation-screening.html">Image Annotation Screening</a>
    </li>
    <li class="dropdown-header">Visualization</li>
    <li>
      <a href="../articles/spata-v2-plotting-surface.html">Surface Plotting</a>
    </li>
    <li>
      <a href="../articles/spata-v2-plotting-misc.html">Miscellaneous</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Code &amp; Concept
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/cc-spatial-measures.html">Spatial Measures</a>
    </li>
    <li>
      <a href="../articles/cc-spatial-segmentation-vs-image-annotation.html">Spatial Segmentation vs. Image Annotations</a>
    </li>
    <li>
      <a href="../articles/cc-spatial-trajectory-screening.html">Spatial Trajectory Screening</a>
    </li>
    <li>
      <a href="../articles/cc-image-annotation-screening.html">Image Annotation Screening</a>
    </li>
    <li>
      <a href="../articles/cc-model-fitting-in-spatial-transcriptomic-studies.html">Model Fitting</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Spatial Segmentation</h1>
            
      
      
      <div class="hidden name"><code>spata-v2-spatial-segmentation.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="prerequisites">1. Prerequisites<a class="anchor" aria-label="anchor" href="#prerequisites"></a>
</h2>
<p>Make sure to be familiar with the following vignettes:</p>
<ul>
<li><a href="cc-spatial-segmentation-vs-image-annotations.html">Spatial
Segmentation vs. Image Annotations</a></li>
</ul>
</div>
<div class="section level2">
<h2 id="introduction-overview">2. Introduction &amp; overview<a class="anchor" aria-label="anchor" href="#introduction-overview"></a>
</h2>
<p>Spatial transcriptomic samples come along with an underlying
histology image. To integrate the histological classification of the
area a specific barcode-spot covers the sample can be manually segmented
with <code>SPATA2</code>.</p>
<p>As an example we are using a spatial transcriptomic sample of a
central nervous system malignancy that features three different,
adjacent histological areas: Tumor, a transition zone as well as
infiltrated cortex.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://themilolab.com/" class="external-link">SPATA2</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SPATAData</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span>

<span class="va">object_t269</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SPATAData/man/downloadSpataObject.html" class="external-link">downloadSpataObject</a></span><span class="op">(</span>sample_name <span class="op">=</span> <span class="st">"269_T"</span><span class="op">)</span>

<span class="co"># load example segmentations</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"spatial_segmentations"</span><span class="op">)</span>

<span class="va">object_t269</span> <span class="op">&lt;-</span> 
  <span class="fu"><a href="../reference/addFeatures.html">addFeatures</a></span><span class="op">(</span>
    object <span class="op">=</span> <span class="va">object_t269</span>, 
    feature_df <span class="op">=</span> <span class="va">spatial_segmentations</span><span class="op">[[</span><span class="st">"269_T"</span><span class="op">]</span><span class="op">]</span>, 
    overwrite <span class="op">=</span> <span class="cn">TRUE</span>
  <span class="op">)</span>

<span class="co"># plot results</span>
<span class="fu"><a href="../reference/plotImageGgplot.html">plotImageGgplot</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">object_t269</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="../reference/ggpLayerFrameByCoords.html">ggpLayerFrameByCoords</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">object_t269</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="../reference/ggpLayerThemeCoords.html">ggpLayerThemeCoords</a></span><span class="op">(</span><span class="op">)</span>

<span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">object_t269</span>, color_by <span class="op">=</span> <span class="st">"histology"</span><span class="op">)</span></code></pre></div>
<div class="figure">
<img src="spata-v2-spatial-segmentation_files/figure-html/unnamed-chunk-2-1.png" alt="Fig.1 Example sample T269." width="50%"><img src="spata-v2-spatial-segmentation_files/figure-html/unnamed-chunk-2-2.png" alt="Fig.1 Example sample T269." width="50%"><p class="caption">
Fig.1 Example sample T269.
</p>
</div>
<p>The spatial segmentation based on histology is stored in grouping
variable <em>‘histology’</em> how to create these histology based
grouping variables is explained in this tutorial.</p>
</div>
<div class="section level2">
<h2 id="how-spatial-segmentation-is-stored">3. How spatial segmentation is stored<a class="anchor" aria-label="anchor" href="#how-spatial-segmentation-is-stored"></a>
</h2>
<p>Grouping variables in transcriptomic studies are usually created by
cluster algorithms and are stored in form of factors in the feature
data.frame (or meta data.frame in case of <code>Seurat</code> objects).
As spatial segmentation is a grouping variable, too, it is stored there
as well.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/getFeatureDf.html">getFeatureDf</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">object_t269</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> 
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">barcodes</span>, <span class="fu"><a href="https://tidyselect.r-lib.org/reference/where.html" class="external-link">where</a></span><span class="op">(</span><span class="va">is.factor</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 3,213 x 3</span></span>
<span class="co">##    barcodes           seurat_clusters histology  </span>
<span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>              <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>      </span>
<span class="co">## <span style="color: #BCBCBC;"> 1</span> AAACAAGTATCTCCCA-1 2               infiltrated</span>
<span class="co">## <span style="color: #BCBCBC;"> 2</span> AAACACCAATAACTGC-1 0               tumor      </span>
<span class="co">## <span style="color: #BCBCBC;"> 3</span> AAACAGAGCGACTCCT-1 5               infiltrated</span>
<span class="co">## <span style="color: #BCBCBC;"> 4</span> AAACATTTCCCGGATT-1 5               infiltrated</span>
<span class="co">## <span style="color: #BCBCBC;"> 5</span> AAACCCGAACGAAATC-1 2               infiltrated</span>
<span class="co">## <span style="color: #BCBCBC;"> 6</span> AAACCGGGTAGGTACC-1 7               tumor      </span>
<span class="co">## <span style="color: #BCBCBC;"> 7</span> AAACCGTTCGTCCAGG-1 0               tumor      </span>
<span class="co">## <span style="color: #BCBCBC;"> 8</span> AAACCTAAGCAGCCGG-1 1               infiltrated</span>
<span class="co">## <span style="color: #BCBCBC;"> 9</span> AAACCTCATGAAGTTG-1 7               tumor      </span>
<span class="co">## <span style="color: #BCBCBC;">10</span> AAACGAGACGGTTGAT-1 1               infiltrated</span>
<span class="co">## <span style="color: #949494;"># i 3,203 more rows</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="how-the-sample-can-be-segmented">4. How the sample can be segmented<a class="anchor" aria-label="anchor" href="#how-the-sample-can-be-segmented"></a>
</h2>
<p>The interactive segmentation can be done with the function
<code><a href="../reference/createSpatialSegmentation.html">createSpatialSegmentation()</a></code>. This function gives access to
an interface in which the respective segments/areas of the sample can be
encircled and labeled such that, eventually, all barcode-spots are
labeled.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># discard the one added above, to create new one</span>
<span class="va">object_t269</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/discardFeatures.html">discardFeatures</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">object_t269</span>, feature_names <span class="op">=</span> <span class="st">"histology"</span><span class="op">)</span>

<span class="va">object_t269</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createSpatialSegmentation.html">createSpatialSegmentation</a></span><span class="op">(</span><span class="va">object_t269</span><span class="op">)</span></code></pre></div>
<div class="figure">
<img src="images/createSpatialSegmentation_interface.JPG" alt="Fig.2 The whole interface of createSpatialSegmentation()." width="100%"><p class="caption">
Fig.2 The whole interface of createSpatialSegmentation().
</p>
</div>
<div class="figure">
<img src="images/createSpatialSegmentation_create_segm_var.JPG" alt="Fig.3 No segmentation variable exists." width="100%"><p class="caption">
Fig.3 No segmentation variable exists.
</p>
</div>
<p>At first, there is no plot displayed on the left. This is because
there is currently no segmentation variable in the feature data.frame
that can be worked on.</p>
<div class="section level3">
<h3 id="create-new-segmentation-variables">3.1 Create new segmentation variables<a class="anchor" aria-label="anchor" href="#create-new-segmentation-variables"></a>
</h3>
<p>To create a new segmentation variable click on ‘Create new
segmentation variable’. You are then prompted to enter the name that you
want to give the new variable. Here, we simply call it
<em>histology</em>.</p>
<div class="figure">
<img src="images/createSpatialSegmentation_enter_segm_var.JPG" alt="Fig.4 Enter segmentation name." width="100%"><p class="caption">
Fig.4 Enter segmentation name.
</p>
</div>
<p>After clicking on <em>Add segmentation variable</em> a new variable
is created in the feature data.frame and all of it’s values are
<em>‘unnamed’</em>. This is the default, has we haven’t segmented
anything yet.</p>
<div class="figure">
<img src="images/createSpatialSegmentation_new_segm_var.JPG" alt="Fig.5 New segmentation variable 'histology' has been created." width="100%"><p class="caption">
Fig.5 New segmentation variable ‘histology’ has been created.
</p>
</div>
</div>
<div class="section level3">
<h3 id="work-on-segmentation-variables">3.2 Work on segmentation variables<a class="anchor" aria-label="anchor" href="#work-on-segmentation-variables"></a>
</h3>
<p>Working on segmentation variables means to consecutively label all
barcode-spots depending on the histological area they cover. To achieve
this, encircle the areas that you want to label on the Interaction-plot.
Start drawing by double clicking on the plot or with the shortcut on
your keyboard <em>d</em>. If you are drawing the plot tells you.</p>
<div class="figure">
<img src="images/interactive_segmentation_1.png" alt="Fig.6 Drawing mode - On." width="100%"><p class="caption">
Fig.6 Drawing mode - On.
</p>
</div>
<p>By double-clicking again you leave the drawing mode. This can be used
to zoom in or out to reorientate. If you double-click again the endpoint
of the line drawn will connect to the position of your cursor and you
can continue drawing. Encircle the area till start and endpoint of the
line are close enough to be connected such that the circle closes. To
close the circle click on ‘Highlight’. This highlights the barcode-spots
that are encircled this way.</p>
<div class="figure">
<img src="images/interactive_segmentation_2.JPG" alt="Fig.7 Area encircled." width="100%"><p class="caption">
Fig.7 Area encircled.
</p>
</div>
<div class="figure">
<img src="images/interactive_segmentation_3.JPG" alt="Highlighted barcode-spots." width="100%"><p class="caption">
Highlighted barcode-spots.
</p>
</div>
<p>After highlighting the barcode-spots the name that you want to label
them with can be entered below the ‘Highlight’-button. Click on ‘Name’
to save the results. The results should be immediately displayed in the
plot on the left.</p>
<div class="figure">
<img src="images/interactive_segmentation_4.JPG" alt="Fig.8 Label is stored in the grouping variable." width="100%"><p class="caption">
Fig.8 Label is stored in the grouping variable.
</p>
</div>
<p>This can be done over and over again till you are satisfied. Note
that the label you gave a barcode-spot is overwritten if you include it
again in another area.</p>
<p>(The example below simply visualizes the concept of segmenting an
area that is located inside another one. As there is nothing to actually
label on the image the results were not saved in the output
segmentation.)</p>
<img src="images/interactive_segmentation_5.JPG" width="100%"><div class="figure">
<img src="images/interactive_segmentation_6.JPG" alt="Fig.9 &amp; 10 Encircle an area within an area." width="100%"><p class="caption">
Fig.9 &amp; 10 Encircle an area within an area.
</p>
</div>
<p>If you want to encircle an area <strong>within</strong> an area you
must label the bigger area first and then go into the details as the
last label is the one that stays.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">object_t269</span>, color_by <span class="op">=</span> <span class="st">"histology"</span>, pt_clrp <span class="op">=</span> <span class="st">"npg"</span><span class="op">)</span></code></pre></div>
<div class="figure">
<img src="spata-v2-spatial-segmentation_files/figure-html/unnamed-chunk-15-1.png" alt="Fig.11 The resulting grouping variable on the surface." width="100%"><p class="caption">
Fig.11 The resulting grouping variable on the surface.
</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="working-with-segmentation-variables">4. Working with segmentation variables<a class="anchor" aria-label="anchor" href="#working-with-segmentation-variables"></a>
</h2>
<p>Segmentation variables can be used like any other grouping variable
and are usually referred to via the arguments
<code>grouping_variable</code>, <code>grouping_variables</code> or
<code>across</code> or both.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># names of grouping variables</span>
<span class="fu"><a href="../reference/getGroupingOptions.html">getGroupingOptions</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">object_t269</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">##            factor            factor </span>
<span class="co">## "seurat_clusters"       "histology"</span></code></pre>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># names of groups within the grouping variable</span>
<span class="fu"><a href="../reference/getGroupNames.html">getGroupNames</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">object_t269</span>, grouping_variable <span class="op">=</span> <span class="st">"histology"</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] "tumor"       "transition"  "infiltrated"</span></code></pre>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotSurfaceComparison.html">plotSurfaceComparison</a></span><span class="op">(</span>
  object <span class="op">=</span> <span class="va">object_t269</span>,
  color_by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"B2M"</span>, <span class="st">"MBP"</span><span class="op">)</span>,
  ncol <span class="op">=</span> <span class="fl">1</span>
<span class="op">)</span>

<span class="fu"><a href="../reference/plotBoxplot.html">plotViolinplot</a></span><span class="op">(</span>
  object <span class="op">=</span> <span class="va">object_t269</span>, 
  variables <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"B2M"</span>, <span class="st">"MBP"</span><span class="op">)</span>, 
  across <span class="op">=</span> <span class="st">"histology"</span>, 
  ncol <span class="op">=</span> <span class="fl">1</span>, 
  clrp <span class="op">=</span> <span class="st">"npg"</span>
<span class="op">)</span></code></pre></div>
<div class="figure">
<img src="spata-v2-spatial-segmentation_files/figure-html/unnamed-chunk-17-1.png" alt="Fig.12 Working with segmentation. Example 1." width="50%"><img src="spata-v2-spatial-segmentation_files/figure-html/unnamed-chunk-17-2.png" alt="Fig.12 Working with segmentation. Example 1." width="50%"><p class="caption">
Fig.12 Working with segmentation. Example 1.
</p>
</div>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotBarchart.html">plotBarchart</a></span><span class="op">(</span>
  object <span class="op">=</span> <span class="va">object_t269</span>, 
  grouping_variables <span class="op">=</span> <span class="st">"seurat_clusters"</span>,
  across <span class="op">=</span> <span class="st">"histology"</span>, 
  clrp <span class="op">=</span> <span class="st">"npg"</span>
<span class="op">)</span>

<span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span>
  object <span class="op">=</span> <span class="va">object_t269</span>,
  color_by <span class="op">=</span> <span class="st">"seurat_clusters"</span>,
  pt_clrp <span class="op">=</span> <span class="st">"lo"</span>
<span class="op">)</span></code></pre></div>
<div class="figure">
<img src="spata-v2-spatial-segmentation_files/figure-html/unnamed-chunk-18-1.png" alt="Fig.13 Working with segmentation. Example 2." width="50%"><img src="spata-v2-spatial-segmentation_files/figure-html/unnamed-chunk-18-2.png" alt="Fig.13 Working with segmentation. Example 2." width="50%"><p class="caption">
Fig.13 Working with segmentation. Example 2.
</p>
</div>
<p><a href="spata-v2-dea.html">Differential expression analysis</a>
based on histology becomes possible, too, this way.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># run DEA based on histology</span>
<span class="va">object_t269</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runDEA.html">runDEA</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">object_t269</span>, across <span class="op">=</span> <span class="st">"histology"</span><span class="op">)</span>

<span class="fu"><a href="../reference/plotDeaVolcano.html">plotDeaVolcano</a></span><span class="op">(</span>
  object <span class="op">=</span> <span class="va">object_t269</span>,
  across <span class="op">=</span> <span class="st">"histology"</span>, 
  use_pseudolog <span class="op">=</span> <span class="cn">TRUE</span>
  <span class="op">)</span></code></pre></div>
<div class="figure">
<img src="spata-v2-spatial-segmentation_files/figure-html/unnamed-chunk-20-1.png" alt="Fig.14 Volcano plot displays DE-results based on histology." width="100%"><p class="caption">
Fig.14 Volcano plot displays DE-results based on histology.
</p>
</div>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotSurfaceComparison.html">plotSurfaceComparison</a></span><span class="op">(</span>
  object <span class="op">=</span> <span class="va">object_t269</span>,
  color_by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"B2M"</span>, <span class="st">"MBP"</span>, <span class="st">"SNAP25"</span><span class="op">)</span>, 
  nrow <span class="op">=</span> <span class="fl">1</span>
<span class="op">)</span></code></pre></div>
<div class="figure">
<img src="spata-v2-spatial-segmentation_files/figure-html/unnamed-chunk-21-1.png" alt="Fig.15 Example surface plots." width="100%"><p class="caption">
Fig.15 Example surface plots.
</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Jan Kueckelhaus, Dieter-Henrik Heiland.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
