<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Spatial Annotation Screening • SPATA2</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Spatial Annotation Screening">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SPATA2</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">3.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/installation.html">Installation</a>
</li>
<li>
  <a href="../articles/spata-data.html">SPATAData</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">Initiating a SPATA2 object</li>
    <li>
      <a href="../articles/initiation-and-preprocessing-customized.html">Initiation &amp; Preprocessing with individual data</a>
    </li>
    <li>
      <a href="../articles/initiation-and-preprocessing-merfish.html">Initiation &amp; Preprocessing with MERFISH data</a>
    </li>
    <li>
      <a href="../articles/initiation-and-preprocessing-visium.html">Initiation &amp; Preprocessing with Visium data</a>
    </li>
    <li>
      <a href="../articles/initiation-and-preprocessing-visium-hd.html">Initiation &amp; Preprocessing with VisiumHD data</a>
    </li>
    <li class="dropdown-header">Basic Manipulation of a SPATA2 object</li>
    <li>
      <a href="../articles/adding-data.html">Adding Data to SPATA2 objects</a>
    </li>
    <li>
      <a href="../articles/package-compatibility.html">Platform Compatibility</a>
    </li>
    <li>
      <a href="../articles/molecular-variables.html">Molecular Variables</a>
    </li>
    <li>
      <a href="../articles/image-handling.html">Image Handling</a>
    </li>
    <li>
      <a href="../articles/spatial-segmentation.html">Spatial Segmentation</a>
    </li>
    <li class="dropdown-header">Implemented Algorithms &amp; Concepts</li>
    <li>
      <a href="../articles/cnv.html">Copy Number Variations</a>
    </li>
    <li>
      <a href="../articles/clustering.html">Clustering</a>
    </li>
    <li>
      <a href="../articles/dimensional-reduction.html">Dimensional Reduction</a>
    </li>
    <li>
      <a href="../articles/dea.html">Differential Expression</a>
    </li>
    <li>
      <a href="../articles/gsea.html">Gene Set Enrichment</a>
    </li>
    <li>
      <a href="../articles/spatial-measures.html">Spatial Measures</a>
    </li>
    <li class="dropdown-header">Spatial Gradient Screening</li>
    <li>
      <a href="../articles/creating-spatial-annotations.html">Creating Spatial Annotations</a>
    </li>
    <li>
      <a href="../articles/using-spatial-annotations.html">Using Spatial Annotations</a>
    </li>
    <li>
      <a href="../articles/spatial-annotation-screening.html">Spatial Annotation Screening</a>
    </li>
    <li>
      <a href="../articles/creating-spatial-trajectories.html">Creating Spatial Trajectories</a>
    </li>
    <li>
      <a href="../articles/spatial-trajectory-screening.html">Spatial Trajectory Screening</a>
    </li>
    <li class="dropdown-header">Visualization</li>
    <li>
      <a href="../articles/plotting-surface.html">Surface Plotting</a>
    </li>
  </ul>
</li>
<li>
  <a href="../articles/newslog.html">Newslog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Spatial Annotation Screening</h1>
            
      

      <div class="hidden name"><code>spatial-annotation-screening.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="introduction">1. Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>This vignette exemplifies how to use Spatial Annotation Screening in
SPATA2. It builds on the spatial annotations created in the vignette on
<a href="creating-spatial-annotation.html">creating spatial
annotations</a>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://themilolab.com/" class="external-link">SPATA2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># load SPATA2 inbuilt data and process</span></span>
<span><span class="va">object_t313</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/loadExampleObject.html">loadExampleObject</a></span><span class="op">(</span><span class="st">"UKF313T"</span>, process <span class="op">=</span> <span class="cn">TRUE</span>, meta <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">necrotic_ids</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"necrotic_area"</span>, <span class="st">"necrotic_edge"</span>, <span class="st">"necrotic_edge2"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plotSpatialAnnotations.html">plotSpatialAnnotations</a></span><span class="op">(</span><span class="va">object_t313</span>, ids <span class="op">=</span> <span class="va">necrotic_ids</span>, nrow <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p><img src="spatial-annotation-screening_files/figure-html/unnamed-chunk-2-1.png" width="100%"></p>
<p>Spatial Annotation Screening (SAS) pursues the hypothesis that
specific genes - or other numeric features for that matter - display
non-random expression patterns in relation to spatial reference
features, such as spatial annotations. SAS utilizes these reference
features to incorporate the integration of potential biological forces
in the identification of spatially variable genes, such as the presence
of necrosis within a tumor. This allows for a supervised,
hypothesis-driven screening for spatial patterns, which, unlike
differential expression analysis (DEA), acknowledges the continuous
nature of gene expression and avoids the limitations of group-based
testing.</p>
</div>
<div class="section level2">
<h2 id="running-the-algorithm">2. Running the algorithm<a class="anchor" aria-label="anchor" href="#running-the-algorithm"></a>
</h2>
<p>The algorithm is wrapped up in the function
<code><a href="../reference/spatialAnnotationScreening.html">spatialAnnotationScreening()</a></code>. Note that parameters
<code>core</code> and <code>distance</code> are essential for the
outcome. Make sure to align either input to your questioning. See the
vignette on <a href="using-spatia-annotation.Rmd">using spatial
annotations</a> for more information.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># prefiltering genes for spatial variability with SPARKX is recommended</span></span>
<span><span class="va">object_t313</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runSPARKX.html">runSPARKX</a></span><span class="op">(</span><span class="va">object_t313</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sparkx_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runSPARKX.html">getSparkxGenes</a></span><span class="op">(</span><span class="va">object_t313</span>, threshold_pval <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sas_out</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="../reference/spatialAnnotationScreening.html">spatialAnnotationScreening</a></span><span class="op">(</span></span>
<span>    object <span class="op">=</span> <span class="va">object_t313</span>, </span>
<span>    ids <span class="op">=</span> <span class="va">necrotic_ids</span>, </span>
<span>    variables <span class="op">=</span> <span class="va">sparkx_genes</span>, </span>
<span>    core <span class="op">=</span> <span class="cn">FALSE</span>, <span class="co"># do not include the core of the annotations</span></span>
<span>    distance <span class="op">=</span> <span class="st">"dte"</span> <span class="co"># distance to edge </span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p>The output of <code><a href="../reference/spatialAnnotationScreening.html">spatialAnnotationScreening()</a></code> is an S4
object of class <code>SpatialAnnotationScreening</code> which contains
the set up as well as the results from the algorithm.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># an S4 object that contains the results </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">sas_out</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "SpatialAnnotationScreening"</span></span>
<span><span class="co">## attr(,"package")</span></span>
<span><span class="co">## [1] "SPATA2"</span></span></code></pre>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/methods/slot.html" class="external-link">slotNames</a></span><span class="op">(</span><span class="va">sas_out</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "annotations" "coordinates" "models"      "qc"          "results"    </span></span>
<span><span class="co">## [6] "sample"      "set_up"</span></span></code></pre>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># prepare plotting </span></span>
<span><span class="va">screening_dir_layer</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ggpLayerScreeningDirectionSAS.html">ggpLayerScreeningDirectionSAS</a></span><span class="op">(</span><span class="va">object_t313</span>, ids <span class="op">=</span> <span class="va">necrotic_ids</span>, line_size <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span>
<span><span class="va">sas_areas</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/color_vector.html">color_vector</a></span><span class="op">(</span>clrp <span class="op">=</span> <span class="st">"npg"</span>, names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"core"</span>, <span class="st">"environment"</span>, <span class="st">"periphery"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># left plot</span></span>
<span><span class="co"># use information within the SAS output to recreate the set up </span></span>
<span><span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span><span class="va">sas_out</span>, color_by <span class="op">=</span> <span class="st">"rel_loc"</span>, clrp_adjust <span class="op">=</span> <span class="va">sas_areas</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="va">screening_dir_layer</span></span>
<span></span>
<span><span class="co"># right plot</span></span>
<span><span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span><span class="va">sas_out</span>, color_by <span class="op">=</span> <span class="st">"dist"</span>, fill <span class="op">=</span> <span class="fu"><a href="https://scales.r-lib.org/reference/alpha.html" class="external-link">alpha</a></span><span class="op">(</span><span class="st">"lightgrey"</span>, <span class="fl">0.25</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>color <span class="op">=</span> <span class="st">"Dist [mm]"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="va">screening_dir_layer</span></span></code></pre></div>
<p><img src="spatial-annotation-screening_files/figure-html/unnamed-chunk-4-1.png" width="50%"><img src="spatial-annotation-screening_files/figure-html/unnamed-chunk-4-2.png" width="50%"></p>
</div>
<div class="section level2">
<h2 id="extracting-results">3. Extracting results<a class="anchor" aria-label="anchor" href="#extracting-results"></a>
</h2>
<p>Results are stored in slot <span class="citation">@results</span>.</p>
<p>Slot <span class="citation">@results$significance</span> contains a
data.frame with one row for each screened variable which provides
information regarding the degree of randomness the inferred pattern
contains as quantified by the total variation (<em>tot_var</em>). The
p-value gives the probability to obtain such a total variation under
complete randomness and indicates the degree of significance. Column
<em>fdr</em> contains the adjusted p-value according to the False
Discovery Rate.</p>
<p>Slot <span class="citation">@results$model_fits</span> contains the
model fitting results. It is a data.frame where each row corresponds to
a variable ~ model pair. The columns <em>mae</em> (mean absolute error)
and <em>rmse</em> (root mean squared error) indicate the quality of the
fit. The lower the value the better.</p>
<p>You can easily subset the results with <code>dplyr</code> verbs.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># e.g. filter output for gradients with a p-value less than 0.05</span></span>
<span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">sas_out</span><span class="op">@</span><span class="va">results</span><span class="op">$</span><span class="va">significance</span>, <span class="va">p_value</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span>  </span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 560 × 6</span></span></span>
<span><span class="co">##    variables  rel_var tot_var p_value norm_var    fdr</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> A2M          0.112    2.34  0.009<span style="text-decoration: underline;">1</span>   0.117  0.055<span style="text-decoration: underline;">9</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> AC245595.1   1        2.33  0.008<span style="text-decoration: underline;">3</span>   0.117  0.051<span style="text-decoration: underline;">3</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> ACTA2        0.516    2.13  0.002<span style="text-decoration: underline;">6</span>   0.107  0.023<span style="text-decoration: underline;">5</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> ACTB         0.886    1.69  0        0.084<span style="text-decoration: underline;">6</span> 0     </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> ACTG1       -<span style="color: #BB0000;">0.569</span>    1.82  0        0.091<span style="text-decoration: underline;">1</span> 0     </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> ADAM12       0.361    2.50  0.018<span style="text-decoration: underline;">5</span>   0.125  0.085<span style="text-decoration: underline;">2</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> ADD1        -<span style="color: #BB0000;">0.426</span>    2.32  0.007<span style="text-decoration: underline;">4</span>   0.116  0.047<span style="text-decoration: underline;">4</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> ADM         -<span style="color: #BB0000;">0.959</span>    1.80  0        0.089<span style="text-decoration: underline;">8</span> 0     </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> ADRM1        0.915    2.69  0.036<span style="text-decoration: underline;">1</span>   0.135  0.125 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> AKR1B1       1        2.26  0.005<span style="text-decoration: underline;">9</span>   0.113  0.041<span style="text-decoration: underline;">8</span></span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 550 more rows</span></span></span></code></pre>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># e.g. filter model fits for best fit by variable</span></span>
<span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">sas_out</span><span class="op">@</span><span class="va">results</span><span class="op">$</span><span class="va">model_fits</span>, <span class="va">variables</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html" class="external-link">slice_min</a></span><span class="op">(</span><span class="va">rmse</span>, n <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 265 × 4</span></span></span>
<span><span class="co">## <span style="color: #949494;"># Groups:   variables [265]</span></span></span>
<span><span class="co">##    variables models                mae   rmse</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>               <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> ACTA2     ascending_linear   0.236  0.266 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> ACTB      ascending_linear   0.230  0.263 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> ACTG1     peak_gradual       0.301  0.388 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> ADD1      peak_gradual       0.290  0.353 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> ADM       descending_gradual 0.072<span style="text-decoration: underline;">5</span> 0.097<span style="text-decoration: underline;">0</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> AKR1B1    ascending_linear   0.153  0.206 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> ANAPC11   ascending_linear   0.097<span style="text-decoration: underline;">6</span> 0.123 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> ANGPTL4   descending_linear  0.130  0.162 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> APOC1     ascending_linear   0.102  0.131 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> APOE      ascending_gradual  0.114  0.141 </span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 255 more rows</span></span></span></code></pre>
<p>Alternatively, you can use <code><a href="../reference/getSgsResultsDf.html">getSgsResultsDf()</a></code> and
<code><a href="../reference/getSgsResultsVec.html">getSgsResultsVec()</a></code> which are convenient wrappers around
conditions with which to subset the results.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># the default: </span></span>
<span><span class="fu"><a href="../reference/getSgsResultsDf.html">getSgsResultsDf</a></span><span class="op">(</span><span class="va">sas_out</span>, pval <span class="op">=</span> <span class="st">"fdr"</span>, threshold_pval <span class="op">=</span> <span class="fl">0.05</span>, eval <span class="op">=</span> <span class="st">"mae"</span>, threshold_eval <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 207 × 9</span></span></span>
<span><span class="co">## <span style="color: #949494;"># Groups:   models [6]</span></span></span>
<span><span class="co">##    variables models          mae   rmse rel_var tot_var p_value norm_var     fdr</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> CD74      ascending_l… 0.055<span style="text-decoration: underline;">0</span> 0.067<span style="text-decoration: underline;">1</span>   0.856    1.31  0        0.065<span style="text-decoration: underline;">5</span> 0      </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> C1QB      ascending_l… 0.057<span style="text-decoration: underline;">0</span> 0.075<span style="text-decoration: underline;">8</span>   0.976    1.49  0        0.074<span style="text-decoration: underline;">3</span> 0      </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> P4HA2     descending_… 0.061<span style="text-decoration: underline;">1</span> 0.078<span style="text-decoration: underline;">2</span>  -<span style="color: #BB0000;">0.822</span>    1.57  0        0.078<span style="text-decoration: underline;">3</span> 0      </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> PSME2     ascending_l… 0.062<span style="text-decoration: underline;">7</span> 0.082<span style="text-decoration: underline;">6</span>   0.783    2.22  0.004<span style="text-decoration: underline;">1</span>   0.111  0.031<span style="text-decoration: underline;">3</span> </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> FN1       descending_… 0.065<span style="text-decoration: underline;">3</span> 0.082<span style="text-decoration: underline;">2</span>  -<span style="color: #BB0000;">0.910</span>    1.77  0        0.088<span style="text-decoration: underline;">4</span> 0      </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> NDUFA13   ascending_l… 0.069<span style="text-decoration: underline;">1</span> 0.088<span style="text-decoration: underline;">3</span>   1        1.77  0        0.088<span style="text-decoration: underline;">3</span> 0      </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> THBS2     descending_… 0.070<span style="text-decoration: underline;">2</span> 0.091<span style="text-decoration: underline;">9</span>  -<span style="color: #BB0000;">0.780</span>    1.53  0        0.076<span style="text-decoration: underline;">3</span> 0      </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> MFAP2     ascending_l… 0.071<span style="text-decoration: underline;">3</span> 0.085<span style="text-decoration: underline;">3</span>   0.994    1.30  0        0.064<span style="text-decoration: underline;">8</span> 0      </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> ADM       descending_… 0.072<span style="text-decoration: underline;">5</span> 0.097<span style="text-decoration: underline;">0</span>  -<span style="color: #BB0000;">0.959</span>    1.80  0        0.089<span style="text-decoration: underline;">8</span> 0      </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> IFITM3    ascending_l… 0.072<span style="text-decoration: underline;">7</span> 0.105    0.720    1.92  0.000<span style="text-decoration: underline;">4</span>   0.096<span style="text-decoration: underline;">2</span> 0.005<span style="text-decoration: underline;">55</span></span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 197 more rows</span></span></span></code></pre>
<p>You can filter the results for significant variables below a certain
threshold for specific models to obtain genes of potential interest. For
instance, to obtain genes that feature a non random, descending gradient
with increasing distance to necrosis - associated with necrosis - use
the following:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># (pval = "fdr", threshold_pval = 0.05)</span></span>
<span><span class="va">desc_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getSgsResultsVec.html">getSgsResultsVec</a></span><span class="op">(</span><span class="va">sas_out</span>, model_subset <span class="op">=</span> <span class="st">"descending"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">desc_genes</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 63</span></span></code></pre>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">desc_genes</span>, <span class="fl">9</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "P4HA2"  "FN1"    "THBS2"  "ADM"    "BRI3"   "EEF1A1" "CA12"   "ENO2"  </span></span>
<span><span class="co">## [9] "SLC2A1"</span></span></code></pre>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">desc_genes_sub</span> <span class="op">&lt;-</span> <span class="va">desc_genes</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">9</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># left plot</span></span>
<span><span class="fu"><a href="../reference/plotSurfaceComparison.html">plotSurfaceComparison</a></span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">object_t313</span>, </span>
<span>  color_by <span class="op">=</span> <span class="va">desc_genes_sub</span>,</span>
<span>  normalize <span class="op">=</span> <span class="cn">T</span>, </span>
<span>  outline <span class="op">=</span> <span class="cn">T</span>, </span>
<span>  pt_clrsp <span class="op">=</span> <span class="st">"BuPu"</span></span>
<span>  <span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="../reference/ggpLayerSpatAnnOutline.html">ggpLayerSpatAnnOutline</a></span><span class="op">(</span><span class="va">object_t313</span>, ids <span class="op">=</span> <span class="va">necrotic_ids</span>, fill <span class="op">=</span> <span class="st">"grey"</span>, incl_edge <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># right plot</span></span>
<span><span class="fu"><a href="../reference/plotSasLineplot.html">plotSasLineplot</a></span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">object_t313</span>, </span>
<span>  variables <span class="op">=</span> <span class="va">desc_genes_sub</span>,</span>
<span>  ids <span class="op">=</span> <span class="va">necrotic_ids</span>, </span>
<span>  line_color <span class="op">=</span> <span class="st">"blue"</span></span>
<span>  <span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Distance to Necrosis (mm)"</span><span class="op">)</span></span></code></pre></div>
<p><img src="spatial-annotation-screening_files/figure-html/unnamed-chunk-7-1.png" width="50%"><img src="spatial-annotation-screening_files/figure-html/unnamed-chunk-7-2.png" width="50%">
Alternatively, you can extract those with an opposite gradient -
ascending ones - which correspond to genes rather <em>repelled by</em>
necrosis.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># (pval = "fdr", threshold_pval = 0.05)</span></span>
<span><span class="va">asc_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getSgsResultsVec.html">getSgsResultsVec</a></span><span class="op">(</span><span class="va">sas_out</span>, model_subset <span class="op">=</span> <span class="st">"ascending"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">asc_genes</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 176</span></span></code></pre>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">asc_genes</span>, <span class="fl">9</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "CD74"    "C1QB"    "PSME2"   "NDUFA13" "MFAP2"   "IFITM3"  "S100A11"</span></span>
<span><span class="co">## [8] "COL5A3"  "TDO2"</span></span></code></pre>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">asc_genes_sub</span> <span class="op">&lt;-</span> <span class="va">asc_genes</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">9</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># show genes</span></span>
<span><span class="va">asc_genes_sub</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "CD74"    "C1QB"    "PSME2"   "NDUFA13" "MFAP2"   "IFITM3"  "S100A11"</span></span>
<span><span class="co">## [8] "COL5A3"  "TDO2"</span></span></code></pre>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># left plot</span></span>
<span><span class="fu"><a href="../reference/plotSurfaceComparison.html">plotSurfaceComparison</a></span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">object_t313</span>,</span>
<span>  color_by <span class="op">=</span> <span class="va">asc_genes_sub</span>, </span>
<span>  normalize <span class="op">=</span> <span class="cn">T</span>, </span>
<span>  outline <span class="op">=</span> <span class="cn">T</span>, </span>
<span>  pt_clrsp <span class="op">=</span> <span class="st">"Reds 3"</span></span>
<span>  <span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="../reference/ggpLayerSpatAnnOutline.html">ggpLayerSpatAnnOutline</a></span><span class="op">(</span><span class="va">object_t313</span>, ids <span class="op">=</span> <span class="va">necrotic_ids</span>, fill <span class="op">=</span> <span class="st">"grey"</span>, incl_edge <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># right plot</span></span>
<span><span class="fu"><a href="../reference/plotSasLineplot.html">plotSasLineplot</a></span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">object_t313</span>, </span>
<span>  variables <span class="op">=</span> <span class="va">asc_genes_sub</span>, </span>
<span>  ids <span class="op">=</span> <span class="va">necrotic_ids</span>, </span>
<span>  line_color <span class="op">=</span> <span class="st">"red"</span></span>
<span>  <span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Distance to Necrosis (mm)"</span><span class="op">)</span></span></code></pre></div>
<p><img src="spatial-annotation-screening_files/figure-html/unnamed-chunk-8-1.png" width="50%"><img src="spatial-annotation-screening_files/figure-html/unnamed-chunk-8-2.png" width="50%"></p>
</div>
<div class="section level2">
<h2 id="pitfalls">4. Pitfalls<a class="anchor" aria-label="anchor" href="#pitfalls"></a>
</h2>
<p>There are some pitfalls regarding spatial annotation screening when
it comes to setting up the screening parameters.</p>
<div class="section level3">
<h3 id="incomplete-annotation">4.1 Incomplete annotation<a class="anchor" aria-label="anchor" href="#incomplete-annotation"></a>
</h3>
<p>Comprehensive annotation is important in order not to include areas
in the screening that are actually of the same nature than the
annotation (based on which the screening is conducted) itself. This
would distort the pattern of the inferred gradient. For instance,
compare the inferred gradient when disregarding the annotated necrotic
area at the right bottom - labeled necrotic_edge. The
<code>distance</code> parameter defines the maximum distance up to which
observations are included in the screening. It defaults to the edge of
the tissue (the maximal distance of all observations that were assigned
to the tissue section it is located on).</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">all_necr_ids</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"necrotic_area"</span>, <span class="st">"necrotic_edge"</span>, <span class="st">"necrotic_edge2"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># plot left, red outlines highlight the omitted annotations</span></span>
<span><span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span><span class="va">object_t313</span>, color_by <span class="op">=</span> <span class="st">"VEGFA"</span>, display_image <span class="op">=</span> <span class="cn">F</span>, pt_clrsp <span class="op">=</span> <span class="st">"PuBu"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="../reference/ggpLayerSpatAnnOutline.html">ggpLayerSpatAnnOutline</a></span><span class="op">(</span><span class="va">object_t313</span>, ids <span class="op">=</span> <span class="st">"necrotic_area"</span>, fill <span class="op">=</span> <span class="fu"><a href="https://scales.r-lib.org/reference/alpha.html" class="external-link">alpha</a></span><span class="op">(</span><span class="st">"lightgrey"</span>, <span class="fl">0.25</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="co"># decrease resolution to 250um for visualization purpose </span></span>
<span>  <span class="fu"><a href="../reference/ggpLayerExprEstimatesSAS.html">ggpLayerExprEstimatesSAS</a></span><span class="op">(</span><span class="va">object_t313</span>, ids  <span class="op">=</span> <span class="st">"necrotic_area"</span>, resolution <span class="op">=</span> <span class="st">"250um"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="../reference/ggpLayerSpatAnnOutline.html">ggpLayerSpatAnnOutline</a></span><span class="op">(</span><span class="va">object_t313</span>, ids <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"necrotic_edge"</span>, <span class="st">"necrotic_edge2"</span><span class="op">)</span>, line_size <span class="op">=</span> <span class="fl">1.5</span>, line_color <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># plot right, distance = "dte": default</span></span>
<span><span class="fu"><a href="../reference/getCoordsDfSA.html">getCoordsDfSA</a></span><span class="op">(</span><span class="va">object_t313</span>, ids <span class="op">=</span> <span class="st">"necrotic_area"</span>, distance <span class="op">=</span> <span class="st">"dte"</span><span class="op">)</span>  <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">.</span>, color_by <span class="op">=</span> <span class="st">"dist"</span>, pt_clrsp <span class="op">=</span> <span class="st">"inferno"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="../reference/ggpLayerSpatAnnOutline.html">ggpLayerSpatAnnOutline</a></span><span class="op">(</span><span class="va">object_t313</span>, ids <span class="op">=</span> <span class="st">"necrotic_area"</span>, fill <span class="op">=</span> <span class="fu"><a href="https://scales.r-lib.org/reference/alpha.html" class="external-link">alpha</a></span><span class="op">(</span><span class="st">"lightgrey"</span>, <span class="fl">0.25</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="spatial-annotation-screening_files/figure-html/unnamed-chunk-9-1.png" width="50%"><img src="spatial-annotation-screening_files/figure-html/unnamed-chunk-9-2.png" width="50%"></p>
<p>Suddenly, the gradient of VEGFA does not look as interesting as with
all necrotic areas included. Note the peak at 2.7mm where the omitted
necrotic edge is located. Therefore, make sure to include all areas in
the screening set up that are big enough to have an impact on the
screening results.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># plot left, only necrotic area, distance = "dte" screens the whole sample and includes other necrotic areas</span></span>
<span><span class="fu"><a href="../reference/plotSasLineplot.html">plotSasLineplot</a></span><span class="op">(</span><span class="va">object_t313</span>, variables <span class="op">=</span> <span class="st">"VEGFA"</span>, ids <span class="op">=</span> <span class="st">"necrotic_area"</span>, line_color <span class="op">=</span> <span class="st">"blue"</span><span class="op">)</span> </span>
<span></span>
<span><span class="co"># plot right, comprehensive annotation as used in the example run above</span></span>
<span><span class="fu"><a href="../reference/plotSasLineplot.html">plotSasLineplot</a></span><span class="op">(</span><span class="va">object_t313</span>, variables <span class="op">=</span> <span class="st">"VEGFA"</span>, ids <span class="op">=</span> <span class="va">all_necr_ids</span>, line_color <span class="op">=</span> <span class="st">"blue"</span><span class="op">)</span></span></code></pre></div>
<p><img src="spatial-annotation-screening_files/figure-html/unnamed-chunk-10-1.png" width="50%"><img src="spatial-annotation-screening_files/figure-html/unnamed-chunk-10-2.png" width="50%"></p>
</div>
<div class="section level3">
<h3 id="the-tissue-edge">4.2 The tissue edge<a class="anchor" aria-label="anchor" href="#the-tissue-edge"></a>
</h3>
<p>Make sure that appropriate results of
<code><a href="../reference/identifyTissueOutline.html">identifyTissueOutline()</a></code> exist in the <code>SPATA2</code>
object. Consider the following sample of two mouse brain tissue sections
with two stab wound injuries located in the cortex. While the clustering
indicates a significant impact of these stab wounds on their local
surrounding, this impact can logically only affect the tissue section
they are located on. Inferring gene expression gradients up to a
distance that transgresses the gab between both tissue sections can
happen if the <code>distance</code> parameter is set carelessely and/or
the tissue outline has not been identified properly!</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># load injured mouse brain data </span></span>
<span><span class="va">object_mouse</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/loadExampleObject.html">loadExampleObject</a></span><span class="op">(</span><span class="st">"LMU_MCI"</span>, process <span class="op">=</span> <span class="cn">TRUE</span>, meta <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># "carelessly" set distance to "3mm"</span></span>
<span><span class="va">dst</span> <span class="op">&lt;-</span> <span class="st">"3mm"</span></span>
<span></span>
<span><span class="co"># with tissue edge in mind (incl_edge = TRUE), only for inj1</span></span>
<span><span class="va">expr_est_with_edge_in_mind</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="../reference/ggpLayerExprEstimatesSAS.html">ggpLayerExprEstimatesSAS</a></span><span class="op">(</span><span class="va">object_mouse</span>, ids <span class="op">=</span> <span class="st">"inj1"</span>, incl_edge <span class="op">=</span> <span class="cn">T</span>, distance <span class="op">=</span> <span class="va">dst</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># without tissue edge in mind (incl_edge = FALSE), only for inj1</span></span>
<span><span class="va">expr_est_without_edge_in_mind</span><span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="../reference/ggpLayerExprEstimatesSAS.html">ggpLayerExprEstimatesSAS</a></span><span class="op">(</span><span class="va">object_mouse</span>, ids <span class="op">=</span> <span class="st">"inj1"</span>, incl_edge <span class="op">=</span> <span class="cn">F</span>, distance <span class="op">=</span> <span class="va">dst</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># left plot</span></span>
<span><span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span><span class="va">object_mouse</span>, color_by <span class="op">=</span> <span class="st">"clusters"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="../reference/ggpLayerSpatAnnOutline.html">ggpLayerSpatAnnOutline</a></span><span class="op">(</span><span class="va">object_mouse</span>, ids <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"inj1"</span>, <span class="st">"inj2"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="va">expr_est_with_edge_in_mind</span></span>
<span>  </span>
<span><span class="co"># right plot</span></span>
<span><span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span><span class="va">object_mouse</span>, color_by <span class="op">=</span> <span class="st">"clusters"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="../reference/ggpLayerSpatAnnOutline.html">ggpLayerSpatAnnOutline</a></span><span class="op">(</span><span class="va">object_mouse</span>, ids <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"inj1"</span>, <span class="st">"inj2"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="va">expr_est_without_edge_in_mind</span></span></code></pre></div>
<p><img src="spatial-annotation-screening_files/figure-html/unnamed-chunk-11-1.png" width="50%"><img src="spatial-annotation-screening_files/figure-html/unnamed-chunk-11-2.png" width="50%"></p>
<p>It goes without saying that the right plot displayes a screening set
up that does not make sense. Gene expression would be estimated in
regions that feature no tissue at all or in regions of a completely
distinct tissue section with no biological connection to the spatial
annotation of interest. Spatial annotation screening always makes sure
that the screening does not transgresses the border of the tissue (in
that sense, <code>incl_edge</code> is always TRUE). But to do that
appropriately it requires proper results in the variable
<em>tissue_section</em> as computed by
<code><a href="../reference/identifyTissueOutline.html">identifyTissueOutline()</a></code>!</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># left plot</span></span>
<span><span class="co"># color by the meta feature that just contains the sample name</span></span>
<span><span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span><span class="va">object_mouse</span>, color_by <span class="op">=</span> <span class="st">"sample"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># mess up the tissue_section variable manually </span></span>
<span><span class="va">object_mouse</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/useVarForTissueOutline.html">useVarForTissueOutline</a></span><span class="op">(</span><span class="va">object_mouse</span>, var_name <span class="op">=</span> <span class="st">"sample"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># right plot, distance = "3mm" transgresses the edge of the upper section</span></span>
<span><span class="co"># because the two sections are not identified as separate sections</span></span>
<span><span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span><span class="va">object_mouse</span>, color_by <span class="op">=</span> <span class="st">"tissue_section"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="../reference/ggpLayerExprEstimatesSAS.html">ggpLayerExprEstimatesSAS</a></span><span class="op">(</span><span class="va">object_mouse</span>, ids <span class="op">=</span> <span class="st">"inj1"</span>, incl_edge <span class="op">=</span> <span class="cn">T</span>, distance <span class="op">=</span> <span class="va">dst</span><span class="op">)</span></span></code></pre></div>
<p><img src="spatial-annotation-screening_files/figure-html/unnamed-chunk-12-1.png" width="50%"><img src="spatial-annotation-screening_files/figure-html/unnamed-chunk-12-2.png" width="50%"></p>
<p>The right plot above shows that even if <code>incl_edge</code> is
TRUE, the screening set up is nonsense because the information provided
regarding the tissue outline and thus the tissue edge are messed up.
Make sure that SPATA2 localizes the spatial annotations on the correct
tissue sections and - in general - that tissue sections are correctly
identified.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># overwrite tissue_section again</span></span>
<span><span class="va">object_mouse</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/identifyTissueOutline.html">identifyTissueOutline</a></span><span class="op">(</span><span class="va">object_mouse</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 14:03:46 Identifying tissue outline with `method = obs`.</span></span></code></pre>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># check on which tissue sections the annotations are located on</span></span>
<span><span class="fu"><a href="../reference/whichTissueSection.html">whichTissueSection</a></span><span class="op">(</span><span class="va">object_mouse</span>, id <span class="op">=</span> <span class="st">"inj1"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "tissue_section_2"</span></span></code></pre>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/whichTissueSection.html">whichTissueSection</a></span><span class="op">(</span><span class="va">object_mouse</span>, id <span class="op">=</span> <span class="st">"inj2"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "tissue_section_1"</span></span></code></pre>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># left plot</span></span>
<span><span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span><span class="va">object_mouse</span>, color_by <span class="op">=</span> <span class="st">"tissue_section"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="../reference/ggpLayerTissueOutline.html">ggpLayerTissueOutline</a></span><span class="op">(</span><span class="va">object_mouse</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># set the distance way too high</span></span>
<span><span class="co"># -&gt; no problem, since it is cut of at the tissue edge (which is now properly defined)</span></span>
<span><span class="va">dst_high</span> <span class="op">&lt;-</span> <span class="st">"10mm"</span></span>
<span></span>
<span><span class="va">ids</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"inj1"</span>, <span class="st">"inj2"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># right plot</span></span>
<span><span class="co"># plot both annotations on the surface</span></span>
<span><span class="co"># a warning informs you that the distance was adjusted based on the tissue edge</span></span>
<span><span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span><span class="va">object_mouse</span>, color_by <span class="op">=</span> <span class="st">"tissue_section"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="../reference/ggpLayerTissueOutline.html">ggpLayerTissueOutline</a></span><span class="op">(</span><span class="va">object_mouse</span>, line_size <span class="op">=</span> <span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="../reference/ggpLayerSpatAnnOutline.html">ggpLayerSpatAnnOutline</a></span><span class="op">(</span><span class="va">object_mouse</span>, ids <span class="op">=</span> <span class="va">ids</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="../reference/ggpLayerExprEstimatesSAS.html">ggpLayerExprEstimatesSAS</a></span><span class="op">(</span><span class="va">object_mouse</span>, ids <span class="op">=</span> <span class="va">ids</span>, incl_edge <span class="op">=</span> <span class="cn">T</span>, distance <span class="op">=</span> <span class="va">dst_high</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="../reference/ggpLayerSpatAnnPointer.html">ggpLayerSpatAnnPointer</a></span><span class="op">(</span></span>
<span>    object <span class="op">=</span> <span class="va">object_mouse</span>, </span>
<span>    ids <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"inj1"</span>, <span class="st">"inj2"</span><span class="op">)</span>,</span>
<span>    ptr_lengths <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"0.75mm"</span>, <span class="st">"1.25mm"</span><span class="op">)</span>,</span>
<span>    ptr_angles <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">70</span>, <span class="fl">270</span><span class="op">)</span>,</span>
<span>    ptr_arrow <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/grid/arrow.html" class="external-link">arrow</a></span><span class="op">(</span>length <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fl">0.1</span>, <span class="st">"inches"</span><span class="op">)</span><span class="op">)</span>, </span>
<span>    text_dist <span class="op">=</span> <span class="fl">25</span>, </span>
<span>    text_size <span class="op">=</span> <span class="fl">6</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning in get_coords_df_sa(object = object, id = id, distance = dist,</span></span>
<span><span class="co">## resolution = resolution, : Parameter `distance` equals ~758.74px and exceeds</span></span>
<span><span class="co">## the distance from spatial annotation 'inj1' to the edge of tissue section</span></span>
<span><span class="co">## 'tissue_section_2' where it is located on: 298.74px. The parameter was adjusted</span></span>
<span><span class="co">## accordingly.</span></span></code></pre>
<pre><code><span><span class="co">## Warning in get_coords_df_sa(object = object, id = id, distance = dist,</span></span>
<span><span class="co">## resolution = resolution, : Parameter `distance` equals ~758.74px and exceeds</span></span>
<span><span class="co">## the distance from spatial annotation 'inj2' to the edge of tissue section</span></span>
<span><span class="co">## 'tissue_section_1' where it is located on: 268.89px. The parameter was adjusted</span></span>
<span><span class="co">## accordingly.</span></span></code></pre>
<p><img src="spatial-annotation-screening_files/figure-html/unnamed-chunk-13-1.png" width="50%"><img src="spatial-annotation-screening_files/figure-html/unnamed-chunk-13-2.png" width="50%"></p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Jan Kueckelhaus, Dieter-Henrik Heiland.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

      </footer>
</div>






  </body>
</html>
