<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Spatial Trajectory Screening • SPATA2</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Spatial Trajectory Screening">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SPATA2</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">3.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/installation.html">Installation</a>
</li>
<li>
  <a href="../articles/spata-data.html">SPATAData</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">Initiating a SPATA2 object</li>
    <li>
      <a href="../articles/initiation-and-preprocessing-customized.html">Initiation &amp; Preprocessing with individual data</a>
    </li>
    <li>
      <a href="../articles/initiation-and-preprocessing-merfish.html">Initiation &amp; Preprocessing with MERFISH data</a>
    </li>
    <li>
      <a href="../articles/initiation-and-preprocessing-visium.html">Initiation &amp; Preprocessing with Visium data</a>
    </li>
    <li>
      <a href="../articles/initiation-and-preprocessing-visium-hd.html">Initiation &amp; Preprocessing with VisiumHD data</a>
    </li>
    <li class="dropdown-header">Basic Manipulation of a SPATA2 object</li>
    <li>
      <a href="../articles/adding-data.html">Adding Data to SPATA2 objects</a>
    </li>
    <li>
      <a href="../articles/package-compatibility.html">Platform Compatibility</a>
    </li>
    <li>
      <a href="../articles/molecular-variables.html">Molecular Variables</a>
    </li>
    <li>
      <a href="../articles/image-handling.html">Image Handling</a>
    </li>
    <li>
      <a href="../articles/spatial-segmentation.html">Spatial Segmentation</a>
    </li>
    <li class="dropdown-header">Implemented Algorithms &amp; Concepts</li>
    <li>
      <a href="../articles/cnv.html">Copy Number Variations</a>
    </li>
    <li>
      <a href="../articles/clustering.html">Clustering</a>
    </li>
    <li>
      <a href="../articles/dimensional-reduction.html">Dimensional Reduction</a>
    </li>
    <li>
      <a href="../articles/dea.html">Differential Expression</a>
    </li>
    <li>
      <a href="../articles/gsea.html">Gene Set Enrichment</a>
    </li>
    <li>
      <a href="../articles/spatial-measures.html">Spatial Measures</a>
    </li>
    <li class="dropdown-header">Spatial Gradient Screening</li>
    <li>
      <a href="../articles/creating-spatial-annotations.html">Creating Spatial Annotations</a>
    </li>
    <li>
      <a href="../articles/using-spatial-annotations.html">Using Spatial Annotations</a>
    </li>
    <li>
      <a href="../articles/spatial-annotation-screening.html">Spatial Annotation Screening</a>
    </li>
    <li>
      <a href="../articles/creating-spatial-trajectories.html">Creating Spatial Trajectories</a>
    </li>
    <li>
      <a href="../articles/spatial-trajectory-screening.html">Spatial Trajectory Screening</a>
    </li>
    <li class="dropdown-header">Visualization</li>
    <li>
      <a href="../articles/plotting-surface.html">Surface Plotting</a>
    </li>
  </ul>
</li>
<li>
  <a href="../articles/newslog.html">Newslog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Spatial Trajectory Screening</h1>
            
      

      <div class="hidden name"><code>spatial-trajectory-screening.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="introduction">1. Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Spatial Trajectory Screening (STS) pursues the hypothesis that
specific genes - or other numeric features for that matter - display
non-random expression patterns in relation to spatial reference
features, such as spatial trajectories. STS utilizes these reference
features to incorporate the integration of potential biological forces
in the identification of spatially variable genes, such as the direction
of tumorous infiltration. This allows for a supervised,
hypothesis-driven screening for spatial patterns, which, unlike
differential expression analysis (DEA), acknowledges the continuous
nature of gene expression and avoids the limitations of group-based
testing. The algorithm is wrapped up in the function
<code><a href="../reference/spatialTrajectoryScreening.html">spatialTrajectoryScreening()</a></code>. See the tutorial we are using
the same example sample that we used in the tutorial on <a href="creating-spatial-trajectories.html">creating spatial
trajectories</a> - the glioblastoma sample UKF269T.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># load required packages</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://themilolab.com/" class="external-link">SPATA2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># load SPATA2 inbuilt data</span></span>
<span><span class="va">object_t269</span> <span class="op">&lt;-</span> <span class="va">object_t269</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/loadExampleObject.html">loadExampleObject</a></span><span class="op">(</span><span class="st">"UKF269T"</span>, process <span class="op">=</span> <span class="cn">TRUE</span>, meta <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># show image </span></span>
<span><span class="fu"><a href="../reference/plotImage.html">plotImage</a></span><span class="op">(</span><span class="va">object_t269</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># created with code </span></span>
<span><span class="fu"><a href="../reference/plotSpatialTrajectories.html">plotSpatialTrajectories</a></span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">object_t269</span>, </span>
<span>  ids <span class="op">=</span> <span class="st">"horizontal_mid"</span>, </span>
<span>  color_by <span class="op">=</span> <span class="st">"histology"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="spatial-trajectory-screening_files/figure-html/unnamed-chunk-2-1.png" width="50%"><img src="spatial-trajectory-screening_files/figure-html/unnamed-chunk-2-2.png" width="50%"></p>
</div>
<div class="section level2">
<h2 id="run-the-algorithm">2. Run the algorithm<a class="anchor" aria-label="anchor" href="#run-the-algorithm"></a>
</h2>
<p>The function to use is called
<code><a href="../reference/spatialTrajectoryScreening.html">spatialTrajectoryScreening()</a></code>. The parameter
<code>variables</code> takes the numeric variables that are supposed to
be included in the screening process. Since all sorts of numeric
variables can be included in the screening, the argument for the input
is simply called <code>variables</code>. Here, we are using the genes
that were already identified as spatially variable by
<code>SPARKX</code>. The goal is to further analyze which of the genes
are expressed in a non-random and biologically meaningful way along the
trajectory.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># this is a wrapper around SPARK::sparkx()</span></span>
<span><span class="va">object_t269</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runSPARKX.html">runSPARKX</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">object_t269</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## ## ===== SPARK-X INPUT INFORMATION ==== </span></span>
<span><span class="co">## ## number of total samples: 3213 </span></span>
<span><span class="co">## ## number of total genes: 15000 </span></span>
<span><span class="co">## ## Running with single core, may take some time </span></span>
<span><span class="co">## ## Testing With Projection Kernel</span></span>
<span><span class="co">## ## Testing With Gaussian Kernel 1</span></span>
<span><span class="co">## ## Testing With Gaussian Kernel 2</span></span>
<span><span class="co">## ## Testing With Gaussian Kernel 3</span></span>
<span><span class="co">## ## Testing With Gaussian Kernel 4</span></span>
<span><span class="co">## ## Testing With Gaussian Kernel 5</span></span>
<span><span class="co">## ## Testing With Cosine Kernel 1</span></span>
<span><span class="co">## ## Testing With Cosine Kernel 2</span></span>
<span><span class="co">## ## Testing With Cosine Kernel 3</span></span>
<span><span class="co">## ## Testing With Cosine Kernel 4</span></span>
<span><span class="co">## ## Testing With Cosine Kernel 5</span></span></code></pre>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># keep genes with a sparkx pvalue of 0.01 or lower</span></span>
<span><span class="va">spark_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runSPARKX.html">getSparkxGeneDf</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">object_t269</span>, threshold_pval <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># show results</span></span>
<span><span class="va">spark_df</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 8,734 × 3</span></span></span>
<span><span class="co">##    genes    combinedPval adjustedPval</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> RPL22               0            0</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> ID3                 0            0</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> MARCKSL1            0            0</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> PHC2                0            0</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> RPS8                0            0</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> WLS                 0            0</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> GNG5                0            0</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> RPL5                0            0</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> CNN3                0            0</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> RHOC                0            0</span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 8,724 more rows</span></span></span></code></pre>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># name of the trajectory</span></span>
<span><span class="va">id</span> <span class="op">&lt;-</span> <span class="st">"horizontal_mid"</span></span>
<span></span>
<span><span class="co"># `getSparkxGenes()` would work, too</span></span>
<span><span class="va">input_genes</span> <span class="op">&lt;-</span> <span class="va">spark_df</span><span class="op">[[</span><span class="st">"genes"</span><span class="op">]</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># define start and end positions of the trajectory directly</span></span>
<span><span class="va">object_t269</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="../reference/addSpatialTrajectory.html">addSpatialTrajectory</a></span><span class="op">(</span></span>
<span>    object <span class="op">=</span> <span class="va">object_t269</span>,</span>
<span>    id <span class="op">=</span> <span class="va">id</span>,</span>
<span>    start <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"1.5mm"</span>, <span class="st">"4mm"</span><span class="op">)</span>,</span>
<span>    end <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"6.5mm"</span>, <span class="st">"4mm"</span><span class="op">)</span>,</span>
<span>    overwrite <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="co"># note: the results are NOT stored in the SPATA2 object but in a separate object</span></span>
<span><span class="va">sts_out</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="../reference/spatialTrajectoryScreening.html">spatialTrajectoryScreening</a></span><span class="op">(</span></span>
<span>    object <span class="op">=</span> <span class="va">object_t269</span>, </span>
<span>    id <span class="op">=</span> <span class="va">id</span>, <span class="co"># ID of the spatial trajectory</span></span>
<span>    variables <span class="op">=</span> <span class="va">input_genes</span> <span class="co"># the variables/genes to include in the screening </span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">sts_out</span><span class="op">)</span></span></code></pre></div>
<p><strong>Note</strong>: The output of
<code><a href="../reference/spatialTrajectoryScreening.html">spatialTrajectoryScreening()</a></code> is <strong>not</strong> saved
in the <code>SPATA2</code> object but returned in a separate S4 object
of class <code>SpatialTrajectoryScreening</code>. Do
<strong>not</strong> overwrite the <code>SPATA2</code> object by writing
<code>object_t269 &lt;- spatialTrajectoryScreening(object = object_t269, id = id, ...)</code>.</p>
</div>
<div class="section level2">
<h2 id="results">3. Results<a class="anchor" aria-label="anchor" href="#results"></a>
</h2>
<p>The first step of the screening identifies pattern that are unlikely
due to random gene expression. The second step fits the non-random gene
expression pattern to predefined models which guides in interpretation
and screening for specific gene expression pattern.</p>
<p>Slot <span class="citation">@results$significance</span> contains a
data.frame with one row for each screened variable which provides
information regarding the degree of randomness the inferred pattern
contains as quantified by the total variation (tot_var). The p-value
gives the probability to obtain such a total variation under complete
randomness and indicates the degree of significance. Column <em>fdr</em>
contains the adjusted p-value according to the False Discovery Rate.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sign_df</span> <span class="op">&lt;-</span> </span>
<span>  <span class="va">sts_out</span><span class="op">@</span><span class="va">results</span><span class="op">$</span><span class="va">significance</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">fdr</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># show significance data.frame</span></span>
<span><span class="va">sign_df</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 1,492 × 6</span></span></span>
<span><span class="co">##    variables  rel_var tot_var p_value norm_var     fdr</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> A1BG        -<span style="color: #BB0000;">0.597</span>    7.77  0.000<span style="text-decoration: underline;">7</span>    0.155 0.001<span style="text-decoration: underline;">43</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> A2M         -<span style="color: #BB0000;">0.348</span>    6.19  0         0.124 0      </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> ABAT        -<span style="color: #BB0000;">0.372</span>    8.68  0.006<span style="text-decoration: underline;">3</span>    0.174 0.009<span style="text-decoration: underline;">88</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> ABCD3       -<span style="color: #BB0000;">0.282</span>    9.26  0.021<span style="text-decoration: underline;">5</span>    0.185 0.029<span style="text-decoration: underline;">2</span> </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> ABHD17C     -<span style="color: #BB0000;">0.471</span>    9.32  0.023<span style="text-decoration: underline;">3</span>    0.186 0.031<span style="text-decoration: underline;">0</span> </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> ABHD2       -<span style="color: #BB0000;">0.552</span>    6.60  0         0.132 0      </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> ABL1        -<span style="color: #BB0000;">0.671</span>    9.59  0.035<span style="text-decoration: underline;">4</span>    0.192 0.044<span style="text-decoration: underline;">4</span> </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> AC104986.2  -<span style="color: #BB0000;">0.458</span>    8.25  0.002<span style="text-decoration: underline;">1</span>    0.165 0.003<span style="text-decoration: underline;">69</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> ACAD8       -<span style="color: #BB0000;">0.631</span>    8.29  0.002<span style="text-decoration: underline;">4</span>    0.166 0.004<span style="text-decoration: underline;">18</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> ACBD6       -<span style="color: #BB0000;">0.761</span>    8.87  0.010<span style="text-decoration: underline;">5</span>    0.177 0.015<span style="text-decoration: underline;">6</span> </span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 1,482 more rows</span></span></span></code></pre>
<div class="section level3">
<h3 id="non-random-gene-expression-gradients">3.1 Non-random gene expression gradients<a class="anchor" aria-label="anchor" href="#non-random-gene-expression-gradients"></a>
</h3>
<p>The figures below show examples for gene expression gradient
identified as non-random.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># extract variables names</span></span>
<span><span class="va">non_random</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getSgsResultsVec.html">getSgsResultsVec</a></span><span class="op">(</span><span class="va">sts_out</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fl">4</span><span class="op">)</span></span>
<span></span>
<span><span class="va">trajectory_add_on</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="../reference/ggpLayerSpatialTrajectories.html">ggpLayerSpatialTrajectories</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">object_t269</span>, ids <span class="op">=</span> <span class="st">"horizontal_mid"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plotSurfaceComparison.html">plotSurfaceComparison</a></span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">object_t269</span>,</span>
<span>  color_by <span class="op">=</span> <span class="va">non_random</span>,</span>
<span>  pt_clrsp <span class="op">=</span> <span class="st">"Reds 3"</span>,</span>
<span>  outline <span class="op">=</span> <span class="cn">T</span>,</span>
<span>  nrow <span class="op">=</span> <span class="fl">1</span></span>
<span>  <span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="va">trajectory_add_on</span></span></code></pre></div>
<p><img src="spatial-trajectory-screening_files/figure-html/unnamed-chunk-6-1.png" width="100%"></p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotStsLineplot.html">plotStsLineplot</a></span><span class="op">(</span><span class="va">object_t269</span>, variables <span class="op">=</span> <span class="va">non_random</span>, id <span class="op">=</span> <span class="st">"horizontal_mid"</span>, line_color <span class="op">=</span> <span class="st">"red"</span>, nrow <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> </span></code></pre></div>
<p><img src="spatial-trajectory-screening_files/figure-html/unnamed-chunk-7-1.png" width="100%"></p>
</div>
<div class="section level3">
<h3 id="random-gene-expression-gradients">3.2 Random gene expression gradients<a class="anchor" aria-label="anchor" href="#random-gene-expression-gradients"></a>
</h3>
<p>The figures below show examples for gene expression gradient
identified as random.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># extract random variable names </span></span>
<span><span class="va">random</span> <span class="op">&lt;-</span> </span>
<span>  <span class="va">sts_out</span><span class="op">@</span><span class="va">results</span><span class="op">$</span><span class="va">significance</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">fdr</span> <span class="op">&gt;</span> <span class="fl">0.05</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html" class="external-link">slice_max</a></span><span class="op">(</span><span class="va">tot_var</span>, n <span class="op">=</span> <span class="fl">4</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html" class="external-link">pull</a></span><span class="op">(</span><span class="va">variables</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fl">4</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plotSurfaceComparison.html">plotSurfaceComparison</a></span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">object_t269</span>,</span>
<span>  color_by <span class="op">=</span> <span class="va">random</span>,</span>
<span>  pt_clrsp <span class="op">=</span> <span class="st">"BuPu"</span>,</span>
<span>  outline <span class="op">=</span> <span class="cn">T</span>, </span>
<span>  nrow <span class="op">=</span> <span class="fl">1</span></span>
<span>  <span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="va">trajectory_add_on</span></span></code></pre></div>
<p><img src="spatial-trajectory-screening_files/figure-html/unnamed-chunk-8-1.png" width="100%"></p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotStsLineplot.html">plotStsLineplot</a></span><span class="op">(</span><span class="va">object_t269</span>, variables <span class="op">=</span> <span class="va">random</span>, id <span class="op">=</span> <span class="st">"horizontal_mid"</span>, line_color <span class="op">=</span> <span class="st">"blue"</span>, nrow <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> </span></code></pre></div>
<p><img src="spatial-trajectory-screening_files/figure-html/unnamed-chunk-9-1.png" width="100%"></p>
</div>
<div class="section level3">
<h3 id="random-vs--non-random">Random vs. non-random<a class="anchor" aria-label="anchor" href="#random-vs--non-random"></a>
</h3>
<p>Looking at the gradient of variable <em>LEPROT</em> (random, blue)
and <em>SHISA5</em> (non-random, red) both feature a descending pattern
along the trajectory. How come that one is identified as most likely
random while the other is not? Spatial gradient screening decides which
variables are most likely random or not random by computing the variance
along the gradient, which is stored in the variable <em>tot_var</em> in
<code>sts_out@results$significance</code>. Consider both gradient plots
with the expression estimates plotted as black points, too. While
<em>SHISA5</em> features a comparatively smooth decline, <em>LEPROT</em>
does not. The variance along its gradient is too high to be deemed
significant.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># left plot</span></span>
<span><span class="fu"><a href="../reference/plotStsLineplot.html">plotStsLineplot</a></span><span class="op">(</span><span class="va">object_t269</span>, variables <span class="op">=</span> <span class="st">"LEPROT"</span>, line_color <span class="op">=</span> <span class="st">"blue"</span>, id <span class="op">=</span> <span class="st">"horizontal_mid"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># right plot</span></span>
<span><span class="fu"><a href="../reference/plotStsLineplot.html">plotStsLineplot</a></span><span class="op">(</span><span class="va">object_t269</span>, variables <span class="op">=</span> <span class="st">"SHISA5"</span>, line_color <span class="op">=</span> <span class="st">"red"</span>, id <span class="op">=</span> <span class="st">"horizontal_mid"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="spatial-trajectory-screening_files/figure-html/unnamed-chunk-10-1.png" width="50%"><img src="spatial-trajectory-screening_files/figure-html/unnamed-chunk-10-2.png" width="50%"></p>
</div>
<div class="section level3">
<h3 id="model-fits">3.2 Model fits<a class="anchor" aria-label="anchor" href="#model-fits"></a>
</h3>
<p>The second step uses predefined models and fits them to the inferred
gradients of the pattern identified as non-random in the previous step.
The figure below shows the default models used by SPATA2. They can be
extended by the user for specific querries with the argument
<code>model_add</code>.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/showModels.html">showModels</a></span><span class="op">(</span>nrow <span class="op">=</span> <span class="fl">4</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Distance along Trajectory [%]"</span><span class="op">)</span></span></code></pre></div>
<p><img src="spatial-trajectory-screening_files/figure-html/unnamed-chunk-11-1.png" width="100%"></p>
<p>Slot <span class="citation">@results$model_fits</span> contains the
model fitting results. It is a data.frame where each row corresponds to
a variable ~ model pair. The columns mae (mean absolute error) and rmse
(root mean squared error) indicate the quality of the fit. The lower the
value the better.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">best_fits</span> <span class="op">&lt;-</span> </span>
<span>  <span class="va">sts_out</span><span class="op">@</span><span class="va">results</span><span class="op">$</span><span class="va">model_fits</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">variables</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">sign_df</span><span class="op">[[</span><span class="st">"variables"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">variables</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html" class="external-link">slice_min</a></span><span class="op">(</span><span class="va">mae</span>, n <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="va">best_fits</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 1,492 × 4</span></span></span>
<span><span class="co">## <span style="color: #949494;"># Groups:   variables [1,492]</span></span></span>
<span><span class="co">##    variables  models              mae  rmse</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> A1BG       descending_linear 0.172 0.216</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> A2M        descending_linear 0.172 0.218</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> ABAT       descending_linear 0.193 0.246</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> ABCD3      descending_linear 0.215 0.253</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> ABHD17C    descending_linear 0.201 0.260</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> ABHD2      descending_linear 0.181 0.217</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> ABL1       descending_linear 0.179 0.225</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> AC104986.2 descending_linear 0.196 0.235</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> ACAD8      descending_linear 0.171 0.212</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> ACBD6      descending_linear 0.240 0.281</span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 1,482 more rows</span></span></span></code></pre>
<p>The following code chunk extracts the genes that followed each model
best.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">best_fits_by_model</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">best_fits</span>, <span class="va">models</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html" class="external-link">slice_min</a></span><span class="op">(</span><span class="va">mae</span>, n <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">rmse</span> <span class="op">&lt;</span> <span class="fl">0.2</span><span class="op">)</span> <span class="co"># threshold suggestions for root mean squared error</span></span>
<span></span>
<span><span class="va">best_fits_by_model</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 6 × 4</span></span></span>
<span><span class="co">## <span style="color: #949494;"># Groups:   models [6]</span></span></span>
<span><span class="co">##   variables models               mae  rmse</span></span>
<span><span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>              <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span> VGF       ascending_gradual  0.110 0.171</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span> ARC       ascending_linear   0.132 0.184</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">3</span> RDH10     descending_gradual 0.133 0.163</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">4</span> SIPA1L2   descending_linear  0.112 0.141</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">5</span> SELENOP   peak_moderate      0.121 0.145</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">6</span> MOBP      peak_sharp         0.124 0.155</span></span></code></pre>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSurfaceComparison.html">plotSurfaceComparison</a></span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">object_t269</span>, </span>
<span>  color_by <span class="op">=</span> <span class="va">best_fits_by_model</span><span class="op">[[</span><span class="st">"variables"</span><span class="op">]</span><span class="op">]</span>, </span>
<span>  outline <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>  display_image <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>  pt_clrsp <span class="op">=</span> <span class="st">"Reds 3"</span>, </span>
<span>  nrow <span class="op">=</span> <span class="fl">2</span></span>
<span>  <span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="va">trajectory_add_on</span></span></code></pre></div>
<p><img src="spatial-trajectory-screening_files/figure-html/unnamed-chunk-14-1.png" width="100%"></p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotStsLineplot.html">plotStsLineplot</a></span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">object_t269</span>, </span>
<span>  variables <span class="op">=</span> <span class="va">best_fits_by_model</span><span class="op">[[</span><span class="st">"variables"</span><span class="op">]</span><span class="op">]</span>, </span>
<span>  id <span class="op">=</span> <span class="st">"horizontal_mid"</span>, </span>
<span>  line_color <span class="op">=</span> <span class="st">"red"</span>, </span>
<span>  nrow <span class="op">=</span> <span class="fl">2</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="spatial-trajectory-screening_files/figure-html/unnamed-chunk-15-1.png" width="100%"></p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Jan Kueckelhaus, Dieter-Henrik Heiland.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

      </footer>
</div>






  </body>
</html>
