<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Gene Set Enrichment Analysis â€¢ SPATA2</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Gene Set Enrichment Analysis">
<meta property="og:description" content="SPATA2">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SPATA2</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/spata-v2-installation.html">Installation</a>
</li>
<li>
  <a href="../articles/spata-v2-spata-data.html">SPATAData</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">Around the spata2 object</li>
    <li>
      <a href="../articles/spata-v2-object-initiation-and-manipulation.html">Object Set Up and Manipulation</a>
    </li>
    <li>
      <a href="../articles/spata-v2-platform-compatibility.html">Platform Compatibility</a>
    </li>
    <li>
      <a href="../articles/spata-v2-autoencoder.html">Autoencoder</a>
    </li>
    <li>
      <a href="../articles/spata-v2-genes-gene-sets-and-features.html">Genes, Gene Sets and Features</a>
    </li>
    <li class="dropdown-header">Image Analysis</li>
    <li>
      <a href="../articles/spata-v2-image-handling.html">Image Handling</a>
    </li>
    <li>
      <a href="../articles/spata-v2-image-annotations.html">Image Annotations</a>
    </li>
    <li>
      <a href="../articles/spata-v2-spatial-segmentation.html">Spatial Segmentation</a>
    </li>
    <li class="dropdown-header">Analysis of Gene Expression</li>
    <li>
      <a href="../articles/spata-v2-cnv-analysis.html">Copy Number Variations</a>
    </li>
    <li>
      <a href="../articles/spata-v2-clustering.html">Clustering</a>
    </li>
    <li>
      <a href="../articles/spata-v2-dea.html">Differential Expression</a>
    </li>
    <li>
      <a href="../articles/spata-v2-gsea.html">Gene Set Enrichment</a>
    </li>
    <li class="dropdown-header">Expression Gradients &amp; Gradient Screening</li>
    <li>
      <a href="../articles/spata-v2-spatial-trajectories.html">Spatial Trajectories</a>
    </li>
    <li>
      <a href="../articles/spata-v2-spatial-trajectory-screening.html">Spatial Trajectory Screening</a>
    </li>
    <li>
      <a href="../articles/spata-v2-image-annotation-screening.html">Image Annotation Screening</a>
    </li>
    <li>
      <a href="../articles/spata-v2-scd.html">Integrate Single Cell Deconvolution</a>
    </li>
    <li class="dropdown-header">Visualization</li>
    <li>
      <a href="../articles/spata-v2-plotting-surface.html">Surface Plotting</a>
    </li>
    <li>
      <a href="../articles/spata-v2-plotting-misc.html">Miscellaneous</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Code &amp; Concept
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/cc-spatial-measures.html">Spatial Measures</a>
    </li>
    <li>
      <a href="../articles/cc-spatial-segmentation-vs-image-annotations.html">Spatial Segmentation vs. Image Annotations</a>
    </li>
    <li>
      <a href="../articles/cc-spatial-trajectory-screening.html">Spatial Trajectory Screening</a>
    </li>
    <li>
      <a href="../articles/cc-image-annotation-screening.html">Image Annotation Screening</a>
    </li>
    <li>
      <a href="../articles/cc-model-fitting-in-spatial-transcriptomic-studies.html">Model Fitting</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Gene Set Enrichment Analysis</h1>
            
      
      
      <div class="hidden name"><code>spata-v2-gsea.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="prerequisites">1. Prerequisites<a class="anchor" aria-label="anchor" href="#prerequisites"></a>
</h2>
<ul>
<li><a href="spata-v2-dea.html">Differential Expression
Analysis</a></li>
</ul>
</div>
<div class="section level2">
<h2 id="introduction">1. Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>For a deeper into the results of the DEA one can look for gene sets
that are enriched in the proposed clustering or created based on
histology. <code>SPATA2</code> implements the <a href="https://github.com/montilab/hypeR" class="external-link"><code>hypeR</code></a> package
which uses hypergeometric testing for enriched gene sets.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># load packages </span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://themilolab.com/" class="external-link">SPATA2</a></span><span class="op">)</span> 
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SPATAData</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span>

<span class="va">object_t275</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/downloadSpataObject.html">downloadSpataObject</a></span><span class="op">(</span>sample_name <span class="op">=</span> <span class="st">"275_T"</span><span class="op">)</span>

<span class="co"># load clustering list</span>
<span class="va">object_t275</span> <span class="op">&lt;-</span> 
  <span class="fu"><a href="../reference/setDefault.html">setDefault</a></span><span class="op">(</span>
    object <span class="op">=</span> <span class="va">object_t275</span>, 
    display_image <span class="op">=</span> <span class="cn">TRUE</span>,
    clrp <span class="op">=</span> <span class="st">"jco"</span>,
    pt_clrp <span class="op">=</span> <span class="st">"jco"</span>
  <span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"snn_clustering"</span><span class="op">)</span>

<span class="va">object_t275</span> <span class="op">&lt;-</span> 
  <span class="fu"><a href="../reference/addFeatures.html">addFeatures</a></span><span class="op">(</span>
    object <span class="op">=</span> <span class="va">object_t275</span>, 
    feature_df <span class="op">=</span> <span class="va">snn_clustering</span><span class="op">[[</span><span class="st">"275_T"</span><span class="op">]</span><span class="op">]</span>,
    overwrite <span class="op">=</span> <span class="cn">TRUE</span>
  <span class="op">)</span>

<span class="co"># histology only</span>
<span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">object_t275</span>, pt_alpha <span class="op">=</span> <span class="fl">0</span><span class="op">)</span>

<span class="co"># colored by clustering</span>
<span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">object_t275</span>, color_by <span class="op">=</span> <span class="st">"snn"</span><span class="op">)</span></code></pre></div>
<div class="figure">
<img src="spata-v2-gsea_files/figure-html/unnamed-chunk-2-1.png" alt="Fig.1 Shared Nearest Neighbour clustering will be analyzed with GSEA." width="50%"><img src="spata-v2-gsea_files/figure-html/unnamed-chunk-2-2.png" alt="Fig.1 Shared Nearest Neighbour clustering will be analyzed with GSEA." width="50%"><p class="caption">
Fig.1 Shared Nearest Neighbour clustering will be analyzed with GSEA.
</p>
</div>
</div>
<div class="section level2">
<h2 id="running-the-analysis">2. Running the analysis<a class="anchor" aria-label="anchor" href="#running-the-analysis"></a>
</h2>
<p>The function <code><a href="../reference/runGSEA.html">runGSEA()</a></code> conducts the computation. It
requires the results from <code><a href="../reference/runDEA.html">runDEA()</a></code> which conducts the
DE-analysis. By default, the function uses all gene sets that are saved
in the <code>spata2</code> object. If there are gene sets that you donâ€™t
want to test against you can provide a subsetted list of gene sets.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/printGeneSetOverview.html">printGeneSetOverview</a></span><span class="op">(</span><span class="va">object_t275</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">##        Class Available Gene Sets</span>
<span class="co">## 1         BC                 289</span>
<span class="co">## 2      BP.GO                7269</span>
<span class="co">## 3      CC.GO                 972</span>
<span class="co">## 4 Cell.types                   3</span>
<span class="co">## 5         HM                  50</span>
<span class="co">## 6      MF.GO                1581</span>
<span class="co">## 7        New                   1</span>
<span class="co">## 8       RCTM                1493</span></code></pre>
<p>The default <code>spata2</code> object contains a variety of gene
sets.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">gs_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getGeneSetDf.html">getGeneSetList</a></span><span class="op">(</span><span class="va">object_t275</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">gs_list</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] 11658</span></code></pre>
<p>For the sake of clarity, his example only uses Biocarta (BC) and
Hallmark (HM) gene sets.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># subset gene sets that start with HM or BC</span>
<span class="va">sub_vec</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_detect.html" class="external-link">str_detect</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">gs_list</span><span class="op">)</span>, pattern <span class="op">=</span> <span class="st">"^HM|^BC"</span><span class="op">)</span>

<span class="va">gs_list_sub</span> <span class="op">&lt;-</span> <span class="va">gs_list</span><span class="op">[</span><span class="va">sub_vec</span><span class="op">]</span>

<span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">gs_list_sub</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] 339</span></code></pre>
<p>Use argument gene_set_list if you want to provide a specific list of
gene sets. Keep it empty if you want to test against all gene sets
stored in the <code>spata2</code> object.</p>
</div>
<div class="section level2">
<h2 id="extracting-results">3. Extracting results<a class="anchor" aria-label="anchor" href="#extracting-results"></a>
</h2>
<p>The results can be manually extracted via the following
functions.</p>
<p><code><a href="../reference/getGseaResults.html">getGseaResults()</a></code> extracts a list of <code>hypeR</code>
objects - on for each group. <code>getGseaResultsDf()</code> extracts a
data.frame that results from merging the data of all <code>hypeR</code>
objects together. The group belonging of each gene set is saved in the
variable/column that is named according to the grouping variable - here
<em>snn</em>.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/getGseaDf.html">getGseaDf</a></span><span class="op">(</span>
  object <span class="op">=</span> <span class="va">object_t275</span>, 
  across <span class="op">=</span> <span class="st">"snn"</span>,
  method_de <span class="op">=</span> <span class="st">"wilcox"</span>, 
  n_gsets <span class="op">=</span> <span class="fl">20</span> <span class="co"># extract top 20 most significant gene sets</span>
<span class="op">)</span> </code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 160 x 10</span></span>
<span class="co">## <span style="color: #949494;"># Groups:   snn [8]</span></span>
<span class="co">##    snn   label          pval      fdr signature geneset overlap background hits </span>
<span class="co">##    <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span> <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span></span>
<span class="co">## <span style="color: #BCBCBC;"> 1</span> 0     HM_EPITHE~ 3.30<span style="color: #949494;">e</span><span style="color: #BB0000;">-19</span> 1.10<span style="color: #949494;">e</span><span style="color: #BB0000;">-16</span>       272     195      26      <span style="text-decoration: underline;">21</span>441 COL5~</span>
<span class="co">## <span style="color: #BCBCBC;"> 2</span> 0     HM_ANGIOG~ 8.3 <span style="color: #949494;">e</span><span style="color: #BB0000;">- 9</span> 1.4 <span style="color: #949494;">e</span><span style="color: #BB0000;">- 6</span>       272      34       8      <span style="text-decoration: underline;">21</span>441 POST~</span>
<span class="co">## <span style="color: #BCBCBC;"> 3</span> 0     HM_COMPLE~ 2.20<span style="color: #949494;">e</span><span style="color: #BB0000;">- 8</span> 2.40<span style="color: #949494;">e</span><span style="color: #BB0000;">- 6</span>       272     190      15      <span style="text-decoration: underline;">21</span>441 ADAM~</span>
<span class="co">## <span style="color: #BCBCBC;"> 4</span> 0     HM_CHOLES~ 2.8 <span style="color: #949494;">e</span><span style="color: #BB0000;">- 8</span> 2.40<span style="color: #949494;">e</span><span style="color: #BB0000;">- 6</span>       272      73      10      <span style="text-decoration: underline;">21</span>441 ALDO~</span>
<span class="co">## <span style="color: #BCBCBC;"> 5</span> 0     HM_COAGUL~ 4.30<span style="color: #949494;">e</span><span style="color: #BB0000;">- 8</span> 2.90<span style="color: #949494;">e</span><span style="color: #BB0000;">- 6</span>       272     120      12      <span style="text-decoration: underline;">21</span>441 TIMP~</span>
<span class="co">## <span style="color: #BCBCBC;"> 6</span> 0     HM_GLYCOL~ 1   <span style="color: #949494;">e</span><span style="color: #BB0000;">- 6</span> 5.10<span style="color: #949494;">e</span><span style="color: #BB0000;">- 5</span>       272     190      13      <span style="text-decoration: underline;">21</span>441 PKM,~</span>
<span class="co">## <span style="color: #BCBCBC;"> 7</span> 0     HM_APICAL~ 1.10<span style="color: #949494;">e</span><span style="color: #BB0000;">- 6</span> 5.10<span style="color: #949494;">e</span><span style="color: #BB0000;">- 5</span>       272     191      13      <span style="text-decoration: underline;">21</span>441 ITGB~</span>
<span class="co">## <span style="color: #BCBCBC;"> 8</span> 0     HM_TNFA_S~ 6   <span style="color: #949494;">e</span><span style="color: #BB0000;">- 6</span> 2.5 <span style="color: #949494;">e</span><span style="color: #BB0000;">- 4</span>       272     190      12      <span style="text-decoration: underline;">21</span>441 F3,J~</span>
<span class="co">## <span style="color: #BCBCBC;"> 9</span> 0     HM_UV_RES~ 1.3 <span style="color: #949494;">e</span><span style="color: #BB0000;">- 5</span> 4.9 <span style="color: #949494;">e</span><span style="color: #BB0000;">- 4</span>       272     141      10      <span style="text-decoration: underline;">21</span>441 ANXA~</span>
<span class="co">## <span style="color: #BCBCBC;">10</span> 0     BC_PLATEL~ 2.3 <span style="color: #949494;">e</span><span style="color: #BB0000;">- 5</span> 7.8 <span style="color: #949494;">e</span><span style="color: #BB0000;">- 4</span>       272      14       4      <span style="text-decoration: underline;">21</span>441 COL4~</span>
<span class="co">## <span style="color: #949494;"># i 150 more rows</span></span>
<span class="co">## <span style="color: #949494;"># i 1 more variable: overlap_perc &lt;dbl&gt;</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="plotting-results">4. Plotting results<a class="anchor" aria-label="anchor" href="#plotting-results"></a>
</h2>
<p>Gene set enrichment results can be visualized via dot plots. This can
be either done by group with by_group = TRUE or merged for all groups.
Figure 2 visualizes the enrichment for cluster 0 and 5 highlighting the
hypoxia associated activity in this area.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotGseaDotPlot.html">plotGseaDotPlot</a></span><span class="op">(</span>
  object <span class="op">=</span> <span class="va">object_t275</span>,
  across <span class="op">=</span> <span class="st">"snn"</span>,
  across_subset <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"0"</span>, <span class="st">"5"</span><span class="op">)</span>,
  n_gsets <span class="op">=</span> <span class="fl">10</span>,
  by_group <span class="op">=</span> <span class="cn">TRUE</span>, 
  transform_with <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>fdr <span class="op">=</span> <span class="va">log10</span><span class="op">)</span>,
  nrow <span class="op">=</span> <span class="fl">2</span>
<span class="op">)</span></code></pre></div>
<div class="figure">
<img src="spata-v2-gsea_files/figure-html/unnamed-chunk-7-1.png" alt="Fig.2 Dot plot to display GSEA results by group." width="100%"><p class="caption">
Fig.2 Dot plot to display GSEA results by group.
</p>
</div>
<p>Using by_group = FALSE the results are merged to one plot. Note that
GSEA did not result in any enrichment fo cluster 1 and 3. This might be
due to the reduced number of gene sets among which no enriched one was
found or due to a low number of upregulated genes.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotGseaDotPlot.html">plotGseaDotPlot</a></span><span class="op">(</span>
  object <span class="op">=</span> <span class="va">object_t275</span>,
  across <span class="op">=</span> <span class="st">"snn"</span>,
  across_subset <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="fl">0</span><span class="op">:</span><span class="fl">5</span><span class="op">)</span>,
  n_gsets <span class="op">=</span> <span class="fl">7</span>,
  pt_alpha <span class="op">=</span> <span class="fl">0.8</span>,
  transform_with <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>fdr <span class="op">=</span> <span class="va">log10</span><span class="op">)</span>,
  by_group <span class="op">=</span> <span class="cn">FALSE</span> <span class="co"># merge in one plot</span>
<span class="op">)</span> </code></pre></div>
<div class="figure">
<img src="spata-v2-gsea_files/figure-html/unnamed-chunk-8-1.png" alt="Fig.3 Merged GSEA dot plot." width="100%"><p class="caption">
Fig.3 Merged GSEA dot plot.
</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Jan Kueckelhaus, Dieter-Henrik Heiland, Simon Frerich.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
