<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Using Spatial Annotations • SPATA2</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Using Spatial Annotations">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SPATA2</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">3.1.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/installation.html">Installation</a>
</li>
<li>
  <a href="../articles/spata-data.html">SPATAData</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">Initiating a SPATA2 object</li>
    <li>
      <a href="../articles/initiation-and-preprocessing-customized.html">Initiation &amp; Preprocessing with individual data</a>
    </li>
    <li>
      <a href="../articles/initiation-and-preprocessing-merfish.html">Initiation &amp; Preprocessing with MERFISH data</a>
    </li>
    <li>
      <a href="../articles/initiation-and-preprocessing-visium.html">Initiation &amp; Preprocessing with Visium data</a>
    </li>
    <li>
      <a href="../articles/initiation-and-preprocessing-visium-hd.html">Initiation &amp; Preprocessing with VisiumHD data</a>
    </li>
    <li class="dropdown-header">Basic Manipulation of a SPATA2 object</li>
    <li>
      <a href="../articles/adding-data.html">Adding Data to SPATA2 objects</a>
    </li>
    <li>
      <a href="../articles/package-compatibility.html">Platform Compatibility</a>
    </li>
    <li>
      <a href="../articles/molecular-variables.html">Molecular Variables</a>
    </li>
    <li>
      <a href="../articles/image-handling.html">Image Handling</a>
    </li>
    <li>
      <a href="../articles/spatial-segmentation.html">Spatial Segmentation</a>
    </li>
    <li class="dropdown-header">Implemented Algorithms &amp; Concepts</li>
    <li>
      <a href="../articles/cnv.html">Copy Number Variations</a>
    </li>
    <li>
      <a href="../articles/clustering.html">Clustering</a>
    </li>
    <li>
      <a href="../articles/dimensional-reduction.html">Dimensional Reduction</a>
    </li>
    <li>
      <a href="../articles/dea.html">Differential Expression</a>
    </li>
    <li>
      <a href="../articles/gsea.html">Gene Set Enrichment</a>
    </li>
    <li>
      <a href="../articles/spatial-measures.html">Spatial Measures</a>
    </li>
    <li class="dropdown-header">Spatial Gradient Screening</li>
    <li>
      <a href="../articles/creating-spatial-annotations.html">Creating Spatial Annotations</a>
    </li>
    <li>
      <a href="../articles/using-spatial-annotations.html">Using Spatial Annotations</a>
    </li>
    <li>
      <a href="../articles/spatial-annotation-screening.html">Spatial Annotation Screening</a>
    </li>
    <li>
      <a href="../articles/creating-spatial-trajectories.html">Creating Spatial Trajectories</a>
    </li>
    <li>
      <a href="../articles/spatial-trajectory-screening.html">Spatial Trajectory Screening</a>
    </li>
    <li class="dropdown-header">Visualization</li>
    <li>
      <a href="../articles/plotting-surface.html">Surface Plotting</a>
    </li>
  </ul>
</li>
<li>
  <a href="../articles/newslog.html">Newslog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Using Spatial Annotations</h1>
            
      

      <div class="hidden name"><code>using-spatial-annotations.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="introduction">1. Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>This vignette exemplifies how to use Spatial Annotations in SPATA2.
It builds on the spatial annotations created in the vignette <a href="creating-spatial-annotations.html">creating spatial
annotations</a>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># load required packages</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://themilolab.com/" class="external-link">SPATA2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># load SPATA2 inbuilt data</span></span>
<span><span class="va">object_t313</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/loadExampleObject.html">loadExampleObject</a></span><span class="op">(</span><span class="st">"UKF313T"</span>, process <span class="op">=</span> <span class="cn">TRUE</span>, meta <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="image-subsetting">2. Image subsetting<a class="anchor" aria-label="anchor" href="#image-subsetting"></a>
</h2>
<p>Spatial annotations annotate space which can be directly translated
into the image regardless of the annotation was drawn based on
histomorphological features (Image Annotations) or numeric and grouping
features (Numeric- and GroupAnnotations). The annotated space can be
used to obtain image sections cropped to only include the annotated
area. <code><a href="../reference/getSpatialAnnotation.html">getSpatialAnnotation()</a></code> extracts an object of class
<code>SpatialAnnotation</code>.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># obtain the IDs of all spatial annotations</span></span>
<span><span class="fu"><a href="../reference/getSpatAnnIds.html">getSpatAnnIds</a></span><span class="op">(</span><span class="va">object_t313</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "necrotic_area"          "necrotic_center"        "necrotic_edge"         </span></span>
<span><span class="co">## [4] "necrotic_edge2"         "necrotic_edge2_transgr"</span></span></code></pre>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">necrotic_area</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="../reference/getSpatialAnnotation.html">getSpatialAnnotation</a></span><span class="op">(</span></span>
<span>    object <span class="op">=</span> <span class="va">object_t313</span>,</span>
<span>    id <span class="op">=</span> <span class="st">"necrotic_area"</span>, </span>
<span>    add_image <span class="op">=</span> <span class="cn">T</span></span>
<span>    <span class="op">)</span></span>
<span></span>
<span><span class="co"># print summary </span></span>
<span><span class="va">necrotic_area</span></span></code></pre></div>
<pre><code><span><span class="co">## An object of class SpatialAnnotation</span></span>
<span><span class="co">## ID:  necrotic_area </span></span>
<span><span class="co">## Sample:  UKF313T </span></span>
<span><span class="co">## Memory: 1.53 Mb</span></span>
<span><span class="co">## Tags:  necrotic, compr </span></span>
<span><span class="co">## Area information:</span></span>
<span><span class="co">##    outer :  286  vertices</span></span>
<span><span class="co">##    inner1 :  235  vertices</span></span>
<span><span class="co">## Cropped image dimensions (WxHxC):  238 x 250 x 3</span></span></code></pre>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># print slot names</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/methods/slot.html" class="external-link">slotNames</a></span><span class="op">(</span><span class="va">necrotic_area</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "parent_name" "area"        "id"          "image"       "image_info" </span></span>
<span><span class="co">## [6] "misc"        "sample"      "tags"        "version"</span></span></code></pre>
<p>By default, the image is cropped in a way that only the annotation is
included.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># visualize from the SPATA2 object, with expand = 0</span></span>
<span><span class="fu"><a href="../reference/plotSpatialAnnotations.html">plotSpatialAnnotations</a></span><span class="op">(</span><span class="va">object_t313</span>, ids <span class="op">=</span> <span class="st">"necrotic_area"</span>, expand <span class="op">=</span> <span class="fl">0</span>, fill <span class="op">=</span> <span class="cn">NA</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># expand = 0 is how the image is extracted when extracting the spatial annotation</span></span>
<span><span class="co"># (in SPATA2, images are plotted upside down to fit into the cartesian coordinate system)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="fu">EBImage</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/EBImage/man/spatial.html" class="external-link">flip</a></span><span class="op">(</span><span class="va">necrotic_area</span><span class="op">@</span><span class="va">image</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="using-spatial-annotations_files/figure-html/unnamed-chunk-3-1.png" width="50%"><img src="using-spatial-annotations_files/figure-html/unnamed-chunk-3-2.png" width="50%"></p>
<p>The argument <code>expand</code> can be used with
<code><a href="../reference/getSpatialAnnotation.html">getSpatialAnnotation()</a></code>, too, in order to manipulate with
how much extra space the image is cropped. This works in pixel as well
as with SPATA2’s SI unit system.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># visualize from the SPATA2 object, with expand = 0</span></span>
<span><span class="fu"><a href="../reference/plotSpatialAnnotations.html">plotSpatialAnnotations</a></span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">object_t313</span>,</span>
<span>  ids <span class="op">=</span> <span class="st">"necrotic_area"</span>, </span>
<span>  expand <span class="op">=</span> <span class="st">"500um"</span>, </span>
<span>  fill <span class="op">=</span> <span class="cn">NA</span>, </span>
<span>  sb_dist <span class="op">=</span> <span class="st">"500um"</span>, </span>
<span>  sb_pos <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"4.5mm"</span>, <span class="st">"3mm"</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="va">necrotic_area_expanded</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="../reference/getSpatialAnnotation.html">getSpatialAnnotation</a></span><span class="op">(</span></span>
<span>    object <span class="op">=</span> <span class="va">object_t313</span>,</span>
<span>    id <span class="op">=</span> <span class="st">"necrotic_area"</span>, </span>
<span>    add_image <span class="op">=</span> <span class="cn">T</span>, </span>
<span>    expand <span class="op">=</span> <span class="st">"500um"</span></span>
<span>    <span class="op">)</span></span>
<span></span>
<span><span class="va">expanded_image</span> <span class="op">&lt;-</span> <span class="fu">EBImage</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/EBImage/man/spatial.html" class="external-link">flip</a></span><span class="op">(</span><span class="va">necrotic_area_expanded</span><span class="op">@</span><span class="va">image</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">expanded_image</span><span class="op">)</span></span></code></pre></div>
<p><img src="using-spatial-annotations_files/figure-html/unnamed-chunk-4-1.png" width="50%"><img src="using-spatial-annotations_files/figure-html/unnamed-chunk-4-2.png" width="50%"></p>
</div>
<div class="section level2">
<h2 id="inferring-expression-gradients">3. Inferring expression gradients<a class="anchor" aria-label="anchor" href="#inferring-expression-gradients"></a>
</h2>
<p>The borders of spatial annotations can be used as reference points to
analyze expression of numeric features (e.g. gene expression) as a
function of distance to the annotated areas. The figure below
illustrates the concept.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">necrotic_ids</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"necrotic_area"</span>, <span class="st">"necrotic_edge"</span>, <span class="st">"necrotic_edge2"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># obtain distance data relative to spatial annotations (SA) with getCoordsDfSA()</span></span>
<span><span class="va">coords_df</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="../reference/getCoordsDfSA.html">getCoordsDfSA</a></span><span class="op">(</span><span class="va">object_t313</span>, ids <span class="op">=</span> <span class="va">necrotic_ids</span>, unit <span class="op">=</span> <span class="st">"mm"</span>, binwidth <span class="op">=</span> <span class="st">"200um"</span>, core0 <span class="op">=</span> <span class="cn">T</span><span class="op">)</span> </span>
<span></span>
<span><span class="co"># add distance to necrosis as meta feature to the SPATA2 object</span></span>
<span><span class="co"># to make them accessible for other functions</span></span>
<span><span class="va">object_t313</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="../reference/addFeatures.html">addFeatures</a></span><span class="op">(</span><span class="va">object_t313</span>, feature_df <span class="op">=</span> <span class="va">coords_df</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"barcodes"</span>, <span class="st">"dist"</span>, <span class="st">"bins_dist"</span><span class="op">)</span><span class="op">]</span>, overwrite <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># create ggproto layer for further surface plots</span></span>
<span><span class="va">necrotic_outline</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="../reference/ggpLayerSpatAnnOutline.html">ggpLayerSpatAnnOutline</a></span><span class="op">(</span><span class="va">object_t313</span>, ids <span class="op">=</span> <span class="va">necrotic_ids</span>, incl_edge <span class="op">=</span> <span class="cn">T</span>, fill <span class="op">=</span> <span class="st">"grey"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sas_screning</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="../reference/ggpLayerScreeningDirectionSAS.html">ggpLayerScreeningDirectionSAS</a></span><span class="op">(</span><span class="va">object_t313</span>, ids <span class="op">=</span> <span class="va">necrotic_ids</span>, line_size <span class="op">=</span> <span class="fl">0.75</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># show plots </span></span>
<span><span class="fu"><a href="../reference/plotImage.html">plotImage</a></span><span class="op">(</span><span class="va">object_t313</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="va">necrotic_outline</span> <span class="op">+</span> </span>
<span>  <span class="va">sas_screning</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span><span class="va">object_t313</span>, color_by <span class="op">=</span> <span class="st">"dist"</span>, pt_clrsp <span class="op">=</span> <span class="st">"plasma"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="va">necrotic_outline</span> <span class="op">+</span> </span>
<span>  <span class="va">sas_screning</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>color <span class="op">=</span> <span class="st">"Dist [mm]"</span><span class="op">)</span></span></code></pre></div>
<p><img src="using-spatial-annotations_files/figure-html/unnamed-chunk-5-1.png" width="50%"><img src="using-spatial-annotations_files/figure-html/unnamed-chunk-5-2.png" width="50%"></p>
<p>Inferring expression as function of distance is part of the Spatial
Gradient Screening algorithm (SAS). Therefore, functions related to this
concept feature the acronym <em>Sas</em> such as
<code><a href="../reference/plotSasLineplot.html">plotSasLineplot()</a></code> or <code><a href="../reference/plotSasHeatmap.html">plotSasHeatmap()</a></code>.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span><span class="va">object_t313</span>, color_by <span class="op">=</span> <span class="st">"HM_HYPOXIA"</span>, display_image <span class="op">=</span> <span class="cn">F</span>, outline <span class="op">=</span> <span class="cn">T</span>, pt_clrsp <span class="op">=</span> <span class="st">"BuPu"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="va">necrotic_outline</span> <span class="op">+</span> </span>
<span>  <span class="va">sas_screning</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plotSasLineplot.html">plotSasLineplot</a></span><span class="op">(</span><span class="va">object_t313</span>, ids <span class="op">=</span> <span class="va">necrotic_ids</span>, variables <span class="op">=</span> <span class="st">"HM_HYPOXIA"</span>, line_color <span class="op">=</span> <span class="st">"blue"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Distance to Necrotic Borders [mm]"</span><span class="op">)</span></span></code></pre></div>
<p><img src="using-spatial-annotations_files/figure-html/unnamed-chunk-6-1.png" width="50%"><img src="using-spatial-annotations_files/figure-html/unnamed-chunk-6-2.png" width="50%"></p>
<p>If the number of features displayed becomes too high for a lineplot
switch to <code><a href="../reference/plotSasHeatmap.html">plotSasHeatmap()</a></code>.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># use spatialAnnotationScreening() to identify non random genes </span></span>
<span><span class="co"># and genes with specific expression pattern related to your spatial annotations</span></span>
<span><span class="va">associated_with_necrosis</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"SLC2A1"</span>, <span class="st">"ADM"</span>, <span class="st">"ERO1A"</span>, <span class="st">"CD44"</span>, <span class="st">"STC2"</span>, <span class="st">"CA12"</span>, <span class="st">"VEGFA"</span>, <span class="st">"NDRG1"</span>, <span class="st">"TMEM158"</span>, <span class="st">"LOX"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">repelled_by_necrosis</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD74"</span>, <span class="st">"CLU"</span>, <span class="st">"IFITM3"</span>, <span class="st">"C1QA"</span>, <span class="st">"C1QB"</span>, <span class="st">"CD68"</span>, <span class="st">"SELENOP"</span>, <span class="st">"SEC61B"</span>, <span class="st">"HLA-DRB1"</span>, <span class="st">"EMC6"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plotSasHeatmap.html">plotSasHeatmap</a></span><span class="op">(</span><span class="va">object_t313</span>, variables <span class="op">=</span> <span class="va">associated_with_necrosis</span>, ids <span class="op">=</span> <span class="va">necrotic_ids</span>, clrsp <span class="op">=</span> <span class="st">"BuPu"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Distance to Necrotic Borders [mm]"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plotSasHeatmap.html">plotSasHeatmap</a></span><span class="op">(</span><span class="va">object_t313</span>, variables <span class="op">=</span> <span class="va">repelled_by_necrosis</span>, ids <span class="op">=</span> <span class="va">necrotic_ids</span>, clrsp <span class="op">=</span> <span class="st">"Reds 3"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Distance to Necrotic Borders [mm]"</span><span class="op">)</span></span></code></pre></div>
<p><img src="using-spatial-annotations_files/figure-html/unnamed-chunk-7-1.png" width="50%"><img src="using-spatial-annotations_files/figure-html/unnamed-chunk-7-2.png" width="50%"></p>
</div>
<div class="section level2">
<h2 id="core-environment-and-periphery">3. Core, Environment and Periphery<a class="anchor" aria-label="anchor" href="#core-environment-and-periphery"></a>
</h2>
<p>The concept of Spatial Annotation Screening differentiates three
types of areas when working with spatial annotations:</p>
<ul>
<li>Core: The area annotated by the outer (and inner, see annotation
<em>necrotic_area</em>) borders of a spatial annotation.</li>
<li>Environemnt: The area by which the annotation (or the annotations)
is surrounded.</li>
<li>Periphery: The area that exceeds the distance up to which the
expression pattern are inferred which is defined by the parameter
<code>distance</code>.</li>
</ul>
<p>By default <code>distance</code> is the maximum of distances between
the objects observations (here, barcoded spots) to their respective
closest annotation border. In the case of the three necrotic annotations
the maximum distance is approximately 2mm.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## in mm, since unit = "mm"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">coords_df</span><span class="op">$</span><span class="va">dist</span><span class="op">)</span>, digits <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 1.99</span></span></code></pre>
<p>This is referred to as <em>distance to edge (dte)</em> as
observations up to the tissue’s edge are included. If your hypothesis
requires a specific distance parameter up to which the expression
gradient is inferred use the <code>distance</code> parameter. Fig.6
visualizes the differences in the set up.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sas_areas</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/color_vector.html">color_vector</a></span><span class="op">(</span>clrp <span class="op">=</span> <span class="st">"npg"</span>, names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"core"</span>, <span class="st">"environment"</span>, <span class="st">"periphery"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">necrotic_outline_transp</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="../reference/ggpLayerSpatAnnOutline.html">ggpLayerSpatAnnOutline</a></span><span class="op">(</span><span class="va">object_t313</span>, ids <span class="op">=</span> <span class="va">necrotic_ids</span>, incl_edge <span class="op">=</span> <span class="cn">T</span>, fill <span class="op">=</span> <span class="cn">NA</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># rel_loc = location of each observation relative to the set up</span></span>
<span><span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span><span class="va">coords_df</span>, color_by <span class="op">=</span> <span class="st">"rel_loc"</span>, clrp_adjust <span class="op">=</span> <span class="va">sas_areas</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="va">necrotic_outline_transp</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>subtitle <span class="op">=</span> <span class="st">"`distance` = 'dte'"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># the scenario changes when distance is specified</span></span>
<span><span class="va">coords_df2</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="../reference/getCoordsDfSA.html">getCoordsDfSA</a></span><span class="op">(</span><span class="va">object_t313</span>, ids <span class="op">=</span> <span class="va">necrotic_ids</span>, distance <span class="op">=</span> <span class="st">"1.5mm"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span><span class="va">coords_df2</span>, color_by <span class="op">=</span> <span class="st">"rel_loc"</span>, clrp_adjust <span class="op">=</span> <span class="va">sas_areas</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="va">necrotic_outline_transp</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>subtitle <span class="op">=</span> <span class="st">"`distance` = 1.5mm"</span><span class="op">)</span></span></code></pre></div>
<p><img src="using-spatial-annotations_files/figure-html/unnamed-chunk-9-1.png" width="50%"><img src="using-spatial-annotations_files/figure-html/unnamed-chunk-9-2.png" width="50%"></p>
<p>Another aspect to consider is the <code>core</code> parameter.
Depending on whether it is set to <code>core = TRUE</code> or
<code>core = FALSE</code> the inside of the annotation is included in
the gradient. It depends on the circumstances and your hypothesis
whether it makes sense to include the core or not. In case of the
necrotic areas we decided to set <code>core = FALSE</code> since we were
particularly interested in the reaction of the surrounding of the
necrotic areas.</p>
</div>
<div class="section level2">
<h2 id="using-external-coordinate-data-frames">4. Using external coordinate data.frames<a class="anchor" aria-label="anchor" href="#using-external-coordinate-data-frames"></a>
</h2>
<p>You can relate spatial annotations to observations that do not live
in the <code>SPATA2</code> object but in a separate data.frame. Only
requirements are that the data.frame contains x- and y- coordinates
scaled to the resolution of the active image (if there is one) and a
variable called <em>barcodes</em> identifying each observation
uniquely.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># get sc deconvolution data</span></span>
<span><span class="va">sc_input</span> <span class="op">&lt;-</span> <span class="va">example_data</span><span class="op">$</span><span class="va">sc_input_lmu_mci</span></span>
<span></span>
<span><span class="va">sc_input</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 4,356 × 5</span></span></span>
<span><span class="co">##    barcodes     x     y cell_type             spot_id           </span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>                 <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>             </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> cell1     442.  378. Neurons               AAACAAGTATCTCCCA-1</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> cell2     412.  140. Mural cells           AAACAGAGCGACTCCT-1</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> cell3     412.  137. Astrocytes            AAACAGAGCGACTCCT-1</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> cell4     413.  142. Astrocytes            AAACAGAGCGACTCCT-1</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> cell5     412.  138. Microglia             AAACAGAGCGACTCCT-1</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> cell6     424.  447. Monocytes/Macrophages AAACATTTCCCGGATT-1</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> cell7     425.  452. Microglia             AAACATTTCCCGGATT-1</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> cell8     426.  450. Microglia             AAACATTTCCCGGATT-1</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> cell9     492.  341. OPCs                  AAACCCGAACGAAATC-1</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> cell10    494.  346. OPCs                  AAACCCGAACGAAATC-1</span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 4,346 more rows</span></span></span></code></pre>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># load mouse brain sectionas as example data</span></span>
<span><span class="va">object_mouse</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/loadExampleObject.html">loadExampleObject</a></span><span class="op">(</span><span class="st">"LMU_MCI"</span>, process <span class="op">=</span> <span class="cn">TRUE</span>, meta <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">hemispheres</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ggpLayerTissueOutline.html">ggpLayerTissueOutline</a></span><span class="op">(</span><span class="va">object_mouse</span><span class="op">)</span></span>
<span><span class="va">injuries</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ggpLayerSpatAnnOutline.html">ggpLayerSpatAnnOutline</a></span><span class="op">(</span><span class="va">object_mouse</span>, ids <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"inj1"</span>, <span class="st">"inj2"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># left plot</span></span>
<span><span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span><span class="va">object_mouse</span>, color_by <span class="op">=</span> <span class="st">"clusters"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="va">injuries</span></span>
<span></span>
<span><span class="co"># right plot</span></span>
<span><span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span><span class="va">sc_input</span>, color_by <span class="op">=</span> <span class="st">"cell_type"</span>, pt_clrp <span class="op">=</span> <span class="st">"tab20b"</span>, pt_size <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="va">injuries</span> <span class="op">+</span> </span>
<span>  <span class="va">hemispheres</span>  </span></code></pre></div>
<p><img src="using-spatial-annotations_files/figure-html/unnamed-chunk-10-1.png" width="50%"><img src="using-spatial-annotations_files/figure-html/unnamed-chunk-10-2.png" width="50%"></p>
<p>Use <code>sc_input</code> as input for argument
<code>coords_df</code> of <code><a href="../reference/getCoordsDfSA.html">getCoordsDfSA()</a></code>. This way, not
the coordinates stored in the <code>SPATA2</code> objects are used but
the ones you provide.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># relate to spatial annotations ...</span></span>
<span><span class="va">sc_input_sa</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="../reference/getCoordsDfSA.html">getCoordsDfSA</a></span><span class="op">(</span></span>
<span>    object <span class="op">=</span> <span class="va">object_mouse</span>,</span>
<span>    ids <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"inj1"</span>, <span class="st">"inj2"</span><span class="op">)</span>,</span>
<span>    resolution <span class="op">=</span> <span class="st">"300um"</span>, </span>
<span>    coords_df <span class="op">=</span> <span class="va">sc_input</span></span>
<span>    <span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">rel_loc</span> <span class="op">!=</span> <span class="st">"core"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># ... which adds the spatial information to the data.frame</span></span>
<span><span class="va">sc_input_sa</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 4,092 × 15</span></span></span>
<span><span class="co">##    barcodes     x     y tissue_section_id cell_type spot_id tissue_section  dist</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> cell1     442.  378. tissue_section_2  Neurons   AAACAA… tissue_sectio… 0.502</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> cell10    494.  346. tissue_section_2  OPCs      AAACCC… tissue_sectio… 1.18 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> cell100   408.  375. tissue_section_2  Oligos    AACGGA… tissue_sectio… 0.659</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> cell1000  370.  171. tissue_section_1  Monocyte… ATGCCG… tissue_sectio… 0.366</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> cell1001  371.  175. tissue_section_1  Oligos    ATGCCG… tissue_sectio… 0.415</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> cell1002  344.  442. tissue_section_2  Neurons   ATGCTC… tissue_sectio… 1.14 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> cell1003  345.  441. tissue_section_2  Monocyte… ATGCTC… tissue_sectio… 1.13 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> cell1004  346.  445. tissue_section_2  Astrocyt… ATGCTC… tissue_sectio… 1.11 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> cell1005  347.  441. tissue_section_2  Astrocyt… ATGCTC… tissue_sectio… 1.10 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> cell1006  481.  419. tissue_section_2  OPCs      ATGCTT… tissue_sectio… 0.567</span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 4,082 more rows</span></span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 7 more variables: border &lt;chr&gt;, dist_unit &lt;chr&gt;, bins_dist &lt;fct&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   rel_loc &lt;fct&gt;, angle &lt;dbl&gt;, bins_angle &lt;fct&gt;, id &lt;fct&gt;</span></span></span></code></pre>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ct</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Monocytes/Macrophages"</span>, <span class="st">"Microglia"</span>, <span class="st">"Neurons"</span>, <span class="st">"Astrocytes"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">sc_input_sa</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_bar</a></span><span class="op">(</span>mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">bins_dist</span>, fill <span class="op">=</span> <span class="va">cell_type</span><span class="op">)</span>, position <span class="op">=</span> <span class="st">"stack"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="../reference/scale_color_add_on.html">scale_color_add_on</a></span><span class="op">(</span>aes <span class="op">=</span> <span class="st">"fill"</span>, variable <span class="op">=</span> <span class="va">sc_input_sa</span><span class="op">$</span><span class="va">cell_type</span>, clrp <span class="op">=</span> <span class="st">"tab20b"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="using-spatial-annotations_files/figure-html/unnamed-chunk-11-1.png" width="100%"></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Jan Kueckelhaus, Dieter-Henrik Heiland.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

      </footer>
</div>






  </body>
</html>
