<!DOCTYPE html>
<<<<<<< HEAD
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Obtain a summarized trajectory data.frame — getTrajectoryDf • SPATA2</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Obtain a summarized trajectory data.frame — getTrajectoryDf"><meta property="og:description" content="Computes the expression trends of all specified variables
along the direction of the spatial trajectory."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
=======
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Obtain a trajectory data.frame — getTrajectoryDf • SPATA2</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Obtain a trajectory data.frame — getTrajectoryDf"><meta property="og:description" content="Extracts a data.frame that contains information about barcode-spots
needed for analysis related to spatialTrajectoryScreening()."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
>>>>>>> f7196fe9dc51c1738b724567a2e437b9d4780e28
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SPATA2</a>
<<<<<<< HEAD
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.0</span>
=======
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.0.0</span>
>>>>>>> f7196fe9dc51c1738b724567a2e437b9d4780e28
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
<<<<<<< HEAD
  <a href="../articles/spata-v2-introduction-and-philosophy.html">About</a>
</li>
<li>
  <a href="../articles/spata-v2-installation.html">Installation</a>
</li>
<li>
  <a href="../articles/spata-v2-terminology.html">Terminology</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
=======
  <a href="../articles/spata-v2-installation.html">Installation</a>
</li>
<li>
  <a href="../articles/spata-v2-spata-data.html">SPATAData</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
>>>>>>> f7196fe9dc51c1738b724567a2e437b9d4780e28
    Articles
     
    <span class="caret"></span>
  </a>
<<<<<<< HEAD
  <ul class="dropdown-menu" role="menu"><li class="dropdown-header">Around the Spata-Object</li>
=======
  <ul class="dropdown-menu" role="menu"><li class="dropdown-header">Around the spata2 object</li>
>>>>>>> f7196fe9dc51c1738b724567a2e437b9d4780e28
    <li>
      <a href="../articles/spata-v2-object-initiation-and-manipulation.html">Object Set Up and Manipulation</a>
    </li>
    <li>
<<<<<<< HEAD
      <a href="../articles/spata-v2-extract-add-and-join-data.html">Extracting, Adding and Joining Data</a>
    </li>
    <li>
      <a href="../articles/">Genes, Gene Sets and Features</a>
    </li>
    <li>
      <a href="../articles/spata-v2-platform-compatibility.html">Integrating Different Platforms</a>
    </li>
    <li class="dropdown-header">Analysis &amp; Computation</li>
    <li>
      <a href="../articles/spata-v2-autoencoder.html">Autoencoder Denoising</a>
    </li>
=======
      <a href="../articles/spata-v2-platform-compatibility.html">Platform Compatibility</a>
    </li>
    <li>
      <a href="../articles/spata-v2-autoencoder.html">Autoencoder</a>
    </li>
    <li>
      <a href="../articles/spata-v2-genes-gene-sets-and-features.html">Genes, Gene Sets and Features</a>
    </li>
    <li class="dropdown-header">Image Analysis</li>
    <li>
      <a href="../articles/spata-v2-image-handling.html">Image Handling</a>
    </li>
    <li>
      <a href="../articles/spata-v2-image-annotations.html">Image Annotations</a>
    </li>
    <li>
      <a href="../articles/spata-v2-spatial-segmentation.html">Spatial Segmentation</a>
    </li>
    <li class="dropdown-header">Analysis of Gene Expression</li>
>>>>>>> f7196fe9dc51c1738b724567a2e437b9d4780e28
    <li>
      <a href="../articles/spata-v2-cnv-analysis.html">Copy Number Variations</a>
    </li>
    <li>
<<<<<<< HEAD
      <a href="../articles/spata-v2-segmentation-and-clustering.html">Spatial Segmentation &amp; Clustering</a>
    </li>
    <li>
      <a href="../articles/spata-v2-de-analysis.html">Differential Expression Analysis</a>
    </li>
    <li>
      <a href="../articles/spata-v2-gsea.html">Gene Set Enrichment Analysis</a>
    </li>
    <li>
      <a href="../articles/spata-v2-trajectory-analysis.html">Spatial Trajectories</a>
=======
      <a href="../articles/spata-v2-clustering.html">Clustering</a>
    </li>
    <li>
      <a href="../articles/spata-v2-dea.html">Differential Expression</a>
    </li>
    <li>
      <a href="../articles/spata-v2-gsea.html">Gene Set Enrichment</a>
    </li>
    <li class="dropdown-header">Expression Gradients &amp; Gradient Screening</li>
    <li>
      <a href="../articles/spata-v2-spatial-trajectories.html">Spatial Trajectories</a>
    </li>
    <li>
      <a href="../articles/spata-v2-spatial-trajectory-screening.html">Spatial Trajectory Screening</a>
    </li>
    <li>
      <a href="../articles/spata-v2-image-annotation-screening.html">Image Annotation Screening</a>
    </li>
    <li>
      <a href="../articles/spata-v2-scd.html">Integrate Single Cell Deconvolution</a>
>>>>>>> f7196fe9dc51c1738b724567a2e437b9d4780e28
    </li>
    <li class="dropdown-header">Visualization</li>
    <li>
      <a href="../articles/spata-v2-plotting-surface.html">Surface Plotting</a>
    </li>
    <li>
<<<<<<< HEAD
      <a href="../articles/spata-v2-plotting-dimensional-reduction.html">Dimensional Reduction</a>
    </li>
    <li>
      <a href="../articles/spata-v2-plotting-descriptive-statistics.html">Descriptive Statistics</a>
    </li>
    <li>
      <a href="../articles/spata-v2-plotting-miscellaneous.html">Miscellaneous</a>
=======
      <a href="../articles/spata-v2-plotting-misc.html">Miscellaneous</a>
    </li>
  </ul></li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Code &amp; Concept
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/cc-spatial-measures.html">Spatial Measures</a>
    </li>
    <li>
      <a href="../articles/cc-spatial-segmentation-vs-image-annotations.html">Spatial Segmentation vs. Image Annotations</a>
    </li>
    <li>
      <a href="../articles/cc-spatial-trajectory-screening.html">Spatial Trajectory Screening</a>
    </li>
    <li>
      <a href="../articles/cc-image-annotation-screening.html">Image Annotation Screening</a>
    </li>
    <li>
      <a href="../articles/cc-model-fitting-in-spatial-transcriptomic-studies.html">Model Fitting</a>
>>>>>>> f7196fe9dc51c1738b724567a2e437b9d4780e28
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
<<<<<<< HEAD
    <h1>Obtain a summarized trajectory data.frame</h1>
=======
    <h1>Obtain a trajectory data.frame</h1>
>>>>>>> f7196fe9dc51c1738b724567a2e437b9d4780e28
    
    <div class="hidden name"><code>getTrajectoryDf.Rd</code></div>
    </div>

    <div class="ref-description">
<<<<<<< HEAD
    <p>Computes the expression trends of all specified variables
along the direction of the spatial trajectory.</p>
=======
    <p>Extracts a data.frame that contains information about barcode-spots
needed for analysis related to <code><a href="spatialTrajectoryScreening.html">spatialTrajectoryScreening()</a></code>.</p>
>>>>>>> f7196fe9dc51c1738b724567a2e437b9d4780e28
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">getTrajectoryDf</span><span class="op">(</span>
  <span class="va">object</span>,
<<<<<<< HEAD
  <span class="va">trajectory_name</span>,
  <span class="va">variables</span>,
  method_gs <span class="op">=</span> <span class="st">"mean"</span>,
  binwidth <span class="op">=</span> <span class="fl">5</span>,
  normalize <span class="op">=</span> <span class="cn">TRUE</span>,
  whole_sample <span class="op">=</span> <span class="cn">FALSE</span>,
  shift_wider <span class="op">=</span> <span class="cn">FALSE</span>,
  verbose <span class="op">=</span> <span class="cn">TRUE</span>,
  of_sample <span class="op">=</span> <span class="cn">NA</span>
=======
  <span class="va">id</span>,
  <span class="va">variables</span>,
  binwidth <span class="op">=</span> <span class="fu"><a href="getCCD.html">getCCD</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span>,
  n_bins <span class="op">=</span> <span class="cn">NA_integer_</span>,
  method_gs <span class="op">=</span> <span class="cn">NULL</span>,
  normalize <span class="op">=</span> <span class="cn">TRUE</span>,
  summarize_with <span class="op">=</span> <span class="cn">FALSE</span>,
  smooth_span <span class="op">=</span> <span class="fl">0</span>,
  format <span class="op">=</span> <span class="st">"wide"</span>,
  verbose <span class="op">=</span> <span class="cn">NULL</span>,
  <span class="va">...</span>
>>>>>>> f7196fe9dc51c1738b724567a2e437b9d4780e28
<span class="op">)</span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>object</dt>
<<<<<<< HEAD
<dd><p>A valid spata-object.</p></dd>
<dt>trajectory_name</dt>
<dd><p>The trajectory of interest specified
as a single character value.</p></dd>
<dt>variables</dt>
<dd><p>Character vector. The variables of interest:</p>
<ul><li><p><strong>Gene sets</strong>: Must be in <code><a href="getGeneSets.html">getGeneSets()</a></code></p></li>
<li><p><strong>Genes</strong>: Must be in <code><a href="getGenes.html">getGenes()</a></code></p></li>
<li><p><strong>Features</strong>: Must be numeric ones of <code><a href="getFeatureNames.html">getFeatureNames()</a></code></p></li>
</ul></dd>
<dt>method_gs</dt>
<dd><p>Character value. The method according to which gene sets will be handled
specified as a character of length one. This can be either <em>'mean'</em> or one
of <em>'gsva', 'ssgsea', 'zscore', or 'plage'</em>. The latter four will be given to
<code>gsva::GSVA()</code>.</p></dd>
<dt>binwidth</dt>
<dd><p>Numeric value. Denotes the binwidth with which to sort all
relevant barcode spots into groups that are then aligned with respect to the
chosen trajectory's direction.#'</p></dd>
<dt>normalize</dt>
<dd><p>Logical. If set to TRUE values will be scaled to 0-1.</p>
<p>Hint: Variables that are uniformly expressed can not be scaled and are discarded.</p></dd>
<dt>whole_sample</dt>
<dd><p>Logical. If TRUE, normalization of the values used
takes place in the light of the complete sample.</p></dd>
<dt>shift_wider</dt>
<dd><p>Logical. If set to TRUE the trajectory data.frame is
shifted to it's wider format. Formats can be changed via <code><a href="shiftTrajectoryDf.html">shiftTrajectoryDf()</a></code>.</p></dd>
=======
<dd><p>An object of class <code>spata2</code>.</p></dd>


<dt>id</dt>
<dd><p>Character value. Denotes the spatial trajectory
of interest.</p></dd>


<dt>variables</dt>
<dd><p>Character vector. The numeric variables of interest. Must be inside:</p>
<ul><li><p><strong>Gene sets</strong> Must be in <code><a href="getGeneSets.html">getGeneSets()</a></code></p></li>
<li><p><strong>Genes</strong> Must be in <code><a href="getGenes.html">getGenes()</a></code></p></li>
<li><p><strong>Features</strong> Must be in <code>getFeatureNames(..., of_class = "numeric")</code></p></li>
</ul></dd>


<dt>binwidth</dt>
<dd><p>Distance value. The width of the bins to which
the barcode-spots are assigned. Defaults to the center-center
distance: <code>binwidth = getCCD(object)</code>.</p></dd>


<dt>n_bins</dt>
<dd><p>Numeric value. Specifies the exact number of bins the barcodes
are binned into.</p></dd>


<dt>method_gs</dt>
<dd><p>Character value. The method according to which gene sets will
be handled specified as a character of length one. This can be either 'mean
or one of 'gsva', 'ssgsea', 'zscore', or 'plage'. The latter four will be given to gsva::GSVA().</p></dd>


<dt>normalize</dt>
<dd><p>Logical. If set to TRUE values will be scaled to 0-1.</p>
<p>Hint: Variables that are uniformly expressed can not be scaled and are discarded.</p></dd>


<dt>summarize_with</dt>
<dd><p>Character value. Name of the function with which to summarize.</p></dd>


<dt>smooth_span</dt>
<dd><p>Numeric value. Controls the degree of smoothing.
Given to argument <code>span</code> of <code><a href="https://rdrr.io/r/stats/loess.html" class="external-link">stats::loess()</a></code>.</p></dd>


>>>>>>> f7196fe9dc51c1738b724567a2e437b9d4780e28
<dt>verbose</dt>
<dd><p>Logical. If set to TRUE informative messages regarding
the computational progress will be printed.</p>
<p>(Warning messages will always be printed.)</p></dd>
<<<<<<< HEAD
<dt>of_sample</dt>
<dd><p>This argument is currently inactive. It might be reactivated when
spata-objects can store more than one sample.</p></dd>
</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>A summarized trajectory data.frame.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>Initially the compiled trajectory data.frame of the specified trajectory
is joined with the respective input of variables via <code><a href="joinWith.html">joinWithVariables()</a></code>.</p>
<p>The argument <code>binwidth</code> refers to the amount of which the barcode-spots of the
given trajectory will be summarized with regards to the trajectory's direction:
The amount of <code>binwidth</code> and the previously specified 'trajectory width' in <code><a href="createTrajectories.html">createTrajectories()</a></code>
determine the length and width of the sub-rectangles in which the rectangle the
trajectory embraces are splitted and in which all barcode-spots are binned.
Via <code><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">dplyr::summarize()</a></code> the variable-means of every sub-rectangle are calculated.
These mean-values are then arranged according to the trajectory's direction.</p>
<p>Eventually the data.frame is shifted via <code><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html" class="external-link">tidyr::pivot_longer()</a></code> to a data.frame in which
every observation refers to the mean-value of one of the specified variable-elements (e.g. a specified
gene set) of the particular sub-rectangle. The returned data.frame contains the following variables:</p>
<ul><li><p><em>trajectory_part</em>: Character. Specifies the trajectory's sub-part of the observation. (Negligible if there is
 only one trajectory part.)</p></li>
<li><p><em>trajectory_part_order</em>: Numeric. Indicates the order within the trajectory-part. (Negligible if there is
 only one trajectory part.)</p></li>
<li><p><em>trajectory_order</em>: Numeric. Indicates the order within the whole trajectory.</p></li>
<li><p><em>gene_sets, genes or features</em>: Character. The respective gene sets, gene or feature the value refers to.</p></li>
<li><p><em>values</em>: Numeric. The actual summarized values.</p></li>
</ul></div>
=======


<dt>...</dt>
<dd><p>Used to absorb deprecated arguments or functions.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>Data.frame. (See details for more.)</p>
    </div>
>>>>>>> f7196fe9dc51c1738b724567a2e437b9d4780e28

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
<<<<<<< HEAD
  <p></p><p>Developed by Jan Kueckelhaus, Dieter-Henrik Heiland.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.3.</p>
=======
  <p></p><p>Developed by Jan Kueckelhaus, Dieter-Henrik Heiland, Simon Frerich.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
>>>>>>> f7196fe9dc51c1738b724567a2e437b9d4780e28
</div>

      </footer></div>

  


  

  </body></html>

