<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>The HistoImage - Class — HistoImage-class • SPATA2</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="The HistoImage - Class — HistoImage-class"><meta property="og:description" content="S4 class that contains an histology image and information and data
about it. Usually live in slot @images of SpatialData objects."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SPATA2</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">3.1.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/installation.html">Installation</a>
</li>
<li>
  <a href="../articles/spata-data.html">SPATAData</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li class="dropdown-header">Initiating a SPATA2 object</li>
    <li>
      <a href="../articles/initiation-and-preprocessing-customized.html">Initiation &amp; Preprocessing with individual data</a>
    </li>
    <li>
      <a href="../articles/initiation-and-preprocessing-merfish.html">Initiation &amp; Preprocessing with MERFISH data</a>
    </li>
    <li>
      <a href="../articles/initiation-and-preprocessing-visium.html">Initiation &amp; Preprocessing with Visium data</a>
    </li>
    <li>
      <a href="../articles/initiation-and-preprocessing-visium-hd.html">Initiation &amp; Preprocessing with VisiumHD data</a>
    </li>
    <li class="dropdown-header">Basic Manipulation of a SPATA2 object</li>
    <li>
      <a href="../articles/adding-data.html">Adding Data to SPATA2 objects</a>
    </li>
    <li>
      <a href="../articles/package-compatibility.html">Platform Compatibility</a>
    </li>
    <li>
      <a href="../articles/molecular-variables.html">Molecular Variables</a>
    </li>
    <li>
      <a href="../articles/image-handling.html">Image Handling</a>
    </li>
    <li>
      <a href="../articles/spatial-segmentation.html">Spatial Segmentation</a>
    </li>
    <li class="dropdown-header">Implemented Algorithms &amp; Concepts</li>
    <li>
      <a href="../articles/cnv.html">Copy Number Variations</a>
    </li>
    <li>
      <a href="../articles/clustering.html">Clustering</a>
    </li>
    <li>
      <a href="../articles/dimensional-reduction.html">Dimensional Reduction</a>
    </li>
    <li>
      <a href="../articles/dea.html">Differential Expression</a>
    </li>
    <li>
      <a href="../articles/gsea.html">Gene Set Enrichment</a>
    </li>
    <li>
      <a href="../articles/spatial-measures.html">Spatial Measures</a>
    </li>
    <li class="dropdown-header">Spatial Gradient Screening</li>
    <li>
      <a href="../articles/creating-spatial-annotations.html">Creating Spatial Annotations</a>
    </li>
    <li>
      <a href="../articles/using-spatial-annotations.html">Using Spatial Annotations</a>
    </li>
    <li>
      <a href="../articles/spatial-annotation-screening.html">Spatial Annotation Screening</a>
    </li>
    <li>
      <a href="../articles/creating-spatial-trajectories.html">Creating Spatial Trajectories</a>
    </li>
    <li>
      <a href="../articles/spatial-trajectory-screening.html">Spatial Trajectory Screening</a>
    </li>
    <li class="dropdown-header">Visualization</li>
    <li>
      <a href="../articles/plotting-surface.html">Surface Plotting</a>
    </li>
  </ul></li>
<li>
  <a href="../articles/newslog.html">Newslog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>The <code>HistoImage</code> - Class</h1>

    <div class="hidden name"><code>HistoImage-class.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>S4 class that contains an histology image and information and data
about it. Usually live in slot @images of <code><a href="SpatialData-class.html">SpatialData</a></code> objects.</p>
    </div>


    <div id="slots">
    <h2>Slots</h2>


<dl><dt><code>active</code></dt>
<dd><p>logical. If <code>TRUE</code>, it is the image used by default. Only one <code>HistoImage</code>
in the <code><a href="SpatialData-class.html">SpatialData</a></code> object can be the active at a time.</p></dd>


<dt><code>aligned</code></dt>
<dd><p>logical. If <code>TRUE</code>, indicates that the image was aligned to
the reference image.</p></dd>


<dt><code>bg_color</code></dt>
<dd><p>character. The color of the background.</p></dd>


<dt><code>dir</code></dt>
<dd><p>character. The directory from where to load the image if slot @image is empty.</p></dd>


<dt><code>image</code></dt>
<dd><p>Image. The image stored as class <code>Image</code> from the package <code>EBImage</code>.</p></dd>


<dt><code>image_info</code></dt>
<dd><p>list. A list of miscellaneous slots that carrie information
regarding the image. Protected slots are:</p><ul><li><p><em>dims</em>: Numeric vector of length three. Output of <code><a href="https://rdrr.io/r/base/dim.html" class="external-link">base::dim()</a></code> on the read image.</p></li>
</ul></dd>


<dt><code>name</code></dt>
<dd><p>character. The name of the image.</p></dd>


<dt><code>outline</code></dt>
<dd><p>list. List of two data.frames in which each row corresponds to
a vertice of the polygon required to outline the whole tissue identified on
the image or single contiguous tissue sections.</p>
<ul><li><p><em>tissue_whole</em>: Data.frame of two variables <em>x_orig</em> and <em>y_orig</em>.</p></li>
<li><p><em>tissue_sections</em>: Data.frame of two variables <em>x_orig</em>, <em>y_orig</em> and <em>section</em> to
outline the tissue section outlined.</p></li>
</ul><p>See the vignette about <code><a href="concept_tissue_outline.html">tissue outline</a></code> for more information.</p></dd>


<dt><code>overlap</code></dt>
<dd><p>numeric. Numeric vector of length two. Quantifies the overlap
of the tissue outline of this image with the tissue outline of the reference image
with a value between 0-1 before and after alignment via <code><a href="alignImage.html">alignImage()</a></code>.</p></dd>


<dt><code>pixel_content</code></dt>
<dd><p>factor. Named factor where names correspond to the
pixels following the naming convention 'px_w1_h1' encoding the width and height value
of the pixel from the original <strong>not transformed</strong> image. Values correspond to
the content the pixel displays. See <code><a href="identifyPixelContent.html">identifyPixelContent()</a></code> for more information.</p></dd>


<dt><code>reference</code></dt>
<dd><p>logical. <code>TRUE</code> if it is the <code>HistoImage</code> used as the reference
with which other histology images are aligned.</p></dd>


<dt><code>sample</code></dt>
<dd><p>character. The name of the tissue portion to which this image belongs.</p></dd>


<dt><code>scale_factors</code></dt>
<dd><p>list. List of single numeric values serving as scale factors for
multiple purposes Reserved slot names:</p><ul><li><p><em>image</em>: Image scale factor with which the original x and y variables (<em>x_orig</em>, <em>y_orig</em>) are multiplied with upon
extraction of the coordinates data.frame (resulting in the <em>x</em> and <em>y</em> variables) ensuring alignment with the image.</p></li>
<li><p><em>pixel</em>: Pixel scale factor used to convert pixel values into SI units. It must have an
attribute called "unit" conforming to the format "SI-unit/px".</p></li>
</ul><p>See <code><a href="concept_scale_factors.html">scale factors</a></code> for more information.</p></dd>


<dt><code>transformations</code></dt>
<dd><p>list. List of transformations to apply upon extracting
the image to ensure alignment with additional images and spatial aspects. In case of default values
no transformation is applied.</p>
<p>The following transformations are applied when the image is <a href="loadImage.html">loaded</a>:</p>
<ul><li><p><em>resize_fct</em>: Numeric value between 0-1. Used to resize the resolution of the image.</p></li>
</ul><p>The following transformations are applied when the image is <a href="getImage.html">extracted</a>:</p>
<ul><li><p><em>angle</em>: Numeric value that ranges from 0-359. Indicates the angle in degrees
by which the image needs to be rotated in <strong>clockwise</strong> direction. Defaults to 0.</p></li>
<li><p><em>flip</em>: List of two logical values named <em>horizontal</em> and <em>vertical</em>. Indicate
if the image is supposed to be flipped around either axis. Both default to <code>FALSE</code></p></li>
<li><p><em>stretch</em>: List of two numerical values named <em>horizontal</em> and <em>vertical</em> serving as scale factors
with which to stretch the respective axis. Both default to 1.</p></li>
<li><p><em>translate</em>: List of two numeric values named <em>horizontal</em> and <em>vertical</em>. Indicate
the number of pixels the image needs to be translated. Positive values shift the image
<strong>downwards</strong> or to the right, respectively. Negative values shift the image <strong>upwards</strong>
or to the left, respectively. Both default to 0.</p></li>
</ul></dd>


</dl></div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Jan Kueckelhaus, Dieter-Henrik Heiland.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

      </footer></div>






  </body></html>

