<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Function reference • SPATA2</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-index">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SPATA2</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/spata-v2-introduction-and-philosophy.html">About</a>
</li>
<li>
  <a href="../articles/spata-v2-installation.html">Installation</a>
</li>
<li>
  <a href="../articles/spata-v2-terminology.html">Terminology</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li class="dropdown-header">Around the Spata-Object</li>
    <li>
      <a href="../articles/spata-v2-object-initiation-and-manipulation.html">Object Set Up and Manipulation</a>
    </li>
    <li>
      <a href="../articles/spata-v2-extract-add-and-join-data.html">Extracting, Adding and Joining Data</a>
    </li>
    <li>
      <a href="../articles/spata-v2-platform-compatibility.html">Integrating Different Platforms</a>
    </li>
    <li class="dropdown-header">Analysis &amp; Computation</li>
    <li>
      <a href="../articles/spata-v2-autoencoder.html">Autoencoder Denoising</a>
    </li>
    <li>
      <a href="../articles/spata-v2-cnv-analysis.html">Copy Number Variations</a>
    </li>
    <li>
      <a href="../articles/spata-v2-segmentation-and-clustering.html">Spatial Segmentation &amp; Clustering</a>
    </li>
    <li>
      <a href="../articles/spata-v2-de-analysis.html">Differential Expression Analysis</a>
    </li>
    <li>
      <a href="../articles/sapta-v2-gsea.html">Gene Set Enrichment Analysis</a>
    </li>
    <li>
      <a href="../articles/spata-v2-trajectory-analysis.html">Spatial Trajectories</a>
    </li>
    <li class="dropdown-header">Visualization</li>
    <li>
      <a href="../articles/spata-v2-plotting-surface.html">Surface Plotting</a>
    </li>
    <li>
      <a href="../articles/spata-v2-plotting-dimensional-reduction.html">Dimensional Reduction</a>
    </li>
    <li>
      <a href="../articles/spata-v2-plotting-descriptive-statistics.html">Descriptive Statistics</a>
    </li>
    <li>
      <a href="../articles/spata-v2-plotting-miscellaneous.html">Miscellaneous</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index"><colgroup><col class="alias"><col class="title"></colgroup><tbody><tr><th colspan="2">
          <h2 id="object-initiation-transformation">Object Initiation &amp; Transformation <a href="#object-initiation-transformation" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>These functions generate the spata-object you will work with throughot this and other platforms.</p>
        </th>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="initiation-updating">Initiation &amp; Updating <a href="#initiation-updating" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p><p>The initiation-functions create a spata-object from scratch.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="initiateSpataObject_10X.html">initiateSpataObject_10X()</a></code> </p>
        </td>
        <td><p>Initiate a spata-object from 10X Visium</p></td>
      </tr><tr><td>
          <p><code><a href="initiateSpataObject_CountMtr.html">initiateSpataObject_CountMtr()</a></code> </p>
        </td>
        <td><p>Initiate a spata-object from a raw count matrix</p></td>
      </tr><tr><td>
          <p><code><a href="initiateSpataObject_Empty.html">initiateSpataObject_Empty()</a></code> </p>
        </td>
        <td><p>Initiate an empty spata-object</p></td>
      </tr><tr><td>
          <p><code><a href="initiateSpataObject_Examples.html">initiateSpataObject_Examples()</a></code> </p>
        </td>
        <td><p>Initiate a spata-object from example data sets</p></td>
      </tr><tr><td>
          <p><code><a href="initiateSpataObject_ExprMtr.html">initiateSpataObject_ExprMtr()</a></code> </p>
        </td>
        <td><p>Initiate spata object from scaled expression matrix</p></td>
      </tr><tr><td>
          <p><code><a href="updateSpataObject.html">updateSpataObject()</a></code> </p>
        </td>
        <td><p>Update spata-object from SPATA to SPATA2</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="transformation">Transformation <a href="#transformation" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p><p>The transform-functions ensure compatibility between different platforms.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="transformSeuratToSpata.html">transformSeuratToSpata()</a></code> </p>
        </td>
        <td><p>Transform seurat-object to spata-object</p></td>
      </tr><tr><td>
          <p><code><a href="transformSpataToCDS.html">transformSpataToCDS()</a></code> </p>
        </td>
        <td><p>Transform spata-object to cell-data-set (Monocle3)</p></td>
      </tr><tr><td>
          <p><code><a href="transformSpataToSeurat.html">transformSpataToSeurat()</a></code> </p>
        </td>
        <td><p>Transform spata-object to a seurat-object</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="object-loading-and-saving">Object Loading and Saving <a href="#object-loading-and-saving" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Some handy functions that make saving and loading corresponding objects more convenient.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="loadGeneSetDf.html">loadGeneSetDf()</a></code> </p>
        </td>
        <td><p>Load gene set data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="loadGSDF.html">loadGSDF()</a></code> </p>
        </td>
        <td><p>Original load gene set data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="loadSpataObject.html">loadSpataObject()</a></code> <code><a href="loadSpataObject.html">loadCorrespondingCDS()</a></code> <code><a href="loadSpataObject.html">loadCorrespondingSeuratObject()</a></code> </p>
        </td>
        <td><p>Load corresponding objects</p></td>
      </tr><tr><td>
          <p><code><a href="saveGeneSetDf.html">saveGeneSetDf()</a></code> </p>
        </td>
        <td><p>Save a gene set data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="saveSpataObject.html">saveSpataObject()</a></code> <code><a href="saveSpataObject.html">saveCorrespondingCDS()</a></code> <code><a href="saveSpataObject.html">saveCorrespondingSeuratObject()</a></code> </p>
        </td>
        <td><p>Save corresponding objects</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="object-subsetting">Object Subsetting <a href="#object-subsetting" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Create spata-objects from data subsets for more in depth analysis.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="subsetBySegment_CountMtr.html">subsetBySegment_CountMtr()</a></code> <code><a href="subsetBySegment_CountMtr.html">subsetByBarcodes_CountMtr()</a></code> <code><a href="subsetBySegment_CountMtr.html">subsetBySegment_ExprMtr()</a></code> <code><a href="subsetBySegment_CountMtr.html">subsetByBarcodes_ExprMtr()</a></code> </p>
        </td>
        <td><p>Subset a spata-object</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="object-manipulation">Object Manipulation <a href="#object-manipulation" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>SPATA invites you to extract your analysis and to add the results of your analysis. Constant change of the object’s content is therefore inevitable. The following functions serve as handy, helping hands to add SPATA-extern content to the object without disturbing it’s integrity regarding SPATA-intern processes.</p>
        </th>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="set-content">Set Content <a href="#set-content" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p><p>The set-functions let you set the content of specific slots.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="setActiveExpressionMatrix.html">setActiveExpressionMatrix()</a></code> </p>
        </td>
        <td><p>Denote the default expression matrix</p></td>
      </tr><tr><td>
          <p><code><a href="setAutoencoderAssessment.html">setAutoencoderAssessment()</a></code> </p>
        </td>
        <td><p>Set results of autoencoder assessment</p></td>
      </tr><tr><td>
          <p><code><a href="setCnvResults.html">setCnvResults()</a></code> </p>
        </td>
        <td><p>Set cnv-results</p></td>
      </tr><tr><td>
          <p><code><a href="setCoordsDf.html">setCoordsDf()</a></code> </p>
        </td>
        <td><p>Set the coordinates</p></td>
      </tr><tr><td>
          <p><code><a href="setCountMatrix.html">setCountMatrix()</a></code> <code><a href="setCountMatrix.html">setDenoisedMatrix()</a></code> <code><a href="setCountMatrix.html">setScaledMatrix()</a></code> </p>
        </td>
        <td><p>Set data matrices</p></td>
      </tr><tr><td>
          <p><code><a href="setCsrCutoff.html">setCsrCutoff()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="setDefaultInstructions.html">setDefaultInstructions()</a></code> <code><a href="setDefaultInstructions.html">setDirectoryInstructions()</a></code> </p>
        </td>
        <td><p>Set default instructions</p></td>
      </tr><tr><td>
          <p><code><a href="setFeatureDf.html">setFeatureDf()</a></code> </p>
        </td>
        <td><p>Set feature data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="setGeneSetDf.html">setGeneSetDf()</a></code> </p>
        </td>
        <td><p>Set the gene-set data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="setHspaResults.html">setHspaResults()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="setInitiationInfo.html">setInitiationInfo()</a></code> </p>
        </td>
        <td><p>Set initiation information</p></td>
      </tr><tr><td>
          <p><code><a href="setPcaDf.html">setPcaDf()</a></code> <code><a href="setPcaDf.html">setTsneDf()</a></code> <code><a href="setPcaDf.html">setUmapDf()</a></code> </p>
        </td>
        <td><p>Set dimensional reductions data.frames</p></td>
      </tr><tr><td>
          <p><code><a href="setPrResults.html">setPrResults()</a></code> </p>
        </td>
        <td><p>Set results of pattern recognition methods</p></td>
      </tr><tr><td>
          <p><code><a href="setSpCorResults.html">setSpCorResults()</a></code> </p>
        </td>
        <td><p>Set results of spatial correlation analysis</p></td>
      </tr><tr><td>
          <p><code><a href="set_dummy.html">set_dummy()</a></code> </p>
        </td>
        <td><p>set</p></td>
      </tr><tr><td>
          <p><code><a href="across.html">across()</a></code> <code><a href="across.html">across_dummy()</a></code> </p>
        </td>
        <td><p>across</p></td>
      </tr><tr><td>
          <p><code><a href="addAutoencoderSetUp.html">addAutoencoderSetUp()</a></code> </p>
        </td>
        <td><p>Add the set up of a neural network</p></td>
      </tr><tr><td>
          <p><code><a href="addExpressionMatrix.html">addExpressionMatrix()</a></code> </p>
        </td>
        <td><p>Add an expression matrix</p></td>
      </tr><tr><td>
          <p><code><a href="addFeatures.html">addFeatures()</a></code> </p>
        </td>
        <td><p>Add a new feature</p></td>
      </tr><tr><td>
          <p><code><a href="addGeneFeatures.html">addGeneFeatures()</a></code> </p>
        </td>
        <td><p>Add new gene features</p></td>
      </tr><tr><td>
          <p><code><a href="addGeneMetaData.html">addGeneMetaData()</a></code> </p>
        </td>
        <td><p>Add gene meta data to the object</p></td>
      </tr><tr><td>
          <p><code><a href="addGeneSet.html">addGeneSet()</a></code> <code><a href="addGeneSet.html">addGeneSetsInteractive()</a></code> </p>
        </td>
        <td><p>Add a new gene set</p></td>
      </tr><tr><td>
          <p><code><a href="addSpCorCluster.html">addSpCorCluster()</a></code> </p>
        </td>
        <td><p>Add cluster results of spatial correlation results</p></td>
      </tr><tr><td>
          <p><code><a href="adjustDefaultInstructions.html">adjustDefaultInstructions()</a></code> <code><a href="adjustDefaultInstructions.html">adjustDirectoryInstructions()</a></code> </p>
        </td>
        <td><p>Adjust default instructions</p></td>
      </tr><tr><td>
          <p><code><a href="adjustGeneSetDf.html">adjustGeneSetDf()</a></code> </p>
        </td>
        <td><p>Filter gene-set data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="adjusting_check_dummy.html">adjusting_check_dummy()</a></code> </p>
        </td>
        <td><p>This is a text dummy</p></td>
      </tr><tr><td>
          <p><code><a href="allColorpalettes.html">allColorpalettes()</a></code> <code><a href="allColorpalettes.html">allColorspectra()</a></code> </p>
        </td>
        <td><p>Valid colorpanels &amp; -spectra</p></td>
      </tr><tr><td>
          <p><code><a href="all_color_palettes.html">all_color_palettes()</a></code> <code><a href="all_color_palettes.html">all_color_spectra()</a></code> </p>
        </td>
        <td><p>Color palette names</p></td>
      </tr><tr><td>
          <p><code><a href="argument_dummy.html">argument_dummy()</a></code> </p>
        </td>
        <td><p>Argument dummy</p></td>
      </tr><tr><td>
          <p><code><a href="assessTrajectoryTrends.html">assessTrajectoryTrends()</a></code> <code><a href="assessTrajectoryTrends.html">assessTrajectoryTrends2()</a></code> <code><a href="assessTrajectoryTrends.html">assessTrajectoryTrendsCustomized()</a></code> <code><a href="assessTrajectoryTrends.html">assessTrajectoryTrendsCustomized2()</a></code> </p>
        </td>
        <td><p>Trajectory trend analysis</p></td>
      </tr><tr><td>
          <p><code><a href="average_genes.html">average_genes()</a></code> </p>
        </td>
        <td><p>average_genes</p></td>
      </tr><tr><td>
          <p><code><a href="binarize_matrix.html">binarize_matrix()</a></code> </p>
        </td>
        <td><p>Binarize expression matrix</p></td>
      </tr><tr><td>
          <p><code><a href="binwidth_dummy.html">binwidth_dummy()</a></code> </p>
        </td>
        <td><p>binwidth</p></td>
      </tr><tr><td>
          <p><code><a href="cds_dummy.html">cds_dummy()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="checkpoint.html">checkpoint()</a></code> </p>
        </td>
        <td><p>Shiny feedback messages</p></td>
      </tr><tr><td>
          <p><code><a href="check_all_barcodes.html">check_all_barcodes()</a></code> </p>
        </td>
        <td><p>Check barcodes of all matrices and data.frames</p></td>
      </tr><tr><td>
          <p><code><a href="check_assign.html">check_assign()</a></code> </p>
        </td>
        <td><p>Check assign input</p></td>
      </tr><tr><td>
          <p><code><a href="check_atdf.html">check_atdf()</a></code> <code><a href="check_atdf.html">check_rtdf()</a></code> </p>
        </td>
        <td><p>Check assessed trajectory data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="check_availability.html">check_availability()</a></code> </p>
        </td>
        <td><p>Gives feedback if object was not found</p></td>
      </tr><tr><td>
          <p><code><a href="check_color_to.html">check_color_to()</a></code> </p>
        </td>
        <td><p>Check color to</p></td>
      </tr><tr><td>
          <p><code><a href="check_compiled_trajectory_df.html">check_compiled_trajectory_df()</a></code> </p>
        </td>
        <td><p>Check compiled trajectory data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="check_coordinate_variables.html">check_coordinate_variables()</a></code> </p>
        </td>
        <td><p>Check coordinate variables</p></td>
      </tr><tr><td>
          <p><code><a href="check_coords_df.html">check_coords_df()</a></code> </p>
        </td>
        <td><p>Check coords data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="check_customized_trends.html">check_customized_trends()</a></code> </p>
        </td>
        <td><p>Check customized trend list</p></td>
      </tr><tr><td>
          <p><code><a href="check_dea_df.html">check_dea_df()</a></code> </p>
        </td>
        <td><p>Check de data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="check_display.html">check_display()</a></code> </p>
        </td>
        <td><p>Check display input</p></td>
      </tr><tr><td>
          <p><code><a href="check_features.html">check_features()</a></code> </p>
        </td>
        <td><p>Check feature variables input</p></td>
      </tr><tr><td>
          <p><code><a href="check_feature_df.html">check_feature_df()</a></code> </p>
        </td>
        <td><p>Check feature data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="check_genes.html">check_genes()</a></code> </p>
        </td>
        <td><p>Check gene input</p></td>
      </tr><tr><td>
          <p><code><a href="check_gene_sets.html">check_gene_sets()</a></code> </p>
        </td>
        <td><p>Check gene set input</p></td>
      </tr><tr><td>
          <p><code><a href="check_method.html">check_method()</a></code> </p>
        </td>
        <td><p>Check method input</p></td>
      </tr><tr><td>
          <p><code><a href="check_monocle_input.html">check_monocle_input()</a></code> </p>
        </td>
        <td><p>Check monocle input</p></td>
      </tr><tr><td>
          <p><code><a href="check_monocle_packages.html">check_monocle_packages()</a></code> </p>
        </td>
        <td><p>Makes sure that both packages are installed</p></td>
      </tr><tr><td>
          <p><code><a href="check_object.html">check_object()</a></code> </p>
        </td>
        <td><p>Check spata object input</p></td>
      </tr><tr><td>
          <p><code><a href="check_pattern.html">check_pattern()</a></code> </p>
        </td>
        <td><p>Check pattern input</p></td>
      </tr><tr><td>
          <p><code><a href="check_pt.html">check_pt()</a></code> </p>
        </td>
        <td><p>Check pt input</p></td>
      </tr><tr><td>
          <p><code><a href="check_sample.html">check_sample()</a></code> </p>
        </td>
        <td><p>Check sample input</p></td>
      </tr><tr><td>
          <p><code><a href="check_saving.html">check_saving()</a></code> </p>
        </td>
        <td><p>Check saving input</p></td>
      </tr><tr><td>
          <p><code><a href="check_segment.html">check_segment()</a></code> </p>
        </td>
        <td><p>Check segment name</p></td>
      </tr><tr><td>
          <p><code><a href="check_slot_coordinates.html">check_slot_coordinates()</a></code> <code><a href="check_slot_coordinates.html">check_slot_data()</a></code> <code><a href="check_slot_coordinates.html">check_slot_version()</a></code> <code><a href="check_slot_coordinates.html">check_slot_scvelo()</a></code> <code><a href="check_slot_coordinates.html">check_slot_trajectories()</a></code> <code><a href="check_slot_coordinates.html">check_trajectory_object()</a></code> <code><a href="check_slot_coordinates.html">check_slot_dim_red()</a></code> <code><a href="check_slot_coordinates.html">check_slot_fdata()</a></code> <code><a href="check_slot_coordinates.html">check_slot_image()</a></code> <code><a href="check_slot_coordinates.html">check_slot_used_genesets()</a></code> <code><a href="check_slot_coordinates.html">check_slot_samples()</a></code> </p>
        </td>
        <td><p>Check spata slots</p></td>
      </tr><tr><td>
          <p><code><a href="check_smooth.html">check_smooth()</a></code> </p>
        </td>
        <td><p>Check smooth input</p></td>
      </tr><tr><td>
          <p><code><a href="check_spata_df.html">check_spata_df()</a></code> </p>
        </td>
        <td><p>Check coordinate data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="check_stdf.html">check_stdf()</a></code> </p>
        </td>
        <td><p>Check summarized trajectory data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="check_summarized_trajectory_df.html">check_summarized_trajectory_df()</a></code> </p>
        </td>
        <td><p>Check stdf-input</p></td>
      </tr><tr><td>
          <p><code><a href="check_to.html">check_to()</a></code> </p>
        </td>
        <td><p>Gives feedback about input validity</p></td>
      </tr><tr><td>
          <p><code><a href="check_trajectory.html">check_trajectory()</a></code> </p>
        </td>
        <td><p>Check trajectory name input</p></td>
      </tr><tr><td>
          <p><code><a href="check_trajectory_binwidth.html">check_trajectory_binwidth()</a></code> </p>
        </td>
        <td><p>Check trajectory binwdith input</p></td>
      </tr><tr><td>
          <p><code><a href="check_uniform_genes.html">check_uniform_genes()</a></code> </p>
        </td>
        <td><p>Check uniform genes input</p></td>
      </tr><tr><td>
          <p><code><a href="check_variables.html">check_variables()</a></code> </p>
        </td>
        <td><p>Check variables</p></td>
      </tr><tr><td>
          <p><code><a href="clrp.html">clrp()</a></code> <code><a href="clrp.html">clrp_dummy()</a></code> </p>
        </td>
        <td><p>clrp</p></td>
      </tr><tr><td>
          <p><code><a href="clusterSpCorResults.html">clusterSpCorResults()</a></code> </p>
        </td>
        <td><p>Cluster genes according to their expression profile</p></td>
      </tr><tr><td>
          <p><code><a href="cnv_ref.html">cnv_ref</a></code> </p>
        </td>
        <td><p>A list of reference data for copy number variation analysis (CNV)</p></td>
      </tr><tr><td>
          <p><code><a href="cnv_regions_df.html">cnv_regions_df</a></code> </p>
        </td>
        <td><p>A data.frame necessary for cnv-analysis. Contains information about start and
end points of chromosomes.</p></td>
      </tr><tr><td>
          <p><code><a href="computeGeneMetaData.html">computeGeneMetaData()</a></code> <code><a href="computeGeneMetaData.html">computeGeneMetaData2()</a></code> </p>
        </td>
        <td><p>Compute gene summary statistics</p></td>
      </tr><tr><td>
          <p><code><a href="compute_pattern_similarity.html">compute_pattern_similarity()</a></code> <code><a href="compute_pattern_similarity.html">compute_pattern_relation()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="createSegmentation.html">createSegmentation()</a></code> </p>
        </td>
        <td><p>Spatial Segmentation</p></td>
      </tr><tr><td>
          <p><code><a href="createTrajectories.html">createTrajectories()</a></code> </p>
        </td>
        <td><p>Spatial Trajectories</p></td>
      </tr><tr><td>
          <p><code><a href="createTrajectoryManually.html">createTrajectoryManually()</a></code> </p>
        </td>
        <td><p>Create spatial trajectories manually</p></td>
      </tr><tr><td>
          <p><code><a href="data_counts-class.html">data_counts-class</a></code> </p>
        </td>
        <td><p>data_counts object</p></td>
      </tr><tr><td>
          <p><code><a href="default_instructions-class.html">default_instructions-class</a></code> </p>
        </td>
        <td><p>default instructions</p></td>
      </tr><tr><td>
          <p><code><a href="dim_red-class.html">dim_red-class</a></code> </p>
        </td>
        <td><p>dim_red object</p></td>
      </tr><tr><td>
          <p><code><a href="discardExpressionMatrix.html">discardExpressionMatrix()</a></code> </p>
        </td>
        <td><p>Discard an expression matrix</p></td>
      </tr><tr><td>
          <p><code><a href="discardFeatures.html">discardFeatures()</a></code> </p>
        </td>
        <td><p>Discard features</p></td>
      </tr><tr><td>
          <p><code><a href="discardGeneFeatures.html">discardGeneFeatures()</a></code> </p>
        </td>
        <td><p>Discard gene features</p></td>
      </tr><tr><td>
          <p><code><a href="discardGenes.html">discardGenes()</a></code> </p>
        </td>
        <td><p>Discard genes</p></td>
      </tr><tr><td>
          <p><code><a href="discardGeneSets.html">discardGeneSets()</a></code> </p>
        </td>
        <td><p>Discard gene sets</p></td>
      </tr><tr><td>
          <p><code><a href="dropped_df_dummy.html">dropped_df_dummy()</a></code> </p>
        </td>
        <td><p>dropped_df</p></td>
      </tr><tr><td>
          <p><code><a href="examineClusterResults.html">examineClusterResults()</a></code> </p>
        </td>
        <td><p>Examine clustering results</p></td>
      </tr><tr><td>
          <p><code><a href="examineTrajectoryAssessment.html">examineTrajectoryAssessment()</a></code> </p>
        </td>
        <td><p>Examine trajectory-moddeling results</p></td>
      </tr><tr><td>
          <p><code><a href="filterDeaDf.html">filterDeaDf()</a></code> </p>
        </td>
        <td><p>Postprocess de-analysis results</p></td>
      </tr><tr><td>
          <p><code><a href="filterTrends.html">filterTrends()</a></code> <code><a href="filterTrends.html">filterTrajectoryTrends()</a></code> </p>
        </td>
        <td><p>Filter variables of a certain trend</p></td>
      </tr><tr><td>
          <p><code><a href="findHierarchicalCluster.html">findHierarchicalCluster()</a></code> <code><a href="findHierarchicalCluster.html">findHierarchicalClusterCnv()</a></code> </p>
        </td>
        <td><p>Cluster sample via hclust</p></td>
      </tr><tr><td>
          <p><code><a href="findMonocleClusters.html">findMonocleClusters()</a></code> </p>
        </td>
        <td><p>Cluster sample via monocle3</p></td>
      </tr><tr><td>
          <p><code><a href="findNearestNeighbourClusters.html">findNearestNeighbourClusters()</a></code> </p>
        </td>
        <td><p>Cluster sample via nearest neighbour analysis</p></td>
      </tr><tr><td>
          <p><code><a href="findSeuratClusters.html">findSeuratClusters()</a></code> </p>
        </td>
        <td><p>Cluster sample via Seurat</p></td>
      </tr><tr><td>
          <p><code><a href="flexible_call_dummy.html">flexible_call_dummy()</a></code> </p>
        </td>
        <td><p>flexible_call_dummy</p></td>
      </tr><tr><td>
          <p><code><a href="flipImage.html">flipImage()</a></code> </p>
        </td>
        <td><p>Flip object image</p></td>
      </tr><tr><td>
          <p><code><a href="gene_set_path.html">gene_set_path()</a></code> </p>
        </td>
        <td><p>gene_set_path</p></td>
      </tr><tr><td>
          <p><code><a href="getActiveMatrixName.html">getActiveMatrixName()</a></code> </p>
        </td>
        <td><p>Obtain name of currently active expression matrix</p></td>
      </tr><tr><td>
          <p><code><a href="getAutoencoderAssessment.html">getAutoencoderAssessment()</a></code> </p>
        </td>
        <td><p>Obtain information about the optimal neural network set up</p></td>
      </tr><tr><td>
          <p><code><a href="getAutoencoderSetUp.html">getAutoencoderSetUp()</a></code> </p>
        </td>
        <td><p>Obtain information on neural network</p></td>
      </tr><tr><td>
          <p><code><a href="getBarcodes.html">getBarcodes()</a></code> </p>
        </td>
        <td><p>Obtain specific barcodes</p></td>
      </tr><tr><td>
          <p><code><a href="getCnvFeatureNames.html">getCnvFeatureNames()</a></code> </p>
        </td>
        <td><p>Obtain features names under which cnv-analysis results are stored.</p></td>
      </tr><tr><td>
          <p><code><a href="getCnvResults.html">getCnvResults()</a></code> </p>
        </td>
        <td><p>Obtain copy-number-variations results</p></td>
      </tr><tr><td>
          <p><code><a href="getCoordsDf.html">getCoordsDf()</a></code> <code><a href="getCoordsDf.html">getSegmentDf()</a></code> </p>
        </td>
        <td><p>Obtain spatial coordinates</p></td>
      </tr><tr><td>
          <p><code><a href="getDeaOverview.html">getDeaOverview()</a></code> </p>
        </td>
        <td><p>Obtain info on de-analysis storage</p></td>
      </tr><tr><td>
          <p><code><a href="getDeaResultsDf.html">getDeaResultsDf()</a></code> <code><a href="getDeaResultsDf.html">getDeaGenes()</a></code> </p>
        </td>
        <td><p>Obtain de-analysis results</p></td>
      </tr><tr><td>
          <p><code><a href="getDefaultInstructions.html">getDefaultInstructions()</a></code> <code><a href="getDefaultInstructions.html">getDirectoryInstructions()</a></code> </p>
        </td>
        <td><p>Obtain default argument inputs</p></td>
      </tr><tr><td>
          <p><code><a href="getDimRedDf.html">getDimRedDf()</a></code> <code><a href="getDimRedDf.html">getPcaDf()</a></code> <code><a href="getDimRedDf.html">getPcaMtr()</a></code> <code><a href="getDimRedDf.html">getUmapDf()</a></code> <code><a href="getDimRedDf.html">getTsneDf()</a></code> </p>
        </td>
        <td><p>Obtain dimensional reduction data</p></td>
      </tr><tr><td>
          <p><code><a href="getExpressionMatrixNames.html">getExpressionMatrixNames()</a></code> </p>
        </td>
        <td><p>Obtain names of stored expression matrices</p></td>
      </tr><tr><td>
          <p><code><a href="getFeatureDf.html">getFeatureDf()</a></code> </p>
        </td>
        <td><p>Obtain feature data</p></td>
      </tr><tr><td>
          <p><code><a href="getFeatureNames.html">getFeatureNames()</a></code> </p>
        </td>
        <td><p>Obtain feature names</p></td>
      </tr><tr><td>
          <p><code><a href="getFeatureValues.html">getFeatureValues()</a></code> </p>
        </td>
        <td><p>Obtain unique categorical feature values</p></td>
      </tr><tr><td>
          <p><code><a href="getFeatureVariables.html">getFeatureVariables()</a></code> </p>
        </td>
        <td><p>Obtain a feature variable</p></td>
      </tr><tr><td>
          <p><code><a href="getFromSeurat.html">getFromSeurat()</a></code> </p>
        </td>
        <td><p>Safe extraction</p></td>
      </tr><tr><td>
          <p><code><a href="getGeneCounts.html">getGeneCounts()</a></code> </p>
        </td>
        <td><p>Obtain total number of gene counts</p></td>
      </tr><tr><td>
          <p><code><a href="getGeneDistMtr.html">getGeneDistMtr()</a></code> </p>
        </td>
        <td><p>Obtain distance measurements of spatially correlated genes</p></td>
      </tr><tr><td>
          <p><code><a href="getGeneFeatureNames.html">getGeneFeatureNames()</a></code> </p>
        </td>
        <td><p>Obtain feature names of the gene meta data</p></td>
      </tr><tr><td>
          <p><code><a href="getGeneMetaData.html">getGeneMetaData()</a></code> <code><a href="getGeneMetaData.html">getGeneMetaDf()</a></code> </p>
        </td>
        <td><p>Obtain gene meta data</p></td>
      </tr><tr><td>
          <p><code><a href="getGenePatternCluster.html">getGenePatternCluster()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="getGenePatternCorrelations.html">getGenePatternCorrelations()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="getGenePatternDistances.html">getGenePatternDistances()</a></code> <code><a href="getGenePatternDistances.html">getGenePatternSimilarities()</a></code> </p>
        </td>
        <td><p>Obtain similarity scores between identified gene patterns</p></td>
      </tr><tr><td>
          <p><code><a href="getGenePatternEvalDf.html">getGenePatternEvalDf()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="getGenePatternExtentDf.html">getGenePatternExtentDf()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="getGenePatternsAboveThreshold.html">getGenePatternsAboveThreshold()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="getGenes.html">getGenes()</a></code> <code><a href="getGenes.html">getGenesInteractive()</a></code> </p>
        </td>
        <td><p>Obtain gene names</p></td>
      </tr><tr><td>
          <p><code><a href="getGeneSetDf.html">getGeneSetDf()</a></code> </p>
        </td>
        <td><p>Obtain gene set data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="getGeneSetOverview.html">getGeneSetOverview()</a></code> </p>
        </td>
        <td><p>Overview about the current gene sets</p></td>
      </tr><tr><td>
          <p><code><a href="getGeneSets.html">getGeneSets()</a></code> <code><a href="getGeneSets.html">getGeneSetsInteractive()</a></code> </p>
        </td>
        <td><p>Obtain gene set names</p></td>
      </tr><tr><td>
          <p><code><a href="getGroupingOptions.html">getGroupingOptions()</a></code> </p>
        </td>
        <td><p>Obtain variable names that group the barcode spots</p></td>
      </tr><tr><td>
          <p><code><a href="getGroupNames.html">getGroupNames()</a></code> </p>
        </td>
        <td><p>Obtain group names a grouping variable contains</p></td>
      </tr><tr><td>
          <p><code><a href="getGseaDf.html">getGseaDf()</a></code> </p>
        </td>
        <td><p>Obtain enrichment data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="getGseaResults.html">getGseaResults()</a></code> </p>
        </td>
        <td><p>Obtain enrichment results</p></td>
      </tr><tr><td>
          <p><code><a href="getHclustObject.html">getHclustObject()</a></code> <code><a href="getHclustObject.html">getHclustObjectCnv()</a></code> </p>
        </td>
        <td><p>Obtain hierarchical cluster results</p></td>
      </tr><tr><td>
          <p><code><a href="getHspaResults.html">getHspaResults()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="getImage.html">getImage()</a></code> </p>
        </td>
        <td><p>Obtain histology image</p></td>
      </tr><tr><td>
          <p><code><a href="getInitiationInfo.html">getInitiationInfo()</a></code> <code><a href="getInitiationInfo.html">getInitiationInput()</a></code> </p>
        </td>
        <td><p>Obtain information about object initiation</p></td>
      </tr><tr><td>
          <p><code><a href="getMatrix.html">getMatrix()</a></code> <code><a href="getMatrix.html">getExpressionMatrix()</a></code> <code><a href="getMatrix.html">getCountMatrix()</a></code> </p>
        </td>
        <td><p>Obtain count and expression matrix</p></td>
      </tr><tr><td>
          <p><code><a href="getPrResults.html">getPrResults()</a></code> <code><a href="getPrResults.html">getPatternNames()</a></code> </p>
        </td>
        <td><p>Obtain pattern recognition results</p></td>
      </tr><tr><td>
          <p><code><a href="getSampleNames.html">getSampleNames()</a></code> </p>
        </td>
        <td><p>Obtain sample names</p></td>
      </tr><tr><td>
          <p><code><a href="getSegmentNames.html">getSegmentNames()</a></code> </p>
        </td>
        <td><p>Obtain segment names</p></td>
      </tr><tr><td>
          <p><code><a href="getSignatureEnrichment.html">getSignatureEnrichment()</a></code> </p>
        </td>
        <td><p>Obtain signature enrichment</p></td>
      </tr><tr><td>
          <p><code><a href="getSpataDf.html">getSpataDf()</a></code> </p>
        </td>
        <td><p>Obtain a spata-data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="getSpCorCluster.html">getSpCorCluster()</a></code> <code><a href="getSpCorCluster.html">getSpCorClusterNames()</a></code> </p>
        </td>
        <td><p>Obtain cluster results based on spatial correlation analysis</p></td>
      </tr><tr><td>
          <p><code><a href="getSpCorResults.html">getSpCorResults()</a></code> </p>
        </td>
        <td><p>Obtain spatial correlation results</p></td>
      </tr><tr><td>
          <p><code><a href="getTrajectoryComment-spata2-method.html">getTrajectoryComment(<i>&lt;spata2&gt;</i>)</a></code> </p>
        </td>
        <td><p>Obtain a trajectory comment</p></td>
      </tr><tr><td>
          <p><code><a href="getTrajectoryComment.html">getTrajectoryComment()</a></code> </p>
        </td>
        <td><p>Obtain a trajectory comment</p></td>
      </tr><tr><td>
          <p><code><a href="getTrajectoryDf.html">getTrajectoryDf()</a></code> </p>
        </td>
        <td><p>Obtain a summarized trajectory data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="getTrajectoryLength.html">getTrajectoryLength()</a></code> </p>
        </td>
        <td><p>Obtain the length of a trajectory</p></td>
      </tr><tr><td>
          <p><code><a href="getTrajectoryNames.html">getTrajectoryNames()</a></code> </p>
        </td>
        <td><p>Obtain trajectory names</p></td>
      </tr><tr><td>
          <p><code><a href="getTrajectoryObject.html">getTrajectoryObject()</a></code> </p>
        </td>
        <td><p>Obtain trajectory object</p></td>
      </tr><tr><td>
          <p><code><a href="ggplot_family.html">ggplot_family()</a></code> </p>
        </td>
        <td><p>ggplot_family</p></td>
      </tr><tr><td>
          <p><code><a href="gsdf.html">gsdf</a></code> </p>
        </td>
        <td><p>The default collection of frequently used gene-sets</p></td>
      </tr><tr><td>
          <p><code><a href="hlpr_add_barcode_suffix.html">hlpr_add_barcode_suffix()</a></code> </p>
        </td>
        <td><p>Make sure that barcodes are spata-like</p></td>
      </tr><tr><td>
          <p><code><a href="hlpr_add_models.html">hlpr_add_models()</a></code> <code><a href="hlpr_add_models.html">hlpr_add_residuals()</a></code> <code><a href="hlpr_add_models.html">hlpr_add_residuals_customized()</a></code> <code><a href="hlpr_add_models.html">hlpr_summarize_residuals()</a></code> <code><a href="hlpr_add_models.html">hlpr_name_models()</a></code> <code><a href="hlpr_add_models.html">hlpr_filter_trend()</a></code> </p>
        </td>
        <td><p>Helper functions for trajectory ranking</p></td>
      </tr><tr><td>
          <p><code><a href="hlpr_add_old_coords.html">hlpr_add_old_coords()</a></code> </p>
        </td>
        <td><p>Adds old coordinates</p></td>
      </tr><tr><td>
          <p><code><a href="hlpr_adjust_legend_size.html">hlpr_adjust_legend_size()</a></code> </p>
        </td>
        <td><p>Adjusts the size of discrete legend points</p></td>
      </tr><tr><td>
          <p><code><a href="hlpr_assess_trajectory_trends.html">hlpr_assess_trajectory_trends()</a></code> <code><a href="hlpr_assess_trajectory_trends.html">hlpr_assess_trajectory_trends_customized()</a></code> </p>
        </td>
        <td><p>Assess trajectory ranking.</p></td>
      </tr><tr><td>
          <p><code><a href="hlpr_assign_arguments.html">hlpr_assign_arguments()</a></code> </p>
        </td>
        <td><p>Assign arguments to calling function</p></td>
      </tr><tr><td>
          <p><code><a href="hlpr_breaks.html">hlpr_breaks()</a></code> </p>
        </td>
        <td><p>Calculates breaks for heatmap according to input matrix</p></td>
      </tr><tr><td>
          <p><code><a href="hlpr_compare_samples.html">hlpr_compare_samples()</a></code> </p>
        </td>
        <td><p>Compare samples within an object</p></td>
      </tr><tr><td>
          <p><code><a href="hlpr_compile_trajectory.html">hlpr_compile_trajectory()</a></code> </p>
        </td>
        <td><p>Compiles a trajectory data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="hlpr_display_title.html">hlpr_display_title()</a></code> <code><a href="hlpr_display_title.html">hlpr_display_subtitle()</a></code> </p>
        </td>
        <td><p>No documentation needed</p></td>
      </tr><tr><td>
          <p><code><a href="hlpr_dist_mtr_to_df.html">hlpr_dist_mtr_to_df()</a></code> </p>
        </td>
        <td><p>Convert distance matrix to distance data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="hlpr_gene_set_name.html">hlpr_gene_set_name()</a></code> </p>
        </td>
        <td><p>Removes the class part of a gene set string</p></td>
      </tr><tr><td>
          <p><code><a href="hlpr_geom_trajectory_fit.html">hlpr_geom_trajectory_fit()</a></code> </p>
        </td>
        <td><p>Easy switch between geom line and and geom smooth</p></td>
      </tr><tr><td>
          <p><code><a href="hlpr_image_add_on.html">hlpr_image_add_on()</a></code> <code><a href="hlpr_image_add_on.html">hlpr_image_add_on2()</a></code> </p>
        </td>
        <td><p>Provides the image as ggplot background</p></td>
      </tr><tr><td>
          <p><code><a href="hlpr_labs_add_on.html">hlpr_labs_add_on()</a></code> </p>
        </td>
        <td><p>Return customized ggplot:labs()</p></td>
      </tr><tr><td>
          <p><code><a href="hlpr_normalize_imap.html">hlpr_normalize_imap()</a></code> </p>
        </td>
        <td><p>Normalize gene or gene set values</p></td>
      </tr><tr><td>
          <p><code><a href="hlpr_normalize_vctr.html">hlpr_normalize_vctr()</a></code> </p>
        </td>
        <td><p>Normalize gene or gene set values</p></td>
      </tr><tr><td>
          <p><code><a href="hlpr_one_distinct.html">hlpr_one_distinct()</a></code> </p>
        </td>
        <td><p>Number of distinct values equal to 1?</p></td>
      </tr><tr><td>
          <p><code><a href="hlpr_process_spatial_correlation_cluster.html">hlpr_process_spatial_correlation_cluster()</a></code> </p>
        </td>
        <td><p>Process spatial correlation results</p></td>
      </tr><tr><td>
          <p><code><a href="hlpr_rank_trajectory_trends.html">hlpr_rank_trajectory_trends()</a></code> <code><a href="hlpr_rank_trajectory_trends.html">hlpr_rank_trajectory_trends_customized()</a></code> </p>
        </td>
        <td><p>Rank trajectory trends.</p></td>
      </tr><tr><td>
          <p><code><a href="hlpr_run_cnva_pca.html">hlpr_run_cnva_pca()</a></code> </p>
        </td>
        <td><p>run pca on the cnv-matrix</p></td>
      </tr><tr><td>
          <p><code><a href="hlpr_save_spata_object.html">hlpr_save_spata_object()</a></code> </p>
        </td>
        <td><p>Save spata object inside functions</p></td>
      </tr><tr><td>
          <p><code><a href="hlpr_scatterplot.html">hlpr_scatterplot()</a></code> </p>
        </td>
        <td><p>Color to + scatterplot helper</p></td>
      </tr><tr><td>
          <p><code><a href="hlpr_smooth.html">hlpr_smooth()</a></code> </p>
        </td>
        <td><p>Smooth variables spatially</p></td>
      </tr><tr><td>
          <p><code><a href="hlpr_smooth_shiny.html">hlpr_smooth_shiny()</a></code> </p>
        </td>
        <td><p>Smooth variable spatially in mini-shiny-apps</p></td>
      </tr><tr><td>
          <p><code><a href="hlpr_subset_across.html">hlpr_subset_across()</a></code> </p>
        </td>
        <td><p>Subset the across-variables</p></td>
      </tr><tr><td>
          <p><code><a href="hlpr_summarize_trajectory_df.html">hlpr_summarize_trajectory_df()</a></code> </p>
        </td>
        <td><p>Join and summarize compiled trajectory data.frames</p></td>
      </tr><tr><td>
          <p><code><a href="hlpr_vector_projection.html">hlpr_vector_projection()</a></code> </p>
        </td>
        <td><p>Perform vector projection</p></td>
      </tr><tr><td>
          <p><code><a href="hlpr_widen_trajectory_df.html">hlpr_widen_trajectory_df()</a></code> </p>
        </td>
        <td><p>Widen trajectory data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="hspaClusterGenePatterns.html">hspaClusterGenePatterns()</a></code> </p>
        </td>
        <td><p>Construct hcluster-tree from correlated gene pattern similarities</p></td>
      </tr><tr><td>
          <p><code><a href="hspaCsrCutoff.html">hspaCsrCutoff()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="hspaCsrTesting.html">hspaCsrTesting()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="hspaDataBinarization.html">hspaDataBinarization()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="hspaPatternIdentification.html">hspaPatternIdentification()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="hspaPatternSimilarity.html">hspaPatternSimilarity()</a></code> <code><a href="hspaPatternSimilarity.html">hspaPatternSimilarity_future()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="hspaSelectGenes.html">hspaSelectGenes()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="identify_gene_patterns_dbscan.html">identify_gene_patterns_dbscan()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="image_dummy.html">image_dummy()</a></code> </p>
        </td>
        <td><p>image_dummy</p></td>
      </tr><tr><td>
          <p><code><a href="initiateSpataObject_10X.html">initiateSpataObject_10X()</a></code> </p>
        </td>
        <td><p>Initiate a spata-object from 10X Visium</p></td>
      </tr><tr><td>
          <p><code><a href="initiateSpataObject_CountMtr.html">initiateSpataObject_CountMtr()</a></code> </p>
        </td>
        <td><p>Initiate a spata-object from a raw count matrix</p></td>
      </tr><tr><td>
          <p><code><a href="initiateSpataObject_Empty.html">initiateSpataObject_Empty()</a></code> </p>
        </td>
        <td><p>Initiate an empty spata-object</p></td>
      </tr><tr><td>
          <p><code><a href="initiateSpataObject_Examples.html">initiateSpataObject_Examples()</a></code> </p>
        </td>
        <td><p>Initiate a spata-object from example data sets</p></td>
      </tr><tr><td>
          <p><code><a href="initiateSpataObject_ExprMtr.html">initiateSpataObject_ExprMtr()</a></code> </p>
        </td>
        <td><p>Initiate spata object from scaled expression matrix</p></td>
      </tr><tr><td>
          <p><code><a href="is_subsetted_by_segment.html">is_subsetted_by_segment()</a></code> </p>
        </td>
        <td><p>Check if object has been subsetted by segment</p></td>
      </tr><tr><td>
          <p><code><a href="joinWith.html">joinWith()</a></code> <code><a href="joinWith.html">joinWithFeatures()</a></code> <code><a href="joinWith.html">joinWithGenes()</a></code> <code><a href="joinWith.html">joinWithGeneSets()</a></code> <code><a href="joinWith.html">joinWithVariables()</a></code> </p>
        </td>
        <td><p>Join barcodes with additional variables</p></td>
      </tr><tr><td>
          <p><code><a href="joinWithPca.html">joinWithPca()</a></code> <code><a href="joinWithPca.html">joinWithTsne()</a></code> <code><a href="joinWithPca.html">joinWithUmap()</a></code> </p>
        </td>
        <td><p>Join barcodes with additional variables</p></td>
      </tr><tr><td>
          <p><code><a href="lazy_check_dummy.html">lazy_check_dummy()</a></code> </p>
        </td>
        <td><p>lazy_check_dummy</p></td>
      </tr><tr><td>
          <p><code><a href="legendBottom.html">legendBottom()</a></code> <code><a href="legendBottom.html">legendNone()</a></code> <code><a href="legendBottom.html">legendRight()</a></code> <code><a href="legendBottom.html">legendTop()</a></code> </p>
        </td>
        <td><p>Ggplot add on wrapper</p></td>
      </tr><tr><td>
          <p><code><a href="loadGeneSetDf.html">loadGeneSetDf()</a></code> </p>
        </td>
        <td><p>Load gene set data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="loadGSDF.html">loadGSDF()</a></code> </p>
        </td>
        <td><p>Original load gene set data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="loadSpataObject.html">loadSpataObject()</a></code> <code><a href="loadSpataObject.html">loadCorrespondingCDS()</a></code> <code><a href="loadSpataObject.html">loadCorrespondingSeuratObject()</a></code> </p>
        </td>
        <td><p>Load corresponding objects</p></td>
      </tr><tr><td>
          <p><code><a href="mergeSpataObjects.html">mergeSpataObjects()</a></code> </p>
        </td>
        <td><p>Merge Spata Objects</p></td>
      </tr><tr><td>
          <p><code><a href="method_hclust.html">method_hclust()</a></code> </p>
        </td>
        <td><p>method_hclust</p></td>
      </tr><tr><td>
          <p><code><a href="moduleAddGeneSetsServer.html">moduleAddGeneSetsServer()</a></code> </p>
        </td>
        <td><p>Server of the add gene sets module</p></td>
      </tr><tr><td>
          <p><code><a href="moduleAddGeneSetsUI.html">moduleAddGeneSetsUI()</a></code> </p>
        </td>
        <td><p>UI of the add gene sets module</p></td>
      </tr><tr><td>
          <p><code><a href="moduleSurfacePlotServer.html">moduleSurfacePlotServer()</a></code> </p>
        </td>
        <td><p>Server of the surface plot module</p></td>
      </tr><tr><td>
          <p><code><a href="moduleSurfacePlotUI.html">moduleSurfacePlotUI()</a></code> </p>
        </td>
        <td><p>UI of the surface plot module</p></td>
      </tr><tr><td>
          <p><code><a href="nGenes.html">nGenes()</a></code> </p>
        </td>
        <td><p>Number of genes</p></td>
      </tr><tr><td>
          <p><code><a href="normalize.html">normalize()</a></code> </p>
        </td>
        <td><p>normalize</p></td>
      </tr><tr><td>
          <p><code><a href="pb_dummy.html">pb_dummy()</a></code> </p>
        </td>
        <td><p>pb</p></td>
      </tr><tr><td>
          <p><code><a href="pheatmap.html">pheatmap()</a></code> </p>
        </td>
        <td><p>de_df</p></td>
      </tr><tr><td>
          <p><code><a href="plotAutoencoderAssessment.html">plotAutoencoderAssessment()</a></code> </p>
        </td>
        <td><p>Plot total variance of different neural networks</p></td>
      </tr><tr><td>
          <p><code><a href="plotAutoencoderResults.html">plotAutoencoderResults()</a></code> </p>
        </td>
        <td><p>Plot scaled vs. denoised expression</p></td>
      </tr><tr><td>
          <p><code><a href="plotBarchart.html">plotBarchart()</a></code> </p>
        </td>
        <td><p>Plot distribution of discrete variables</p></td>
      </tr><tr><td>
          <p><code><a href="plotBoxplot.html">plotBoxplot()</a></code> <code><a href="plotBoxplot.html">plotDensityplot()</a></code> <code><a href="plotBoxplot.html">plotHistogram()</a></code> <code><a href="plotBoxplot.html">plotRidgeplot()</a></code> <code><a href="plotBoxplot.html">plotViolinplot()</a></code> </p>
        </td>
        <td><p>Plot numeric distribution and statistical tests</p></td>
      </tr><tr><td>
          <p><code><a href="plotCnvResults.html">plotCnvResults()</a></code> </p>
        </td>
        <td><p>Visualize copy-number-variations results</p></td>
      </tr><tr><td>
          <p><code><a href="plotCsrCutoffSimulations.html">plotCsrCutoffSimulations()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="plotCsrResults.html">plotCsrResults()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="plotCustomizedTrajectoryTrends.html">plotCustomizedTrajectoryTrends()</a></code> </p>
        </td>
        <td><p>Plot customized trajectory trends</p></td>
      </tr><tr><td>
          <p><code><a href="plotDeaDotPlot.html">plotDeaDotPlot()</a></code> </p>
        </td>
        <td><p>Plot differentially expressed genes</p></td>
      </tr><tr><td>
          <p><code><a href="plotDeaGeneCount.html">plotDeaGeneCount()</a></code> </p>
        </td>
        <td><p>Plot the number of differently expressed genes of certain groups</p></td>
      </tr><tr><td>
          <p><code><a href="plotDeaHeatmap.html">plotDeaHeatmap()</a></code> </p>
        </td>
        <td><p>Plot differentially expressed genes</p></td>
      </tr><tr><td>
          <p><code><a href="plotDeaPvalues.html">plotDeaPvalues()</a></code> <code><a href="plotDeaPvalues.html">plotDeaLogFC()</a></code> </p>
        </td>
        <td><p>Plot the p-value and log fold change distribution of de-analysis results</p></td>
      </tr><tr><td>
          <p><code><a href="plotDeaSummary.html">plotDeaSummary()</a></code> </p>
        </td>
        <td><p>Plot a summary of differential expression analysis results</p></td>
      </tr><tr><td>
          <p><code><a href="plotDendrogram.html">plotDendrogram()</a></code> <code><a href="plotDendrogram.html">plotDendrogramCnv()</a></code> </p>
        </td>
        <td><p>Visualize hierarchical clustering results</p></td>
      </tr><tr><td>
          <p><code><a href="plotDistribution.html">plotDistribution()</a></code> <code><a href="plotDistribution.html">plotDistribution2()</a></code> <code><a href="plotDistribution.html">plotDistributionAcross()</a></code> <code><a href="plotDistribution.html">plotDistributionAcross2()</a></code> <code><a href="plotDistribution.html">plotDistributionDiscrete()</a></code> </p>
        </td>
        <td><p>Distribution of continuous values (Deprecated)</p></td>
      </tr><tr><td>
          <p><code><a href="plotFourStates.html">plotFourStates()</a></code> <code><a href="plotFourStates.html">plotFourStates2()</a></code> </p>
        </td>
        <td><p>Gene set state plot</p></td>
      </tr><tr><td>
          <p><code><a href="plotGeneDendrogram.html">plotGeneDendrogram()</a></code> </p>
        </td>
        <td><p>Visualize clustering results</p></td>
      </tr><tr><td>
          <p><code><a href="plotGeneMetaData.html">plotGeneMetaData()</a></code> </p>
        </td>
        <td><p>Plot gene meta data</p></td>
      </tr><tr><td>
          <p><code><a href="plotGenePatternBinarized.html">plotGenePatternBinarized()</a></code> <code><a href="plotGenePatternBinarized.html">plotGenePatternDenoised()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="plotGenePatternCluster.html">plotGenePatternCluster()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="plotGenePatternDendrogram.html">plotGenePatternDendrogram()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="plotGenePatternHeatmap.html">plotGenePatternHeatmap()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="plotGenePatterns.html">plotGenePatterns()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="plotGseaDotPlot.html">plotGseaDotPlot()</a></code> </p>
        </td>
        <td><p>Plot gene set enrichment</p></td>
      </tr><tr><td>
          <p><code><a href="plotHspaSummary.html">plotHspaSummary()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="plotPcaVariation.html">plotPcaVariation()</a></code> </p>
        </td>
        <td><p>Plot Pca Variation</p></td>
      </tr><tr><td>
          <p><code><a href="plotPseudotime.html">plotPseudotime()</a></code> </p>
        </td>
        <td><p>Monocle3 Pseudotime</p></td>
      </tr><tr><td>
          <p><code><a href="plotScatterplot.html">plotScatterplot()</a></code> </p>
        </td>
        <td><p>Plot numeric variables as a scatterplot</p></td>
      </tr><tr><td>
          <p><code><a href="plotSegmentation.html">plotSegmentation()</a></code> </p>
        </td>
        <td><p>Plot segmentation</p></td>
      </tr><tr><td>
          <p><code><a href="plotStatisticsInteractive.html">plotStatisticsInteractive()</a></code> </p>
        </td>
        <td><p>Plot distribution of variables interactively</p></td>
      </tr><tr><td>
          <p><code><a href="plotSurface.html">plotSurface()</a></code> <code><a href="plotSurface.html">plotSurface2()</a></code> <code><a href="plotSurface.html">plotSurfaceInteractive()</a></code> </p>
        </td>
        <td><p>Plot the surface of the sample</p></td>
      </tr><tr><td>
          <p><code><a href="plotSurfaceAverage.html">plotSurfaceAverage()</a></code> </p>
        </td>
        <td><p>Plot several surface plots colored by gene averages</p></td>
      </tr><tr><td>
          <p><code><a href="plotSurfaceComparison.html">plotSurfaceComparison()</a></code> <code><a href="plotSurfaceComparison.html">plotSurfaceComparison2()</a></code> </p>
        </td>
        <td><p>Plot several surface plots colored by numeric variables</p></td>
      </tr><tr><td>
          <p><code><a href="plotSurfaceHotspots.html">plotSurfaceHotspots()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="plotSurfaceQuantiles.html">plotSurfaceQuantiles()</a></code> </p>
        </td>
        <td><p>Plot a surface plot colored by binned numeric variables</p></td>
      </tr><tr><td>
          <p><code><a href="plotTrajectory.html">plotTrajectory()</a></code> </p>
        </td>
        <td><p>Plot trajectory</p></td>
      </tr><tr><td>
          <p><code><a href="plotTrajectoryFeatures.html">plotTrajectoryFeatures()</a></code> <code><a href="plotTrajectoryFeatures.html">plotTrajectoryGenes()</a></code> <code><a href="plotTrajectoryFeatures.html">plotTrajectoryGeneSets()</a></code> </p>
        </td>
        <td><p>Plot continuous trajectory dynamics in lineplots</p></td>
      </tr><tr><td>
          <p><code><a href="plotTrajectoryFeaturesDiscrete.html">plotTrajectoryFeaturesDiscrete()</a></code> </p>
        </td>
        <td><p>Plot discrete trajectory dynamics</p></td>
      </tr><tr><td>
          <p><code><a href="plotTrajectoryFit.html">plotTrajectoryFit()</a></code> <code><a href="plotTrajectoryFit.html">plotTrajectoryFitCustomized()</a></code> </p>
        </td>
        <td><p>Plot trajectory fit</p></td>
      </tr><tr><td>
          <p><code><a href="plotTrajectoryHeatmap.html">plotTrajectoryHeatmap()</a></code> </p>
        </td>
        <td><p>Plot trajectory expression dynamic in heatmap</p></td>
      </tr><tr><td>
          <p><code><a href="plotUmap.html">plotUmap()</a></code> <code><a href="plotUmap.html">plotTsne()</a></code> <code><a href="plotUmap.html">plotPca()</a></code> </p>
        </td>
        <td><p>Plot dimensional reduction</p></td>
      </tr><tr><td>
          <p><code><a href="plot_type_dummy.html">plot_type_dummy()</a></code> </p>
        </td>
        <td><p>plot_type</p></td>
      </tr><tr><td>
          <p><code><a href="printAutoencoderSummary.html">printAutoencoderSummary()</a></code> </p>
        </td>
        <td><p>Print autoencoder summary</p></td>
      </tr><tr><td>
          <p><code><a href="printDeaOverview.html">printDeaOverview()</a></code> </p>
        </td>
        <td><p>Print overview of all conducted de-analysis</p></td>
      </tr><tr><td>
          <p><code><a href="printDefaultInstructions.html">printDefaultInstructions()</a></code> </p>
        </td>
        <td><p>Print current default settings</p></td>
      </tr><tr><td>
          <p><code><a href="printGeneSetOverview.html">printGeneSetOverview()</a></code> </p>
        </td>
        <td><p>Print overview about the current gene sets</p></td>
      </tr><tr><td>
          <p><code><a href="print_family.html">print_family()</a></code> </p>
        </td>
        <td><p>print</p></td>
      </tr><tr><td>
          <p><code><a href="process_seurat_object.html">process_seurat_object()</a></code> </p>
        </td>
        <td><p>Wrapper around Seurat processing functions</p></td>
      </tr><tr><td>
          <p><code><a href="renameFeatures.html">renameFeatures()</a></code> </p>
        </td>
        <td><p>Rename features</p></td>
      </tr><tr><td>
          <p><code><a href="renameGroups.html">renameGroups()</a></code> <code><a href="renameGroups.html">renameSegments()</a></code> </p>
        </td>
        <td><p>Rename cluster/group names</p></td>
      </tr><tr><td>
          <p><code><a href="runAutoencoderAssessment.html">runAutoencoderAssessment()</a></code> <code><a href="runAutoencoderAssessment.html">assessAutoencoderOptions()</a></code> </p>
        </td>
        <td><p>Assessment of Neural Network Set Up</p></td>
      </tr><tr><td>
          <p><code><a href="runAutoencoderDenoising.html">runAutoencoderDenoising()</a></code> </p>
        </td>
        <td><p>Denoise expression matrix</p></td>
      </tr><tr><td>
          <p><code><a href="runCnvAnalysis.html">runCnvAnalysis()</a></code> </p>
        </td>
        <td><p>Identify large-scale chromosomal copy number variations</p></td>
      </tr><tr><td>
          <p><code><a href="runDeAnalysis.html">runDeAnalysis()</a></code> </p>
        </td>
        <td><p>Find differently expressed genes</p></td>
      </tr><tr><td>
          <p><code><a href="runGSEA.html">runGSEA()</a></code> </p>
        </td>
        <td><p>Compute gene set enrichment</p></td>
      </tr><tr><td>
          <p><code><a href="runHclust.html">runHclust()</a></code> <code><a href="runHclust.html">runHclustCnv()</a></code> </p>
        </td>
        <td><p>Run hierarchical clustering</p></td>
      </tr><tr><td>
          <p><code><a href="runHspa.html">runHspa()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="runPca.html">runPca()</a></code> <code><a href="runPca.html">runPca2()</a></code> </p>
        </td>
        <td><p>Run Principal Component Analysis</p></td>
      </tr><tr><td>
          <p><code><a href="runSpatialCorrelationAnalysis.html">runSpatialCorrelationAnalysis()</a></code> </p>
        </td>
        <td><p>Initiate gene clustering analysis based on spatial patterns</p></td>
      </tr><tr><td>
          <p><code><a href="runTsne.html">runTsne()</a></code> <code><a href="runTsne.html">runTsne2()</a></code> </p>
        </td>
        <td><p>Run t-Stochastic Neighbour Embedding</p></td>
      </tr><tr><td>
          <p><code><a href="runUmap.html">runUmap()</a></code> <code><a href="runUmap.html">runUmap2()</a></code> </p>
        </td>
        <td><p>Run UMAP-Analysis</p></td>
      </tr><tr><td>
          <p><code><a href="sample_name.html">sample_name()</a></code> </p>
        </td>
        <td><p>sample_name</p></td>
      </tr><tr><td>
          <p><code><a href="saveGeneSetDf.html">saveGeneSetDf()</a></code> </p>
        </td>
        <td><p>Save a gene set data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="saveSpataObject.html">saveSpataObject()</a></code> <code><a href="saveSpataObject.html">saveCorrespondingCDS()</a></code> <code><a href="saveSpataObject.html">saveCorrespondingSeuratObject()</a></code> </p>
        </td>
        <td><p>Save corresponding objects</p></td>
      </tr><tr><td>
          <p><code><a href="scale_color_add_on.html">scale_color_add_on()</a></code> </p>
        </td>
        <td><p>ggplot2 - color add on</p></td>
      </tr><tr><td>
          <p><code><a href="seurat_object_dummy.html">seurat_object_dummy()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="shiftTrajectoryDf.html">shiftTrajectoryDf()</a></code> </p>
        </td>
        <td><p>Shift trajectory data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="smoothSpatially.html">smoothSpatially()</a></code> </p>
        </td>
        <td><p>Smooth numeric variables spatially</p></td>
      </tr><tr><td>
          <p><code><a href="spata2-class.html">spata2-class</a></code> </p>
        </td>
        <td><p>The spata-object</p></td>
      </tr><tr><td>
          <p><code><a href="spatial_trajectory-class.html">spatial_trajectory-class</a></code> </p>
        </td>
        <td><p>spatial_trajectory object</p></td>
      </tr><tr><td>
          <p><code><a href="subsetBySegment_CountMtr.html">subsetBySegment_CountMtr()</a></code> <code><a href="subsetBySegment_CountMtr.html">subsetByBarcodes_CountMtr()</a></code> <code><a href="subsetBySegment_CountMtr.html">subsetBySegment_ExprMtr()</a></code> <code><a href="subsetBySegment_CountMtr.html">subsetByBarcodes_ExprMtr()</a></code> </p>
        </td>
        <td><p>Subset a spata-object</p></td>
      </tr><tr><td>
          <p><code><a href="surfaceFrame.html">surfaceFrame()</a></code> </p>
        </td>
        <td><p>Add frame of the sample</p></td>
      </tr><tr><td>
          <p><code><a href="tab_create_segmentation_return.html">tab_create_segmentation_return()</a></code> </p>
        </td>
        <td><p>Segmentation plot tab - return</p></td>
      </tr><tr><td>
          <p><code><a href="tab_create_trajectories_return.html">tab_create_trajectories_return()</a></code> </p>
        </td>
        <td><p>Trajectory plot tab - return</p></td>
      </tr><tr><td>
          <p><code><a href="tab_surface_plots_app.html">tab_surface_plots_app()</a></code> </p>
        </td>
        <td><p>Surface plot tab - classic</p></td>
      </tr><tr><td>
          <p><code><a href="tab_surface_plots_return.html">tab_surface_plots_return()</a></code> </p>
        </td>
        <td><p>Surface plot tab - return</p></td>
      </tr><tr><td>
          <p><code><a href="trajectory_patterns.html">trajectory_patterns</a></code> </p>
        </td>
        <td><p>Trajectory patterns</p></td>
      </tr><tr><td>
          <p><code><a href="transformSeuratToSpata.html">transformSeuratToSpata()</a></code> </p>
        </td>
        <td><p>Transform seurat-object to spata-object</p></td>
      </tr><tr><td>
          <p><code><a href="transformSpataToCDS.html">transformSpataToCDS()</a></code> </p>
        </td>
        <td><p>Transform spata-object to cell-data-set (Monocle3)</p></td>
      </tr><tr><td>
          <p><code><a href="transformSpataToSeurat.html">transformSpataToSeurat()</a></code> </p>
        </td>
        <td><p>Transform spata-object to a seurat-object</p></td>
      </tr><tr><td>
          <p><code><a href="updateSpataObject.html">updateSpataObject()</a></code> </p>
        </td>
        <td><p>Update spata-object from SPATA to SPATA2</p></td>
      </tr><tr><td>
          <p><code><a href="validActivationFunctions.html">validActivationFunctions()</a></code> <code><a href="validActivationFunctions.html">validAgglomerationMethods()</a></code> <code><a href="validActivationFunctions.html">validColorPalettes()</a></code> <code><a href="validActivationFunctions.html">validColorSpectra()</a></code> <code><a href="validActivationFunctions.html">validDeAnalysisMethods()</a></code> <code><a href="validActivationFunctions.html">validDefaultInstructionSlots()</a></code> <code><a href="validActivationFunctions.html">validDimRedMethods()</a></code> <code><a href="validActivationFunctions.html">validDirectoryInstructionSlots()</a></code> <code><a href="validActivationFunctions.html">validDistanceMethods()</a></code> <code><a href="validActivationFunctions.html">validHierarchicalClusterMethods()</a></code> <code><a href="validActivationFunctions.html">validPatternRecognitionMethods()</a></code> <code><a href="validActivationFunctions.html">validPadjMethods()</a></code> <code><a href="validActivationFunctions.html">validPlotTypes()</a></code> </p>
        </td>
        <td><p>Obtain valid argument inputs</p></td>
      </tr><tr><td>
          <p><code><a href="validateSpataObject.html">validateSpataObject()</a></code> </p>
        </td>
        <td><p>Validate a spata object</p></td>
      </tr><tr><td>
          <p><code><a href="validation.html">validation()</a></code> </p>
        </td>
        <td><p>Validate object input</p></td>
      </tr><tr><td>
          <p><code><a href="variable.html">variable()</a></code> </p>
        </td>
        <td><p>variable</p></td>
      </tr><tr><td>
          <p><code><a href="variables_num.html">variables_num()</a></code> </p>
        </td>
        <td><p>variables_num</p></td>
      </tr><tr><td>
          <p><code><a href="variable_num.html">variable_num()</a></code> </p>
        </td>
        <td><p>variable_num</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="add-content">Add Content <a href="#add-content" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p><p>The add-functions let you add content to the spata-object savely.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="addAutoencoderSetUp.html">addAutoencoderSetUp()</a></code> </p>
        </td>
        <td><p>Add the set up of a neural network</p></td>
      </tr><tr><td>
          <p><code><a href="addExpressionMatrix.html">addExpressionMatrix()</a></code> </p>
        </td>
        <td><p>Add an expression matrix</p></td>
      </tr><tr><td>
          <p><code><a href="addFeatures.html">addFeatures()</a></code> </p>
        </td>
        <td><p>Add a new feature</p></td>
      </tr><tr><td>
          <p><code><a href="addGeneFeatures.html">addGeneFeatures()</a></code> </p>
        </td>
        <td><p>Add new gene features</p></td>
      </tr><tr><td>
          <p><code><a href="addGeneMetaData.html">addGeneMetaData()</a></code> </p>
        </td>
        <td><p>Add gene meta data to the object</p></td>
      </tr><tr><td>
          <p><code><a href="addGeneSet.html">addGeneSet()</a></code> <code><a href="addGeneSet.html">addGeneSetsInteractive()</a></code> </p>
        </td>
        <td><p>Add a new gene set</p></td>
      </tr><tr><td>
          <p><code><a href="addSpCorCluster.html">addSpCorCluster()</a></code> </p>
        </td>
        <td><p>Add cluster results of spatial correlation results</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="discard-content">Discard Content <a href="#discard-content" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p><p>The discard-functions let you delete unwanted information savely.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="discardExpressionMatrix.html">discardExpressionMatrix()</a></code> </p>
        </td>
        <td><p>Discard an expression matrix</p></td>
      </tr><tr><td>
          <p><code><a href="discardFeatures.html">discardFeatures()</a></code> </p>
        </td>
        <td><p>Discard features</p></td>
      </tr><tr><td>
          <p><code><a href="discardGeneFeatures.html">discardGeneFeatures()</a></code> </p>
        </td>
        <td><p>Discard gene features</p></td>
      </tr><tr><td>
          <p><code><a href="discardGenes.html">discardGenes()</a></code> </p>
        </td>
        <td><p>Discard genes</p></td>
      </tr><tr><td>
          <p><code><a href="discardGeneSets.html">discardGeneSets()</a></code> </p>
        </td>
        <td><p>Discard gene sets</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="create-content-interactively">Create Content Interactively <a href="#create-content-interactively" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p><p>The create-functions open an interactive application in which the content to be added needs to be created at first.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="createSegmentation.html">createSegmentation()</a></code> </p>
        </td>
        <td><p>Spatial Segmentation</p></td>
      </tr><tr><td>
          <p><code><a href="createTrajectories.html">createTrajectories()</a></code> </p>
        </td>
        <td><p>Spatial Trajectories</p></td>
      </tr><tr><td>
          <p><code><a href="createTrajectoryManually.html">createTrajectoryManually()</a></code> </p>
        </td>
        <td><p>Create spatial trajectories manually</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="adjust-content">Adjust Content <a href="#adjust-content" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p><p>The adjust-functions let you adjust content of the spata-object without overwriting it in it’s essence.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="adjustDefaultInstructions.html">adjustDefaultInstructions()</a></code> <code><a href="adjustDefaultInstructions.html">adjustDirectoryInstructions()</a></code> </p>
        </td>
        <td><p>Adjust default instructions</p></td>
      </tr><tr><td>
          <p><code><a href="adjustGeneSetDf.html">adjustGeneSetDf()</a></code> </p>
        </td>
        <td><p>Filter gene-set data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="adjusting_check_dummy.html">adjusting_check_dummy()</a></code> </p>
        </td>
        <td><p>This is a text dummy</p></td>
      </tr><tr><td>
          <p><code><a href="across.html">across()</a></code> <code><a href="across.html">across_dummy()</a></code> </p>
        </td>
        <td><p>across</p></td>
      </tr><tr><td>
          <p><code><a href="addAutoencoderSetUp.html">addAutoencoderSetUp()</a></code> </p>
        </td>
        <td><p>Add the set up of a neural network</p></td>
      </tr><tr><td>
          <p><code><a href="addExpressionMatrix.html">addExpressionMatrix()</a></code> </p>
        </td>
        <td><p>Add an expression matrix</p></td>
      </tr><tr><td>
          <p><code><a href="addFeatures.html">addFeatures()</a></code> </p>
        </td>
        <td><p>Add a new feature</p></td>
      </tr><tr><td>
          <p><code><a href="addGeneFeatures.html">addGeneFeatures()</a></code> </p>
        </td>
        <td><p>Add new gene features</p></td>
      </tr><tr><td>
          <p><code><a href="addGeneMetaData.html">addGeneMetaData()</a></code> </p>
        </td>
        <td><p>Add gene meta data to the object</p></td>
      </tr><tr><td>
          <p><code><a href="addGeneSet.html">addGeneSet()</a></code> <code><a href="addGeneSet.html">addGeneSetsInteractive()</a></code> </p>
        </td>
        <td><p>Add a new gene set</p></td>
      </tr><tr><td>
          <p><code><a href="addSpCorCluster.html">addSpCorCluster()</a></code> </p>
        </td>
        <td><p>Add cluster results of spatial correlation results</p></td>
      </tr><tr><td>
          <p><code><a href="allColorpalettes.html">allColorpalettes()</a></code> <code><a href="allColorpalettes.html">allColorspectra()</a></code> </p>
        </td>
        <td><p>Valid colorpanels &amp; -spectra</p></td>
      </tr><tr><td>
          <p><code><a href="all_color_palettes.html">all_color_palettes()</a></code> <code><a href="all_color_palettes.html">all_color_spectra()</a></code> </p>
        </td>
        <td><p>Color palette names</p></td>
      </tr><tr><td>
          <p><code><a href="argument_dummy.html">argument_dummy()</a></code> </p>
        </td>
        <td><p>Argument dummy</p></td>
      </tr><tr><td>
          <p><code><a href="assessTrajectoryTrends.html">assessTrajectoryTrends()</a></code> <code><a href="assessTrajectoryTrends.html">assessTrajectoryTrends2()</a></code> <code><a href="assessTrajectoryTrends.html">assessTrajectoryTrendsCustomized()</a></code> <code><a href="assessTrajectoryTrends.html">assessTrajectoryTrendsCustomized2()</a></code> </p>
        </td>
        <td><p>Trajectory trend analysis</p></td>
      </tr><tr><td>
          <p><code><a href="average_genes.html">average_genes()</a></code> </p>
        </td>
        <td><p>average_genes</p></td>
      </tr><tr><td>
          <p><code><a href="binarize_matrix.html">binarize_matrix()</a></code> </p>
        </td>
        <td><p>Binarize expression matrix</p></td>
      </tr><tr><td>
          <p><code><a href="binwidth_dummy.html">binwidth_dummy()</a></code> </p>
        </td>
        <td><p>binwidth</p></td>
      </tr><tr><td>
          <p><code><a href="cds_dummy.html">cds_dummy()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="checkpoint.html">checkpoint()</a></code> </p>
        </td>
        <td><p>Shiny feedback messages</p></td>
      </tr><tr><td>
          <p><code><a href="check_all_barcodes.html">check_all_barcodes()</a></code> </p>
        </td>
        <td><p>Check barcodes of all matrices and data.frames</p></td>
      </tr><tr><td>
          <p><code><a href="check_assign.html">check_assign()</a></code> </p>
        </td>
        <td><p>Check assign input</p></td>
      </tr><tr><td>
          <p><code><a href="check_atdf.html">check_atdf()</a></code> <code><a href="check_atdf.html">check_rtdf()</a></code> </p>
        </td>
        <td><p>Check assessed trajectory data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="check_availability.html">check_availability()</a></code> </p>
        </td>
        <td><p>Gives feedback if object was not found</p></td>
      </tr><tr><td>
          <p><code><a href="check_color_to.html">check_color_to()</a></code> </p>
        </td>
        <td><p>Check color to</p></td>
      </tr><tr><td>
          <p><code><a href="check_compiled_trajectory_df.html">check_compiled_trajectory_df()</a></code> </p>
        </td>
        <td><p>Check compiled trajectory data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="check_coordinate_variables.html">check_coordinate_variables()</a></code> </p>
        </td>
        <td><p>Check coordinate variables</p></td>
      </tr><tr><td>
          <p><code><a href="check_coords_df.html">check_coords_df()</a></code> </p>
        </td>
        <td><p>Check coords data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="check_customized_trends.html">check_customized_trends()</a></code> </p>
        </td>
        <td><p>Check customized trend list</p></td>
      </tr><tr><td>
          <p><code><a href="check_dea_df.html">check_dea_df()</a></code> </p>
        </td>
        <td><p>Check de data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="check_display.html">check_display()</a></code> </p>
        </td>
        <td><p>Check display input</p></td>
      </tr><tr><td>
          <p><code><a href="check_features.html">check_features()</a></code> </p>
        </td>
        <td><p>Check feature variables input</p></td>
      </tr><tr><td>
          <p><code><a href="check_feature_df.html">check_feature_df()</a></code> </p>
        </td>
        <td><p>Check feature data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="check_genes.html">check_genes()</a></code> </p>
        </td>
        <td><p>Check gene input</p></td>
      </tr><tr><td>
          <p><code><a href="check_gene_sets.html">check_gene_sets()</a></code> </p>
        </td>
        <td><p>Check gene set input</p></td>
      </tr><tr><td>
          <p><code><a href="check_method.html">check_method()</a></code> </p>
        </td>
        <td><p>Check method input</p></td>
      </tr><tr><td>
          <p><code><a href="check_monocle_input.html">check_monocle_input()</a></code> </p>
        </td>
        <td><p>Check monocle input</p></td>
      </tr><tr><td>
          <p><code><a href="check_monocle_packages.html">check_monocle_packages()</a></code> </p>
        </td>
        <td><p>Makes sure that both packages are installed</p></td>
      </tr><tr><td>
          <p><code><a href="check_object.html">check_object()</a></code> </p>
        </td>
        <td><p>Check spata object input</p></td>
      </tr><tr><td>
          <p><code><a href="check_pattern.html">check_pattern()</a></code> </p>
        </td>
        <td><p>Check pattern input</p></td>
      </tr><tr><td>
          <p><code><a href="check_pt.html">check_pt()</a></code> </p>
        </td>
        <td><p>Check pt input</p></td>
      </tr><tr><td>
          <p><code><a href="check_sample.html">check_sample()</a></code> </p>
        </td>
        <td><p>Check sample input</p></td>
      </tr><tr><td>
          <p><code><a href="check_saving.html">check_saving()</a></code> </p>
        </td>
        <td><p>Check saving input</p></td>
      </tr><tr><td>
          <p><code><a href="check_segment.html">check_segment()</a></code> </p>
        </td>
        <td><p>Check segment name</p></td>
      </tr><tr><td>
          <p><code><a href="check_slot_coordinates.html">check_slot_coordinates()</a></code> <code><a href="check_slot_coordinates.html">check_slot_data()</a></code> <code><a href="check_slot_coordinates.html">check_slot_version()</a></code> <code><a href="check_slot_coordinates.html">check_slot_scvelo()</a></code> <code><a href="check_slot_coordinates.html">check_slot_trajectories()</a></code> <code><a href="check_slot_coordinates.html">check_trajectory_object()</a></code> <code><a href="check_slot_coordinates.html">check_slot_dim_red()</a></code> <code><a href="check_slot_coordinates.html">check_slot_fdata()</a></code> <code><a href="check_slot_coordinates.html">check_slot_image()</a></code> <code><a href="check_slot_coordinates.html">check_slot_used_genesets()</a></code> <code><a href="check_slot_coordinates.html">check_slot_samples()</a></code> </p>
        </td>
        <td><p>Check spata slots</p></td>
      </tr><tr><td>
          <p><code><a href="check_smooth.html">check_smooth()</a></code> </p>
        </td>
        <td><p>Check smooth input</p></td>
      </tr><tr><td>
          <p><code><a href="check_spata_df.html">check_spata_df()</a></code> </p>
        </td>
        <td><p>Check coordinate data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="check_stdf.html">check_stdf()</a></code> </p>
        </td>
        <td><p>Check summarized trajectory data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="check_summarized_trajectory_df.html">check_summarized_trajectory_df()</a></code> </p>
        </td>
        <td><p>Check stdf-input</p></td>
      </tr><tr><td>
          <p><code><a href="check_to.html">check_to()</a></code> </p>
        </td>
        <td><p>Gives feedback about input validity</p></td>
      </tr><tr><td>
          <p><code><a href="check_trajectory.html">check_trajectory()</a></code> </p>
        </td>
        <td><p>Check trajectory name input</p></td>
      </tr><tr><td>
          <p><code><a href="check_trajectory_binwidth.html">check_trajectory_binwidth()</a></code> </p>
        </td>
        <td><p>Check trajectory binwdith input</p></td>
      </tr><tr><td>
          <p><code><a href="check_uniform_genes.html">check_uniform_genes()</a></code> </p>
        </td>
        <td><p>Check uniform genes input</p></td>
      </tr><tr><td>
          <p><code><a href="check_variables.html">check_variables()</a></code> </p>
        </td>
        <td><p>Check variables</p></td>
      </tr><tr><td>
          <p><code><a href="clrp.html">clrp()</a></code> <code><a href="clrp.html">clrp_dummy()</a></code> </p>
        </td>
        <td><p>clrp</p></td>
      </tr><tr><td>
          <p><code><a href="clusterSpCorResults.html">clusterSpCorResults()</a></code> </p>
        </td>
        <td><p>Cluster genes according to their expression profile</p></td>
      </tr><tr><td>
          <p><code><a href="cnv_ref.html">cnv_ref</a></code> </p>
        </td>
        <td><p>A list of reference data for copy number variation analysis (CNV)</p></td>
      </tr><tr><td>
          <p><code><a href="cnv_regions_df.html">cnv_regions_df</a></code> </p>
        </td>
        <td><p>A data.frame necessary for cnv-analysis. Contains information about start and
end points of chromosomes.</p></td>
      </tr><tr><td>
          <p><code><a href="computeGeneMetaData.html">computeGeneMetaData()</a></code> <code><a href="computeGeneMetaData.html">computeGeneMetaData2()</a></code> </p>
        </td>
        <td><p>Compute gene summary statistics</p></td>
      </tr><tr><td>
          <p><code><a href="compute_pattern_similarity.html">compute_pattern_similarity()</a></code> <code><a href="compute_pattern_similarity.html">compute_pattern_relation()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="createSegmentation.html">createSegmentation()</a></code> </p>
        </td>
        <td><p>Spatial Segmentation</p></td>
      </tr><tr><td>
          <p><code><a href="createTrajectories.html">createTrajectories()</a></code> </p>
        </td>
        <td><p>Spatial Trajectories</p></td>
      </tr><tr><td>
          <p><code><a href="createTrajectoryManually.html">createTrajectoryManually()</a></code> </p>
        </td>
        <td><p>Create spatial trajectories manually</p></td>
      </tr><tr><td>
          <p><code><a href="data_counts-class.html">data_counts-class</a></code> </p>
        </td>
        <td><p>data_counts object</p></td>
      </tr><tr><td>
          <p><code><a href="default_instructions-class.html">default_instructions-class</a></code> </p>
        </td>
        <td><p>default instructions</p></td>
      </tr><tr><td>
          <p><code><a href="dim_red-class.html">dim_red-class</a></code> </p>
        </td>
        <td><p>dim_red object</p></td>
      </tr><tr><td>
          <p><code><a href="discardExpressionMatrix.html">discardExpressionMatrix()</a></code> </p>
        </td>
        <td><p>Discard an expression matrix</p></td>
      </tr><tr><td>
          <p><code><a href="discardFeatures.html">discardFeatures()</a></code> </p>
        </td>
        <td><p>Discard features</p></td>
      </tr><tr><td>
          <p><code><a href="discardGeneFeatures.html">discardGeneFeatures()</a></code> </p>
        </td>
        <td><p>Discard gene features</p></td>
      </tr><tr><td>
          <p><code><a href="discardGenes.html">discardGenes()</a></code> </p>
        </td>
        <td><p>Discard genes</p></td>
      </tr><tr><td>
          <p><code><a href="discardGeneSets.html">discardGeneSets()</a></code> </p>
        </td>
        <td><p>Discard gene sets</p></td>
      </tr><tr><td>
          <p><code><a href="dropped_df_dummy.html">dropped_df_dummy()</a></code> </p>
        </td>
        <td><p>dropped_df</p></td>
      </tr><tr><td>
          <p><code><a href="examineClusterResults.html">examineClusterResults()</a></code> </p>
        </td>
        <td><p>Examine clustering results</p></td>
      </tr><tr><td>
          <p><code><a href="examineTrajectoryAssessment.html">examineTrajectoryAssessment()</a></code> </p>
        </td>
        <td><p>Examine trajectory-moddeling results</p></td>
      </tr><tr><td>
          <p><code><a href="filterDeaDf.html">filterDeaDf()</a></code> </p>
        </td>
        <td><p>Postprocess de-analysis results</p></td>
      </tr><tr><td>
          <p><code><a href="filterTrends.html">filterTrends()</a></code> <code><a href="filterTrends.html">filterTrajectoryTrends()</a></code> </p>
        </td>
        <td><p>Filter variables of a certain trend</p></td>
      </tr><tr><td>
          <p><code><a href="findHierarchicalCluster.html">findHierarchicalCluster()</a></code> <code><a href="findHierarchicalCluster.html">findHierarchicalClusterCnv()</a></code> </p>
        </td>
        <td><p>Cluster sample via hclust</p></td>
      </tr><tr><td>
          <p><code><a href="findMonocleClusters.html">findMonocleClusters()</a></code> </p>
        </td>
        <td><p>Cluster sample via monocle3</p></td>
      </tr><tr><td>
          <p><code><a href="findNearestNeighbourClusters.html">findNearestNeighbourClusters()</a></code> </p>
        </td>
        <td><p>Cluster sample via nearest neighbour analysis</p></td>
      </tr><tr><td>
          <p><code><a href="findSeuratClusters.html">findSeuratClusters()</a></code> </p>
        </td>
        <td><p>Cluster sample via Seurat</p></td>
      </tr><tr><td>
          <p><code><a href="flexible_call_dummy.html">flexible_call_dummy()</a></code> </p>
        </td>
        <td><p>flexible_call_dummy</p></td>
      </tr><tr><td>
          <p><code><a href="flipImage.html">flipImage()</a></code> </p>
        </td>
        <td><p>Flip object image</p></td>
      </tr><tr><td>
          <p><code><a href="gene_set_path.html">gene_set_path()</a></code> </p>
        </td>
        <td><p>gene_set_path</p></td>
      </tr><tr><td>
          <p><code><a href="getActiveMatrixName.html">getActiveMatrixName()</a></code> </p>
        </td>
        <td><p>Obtain name of currently active expression matrix</p></td>
      </tr><tr><td>
          <p><code><a href="getAutoencoderAssessment.html">getAutoencoderAssessment()</a></code> </p>
        </td>
        <td><p>Obtain information about the optimal neural network set up</p></td>
      </tr><tr><td>
          <p><code><a href="getAutoencoderSetUp.html">getAutoencoderSetUp()</a></code> </p>
        </td>
        <td><p>Obtain information on neural network</p></td>
      </tr><tr><td>
          <p><code><a href="getBarcodes.html">getBarcodes()</a></code> </p>
        </td>
        <td><p>Obtain specific barcodes</p></td>
      </tr><tr><td>
          <p><code><a href="getCnvFeatureNames.html">getCnvFeatureNames()</a></code> </p>
        </td>
        <td><p>Obtain features names under which cnv-analysis results are stored.</p></td>
      </tr><tr><td>
          <p><code><a href="getCnvResults.html">getCnvResults()</a></code> </p>
        </td>
        <td><p>Obtain copy-number-variations results</p></td>
      </tr><tr><td>
          <p><code><a href="getCoordsDf.html">getCoordsDf()</a></code> <code><a href="getCoordsDf.html">getSegmentDf()</a></code> </p>
        </td>
        <td><p>Obtain spatial coordinates</p></td>
      </tr><tr><td>
          <p><code><a href="getDeaOverview.html">getDeaOverview()</a></code> </p>
        </td>
        <td><p>Obtain info on de-analysis storage</p></td>
      </tr><tr><td>
          <p><code><a href="getDeaResultsDf.html">getDeaResultsDf()</a></code> <code><a href="getDeaResultsDf.html">getDeaGenes()</a></code> </p>
        </td>
        <td><p>Obtain de-analysis results</p></td>
      </tr><tr><td>
          <p><code><a href="getDefaultInstructions.html">getDefaultInstructions()</a></code> <code><a href="getDefaultInstructions.html">getDirectoryInstructions()</a></code> </p>
        </td>
        <td><p>Obtain default argument inputs</p></td>
      </tr><tr><td>
          <p><code><a href="getDimRedDf.html">getDimRedDf()</a></code> <code><a href="getDimRedDf.html">getPcaDf()</a></code> <code><a href="getDimRedDf.html">getPcaMtr()</a></code> <code><a href="getDimRedDf.html">getUmapDf()</a></code> <code><a href="getDimRedDf.html">getTsneDf()</a></code> </p>
        </td>
        <td><p>Obtain dimensional reduction data</p></td>
      </tr><tr><td>
          <p><code><a href="getExpressionMatrixNames.html">getExpressionMatrixNames()</a></code> </p>
        </td>
        <td><p>Obtain names of stored expression matrices</p></td>
      </tr><tr><td>
          <p><code><a href="getFeatureDf.html">getFeatureDf()</a></code> </p>
        </td>
        <td><p>Obtain feature data</p></td>
      </tr><tr><td>
          <p><code><a href="getFeatureNames.html">getFeatureNames()</a></code> </p>
        </td>
        <td><p>Obtain feature names</p></td>
      </tr><tr><td>
          <p><code><a href="getFeatureValues.html">getFeatureValues()</a></code> </p>
        </td>
        <td><p>Obtain unique categorical feature values</p></td>
      </tr><tr><td>
          <p><code><a href="getFeatureVariables.html">getFeatureVariables()</a></code> </p>
        </td>
        <td><p>Obtain a feature variable</p></td>
      </tr><tr><td>
          <p><code><a href="getFromSeurat.html">getFromSeurat()</a></code> </p>
        </td>
        <td><p>Safe extraction</p></td>
      </tr><tr><td>
          <p><code><a href="getGeneCounts.html">getGeneCounts()</a></code> </p>
        </td>
        <td><p>Obtain total number of gene counts</p></td>
      </tr><tr><td>
          <p><code><a href="getGeneDistMtr.html">getGeneDistMtr()</a></code> </p>
        </td>
        <td><p>Obtain distance measurements of spatially correlated genes</p></td>
      </tr><tr><td>
          <p><code><a href="getGeneFeatureNames.html">getGeneFeatureNames()</a></code> </p>
        </td>
        <td><p>Obtain feature names of the gene meta data</p></td>
      </tr><tr><td>
          <p><code><a href="getGeneMetaData.html">getGeneMetaData()</a></code> <code><a href="getGeneMetaData.html">getGeneMetaDf()</a></code> </p>
        </td>
        <td><p>Obtain gene meta data</p></td>
      </tr><tr><td>
          <p><code><a href="getGenePatternCluster.html">getGenePatternCluster()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="getGenePatternCorrelations.html">getGenePatternCorrelations()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="getGenePatternDistances.html">getGenePatternDistances()</a></code> <code><a href="getGenePatternDistances.html">getGenePatternSimilarities()</a></code> </p>
        </td>
        <td><p>Obtain similarity scores between identified gene patterns</p></td>
      </tr><tr><td>
          <p><code><a href="getGenePatternEvalDf.html">getGenePatternEvalDf()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="getGenePatternExtentDf.html">getGenePatternExtentDf()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="getGenePatternsAboveThreshold.html">getGenePatternsAboveThreshold()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="getGenes.html">getGenes()</a></code> <code><a href="getGenes.html">getGenesInteractive()</a></code> </p>
        </td>
        <td><p>Obtain gene names</p></td>
      </tr><tr><td>
          <p><code><a href="getGeneSetDf.html">getGeneSetDf()</a></code> </p>
        </td>
        <td><p>Obtain gene set data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="getGeneSetOverview.html">getGeneSetOverview()</a></code> </p>
        </td>
        <td><p>Overview about the current gene sets</p></td>
      </tr><tr><td>
          <p><code><a href="getGeneSets.html">getGeneSets()</a></code> <code><a href="getGeneSets.html">getGeneSetsInteractive()</a></code> </p>
        </td>
        <td><p>Obtain gene set names</p></td>
      </tr><tr><td>
          <p><code><a href="getGroupingOptions.html">getGroupingOptions()</a></code> </p>
        </td>
        <td><p>Obtain variable names that group the barcode spots</p></td>
      </tr><tr><td>
          <p><code><a href="getGroupNames.html">getGroupNames()</a></code> </p>
        </td>
        <td><p>Obtain group names a grouping variable contains</p></td>
      </tr><tr><td>
          <p><code><a href="getGseaDf.html">getGseaDf()</a></code> </p>
        </td>
        <td><p>Obtain enrichment data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="getGseaResults.html">getGseaResults()</a></code> </p>
        </td>
        <td><p>Obtain enrichment results</p></td>
      </tr><tr><td>
          <p><code><a href="getHclustObject.html">getHclustObject()</a></code> <code><a href="getHclustObject.html">getHclustObjectCnv()</a></code> </p>
        </td>
        <td><p>Obtain hierarchical cluster results</p></td>
      </tr><tr><td>
          <p><code><a href="getHspaResults.html">getHspaResults()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="getImage.html">getImage()</a></code> </p>
        </td>
        <td><p>Obtain histology image</p></td>
      </tr><tr><td>
          <p><code><a href="getInitiationInfo.html">getInitiationInfo()</a></code> <code><a href="getInitiationInfo.html">getInitiationInput()</a></code> </p>
        </td>
        <td><p>Obtain information about object initiation</p></td>
      </tr><tr><td>
          <p><code><a href="getMatrix.html">getMatrix()</a></code> <code><a href="getMatrix.html">getExpressionMatrix()</a></code> <code><a href="getMatrix.html">getCountMatrix()</a></code> </p>
        </td>
        <td><p>Obtain count and expression matrix</p></td>
      </tr><tr><td>
          <p><code><a href="getPrResults.html">getPrResults()</a></code> <code><a href="getPrResults.html">getPatternNames()</a></code> </p>
        </td>
        <td><p>Obtain pattern recognition results</p></td>
      </tr><tr><td>
          <p><code><a href="getSampleNames.html">getSampleNames()</a></code> </p>
        </td>
        <td><p>Obtain sample names</p></td>
      </tr><tr><td>
          <p><code><a href="getSegmentNames.html">getSegmentNames()</a></code> </p>
        </td>
        <td><p>Obtain segment names</p></td>
      </tr><tr><td>
          <p><code><a href="getSignatureEnrichment.html">getSignatureEnrichment()</a></code> </p>
        </td>
        <td><p>Obtain signature enrichment</p></td>
      </tr><tr><td>
          <p><code><a href="getSpataDf.html">getSpataDf()</a></code> </p>
        </td>
        <td><p>Obtain a spata-data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="getSpCorCluster.html">getSpCorCluster()</a></code> <code><a href="getSpCorCluster.html">getSpCorClusterNames()</a></code> </p>
        </td>
        <td><p>Obtain cluster results based on spatial correlation analysis</p></td>
      </tr><tr><td>
          <p><code><a href="getSpCorResults.html">getSpCorResults()</a></code> </p>
        </td>
        <td><p>Obtain spatial correlation results</p></td>
      </tr><tr><td>
          <p><code><a href="getTrajectoryComment-spata2-method.html">getTrajectoryComment(<i>&lt;spata2&gt;</i>)</a></code> </p>
        </td>
        <td><p>Obtain a trajectory comment</p></td>
      </tr><tr><td>
          <p><code><a href="getTrajectoryComment.html">getTrajectoryComment()</a></code> </p>
        </td>
        <td><p>Obtain a trajectory comment</p></td>
      </tr><tr><td>
          <p><code><a href="getTrajectoryDf.html">getTrajectoryDf()</a></code> </p>
        </td>
        <td><p>Obtain a summarized trajectory data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="getTrajectoryLength.html">getTrajectoryLength()</a></code> </p>
        </td>
        <td><p>Obtain the length of a trajectory</p></td>
      </tr><tr><td>
          <p><code><a href="getTrajectoryNames.html">getTrajectoryNames()</a></code> </p>
        </td>
        <td><p>Obtain trajectory names</p></td>
      </tr><tr><td>
          <p><code><a href="getTrajectoryObject.html">getTrajectoryObject()</a></code> </p>
        </td>
        <td><p>Obtain trajectory object</p></td>
      </tr><tr><td>
          <p><code><a href="ggplot_family.html">ggplot_family()</a></code> </p>
        </td>
        <td><p>ggplot_family</p></td>
      </tr><tr><td>
          <p><code><a href="gsdf.html">gsdf</a></code> </p>
        </td>
        <td><p>The default collection of frequently used gene-sets</p></td>
      </tr><tr><td>
          <p><code><a href="hlpr_add_barcode_suffix.html">hlpr_add_barcode_suffix()</a></code> </p>
        </td>
        <td><p>Make sure that barcodes are spata-like</p></td>
      </tr><tr><td>
          <p><code><a href="hlpr_add_models.html">hlpr_add_models()</a></code> <code><a href="hlpr_add_models.html">hlpr_add_residuals()</a></code> <code><a href="hlpr_add_models.html">hlpr_add_residuals_customized()</a></code> <code><a href="hlpr_add_models.html">hlpr_summarize_residuals()</a></code> <code><a href="hlpr_add_models.html">hlpr_name_models()</a></code> <code><a href="hlpr_add_models.html">hlpr_filter_trend()</a></code> </p>
        </td>
        <td><p>Helper functions for trajectory ranking</p></td>
      </tr><tr><td>
          <p><code><a href="hlpr_add_old_coords.html">hlpr_add_old_coords()</a></code> </p>
        </td>
        <td><p>Adds old coordinates</p></td>
      </tr><tr><td>
          <p><code><a href="hlpr_adjust_legend_size.html">hlpr_adjust_legend_size()</a></code> </p>
        </td>
        <td><p>Adjusts the size of discrete legend points</p></td>
      </tr><tr><td>
          <p><code><a href="hlpr_assess_trajectory_trends.html">hlpr_assess_trajectory_trends()</a></code> <code><a href="hlpr_assess_trajectory_trends.html">hlpr_assess_trajectory_trends_customized()</a></code> </p>
        </td>
        <td><p>Assess trajectory ranking.</p></td>
      </tr><tr><td>
          <p><code><a href="hlpr_assign_arguments.html">hlpr_assign_arguments()</a></code> </p>
        </td>
        <td><p>Assign arguments to calling function</p></td>
      </tr><tr><td>
          <p><code><a href="hlpr_breaks.html">hlpr_breaks()</a></code> </p>
        </td>
        <td><p>Calculates breaks for heatmap according to input matrix</p></td>
      </tr><tr><td>
          <p><code><a href="hlpr_compare_samples.html">hlpr_compare_samples()</a></code> </p>
        </td>
        <td><p>Compare samples within an object</p></td>
      </tr><tr><td>
          <p><code><a href="hlpr_compile_trajectory.html">hlpr_compile_trajectory()</a></code> </p>
        </td>
        <td><p>Compiles a trajectory data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="hlpr_display_title.html">hlpr_display_title()</a></code> <code><a href="hlpr_display_title.html">hlpr_display_subtitle()</a></code> </p>
        </td>
        <td><p>No documentation needed</p></td>
      </tr><tr><td>
          <p><code><a href="hlpr_dist_mtr_to_df.html">hlpr_dist_mtr_to_df()</a></code> </p>
        </td>
        <td><p>Convert distance matrix to distance data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="hlpr_gene_set_name.html">hlpr_gene_set_name()</a></code> </p>
        </td>
        <td><p>Removes the class part of a gene set string</p></td>
      </tr><tr><td>
          <p><code><a href="hlpr_geom_trajectory_fit.html">hlpr_geom_trajectory_fit()</a></code> </p>
        </td>
        <td><p>Easy switch between geom line and and geom smooth</p></td>
      </tr><tr><td>
          <p><code><a href="hlpr_image_add_on.html">hlpr_image_add_on()</a></code> <code><a href="hlpr_image_add_on.html">hlpr_image_add_on2()</a></code> </p>
        </td>
        <td><p>Provides the image as ggplot background</p></td>
      </tr><tr><td>
          <p><code><a href="hlpr_labs_add_on.html">hlpr_labs_add_on()</a></code> </p>
        </td>
        <td><p>Return customized ggplot:labs()</p></td>
      </tr><tr><td>
          <p><code><a href="hlpr_normalize_imap.html">hlpr_normalize_imap()</a></code> </p>
        </td>
        <td><p>Normalize gene or gene set values</p></td>
      </tr><tr><td>
          <p><code><a href="hlpr_normalize_vctr.html">hlpr_normalize_vctr()</a></code> </p>
        </td>
        <td><p>Normalize gene or gene set values</p></td>
      </tr><tr><td>
          <p><code><a href="hlpr_one_distinct.html">hlpr_one_distinct()</a></code> </p>
        </td>
        <td><p>Number of distinct values equal to 1?</p></td>
      </tr><tr><td>
          <p><code><a href="hlpr_process_spatial_correlation_cluster.html">hlpr_process_spatial_correlation_cluster()</a></code> </p>
        </td>
        <td><p>Process spatial correlation results</p></td>
      </tr><tr><td>
          <p><code><a href="hlpr_rank_trajectory_trends.html">hlpr_rank_trajectory_trends()</a></code> <code><a href="hlpr_rank_trajectory_trends.html">hlpr_rank_trajectory_trends_customized()</a></code> </p>
        </td>
        <td><p>Rank trajectory trends.</p></td>
      </tr><tr><td>
          <p><code><a href="hlpr_run_cnva_pca.html">hlpr_run_cnva_pca()</a></code> </p>
        </td>
        <td><p>run pca on the cnv-matrix</p></td>
      </tr><tr><td>
          <p><code><a href="hlpr_save_spata_object.html">hlpr_save_spata_object()</a></code> </p>
        </td>
        <td><p>Save spata object inside functions</p></td>
      </tr><tr><td>
          <p><code><a href="hlpr_scatterplot.html">hlpr_scatterplot()</a></code> </p>
        </td>
        <td><p>Color to + scatterplot helper</p></td>
      </tr><tr><td>
          <p><code><a href="hlpr_smooth.html">hlpr_smooth()</a></code> </p>
        </td>
        <td><p>Smooth variables spatially</p></td>
      </tr><tr><td>
          <p><code><a href="hlpr_smooth_shiny.html">hlpr_smooth_shiny()</a></code> </p>
        </td>
        <td><p>Smooth variable spatially in mini-shiny-apps</p></td>
      </tr><tr><td>
          <p><code><a href="hlpr_subset_across.html">hlpr_subset_across()</a></code> </p>
        </td>
        <td><p>Subset the across-variables</p></td>
      </tr><tr><td>
          <p><code><a href="hlpr_summarize_trajectory_df.html">hlpr_summarize_trajectory_df()</a></code> </p>
        </td>
        <td><p>Join and summarize compiled trajectory data.frames</p></td>
      </tr><tr><td>
          <p><code><a href="hlpr_vector_projection.html">hlpr_vector_projection()</a></code> </p>
        </td>
        <td><p>Perform vector projection</p></td>
      </tr><tr><td>
          <p><code><a href="hlpr_widen_trajectory_df.html">hlpr_widen_trajectory_df()</a></code> </p>
        </td>
        <td><p>Widen trajectory data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="hspaClusterGenePatterns.html">hspaClusterGenePatterns()</a></code> </p>
        </td>
        <td><p>Construct hcluster-tree from correlated gene pattern similarities</p></td>
      </tr><tr><td>
          <p><code><a href="hspaCsrCutoff.html">hspaCsrCutoff()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="hspaCsrTesting.html">hspaCsrTesting()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="hspaDataBinarization.html">hspaDataBinarization()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="hspaPatternIdentification.html">hspaPatternIdentification()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="hspaPatternSimilarity.html">hspaPatternSimilarity()</a></code> <code><a href="hspaPatternSimilarity.html">hspaPatternSimilarity_future()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="hspaSelectGenes.html">hspaSelectGenes()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="identify_gene_patterns_dbscan.html">identify_gene_patterns_dbscan()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="image_dummy.html">image_dummy()</a></code> </p>
        </td>
        <td><p>image_dummy</p></td>
      </tr><tr><td>
          <p><code><a href="initiateSpataObject_10X.html">initiateSpataObject_10X()</a></code> </p>
        </td>
        <td><p>Initiate a spata-object from 10X Visium</p></td>
      </tr><tr><td>
          <p><code><a href="initiateSpataObject_CountMtr.html">initiateSpataObject_CountMtr()</a></code> </p>
        </td>
        <td><p>Initiate a spata-object from a raw count matrix</p></td>
      </tr><tr><td>
          <p><code><a href="initiateSpataObject_Empty.html">initiateSpataObject_Empty()</a></code> </p>
        </td>
        <td><p>Initiate an empty spata-object</p></td>
      </tr><tr><td>
          <p><code><a href="initiateSpataObject_Examples.html">initiateSpataObject_Examples()</a></code> </p>
        </td>
        <td><p>Initiate a spata-object from example data sets</p></td>
      </tr><tr><td>
          <p><code><a href="initiateSpataObject_ExprMtr.html">initiateSpataObject_ExprMtr()</a></code> </p>
        </td>
        <td><p>Initiate spata object from scaled expression matrix</p></td>
      </tr><tr><td>
          <p><code><a href="is_subsetted_by_segment.html">is_subsetted_by_segment()</a></code> </p>
        </td>
        <td><p>Check if object has been subsetted by segment</p></td>
      </tr><tr><td>
          <p><code><a href="joinWith.html">joinWith()</a></code> <code><a href="joinWith.html">joinWithFeatures()</a></code> <code><a href="joinWith.html">joinWithGenes()</a></code> <code><a href="joinWith.html">joinWithGeneSets()</a></code> <code><a href="joinWith.html">joinWithVariables()</a></code> </p>
        </td>
        <td><p>Join barcodes with additional variables</p></td>
      </tr><tr><td>
          <p><code><a href="joinWithPca.html">joinWithPca()</a></code> <code><a href="joinWithPca.html">joinWithTsne()</a></code> <code><a href="joinWithPca.html">joinWithUmap()</a></code> </p>
        </td>
        <td><p>Join barcodes with additional variables</p></td>
      </tr><tr><td>
          <p><code><a href="lazy_check_dummy.html">lazy_check_dummy()</a></code> </p>
        </td>
        <td><p>lazy_check_dummy</p></td>
      </tr><tr><td>
          <p><code><a href="legendBottom.html">legendBottom()</a></code> <code><a href="legendBottom.html">legendNone()</a></code> <code><a href="legendBottom.html">legendRight()</a></code> <code><a href="legendBottom.html">legendTop()</a></code> </p>
        </td>
        <td><p>Ggplot add on wrapper</p></td>
      </tr><tr><td>
          <p><code><a href="loadGeneSetDf.html">loadGeneSetDf()</a></code> </p>
        </td>
        <td><p>Load gene set data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="loadGSDF.html">loadGSDF()</a></code> </p>
        </td>
        <td><p>Original load gene set data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="loadSpataObject.html">loadSpataObject()</a></code> <code><a href="loadSpataObject.html">loadCorrespondingCDS()</a></code> <code><a href="loadSpataObject.html">loadCorrespondingSeuratObject()</a></code> </p>
        </td>
        <td><p>Load corresponding objects</p></td>
      </tr><tr><td>
          <p><code><a href="mergeSpataObjects.html">mergeSpataObjects()</a></code> </p>
        </td>
        <td><p>Merge Spata Objects</p></td>
      </tr><tr><td>
          <p><code><a href="method_hclust.html">method_hclust()</a></code> </p>
        </td>
        <td><p>method_hclust</p></td>
      </tr><tr><td>
          <p><code><a href="moduleAddGeneSetsServer.html">moduleAddGeneSetsServer()</a></code> </p>
        </td>
        <td><p>Server of the add gene sets module</p></td>
      </tr><tr><td>
          <p><code><a href="moduleAddGeneSetsUI.html">moduleAddGeneSetsUI()</a></code> </p>
        </td>
        <td><p>UI of the add gene sets module</p></td>
      </tr><tr><td>
          <p><code><a href="moduleSurfacePlotServer.html">moduleSurfacePlotServer()</a></code> </p>
        </td>
        <td><p>Server of the surface plot module</p></td>
      </tr><tr><td>
          <p><code><a href="moduleSurfacePlotUI.html">moduleSurfacePlotUI()</a></code> </p>
        </td>
        <td><p>UI of the surface plot module</p></td>
      </tr><tr><td>
          <p><code><a href="nGenes.html">nGenes()</a></code> </p>
        </td>
        <td><p>Number of genes</p></td>
      </tr><tr><td>
          <p><code><a href="normalize.html">normalize()</a></code> </p>
        </td>
        <td><p>normalize</p></td>
      </tr><tr><td>
          <p><code><a href="pb_dummy.html">pb_dummy()</a></code> </p>
        </td>
        <td><p>pb</p></td>
      </tr><tr><td>
          <p><code><a href="pheatmap.html">pheatmap()</a></code> </p>
        </td>
        <td><p>de_df</p></td>
      </tr><tr><td>
          <p><code><a href="plotAutoencoderAssessment.html">plotAutoencoderAssessment()</a></code> </p>
        </td>
        <td><p>Plot total variance of different neural networks</p></td>
      </tr><tr><td>
          <p><code><a href="plotAutoencoderResults.html">plotAutoencoderResults()</a></code> </p>
        </td>
        <td><p>Plot scaled vs. denoised expression</p></td>
      </tr><tr><td>
          <p><code><a href="plotBarchart.html">plotBarchart()</a></code> </p>
        </td>
        <td><p>Plot distribution of discrete variables</p></td>
      </tr><tr><td>
          <p><code><a href="plotBoxplot.html">plotBoxplot()</a></code> <code><a href="plotBoxplot.html">plotDensityplot()</a></code> <code><a href="plotBoxplot.html">plotHistogram()</a></code> <code><a href="plotBoxplot.html">plotRidgeplot()</a></code> <code><a href="plotBoxplot.html">plotViolinplot()</a></code> </p>
        </td>
        <td><p>Plot numeric distribution and statistical tests</p></td>
      </tr><tr><td>
          <p><code><a href="plotCnvResults.html">plotCnvResults()</a></code> </p>
        </td>
        <td><p>Visualize copy-number-variations results</p></td>
      </tr><tr><td>
          <p><code><a href="plotCsrCutoffSimulations.html">plotCsrCutoffSimulations()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="plotCsrResults.html">plotCsrResults()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="plotCustomizedTrajectoryTrends.html">plotCustomizedTrajectoryTrends()</a></code> </p>
        </td>
        <td><p>Plot customized trajectory trends</p></td>
      </tr><tr><td>
          <p><code><a href="plotDeaDotPlot.html">plotDeaDotPlot()</a></code> </p>
        </td>
        <td><p>Plot differentially expressed genes</p></td>
      </tr><tr><td>
          <p><code><a href="plotDeaGeneCount.html">plotDeaGeneCount()</a></code> </p>
        </td>
        <td><p>Plot the number of differently expressed genes of certain groups</p></td>
      </tr><tr><td>
          <p><code><a href="plotDeaHeatmap.html">plotDeaHeatmap()</a></code> </p>
        </td>
        <td><p>Plot differentially expressed genes</p></td>
      </tr><tr><td>
          <p><code><a href="plotDeaPvalues.html">plotDeaPvalues()</a></code> <code><a href="plotDeaPvalues.html">plotDeaLogFC()</a></code> </p>
        </td>
        <td><p>Plot the p-value and log fold change distribution of de-analysis results</p></td>
      </tr><tr><td>
          <p><code><a href="plotDeaSummary.html">plotDeaSummary()</a></code> </p>
        </td>
        <td><p>Plot a summary of differential expression analysis results</p></td>
      </tr><tr><td>
          <p><code><a href="plotDendrogram.html">plotDendrogram()</a></code> <code><a href="plotDendrogram.html">plotDendrogramCnv()</a></code> </p>
        </td>
        <td><p>Visualize hierarchical clustering results</p></td>
      </tr><tr><td>
          <p><code><a href="plotDistribution.html">plotDistribution()</a></code> <code><a href="plotDistribution.html">plotDistribution2()</a></code> <code><a href="plotDistribution.html">plotDistributionAcross()</a></code> <code><a href="plotDistribution.html">plotDistributionAcross2()</a></code> <code><a href="plotDistribution.html">plotDistributionDiscrete()</a></code> </p>
        </td>
        <td><p>Distribution of continuous values (Deprecated)</p></td>
      </tr><tr><td>
          <p><code><a href="plotFourStates.html">plotFourStates()</a></code> <code><a href="plotFourStates.html">plotFourStates2()</a></code> </p>
        </td>
        <td><p>Gene set state plot</p></td>
      </tr><tr><td>
          <p><code><a href="plotGeneDendrogram.html">plotGeneDendrogram()</a></code> </p>
        </td>
        <td><p>Visualize clustering results</p></td>
      </tr><tr><td>
          <p><code><a href="plotGeneMetaData.html">plotGeneMetaData()</a></code> </p>
        </td>
        <td><p>Plot gene meta data</p></td>
      </tr><tr><td>
          <p><code><a href="plotGenePatternBinarized.html">plotGenePatternBinarized()</a></code> <code><a href="plotGenePatternBinarized.html">plotGenePatternDenoised()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="plotGenePatternCluster.html">plotGenePatternCluster()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="plotGenePatternDendrogram.html">plotGenePatternDendrogram()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="plotGenePatternHeatmap.html">plotGenePatternHeatmap()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="plotGenePatterns.html">plotGenePatterns()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="plotGseaDotPlot.html">plotGseaDotPlot()</a></code> </p>
        </td>
        <td><p>Plot gene set enrichment</p></td>
      </tr><tr><td>
          <p><code><a href="plotHspaSummary.html">plotHspaSummary()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="plotPcaVariation.html">plotPcaVariation()</a></code> </p>
        </td>
        <td><p>Plot Pca Variation</p></td>
      </tr><tr><td>
          <p><code><a href="plotPseudotime.html">plotPseudotime()</a></code> </p>
        </td>
        <td><p>Monocle3 Pseudotime</p></td>
      </tr><tr><td>
          <p><code><a href="plotScatterplot.html">plotScatterplot()</a></code> </p>
        </td>
        <td><p>Plot numeric variables as a scatterplot</p></td>
      </tr><tr><td>
          <p><code><a href="plotSegmentation.html">plotSegmentation()</a></code> </p>
        </td>
        <td><p>Plot segmentation</p></td>
      </tr><tr><td>
          <p><code><a href="plotStatisticsInteractive.html">plotStatisticsInteractive()</a></code> </p>
        </td>
        <td><p>Plot distribution of variables interactively</p></td>
      </tr><tr><td>
          <p><code><a href="plotSurface.html">plotSurface()</a></code> <code><a href="plotSurface.html">plotSurface2()</a></code> <code><a href="plotSurface.html">plotSurfaceInteractive()</a></code> </p>
        </td>
        <td><p>Plot the surface of the sample</p></td>
      </tr><tr><td>
          <p><code><a href="plotSurfaceAverage.html">plotSurfaceAverage()</a></code> </p>
        </td>
        <td><p>Plot several surface plots colored by gene averages</p></td>
      </tr><tr><td>
          <p><code><a href="plotSurfaceComparison.html">plotSurfaceComparison()</a></code> <code><a href="plotSurfaceComparison.html">plotSurfaceComparison2()</a></code> </p>
        </td>
        <td><p>Plot several surface plots colored by numeric variables</p></td>
      </tr><tr><td>
          <p><code><a href="plotSurfaceHotspots.html">plotSurfaceHotspots()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="plotSurfaceQuantiles.html">plotSurfaceQuantiles()</a></code> </p>
        </td>
        <td><p>Plot a surface plot colored by binned numeric variables</p></td>
      </tr><tr><td>
          <p><code><a href="plotTrajectory.html">plotTrajectory()</a></code> </p>
        </td>
        <td><p>Plot trajectory</p></td>
      </tr><tr><td>
          <p><code><a href="plotTrajectoryFeatures.html">plotTrajectoryFeatures()</a></code> <code><a href="plotTrajectoryFeatures.html">plotTrajectoryGenes()</a></code> <code><a href="plotTrajectoryFeatures.html">plotTrajectoryGeneSets()</a></code> </p>
        </td>
        <td><p>Plot continuous trajectory dynamics in lineplots</p></td>
      </tr><tr><td>
          <p><code><a href="plotTrajectoryFeaturesDiscrete.html">plotTrajectoryFeaturesDiscrete()</a></code> </p>
        </td>
        <td><p>Plot discrete trajectory dynamics</p></td>
      </tr><tr><td>
          <p><code><a href="plotTrajectoryFit.html">plotTrajectoryFit()</a></code> <code><a href="plotTrajectoryFit.html">plotTrajectoryFitCustomized()</a></code> </p>
        </td>
        <td><p>Plot trajectory fit</p></td>
      </tr><tr><td>
          <p><code><a href="plotTrajectoryHeatmap.html">plotTrajectoryHeatmap()</a></code> </p>
        </td>
        <td><p>Plot trajectory expression dynamic in heatmap</p></td>
      </tr><tr><td>
          <p><code><a href="plotUmap.html">plotUmap()</a></code> <code><a href="plotUmap.html">plotTsne()</a></code> <code><a href="plotUmap.html">plotPca()</a></code> </p>
        </td>
        <td><p>Plot dimensional reduction</p></td>
      </tr><tr><td>
          <p><code><a href="plot_type_dummy.html">plot_type_dummy()</a></code> </p>
        </td>
        <td><p>plot_type</p></td>
      </tr><tr><td>
          <p><code><a href="printAutoencoderSummary.html">printAutoencoderSummary()</a></code> </p>
        </td>
        <td><p>Print autoencoder summary</p></td>
      </tr><tr><td>
          <p><code><a href="printDeaOverview.html">printDeaOverview()</a></code> </p>
        </td>
        <td><p>Print overview of all conducted de-analysis</p></td>
      </tr><tr><td>
          <p><code><a href="printDefaultInstructions.html">printDefaultInstructions()</a></code> </p>
        </td>
        <td><p>Print current default settings</p></td>
      </tr><tr><td>
          <p><code><a href="printGeneSetOverview.html">printGeneSetOverview()</a></code> </p>
        </td>
        <td><p>Print overview about the current gene sets</p></td>
      </tr><tr><td>
          <p><code><a href="print_family.html">print_family()</a></code> </p>
        </td>
        <td><p>print</p></td>
      </tr><tr><td>
          <p><code><a href="process_seurat_object.html">process_seurat_object()</a></code> </p>
        </td>
        <td><p>Wrapper around Seurat processing functions</p></td>
      </tr><tr><td>
          <p><code><a href="renameFeatures.html">renameFeatures()</a></code> </p>
        </td>
        <td><p>Rename features</p></td>
      </tr><tr><td>
          <p><code><a href="renameGroups.html">renameGroups()</a></code> <code><a href="renameGroups.html">renameSegments()</a></code> </p>
        </td>
        <td><p>Rename cluster/group names</p></td>
      </tr><tr><td>
          <p><code><a href="runAutoencoderAssessment.html">runAutoencoderAssessment()</a></code> <code><a href="runAutoencoderAssessment.html">assessAutoencoderOptions()</a></code> </p>
        </td>
        <td><p>Assessment of Neural Network Set Up</p></td>
      </tr><tr><td>
          <p><code><a href="runAutoencoderDenoising.html">runAutoencoderDenoising()</a></code> </p>
        </td>
        <td><p>Denoise expression matrix</p></td>
      </tr><tr><td>
          <p><code><a href="runCnvAnalysis.html">runCnvAnalysis()</a></code> </p>
        </td>
        <td><p>Identify large-scale chromosomal copy number variations</p></td>
      </tr><tr><td>
          <p><code><a href="runDeAnalysis.html">runDeAnalysis()</a></code> </p>
        </td>
        <td><p>Find differently expressed genes</p></td>
      </tr><tr><td>
          <p><code><a href="runGSEA.html">runGSEA()</a></code> </p>
        </td>
        <td><p>Compute gene set enrichment</p></td>
      </tr><tr><td>
          <p><code><a href="runHclust.html">runHclust()</a></code> <code><a href="runHclust.html">runHclustCnv()</a></code> </p>
        </td>
        <td><p>Run hierarchical clustering</p></td>
      </tr><tr><td>
          <p><code><a href="runHspa.html">runHspa()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="runPca.html">runPca()</a></code> <code><a href="runPca.html">runPca2()</a></code> </p>
        </td>
        <td><p>Run Principal Component Analysis</p></td>
      </tr><tr><td>
          <p><code><a href="runSpatialCorrelationAnalysis.html">runSpatialCorrelationAnalysis()</a></code> </p>
        </td>
        <td><p>Initiate gene clustering analysis based on spatial patterns</p></td>
      </tr><tr><td>
          <p><code><a href="runTsne.html">runTsne()</a></code> <code><a href="runTsne.html">runTsne2()</a></code> </p>
        </td>
        <td><p>Run t-Stochastic Neighbour Embedding</p></td>
      </tr><tr><td>
          <p><code><a href="runUmap.html">runUmap()</a></code> <code><a href="runUmap.html">runUmap2()</a></code> </p>
        </td>
        <td><p>Run UMAP-Analysis</p></td>
      </tr><tr><td>
          <p><code><a href="sample_name.html">sample_name()</a></code> </p>
        </td>
        <td><p>sample_name</p></td>
      </tr><tr><td>
          <p><code><a href="saveGeneSetDf.html">saveGeneSetDf()</a></code> </p>
        </td>
        <td><p>Save a gene set data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="saveSpataObject.html">saveSpataObject()</a></code> <code><a href="saveSpataObject.html">saveCorrespondingCDS()</a></code> <code><a href="saveSpataObject.html">saveCorrespondingSeuratObject()</a></code> </p>
        </td>
        <td><p>Save corresponding objects</p></td>
      </tr><tr><td>
          <p><code><a href="scale_color_add_on.html">scale_color_add_on()</a></code> </p>
        </td>
        <td><p>ggplot2 - color add on</p></td>
      </tr><tr><td>
          <p><code><a href="setActiveExpressionMatrix.html">setActiveExpressionMatrix()</a></code> </p>
        </td>
        <td><p>Denote the default expression matrix</p></td>
      </tr><tr><td>
          <p><code><a href="setAutoencoderAssessment.html">setAutoencoderAssessment()</a></code> </p>
        </td>
        <td><p>Set results of autoencoder assessment</p></td>
      </tr><tr><td>
          <p><code><a href="setCnvResults.html">setCnvResults()</a></code> </p>
        </td>
        <td><p>Set cnv-results</p></td>
      </tr><tr><td>
          <p><code><a href="setCoordsDf.html">setCoordsDf()</a></code> </p>
        </td>
        <td><p>Set the coordinates</p></td>
      </tr><tr><td>
          <p><code><a href="setCountMatrix.html">setCountMatrix()</a></code> <code><a href="setCountMatrix.html">setDenoisedMatrix()</a></code> <code><a href="setCountMatrix.html">setScaledMatrix()</a></code> </p>
        </td>
        <td><p>Set data matrices</p></td>
      </tr><tr><td>
          <p><code><a href="setCsrCutoff.html">setCsrCutoff()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="setDefaultInstructions.html">setDefaultInstructions()</a></code> <code><a href="setDefaultInstructions.html">setDirectoryInstructions()</a></code> </p>
        </td>
        <td><p>Set default instructions</p></td>
      </tr><tr><td>
          <p><code><a href="setFeatureDf.html">setFeatureDf()</a></code> </p>
        </td>
        <td><p>Set feature data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="setGeneSetDf.html">setGeneSetDf()</a></code> </p>
        </td>
        <td><p>Set the gene-set data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="setHspaResults.html">setHspaResults()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="setInitiationInfo.html">setInitiationInfo()</a></code> </p>
        </td>
        <td><p>Set initiation information</p></td>
      </tr><tr><td>
          <p><code><a href="setPcaDf.html">setPcaDf()</a></code> <code><a href="setPcaDf.html">setTsneDf()</a></code> <code><a href="setPcaDf.html">setUmapDf()</a></code> </p>
        </td>
        <td><p>Set dimensional reductions data.frames</p></td>
      </tr><tr><td>
          <p><code><a href="setPrResults.html">setPrResults()</a></code> </p>
        </td>
        <td><p>Set results of pattern recognition methods</p></td>
      </tr><tr><td>
          <p><code><a href="setSpCorResults.html">setSpCorResults()</a></code> </p>
        </td>
        <td><p>Set results of spatial correlation analysis</p></td>
      </tr><tr><td>
          <p><code><a href="set_dummy.html">set_dummy()</a></code> </p>
        </td>
        <td><p>set</p></td>
      </tr><tr><td>
          <p><code><a href="seurat_object_dummy.html">seurat_object_dummy()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="shiftTrajectoryDf.html">shiftTrajectoryDf()</a></code> </p>
        </td>
        <td><p>Shift trajectory data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="smoothSpatially.html">smoothSpatially()</a></code> </p>
        </td>
        <td><p>Smooth numeric variables spatially</p></td>
      </tr><tr><td>
          <p><code><a href="spata2-class.html">spata2-class</a></code> </p>
        </td>
        <td><p>The spata-object</p></td>
      </tr><tr><td>
          <p><code><a href="spatial_trajectory-class.html">spatial_trajectory-class</a></code> </p>
        </td>
        <td><p>spatial_trajectory object</p></td>
      </tr><tr><td>
          <p><code><a href="subsetBySegment_CountMtr.html">subsetBySegment_CountMtr()</a></code> <code><a href="subsetBySegment_CountMtr.html">subsetByBarcodes_CountMtr()</a></code> <code><a href="subsetBySegment_CountMtr.html">subsetBySegment_ExprMtr()</a></code> <code><a href="subsetBySegment_CountMtr.html">subsetByBarcodes_ExprMtr()</a></code> </p>
        </td>
        <td><p>Subset a spata-object</p></td>
      </tr><tr><td>
          <p><code><a href="surfaceFrame.html">surfaceFrame()</a></code> </p>
        </td>
        <td><p>Add frame of the sample</p></td>
      </tr><tr><td>
          <p><code><a href="tab_create_segmentation_return.html">tab_create_segmentation_return()</a></code> </p>
        </td>
        <td><p>Segmentation plot tab - return</p></td>
      </tr><tr><td>
          <p><code><a href="tab_create_trajectories_return.html">tab_create_trajectories_return()</a></code> </p>
        </td>
        <td><p>Trajectory plot tab - return</p></td>
      </tr><tr><td>
          <p><code><a href="tab_surface_plots_app.html">tab_surface_plots_app()</a></code> </p>
        </td>
        <td><p>Surface plot tab - classic</p></td>
      </tr><tr><td>
          <p><code><a href="tab_surface_plots_return.html">tab_surface_plots_return()</a></code> </p>
        </td>
        <td><p>Surface plot tab - return</p></td>
      </tr><tr><td>
          <p><code><a href="trajectory_patterns.html">trajectory_patterns</a></code> </p>
        </td>
        <td><p>Trajectory patterns</p></td>
      </tr><tr><td>
          <p><code><a href="transformSeuratToSpata.html">transformSeuratToSpata()</a></code> </p>
        </td>
        <td><p>Transform seurat-object to spata-object</p></td>
      </tr><tr><td>
          <p><code><a href="transformSpataToCDS.html">transformSpataToCDS()</a></code> </p>
        </td>
        <td><p>Transform spata-object to cell-data-set (Monocle3)</p></td>
      </tr><tr><td>
          <p><code><a href="transformSpataToSeurat.html">transformSpataToSeurat()</a></code> </p>
        </td>
        <td><p>Transform spata-object to a seurat-object</p></td>
      </tr><tr><td>
          <p><code><a href="updateSpataObject.html">updateSpataObject()</a></code> </p>
        </td>
        <td><p>Update spata-object from SPATA to SPATA2</p></td>
      </tr><tr><td>
          <p><code><a href="validActivationFunctions.html">validActivationFunctions()</a></code> <code><a href="validActivationFunctions.html">validAgglomerationMethods()</a></code> <code><a href="validActivationFunctions.html">validColorPalettes()</a></code> <code><a href="validActivationFunctions.html">validColorSpectra()</a></code> <code><a href="validActivationFunctions.html">validDeAnalysisMethods()</a></code> <code><a href="validActivationFunctions.html">validDefaultInstructionSlots()</a></code> <code><a href="validActivationFunctions.html">validDimRedMethods()</a></code> <code><a href="validActivationFunctions.html">validDirectoryInstructionSlots()</a></code> <code><a href="validActivationFunctions.html">validDistanceMethods()</a></code> <code><a href="validActivationFunctions.html">validHierarchicalClusterMethods()</a></code> <code><a href="validActivationFunctions.html">validPatternRecognitionMethods()</a></code> <code><a href="validActivationFunctions.html">validPadjMethods()</a></code> <code><a href="validActivationFunctions.html">validPlotTypes()</a></code> </p>
        </td>
        <td><p>Obtain valid argument inputs</p></td>
      </tr><tr><td>
          <p><code><a href="validateSpataObject.html">validateSpataObject()</a></code> </p>
        </td>
        <td><p>Validate a spata object</p></td>
      </tr><tr><td>
          <p><code><a href="validation.html">validation()</a></code> </p>
        </td>
        <td><p>Validate object input</p></td>
      </tr><tr><td>
          <p><code><a href="variable.html">variable()</a></code> </p>
        </td>
        <td><p>variable</p></td>
      </tr><tr><td>
          <p><code><a href="variables_num.html">variables_num()</a></code> </p>
        </td>
        <td><p>variables_num</p></td>
      </tr><tr><td>
          <p><code><a href="variable_num.html">variable_num()</a></code> </p>
        </td>
        <td><p>variable_num</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="rename-content">Rename Content <a href="#rename-content" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p><p>The rename-functions let you adjust the name of certain content with which you refer to it via arguments</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="renameFeatures.html">renameFeatures()</a></code> </p>
        </td>
        <td><p>Rename features</p></td>
      </tr><tr><td>
          <p><code><a href="renameGroups.html">renameGroups()</a></code> <code><a href="renameGroups.html">renameSegments()</a></code> </p>
        </td>
        <td><p>Rename cluster/group names</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="extract-data">Extract Data <a href="#extract-data" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>SPATA invites you to extract your analysis. get-functions let you access every imaginable aspect of information conveniently.</p>
        </th>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="dimensional-reduction-and-spatial-coordinates">Dimensional Reduction and Spatial Coordinates <a href="#dimensional-reduction-and-spatial-coordinates" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="getCoordsDf.html">getCoordsDf()</a></code> <code><a href="getCoordsDf.html">getSegmentDf()</a></code> </p>
        </td>
        <td><p>Obtain spatial coordinates</p></td>
      </tr><tr><td>
          <p><code><a href="getSegmentNames.html">getSegmentNames()</a></code> </p>
        </td>
        <td><p>Obtain segment names</p></td>
      </tr><tr><td>
          <p><code><a href="getDimRedDf.html">getDimRedDf()</a></code> <code><a href="getDimRedDf.html">getPcaDf()</a></code> <code><a href="getDimRedDf.html">getPcaMtr()</a></code> <code><a href="getDimRedDf.html">getUmapDf()</a></code> <code><a href="getDimRedDf.html">getTsneDf()</a></code> </p>
        </td>
        <td><p>Obtain dimensional reduction data</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="features">Features <a href="#features" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="getFeatureNames.html">getFeatureNames()</a></code> </p>
        </td>
        <td><p>Obtain feature names</p></td>
      </tr><tr><td>
          <p><code><a href="getCnvFeatureNames.html">getCnvFeatureNames()</a></code> </p>
        </td>
        <td><p>Obtain features names under which cnv-analysis results are stored.</p></td>
      </tr><tr><td>
          <p><code><a href="getFeatureDf.html">getFeatureDf()</a></code> </p>
        </td>
        <td><p>Obtain feature data</p></td>
      </tr><tr><td>
          <p><code><a href="getFeatureValues.html">getFeatureValues()</a></code> </p>
        </td>
        <td><p>Obtain unique categorical feature values</p></td>
      </tr><tr><td>
          <p><code><a href="getFeatureVariables.html">getFeatureVariables()</a></code> </p>
        </td>
        <td><p>Obtain a feature variable</p></td>
      </tr><tr><td>
          <p><code><a href="getGroupingOptions.html">getGroupingOptions()</a></code> </p>
        </td>
        <td><p>Obtain variable names that group the barcode spots</p></td>
      </tr><tr><td>
          <p><code><a href="getGroupNames.html">getGroupNames()</a></code> </p>
        </td>
        <td><p>Obtain group names a grouping variable contains</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="genes-gene-sets">Genes &amp; Gene-sets <a href="#genes-gene-sets" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="getGeneCounts.html">getGeneCounts()</a></code> </p>
        </td>
        <td><p>Obtain total number of gene counts</p></td>
      </tr><tr><td>
          <p><code><a href="getGeneDistMtr.html">getGeneDistMtr()</a></code> </p>
        </td>
        <td><p>Obtain distance measurements of spatially correlated genes</p></td>
      </tr><tr><td>
          <p><code><a href="getGeneFeatureNames.html">getGeneFeatureNames()</a></code> </p>
        </td>
        <td><p>Obtain feature names of the gene meta data</p></td>
      </tr><tr><td>
          <p><code><a href="getGeneMetaData.html">getGeneMetaData()</a></code> <code><a href="getGeneMetaData.html">getGeneMetaDf()</a></code> </p>
        </td>
        <td><p>Obtain gene meta data</p></td>
      </tr><tr><td>
          <p><code><a href="getGenePatternCluster.html">getGenePatternCluster()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="getGenePatternCorrelations.html">getGenePatternCorrelations()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="getGenePatternDistances.html">getGenePatternDistances()</a></code> <code><a href="getGenePatternDistances.html">getGenePatternSimilarities()</a></code> </p>
        </td>
        <td><p>Obtain similarity scores between identified gene patterns</p></td>
      </tr><tr><td>
          <p><code><a href="getGenePatternEvalDf.html">getGenePatternEvalDf()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="getGenePatternExtentDf.html">getGenePatternExtentDf()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="getGenePatternsAboveThreshold.html">getGenePatternsAboveThreshold()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="getGenes.html">getGenes()</a></code> <code><a href="getGenes.html">getGenesInteractive()</a></code> </p>
        </td>
        <td><p>Obtain gene names</p></td>
      </tr><tr><td>
          <p><code><a href="getGeneSetDf.html">getGeneSetDf()</a></code> </p>
        </td>
        <td><p>Obtain gene set data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="getGeneSetOverview.html">getGeneSetOverview()</a></code> </p>
        </td>
        <td><p>Overview about the current gene sets</p></td>
      </tr><tr><td>
          <p><code><a href="getGeneSets.html">getGeneSets()</a></code> <code><a href="getGeneSets.html">getGeneSetsInteractive()</a></code> </p>
        </td>
        <td><p>Obtain gene set names</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="data-matrices">Data matrices <a href="#data-matrices" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="getBarcodes.html">getBarcodes()</a></code> </p>
        </td>
        <td><p>Obtain specific barcodes</p></td>
      </tr><tr><td>
          <p><code><a href="getActiveMatrixName.html">getActiveMatrixName()</a></code> </p>
        </td>
        <td><p>Obtain name of currently active expression matrix</p></td>
      </tr><tr><td>
          <p><code><a href="getMatrix.html">getMatrix()</a></code> <code><a href="getMatrix.html">getExpressionMatrix()</a></code> <code><a href="getMatrix.html">getCountMatrix()</a></code> </p>
        </td>
        <td><p>Obtain count and expression matrix</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="spatial-trajectories">Spatial Trajectories <a href="#spatial-trajectories" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="getTrajectoryComment-spata2-method.html">getTrajectoryComment(<i>&lt;spata2&gt;</i>)</a></code> </p>
        </td>
        <td><p>Obtain a trajectory comment</p></td>
      </tr><tr><td>
          <p><code><a href="getTrajectoryComment.html">getTrajectoryComment()</a></code> </p>
        </td>
        <td><p>Obtain a trajectory comment</p></td>
      </tr><tr><td>
          <p><code><a href="getTrajectoryDf.html">getTrajectoryDf()</a></code> </p>
        </td>
        <td><p>Obtain a summarized trajectory data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="getTrajectoryLength.html">getTrajectoryLength()</a></code> </p>
        </td>
        <td><p>Obtain the length of a trajectory</p></td>
      </tr><tr><td>
          <p><code><a href="getTrajectoryNames.html">getTrajectoryNames()</a></code> </p>
        </td>
        <td><p>Obtain trajectory names</p></td>
      </tr><tr><td>
          <p><code><a href="getTrajectoryObject.html">getTrajectoryObject()</a></code> </p>
        </td>
        <td><p>Obtain trajectory object</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="miscellaneous">Miscellaneous <a href="#miscellaneous" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="getAutoencoderAssessment.html">getAutoencoderAssessment()</a></code> </p>
        </td>
        <td><p>Obtain information about the optimal neural network set up</p></td>
      </tr><tr><td>
          <p><code><a href="getAutoencoderSetUp.html">getAutoencoderSetUp()</a></code> </p>
        </td>
        <td><p>Obtain information on neural network</p></td>
      </tr><tr><td>
          <p><code><a href="getCnvResults.html">getCnvResults()</a></code> </p>
        </td>
        <td><p>Obtain copy-number-variations results</p></td>
      </tr><tr><td>
          <p><code><a href="getDeaOverview.html">getDeaOverview()</a></code> </p>
        </td>
        <td><p>Obtain info on de-analysis storage</p></td>
      </tr><tr><td>
          <p><code><a href="getDeaResultsDf.html">getDeaResultsDf()</a></code> <code><a href="getDeaResultsDf.html">getDeaGenes()</a></code> </p>
        </td>
        <td><p>Obtain de-analysis results</p></td>
      </tr><tr><td>
          <p><code><a href="getDefaultInstructions.html">getDefaultInstructions()</a></code> <code><a href="getDefaultInstructions.html">getDirectoryInstructions()</a></code> </p>
        </td>
        <td><p>Obtain default argument inputs</p></td>
      </tr><tr><td>
          <p><code><a href="getGseaDf.html">getGseaDf()</a></code> </p>
        </td>
        <td><p>Obtain enrichment data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="getGseaResults.html">getGseaResults()</a></code> </p>
        </td>
        <td><p>Obtain enrichment results</p></td>
      </tr><tr><td>
          <p><code><a href="getImage.html">getImage()</a></code> </p>
        </td>
        <td><p>Obtain histology image</p></td>
      </tr><tr><td>
          <p><code><a href="getInitiationInfo.html">getInitiationInfo()</a></code> <code><a href="getInitiationInfo.html">getInitiationInput()</a></code> </p>
        </td>
        <td><p>Obtain information about object initiation</p></td>
      </tr><tr><td>
          <p><code><a href="getSampleNames.html">getSampleNames()</a></code> </p>
        </td>
        <td><p>Obtain sample names</p></td>
      </tr><tr><td>
          <p><code><a href="getSpataDf.html">getSpataDf()</a></code> </p>
        </td>
        <td><p>Obtain a spata-data.frame</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="information-summaries">Information &amp; Summaries <a href="#information-summaries" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Some handy functions that quickly give you an overwiew about the spata-object’s content and the progress you have made.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="printAutoencoderSummary.html">printAutoencoderSummary()</a></code> </p>
        </td>
        <td><p>Print autoencoder summary</p></td>
      </tr><tr><td>
          <p><code><a href="printDeaOverview.html">printDeaOverview()</a></code> </p>
        </td>
        <td><p>Print overview of all conducted de-analysis</p></td>
      </tr><tr><td>
          <p><code><a href="printDefaultInstructions.html">printDefaultInstructions()</a></code> </p>
        </td>
        <td><p>Print current default settings</p></td>
      </tr><tr><td>
          <p><code><a href="printGeneSetOverview.html">printGeneSetOverview()</a></code> </p>
        </td>
        <td><p>Print overview about the current gene sets</p></td>
      </tr><tr><td>
          <p><code><a href="print_family.html">print_family()</a></code> </p>
        </td>
        <td><p>print</p></td>
      </tr><tr><td>
          <p><code><a href="across.html">across()</a></code> <code><a href="across.html">across_dummy()</a></code> </p>
        </td>
        <td><p>across</p></td>
      </tr><tr><td>
          <p><code><a href="addAutoencoderSetUp.html">addAutoencoderSetUp()</a></code> </p>
        </td>
        <td><p>Add the set up of a neural network</p></td>
      </tr><tr><td>
          <p><code><a href="addExpressionMatrix.html">addExpressionMatrix()</a></code> </p>
        </td>
        <td><p>Add an expression matrix</p></td>
      </tr><tr><td>
          <p><code><a href="addFeatures.html">addFeatures()</a></code> </p>
        </td>
        <td><p>Add a new feature</p></td>
      </tr><tr><td>
          <p><code><a href="addGeneFeatures.html">addGeneFeatures()</a></code> </p>
        </td>
        <td><p>Add new gene features</p></td>
      </tr><tr><td>
          <p><code><a href="addGeneMetaData.html">addGeneMetaData()</a></code> </p>
        </td>
        <td><p>Add gene meta data to the object</p></td>
      </tr><tr><td>
          <p><code><a href="addGeneSet.html">addGeneSet()</a></code> <code><a href="addGeneSet.html">addGeneSetsInteractive()</a></code> </p>
        </td>
        <td><p>Add a new gene set</p></td>
      </tr><tr><td>
          <p><code><a href="addSpCorCluster.html">addSpCorCluster()</a></code> </p>
        </td>
        <td><p>Add cluster results of spatial correlation results</p></td>
      </tr><tr><td>
          <p><code><a href="adjustDefaultInstructions.html">adjustDefaultInstructions()</a></code> <code><a href="adjustDefaultInstructions.html">adjustDirectoryInstructions()</a></code> </p>
        </td>
        <td><p>Adjust default instructions</p></td>
      </tr><tr><td>
          <p><code><a href="adjustGeneSetDf.html">adjustGeneSetDf()</a></code> </p>
        </td>
        <td><p>Filter gene-set data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="adjusting_check_dummy.html">adjusting_check_dummy()</a></code> </p>
        </td>
        <td><p>This is a text dummy</p></td>
      </tr><tr><td>
          <p><code><a href="allColorpalettes.html">allColorpalettes()</a></code> <code><a href="allColorpalettes.html">allColorspectra()</a></code> </p>
        </td>
        <td><p>Valid colorpanels &amp; -spectra</p></td>
      </tr><tr><td>
          <p><code><a href="all_color_palettes.html">all_color_palettes()</a></code> <code><a href="all_color_palettes.html">all_color_spectra()</a></code> </p>
        </td>
        <td><p>Color palette names</p></td>
      </tr><tr><td>
          <p><code><a href="argument_dummy.html">argument_dummy()</a></code> </p>
        </td>
        <td><p>Argument dummy</p></td>
      </tr><tr><td>
          <p><code><a href="assessTrajectoryTrends.html">assessTrajectoryTrends()</a></code> <code><a href="assessTrajectoryTrends.html">assessTrajectoryTrends2()</a></code> <code><a href="assessTrajectoryTrends.html">assessTrajectoryTrendsCustomized()</a></code> <code><a href="assessTrajectoryTrends.html">assessTrajectoryTrendsCustomized2()</a></code> </p>
        </td>
        <td><p>Trajectory trend analysis</p></td>
      </tr><tr><td>
          <p><code><a href="average_genes.html">average_genes()</a></code> </p>
        </td>
        <td><p>average_genes</p></td>
      </tr><tr><td>
          <p><code><a href="binarize_matrix.html">binarize_matrix()</a></code> </p>
        </td>
        <td><p>Binarize expression matrix</p></td>
      </tr><tr><td>
          <p><code><a href="binwidth_dummy.html">binwidth_dummy()</a></code> </p>
        </td>
        <td><p>binwidth</p></td>
      </tr><tr><td>
          <p><code><a href="cds_dummy.html">cds_dummy()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="checkpoint.html">checkpoint()</a></code> </p>
        </td>
        <td><p>Shiny feedback messages</p></td>
      </tr><tr><td>
          <p><code><a href="check_all_barcodes.html">check_all_barcodes()</a></code> </p>
        </td>
        <td><p>Check barcodes of all matrices and data.frames</p></td>
      </tr><tr><td>
          <p><code><a href="check_assign.html">check_assign()</a></code> </p>
        </td>
        <td><p>Check assign input</p></td>
      </tr><tr><td>
          <p><code><a href="check_atdf.html">check_atdf()</a></code> <code><a href="check_atdf.html">check_rtdf()</a></code> </p>
        </td>
        <td><p>Check assessed trajectory data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="check_availability.html">check_availability()</a></code> </p>
        </td>
        <td><p>Gives feedback if object was not found</p></td>
      </tr><tr><td>
          <p><code><a href="check_color_to.html">check_color_to()</a></code> </p>
        </td>
        <td><p>Check color to</p></td>
      </tr><tr><td>
          <p><code><a href="check_compiled_trajectory_df.html">check_compiled_trajectory_df()</a></code> </p>
        </td>
        <td><p>Check compiled trajectory data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="check_coordinate_variables.html">check_coordinate_variables()</a></code> </p>
        </td>
        <td><p>Check coordinate variables</p></td>
      </tr><tr><td>
          <p><code><a href="check_coords_df.html">check_coords_df()</a></code> </p>
        </td>
        <td><p>Check coords data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="check_customized_trends.html">check_customized_trends()</a></code> </p>
        </td>
        <td><p>Check customized trend list</p></td>
      </tr><tr><td>
          <p><code><a href="check_dea_df.html">check_dea_df()</a></code> </p>
        </td>
        <td><p>Check de data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="check_display.html">check_display()</a></code> </p>
        </td>
        <td><p>Check display input</p></td>
      </tr><tr><td>
          <p><code><a href="check_features.html">check_features()</a></code> </p>
        </td>
        <td><p>Check feature variables input</p></td>
      </tr><tr><td>
          <p><code><a href="check_feature_df.html">check_feature_df()</a></code> </p>
        </td>
        <td><p>Check feature data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="check_genes.html">check_genes()</a></code> </p>
        </td>
        <td><p>Check gene input</p></td>
      </tr><tr><td>
          <p><code><a href="check_gene_sets.html">check_gene_sets()</a></code> </p>
        </td>
        <td><p>Check gene set input</p></td>
      </tr><tr><td>
          <p><code><a href="check_method.html">check_method()</a></code> </p>
        </td>
        <td><p>Check method input</p></td>
      </tr><tr><td>
          <p><code><a href="check_monocle_input.html">check_monocle_input()</a></code> </p>
        </td>
        <td><p>Check monocle input</p></td>
      </tr><tr><td>
          <p><code><a href="check_monocle_packages.html">check_monocle_packages()</a></code> </p>
        </td>
        <td><p>Makes sure that both packages are installed</p></td>
      </tr><tr><td>
          <p><code><a href="check_object.html">check_object()</a></code> </p>
        </td>
        <td><p>Check spata object input</p></td>
      </tr><tr><td>
          <p><code><a href="check_pattern.html">check_pattern()</a></code> </p>
        </td>
        <td><p>Check pattern input</p></td>
      </tr><tr><td>
          <p><code><a href="check_pt.html">check_pt()</a></code> </p>
        </td>
        <td><p>Check pt input</p></td>
      </tr><tr><td>
          <p><code><a href="check_sample.html">check_sample()</a></code> </p>
        </td>
        <td><p>Check sample input</p></td>
      </tr><tr><td>
          <p><code><a href="check_saving.html">check_saving()</a></code> </p>
        </td>
        <td><p>Check saving input</p></td>
      </tr><tr><td>
          <p><code><a href="check_segment.html">check_segment()</a></code> </p>
        </td>
        <td><p>Check segment name</p></td>
      </tr><tr><td>
          <p><code><a href="check_slot_coordinates.html">check_slot_coordinates()</a></code> <code><a href="check_slot_coordinates.html">check_slot_data()</a></code> <code><a href="check_slot_coordinates.html">check_slot_version()</a></code> <code><a href="check_slot_coordinates.html">check_slot_scvelo()</a></code> <code><a href="check_slot_coordinates.html">check_slot_trajectories()</a></code> <code><a href="check_slot_coordinates.html">check_trajectory_object()</a></code> <code><a href="check_slot_coordinates.html">check_slot_dim_red()</a></code> <code><a href="check_slot_coordinates.html">check_slot_fdata()</a></code> <code><a href="check_slot_coordinates.html">check_slot_image()</a></code> <code><a href="check_slot_coordinates.html">check_slot_used_genesets()</a></code> <code><a href="check_slot_coordinates.html">check_slot_samples()</a></code> </p>
        </td>
        <td><p>Check spata slots</p></td>
      </tr><tr><td>
          <p><code><a href="check_smooth.html">check_smooth()</a></code> </p>
        </td>
        <td><p>Check smooth input</p></td>
      </tr><tr><td>
          <p><code><a href="check_spata_df.html">check_spata_df()</a></code> </p>
        </td>
        <td><p>Check coordinate data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="check_stdf.html">check_stdf()</a></code> </p>
        </td>
        <td><p>Check summarized trajectory data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="check_summarized_trajectory_df.html">check_summarized_trajectory_df()</a></code> </p>
        </td>
        <td><p>Check stdf-input</p></td>
      </tr><tr><td>
          <p><code><a href="check_to.html">check_to()</a></code> </p>
        </td>
        <td><p>Gives feedback about input validity</p></td>
      </tr><tr><td>
          <p><code><a href="check_trajectory.html">check_trajectory()</a></code> </p>
        </td>
        <td><p>Check trajectory name input</p></td>
      </tr><tr><td>
          <p><code><a href="check_trajectory_binwidth.html">check_trajectory_binwidth()</a></code> </p>
        </td>
        <td><p>Check trajectory binwdith input</p></td>
      </tr><tr><td>
          <p><code><a href="check_uniform_genes.html">check_uniform_genes()</a></code> </p>
        </td>
        <td><p>Check uniform genes input</p></td>
      </tr><tr><td>
          <p><code><a href="check_variables.html">check_variables()</a></code> </p>
        </td>
        <td><p>Check variables</p></td>
      </tr><tr><td>
          <p><code><a href="clrp.html">clrp()</a></code> <code><a href="clrp.html">clrp_dummy()</a></code> </p>
        </td>
        <td><p>clrp</p></td>
      </tr><tr><td>
          <p><code><a href="clusterSpCorResults.html">clusterSpCorResults()</a></code> </p>
        </td>
        <td><p>Cluster genes according to their expression profile</p></td>
      </tr><tr><td>
          <p><code><a href="cnv_ref.html">cnv_ref</a></code> </p>
        </td>
        <td><p>A list of reference data for copy number variation analysis (CNV)</p></td>
      </tr><tr><td>
          <p><code><a href="cnv_regions_df.html">cnv_regions_df</a></code> </p>
        </td>
        <td><p>A data.frame necessary for cnv-analysis. Contains information about start and
end points of chromosomes.</p></td>
      </tr><tr><td>
          <p><code><a href="computeGeneMetaData.html">computeGeneMetaData()</a></code> <code><a href="computeGeneMetaData.html">computeGeneMetaData2()</a></code> </p>
        </td>
        <td><p>Compute gene summary statistics</p></td>
      </tr><tr><td>
          <p><code><a href="compute_pattern_similarity.html">compute_pattern_similarity()</a></code> <code><a href="compute_pattern_similarity.html">compute_pattern_relation()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="createSegmentation.html">createSegmentation()</a></code> </p>
        </td>
        <td><p>Spatial Segmentation</p></td>
      </tr><tr><td>
          <p><code><a href="createTrajectories.html">createTrajectories()</a></code> </p>
        </td>
        <td><p>Spatial Trajectories</p></td>
      </tr><tr><td>
          <p><code><a href="createTrajectoryManually.html">createTrajectoryManually()</a></code> </p>
        </td>
        <td><p>Create spatial trajectories manually</p></td>
      </tr><tr><td>
          <p><code><a href="data_counts-class.html">data_counts-class</a></code> </p>
        </td>
        <td><p>data_counts object</p></td>
      </tr><tr><td>
          <p><code><a href="default_instructions-class.html">default_instructions-class</a></code> </p>
        </td>
        <td><p>default instructions</p></td>
      </tr><tr><td>
          <p><code><a href="dim_red-class.html">dim_red-class</a></code> </p>
        </td>
        <td><p>dim_red object</p></td>
      </tr><tr><td>
          <p><code><a href="discardExpressionMatrix.html">discardExpressionMatrix()</a></code> </p>
        </td>
        <td><p>Discard an expression matrix</p></td>
      </tr><tr><td>
          <p><code><a href="discardFeatures.html">discardFeatures()</a></code> </p>
        </td>
        <td><p>Discard features</p></td>
      </tr><tr><td>
          <p><code><a href="discardGeneFeatures.html">discardGeneFeatures()</a></code> </p>
        </td>
        <td><p>Discard gene features</p></td>
      </tr><tr><td>
          <p><code><a href="discardGenes.html">discardGenes()</a></code> </p>
        </td>
        <td><p>Discard genes</p></td>
      </tr><tr><td>
          <p><code><a href="discardGeneSets.html">discardGeneSets()</a></code> </p>
        </td>
        <td><p>Discard gene sets</p></td>
      </tr><tr><td>
          <p><code><a href="dropped_df_dummy.html">dropped_df_dummy()</a></code> </p>
        </td>
        <td><p>dropped_df</p></td>
      </tr><tr><td>
          <p><code><a href="examineClusterResults.html">examineClusterResults()</a></code> </p>
        </td>
        <td><p>Examine clustering results</p></td>
      </tr><tr><td>
          <p><code><a href="examineTrajectoryAssessment.html">examineTrajectoryAssessment()</a></code> </p>
        </td>
        <td><p>Examine trajectory-moddeling results</p></td>
      </tr><tr><td>
          <p><code><a href="filterDeaDf.html">filterDeaDf()</a></code> </p>
        </td>
        <td><p>Postprocess de-analysis results</p></td>
      </tr><tr><td>
          <p><code><a href="filterTrends.html">filterTrends()</a></code> <code><a href="filterTrends.html">filterTrajectoryTrends()</a></code> </p>
        </td>
        <td><p>Filter variables of a certain trend</p></td>
      </tr><tr><td>
          <p><code><a href="findHierarchicalCluster.html">findHierarchicalCluster()</a></code> <code><a href="findHierarchicalCluster.html">findHierarchicalClusterCnv()</a></code> </p>
        </td>
        <td><p>Cluster sample via hclust</p></td>
      </tr><tr><td>
          <p><code><a href="findMonocleClusters.html">findMonocleClusters()</a></code> </p>
        </td>
        <td><p>Cluster sample via monocle3</p></td>
      </tr><tr><td>
          <p><code><a href="findNearestNeighbourClusters.html">findNearestNeighbourClusters()</a></code> </p>
        </td>
        <td><p>Cluster sample via nearest neighbour analysis</p></td>
      </tr><tr><td>
          <p><code><a href="findSeuratClusters.html">findSeuratClusters()</a></code> </p>
        </td>
        <td><p>Cluster sample via Seurat</p></td>
      </tr><tr><td>
          <p><code><a href="flexible_call_dummy.html">flexible_call_dummy()</a></code> </p>
        </td>
        <td><p>flexible_call_dummy</p></td>
      </tr><tr><td>
          <p><code><a href="flipImage.html">flipImage()</a></code> </p>
        </td>
        <td><p>Flip object image</p></td>
      </tr><tr><td>
          <p><code><a href="gene_set_path.html">gene_set_path()</a></code> </p>
        </td>
        <td><p>gene_set_path</p></td>
      </tr><tr><td>
          <p><code><a href="getActiveMatrixName.html">getActiveMatrixName()</a></code> </p>
        </td>
        <td><p>Obtain name of currently active expression matrix</p></td>
      </tr><tr><td>
          <p><code><a href="getAutoencoderAssessment.html">getAutoencoderAssessment()</a></code> </p>
        </td>
        <td><p>Obtain information about the optimal neural network set up</p></td>
      </tr><tr><td>
          <p><code><a href="getAutoencoderSetUp.html">getAutoencoderSetUp()</a></code> </p>
        </td>
        <td><p>Obtain information on neural network</p></td>
      </tr><tr><td>
          <p><code><a href="getBarcodes.html">getBarcodes()</a></code> </p>
        </td>
        <td><p>Obtain specific barcodes</p></td>
      </tr><tr><td>
          <p><code><a href="getCnvFeatureNames.html">getCnvFeatureNames()</a></code> </p>
        </td>
        <td><p>Obtain features names under which cnv-analysis results are stored.</p></td>
      </tr><tr><td>
          <p><code><a href="getCnvResults.html">getCnvResults()</a></code> </p>
        </td>
        <td><p>Obtain copy-number-variations results</p></td>
      </tr><tr><td>
          <p><code><a href="getCoordsDf.html">getCoordsDf()</a></code> <code><a href="getCoordsDf.html">getSegmentDf()</a></code> </p>
        </td>
        <td><p>Obtain spatial coordinates</p></td>
      </tr><tr><td>
          <p><code><a href="getDeaOverview.html">getDeaOverview()</a></code> </p>
        </td>
        <td><p>Obtain info on de-analysis storage</p></td>
      </tr><tr><td>
          <p><code><a href="getDeaResultsDf.html">getDeaResultsDf()</a></code> <code><a href="getDeaResultsDf.html">getDeaGenes()</a></code> </p>
        </td>
        <td><p>Obtain de-analysis results</p></td>
      </tr><tr><td>
          <p><code><a href="getDefaultInstructions.html">getDefaultInstructions()</a></code> <code><a href="getDefaultInstructions.html">getDirectoryInstructions()</a></code> </p>
        </td>
        <td><p>Obtain default argument inputs</p></td>
      </tr><tr><td>
          <p><code><a href="getDimRedDf.html">getDimRedDf()</a></code> <code><a href="getDimRedDf.html">getPcaDf()</a></code> <code><a href="getDimRedDf.html">getPcaMtr()</a></code> <code><a href="getDimRedDf.html">getUmapDf()</a></code> <code><a href="getDimRedDf.html">getTsneDf()</a></code> </p>
        </td>
        <td><p>Obtain dimensional reduction data</p></td>
      </tr><tr><td>
          <p><code><a href="getExpressionMatrixNames.html">getExpressionMatrixNames()</a></code> </p>
        </td>
        <td><p>Obtain names of stored expression matrices</p></td>
      </tr><tr><td>
          <p><code><a href="getFeatureDf.html">getFeatureDf()</a></code> </p>
        </td>
        <td><p>Obtain feature data</p></td>
      </tr><tr><td>
          <p><code><a href="getFeatureNames.html">getFeatureNames()</a></code> </p>
        </td>
        <td><p>Obtain feature names</p></td>
      </tr><tr><td>
          <p><code><a href="getFeatureValues.html">getFeatureValues()</a></code> </p>
        </td>
        <td><p>Obtain unique categorical feature values</p></td>
      </tr><tr><td>
          <p><code><a href="getFeatureVariables.html">getFeatureVariables()</a></code> </p>
        </td>
        <td><p>Obtain a feature variable</p></td>
      </tr><tr><td>
          <p><code><a href="getFromSeurat.html">getFromSeurat()</a></code> </p>
        </td>
        <td><p>Safe extraction</p></td>
      </tr><tr><td>
          <p><code><a href="getGeneCounts.html">getGeneCounts()</a></code> </p>
        </td>
        <td><p>Obtain total number of gene counts</p></td>
      </tr><tr><td>
          <p><code><a href="getGeneDistMtr.html">getGeneDistMtr()</a></code> </p>
        </td>
        <td><p>Obtain distance measurements of spatially correlated genes</p></td>
      </tr><tr><td>
          <p><code><a href="getGeneFeatureNames.html">getGeneFeatureNames()</a></code> </p>
        </td>
        <td><p>Obtain feature names of the gene meta data</p></td>
      </tr><tr><td>
          <p><code><a href="getGeneMetaData.html">getGeneMetaData()</a></code> <code><a href="getGeneMetaData.html">getGeneMetaDf()</a></code> </p>
        </td>
        <td><p>Obtain gene meta data</p></td>
      </tr><tr><td>
          <p><code><a href="getGenePatternCluster.html">getGenePatternCluster()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="getGenePatternCorrelations.html">getGenePatternCorrelations()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="getGenePatternDistances.html">getGenePatternDistances()</a></code> <code><a href="getGenePatternDistances.html">getGenePatternSimilarities()</a></code> </p>
        </td>
        <td><p>Obtain similarity scores between identified gene patterns</p></td>
      </tr><tr><td>
          <p><code><a href="getGenePatternEvalDf.html">getGenePatternEvalDf()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="getGenePatternExtentDf.html">getGenePatternExtentDf()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="getGenePatternsAboveThreshold.html">getGenePatternsAboveThreshold()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="getGenes.html">getGenes()</a></code> <code><a href="getGenes.html">getGenesInteractive()</a></code> </p>
        </td>
        <td><p>Obtain gene names</p></td>
      </tr><tr><td>
          <p><code><a href="getGeneSetDf.html">getGeneSetDf()</a></code> </p>
        </td>
        <td><p>Obtain gene set data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="getGeneSetOverview.html">getGeneSetOverview()</a></code> </p>
        </td>
        <td><p>Overview about the current gene sets</p></td>
      </tr><tr><td>
          <p><code><a href="getGeneSets.html">getGeneSets()</a></code> <code><a href="getGeneSets.html">getGeneSetsInteractive()</a></code> </p>
        </td>
        <td><p>Obtain gene set names</p></td>
      </tr><tr><td>
          <p><code><a href="getGroupingOptions.html">getGroupingOptions()</a></code> </p>
        </td>
        <td><p>Obtain variable names that group the barcode spots</p></td>
      </tr><tr><td>
          <p><code><a href="getGroupNames.html">getGroupNames()</a></code> </p>
        </td>
        <td><p>Obtain group names a grouping variable contains</p></td>
      </tr><tr><td>
          <p><code><a href="getGseaDf.html">getGseaDf()</a></code> </p>
        </td>
        <td><p>Obtain enrichment data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="getGseaResults.html">getGseaResults()</a></code> </p>
        </td>
        <td><p>Obtain enrichment results</p></td>
      </tr><tr><td>
          <p><code><a href="getHclustObject.html">getHclustObject()</a></code> <code><a href="getHclustObject.html">getHclustObjectCnv()</a></code> </p>
        </td>
        <td><p>Obtain hierarchical cluster results</p></td>
      </tr><tr><td>
          <p><code><a href="getHspaResults.html">getHspaResults()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="getImage.html">getImage()</a></code> </p>
        </td>
        <td><p>Obtain histology image</p></td>
      </tr><tr><td>
          <p><code><a href="getInitiationInfo.html">getInitiationInfo()</a></code> <code><a href="getInitiationInfo.html">getInitiationInput()</a></code> </p>
        </td>
        <td><p>Obtain information about object initiation</p></td>
      </tr><tr><td>
          <p><code><a href="getMatrix.html">getMatrix()</a></code> <code><a href="getMatrix.html">getExpressionMatrix()</a></code> <code><a href="getMatrix.html">getCountMatrix()</a></code> </p>
        </td>
        <td><p>Obtain count and expression matrix</p></td>
      </tr><tr><td>
          <p><code><a href="getPrResults.html">getPrResults()</a></code> <code><a href="getPrResults.html">getPatternNames()</a></code> </p>
        </td>
        <td><p>Obtain pattern recognition results</p></td>
      </tr><tr><td>
          <p><code><a href="getSampleNames.html">getSampleNames()</a></code> </p>
        </td>
        <td><p>Obtain sample names</p></td>
      </tr><tr><td>
          <p><code><a href="getSegmentNames.html">getSegmentNames()</a></code> </p>
        </td>
        <td><p>Obtain segment names</p></td>
      </tr><tr><td>
          <p><code><a href="getSignatureEnrichment.html">getSignatureEnrichment()</a></code> </p>
        </td>
        <td><p>Obtain signature enrichment</p></td>
      </tr><tr><td>
          <p><code><a href="getSpataDf.html">getSpataDf()</a></code> </p>
        </td>
        <td><p>Obtain a spata-data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="getSpCorCluster.html">getSpCorCluster()</a></code> <code><a href="getSpCorCluster.html">getSpCorClusterNames()</a></code> </p>
        </td>
        <td><p>Obtain cluster results based on spatial correlation analysis</p></td>
      </tr><tr><td>
          <p><code><a href="getSpCorResults.html">getSpCorResults()</a></code> </p>
        </td>
        <td><p>Obtain spatial correlation results</p></td>
      </tr><tr><td>
          <p><code><a href="getTrajectoryComment-spata2-method.html">getTrajectoryComment(<i>&lt;spata2&gt;</i>)</a></code> </p>
        </td>
        <td><p>Obtain a trajectory comment</p></td>
      </tr><tr><td>
          <p><code><a href="getTrajectoryComment.html">getTrajectoryComment()</a></code> </p>
        </td>
        <td><p>Obtain a trajectory comment</p></td>
      </tr><tr><td>
          <p><code><a href="getTrajectoryDf.html">getTrajectoryDf()</a></code> </p>
        </td>
        <td><p>Obtain a summarized trajectory data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="getTrajectoryLength.html">getTrajectoryLength()</a></code> </p>
        </td>
        <td><p>Obtain the length of a trajectory</p></td>
      </tr><tr><td>
          <p><code><a href="getTrajectoryNames.html">getTrajectoryNames()</a></code> </p>
        </td>
        <td><p>Obtain trajectory names</p></td>
      </tr><tr><td>
          <p><code><a href="getTrajectoryObject.html">getTrajectoryObject()</a></code> </p>
        </td>
        <td><p>Obtain trajectory object</p></td>
      </tr><tr><td>
          <p><code><a href="ggplot_family.html">ggplot_family()</a></code> </p>
        </td>
        <td><p>ggplot_family</p></td>
      </tr><tr><td>
          <p><code><a href="gsdf.html">gsdf</a></code> </p>
        </td>
        <td><p>The default collection of frequently used gene-sets</p></td>
      </tr><tr><td>
          <p><code><a href="hlpr_add_barcode_suffix.html">hlpr_add_barcode_suffix()</a></code> </p>
        </td>
        <td><p>Make sure that barcodes are spata-like</p></td>
      </tr><tr><td>
          <p><code><a href="hlpr_add_models.html">hlpr_add_models()</a></code> <code><a href="hlpr_add_models.html">hlpr_add_residuals()</a></code> <code><a href="hlpr_add_models.html">hlpr_add_residuals_customized()</a></code> <code><a href="hlpr_add_models.html">hlpr_summarize_residuals()</a></code> <code><a href="hlpr_add_models.html">hlpr_name_models()</a></code> <code><a href="hlpr_add_models.html">hlpr_filter_trend()</a></code> </p>
        </td>
        <td><p>Helper functions for trajectory ranking</p></td>
      </tr><tr><td>
          <p><code><a href="hlpr_add_old_coords.html">hlpr_add_old_coords()</a></code> </p>
        </td>
        <td><p>Adds old coordinates</p></td>
      </tr><tr><td>
          <p><code><a href="hlpr_adjust_legend_size.html">hlpr_adjust_legend_size()</a></code> </p>
        </td>
        <td><p>Adjusts the size of discrete legend points</p></td>
      </tr><tr><td>
          <p><code><a href="hlpr_assess_trajectory_trends.html">hlpr_assess_trajectory_trends()</a></code> <code><a href="hlpr_assess_trajectory_trends.html">hlpr_assess_trajectory_trends_customized()</a></code> </p>
        </td>
        <td><p>Assess trajectory ranking.</p></td>
      </tr><tr><td>
          <p><code><a href="hlpr_assign_arguments.html">hlpr_assign_arguments()</a></code> </p>
        </td>
        <td><p>Assign arguments to calling function</p></td>
      </tr><tr><td>
          <p><code><a href="hlpr_breaks.html">hlpr_breaks()</a></code> </p>
        </td>
        <td><p>Calculates breaks for heatmap according to input matrix</p></td>
      </tr><tr><td>
          <p><code><a href="hlpr_compare_samples.html">hlpr_compare_samples()</a></code> </p>
        </td>
        <td><p>Compare samples within an object</p></td>
      </tr><tr><td>
          <p><code><a href="hlpr_compile_trajectory.html">hlpr_compile_trajectory()</a></code> </p>
        </td>
        <td><p>Compiles a trajectory data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="hlpr_display_title.html">hlpr_display_title()</a></code> <code><a href="hlpr_display_title.html">hlpr_display_subtitle()</a></code> </p>
        </td>
        <td><p>No documentation needed</p></td>
      </tr><tr><td>
          <p><code><a href="hlpr_dist_mtr_to_df.html">hlpr_dist_mtr_to_df()</a></code> </p>
        </td>
        <td><p>Convert distance matrix to distance data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="hlpr_gene_set_name.html">hlpr_gene_set_name()</a></code> </p>
        </td>
        <td><p>Removes the class part of a gene set string</p></td>
      </tr><tr><td>
          <p><code><a href="hlpr_geom_trajectory_fit.html">hlpr_geom_trajectory_fit()</a></code> </p>
        </td>
        <td><p>Easy switch between geom line and and geom smooth</p></td>
      </tr><tr><td>
          <p><code><a href="hlpr_image_add_on.html">hlpr_image_add_on()</a></code> <code><a href="hlpr_image_add_on.html">hlpr_image_add_on2()</a></code> </p>
        </td>
        <td><p>Provides the image as ggplot background</p></td>
      </tr><tr><td>
          <p><code><a href="hlpr_labs_add_on.html">hlpr_labs_add_on()</a></code> </p>
        </td>
        <td><p>Return customized ggplot:labs()</p></td>
      </tr><tr><td>
          <p><code><a href="hlpr_normalize_imap.html">hlpr_normalize_imap()</a></code> </p>
        </td>
        <td><p>Normalize gene or gene set values</p></td>
      </tr><tr><td>
          <p><code><a href="hlpr_normalize_vctr.html">hlpr_normalize_vctr()</a></code> </p>
        </td>
        <td><p>Normalize gene or gene set values</p></td>
      </tr><tr><td>
          <p><code><a href="hlpr_one_distinct.html">hlpr_one_distinct()</a></code> </p>
        </td>
        <td><p>Number of distinct values equal to 1?</p></td>
      </tr><tr><td>
          <p><code><a href="hlpr_process_spatial_correlation_cluster.html">hlpr_process_spatial_correlation_cluster()</a></code> </p>
        </td>
        <td><p>Process spatial correlation results</p></td>
      </tr><tr><td>
          <p><code><a href="hlpr_rank_trajectory_trends.html">hlpr_rank_trajectory_trends()</a></code> <code><a href="hlpr_rank_trajectory_trends.html">hlpr_rank_trajectory_trends_customized()</a></code> </p>
        </td>
        <td><p>Rank trajectory trends.</p></td>
      </tr><tr><td>
          <p><code><a href="hlpr_run_cnva_pca.html">hlpr_run_cnva_pca()</a></code> </p>
        </td>
        <td><p>run pca on the cnv-matrix</p></td>
      </tr><tr><td>
          <p><code><a href="hlpr_save_spata_object.html">hlpr_save_spata_object()</a></code> </p>
        </td>
        <td><p>Save spata object inside functions</p></td>
      </tr><tr><td>
          <p><code><a href="hlpr_scatterplot.html">hlpr_scatterplot()</a></code> </p>
        </td>
        <td><p>Color to + scatterplot helper</p></td>
      </tr><tr><td>
          <p><code><a href="hlpr_smooth.html">hlpr_smooth()</a></code> </p>
        </td>
        <td><p>Smooth variables spatially</p></td>
      </tr><tr><td>
          <p><code><a href="hlpr_smooth_shiny.html">hlpr_smooth_shiny()</a></code> </p>
        </td>
        <td><p>Smooth variable spatially in mini-shiny-apps</p></td>
      </tr><tr><td>
          <p><code><a href="hlpr_subset_across.html">hlpr_subset_across()</a></code> </p>
        </td>
        <td><p>Subset the across-variables</p></td>
      </tr><tr><td>
          <p><code><a href="hlpr_summarize_trajectory_df.html">hlpr_summarize_trajectory_df()</a></code> </p>
        </td>
        <td><p>Join and summarize compiled trajectory data.frames</p></td>
      </tr><tr><td>
          <p><code><a href="hlpr_vector_projection.html">hlpr_vector_projection()</a></code> </p>
        </td>
        <td><p>Perform vector projection</p></td>
      </tr><tr><td>
          <p><code><a href="hlpr_widen_trajectory_df.html">hlpr_widen_trajectory_df()</a></code> </p>
        </td>
        <td><p>Widen trajectory data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="hspaClusterGenePatterns.html">hspaClusterGenePatterns()</a></code> </p>
        </td>
        <td><p>Construct hcluster-tree from correlated gene pattern similarities</p></td>
      </tr><tr><td>
          <p><code><a href="hspaCsrCutoff.html">hspaCsrCutoff()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="hspaCsrTesting.html">hspaCsrTesting()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="hspaDataBinarization.html">hspaDataBinarization()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="hspaPatternIdentification.html">hspaPatternIdentification()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="hspaPatternSimilarity.html">hspaPatternSimilarity()</a></code> <code><a href="hspaPatternSimilarity.html">hspaPatternSimilarity_future()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="hspaSelectGenes.html">hspaSelectGenes()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="identify_gene_patterns_dbscan.html">identify_gene_patterns_dbscan()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="image_dummy.html">image_dummy()</a></code> </p>
        </td>
        <td><p>image_dummy</p></td>
      </tr><tr><td>
          <p><code><a href="initiateSpataObject_10X.html">initiateSpataObject_10X()</a></code> </p>
        </td>
        <td><p>Initiate a spata-object from 10X Visium</p></td>
      </tr><tr><td>
          <p><code><a href="initiateSpataObject_CountMtr.html">initiateSpataObject_CountMtr()</a></code> </p>
        </td>
        <td><p>Initiate a spata-object from a raw count matrix</p></td>
      </tr><tr><td>
          <p><code><a href="initiateSpataObject_Empty.html">initiateSpataObject_Empty()</a></code> </p>
        </td>
        <td><p>Initiate an empty spata-object</p></td>
      </tr><tr><td>
          <p><code><a href="initiateSpataObject_Examples.html">initiateSpataObject_Examples()</a></code> </p>
        </td>
        <td><p>Initiate a spata-object from example data sets</p></td>
      </tr><tr><td>
          <p><code><a href="initiateSpataObject_ExprMtr.html">initiateSpataObject_ExprMtr()</a></code> </p>
        </td>
        <td><p>Initiate spata object from scaled expression matrix</p></td>
      </tr><tr><td>
          <p><code><a href="is_subsetted_by_segment.html">is_subsetted_by_segment()</a></code> </p>
        </td>
        <td><p>Check if object has been subsetted by segment</p></td>
      </tr><tr><td>
          <p><code><a href="joinWith.html">joinWith()</a></code> <code><a href="joinWith.html">joinWithFeatures()</a></code> <code><a href="joinWith.html">joinWithGenes()</a></code> <code><a href="joinWith.html">joinWithGeneSets()</a></code> <code><a href="joinWith.html">joinWithVariables()</a></code> </p>
        </td>
        <td><p>Join barcodes with additional variables</p></td>
      </tr><tr><td>
          <p><code><a href="joinWithPca.html">joinWithPca()</a></code> <code><a href="joinWithPca.html">joinWithTsne()</a></code> <code><a href="joinWithPca.html">joinWithUmap()</a></code> </p>
        </td>
        <td><p>Join barcodes with additional variables</p></td>
      </tr><tr><td>
          <p><code><a href="lazy_check_dummy.html">lazy_check_dummy()</a></code> </p>
        </td>
        <td><p>lazy_check_dummy</p></td>
      </tr><tr><td>
          <p><code><a href="legendBottom.html">legendBottom()</a></code> <code><a href="legendBottom.html">legendNone()</a></code> <code><a href="legendBottom.html">legendRight()</a></code> <code><a href="legendBottom.html">legendTop()</a></code> </p>
        </td>
        <td><p>Ggplot add on wrapper</p></td>
      </tr><tr><td>
          <p><code><a href="loadGeneSetDf.html">loadGeneSetDf()</a></code> </p>
        </td>
        <td><p>Load gene set data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="loadGSDF.html">loadGSDF()</a></code> </p>
        </td>
        <td><p>Original load gene set data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="loadSpataObject.html">loadSpataObject()</a></code> <code><a href="loadSpataObject.html">loadCorrespondingCDS()</a></code> <code><a href="loadSpataObject.html">loadCorrespondingSeuratObject()</a></code> </p>
        </td>
        <td><p>Load corresponding objects</p></td>
      </tr><tr><td>
          <p><code><a href="mergeSpataObjects.html">mergeSpataObjects()</a></code> </p>
        </td>
        <td><p>Merge Spata Objects</p></td>
      </tr><tr><td>
          <p><code><a href="method_hclust.html">method_hclust()</a></code> </p>
        </td>
        <td><p>method_hclust</p></td>
      </tr><tr><td>
          <p><code><a href="moduleAddGeneSetsServer.html">moduleAddGeneSetsServer()</a></code> </p>
        </td>
        <td><p>Server of the add gene sets module</p></td>
      </tr><tr><td>
          <p><code><a href="moduleAddGeneSetsUI.html">moduleAddGeneSetsUI()</a></code> </p>
        </td>
        <td><p>UI of the add gene sets module</p></td>
      </tr><tr><td>
          <p><code><a href="moduleSurfacePlotServer.html">moduleSurfacePlotServer()</a></code> </p>
        </td>
        <td><p>Server of the surface plot module</p></td>
      </tr><tr><td>
          <p><code><a href="moduleSurfacePlotUI.html">moduleSurfacePlotUI()</a></code> </p>
        </td>
        <td><p>UI of the surface plot module</p></td>
      </tr><tr><td>
          <p><code><a href="nGenes.html">nGenes()</a></code> </p>
        </td>
        <td><p>Number of genes</p></td>
      </tr><tr><td>
          <p><code><a href="normalize.html">normalize()</a></code> </p>
        </td>
        <td><p>normalize</p></td>
      </tr><tr><td>
          <p><code><a href="pb_dummy.html">pb_dummy()</a></code> </p>
        </td>
        <td><p>pb</p></td>
      </tr><tr><td>
          <p><code><a href="pheatmap.html">pheatmap()</a></code> </p>
        </td>
        <td><p>de_df</p></td>
      </tr><tr><td>
          <p><code><a href="plotAutoencoderAssessment.html">plotAutoencoderAssessment()</a></code> </p>
        </td>
        <td><p>Plot total variance of different neural networks</p></td>
      </tr><tr><td>
          <p><code><a href="plotAutoencoderResults.html">plotAutoencoderResults()</a></code> </p>
        </td>
        <td><p>Plot scaled vs. denoised expression</p></td>
      </tr><tr><td>
          <p><code><a href="plotBarchart.html">plotBarchart()</a></code> </p>
        </td>
        <td><p>Plot distribution of discrete variables</p></td>
      </tr><tr><td>
          <p><code><a href="plotBoxplot.html">plotBoxplot()</a></code> <code><a href="plotBoxplot.html">plotDensityplot()</a></code> <code><a href="plotBoxplot.html">plotHistogram()</a></code> <code><a href="plotBoxplot.html">plotRidgeplot()</a></code> <code><a href="plotBoxplot.html">plotViolinplot()</a></code> </p>
        </td>
        <td><p>Plot numeric distribution and statistical tests</p></td>
      </tr><tr><td>
          <p><code><a href="plotCnvResults.html">plotCnvResults()</a></code> </p>
        </td>
        <td><p>Visualize copy-number-variations results</p></td>
      </tr><tr><td>
          <p><code><a href="plotCsrCutoffSimulations.html">plotCsrCutoffSimulations()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="plotCsrResults.html">plotCsrResults()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="plotCustomizedTrajectoryTrends.html">plotCustomizedTrajectoryTrends()</a></code> </p>
        </td>
        <td><p>Plot customized trajectory trends</p></td>
      </tr><tr><td>
          <p><code><a href="plotDeaDotPlot.html">plotDeaDotPlot()</a></code> </p>
        </td>
        <td><p>Plot differentially expressed genes</p></td>
      </tr><tr><td>
          <p><code><a href="plotDeaGeneCount.html">plotDeaGeneCount()</a></code> </p>
        </td>
        <td><p>Plot the number of differently expressed genes of certain groups</p></td>
      </tr><tr><td>
          <p><code><a href="plotDeaHeatmap.html">plotDeaHeatmap()</a></code> </p>
        </td>
        <td><p>Plot differentially expressed genes</p></td>
      </tr><tr><td>
          <p><code><a href="plotDeaPvalues.html">plotDeaPvalues()</a></code> <code><a href="plotDeaPvalues.html">plotDeaLogFC()</a></code> </p>
        </td>
        <td><p>Plot the p-value and log fold change distribution of de-analysis results</p></td>
      </tr><tr><td>
          <p><code><a href="plotDeaSummary.html">plotDeaSummary()</a></code> </p>
        </td>
        <td><p>Plot a summary of differential expression analysis results</p></td>
      </tr><tr><td>
          <p><code><a href="plotDendrogram.html">plotDendrogram()</a></code> <code><a href="plotDendrogram.html">plotDendrogramCnv()</a></code> </p>
        </td>
        <td><p>Visualize hierarchical clustering results</p></td>
      </tr><tr><td>
          <p><code><a href="plotDistribution.html">plotDistribution()</a></code> <code><a href="plotDistribution.html">plotDistribution2()</a></code> <code><a href="plotDistribution.html">plotDistributionAcross()</a></code> <code><a href="plotDistribution.html">plotDistributionAcross2()</a></code> <code><a href="plotDistribution.html">plotDistributionDiscrete()</a></code> </p>
        </td>
        <td><p>Distribution of continuous values (Deprecated)</p></td>
      </tr><tr><td>
          <p><code><a href="plotFourStates.html">plotFourStates()</a></code> <code><a href="plotFourStates.html">plotFourStates2()</a></code> </p>
        </td>
        <td><p>Gene set state plot</p></td>
      </tr><tr><td>
          <p><code><a href="plotGeneDendrogram.html">plotGeneDendrogram()</a></code> </p>
        </td>
        <td><p>Visualize clustering results</p></td>
      </tr><tr><td>
          <p><code><a href="plotGeneMetaData.html">plotGeneMetaData()</a></code> </p>
        </td>
        <td><p>Plot gene meta data</p></td>
      </tr><tr><td>
          <p><code><a href="plotGenePatternBinarized.html">plotGenePatternBinarized()</a></code> <code><a href="plotGenePatternBinarized.html">plotGenePatternDenoised()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="plotGenePatternCluster.html">plotGenePatternCluster()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="plotGenePatternDendrogram.html">plotGenePatternDendrogram()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="plotGenePatternHeatmap.html">plotGenePatternHeatmap()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="plotGenePatterns.html">plotGenePatterns()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="plotGseaDotPlot.html">plotGseaDotPlot()</a></code> </p>
        </td>
        <td><p>Plot gene set enrichment</p></td>
      </tr><tr><td>
          <p><code><a href="plotHspaSummary.html">plotHspaSummary()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="plotPcaVariation.html">plotPcaVariation()</a></code> </p>
        </td>
        <td><p>Plot Pca Variation</p></td>
      </tr><tr><td>
          <p><code><a href="plotPseudotime.html">plotPseudotime()</a></code> </p>
        </td>
        <td><p>Monocle3 Pseudotime</p></td>
      </tr><tr><td>
          <p><code><a href="plotScatterplot.html">plotScatterplot()</a></code> </p>
        </td>
        <td><p>Plot numeric variables as a scatterplot</p></td>
      </tr><tr><td>
          <p><code><a href="plotSegmentation.html">plotSegmentation()</a></code> </p>
        </td>
        <td><p>Plot segmentation</p></td>
      </tr><tr><td>
          <p><code><a href="plotStatisticsInteractive.html">plotStatisticsInteractive()</a></code> </p>
        </td>
        <td><p>Plot distribution of variables interactively</p></td>
      </tr><tr><td>
          <p><code><a href="plotSurface.html">plotSurface()</a></code> <code><a href="plotSurface.html">plotSurface2()</a></code> <code><a href="plotSurface.html">plotSurfaceInteractive()</a></code> </p>
        </td>
        <td><p>Plot the surface of the sample</p></td>
      </tr><tr><td>
          <p><code><a href="plotSurfaceAverage.html">plotSurfaceAverage()</a></code> </p>
        </td>
        <td><p>Plot several surface plots colored by gene averages</p></td>
      </tr><tr><td>
          <p><code><a href="plotSurfaceComparison.html">plotSurfaceComparison()</a></code> <code><a href="plotSurfaceComparison.html">plotSurfaceComparison2()</a></code> </p>
        </td>
        <td><p>Plot several surface plots colored by numeric variables</p></td>
      </tr><tr><td>
          <p><code><a href="plotSurfaceHotspots.html">plotSurfaceHotspots()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="plotSurfaceQuantiles.html">plotSurfaceQuantiles()</a></code> </p>
        </td>
        <td><p>Plot a surface plot colored by binned numeric variables</p></td>
      </tr><tr><td>
          <p><code><a href="plotTrajectory.html">plotTrajectory()</a></code> </p>
        </td>
        <td><p>Plot trajectory</p></td>
      </tr><tr><td>
          <p><code><a href="plotTrajectoryFeatures.html">plotTrajectoryFeatures()</a></code> <code><a href="plotTrajectoryFeatures.html">plotTrajectoryGenes()</a></code> <code><a href="plotTrajectoryFeatures.html">plotTrajectoryGeneSets()</a></code> </p>
        </td>
        <td><p>Plot continuous trajectory dynamics in lineplots</p></td>
      </tr><tr><td>
          <p><code><a href="plotTrajectoryFeaturesDiscrete.html">plotTrajectoryFeaturesDiscrete()</a></code> </p>
        </td>
        <td><p>Plot discrete trajectory dynamics</p></td>
      </tr><tr><td>
          <p><code><a href="plotTrajectoryFit.html">plotTrajectoryFit()</a></code> <code><a href="plotTrajectoryFit.html">plotTrajectoryFitCustomized()</a></code> </p>
        </td>
        <td><p>Plot trajectory fit</p></td>
      </tr><tr><td>
          <p><code><a href="plotTrajectoryHeatmap.html">plotTrajectoryHeatmap()</a></code> </p>
        </td>
        <td><p>Plot trajectory expression dynamic in heatmap</p></td>
      </tr><tr><td>
          <p><code><a href="plotUmap.html">plotUmap()</a></code> <code><a href="plotUmap.html">plotTsne()</a></code> <code><a href="plotUmap.html">plotPca()</a></code> </p>
        </td>
        <td><p>Plot dimensional reduction</p></td>
      </tr><tr><td>
          <p><code><a href="plot_type_dummy.html">plot_type_dummy()</a></code> </p>
        </td>
        <td><p>plot_type</p></td>
      </tr><tr><td>
          <p><code><a href="process_seurat_object.html">process_seurat_object()</a></code> </p>
        </td>
        <td><p>Wrapper around Seurat processing functions</p></td>
      </tr><tr><td>
          <p><code><a href="renameFeatures.html">renameFeatures()</a></code> </p>
        </td>
        <td><p>Rename features</p></td>
      </tr><tr><td>
          <p><code><a href="renameGroups.html">renameGroups()</a></code> <code><a href="renameGroups.html">renameSegments()</a></code> </p>
        </td>
        <td><p>Rename cluster/group names</p></td>
      </tr><tr><td>
          <p><code><a href="runAutoencoderAssessment.html">runAutoencoderAssessment()</a></code> <code><a href="runAutoencoderAssessment.html">assessAutoencoderOptions()</a></code> </p>
        </td>
        <td><p>Assessment of Neural Network Set Up</p></td>
      </tr><tr><td>
          <p><code><a href="runAutoencoderDenoising.html">runAutoencoderDenoising()</a></code> </p>
        </td>
        <td><p>Denoise expression matrix</p></td>
      </tr><tr><td>
          <p><code><a href="runCnvAnalysis.html">runCnvAnalysis()</a></code> </p>
        </td>
        <td><p>Identify large-scale chromosomal copy number variations</p></td>
      </tr><tr><td>
          <p><code><a href="runDeAnalysis.html">runDeAnalysis()</a></code> </p>
        </td>
        <td><p>Find differently expressed genes</p></td>
      </tr><tr><td>
          <p><code><a href="runGSEA.html">runGSEA()</a></code> </p>
        </td>
        <td><p>Compute gene set enrichment</p></td>
      </tr><tr><td>
          <p><code><a href="runHclust.html">runHclust()</a></code> <code><a href="runHclust.html">runHclustCnv()</a></code> </p>
        </td>
        <td><p>Run hierarchical clustering</p></td>
      </tr><tr><td>
          <p><code><a href="runHspa.html">runHspa()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="runPca.html">runPca()</a></code> <code><a href="runPca.html">runPca2()</a></code> </p>
        </td>
        <td><p>Run Principal Component Analysis</p></td>
      </tr><tr><td>
          <p><code><a href="runSpatialCorrelationAnalysis.html">runSpatialCorrelationAnalysis()</a></code> </p>
        </td>
        <td><p>Initiate gene clustering analysis based on spatial patterns</p></td>
      </tr><tr><td>
          <p><code><a href="runTsne.html">runTsne()</a></code> <code><a href="runTsne.html">runTsne2()</a></code> </p>
        </td>
        <td><p>Run t-Stochastic Neighbour Embedding</p></td>
      </tr><tr><td>
          <p><code><a href="runUmap.html">runUmap()</a></code> <code><a href="runUmap.html">runUmap2()</a></code> </p>
        </td>
        <td><p>Run UMAP-Analysis</p></td>
      </tr><tr><td>
          <p><code><a href="sample_name.html">sample_name()</a></code> </p>
        </td>
        <td><p>sample_name</p></td>
      </tr><tr><td>
          <p><code><a href="saveGeneSetDf.html">saveGeneSetDf()</a></code> </p>
        </td>
        <td><p>Save a gene set data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="saveSpataObject.html">saveSpataObject()</a></code> <code><a href="saveSpataObject.html">saveCorrespondingCDS()</a></code> <code><a href="saveSpataObject.html">saveCorrespondingSeuratObject()</a></code> </p>
        </td>
        <td><p>Save corresponding objects</p></td>
      </tr><tr><td>
          <p><code><a href="scale_color_add_on.html">scale_color_add_on()</a></code> </p>
        </td>
        <td><p>ggplot2 - color add on</p></td>
      </tr><tr><td>
          <p><code><a href="setActiveExpressionMatrix.html">setActiveExpressionMatrix()</a></code> </p>
        </td>
        <td><p>Denote the default expression matrix</p></td>
      </tr><tr><td>
          <p><code><a href="setAutoencoderAssessment.html">setAutoencoderAssessment()</a></code> </p>
        </td>
        <td><p>Set results of autoencoder assessment</p></td>
      </tr><tr><td>
          <p><code><a href="setCnvResults.html">setCnvResults()</a></code> </p>
        </td>
        <td><p>Set cnv-results</p></td>
      </tr><tr><td>
          <p><code><a href="setCoordsDf.html">setCoordsDf()</a></code> </p>
        </td>
        <td><p>Set the coordinates</p></td>
      </tr><tr><td>
          <p><code><a href="setCountMatrix.html">setCountMatrix()</a></code> <code><a href="setCountMatrix.html">setDenoisedMatrix()</a></code> <code><a href="setCountMatrix.html">setScaledMatrix()</a></code> </p>
        </td>
        <td><p>Set data matrices</p></td>
      </tr><tr><td>
          <p><code><a href="setCsrCutoff.html">setCsrCutoff()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="setDefaultInstructions.html">setDefaultInstructions()</a></code> <code><a href="setDefaultInstructions.html">setDirectoryInstructions()</a></code> </p>
        </td>
        <td><p>Set default instructions</p></td>
      </tr><tr><td>
          <p><code><a href="setFeatureDf.html">setFeatureDf()</a></code> </p>
        </td>
        <td><p>Set feature data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="setGeneSetDf.html">setGeneSetDf()</a></code> </p>
        </td>
        <td><p>Set the gene-set data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="setHspaResults.html">setHspaResults()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="setInitiationInfo.html">setInitiationInfo()</a></code> </p>
        </td>
        <td><p>Set initiation information</p></td>
      </tr><tr><td>
          <p><code><a href="setPcaDf.html">setPcaDf()</a></code> <code><a href="setPcaDf.html">setTsneDf()</a></code> <code><a href="setPcaDf.html">setUmapDf()</a></code> </p>
        </td>
        <td><p>Set dimensional reductions data.frames</p></td>
      </tr><tr><td>
          <p><code><a href="setPrResults.html">setPrResults()</a></code> </p>
        </td>
        <td><p>Set results of pattern recognition methods</p></td>
      </tr><tr><td>
          <p><code><a href="setSpCorResults.html">setSpCorResults()</a></code> </p>
        </td>
        <td><p>Set results of spatial correlation analysis</p></td>
      </tr><tr><td>
          <p><code><a href="set_dummy.html">set_dummy()</a></code> </p>
        </td>
        <td><p>set</p></td>
      </tr><tr><td>
          <p><code><a href="seurat_object_dummy.html">seurat_object_dummy()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="shiftTrajectoryDf.html">shiftTrajectoryDf()</a></code> </p>
        </td>
        <td><p>Shift trajectory data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="smoothSpatially.html">smoothSpatially()</a></code> </p>
        </td>
        <td><p>Smooth numeric variables spatially</p></td>
      </tr><tr><td>
          <p><code><a href="spata2-class.html">spata2-class</a></code> </p>
        </td>
        <td><p>The spata-object</p></td>
      </tr><tr><td>
          <p><code><a href="spatial_trajectory-class.html">spatial_trajectory-class</a></code> </p>
        </td>
        <td><p>spatial_trajectory object</p></td>
      </tr><tr><td>
          <p><code><a href="subsetBySegment_CountMtr.html">subsetBySegment_CountMtr()</a></code> <code><a href="subsetBySegment_CountMtr.html">subsetByBarcodes_CountMtr()</a></code> <code><a href="subsetBySegment_CountMtr.html">subsetBySegment_ExprMtr()</a></code> <code><a href="subsetBySegment_CountMtr.html">subsetByBarcodes_ExprMtr()</a></code> </p>
        </td>
        <td><p>Subset a spata-object</p></td>
      </tr><tr><td>
          <p><code><a href="surfaceFrame.html">surfaceFrame()</a></code> </p>
        </td>
        <td><p>Add frame of the sample</p></td>
      </tr><tr><td>
          <p><code><a href="tab_create_segmentation_return.html">tab_create_segmentation_return()</a></code> </p>
        </td>
        <td><p>Segmentation plot tab - return</p></td>
      </tr><tr><td>
          <p><code><a href="tab_create_trajectories_return.html">tab_create_trajectories_return()</a></code> </p>
        </td>
        <td><p>Trajectory plot tab - return</p></td>
      </tr><tr><td>
          <p><code><a href="tab_surface_plots_app.html">tab_surface_plots_app()</a></code> </p>
        </td>
        <td><p>Surface plot tab - classic</p></td>
      </tr><tr><td>
          <p><code><a href="tab_surface_plots_return.html">tab_surface_plots_return()</a></code> </p>
        </td>
        <td><p>Surface plot tab - return</p></td>
      </tr><tr><td>
          <p><code><a href="trajectory_patterns.html">trajectory_patterns</a></code> </p>
        </td>
        <td><p>Trajectory patterns</p></td>
      </tr><tr><td>
          <p><code><a href="transformSeuratToSpata.html">transformSeuratToSpata()</a></code> </p>
        </td>
        <td><p>Transform seurat-object to spata-object</p></td>
      </tr><tr><td>
          <p><code><a href="transformSpataToCDS.html">transformSpataToCDS()</a></code> </p>
        </td>
        <td><p>Transform spata-object to cell-data-set (Monocle3)</p></td>
      </tr><tr><td>
          <p><code><a href="transformSpataToSeurat.html">transformSpataToSeurat()</a></code> </p>
        </td>
        <td><p>Transform spata-object to a seurat-object</p></td>
      </tr><tr><td>
          <p><code><a href="updateSpataObject.html">updateSpataObject()</a></code> </p>
        </td>
        <td><p>Update spata-object from SPATA to SPATA2</p></td>
      </tr><tr><td>
          <p><code><a href="validActivationFunctions.html">validActivationFunctions()</a></code> <code><a href="validActivationFunctions.html">validAgglomerationMethods()</a></code> <code><a href="validActivationFunctions.html">validColorPalettes()</a></code> <code><a href="validActivationFunctions.html">validColorSpectra()</a></code> <code><a href="validActivationFunctions.html">validDeAnalysisMethods()</a></code> <code><a href="validActivationFunctions.html">validDefaultInstructionSlots()</a></code> <code><a href="validActivationFunctions.html">validDimRedMethods()</a></code> <code><a href="validActivationFunctions.html">validDirectoryInstructionSlots()</a></code> <code><a href="validActivationFunctions.html">validDistanceMethods()</a></code> <code><a href="validActivationFunctions.html">validHierarchicalClusterMethods()</a></code> <code><a href="validActivationFunctions.html">validPatternRecognitionMethods()</a></code> <code><a href="validActivationFunctions.html">validPadjMethods()</a></code> <code><a href="validActivationFunctions.html">validPlotTypes()</a></code> </p>
        </td>
        <td><p>Obtain valid argument inputs</p></td>
      </tr><tr><td>
          <p><code><a href="validateSpataObject.html">validateSpataObject()</a></code> </p>
        </td>
        <td><p>Validate a spata object</p></td>
      </tr><tr><td>
          <p><code><a href="validation.html">validation()</a></code> </p>
        </td>
        <td><p>Validate object input</p></td>
      </tr><tr><td>
          <p><code><a href="variable.html">variable()</a></code> </p>
        </td>
        <td><p>variable</p></td>
      </tr><tr><td>
          <p><code><a href="variables_num.html">variables_num()</a></code> </p>
        </td>
        <td><p>variables_num</p></td>
      </tr><tr><td>
          <p><code><a href="variable_num.html">variable_num()</a></code> </p>
        </td>
        <td><p>variable_num</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="join-information">Join Information <a href="#join-information" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>These functions access the information of interest, perform additional customized computations and join it as additional variables to already existing data.frames in a tidy-data fashion.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="joinWith.html">joinWith()</a></code> <code><a href="joinWith.html">joinWithFeatures()</a></code> <code><a href="joinWith.html">joinWithGenes()</a></code> <code><a href="joinWith.html">joinWithGeneSets()</a></code> <code><a href="joinWith.html">joinWithVariables()</a></code> </p>
        </td>
        <td><p>Join barcodes with additional variables</p></td>
      </tr><tr><td>
          <p><code><a href="joinWithPca.html">joinWithPca()</a></code> <code><a href="joinWithPca.html">joinWithTsne()</a></code> <code><a href="joinWithPca.html">joinWithUmap()</a></code> </p>
        </td>
        <td><p>Join barcodes with additional variables</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="analysis">Analysis <a href="#analysis" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="computationally-heavy-calculations">Computationally heavy calculations <a href="#computationally-heavy-calculations" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p><p>These functions perform calculations that might take some time. The results are stored inside the spata-object and can be obtained via get-functions.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="computeGeneMetaData.html">computeGeneMetaData()</a></code> <code><a href="computeGeneMetaData.html">computeGeneMetaData2()</a></code> </p>
        </td>
        <td><p>Compute gene summary statistics</p></td>
      </tr><tr><td>
          <p><code><a href="compute_pattern_similarity.html">compute_pattern_similarity()</a></code> <code><a href="compute_pattern_similarity.html">compute_pattern_relation()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="runAutoencoderAssessment.html">runAutoencoderAssessment()</a></code> <code><a href="runAutoencoderAssessment.html">assessAutoencoderOptions()</a></code> </p>
        </td>
        <td><p>Assessment of Neural Network Set Up</p></td>
      </tr><tr><td>
          <p><code><a href="runAutoencoderDenoising.html">runAutoencoderDenoising()</a></code> </p>
        </td>
        <td><p>Denoise expression matrix</p></td>
      </tr><tr><td>
          <p><code><a href="runCnvAnalysis.html">runCnvAnalysis()</a></code> </p>
        </td>
        <td><p>Identify large-scale chromosomal copy number variations</p></td>
      </tr><tr><td>
          <p><code><a href="runDeAnalysis.html">runDeAnalysis()</a></code> </p>
        </td>
        <td><p>Find differently expressed genes</p></td>
      </tr><tr><td>
          <p><code><a href="runGSEA.html">runGSEA()</a></code> </p>
        </td>
        <td><p>Compute gene set enrichment</p></td>
      </tr><tr><td>
          <p><code><a href="runHclust.html">runHclust()</a></code> <code><a href="runHclust.html">runHclustCnv()</a></code> </p>
        </td>
        <td><p>Run hierarchical clustering</p></td>
      </tr><tr><td>
          <p><code><a href="runHspa.html">runHspa()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="runPca.html">runPca()</a></code> <code><a href="runPca.html">runPca2()</a></code> </p>
        </td>
        <td><p>Run Principal Component Analysis</p></td>
      </tr><tr><td>
          <p><code><a href="runSpatialCorrelationAnalysis.html">runSpatialCorrelationAnalysis()</a></code> </p>
        </td>
        <td><p>Initiate gene clustering analysis based on spatial patterns</p></td>
      </tr><tr><td>
          <p><code><a href="runTsne.html">runTsne()</a></code> <code><a href="runTsne.html">runTsne2()</a></code> </p>
        </td>
        <td><p>Run t-Stochastic Neighbour Embedding</p></td>
      </tr><tr><td>
          <p><code><a href="runUmap.html">runUmap()</a></code> <code><a href="runUmap.html">runUmap2()</a></code> </p>
        </td>
        <td><p>Run UMAP-Analysis</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="trajectories">Trajectories <a href="#trajectories" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="assessTrajectoryTrends.html">assessTrajectoryTrends()</a></code> <code><a href="assessTrajectoryTrends.html">assessTrajectoryTrends2()</a></code> <code><a href="assessTrajectoryTrends.html">assessTrajectoryTrendsCustomized()</a></code> <code><a href="assessTrajectoryTrends.html">assessTrajectoryTrendsCustomized2()</a></code> </p>
        </td>
        <td><p>Trajectory trend analysis</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="clustering">Clustering <a href="#clustering" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="findHierarchicalCluster.html">findHierarchicalCluster()</a></code> <code><a href="findHierarchicalCluster.html">findHierarchicalClusterCnv()</a></code> </p>
        </td>
        <td><p>Cluster sample via hclust</p></td>
      </tr><tr><td>
          <p><code><a href="findMonocleClusters.html">findMonocleClusters()</a></code> </p>
        </td>
        <td><p>Cluster sample via monocle3</p></td>
      </tr><tr><td>
          <p><code><a href="findNearestNeighbourClusters.html">findNearestNeighbourClusters()</a></code> </p>
        </td>
        <td><p>Cluster sample via nearest neighbour analysis</p></td>
      </tr><tr><td>
          <p><code><a href="findSeuratClusters.html">findSeuratClusters()</a></code> </p>
        </td>
        <td><p>Cluster sample via Seurat</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="process-examine">Process &amp; Examine <a href="#process-examine" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="filterDeaDf.html">filterDeaDf()</a></code> </p>
        </td>
        <td><p>Postprocess de-analysis results</p></td>
      </tr><tr><td>
          <p><code><a href="filterTrends.html">filterTrends()</a></code> <code><a href="filterTrends.html">filterTrajectoryTrends()</a></code> </p>
        </td>
        <td><p>Filter variables of a certain trend</p></td>
      </tr><tr><td>
          <p><code><a href="examineClusterResults.html">examineClusterResults()</a></code> </p>
        </td>
        <td><p>Examine clustering results</p></td>
      </tr><tr><td>
          <p><code><a href="examineTrajectoryAssessment.html">examineTrajectoryAssessment()</a></code> </p>
        </td>
        <td><p>Examine trajectory-moddeling results</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="visualization">Visualization <a href="#visualization" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>The plot-family is again split in several subfamilies.</p>
        </th>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="surface-plots">Surface Plots <a href="#surface-plots" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p><p>Visualize the barcode-spots information with a spatial dimension.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="plotSurface.html">plotSurface()</a></code> <code><a href="plotSurface.html">plotSurface2()</a></code> <code><a href="plotSurface.html">plotSurfaceInteractive()</a></code> </p>
        </td>
        <td><p>Plot the surface of the sample</p></td>
      </tr><tr><td>
          <p><code><a href="plotSurfaceAverage.html">plotSurfaceAverage()</a></code> </p>
        </td>
        <td><p>Plot several surface plots colored by gene averages</p></td>
      </tr><tr><td>
          <p><code><a href="plotSurfaceComparison.html">plotSurfaceComparison()</a></code> <code><a href="plotSurfaceComparison.html">plotSurfaceComparison2()</a></code> </p>
        </td>
        <td><p>Plot several surface plots colored by numeric variables</p></td>
      </tr><tr><td>
          <p><code><a href="plotSurfaceHotspots.html">plotSurfaceHotspots()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="plotSurfaceQuantiles.html">plotSurfaceQuantiles()</a></code> </p>
        </td>
        <td><p>Plot a surface plot colored by binned numeric variables</p></td>
      </tr><tr><td>
          <p><code><a href="plotSegmentation.html">plotSegmentation()</a></code> </p>
        </td>
        <td><p>Plot segmentation</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="differential-expression-analysis-gsea">Differential Expression Analysis &amp; GSEA <a href="#differential-expression-analysis-gsea" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p><p>Visualize the results of differential expression nalysis</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="plotDeaDotPlot.html">plotDeaDotPlot()</a></code> </p>
        </td>
        <td><p>Plot differentially expressed genes</p></td>
      </tr><tr><td>
          <p><code><a href="plotDeaGeneCount.html">plotDeaGeneCount()</a></code> </p>
        </td>
        <td><p>Plot the number of differently expressed genes of certain groups</p></td>
      </tr><tr><td>
          <p><code><a href="plotDeaHeatmap.html">plotDeaHeatmap()</a></code> </p>
        </td>
        <td><p>Plot differentially expressed genes</p></td>
      </tr><tr><td>
          <p><code><a href="plotDeaPvalues.html">plotDeaPvalues()</a></code> <code><a href="plotDeaPvalues.html">plotDeaLogFC()</a></code> </p>
        </td>
        <td><p>Plot the p-value and log fold change distribution of de-analysis results</p></td>
      </tr><tr><td>
          <p><code><a href="plotDeaSummary.html">plotDeaSummary()</a></code> </p>
        </td>
        <td><p>Plot a summary of differential expression analysis results</p></td>
      </tr><tr><td>
          <p><code><a href="plotGseaDotPlot.html">plotGseaDotPlot()</a></code> </p>
        </td>
        <td><p>Plot gene set enrichment</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="dimensional-reduction">Dimensional Reduction <a href="#dimensional-reduction" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p><p>Visualize the barcode spots in a reduced dimension fashion.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="plotUmap.html">plotUmap()</a></code> <code><a href="plotUmap.html">plotTsne()</a></code> <code><a href="plotUmap.html">plotPca()</a></code> </p>
        </td>
        <td><p>Plot dimensional reduction</p></td>
      </tr><tr><td>
          <p><code><a href="plotPcaVariation.html">plotPcaVariation()</a></code> </p>
        </td>
        <td><p>Plot Pca Variation</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="spatial-trajectories">Spatial Trajectories <a href="#spatial-trajectories" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p><p>Visualize the barcode-spots information with a directional dimension.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="plotTrajectory.html">plotTrajectory()</a></code> </p>
        </td>
        <td><p>Plot trajectory</p></td>
      </tr><tr><td>
          <p><code><a href="plotTrajectoryFeatures.html">plotTrajectoryFeatures()</a></code> <code><a href="plotTrajectoryFeatures.html">plotTrajectoryGenes()</a></code> <code><a href="plotTrajectoryFeatures.html">plotTrajectoryGeneSets()</a></code> </p>
        </td>
        <td><p>Plot continuous trajectory dynamics in lineplots</p></td>
      </tr><tr><td>
          <p><code><a href="plotTrajectoryFeaturesDiscrete.html">plotTrajectoryFeaturesDiscrete()</a></code> </p>
        </td>
        <td><p>Plot discrete trajectory dynamics</p></td>
      </tr><tr><td>
          <p><code><a href="plotTrajectoryFit.html">plotTrajectoryFit()</a></code> <code><a href="plotTrajectoryFit.html">plotTrajectoryFitCustomized()</a></code> </p>
        </td>
        <td><p>Plot trajectory fit</p></td>
      </tr><tr><td>
          <p><code><a href="plotTrajectoryHeatmap.html">plotTrajectoryHeatmap()</a></code> </p>
        </td>
        <td><p>Plot trajectory expression dynamic in heatmap</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="statistics">Statistics <a href="#statistics" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p><p>Visualize the distribution of variables across groups or the sample and perform statistical tests.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="plotBarchart.html">plotBarchart()</a></code> </p>
        </td>
        <td><p>Plot distribution of discrete variables</p></td>
      </tr><tr><td>
          <p><code><a href="plotBoxplot.html">plotBoxplot()</a></code> <code><a href="plotBoxplot.html">plotDensityplot()</a></code> <code><a href="plotBoxplot.html">plotHistogram()</a></code> <code><a href="plotBoxplot.html">plotRidgeplot()</a></code> <code><a href="plotBoxplot.html">plotViolinplot()</a></code> </p>
        </td>
        <td><p>Plot numeric distribution and statistical tests</p></td>
      </tr><tr><td>
          <p><code><a href="plotStatisticsInteractive.html">plotStatisticsInteractive()</a></code> </p>
        </td>
        <td><p>Plot distribution of variables interactively</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="miscellaneous">Miscellaneous <a href="#miscellaneous" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p><p>Visualize a variety of aspects.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="plotCnvResults.html">plotCnvResults()</a></code> </p>
        </td>
        <td><p>Visualize copy-number-variations results</p></td>
      </tr><tr><td>
          <p><code><a href="plotDendrogram.html">plotDendrogram()</a></code> <code><a href="plotDendrogram.html">plotDendrogramCnv()</a></code> </p>
        </td>
        <td><p>Visualize hierarchical clustering results</p></td>
      </tr><tr><td>
          <p><code><a href="plotGeneMetaData.html">plotGeneMetaData()</a></code> </p>
        </td>
        <td><p>Plot gene meta data</p></td>
      </tr><tr><td>
          <p><code><a href="plotFourStates.html">plotFourStates()</a></code> <code><a href="plotFourStates.html">plotFourStates2()</a></code> </p>
        </td>
        <td><p>Gene set state plot</p></td>
      </tr><tr><td>
          <p><code><a href="plotScatterplot.html">plotScatterplot()</a></code> </p>
        </td>
        <td><p>Plot numeric variables as a scatterplot</p></td>
      </tr></tbody></table></div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Jan Kueckelhaus, Dieter-Henrik Heiland.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.1.</p>
</div>

      </footer></div>

  


  

  </body></html>

