<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Function reference â€¢ SPATA2</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-index">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SPATA2</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.0.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/spata-v2-installation.html">Installation</a>
</li>
<li>
  <a href="../articles/spata-v2-spata-data.html">SPATAData</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li class="dropdown-header">Around the spata2 object</li>
    <li>
      <a href="../articles/spata-v2-object-initiation-and-manipulation.html">Object Set Up and Manipulation</a>
    </li>
    <li>
      <a href="../articles/spata-v2-platform-compatibility.html">Platform Compatibility</a>
    </li>
    <li>
      <a href="../articles/spata-v2-autoencoder.html">Autoencoder</a>
    </li>
    <li>
      <a href="../articles/spata-v2-genes-gene-sets-and-features.html">Genes, Gene Sets and Features</a>
    </li>
    <li class="dropdown-header">Image Analysis</li>
    <li>
      <a href="../articles/spata-v2-image-handling.html">Image Handling</a>
    </li>
    <li>
      <a href="../articles/spata-v2-image-annotations.html">Image Annotations</a>
    </li>
    <li>
      <a href="../articles/spata-v2-spatial-segmentation.html">Spatial Segmentation</a>
    </li>
    <li class="dropdown-header">Analysis of Gene Expression</li>
    <li>
      <a href="../articles/spata-v2-cnv-analysis.html">Copy Number Variations</a>
    </li>
    <li>
      <a href="../articles/spata-v2-clustering.html">Clustering</a>
    </li>
    <li>
      <a href="../articles/spata-v2-dea.html">Differential Expression</a>
    </li>
    <li>
      <a href="../articles/spata-v2-gsea.html">Gene Set Enrichment</a>
    </li>
    <li class="dropdown-header">Expression Gradients &amp; Gradient Screening</li>
    <li>
      <a href="../articles/spata-v2-spatial-trajectories.html">Spatial Trajectories</a>
    </li>
    <li>
      <a href="../articles/spata-v2-spatial-trajectory-screening.html">Spatial Trajectory Screening</a>
    </li>
    <li>
      <a href="../articles/spata-v2-image-annotation-screening.html">Image Annotation Screening</a>
    </li>
    <li>
      <a href="../articles/spata-v2-scd.html">Integrate Single Cell Deconvolution</a>
    </li>
    <li class="dropdown-header">Visualization</li>
    <li>
      <a href="../articles/spata-v2-plotting-surface.html">Surface Plotting</a>
    </li>
    <li>
      <a href="../articles/spata-v2-plotting-misc.html">Miscellaneous</a>
    </li>
  </ul></li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Code &amp; Concept
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/cc-spatial-measures.html">Spatial Measures</a>
    </li>
    <li>
      <a href="../articles/cc-spatial-segmentation-vs-image-annotations.html">Spatial Segmentation vs. Image Annotations</a>
    </li>
    <li>
      <a href="../articles/cc-spatial-trajectory-screening.html">Spatial Trajectory Screening</a>
    </li>
    <li>
      <a href="../articles/cc-image-annotation-screening.html">Image Annotation Screening</a>
    </li>
    <li>
      <a href="../articles/cc-model-fitting-in-spatial-transcriptomic-studies.html">Model Fitting</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index"><colgroup><col class="alias"><col class="title"></colgroup><tbody><tr><th colspan="2">
          <h2 id="object-initiation-transformation">Object Initiation &amp; Transformation <a href="#object-initiation-transformation" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>These functions generate the spata-object you will work with throughot this and other platforms.</p>
        </th>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="object-initiation-updating">Object Initiation &amp; Updating <a href="#object-initiation-updating" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p><p>The initiation-functions create a spata-object from scratch.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="initiateSpataObject_10X.html">initiateSpataObject_10X()</a></code> </p>
        </td>
        <td><p>Initiate a <code>spata2</code> object from 10X Visium</p></td>
      </tr><tr><td>
          <p><code><a href="initiateSpataObject_CountMtr.html">initiateSpataObject_CountMtr()</a></code> </p>
        </td>
        <td><p>Initiate a <code>spata2</code> object from a raw count matrix</p></td>
      </tr><tr><td>
          <p><code><a href="initiateSpataObject_Empty.html">initiateSpataObject_Empty()</a></code> </p>
        </td>
        <td><p>Initiate an empty <code>spata2</code> object</p></td>
      </tr><tr><td>
          <p><code><a href="initiateSpataObject_Examples.html">initiateSpataObject_Examples()</a></code> </p>
        </td>
        <td><p>Initiate a <code>spata2</code> object from example data sets</p></td>
      </tr><tr><td>
          <p><code><a href="initiateSpataObject_ExprMtr.html">initiateSpataObject_ExprMtr()</a></code> </p>
        </td>
        <td><p>Initiate spata object from scaled expression matrix</p></td>
      </tr><tr><td>
          <p><code><a href="updateSpataObject.html">updateSpataObject()</a></code> </p>
        </td>
        <td><p>Update spata-object from SPATA to SPATA2</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="transformation">Transformation <a href="#transformation" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p><p>The transform-functions ensure compatibility between different platforms.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="asGiotto.html">asGiotto()</a></code> </p>
        </td>
        <td><p>Transform <code>spata2</code> object to <code>Giotto</code></p></td>
      </tr><tr><td>
          <p><code><a href="asHistologyImage.html">asHistologyImage()</a></code> </p>
        </td>
        <td><p>Convert to class <code>HistologyImage</code></p></td>
      </tr><tr><td>
          <p><code><a href="asHistologyImaging.html">asHistologyImaging()</a></code> </p>
        </td>
        <td><p>Convert to class <code>HistologyImaging</code></p></td>
      </tr><tr><td>
          <p><code><a href="asSeurat.html">asSeurat()</a></code> </p>
        </td>
        <td><p>Transform <code>SPATA2</code> to <code>Seurat</code></p></td>
      </tr><tr><td>
          <p><code><a href="asSingleCellExperiment.html">asSingleCellExperiment()</a></code> </p>
        </td>
        <td><p>Transform to <code>SingleCellExperiment</code></p></td>
      </tr><tr><td>
          <p><code><a href="asSPATA2.html">asSPATA2()</a></code> </p>
        </td>
        <td><p>Transform to <code>spata2</code> object object</p></td>
      </tr><tr><td>
          <p><code><a href="asSpatialTrajectory.html">asSpatialTrajectory()</a></code> </p>
        </td>
        <td><p>Transform to <code>SpatialTrajectory</code></p></td>
      </tr><tr><td>
          <p><code><a href="asSummarizedExperiment.html">asSummarizedExperiment()</a></code> </p>
        </td>
        <td><p>Transform to <code>SummarizedExperiment</code></p></td>
      </tr><tr><td>
          <p><code><a href="asVisiumV1.html">asVisiumV1()</a></code> </p>
        </td>
        <td><p>Transform <code>HistologyImage</code> to <code>VisiumV1</code></p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="object-loading-and-saving">Object Loading and Saving <a href="#object-loading-and-saving" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Some handy functions that make saving and loading corresponding objects more convenient.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="loadImageLowres.html">loadImage()</a></code> <code><a href="loadImageLowres.html">loadImageDefault()</a></code> <code><a href="loadImageLowres.html">loadImageHighres()</a></code> <code><a href="loadImageLowres.html">loadImageLowres()</a></code> </p>
        </td>
        <td><p>Load known images</p></td>
      </tr><tr><td>
          <p><code><a href="loadSpataObject.html">loadSpataObject()</a></code> </p>
        </td>
        <td><p>Load corresponding objects</p></td>
      </tr><tr><td>
          <p><code><a href="saveGeneSetDf.html">saveGeneSetDf()</a></code> </p>
        </td>
        <td><p>Save a gene set data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="saveSpataObject.html">saveCorrespondingCDS()</a></code> <code><a href="saveSpataObject.html">saveCorrespondingSeuratObject()</a></code> <code><a href="saveSpataObject.html">saveSpataObject()</a></code> </p>
        </td>
        <td><p>Save corresponding objects</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="object-subsetting">Object Subsetting <a href="#object-subsetting" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Create spata-objects from data subsets for more in depth analysis.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="subsetByBarcodes.html">subsetByBarcodes()</a></code> </p>
        </td>
        <td><p>Subsetting by barcodes</p></td>
      </tr><tr><td>
          <p><code><a href="subsetByGenes.html">subsetByGenes()</a></code> </p>
        </td>
        <td><p>Subset by genes</p></td>
      </tr><tr><td>
          <p><code><a href="cropSpataObject.html">cropSpataObject()</a></code> </p>
        </td>
        <td><p>Subset by x- and y-range</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="object-manipulation">Object Manipulation <a href="#object-manipulation" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="add-content">Add Content <a href="#add-content" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p><p>The add-functions let you add content to the spata-object savely.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="addAutoencoderSetUp.html">addAutoencoderSetUp()</a></code> </p>
        </td>
        <td><p>Add the set up of a neural network</p></td>
      </tr><tr><td>
          <p><code><a href="addExpressionMatrix.html">addExpressionMatrix()</a></code> </p>
        </td>
        <td><p>Add an expression matrix</p></td>
      </tr><tr><td>
          <p><code><a href="addFeatures.html">addFeatures()</a></code> </p>
        </td>
        <td><p>Add a new feature</p></td>
      </tr><tr><td>
          <p><code><a href="addGeneFeatures.html">addGeneFeatures()</a></code> </p>
        </td>
        <td><p>Add new gene features</p></td>
      </tr><tr><td>
          <p><code><a href="addGeneMetaData.html">addGeneMetaData()</a></code> </p>
        </td>
        <td><p>Add gene meta data to the object</p></td>
      </tr><tr><td>
          <p><code><a href="addGeneSet.html">addGeneSet()</a></code> <code><a href="addGeneSet.html">addGeneSetsInteractive()</a></code> </p>
        </td>
        <td><p>Add a new gene set</p></td>
      </tr><tr><td>
          <p><code><a href="addImageDir.html">addImageDir()</a></code> </p>
        </td>
        <td><p>Add individual image directories</p></td>
      </tr><tr><td>
          <p><code><a href="createImageAnnotations.html">addImageAnnotation()</a></code> <code><a href="createImageAnnotations.html">createImageAnnotations()</a></code> </p>
        </td>
        <td><p>Add image annotations</p></td>
      </tr><tr><td>
          <p><code><a href="createSpatialTrajectories.html">addSpatialTrajectory()</a></code> <code><a href="createSpatialTrajectories.html">createSpatialTrajectories()</a></code> </p>
        </td>
        <td><p>Add spatial trajectories</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="create-content-interactively">Create Content Interactively <a href="#create-content-interactively" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="createHistologyImaging.html">createHistologyImaging()</a></code> </p>
        </td>
        <td><p>Create S4 Image object for SPATA2</p></td>
      </tr><tr><td>
          <p><code><a href="createImageAnnotations.html">addImageAnnotation()</a></code> <code><a href="createImageAnnotations.html">createImageAnnotations()</a></code> </p>
        </td>
        <td><p>Add image annotations</p></td>
      </tr><tr><td>
          <p><code><a href="createSpatialSegmentation.html">createSpatialSegmentation()</a></code> </p>
        </td>
        <td><p>Interactive sample segmentation</p></td>
      </tr><tr><td>
          <p><code><a href="createSpatialTrajectories.html">addSpatialTrajectory()</a></code> <code><a href="createSpatialTrajectories.html">createSpatialTrajectories()</a></code> </p>
        </td>
        <td><p>Add spatial trajectories</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="discard-content">Discard Content <a href="#discard-content" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p><p>The discard-functions let you delete unwanted information savely.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="discardExpressionMatrix.html">discardExpressionMatrix()</a></code> </p>
        </td>
        <td><p>Discard an expression matrix</p></td>
      </tr><tr><td>
          <p><code><a href="discardFeatures.html">discardFeatures()</a></code> </p>
        </td>
        <td><p>Discard features</p></td>
      </tr><tr><td>
          <p><code><a href="discardGeneFeatures.html">discardGeneFeatures()</a></code> </p>
        </td>
        <td><p>Discard gene features</p></td>
      </tr><tr><td>
          <p><code><a href="discardGenes.html">discardGenes()</a></code> </p>
        </td>
        <td><p>Discard genes</p></td>
      </tr><tr><td>
          <p><code><a href="discardGeneSets.html">discardGeneSets()</a></code> </p>
        </td>
        <td><p>Discard gene sets</p></td>
      </tr><tr><td>
          <p><code><a href="discardImageAnnotations.html">discardImageAnnotations()</a></code> </p>
        </td>
        <td><p>Discard image annotations</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="rename-content">Rename Content <a href="#rename-content" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p><p>The rename-functions let you adjust the name of certain content with which you refer to it via arguments</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="renameFeatures.html">renameFeatures()</a></code> </p>
        </td>
        <td><p>Rename features</p></td>
      </tr><tr><td>
          <p><code><a href="renameGroups.html">renameGroups()</a></code> <code><a href="renameGroups.html">renameSegments()</a></code> </p>
        </td>
        <td><p>Rename cluster/group names</p></td>
      </tr><tr><td>
          <p><code><a href="renameImgAnn.html">renameImgAnn()</a></code> </p>
        </td>
        <td><p>Rename image annotation ID</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="set-content">Set Content <a href="#set-content" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p><p>The set-functions let you set the content of specific slots.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="setActiveExpressionMatrix.html">setActiveMatrix()</a></code> <code><a href="setActiveExpressionMatrix.html">setActiveExpressionMatrix()</a></code> </p>
        </td>
        <td><p>Denote the default expression matrix</p></td>
      </tr><tr><td>
          <p><code><a href="setAutoencoderAssessment.html">setAutoencoderAssessment()</a></code> </p>
        </td>
        <td><p>Set results of autoencoder assessment</p></td>
      </tr><tr><td>
          <p><code><a href="setBarcodes.html">setBarcodes()</a></code> </p>
        </td>
        <td><p>Set barcodes</p></td>
      </tr><tr><td>
          <p><code><a href="setCnvResults.html">setCnvResults()</a></code> </p>
        </td>
        <td><p>Set cnv-results</p></td>
      </tr><tr><td>
          <p><code><a href="setCoordsDf.html">setCoordsDf()</a></code> </p>
        </td>
        <td><p>Set the coordinates</p></td>
      </tr><tr><td>
          <p><code><a href="setCountMatrix.html">setCountMatrix()</a></code> <code><a href="setCountMatrix.html">setDenoisedMatrix()</a></code> <code><a href="setCountMatrix.html">setNormalizedMatrix()</a></code> <code><a href="setCountMatrix.html">setScaledMatrix()</a></code> </p>
        </td>
        <td><p>Set data matrices</p></td>
      </tr><tr><td>
          <p><code><a href="setDefault.html">setDefault()</a></code> </p>
        </td>
        <td><p>Set object specific default</p></td>
      </tr><tr><td>
          <p><code><a href="setDefaultGrouping.html">getDefaultGrouping()</a></code> <code><a href="setDefaultGrouping.html">setDefaultGrouping()</a></code> </p>
        </td>
        <td><p>Default grouping</p></td>
      </tr><tr><td>
          <p><code><a href="setDefaultInstructions.html">setDefaultInstructions()</a></code> <code><a href="setDefaultInstructions.html">setDirectoryInstructions()</a></code> </p>
        </td>
        <td><p>Set default instructions</p></td>
      </tr><tr><td>
          <p><code><a href="setDefaultTrajectory.html">getDefaultTrajectory()</a></code> <code><a href="setDefaultTrajectory.html">getDefaultTrajectoryId()</a></code> <code><a href="setDefaultTrajectory.html">setDefaultTrajectory()</a></code> <code><a href="setDefaultTrajectory.html">setDefaultTrajectoryId()</a></code> </p>
        </td>
        <td><p>Obtain dimensional reduction data</p></td>
      </tr><tr><td>
          <p><code><a href="setFeatureDf.html">setFeatureDf()</a></code> </p>
        </td>
        <td><p>Set feature data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="setGeneSetDf.html">setGeneSetDf()</a></code> </p>
        </td>
        <td><p>Set the gene-set data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="setImageAnnotation.html">setImageAnnotation()</a></code> <code><a href="setImageAnnotation.html">setImageAnnotations()</a></code> </p>
        </td>
        <td><p>Set image annotations</p></td>
      </tr><tr><td>
          <p><code><a href="setImageDirLowres.html">setImageDirDefault()</a></code> <code><a href="setImageDirLowres.html">setImageDirHighres()</a></code> <code><a href="setImageDirLowres.html">setImageDirLowres()</a></code> </p>
        </td>
        <td><p>Set image directories</p></td>
      </tr><tr><td>
          <p><code><a href="setImageObject.html">setImageObject()</a></code> </p>
        </td>
        <td><p>Set image object</p></td>
      </tr><tr><td>
          <p><code><a href="setImageOrigin.html">setImageOrigin()</a></code> </p>
        </td>
        <td><p>Set image origin</p></td>
      </tr><tr><td>
          <p><code><a href="setInitiationInfo.html">setInitiationInfo()</a></code> </p>
        </td>
        <td><p>Set initiation information</p></td>
      </tr><tr><td>
          <p><code><a href="setPcaDf.html">setPcaDf()</a></code> <code><a href="setPcaDf.html">setTsneDf()</a></code> <code><a href="setPcaDf.html">setUmapDf()</a></code> </p>
        </td>
        <td><p>Set dimensional reductions data.frames</p></td>
      </tr><tr><td>
          <p><code><a href="setPixelScaleFactor.html">setPixelScaleFactor()</a></code> </p>
        </td>
        <td><p>Set pixel scale factor</p></td>
      </tr><tr><td>
          <p><code><a href="setSpataDir.html">setSpataDir()</a></code> </p>
        </td>
        <td><p>Set SPATA2 directory</p></td>
      </tr><tr><td>
          <p><code><a href="setTissueOutline.html">setTissueOutline()</a></code> </p>
        </td>
        <td><p>Set tissue outline</p></td>
      </tr><tr><td>
          <p><code><a href="setTrajectory.html">setTrajectory()</a></code> <code><a href="setTrajectory.html">setTrajectories()</a></code> </p>
        </td>
        <td><p>Set trajectories</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="availability-checks">Availability Checks <a href="#availability-checks" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="containsCNV.html">containsCNV()</a></code> <code><a href="containsCNV.html">containsHistologyImage()</a></code> <code><a href="containsCNV.html">containsTissueOutline()</a></code> </p>
        </td>
        <td><p>Check availability of miscellaneous content</p></td>
      </tr><tr><td>
          <p><code><a href="containsHistologyImaging.html">containsHistologyImaging()</a></code> <code><a href="containsHistologyImaging.html">containsImageObject()</a></code> </p>
        </td>
        <td><p>Check availability of <code>HistologyImaging</code> object</p></td>
      </tr><tr><td>
          <p><code><a href="containsImage.html">containsImage()</a></code> </p>
        </td>
        <td><p>Check availability of an image</p></td>
      </tr><tr><td>
          <p><code><a href="containsPixelScaleFactor.html">containsPixelScaleFactor()</a></code> </p>
        </td>
        <td><p>Check availability of pixel scale factor</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="built-in-data">Built-in Data <a href="#built-in-data" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="clustering.html">clustering</a></code> </p>
        </td>
        <td><p>A list of data.frames that contain clustering variables for specific <code>SPATA2</code>
objects - available via the <code>SPATAData</code> package. Names of the list slots
correspond to the sample names in variable  <em>sample</em> of <code><a href="https://rdrr.io/pkg/SPATAData/man/source_df.html" class="external-link">SPATAData::source_df</a></code>.
Names of the cluster variables in the data.frames corresponds to the algorithm used.</p></td>
      </tr><tr><td>
          <p><code><a href="cnv_ref.html">cnv_ref</a></code> </p>
        </td>
        <td><p>A list of reference data for copy number variation analysis (CNV)</p></td>
      </tr><tr><td>
          <p><code><a href="cnv_regions_df.html">cnv_regions_df</a></code> </p>
        </td>
        <td><p>A data.frame necessary for cnv-analysis. Contains information about start and
end points of chromosomes.</p></td>
      </tr><tr><td>
          <p><code><a href="dea.html">dea</a></code> </p>
        </td>
        <td><p>A nested list. First layer is named by the sample name. Second layer is named
by the grouping variable. Third layer is named by the method. Contains
data.frames of differential gene expression analysis results from the
function <code><a href="https://satijalab.org/seurat/reference/FindAllMarkers.html" class="external-link">Seurat::FindAllMarkers()</a></code>. Set with <code>setDeaResultsDf()</code>.</p></td>
      </tr><tr><td>
          <p><code><a href="gene_pos_df.html">gene_pos_df</a></code> </p>
        </td>
        <td><p>A data.frame necessary for cnv-analysis. Contains information about the gene positions
on chromosomes. Contains the following variables:</p></td>
      </tr><tr><td>
          <p><code><a href="gsdf.html">gsdf</a></code> </p>
        </td>
        <td><p>The default collection of frequently used gene-sets</p></td>
      </tr><tr><td>
          <p><code><a href="image_annotations.html">image_annotations</a></code> </p>
        </td>
        <td><p>A list of lists of image annotations for specific <code>SPATA2</code>
objects - available via the <code>SPATAData</code> package. Names of the list slots
correspond to the sample names in variable  <em>sample</em> of <code><a href="https://rdrr.io/pkg/SPATAData/man/source_df.html" class="external-link">SPATAData::source_df</a></code>.</p></td>
      </tr><tr><td>
          <p><code><a href="sc_deconvolution.html">sc_deconvolution</a></code> </p>
        </td>
        <td><p>List of data.frames of single cell deconvolution. Names of
the list correspond to the sample name. Obtain data.frame
using <code>sc_deconvolution[[*&lt;sample.name&gt;*]]</code></p></td>
      </tr><tr><td>
          <p><code><a href="scale_factors.html">scale_factors</a></code> </p>
        </td>
        <td><p>List of scale factors for Visium input.</p></td>
      </tr><tr><td>
          <p><code><a href="spatial_segmentations.html">spatial_segmentations</a></code> </p>
        </td>
        <td><p>A list of lists of spatial segmentations for specific <code>SPATA2</code>
objects - available via the <code>SPATAData</code> package. Names of the list slots
correspond to the sample names in variable  <em>sample</em> of <code><a href="https://rdrr.io/pkg/SPATAData/man/source_df.html" class="external-link">SPATAData::source_df</a></code>.
Every sample specific slot is a data.frame with two variables: <em>barcodes</em>
and the spatial segmentation variable (usually named <em>histolgoy</em>).</p></td>
      </tr><tr><td>
          <p><code><a href="spatial_trajectories.html">spatial_trajectories</a></code> </p>
        </td>
        <td><p>A list of lists of spatial spatial_trajectories for specific <code>SPATA2</code>
objects - available via the <code>SPATAData</code> package. Names of the list slots
correspond to the sample names in variable  <em>sample</em> of <code><a href="https://rdrr.io/pkg/SPATAData/man/source_df.html" class="external-link">SPATAData::source_df</a></code>.</p></td>
      </tr><tr><td>
          <p><code><a href="summarize_formulas.html">summarize_formulas</a></code> </p>
        </td>
        <td><p>List of summarizing formulas</p></td>
      </tr><tr><td>
          <p><code><a href="visium_coords.html">visium_coords</a></code> </p>
        </td>
        <td><p>A data.frame that contains all barcode spots of the Visium method including
their spatial positioning. Variable names: <em>barcodes</em>, <em>col</em>, <em>row</em>, <em>imagecol</em>,
<em>imagerow</em>, <em>xlr</em>, <em>ylr</em>, <em>xhr</em>, <em>yhr</em>. lr = low resolution, hr = high resolution</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="counting">Counting <a href="#counting" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="nBarcodes.html">nBarcodes()</a></code> </p>
        </td>
        <td><p>Number of barcodes</p></td>
      </tr><tr><td>
          <p><code><a href="nCounts.html">nCounts()</a></code> </p>
        </td>
        <td><p>Number of counts</p></td>
      </tr><tr><td>
          <p><code><a href="nGenes.html">nGenes()</a></code> </p>
        </td>
        <td><p>Number of genes</p></td>
      </tr><tr><td>
          <p><code><a href="nImageAnnotations.html">nImageAnnotations()</a></code> </p>
        </td>
        <td><p>Number of image annotations</p></td>
      </tr><tr><td>
          <p><code><a href="nSpatialTrajectories.html">nSpatialTrajectories()</a></code> <code><a href="nSpatialTrajectories.html">nTrajectories()</a></code> </p>
        </td>
        <td><p>Number of spatial trajectories</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="download">Download <a href="#download" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="downloadPubExample.html">downloadPubExample()</a></code> </p>
        </td>
        <td><p>Download data from the publication</p></td>
      </tr><tr><td>
          <p><code><a href="downloadSpataObject.html">downloadSpataObject()</a></code> <code><a href="downloadSpataObject.html">downloadSpataObjects()</a></code> </p>
        </td>
        <td><p>Download <code>spata2</code> objects</p></td>
      </tr><tr><td>
          <p><code><a href="downloadRawData.html">downloadRawData()</a></code> </p>
        </td>
        <td><p>Download raw Visium output</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="extract-data">Extract Data <a href="#extract-data" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="autoencoder-related">Autoencoder related <a href="#autoencoder-related" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="getAutoencoderAssessment.html">getAutoencoderAssessment()</a></code> </p>
        </td>
        <td><p>Obtain information about the optimal neural network set up</p></td>
      </tr><tr><td>
          <p><code><a href="getAutoencoderSetUp.html">getAutoencoderSetUp()</a></code> </p>
        </td>
        <td><p>Obtain information on neural network</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="copy-number-variations">Copy Number Variations <a href="#copy-number-variations" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="getChrRegionsDf.html">getChrRegionsDf()</a></code> </p>
        </td>
        <td><p>Obtain chromosome information</p></td>
      </tr><tr><td>
          <p><code><a href="getCnvFeatureNames.html">getCnvFeatureNames()</a></code> </p>
        </td>
        <td><p>Obtain features names under which cnv-analysis results are stored.</p></td>
      </tr><tr><td>
          <p><code><a href="getCnvGenesDf.html">getCnvGenesDf()</a></code> </p>
        </td>
        <td><p>Obtain CNV results by gene</p></td>
      </tr><tr><td>
          <p><code><a href="getCnvResults.html">getCnvResults()</a></code> </p>
        </td>
        <td><p>Obtain copy-number-variations results</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="data-matrices">Data matrices <a href="#data-matrices" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="getActiveMatrixName.html">getActiveMatrixName()</a></code> <code><a href="getActiveMatrixName.html">getActiveMatrixName()</a></code> <code><a href="getActiveMatrixName.html">getActiveExpressionMatrixName()</a></code> </p>
        </td>
        <td><p>Obtain name of currently active expression matrix</p></td>
      </tr><tr><td>
          <p><code><a href="getExpressionMatrixNames.html">getExpressionMatrixNames()</a></code> <code><a href="getExpressionMatrixNames.html">getExpressionMatrixNames()</a></code> </p>
        </td>
        <td><p>Obtain names of stored expression matrices</p></td>
      </tr><tr><td>
          <p><code><a href="getMatrix.html">getMatrix()</a></code> <code><a href="getMatrix.html">getExpressionMatrix()</a></code> <code><a href="getMatrix.html">getCountMatrix()</a></code> <code><a href="getMatrix.html">getCountMatrix()</a></code> <code><a href="getMatrix.html">getExpressionMatrix()</a></code> <code><a href="getMatrix.html">getMatrix()</a></code> <code><a href="getMatrix.html">getMatrices()</a></code> </p>
        </td>
        <td><p>Obtain count and expression matrix</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="dimensional-reduction-and-spatial-coordinates">Dimensional Reduction and Spatial Coordinates <a href="#dimensional-reduction-and-spatial-coordinates" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="getBarcodesInPolygon.html">getBarcodesInPolygon()</a></code> <code><a href="getBarcodesInPolygon.html">getBarcodesInPolygonList()</a></code> </p>
        </td>
        <td><p>Obtain barcodes in polygon</p></td>
      </tr><tr><td>
          <p><code><a href="getBarcodeSpotDistance.html">getBarcodeSpotDistance()</a></code> </p>
        </td>
        <td><p>Obtain barcode spot distances</p></td>
      </tr><tr><td>
          <p><code><a href="getBarcodeSpotDistances.html">getBarcodeSpotDistances()</a></code> </p>
        </td>
        <td><p>Obtain distances between barcodes</p></td>
      </tr><tr><td>
          <p><code><a href="getCoordsCenter.html">getCoordsCenter()</a></code> </p>
        </td>
        <td><p>Obtain coordinate center</p></td>
      </tr><tr><td>
          <p><code><a href="getCoordsRange.html">getCoordsRange()</a></code> </p>
        </td>
        <td><p>Obtain coordinate range</p></td>
      </tr><tr><td>
          <p><code><a href="getDimRedDf.html">getDimRedDf()</a></code> <code><a href="getDimRedDf.html">getPcaDf()</a></code> <code><a href="getDimRedDf.html">getPcaMtr()</a></code> <code><a href="getDimRedDf.html">getUmapDf()</a></code> <code><a href="getDimRedDf.html">getTsneDf()</a></code> <code><a href="getDimRedDf.html">getDimRedDf()</a></code> <code><a href="getDimRedDf.html">getPcaDf()</a></code> <code><a href="getDimRedDf.html">getPcaMtr()</a></code> <code><a href="getDimRedDf.html">getTsneDf()</a></code> <code><a href="getDimRedDf.html">getUmapDf()</a></code> </p>
        </td>
        <td><p>Obtain dimensional reduction data</p></td>
      </tr><tr><td>
          <p><code><a href="getTissueOutlineDf.html">getTissueOutlineDf()</a></code> </p>
        </td>
        <td><p>Obtain outer barcode spots</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="features-grouping">Features &amp; Grouping <a href="#features-grouping" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="getFeatureNames.html">getFeatureNames()</a></code> <code><a href="getFeatureNames.html">getFeatureNames()</a></code> </p>
        </td>
        <td><p>Obtain feature names</p></td>
      </tr><tr><td>
          <p><code><a href="getCnvFeatureNames.html">getCnvFeatureNames()</a></code> </p>
        </td>
        <td><p>Obtain features names under which cnv-analysis results are stored.</p></td>
      </tr><tr><td>
          <p><code><a href="getFeatureDf.html">getFeatureDf()</a></code> <code><a href="getFeatureDf.html">getFeatureDf()</a></code> </p>
        </td>
        <td><p>Obtain feature data</p></td>
      </tr><tr><td>
          <p><code><a href="getFeatureValues.html">getFeatureValues()</a></code> <code><a href="getFeatureValues.html">getFeatureValues()</a></code> </p>
        </td>
        <td><p>Obtain unique categorical feature values</p></td>
      </tr><tr><td>
          <p><code><a href="getFeatureVariables.html">getFeatureVariables()</a></code> <code><a href="getFeatureVariables.html">getFeatureVariables()</a></code> </p>
        </td>
        <td><p>Obtain a feature variable</p></td>
      </tr><tr><td>
          <p><code><a href="getGroupingOptions.html">getGroupingOptions()</a></code> <code><a href="getGroupingOptions.html">getGroupingOptions()</a></code> </p>
        </td>
        <td><p>Obtain variable names that group the barcode spots</p></td>
      </tr><tr><td>
          <p><code><a href="getGroupNames.html">getGroupNames()</a></code> <code><a href="getGroupNames.html">getGroupNames()</a></code> </p>
        </td>
        <td><p>Obtain group names a grouping variable contains</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="differential-expression-analysis">Differential Expression Analysis <a href="#differential-expression-analysis" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="getDeaLfcName.html">getDeaLfcName()</a></code> </p>
        </td>
        <td><p>Obtain LFC name</p></td>
      </tr><tr><td>
          <p><code><a href="getDeaOverview.html">getDeaOverview()</a></code> <code><a href="getDeaOverview.html">getDeaOverview()</a></code> </p>
        </td>
        <td><p>Obtain info on de-analysis storage</p></td>
      </tr><tr><td>
          <p><code><a href="getDeaResultsDf.html">getDeaResultsDf()</a></code> <code><a href="getDeaResultsDf.html">getDeaGenes()</a></code> <code><a href="getDeaResultsDf.html">getDeaGenes()</a></code> <code><a href="getDeaResultsDf.html">getDeaResultsDf()</a></code> </p>
        </td>
        <td><p>Obtain de-analysis results</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="gene-set-enrichment-analysis">Gene Set Enrichment Analysis <a href="#gene-set-enrichment-analysis" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="getGseaDf.html">getGseaDf()</a></code> </p>
        </td>
        <td><p>Obtain enrichment data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="getGseaResults.html">getGseaResults()</a></code> </p>
        </td>
        <td><p>Obtain enrichment results</p></td>
      </tr><tr><td>
          <p><code><a href="getSignatureEnrichment.html">getSignatureEnrichment()</a></code> </p>
        </td>
        <td><p>Obtain signature enrichment</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="genes-gene-sets">Genes &amp; Gene-sets <a href="#genes-gene-sets" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="getGeneCounts.html">getGeneCounts()</a></code> <code><a href="getGeneCounts.html">getGeneCounts()</a></code> </p>
        </td>
        <td><p>Obtain total number of gene counts</p></td>
      </tr><tr><td>
          <p><code><a href="getGeneFeatureNames.html">getGeneFeatureNames()</a></code> <code><a href="getGeneFeatureNames.html">getGeneFeatureNames()</a></code> </p>
        </td>
        <td><p>Obtain feature names of the gene meta data</p></td>
      </tr><tr><td>
          <p><code><a href="getGeneMetaData.html">getGeneMetaData()</a></code> <code><a href="getGeneMetaData.html">getGeneMetaDf()</a></code> <code><a href="getGeneMetaData.html">getGeneMetaData()</a></code> <code><a href="getGeneMetaData.html">getGeneMetaDf()</a></code> </p>
        </td>
        <td><p>Obtain gene meta data</p></td>
      </tr><tr><td>
          <p><code><a href="getGenePosDf.html">getGenePosDf()</a></code> </p>
        </td>
        <td><p>Obtain gene information</p></td>
      </tr><tr><td>
          <p><code><a href="getGenes.html">getGenes()</a></code> <code><a href="getGenes.html">getGenesInteractive()</a></code> <code><a href="getGenes.html">getGenes()</a></code> <code><a href="getGenes.html">getGenesInteractive()</a></code> </p>
        </td>
        <td><p>Obtain gene names</p></td>
      </tr><tr><td>
          <p><code><a href="getGeneSetDf.html">getGeneSetDf()</a></code> <code><a href="getGeneSetDf.html">getGeneSetDf()</a></code> <code><a href="getGeneSetDf.html">getGeneSetList()</a></code> </p>
        </td>
        <td><p>Obtain gene set data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="getGeneSetOverview.html">getGeneSetOverview()</a></code> <code><a href="getGeneSetOverview.html">getGeneSetOverview()</a></code> </p>
        </td>
        <td><p>Overview about the current gene sets</p></td>
      </tr><tr><td>
          <p><code><a href="getGeneSets.html">getGeneSets()</a></code> <code><a href="getGeneSets.html">getGeneSetsInteractive()</a></code> <code><a href="getGeneSets.html">getGeneSets()</a></code> <code><a href="getGeneSets.html">getGeneSetsInteractive()</a></code> </p>
        </td>
        <td><p>Obtain gene set names</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="gradient-screening">Gradient Screening <a href="#gradient-screening" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="getResultsDf.html">getResultsDf()</a></code> </p>
        </td>
        <td><p>Obtain results stored in data.frames</p></td>
      </tr><tr><td>
          <p><code><a href="getResultsVec.html">getResultsVec()</a></code> </p>
        </td>
        <td><p>Obtain screening results stored in vectors</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="image">Image <a href="#image" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="getIasDf.html">getIasDf()</a></code> <code><a href="getIasDf.html">getImageAnnotationScreeningDf()</a></code> </p>
        </td>
        <td><p>Obtain image annotation screening data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="getImage.html">getImage()</a></code> <code><a href="getImage.html">getImage()</a></code> </p>
        </td>
        <td><p>Obtain histology image</p></td>
      </tr><tr><td>
          <p><code><a href="getImageAnnotation.html">getImageAnnotation()</a></code> </p>
        </td>
        <td><p>Obtain object of class <code>ImageAnnotation</code></p></td>
      </tr><tr><td>
          <p><code><a href="getImageAnnotations.html">getImageAnnotations()</a></code> </p>
        </td>
        <td><p>Obtain list of <code>ImageAnnotation</code>-objects</p></td>
      </tr><tr><td>
          <p><code><a href="getImageCenter.html">getImageCenter()</a></code> </p>
        </td>
        <td><p>Obtain image center</p></td>
      </tr><tr><td>
          <p><code><a href="getImageDf.html">getImageDf()</a></code> </p>
        </td>
        <td><p>Obtain melted image</p></td>
      </tr><tr><td>
          <p><code><a href="getImageDims.html">getImageDims()</a></code> <code><a href="getImageDims.html">getImageRange()</a></code> </p>
        </td>
        <td><p>Obtain image dimensions/ranges</p></td>
      </tr><tr><td>
          <p><code><a href="getImageDirLowres.html">getImageDir()</a></code> <code><a href="getImageDirLowres.html">getImageDirDefault()</a></code> <code><a href="getImageDirLowres.html">getImageDirectories()</a></code> <code><a href="getImageDirLowres.html">getImageDirHighres()</a></code> <code><a href="getImageDirLowres.html">getImageDirLowres()</a></code> </p>
        </td>
        <td><p>Obtain image directories</p></td>
      </tr><tr><td>
          <p><code><a href="getImageInfo.html">getImageInfo()</a></code> </p>
        </td>
        <td><p>Obatain image information</p></td>
      </tr><tr><td>
          <p><code><a href="getImageObject.html">getImageObject()</a></code> </p>
        </td>
        <td><p>Obtain object of class <code>HistologyImage</code></p></td>
      </tr><tr><td>
          <p><code><a href="getImageOrigin.html">getImageOrigin()</a></code> </p>
        </td>
        <td><p>Obtain image origin</p></td>
      </tr><tr><td>
          <p><code><a href="getImageRaster.html">getImageRaster()</a></code> <code><a href="getImageRaster.html">getImageRasterInfo()</a></code> </p>
        </td>
        <td><p>Obtain image raster-(information)</p></td>
      </tr><tr><td>
          <p><code><a href="getImageSectionsByBarcode.html">getImageSectionsByBarcode()</a></code> </p>
        </td>
        <td><p>Obtain image sections by barcode spot</p></td>
      </tr><tr><td>
          <p><code><a href="getPixelDf.html">getPixelDf()</a></code> </p>
        </td>
        <td><p>Obtain pixel data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="getPixelScaleFactor.html">getPixelScaleFactor()</a></code> </p>
        </td>
        <td><p>Obtain scale factor for pixel to Euol conversion</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="image-annotations">Image Annotations <a href="#image-annotations" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="getIasDf.html">getIasDf()</a></code> <code><a href="getIasDf.html">getImageAnnotationScreeningDf()</a></code> </p>
        </td>
        <td><p>Obtain image annotation screening data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="getImageAnnotation.html">getImageAnnotation()</a></code> </p>
        </td>
        <td><p>Obtain object of class <code>ImageAnnotation</code></p></td>
      </tr><tr><td>
          <p><code><a href="getImageAnnotations.html">getImageAnnotations()</a></code> </p>
        </td>
        <td><p>Obtain list of <code>ImageAnnotation</code>-objects</p></td>
      </tr><tr><td>
          <p><code><a href="getImgAnnArea.html">getImgAnnArea()</a></code> </p>
        </td>
        <td><p>Obtain area of image annotation</p></td>
      </tr><tr><td>
          <p><code><a href="getImgAnnBarcodes.html">getImgAnnBarcodes()</a></code> </p>
        </td>
        <td><p>Obtain barcodes by image annotation tag</p></td>
      </tr><tr><td>
          <p><code><a href="getImgAnnCenterBcsp.html">getImgAnnCenterBcsp()</a></code> </p>
        </td>
        <td><p>Obtain center barcode-spot</p></td>
      </tr><tr><td>
          <p><code><a href="getImgAnnRange.html">getImgAnnRange()</a></code> </p>
        </td>
        <td><p>Obtain image annotations range</p></td>
      </tr><tr><td>
          <p><code><a href="getImgAnnSf.html">getImgAnnSf()</a></code> </p>
        </td>
        <td><p>Obtain simple feature</p></td>
      </tr><tr><td>
          <p><code><a href="getImgAnnSummaryDf.html">getImgAnnSummaryDf()</a></code> </p>
        </td>
        <td><p>Obtain image annotation summary</p></td>
      </tr><tr><td>
          <p><code><a href="getImgAnnTags.html">getImgAnnTags()</a></code> </p>
        </td>
        <td><p>Obtain image annotations tags</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="miscellaneous">Miscellaneous <a href="#miscellaneous" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="getBarcodes.html">getBarcodes()</a></code> <code><a href="getBarcodes.html">getBarcodes()</a></code> </p>
        </td>
        <td><p>Obtain specific barcodes</p></td>
      </tr><tr><td>
          <p><code><a href="getCnvResults.html">getCnvResults()</a></code> </p>
        </td>
        <td><p>Obtain copy-number-variations results</p></td>
      </tr><tr><td>
          <p><code><a href="getDefaultInstructions.html">getDefaultInstructions()</a></code> <code><a href="getDefaultInstructions.html">getDefault()</a></code> <code><a href="getDefaultInstructions.html">getDirectoryInstructions()</a></code> <code><a href="getDefaultInstructions.html">getDefault()</a></code> <code><a href="getDefaultInstructions.html">getDefaultInstructions()</a></code> <code><a href="getDefaultInstructions.html">getDirectoryInstructions()</a></code> </p>
        </td>
        <td><p>Obtain default argument inputs</p></td>
      </tr><tr><td>
          <p><code><a href="getInitiationInfo.html">getInitiationInfo()</a></code> <code><a href="getInitiationInfo.html">getInitiationInput()</a></code> <code><a href="getInitiationInfo.html">getInitiationInfo()</a></code> <code><a href="getInitiationInfo.html">getInitiationInput()</a></code> </p>
        </td>
        <td><p>Obtain information about object initiation</p></td>
      </tr><tr><td>
          <p><code><a href="getSampleAreaSize.html">getSampleAreaSize()</a></code> </p>
        </td>
        <td><p>Obtain sample area size</p></td>
      </tr><tr><td>
          <p><code><a href="getSampleName.html">getSampleNames()</a></code> <code><a href="getSampleName.html">getSampleName()</a></code> </p>
        </td>
        <td><p>Obtain name of <code>SPATA2</code> object</p></td>
      </tr><tr><td>
          <p><code><a href="getSpataDf.html">getSpataDf()</a></code> <code><a href="getSpataDf.html">getSpataDf()</a></code> </p>
        </td>
        <td><p>Obtain a spata-data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="getSpataDir.html">getSpataDir()</a></code> </p>
        </td>
        <td><p>Obtain SPATA2 object directory</p></td>
      </tr><tr><td>
          <p><code><a href="getVariableNames.html">getVariableNames()</a></code> </p>
        </td>
        <td><p>Obtain variable names</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="spatial-measures">Spatial Measures <a href="#spatial-measures" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="getCCD.html">getCCD()</a></code> </p>
        </td>
        <td><p>Obtain center to center distance</p></td>
      </tr><tr><td>
          <p><code><a href="getSampleAreaSize.html">getSampleAreaSize()</a></code> </p>
        </td>
        <td><p>Obtain sample area size</p></td>
      </tr><tr><td>
          <p><code><a href="getSpatialMethod.html">getSpatialMethod()</a></code> </p>
        </td>
        <td><p>Obtain spatial method</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="spatial-trajectories">Spatial Trajectories <a href="#spatial-trajectories" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="getProjectionDf.html">getProjectionDf()</a></code> </p>
        </td>
        <td><p>Obtain trajectory projection</p></td>
      </tr><tr><td>
          <p><code><a href="getSpatialTrajectory.html">getSpatialTrajectory()</a></code> <code><a href="getSpatialTrajectory.html">getSpatialTrajectories()</a></code> </p>
        </td>
        <td><p>Obtain objects of class <code>SpatialTrajectory</code>.</p></td>
      </tr><tr><td>
          <p><code><a href="getSpatialTrajectoryIds.html">getSpatialTrajectoryIds()</a></code> </p>
        </td>
        <td><p>Obtain spatial trajectory IDs</p></td>
      </tr><tr><td>
          <p><code><a href="getTrajectoryDf.html">getTrajectoryDf()</a></code> <code><a href="getTrajectoryDf.html">getTrajectoryDf()</a></code> </p>
        </td>
        <td><p>Obtain a summarized trajectory data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="getTrajectoryIds.html">getTrajectoryIds()</a></code> </p>
        </td>
        <td><p>Obtain trajectory ids</p></td>
      </tr><tr><td>
          <p><code><a href="getTrajectoryLength.html">getTrajectoryLength()</a></code> <code><a href="getTrajectoryLength.html">getTrajectoryLength()</a></code> </p>
        </td>
        <td><p>Obtain the length of a trajectory</p></td>
      </tr><tr><td>
          <p><code><a href="getTrajectoryObject.html">getTrajectoryObject()</a></code> </p>
        </td>
        <td><p>Obtain trajectory object</p></td>
      </tr><tr><td>
          <p><code><a href="getTrajectorySegmentDf.html">getTrajectorySegmentDf()</a></code> <code><a href="getTrajectorySegmentDf.html">getTrajectorySegmentDf()</a></code> </p>
        </td>
        <td><p>Obtain trjectory course</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="grouping">Grouping <a href="#grouping" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="mergeGroups.html">mergeGroups()</a></code> </p>
        </td>
        <td><p>Lump groups together</p></td>
      </tr><tr><td>
          <p><code><a href="renameGroups.html">renameGroups()</a></code> <code><a href="renameGroups.html">renameSegments()</a></code> </p>
        </td>
        <td><p>Rename cluster/group names</p></td>
      </tr><tr><td>
          <p><code><a href="relevelGroups.html">relevelGroups()</a></code> </p>
        </td>
        <td><p>Relevel groups of grouping variable</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="image-annotion-related">Image Annotion related <a href="#image-annotion-related" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="bin_by_angle.html">bin_by_angle()</a></code> </p>
        </td>
        <td><p>Bin barcode-spots by angle</p></td>
      </tr><tr><td>
          <p><code><a href="bin_by_expansion.html">bin_by_expansion()</a></code> </p>
        </td>
        <td><p>Bin barcode-spots by area extension</p></td>
      </tr><tr><td>
          <p><code><a href="countImageAnnotationTags.html">countImageAnnotationTags()</a></code> </p>
        </td>
        <td><p>Count image annotation tags</p></td>
      </tr><tr><td>
          <p><code><a href="imageAnnotationToSegmentation.html">imageAnnotationToSegmentation()</a></code> </p>
        </td>
        <td><p>Convert image annotation to segmentation</p></td>
      </tr><tr><td>
          <p><code><a href="mapImageAnnotationTags.html">mapImageAnnotationTags()</a></code> </p>
        </td>
        <td><p>Image annotation and barcode intersection</p></td>
      </tr><tr><td>
          <p><code><a href="relateToImageAnnotation.html">relateToImageAnnotation()</a></code> </p>
        </td>
        <td><p>Relate observations to an image annotation</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="image-handling-and-image-coordinate-alignment">Image handling and Image-Coordinate Alignment <a href="#image-handling-and-image-coordinate-alignment" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="alignImageAnnotation.html">alignImageAnnotation()</a></code> <code><a href="alignImageAnnotation.html">alignSpatialTrajectory()</a></code> </p>
        </td>
        <td><p>Align image annotation</p></td>
      </tr><tr><td>
          <p><code><a href="exchangeImage.html">exchangeImage()</a></code> </p>
        </td>
        <td><p>Exchange image</p></td>
      </tr><tr><td>
          <p><code><a href="flipAll.html">flipAll()</a></code> <code><a href="flipAll.html">flipCoordinates()</a></code> <code><a href="flipAll.html">flipCoordsDf()</a></code> <code><a href="flipAll.html">flipImage()</a></code> <code><a href="flipAll.html">flipImageAnnotations()</a></code> <code><a href="flipAll.html">flipSpatialTrajectories()</a></code> </p>
        </td>
        <td><p>Mirror invert image and coordinates</p></td>
      </tr><tr><td>
          <p><code><a href="resetImageJustification.html">resetImageJustification()</a></code> </p>
        </td>
        <td><p>Reset image justification</p></td>
      </tr><tr><td>
          <p><code><a href="rotateAll.html">rotateAll()</a></code> <code><a href="rotateAll.html">rotateImage()</a></code> <code><a href="rotateAll.html">rotateCoordinates()</a></code> <code><a href="rotateAll.html">rotateCoordsDf()</a></code> <code><a href="rotateAll.html">rotateImageAnnotations()</a></code> <code><a href="rotateAll.html">rotateSpatialTrajectories()</a></code> </p>
        </td>
        <td><p>Rotate image and coordinates</p></td>
      </tr><tr><td>
          <p><code><a href="scaleAll.html">scaleAll()</a></code> <code><a href="scaleAll.html">scaleImage()</a></code> <code><a href="scaleAll.html">scaleCoordinates()</a></code> <code><a href="scaleAll.html">scaleCoordsDf()</a></code> <code><a href="scaleAll.html">scaleImageAnnotations()</a></code> <code><a href="scaleAll.html">scaleSpatialTrajectories()</a></code> </p>
        </td>
        <td><p>Scale image and coordinates</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="information-summaries">Information &amp; Summaries <a href="#information-summaries" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Some handy functions that quickly give you an overwiew about the spata-objectâ€™s content and the progress you have made.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="printAutoencoderSummary.html">printAutoencoderSummary()</a></code> </p>
        </td>
        <td><p>Print autoencoder summary</p></td>
      </tr><tr><td>
          <p><code><a href="printDeaOverview.html">printDeaOverview()</a></code> </p>
        </td>
        <td><p>Print overview of all conducted de-analysis</p></td>
      </tr><tr><td>
          <p><code><a href="printDefaultInstructions.html">printDefaultInstructions()</a></code> </p>
        </td>
        <td><p>Print current default settings</p></td>
      </tr><tr><td>
          <p><code><a href="printGeneSetOverview.html">printGeneSetOverview()</a></code> </p>
        </td>
        <td><p>Print overview about the current gene sets</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="join-information">Join Information <a href="#join-information" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>These functions access the information of interest, perform additional customized computations and join it as additional variables to already existing data.frames in a tidy-data fashion.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="joinWith.html">joinWith()</a></code> <code><a href="joinWith.html">joinWithFeatures()</a></code> <code><a href="joinWith.html">joinWithGenes()</a></code> <code><a href="joinWith.html">joinWithGeneSets()</a></code> <code><a href="joinWith.html">joinWithVariables()</a></code> </p>
        </td>
        <td><p>Join barcodes with additional variables</p></td>
      </tr><tr><td>
          <p><code><a href="joinWithPca.html">joinWithPca()</a></code> <code><a href="joinWithPca.html">joinWithTsne()</a></code> <code><a href="joinWithPca.html">joinWithUmap()</a></code> </p>
        </td>
        <td><p>Join barcodes with additional variables</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="miscellaneous-algorithms">Miscellaneous Algorithms <a href="#miscellaneous-algorithms" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="computationally-heavy-calculations">Computationally heavy calculations <a href="#computationally-heavy-calculations" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p><p>These functions perform calculations that might take some time. The results are stored inside the spata-object and can be obtained via get-functions.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="computeCnvByChrArm.html">computeCnvByChrArm()</a></code> </p>
        </td>
        <td><p>Compute CNV by chromosome arm</p></td>
      </tr><tr><td>
          <p><code><a href="computeGeneMetaData.html">computeGeneMetaData()</a></code> <code><a href="computeGeneMetaData.html">computeGeneMetaData2()</a></code> </p>
        </td>
        <td><p>Compute gene summary statistics</p></td>
      </tr><tr><td>
          <p><code><a href="runAutoencoderAssessment.html">assessAutoencoderOptions()</a></code> <code><a href="runAutoencoderAssessment.html">runAutoencoderAssessment()</a></code> </p>
        </td>
        <td><p>Assessment of Neural Network Set Up</p></td>
      </tr><tr><td>
          <p><code><a href="runAutoencoderDenoising.html">runAutoencoderDenoising()</a></code> </p>
        </td>
        <td><p>Denoise expression matrix</p></td>
      </tr><tr><td>
          <p><code><a href="runBayesSpaceClustering.html">runBayesSpaceClustering()</a></code> </p>
        </td>
        <td><p>Clustering with BayesSpace</p></td>
      </tr><tr><td>
          <p><code><a href="runCnvAnalysis.html">runCnvAnalysis()</a></code> </p>
        </td>
        <td><p>Identify large-scale chromosomal copy number variations</p></td>
      </tr><tr><td>
          <p><code><a href="runDEA.html">runDEA()</a></code> <code><a href="runDEA.html">runDeAnalysis()</a></code> </p>
        </td>
        <td><p>Find differently expressed genes</p></td>
      </tr><tr><td>
          <p><code><a href="runGSEA.html">runGSEA()</a></code> </p>
        </td>
        <td><p>Compute gene set enrichment</p></td>
      </tr><tr><td>
          <p><code><a href="runPca.html">runPca()</a></code> <code><a href="runPca.html">runPca2()</a></code> </p>
        </td>
        <td><p>Run Principal Component Analysis</p></td>
      </tr><tr><td>
          <p><code><a href="runSeuratClustering.html">runSeuratClustering()</a></code> </p>
        </td>
        <td><p>Clustering with Seurat</p></td>
      </tr><tr><td>
          <p><code><a href="runSparkx.html">getSparkxGeneDf()</a></code> <code><a href="runSparkx.html">getSparkxGenes()</a></code> <code><a href="runSparkx.html">getSparkxResults()</a></code> <code><a href="runSparkx.html">runSparkx()</a></code> </p>
        </td>
        <td><p>Identify genes of interest with SPARKX</p></td>
      </tr><tr><td>
          <p><code><a href="runTsne.html">runTsne()</a></code> <code><a href="runTsne.html">runTsne2()</a></code> </p>
        </td>
        <td><p>Run t-Stochastic Neighbour Embedding</p></td>
      </tr><tr><td>
          <p><code><a href="runUmap.html">runUmap()</a></code> <code><a href="runUmap.html">runUmap2()</a></code> </p>
        </td>
        <td><p>Run UMAP-Analysis</p></td>
      </tr><tr><td>
          <p><code><a href="findMonocleClusters.html">findMonocleClusters()</a></code> </p>
        </td>
        <td><p>Cluster sample via monocle3</p></td>
      </tr><tr><td>
          <p><code><a href="findNearestNeighbourClusters.html">findNearestNeighbourClusters()</a></code> </p>
        </td>
        <td><p>Cluster sample via nearest neighbour analysis</p></td>
      </tr><tr><td>
          <p><code><a href="findSeuratClusters.html">findSeuratClusters()</a></code> </p>
        </td>
        <td><p>Cluster sample via Seurat</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="process-examine">Process &amp; Examine <a href="#process-examine" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="filterDeaDf.html">filterDeaDf()</a></code> </p>
        </td>
        <td><p>Postprocess de-analysis results</p></td>
      </tr><tr><td>
          <p><code><a href="examineClusterResults.html">examineClusterResults()</a></code> </p>
        </td>
        <td><p>Examine clustering results</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="miscellaneous-functions">Miscellaneous Functions <a href="#miscellaneous-functions" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="add_outline_variable.html">add_outline_variable()</a></code> </p>
        </td>
        <td><p>Add outline variable</p></td>
      </tr><tr><td>
          <p><code><a href="add_dbscan_variable.html">add_dbscan_variable()</a></code> <code><a href="add_dbscan_variable.html">add_tissue_section_variable()</a></code> </p>
        </td>
        <td><p>Add tissue section variable</p></td>
      </tr><tr><td>
          <p><code><a href="all_bcsp_distances.html">all_bcsp_distances()</a></code> </p>
        </td>
        <td><p>Obtain a all barcode-spots distances</p></td>
      </tr><tr><td>
          <p><code><a href="arrange_as_polygon.html">arrange_as_polygon()</a></code> </p>
        </td>
        <td><p>Arrange observations as polygon</p></td>
      </tr><tr><td>
          <p><code><a href="buffer_area.html">buffer_area()</a></code> </p>
        </td>
        <td><p>Buffer area</p></td>
      </tr><tr><td>
          <p><code><a href="create_model_df.html">create_model_df()</a></code> </p>
        </td>
        <td><p>Create model data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="include_tissue_outline.html">include_tissue_outline()</a></code> </p>
        </td>
        <td><p>Include spatial extent of tissue sections in analysis</p></td>
      </tr><tr><td>
          <p><code><a href="intersect_polygons.html">intersect_polygons()</a></code> </p>
        </td>
        <td><p>Test polygon intersection</p></td>
      </tr><tr><td>
          <p><code><a href="validActivationFunctions.html">validActivationFunctions()</a></code> <code><a href="validActivationFunctions.html">validAgglomerationMethods()</a></code> <code><a href="validActivationFunctions.html">validAlluvialTypes()</a></code> <code><a href="validActivationFunctions.html">validColorPalettes()</a></code> <code><a href="validActivationFunctions.html">validColorSpectra()</a></code> <code><a href="validActivationFunctions.html">validDeAnalysisMethods()</a></code> <code><a href="validActivationFunctions.html">validDefaultInstructionSlots()</a></code> <code><a href="validActivationFunctions.html">validDimRedMethods()</a></code> <code><a href="validActivationFunctions.html">validDirectoryInstructionSlots()</a></code> <code><a href="validActivationFunctions.html">validDistanceMethods()</a></code> <code><a href="validActivationFunctions.html">validHierarchicalClusterMethods()</a></code> <code><a href="validActivationFunctions.html">validImageClasses()</a></code> <code><a href="validActivationFunctions.html">validModelNames()</a></code> <code><a href="validActivationFunctions.html">validPatternRecognitionMethods()</a></code> <code><a href="validActivationFunctions.html">validPadjMethods()</a></code> <code><a href="validActivationFunctions.html">validPlotTypes()</a></code> <code><a href="validActivationFunctions.html">validPubExamples()</a></code> <code><a href="validActivationFunctions.html">validSpatialMethods()</a></code> <code><a href="validActivationFunctions.html">validTrajectoryTrends()</a></code> <code><a href="validActivationFunctions.html">validUnits()</a></code> <code><a href="validActivationFunctions.html">validUnitsOfArea()</a></code> <code><a href="validActivationFunctions.html">validUnitsOfAreaSI()</a></code> <code><a href="validActivationFunctions.html">validUnitsOfLength()</a></code> <code><a href="validActivationFunctions.html">validUnitsOfLengthSI()</a></code> <code><a href="validActivationFunctions.html">validEuropeanUnitsOfLength()</a></code> </p>
        </td>
        <td><p>Obtain valid argument inputs</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="spatial-gradient-screening">Spatial Gradient Screening <a href="#spatial-gradient-screening" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Functions around Spatial Trajectory Screening and Image Annotation Screening.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="gradientToModelIAS.html">gradientToModelIAS()</a></code> <code><a href="gradientToModelIAS.html">gradientToModelSTS()</a></code> </p>
        </td>
        <td><p>Create input for <code>model_add</code></p></td>
      </tr><tr><td>
          <p><code><a href="imageAnnotationScreening.html">imageAnnotationScreening()</a></code> </p>
        </td>
        <td><p>Implementation of the IAS-algorithm</p></td>
      </tr><tr><td>
          <p><code><a href="inferSingleCellGradient.html">inferSingleCellGradient()</a></code> </p>
        </td>
        <td><p>Count cells depending on distance to image annotation</p></td>
      </tr><tr><td>
          <p><code><a href="spatialTrajectoryScreening.html">spatialTrajectoryScreening()</a></code> </p>
        </td>
        <td><p>The Spatial Trajectory Screening algorithm</p></td>
      </tr><tr><td>
          <p><code><a href="getStsDf.html">getStsDf()</a></code> </p>
        </td>
        <td><p>Obtain spatial trajectory screening data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="getIasBinAreas.html">getIasBinAreas()</a></code> </p>
        </td>
        <td><p>Calculate IAS bin area</p></td>
      </tr><tr><td>
          <p><code><a href="getIasDf.html">getIasDf()</a></code> <code><a href="getIasDf.html">getImageAnnotationScreeningDf()</a></code> </p>
        </td>
        <td><p>Obtain image annotation screening data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="getIasExpansion.html">getIasExpansion()</a></code> </p>
        </td>
        <td><p>Obtain expanded Image Annotation polygons</p></td>
      </tr><tr><td>
          <p><code><a href="summarizeIAS.html">summarizeIAS()</a></code> </p>
        </td>
        <td><p>Summarize IAS-results</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="spatial-measures">Spatial Measures <a href="#spatial-measures" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="are_all_area_or_dist.html">are_all_area_or_dist()</a></code> </p>
        </td>
        <td><p>Test area or distance input</p></td>
      </tr><tr><td>
          <p><code><a href="as_unit.html">as_meter()</a></code> <code><a href="as_unit.html">as_meter2()</a></code> <code><a href="as_unit.html">as_micrometer()</a></code> <code><a href="as_unit.html">as_micrometer2()</a></code> <code><a href="as_unit.html">as_millimeter()</a></code> <code><a href="as_unit.html">as_millimeter2()</a></code> <code><a href="as_unit.html">as_nanometer()</a></code> <code><a href="as_unit.html">as_nanometer2()</a></code> <code><a href="as_unit.html">as_pixel()</a></code> <code><a href="as_unit.html">as_unit()</a></code> <code><a href="as_unit.html">as_centimeter()</a></code> <code><a href="as_unit.html">as_centimeter2()</a></code> <code><a href="as_unit.html">as_decimeter()</a></code> <code><a href="as_unit.html">as_decimeter2()</a></code> </p>
        </td>
        <td><p>Transform distance and area values</p></td>
      </tr><tr><td>
          <p><code><a href="as_SPATA2_dist.html">as_SPATA2_dist()</a></code> </p>
        </td>
        <td><p>Distance transformation</p></td>
      </tr><tr><td>
          <p><code><a href="extract_unit.html">extract_unit()</a></code> </p>
        </td>
        <td><p>Extract distance units</p></td>
      </tr><tr><td>
          <p><code><a href="extract_value.html">extract_value()</a></code> </p>
        </td>
        <td><p>Extract distance value</p></td>
      </tr><tr><td>
          <p><code><a href="is_area.html">is_area()</a></code> <code><a href="is_area.html">is_area_pixel()</a></code> <code><a href="is_area.html">is_area_si()</a></code> </p>
        </td>
        <td><p>Test area input</p></td>
      </tr><tr><td>
          <p><code><a href="is_dist.html">is_dist()</a></code> <code><a href="is_dist.html">is_dist_si()</a></code> <code><a href="is_dist.html">is_dist_pixel()</a></code> </p>
        </td>
        <td><p>Test distance input</p></td>
      </tr><tr><td>
          <p><code><a href="is_unit_area.html">is_unit_area()</a></code> </p>
        </td>
        <td><p>Test unit of area input</p></td>
      </tr><tr><td>
          <p><code><a href="is_unit_dist.html">is_unit_dist()</a></code> </p>
        </td>
        <td><p>Test unit of length input</p></td>
      </tr><tr><td>
          <p><code><a href="transform_area_si_to_pixel.html">transform_area_si_to_pixel()</a></code> <code><a href="transform_area_si_to_pixel.html">transform_area_si_to_pixels()</a></code> </p>
        </td>
        <td><p>Convert area in SI units to pixel</p></td>
      </tr><tr><td>
          <p><code><a href="transform_dist_si_to_pixel.html">transform_dist_si_to_pixel()</a></code> <code><a href="transform_dist_si_to_pixel.html">transform_dist_si_to_pixels()</a></code> </p>
        </td>
        <td><p>Convert from European Units of Length to pixels</p></td>
      </tr><tr><td>
          <p><code><a href="transform_pixel_to_area_si.html">transform_pixel_to_area_si()</a></code> </p>
        </td>
        <td><p>Converts from pixel to area in SI units</p></td>
      </tr><tr><td>
          <p><code><a href="transform_pixel_to_dist_si.html">transform_pixel_to_dist_si()</a></code> <code><a href="transform_pixel_to_dist_si.html">transform_pixels_to_dist_si()</a></code> </p>
        </td>
        <td><p>Convert from pixels to European units of length</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="s-classes">S4 Classes <a href="#s-classes" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="HistologyImaging-class.html">HistologyImaging-class</a></code> <code><a href="HistologyImaging-class.html">HistologyImaging</a></code> </p>
        </td>
        <td><p>The <code>HistologyImaging</code> - Class</p></td>
      </tr><tr><td>
          <p><code><a href="ImageAnnotation-class.html">ImageAnnotation-class</a></code> <code><a href="ImageAnnotation-class.html">ImageAnnotation</a></code> </p>
        </td>
        <td><p>The <code>ImageAnnotation</code> - Class</p></td>
      </tr><tr><td>
          <p><code><a href="ImageAnnotationScreening-class.html">ImageAnnotationScreening-class</a></code> <code><a href="ImageAnnotationScreening-class.html">ImageAnnotationScreening</a></code> </p>
        </td>
        <td><p>The <code>ImageAnnotationScreening</code> - Class</p></td>
      </tr><tr><td>
          <p><code><a href="spata2-class.html">spata2-class</a></code> <code><a href="spata2-class.html">spata2</a></code> </p>
        </td>
        <td><p>The <code>spata2</code>- Class</p></td>
      </tr><tr><td>
          <p><code><a href="SpatialMethod-class.html">SpatialMethod-class</a></code> <code><a href="SpatialMethod-class.html">SpatialMethod</a></code> </p>
        </td>
        <td><p>The <code>SpatialMethod</code> - Class</p></td>
      </tr><tr><td>
          <p><code><a href="SpatialTrajectory-class.html">SpatialTrajectory-class</a></code> <code><a href="SpatialTrajectory-class.html">SpatialTrajectory</a></code> </p>
        </td>
        <td><p>The <code>SpatialTrajectory</code> - Class</p></td>
      </tr><tr><td>
          <p><code><a href="SpatialTrajectoryScreening-class.html">SpatialTrajectoryScreening-class</a></code> <code><a href="SpatialTrajectoryScreening-class.html">SpatialTrajectoryScreening</a></code> </p>
        </td>
        <td><p>The <code>SpatialTrajectoryScreening</code> - class</p></td>
      </tr><tr><td>
          <p><code><a href="Trajectory-class.html">Trajectory-class</a></code> <code><a href="Trajectory-class.html">Trajectory</a></code> </p>
        </td>
        <td><p>The <code>Trajectory</code> - Class</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="visualization">Visualization <a href="#visualization" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="autoencoder-related">Autoencoder related <a href="#autoencoder-related" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="plotAutoencoderAssessment.html">plotAutoencoderAssessment()</a></code> </p>
        </td>
        <td><p>Plot total variance of different neural networks</p></td>
      </tr><tr><td>
          <p><code><a href="plotAutoencoderResults.html">plotAutoencoderResults()</a></code> </p>
        </td>
        <td><p>Plot scaled vs. denoised expression</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="copy-number-variation">Copy Number Variation <a href="#copy-number-variation" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="plotCnvHeatmap.html">plotCnvHeatmap()</a></code> </p>
        </td>
        <td><p>Plot CNV Heatmap</p></td>
      </tr><tr><td>
          <p><code><a href="plotCnvLineplot.html">plotCnvLineplot()</a></code> </p>
        </td>
        <td><p>Plot CNV Lineplot</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="differential-expression-analysis-gsea">Differential Expression Analysis &amp; GSEA <a href="#differential-expression-analysis-gsea" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p><p>Visualize the results of differential expression nalysis</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="plotDeaDotPlot.html">plotDeaDotPlot()</a></code> </p>
        </td>
        <td><p>Plot differentially expressed genes</p></td>
      </tr><tr><td>
          <p><code><a href="plotDeaGeneCount.html">plotDeaGeneCount()</a></code> </p>
        </td>
        <td><p>Plot the number of differently expressed genes of certain groups</p></td>
      </tr><tr><td>
          <p><code><a href="plotDeaHeatmap.html">plotDeaHeatmap()</a></code> </p>
        </td>
        <td><p>Plot differentially expressed genes</p></td>
      </tr><tr><td>
          <p><code><a href="plotDeaPvalues.html">plotDeaLogFC()</a></code> <code><a href="plotDeaPvalues.html">plotDeaPvalues()</a></code> </p>
        </td>
        <td><p>Plot the p-value and log fold change distribution of de-analysis results</p></td>
      </tr><tr><td>
          <p><code><a href="plotDeaSummary.html">plotDeaSummary()</a></code> </p>
        </td>
        <td><p>Plot a summary of differential expression analysis results</p></td>
      </tr><tr><td>
          <p><code><a href="plotDeaVolcano.html">plotDeaVolcano()</a></code> <code><a href="plotDeaVolcano.html">plotDeaVolcano1v1()</a></code> </p>
        </td>
        <td><p>Plot gene expression testing results</p></td>
      </tr><tr><td>
          <p><code><a href="plotGseaDotPlot.html">plotGseaDotPlot()</a></code> </p>
        </td>
        <td><p>Plot gene set enrichment</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="dimensional-reduction">Dimensional Reduction <a href="#dimensional-reduction" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p><p>Visualize the barcode spots in a reduced dimension fashion.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="plotUmap.html">plotPca()</a></code> <code><a href="plotUmap.html">plotTsne()</a></code> <code><a href="plotUmap.html">plotTsneComparison()</a></code> <code><a href="plotUmap.html">plotUmap()</a></code> <code><a href="plotUmap.html">plotUmapComparison()</a></code> </p>
        </td>
        <td><p>Plot dimensional reduction</p></td>
      </tr><tr><td>
          <p><code><a href="plotPcaVariation.html">plotPcaVariation()</a></code> </p>
        </td>
        <td><p>Plot Pca Variation</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="ggplayer-and-other-gg-objects">ggpLayer and other gg objects <a href="#ggplayer-and-other-gg-objects" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p><p>Add ggroto objects to ggplot2 plots.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="ggpInit.html">ggpInit()</a></code> </p>
        </td>
        <td><p>Initiate ggplot2 layering</p></td>
      </tr><tr><td>
          <p><code><a href="ggpLayerAxesClean.html">ggpLayerAxesClean()</a></code> </p>
        </td>
        <td><p>Display clean axes</p></td>
      </tr><tr><td>
          <p><code><a href="ggpLayerAxesSI.html">ggpLayerAxesSI()</a></code> </p>
        </td>
        <td><p>Display axes with SI units of length</p></td>
      </tr><tr><td>
          <p><code><a href="ggpLayerColorGroupScale.html">ggpLayerColorGroupScale()</a></code> </p>
        </td>
        <td><p>Add group specific color spectrum</p></td>
      </tr><tr><td>
          <p><code><a href="ggpLayerEncirclingIAS.html">ggpLayerEncirclingIAS()</a></code> </p>
        </td>
        <td><p>Add IAS area expansion</p></td>
      </tr><tr><td>
          <p><code><a href="ggpLayerFrame.html">ggpLayerFrame()</a></code> <code><a href="ggpLayerFrame.html">ggpLayerFixFrame()</a></code> </p>
        </td>
        <td><p>Fix ggplot frame</p></td>
      </tr><tr><td>
          <p><code><a href="ggpLayerFrameByCoords.html">ggpLayerFrameByCoords()</a></code> <code><a href="ggpLayerFrameByCoords.html">ggpLayerFrameByImage()</a></code> </p>
        </td>
        <td><p>Set plot limits</p></td>
      </tr><tr><td>
          <p><code><a href="ggpLayerGroupOutline.html">ggpLayerGroupOutline()</a></code> </p>
        </td>
        <td><p>Add group outline</p></td>
      </tr><tr><td>
          <p><code><a href="ggpLayerHorizonIAS.html">ggpLayerHorizonIAS()</a></code> </p>
        </td>
        <td><p>Add IAS area horizon</p></td>
      </tr><tr><td>
          <p><code><a href="ggpLayerImage.html">ggpLayerImage()</a></code> </p>
        </td>
        <td><p>Add histology image</p></td>
      </tr><tr><td>
          <p><code><a href="ggpLayerImgAnnPointer.html">ggpLayerImgAnnPointer()</a></code> </p>
        </td>
        <td><p>Add pointer towards image annotations</p></td>
      </tr><tr><td>
          <p><code><a href="ggpLayerImgAnnRect.html">ggpLayerImgAnnRect()</a></code> </p>
        </td>
        <td><p>Add a rectangular around an image annotation</p></td>
      </tr><tr><td>
          <p><code><a href="ggpLayerLineplotAid.html">ggpLayerLineplotAid()</a></code> </p>
        </td>
        <td><p>Add horizontal and vertical lines</p></td>
      </tr><tr><td>
          <p><code><a href="ggpLayerRect.html">ggpLayerRect()</a></code> </p>
        </td>
        <td><p>Add a rectangular to the plot</p></td>
      </tr><tr><td>
          <p><code><a href="ggpLayerScaleBarSI.html">ggpLayerScaleBarSI()</a></code> </p>
        </td>
        <td><p>Add a scale bar in SI units</p></td>
      </tr><tr><td>
          <p><code><a href="ggpLayerThemeCoords.html">ggpLayerThemeCoords()</a></code> </p>
        </td>
        <td><p>Add coordinates theme</p></td>
      </tr><tr><td>
          <p><code><a href="ggpLayerTissueOutline.html">ggpLayerTissueOutline()</a></code> </p>
        </td>
        <td><p>Add a hull that outlines the tissue</p></td>
      </tr><tr><td>
          <p><code><a href="ggpLayerTrajectories.html">ggpLayerTrajectories()</a></code> </p>
        </td>
        <td><p>Add trajectory layer</p></td>
      </tr><tr><td>
          <p><code><a href="ggpLayerZoom.html">ggpLayerZoom()</a></code> </p>
        </td>
        <td><p>Set plot limits manually</p></td>
      </tr><tr><td>
          <p><code><a href="theme_lineplot_gradient.html">theme_lineplot_gradient()</a></code> <code><a href="theme_lineplot_gradient.html">theme_ridgeplot_gradient()</a></code> <code><a href="theme_lineplot_gradient.html">theme_trajectory_fit()</a></code> </p>
        </td>
        <td><p>ggplot2 themes</p></td>
      </tr><tr><td>
          <p><code><a href="legendBottom.html">legendBottom()</a></code> <code><a href="legendBottom.html">legendColor()</a></code> <code><a href="legendBottom.html">legendLeft()</a></code> <code><a href="legendBottom.html">legendNone()</a></code> <code><a href="legendBottom.html">legendRight()</a></code> <code><a href="legendBottom.html">legendTop()</a></code> </p>
        </td>
        <td><p>ggplot2 legend manipulation</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="geoms">geoms <a href="#geoms" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="geom_point_fixed.html">geom_point_fixed()</a></code> </p>
        </td>
        <td><p>Points (fixed size ~ window ratio)</p></td>
      </tr><tr><td>
          <p><code><a href="geom_segment_fixed.html">geom_segment_fixed()</a></code> </p>
        </td>
        <td><p>Segments (fixed size ~ window ratio)</p></td>
      </tr><tr><td>
          <p><code><a href="geom_text_fixed.html">geom_text_fixed()</a></code> </p>
        </td>
        <td><p>Text (fixed size ~ window ratio)</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="image-image-annotations">Image &amp; Image Annotations <a href="#image-image-annotations" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="addPointsBase.html">addPointsBase()</a></code> </p>
        </td>
        <td><p>Add points to base surface plot</p></td>
      </tr><tr><td>
          <p><code><a href="plotImage.html">plotImage()</a></code> </p>
        </td>
        <td><p>Plot histology image</p></td>
      </tr><tr><td>
          <p><code><a href="plotImageGgplot.html">plotImageGgplot()</a></code> </p>
        </td>
        <td><p>Plot histology image (ggplot2)</p></td>
      </tr><tr><td>
          <p><code><a href="plotImagesGgplot.html">plotImagesGgplot()</a></code> </p>
        </td>
        <td><p>Plot histology images (ggplot2)</p></td>
      </tr><tr><td>
          <p><code><a href="plotImageAnnotations.html">plotImageAnnotations()</a></code> </p>
        </td>
        <td><p>Plot image annotations</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="miscellaneous">Miscellaneous <a href="#miscellaneous" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p><p>Visualize a variety of aspects.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="plotFourStates.html">plotFourStates()</a></code> <code><a href="plotFourStates.html">plotFourStates2()</a></code> </p>
        </td>
        <td><p>Plot state plot</p></td>
      </tr><tr><td>
          <p><code><a href="plotScatterplot.html">plotScatterplot()</a></code> </p>
        </td>
        <td><p>Plot numeric variables as a scatterplot</p></td>
      </tr><tr><td>
          <p><code><a href="showColors.html">showColors()</a></code> <code><a href="showColors.html">showColorPalettes()</a></code> <code><a href="showColors.html">showColorSpectra()</a></code> </p>
        </td>
        <td><p>Show color palettes and spectra</p></td>
      </tr><tr><td>
          <p><code><a href="showModels.html">showModels()</a></code> </p>
        </td>
        <td><p>Show spatial gradient screening models</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="spatial-gradient">Spatial Gradient <a href="#spatial-gradient" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="plotClockplot.html">plotClockplot()</a></code> </p>
        </td>
        <td><p>Plota clockplot</p></td>
      </tr><tr><td>
          <p><code><a href="plotIasBarplot.html">plotIasBarplot()</a></code> <code><a href="plotIasBarplot.html">plotIasBarplotSC()</a></code> </p>
        </td>
        <td><p>Plot IAS barplot</p></td>
      </tr><tr><td>
          <p><code><a href="plotIasEvaluation.html">plotIasEvaluation()</a></code> </p>
        </td>
        <td><p>Plot IAS evaluation per variable-model pair</p></td>
      </tr><tr><td>
          <p><code><a href="plotIasHeatmap.html">plotIasHeatmap()</a></code> </p>
        </td>
        <td><p>Plot IAS heatmap</p></td>
      </tr><tr><td>
          <p><code><a href="plotIasLineplot.html">plotIasLineplot()</a></code> <code><a href="plotIasLineplot.html">plotIasLineplotSC()</a></code> </p>
        </td>
        <td><p>Plot IAS lineplot</p></td>
      </tr><tr><td>
          <p><code><a href="plotIasLineplotFitted.html">plotIasLineplotFitted()</a></code> </p>
        </td>
        <td><p>Plot IAS model fitting</p></td>
      </tr><tr><td>
          <p><code><a href="plotIasRidgeplot.html">plotIasRidgeplot()</a></code> <code><a href="plotIasRidgeplot.html">plotIasRidgeplotSC()</a></code> </p>
        </td>
        <td><p>Plot IAS rideplot</p></td>
      </tr><tr><td>
          <p><code><a href="plotOverview.html">plotOverview()</a></code> </p>
        </td>
        <td><p>Plot overview of S4 objects</p></td>
      </tr><tr><td>
          <p><code><a href="plotVolcano.html">plotVolcano()</a></code> </p>
        </td>
        <td><p>Compare evaluation of spatially opposing fits</p></td>
      </tr><tr><td>
          <p><code><a href="plotTrajectoryBarplot.html">plotTrajectoryBarplot()</a></code> </p>
        </td>
        <td><p>Plot categorical trajectory dynamics</p></td>
      </tr><tr><td>
          <p><code><a href="plotTrajectoryEvaluation.html">plotTrajectoryEvaluation()</a></code> </p>
        </td>
        <td><p>Plot STS evaluation per variable-model pair</p></td>
      </tr><tr><td>
          <p><code><a href="plotTrajectoryHeatmap.html">plotTrajectoryHeatmap()</a></code> </p>
        </td>
        <td><p>Plot trajectory expression dynamic in heatmap</p></td>
      </tr><tr><td>
          <p><code><a href="plotTrajectoryLineplot.html">plotTrajectoryLineplot()</a></code> <code><a href="plotTrajectoryLineplot.html">plotTrajectoryRidgeplot()</a></code> </p>
        </td>
        <td><p>Plot continuous trajectory dynamics</p></td>
      </tr><tr><td>
          <p><code><a href="plotTrajectoryLineplotFitted.html">plotTrajectoryLineplotFitted()</a></code> </p>
        </td>
        <td><p>Plot trajectory model fitting</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="statistics">Statistics <a href="#statistics" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="plotBarchart.html">plotBarchart()</a></code> </p>
        </td>
        <td><p>Plot distribution of discrete variables</p></td>
      </tr><tr><td>
          <p><code><a href="plotBoxplot.html">plotBoxplot()</a></code> <code><a href="plotBoxplot.html">plotDensityplot()</a></code> <code><a href="plotBoxplot.html">plotHistogram()</a></code> <code><a href="plotBoxplot.html">plotRidgeplot()</a></code> <code><a href="plotBoxplot.html">plotVioBoxplot()</a></code> <code><a href="plotBoxplot.html">plotViolinplot()</a></code> </p>
        </td>
        <td><p>Plot numeric distribution and statistical tests</p></td>
      </tr><tr><td>
          <p><code><a href="plotMosaicplot.html">plotMosaicplot()</a></code> </p>
        </td>
        <td><p>Plot mosaic plot</p></td>
      </tr><tr><td>
          <p><code><a href="plotRiverplot.html">plotRiverplot()</a></code> </p>
        </td>
        <td><p>Plot riverplots</p></td>
      </tr><tr><td>
          <p><code><a href="plotStatisticsInteractive.html">plotStatisticsInteractive()</a></code> </p>
        </td>
        <td><p>Plot distribution of variables interactively</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="surface-plots">Surface Plots <a href="#surface-plots" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="plotSurface.html">plotSurface()</a></code> <code><a href="plotSurface.html">plotSurfaceInteractive()</a></code> </p>
        </td>
        <td><p>Plot the surface of the sample</p></td>
      </tr><tr><td>
          <p><code><a href="plotSurfaceAverage.html">plotSurfaceAverage()</a></code> </p>
        </td>
        <td><p>Plot several surface plots colored by gene averages</p></td>
      </tr><tr><td>
          <p><code><a href="plotSurfaceBase.html">plotSurfaceBase()</a></code> </p>
        </td>
        <td><p>Plot surface with R base plotting</p></td>
      </tr><tr><td>
          <p><code><a href="plotSurfaceComparison.html">plotSurfaceComparison()</a></code> <code><a href="plotSurfaceComparison.html">plotSurfaceComparison2()</a></code> </p>
        </td>
        <td><p>Plot several surface plots colored by numeric variables</p></td>
      </tr><tr><td>
          <p><code><a href="plotSurfaceIAS.html">plotSurfaceIAS()</a></code> </p>
        </td>
        <td><p>Plot screening area of IAS-algorithm</p></td>
      </tr><tr><td>
          <p><code><a href="plotSurfaceQuantiles.html">plotSurfaceQuantiles()</a></code> </p>
        </td>
        <td><p>Plot a surface plot colored by binned numeric variables</p></td>
      </tr><tr><td>
          <p><code><a href="plotSurfaceSC.html">plotSurfaceSC()</a></code> </p>
        </td>
        <td><p>Plot single cells on surface</p></td>
      </tr></tbody></table></div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Jan Kueckelhaus, Dieter-Henrik Heiland, Simon Frerich.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

