<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Plot image with ggplot2 — plotImage • SPATA2</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Plot image with ggplot2 — plotImage"><meta property="og:description" content="Plots the histology image with ggplot2."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SPATA2</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">3.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/adding-data.html">Adding Data to SPATA2 objects</a>
    </li>
    <li>
      <a href="../articles/clustering.html">Clustering</a>
    </li>
    <li>
      <a href="../articles/cnv.html">Copy Number Variations (CNV)</a>
    </li>
    <li>
      <a href="../articles/creating-spatial-annotations.html">Creating Spatial Annotations</a>
    </li>
    <li>
      <a href="../articles/creating-spatial-trajectories.html">Spatial Trajectories</a>
    </li>
    <li>
      <a href="../articles/dea.html">Differential Expression Analysis (DEA)</a>
    </li>
    <li>
      <a href="../articles/gsea.html">Gene Set Enrichment Analysis</a>
    </li>
    <li>
      <a href="../articles/image-handling.html">Image Handling</a>
    </li>
    <li>
      <a href="../articles/initiation-and-preprocessing-customized.html">Initiation of individual SPATA2 objects</a>
    </li>
    <li>
      <a href="../articles/initiation-and-preprocessing-visium.html">Initiation &amp; Processing: Visium</a>
    </li>
    <li>
      <a href="../articles/newslog.html">Newslog</a>
    </li>
    <li>
      <a href="../articles/spatial-annotation-screening.html">Spatial Annotation Screening</a>
    </li>
    <li>
      <a href="../articles/spatial-measures.html">Spatial measures</a>
    </li>
    <li>
      <a href="../articles/spatial-segmentation.html">Spatial Segmentation</a>
    </li>
    <li>
      <a href="../articles/spatial-trajectory-screening.html">Spatial Trajectory Screening</a>
    </li>
    <li>
      <a href="../articles/surface-plotting.html">Surface Plotting</a>
    </li>
    <li>
      <a href="../articles/using-spatial-annotations.html">Using Spatial Annotations</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Plot image with ggplot2</h1>
    
    <div class="hidden name"><code>plotImage.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Plots the histology image with <code>ggplot2</code>.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">plotImage</span><span class="op">(</span><span class="va">object</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for SPATA2</span></span>
<span><span class="fu">plotImage</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  img_name <span class="op">=</span> <span class="fu"><a href="activateImage.html">activeImage</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span>,</span>
<span>  outline <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  by_section <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  fragments <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  line_alpha <span class="op">=</span> <span class="fl">0.9</span>,</span>
<span>  line_color <span class="op">=</span> <span class="st">"black"</span>,</span>
<span>  line_size <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  line_type <span class="op">=</span> <span class="st">"solid"</span>,</span>
<span>  transform <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  img_alpha <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  scale_fct <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  xrange <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  yrange <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for SpatialData</span></span>
<span><span class="fu">plotImage</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  img_name <span class="op">=</span> <span class="fu"><a href="activateImage.html">activeImage</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span>,</span>
<span>  outline <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  by_section <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  fragments <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  line_alpha <span class="op">=</span> <span class="fl">0.9</span>,</span>
<span>  line_color <span class="op">=</span> <span class="st">"black"</span>,</span>
<span>  line_size <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  line_type <span class="op">=</span> <span class="st">"solid"</span>,</span>
<span>  transform <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  img_alpha <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  scale_fct <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  xrange <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  yrange <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for HistoImage</span></span>
<span><span class="fu">plotImage</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  outline <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  by_section <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  fragments <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  line_alpha <span class="op">=</span> <span class="fl">0.9</span>,</span>
<span>  line_color <span class="op">=</span> <span class="st">"black"</span>,</span>
<span>  line_size <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  line_type <span class="op">=</span> <span class="st">"solid"</span>,</span>
<span>  transform <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  img_alpha <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  scale_fct <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  xrange <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  yrange <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  display_subtitle <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for Image</span></span>
<span><span class="fu">plotImage</span><span class="op">(</span><span class="va">object</span>, scale_fct <span class="op">=</span> <span class="fl">1</span>, img_alpha <span class="op">=</span> <span class="fl">1</span>, <span class="va">...</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>object</dt>
<dd><p>An object of class <code><a href="SPATA2.html">SPATA2</a></code> or, in case of S4 generics,
objects of classes for which a method has been defined.</p></dd>


<dt>...</dt>
<dd><p>Additional arguments given to <code><a href="ggpLayerZoom.html">ggpLayerZoom()</a></code>.</p></dd>


<dt>img_name</dt>
<dd><p>Character value. The name of the image of interest.
If <code>NULL</code>, the <a href="concept_active.html">active</a> image is chosen by default. Either way, must
be one of <code><a href="getImageNames.html">getImageNames()</a></code>.</p></dd>


<dt>by_section</dt>
<dd><p>Logical value. If <code>TRUE</code>, the outline is used which acknowledges
the potential for multiple tissue sections. If <code>FALSE</code>, the outline is used which
simply outlines everything with one single polygon.</p></dd>


<dt>line_alpha</dt>
<dd><p>Numeric. Affects alpha of main lines of the plot.</p></dd>


<dt>line_color</dt>
<dd><p>Character. Affects color of the main lines of the plot.</p></dd>


<dt>line_size</dt>
<dd><p>Numeric. Affects size of the main lines of the plot.</p></dd>


<dt>line_type</dt>
<dd><p>Character. The line type. One of <em>'blank'</em>, <em>'solid'</em>,
<em>'dashed'</em>, <em>'dotted'</em>, <em>'dotdash'</em>, <em>'longdash'</em> and <em>'twodash'</em>.</p></dd>


<dt>transform</dt>
<dd><p>Logical value. Ignored if no images exist in the object. If <code>TRUE</code>,
spatial transformation instructions saved during image alignment of
the image <code>img_name</code> are applied. Only relevant if there are any images registered
in the <code>SPATA2</code> object.</p></dd>


<dt>xrange, yrange</dt>
<dd><p><a href="concept_distance_measure.html">Distance vector</a> of length
two or <code>NULL</code>. If not <code>NULL</code>, specifies the x- and y-range to which
the spatial output is cropped. E.g. <code>xrange = c(200, 500)</code> results in
the two dimensional space being cropped from x-coordinate 200px up to
x-coordinate 500px. If <code>NULL</code>, the original range is used.</p></dd>


<dt>unit</dt>
<dd><p>Character value. Units of x- and y-axes. Defaults
to <em>'px'</em>.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A ggplot.</p>
    </div>
    <div id="distance-measures">
    <h2>Distance measures</h2>
    


<p>The vignette on distance measures in SPATA2 has been replaced. Click
<code><a href="concept_distance_measure.html">here</a></code> to read it.</p>
    </div>
    <div id="image-visualization-with-ggplot-">
    <h2>Image visualization with ggplot2</h2>
    


<p>When comparing the output of <code>ggplot() + ggpLayerImage()</code> with other image plotting functions,
you may notice that the image appears horizontally flipped when plotted using <code><a href="ggpLayerImage.html">ggpLayerImage()</a></code>.
This behavior is due to the use of a Cartesian coordinate system in <code>SPATA2</code>, where a pixel
with coordinates c(width = 1, height = 1) is plotted on the left side at the bottom.
In contrast, functions like <code><a href="https://rdrr.io/pkg/EBImage/man/display.html" class="external-link">EBImage::display()</a></code> or <code><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">graphics::plot()</a></code> use an <em>image space</em> coordinate system,
where pixel heights start from the top. Consequently, in <em>image space</em>, pixel c(width = 1, height = 1)
is displayed on the top resulting in mirror inverted visualization of the image.</p>
<p>We chose to use a Cartesian coordinate system in <code>SPATA2</code> because we believe it provides a more intuitive
framework for the spatial alignment of tissue, spatial annotations, spatial trajectories,
barcoded sots, single cells, etc. where coordinates in the corresponding data.frames are provided
in form of <em>x</em>- and <em>y</em>-variables. See <code><a href="getCoordsDf.html">getCoordsDf()</a></code>, <code><a href="getImgAnnOutlineDf.html">getImgAnnOutlineDf()</a></code>, <code><a href="getTissueOutlineDf.html">getTissueOutlineDf()</a></code> etc.</p>
<p>If you prefer to view your image in the regular orientation, you can use the <code><a href="flipAll.html">flipAll()</a></code> function on your object,
specifying <code>axis = "horizontal"</code>, to reverse this effect.</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://themilolab.com/" class="external-link">SPATA2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"example_data"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">object</span> <span class="op">&lt;-</span> <span class="va">example_data</span><span class="op">$</span><span class="va">object_UKF275T_diet</span></span>
<span></span>
<span><span class="va">object</span> <span class="op">&lt;-</span></span>
<span> <span class="fu"><a href="createNumericAnnotations.html">createNumericAnnotations</a></span><span class="op">(</span></span>
<span>   object <span class="op">=</span> <span class="va">object</span>,</span>
<span>   variable <span class="op">=</span> <span class="st">"HM_HYPOXIA"</span>,</span>
<span>   threshold <span class="op">=</span> <span class="st">"kmeans_high"</span>,</span>
<span>   id <span class="op">=</span> <span class="st">"hypoxia_ann"</span>,</span>
<span>   inner_borders <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>   force1 <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>   <span class="op">)</span></span>
<span></span>
<span><span class="co"># range of hypoxia annotation</span></span>
<span><span class="va">hr</span> <span class="op">&lt;-</span> <span class="fu"><a href="getSpatAnnRange.html">getSpatAnnRange</a></span><span class="op">(</span><span class="va">object</span>, id <span class="op">=</span> <span class="st">"hypoxia_ann"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">hr_ggpL</span> <span class="op">&lt;-</span> <span class="fu"><a href="ggpLayerSpatAnnOutline.html">ggpLayerSpatAnnOutline</a></span><span class="op">(</span><span class="va">object</span>, ids <span class="op">=</span> <span class="st">"hypoxia_ann"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="plotSurface.html">plotSurface</a></span><span class="op">(</span><span class="va">object</span>, <span class="st">"HM_HYPOXIA"</span><span class="op">)</span> <span class="op">+</span> <span class="va">hr_ggpL</span></span>
<span></span>
<span><span class="va">p_rect</span> <span class="op">&lt;-</span></span>
<span> <span class="fu">plotImage</span><span class="op">(</span><span class="va">object</span><span class="op">)</span> <span class="op">+</span></span>
<span> <span class="va">hr_ggpL</span> <span class="op">+</span></span>
<span> <span class="fu"><a href="ggpLayerRect.html">ggpLayerRect</a></span><span class="op">(</span><span class="va">object</span>, xrange <span class="op">=</span> <span class="va">hr</span><span class="op">$</span><span class="va">x</span>, yrange <span class="op">=</span> <span class="va">hr</span><span class="op">$</span><span class="va">y</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p_zoomed</span> <span class="op">&lt;-</span></span>
<span> <span class="fu">plotImage</span><span class="op">(</span><span class="va">object</span><span class="op">)</span> <span class="op">+</span></span>
<span> <span class="fu"><a href="ggpLayerZoom.html">ggpLayerZoom</a></span><span class="op">(</span><span class="va">object</span>, xrange <span class="op">=</span> <span class="va">hr</span><span class="op">$</span><span class="va">x</span>, yrange <span class="op">=</span> <span class="va">hr</span><span class="op">$</span><span class="va">y</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p_rect</span> <span class="op">+</span> <span class="va">p_zoomed</span></span>
<span></span></code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Jan Kueckelhaus, Dieter-Henrik Heiland, Simon Frerich.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer></div>

  


  

  </body></html>

