<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Plot gene expression testing results — plotDeaVolcano • SPATA2</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Plot gene expression testing results — plotDeaVolcano"><meta property="og:description" content="Plots a common volcano plot with p-value on the y- and logfold
change on the x-axis."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SPATA2</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.0.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/spata-v2-installation.html">Installation</a>
</li>
<li>
  <a href="../articles/spata-v2-spata-data.html">SPATAData</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li class="dropdown-header">Around the spata2 object</li>
    <li>
      <a href="../articles/spata-v2-object-initiation-and-manipulation.html">Object Set Up and Manipulation</a>
    </li>
    <li>
      <a href="../articles/spata-v2-platform-compatibility.html">Platform Compatibility</a>
    </li>
    <li>
      <a href="../articles/spata-v2-autoencoder.html">Autoencoder</a>
    </li>
    <li>
      <a href="../articles/spata-v2-genes-gene-sets-and-features.html">Genes, Gene Sets and Features</a>
    </li>
    <li class="dropdown-header">Image Analysis</li>
    <li>
      <a href="../articles/spata-v2-image-handling.html">Image Handling</a>
    </li>
    <li>
      <a href="../articles/spata-v2-image-annotations.html">Image Annotations</a>
    </li>
    <li>
      <a href="../articles/spata-v2-spatial-segmentation.html">Spatial Segmentation</a>
    </li>
    <li class="dropdown-header">Analysis of Gene Expression</li>
    <li>
      <a href="../articles/spata-v2-cnv-analysis.html">Copy Number Variations</a>
    </li>
    <li>
      <a href="../articles/spata-v2-clustering.html">Clustering</a>
    </li>
    <li>
      <a href="../articles/spata-v2-dea.html">Differential Expression</a>
    </li>
    <li>
      <a href="../articles/spata-v2-gsea.html">Gene Set Enrichment</a>
    </li>
    <li class="dropdown-header">Expression Gradients &amp; Gradient Screening</li>
    <li>
      <a href="../articles/spata-v2-spatial-trajectories.html">Spatial Trajectories</a>
    </li>
    <li>
      <a href="../articles/spata-v2-spatial-trajectory-screening.html">Spatial Trajectory Screening</a>
    </li>
    <li>
      <a href="../articles/spata-v2-image-annotation-screening.html">Image Annotation Screening</a>
    </li>
    <li>
      <a href="../articles/spata-v2-scd.html">Integrate Single Cell Deconvolution</a>
    </li>
    <li class="dropdown-header">Visualization</li>
    <li>
      <a href="../articles/spata-v2-plotting-surface.html">Surface Plotting</a>
    </li>
    <li>
      <a href="../articles/spata-v2-plotting-misc.html">Miscellaneous</a>
    </li>
  </ul></li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Code &amp; Concept
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/cc-spatial-measures.html">Spatial Measures</a>
    </li>
    <li>
      <a href="../articles/cc-spatial-segmentation-vs-image-annotations.html">Spatial Segmentation vs. Image Annotations</a>
    </li>
    <li>
      <a href="../articles/cc-spatial-trajectory-screening.html">Spatial Trajectory Screening</a>
    </li>
    <li>
      <a href="../articles/cc-image-annotation-screening.html">Image Annotation Screening</a>
    </li>
    <li>
      <a href="../articles/cc-model-fitting-in-spatial-transcriptomic-studies.html">Model Fitting</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Plot gene expression testing results</h1>
    
    <div class="hidden name"><code>plotDeaVolcano.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Plots a common volcano plot with p-value on the y- and logfold
change on the x-axis.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">plotDeaVolcano</span><span class="op">(</span>
  <span class="va">object</span>,
  across <span class="op">=</span> <span class="fu"><a href="setDefaultGrouping.html">getDefaultGrouping</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span>,
  across_subset <span class="op">=</span> <span class="cn">NULL</span>,
  relevel <span class="op">=</span> <span class="cn">TRUE</span>,
  method_de <span class="op">=</span> <span class="cn">NULL</span>,
  color_up <span class="op">=</span> <span class="st">"tomato"</span>,
  color_down <span class="op">=</span> <span class="st">"steelblue"</span>,
  color_insignif <span class="op">=</span> <span class="st">"lightgrey"</span>,
  pt_alpha <span class="op">=</span> <span class="fl">0.9</span>,
  pt_size <span class="op">=</span> <span class="fl">1</span>,
  threshold_logFC <span class="op">=</span> <span class="fl">1</span>,
  threshold_pval <span class="op">=</span> <span class="fl">0.01</span>,
  label_genes <span class="op">=</span> <span class="fl">5</span>,
  label_insignificant <span class="op">=</span> <span class="cn">TRUE</span>,
  label_side <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"up"</span>, <span class="st">"down"</span><span class="op">)</span>,
  label_size <span class="op">=</span> <span class="fl">1</span>,
  nrow <span class="op">=</span> <span class="cn">NULL</span>,
  ncol <span class="op">=</span> <span class="cn">NULL</span>,
  scales <span class="op">=</span> <span class="st">"fixed"</span>,
  use_pseudolog <span class="op">=</span> <span class="cn">FALSE</span>,
  <span class="va">...</span>
<span class="op">)</span>

<span class="fu">plotDeaVolcano1v1</span><span class="op">(</span>
  <span class="va">object</span>,
  <span class="va">across</span>,
  method_de <span class="op">=</span> <span class="cn">NULL</span>,
  clrp <span class="op">=</span> <span class="cn">NULL</span>,
  clrp_adjust <span class="op">=</span> <span class="cn">NULL</span>,
  color_insignif <span class="op">=</span> <span class="st">"lightgrey"</span>,
  pt_alpha <span class="op">=</span> <span class="fl">0.9</span>,
  pt_size <span class="op">=</span> <span class="fl">1</span>,
  threshold_logFC <span class="op">=</span> <span class="fl">1</span>,
  threshold_pval <span class="op">=</span> <span class="fl">0.01</span>,
  col_pval <span class="op">=</span> <span class="st">"p_val_adj"</span>,
  label_genes <span class="op">=</span> <span class="fl">5</span>,
  label_size <span class="op">=</span> <span class="fl">1</span>,
  use_pseudolog <span class="op">=</span> <span class="cn">FALSE</span>,
  limits <span class="op">=</span> <span class="cn">NULL</span>,
  display_title <span class="op">=</span> <span class="cn">TRUE</span>,
  title_size <span class="op">=</span> <span class="fl">2</span>,
  digits <span class="op">=</span> <span class="fl">2</span>,
  <span class="va">...</span>
<span class="op">)</span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>object</dt>
<dd><p>An object of class <code>spata2</code>.</p></dd>


<dt>across</dt>
<dd><p>Character value or NULL. Specifies the grouping variable of interest.</p>
<p>Use <code><a href="getGroupingOptions.html">getGroupingOptions()</a></code> to obtain all variable names that group the
barcode spots of your object in a certain manner.</p></dd>


<dt>across_subset</dt>
<dd><p>Character vector or NULL. Specifies the particular groups
of interest the grouping variable specified in argument <code>across</code> contains.</p>
<p>If set to NULL all of them are chosen. You can prefix groups you are NOT interested in
with a <em>'-'</em>. (Saves writing if there are more groups you are interested in
than groups you are not interested in.)</p>
<p>Use <code><a href="getGroupNames.html">getGroupNames()</a></code> to obtain all valid input options.</p></dd>


<dt>relevel</dt>
<dd><p>Logical value. If set to TRUE the input order of <code>across_subset</code>
determines the order in which the groups of interest are displayed. Groups that
are not included are dropped which affects the colors with which they are displayed.</p></dd>


<dt>method_de</dt>
<dd><p>Character value. Denotes the method to according to which the de-analysis is performed.
Given to argument <code>test.use</code> of the <code><a href="https://satijalab.org/seurat/reference/FindAllMarkers.html" class="external-link">Seurat::FindAllMarkers()</a></code>-function. Run <code>SPATA::dea_methods</code>
to obtain all valid input options.</p></dd>


<dt>color_up</dt>
<dd><p>Character value. Color of points that represent upregulated genes.</p></dd>


<dt>color_down</dt>
<dd><p>Character value. Color of poitns that represent downregulated genes.</p></dd>


<dt>color_insignif</dt>
<dd><p>Character value. Color of points that fall below the threshold set
via argument <code>threshold_pval</code>.</p></dd>


<dt>pt_alpha</dt>
<dd><p>Numeric value. Specifies the degree of transparency of all points.</p></dd>


<dt>pt_size</dt>
<dd><p>Numeric value. Specifies the size of all points.</p></dd>


<dt>threshold_logFC</dt>
<dd><p>Numeric value. Denotes the threshold for the logFC. Genes with
a logFC below are displayed as not significant.</p></dd>


<dt>threshold_pval</dt>
<dd><p>Numeric value. Denotes the threshold for the p-value. Genes with
a p-value above are displayed as not significant.</p></dd>


<dt>label_genes</dt>
<dd><p>Specify which genes are labeled in the plot. If numeric,
specifies the number of genes that are labeled. E.g. if <code>label_genes</code> = 5,
the default, the top 5 genes are labeled. If character, specifies the genes
that are supposed to be labeled by name. If <code>NULL</code> or <code>FALSE</code>, no genes are labeled.</p></dd>


<dt>label_insignificant</dt>
<dd><p>Logical value. If <code>FALSE</code>, insignifcant genes are
not labeled.</p></dd>


<dt>label_side</dt>
<dd><p>Character vector. Decides on which side to label genes. Valid input
are <em>'up'</em> and/or <em>'down'</em>.</p></dd>


<dt>nrow, ncol</dt>
<dd><p>Numeric values or NULL. Used to arrange multiple plots.</p></dd>


<dt>use_pseudolog</dt>
<dd><p>Logical value. If TRUE, avglogFC is transformed with log10. Requires
package <code>ggallin</code> to be installed.</p></dd>


<dt>...</dt>
<dd><p>Used to absorb deprecated arguments or functions.</p></dd>


<dt>clrp</dt>
<dd><p>Character value. Specifies the color palette to be used to represent
groups of discrete variables. Run <code><a href="validActivationFunctions.html">validColorPalettes()</a></code> to obtain valid
input options.</p></dd>


<dt>clrp_adjust</dt>
<dd><p>Named character vector or NULL. If character, it adjusts the
color palette that is used to represent the groups. Names of the input vector must refer
to the group and the respective named element denotes the color with which to
represent the group.</p></dd>


<dt>display_title</dt>
<dd><p>Logical value. If set to TRUE an informative title is displayed.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A ggplot.</p>
    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Jan Kueckelhaus, Dieter-Henrik Heiland, Simon Frerich.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

