<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Plot differentially expressed genes — plotDeaHeatmap • SPATA2</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Plot differentially expressed genes — plotDeaHeatmap"><meta property="og:description" content="Visualizes the expression of genes across subgroups in a heatmap. It either takes the results
from previously conducted de-analysis or uses the expression information of specific genes to plot a heatmap."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SPATA2</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.0.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/spata-v2-installation.html">Installation</a>
</li>
<li>
  <a href="../articles/spata-v2-spata-data.html">SPATAData</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li class="dropdown-header">Around the spata2 object</li>
    <li>
      <a href="../articles/spata-v2-object-initiation-and-manipulation.html">Object Set Up and Manipulation</a>
    </li>
    <li>
      <a href="../articles/spata-v2-platform-compatibility.html">Platform Compatibility</a>
    </li>
    <li>
      <a href="../articles/spata-v2-autoencoder.html">Autoencoder</a>
    </li>
    <li>
      <a href="../articles/spata-v2-genes-gene-sets-and-features.html">Genes, Gene Sets and Features</a>
    </li>
    <li class="dropdown-header">Image Analysis</li>
    <li>
      <a href="../articles/spata-v2-image-handling.html">Image Handling</a>
    </li>
    <li>
      <a href="../articles/spata-v2-image-annotations.html">Image Annotations</a>
    </li>
    <li>
      <a href="../articles/spata-v2-spatial-segmentation.html">Spatial Segmentation</a>
    </li>
    <li class="dropdown-header">Analysis of Gene Expression</li>
    <li>
      <a href="../articles/spata-v2-cnv-analysis.html">Copy Number Variations</a>
    </li>
    <li>
      <a href="../articles/spata-v2-clustering.html">Clustering</a>
    </li>
    <li>
      <a href="../articles/spata-v2-dea.html">Differential Expression</a>
    </li>
    <li>
      <a href="../articles/spata-v2-gsea.html">Gene Set Enrichment</a>
    </li>
    <li class="dropdown-header">Expression Gradients &amp; Gradient Screening</li>
    <li>
      <a href="../articles/spata-v2-spatial-trajectories.html">Spatial Trajectories</a>
    </li>
    <li>
      <a href="../articles/spata-v2-spatial-trajectory-screening.html">Spatial Trajectory Screening</a>
    </li>
    <li>
      <a href="../articles/spata-v2-image-annotation-screening.html">Image Annotation Screening</a>
    </li>
    <li>
      <a href="../articles/spata-v2-scd.html">Integrate Single Cell Deconvolution</a>
    </li>
    <li class="dropdown-header">Visualization</li>
    <li>
      <a href="../articles/spata-v2-plotting-surface.html">Surface Plotting</a>
    </li>
    <li>
      <a href="../articles/spata-v2-plotting-misc.html">Miscellaneous</a>
    </li>
  </ul></li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Code &amp; Concept
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/cc-spatial-measures.html">Spatial Measures</a>
    </li>
    <li>
      <a href="../articles/cc-spatial-segmentation-vs-image-annotations.html">Spatial Segmentation vs. Image Annotations</a>
    </li>
    <li>
      <a href="../articles/cc-spatial-trajectory-screening.html">Spatial Trajectory Screening</a>
    </li>
    <li>
      <a href="../articles/cc-image-annotation-screening.html">Image Annotation Screening</a>
    </li>
    <li>
      <a href="../articles/cc-model-fitting-in-spatial-transcriptomic-studies.html">Model Fitting</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Plot differentially expressed genes</h1>
    
    <div class="hidden name"><code>plotDeaHeatmap.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Visualizes the expression of genes across subgroups in a heatmap. It either takes the results
from previously conducted de-analysis or uses the expression information of specific genes to plot a heatmap.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">plotDeaHeatmap</span><span class="op">(</span>
  <span class="va">object</span>,
  <span class="va">across</span>,
  across_subset <span class="op">=</span> <span class="cn">NULL</span>,
  relevel <span class="op">=</span> <span class="cn">NULL</span>,
  method_de <span class="op">=</span> <span class="cn">NULL</span>,
  max_adj_pval <span class="op">=</span> <span class="cn">NULL</span>,
  min_lfc <span class="op">=</span> <span class="cn">NULL</span>,
  n_highest_lfc <span class="op">=</span> <span class="cn">NULL</span>,
  n_lowest_pval <span class="op">=</span> <span class="cn">NULL</span>,
  breaks <span class="op">=</span> <span class="cn">NULL</span>,
  genes <span class="op">=</span> <span class="cn">NULL</span>,
  n_bcsp <span class="op">=</span> <span class="cn">NULL</span>,
  clrp <span class="op">=</span> <span class="cn">NULL</span>,
  colors <span class="op">=</span> <span class="cn">NULL</span>,
  verbose <span class="op">=</span> <span class="cn">NULL</span>,
  of_sample <span class="op">=</span> <span class="cn">NA</span>,
  <span class="va">...</span>
<span class="op">)</span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>object</dt>
<dd><p>An object of class <code>spata2</code>.</p></dd>


<dt>across</dt>
<dd><p>Character value or NULL. Specifies the grouping variable of interest.</p>
<p>Use <code><a href="getGroupingOptions.html">getGroupingOptions()</a></code> to obtain all variable names that group the
barcode spots of your object in a certain manner.</p></dd>


<dt>across_subset</dt>
<dd><p>Character vector or NULL. Specifies the particular groups
of interest the grouping variable specified in argument <code>across</code> contains.</p>
<p>If set to NULL all of them are chosen. You can prefix groups you are NOT interested in
with a <em>'-'</em>. (Saves writing if there are more groups you are interested in
than groups you are not interested in.)</p>
<p>Use <code><a href="getGroupNames.html">getGroupNames()</a></code> to obtain all valid input options.</p></dd>


<dt>relevel</dt>
<dd><p>Logical value. If set to TRUE the input order of <code>across_subset</code>
determines the order in which the groups of interest are displayed. Groups that
are not included are dropped which affects the colors with which they are displayed.</p></dd>


<dt>method_de</dt>
<dd><p>Character value. Denotes the method to according to which the de-analysis is performed.
Given to argument <code>test.use</code> of the <code><a href="https://satijalab.org/seurat/reference/FindAllMarkers.html" class="external-link">Seurat::FindAllMarkers()</a></code>-function. Run <code>SPATA::dea_methods</code>
to obtain all valid input options.</p></dd>


<dt>max_adj_pval</dt>
<dd><p>Numeric value. Sets the threshold for adjusted p-values. All genes
with adjusted p-values above that threshold are ignored.</p></dd>


<dt>min_lfc</dt>
<dd><p>Numeric value. Sets the threshold for average log fold change. All genes
with an average log fold change below that threshold are ignored.</p></dd>


<dt>n_highest_lfc</dt>
<dd><p>Numeric value. Affects the total number of genes that are kept. See details.</p></dd>


<dt>n_lowest_pval</dt>
<dd><p>Numeric value. Affects the total number of genes that are kept. See details.</p></dd>


<dt>breaks</dt>
<dd><p>Denotes the colorspectrum breaks. If set to NULL the breaks are set automatically. If a
numeric vector is specified it is taken as input. If a function is specified the expression matrix is
passed to it as the first argument and the length of <code>colors</code> as the second argument.</p></dd>


<dt>genes</dt>
<dd><p>Character vector or NULL. If you want to display specific genes irrespective of de-anaylsis results you
can specifiy them in <code>genes</code>. If <code>genes</code> is specified that way arguments referring to de-anylsis results are
ignored and only the genes specified are taken and displayed.</p></dd>


<dt>n_bcsp</dt>
<dd><p>The number of barcode-spots belonging to each cluster you want to
include in the matrix. Should be lower than the total number of barcode-spots of every cluster
and can be deployed in order to keep the heatmap clear and aesthetically pleasing.</p>
<p>If set to NULL (the default) it is automatically computed according to the number of genes that
are displayed in the heatmap.</p></dd>


<dt>clrp</dt>
<dd><p>Character value. Specifies the color palette to be used to represent
groups of discrete variables. Run <code><a href="validActivationFunctions.html">validColorPalettes()</a></code> to obtain valid
input options.</p></dd>


<dt>verbose</dt>
<dd><p>Logical. If set to TRUE informative messages regarding
the computational progress will be printed.</p>
<p>(Warning messages will always be printed.)</p></dd>


<dt>of_sample</dt>
<dd><p>This argument is currently inactive. It might be reactivated when
spata-objects can store more than one sample.</p></dd>


<dt>...</dt>
<dd><p>Additional arguments given to <code><a href="https://rdrr.io/pkg/pheatmap/man/pheatmap.html" class="external-link">pheatmap::pheatmap()</a></code>.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A heatmap of class 'pheatmap'.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>The de-data.frame is processed such that the following steps are performed for every experimental
group.</p>
<ol><li><p>Discards genes with <em>avg_logFC</em>-values that are either infinite or negative</p></li>
<li><p>Discards genes with adjusted p-values above the threshold set with <code>max_adj_pval</code></p></li>
<li><p>Discard genes with average log fold change below the treshold set with <code>min_lfc</code></p></li>
<li><p>Slices the data.frame in order that for every experimental group:</p><ol><li><p>the n genes with the highest <em>avg_logFC</em>-values are kept where n = <code>n_highest_lfc</code></p></li>
<li><p>the n genes with the lowest <em>p_val_adj</em>-values are kept where n = <code>n_lowest_pval</code></p></li>
</ol></li>
<li><p>Arranges the genes according to the highest <em>avg_logFC</em>-values</p></li>
</ol></div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Jan Kueckelhaus, Dieter-Henrik Heiland, Simon Frerich.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

