<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Transform spata-object to cell-data-set (Monocle3) — transformSpataToCDS • SPATA2</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Transform spata-object to cell-data-set (Monocle3) — transformSpataToCDS"><meta property="og:description" content="Takes the count matrix of your spata-object and creates a
cell_data_set-object with it. See details for more information on how to use
the arguments."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SPATA2</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/spata-v2-installation.html">Installation</a>
</li>
<li>
  <a href="../articles/spata-v2-spata-data.html">SPATAData</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li class="dropdown-header">Around the spata2 object</li>
    <li>
      <a href="../articles/spata-v2-object-initiation-and-manipulation.html">Object Set Up and Manipulation</a>
    </li>
    <li>
      <a href="../articles/spata-v2-platform-compatibility.html">Platform Compatibility</a>
    </li>
    <li>
      <a href="../articles/spata-v2-autoencoder.html">Autoencoder</a>
    </li>
    <li>
      <a href="../articles/spata-v2-genes-gene-sets-and-features.html">Genes, Gene Sets and Features</a>
    </li>
    <li class="dropdown-header">Image Analysis</li>
    <li>
      <a href="../articles/spata-v2-image-handling.html">Image Handling</a>
    </li>
    <li>
      <a href="../articles/spata-v2-image-annotations.html">Image Annotations</a>
    </li>
    <li>
      <a href="../articles/spata-v2-spatial-segmentation.html">Spatial Segmentation</a>
    </li>
    <li class="dropdown-header">Analysis of Gene Expression</li>
    <li>
      <a href="../articles/spata-v2-cnv-analysis.html">Copy Number Variations</a>
    </li>
    <li>
      <a href="../articles/spata-v2-clustering.html">Clustering</a>
    </li>
    <li>
      <a href="../articles/spata-v2-dea.html">Differential Expression</a>
    </li>
    <li>
      <a href="../articles/spata-v2-gsea.html">Gene Set Enrichment</a>
    </li>
    <li class="dropdown-header">Expression Gradients &amp; Gradient Screening</li>
    <li>
      <a href="../articles/spata-v2-spatial-trajectories.html">Spatial Trajectories</a>
    </li>
    <li>
      <a href="../articles/spata-v2-spatial-trajectory-screening.html">Spatial Trajectory Screening</a>
    </li>
    <li>
      <a href="../articles/spata-v2-image-annotation-screening.html">Image Annotation Screening</a>
    </li>
    <li>
      <a href="../articles/spata-v2-scd.html">Integrate Single Cell Deconvolution</a>
    </li>
    <li class="dropdown-header">Visualization</li>
    <li>
      <a href="../articles/spata-v2-plotting-surface.html">Surface Plotting</a>
    </li>
    <li>
      <a href="../articles/spata-v2-plotting-misc.html">Miscellaneous</a>
    </li>
  </ul></li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Code &amp; Concept
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/cc-spatial-measures.html">Spatial Measures</a>
    </li>
    <li>
      <a href="../articles/cc-spatial-segmentation-vs-image-annotations.html">Spatial Segmentation vs. Image Annotations</a>
    </li>
    <li>
      <a href="../articles/cc-spatial-trajectory-screening.html">Spatial Trajectory Screening</a>
    </li>
    <li>
      <a href="../articles/cc-image-annotation-screening.html">Image Annotation Screening</a>
    </li>
    <li>
      <a href="../articles/cc-model-fitting-in-spatial-transcriptomic-studies.html">Model Fitting</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Transform spata-object to cell-data-set (Monocle3)</h1>
    
    <div class="hidden name"><code>transformSpataToCDS.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Takes the count matrix of your spata-object and creates a
cell_data_set-object with it. See details for more information on how to use
the arguments.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">transformSpataToCDS</span><span class="op">(</span>
  <span class="va">object</span>,
  preprocess_method <span class="op">=</span> <span class="st">"PCA"</span>,
  reduction_method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"PCA"</span>, <span class="st">"UMAP"</span><span class="op">)</span>,
  cluster_method <span class="op">=</span> <span class="st">"leiden"</span>,
  estimate_size_factors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span>,
  preprocess_cds <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span>,
  reduce_dimension <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span>,
  cluster_cells <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span>,
  learn_graph <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span>,
  order_cells <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span>,
  of_sample <span class="op">=</span> <span class="cn">NA</span>,
  verbose <span class="op">=</span> <span class="cn">TRUE</span>
<span class="op">)</span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>object</dt>
<dd><p>An object of class <code>spata2</code>.</p></dd>


<dt>preprocess_method</dt>
<dd><p>Monocle3 - description:</p>
<p>A string specifying the initial dimension method to use,
currently either PCA or LSI. For LSI (latent semantic
indexing), it converts the (sparse) expression matrix into
tf-idf matrix and then performs SVD to decompose the gene
expression / cells into certain modules / topics. Default
is "PCA".</p></dd>


<dt>reduction_method</dt>
<dd><p>Monocle3 - description:</p>
<p>A character string specifying the algorithm to use for
dimensionality reduction. Currently "UMAP", "tSNE", "PCA"
and "LSI" are supported.</p></dd>


<dt>cluster_method</dt>
<dd><p>Monocle3 - description:</p>
<p>String indicating the clustering method to use. Options are
"louvain" or "leiden". Default is "leiden". Resolution parameter
is ignored if set to "louvain".</p></dd>


<dt>verbose</dt>
<dd><p>Logical. If set to TRUE informative messages regarding
the computational progress will be printed.</p>
<p>(Warning messages will always be printed.)</p></dd>


<dt>estimate_size_factors_args</dt>
<dd><p>A list of arguments given to <code><a href="https://rdrr.io/pkg/monocle3/man/estimate_size_factors.html" class="external-link">monocle3::estimate_size_factors()</a></code>.</p></dd>


<dt>preprocess_cds_args</dt>
<dd><p>A list of arguments given to <code><a href="https://rdrr.io/pkg/monocle3/man/preprocess_cds.html" class="external-link">monocle3::preprocess_cds()</a></code>.</p></dd>


<dt>reduce_dimension_args</dt>
<dd><p>A list of arguments given to <code><a href="https://rdrr.io/pkg/monocle3/man/reduce_dimension.html" class="external-link">monocle3::reduce_dimension()</a></code>.</p></dd>


<dt>cluster_cells_args</dt>
<dd><p>A list of arguments given to <code><a href="https://rdrr.io/pkg/monocle3/man/cluster_cells.html" class="external-link">monocle3::cluster_cells()</a></code>.</p></dd>


<dt>learn_graph_args</dt>
<dd><p>A list of arguments given to <code><a href="https://rdrr.io/pkg/monocle3/man/learn_graph.html" class="external-link">monocle3::learn_graph()</a></code>.</p></dd>


<dt>order_cells_args</dt>
<dd><p>A list of arguments given to <code><a href="https://rdrr.io/pkg/monocle3/man/order_cells.html" class="external-link">monocle3::order_cells()</a></code>.</p></dd>


<dt>save_cds_file</dt>
<dd><p>Character value or NULL. A file-directory (that does not already exists) under which created cell_data_set-object
is saved. Should end with <em>'.RDS'</em>.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A monocle3::cell_data_set object.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p><code>compileCellDataSet()</code> is a convenient wrapper around all pre processing functions
monocle3 provides to handle it's core object - the cell_data_set - after it's initiation. Apart from <code>object</code>
and <code>of_sample</code> arguments this function has two argument families.</p>
<p>Handling <code>*_method</code>-arguments:</p>
<p>Monocle3 allows to use different methods for dimensional-reduction or clustering which depend
on each other. These arguments take a character vector of all valid inputs. <code>transformSpataToCDS()</code> iterates
over all valid combinations and returns the cell_data_set with the computed information inside.</p>
<p>Handling monocle-function-arguments:</p>
<p>These arguments take named lists of arguments that are given to the respective function. The <code>_method</code>-arguments
as well as the argument <code>cds</code> are automatically defined and must not be included in the given lists!!! Empty lists - the default -
result in running the function with it's default parameters.</p>
<p>The spata-objects feature data (@fdata) is passed to the cell_data_set for it's slot <code>cell_meta_data</code>.</p>
    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Jan Kueckelhaus, Dieter-Henrik Heiland, Simon Frerich.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

